{
 "cells": [
  {
   "cell_type": "markdown",
   "source": [
    "# Setup"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true,
    "ExecuteTime": {
     "end_time": "2023-08-05T19:20:56.329946Z",
     "start_time": "2023-08-05T19:20:56.290771Z"
    }
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "\n",
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "import os\n",
    "import pprint\n",
    "import ast\n",
    "import re\n",
    "import matplotlib.pyplot as plt\n",
    "import dill\n",
    "import requests\n",
    "import xmltodict\n",
    "import json\n",
    "\n",
    "pp = pprint.PrettyPrinter(depth=6)\n",
    "\n",
    "os.chdir(os.path.expanduser('~/vivarium-ecoli'))\n",
    "\n",
    "ALLOWED_METAL_NAMES =   {'Iron': 'FE+2', 'Cobalt': 'CO+2', 'Copper': 'CU+2', 'Manganese': 'MN+2', 'Molybdenum': 'CPD-8123', 'Nickel': 'NI+2', 'Zinc': 'ZN+2',\n",
    "                        'Calcium': 'CA+2', 'Magnesium': 'MG+2', 'Sodium': 'NA+', 'Potassium': 'K+',\n",
    "                        'Iron-sulfur \\(4Fe-4S\\)': 'CPD-7', 'Iron-sulfur \\(2Fe-2S\\)': 'CPD-6',\n",
    "                         'Iron-sulfur \\(4Fe-4S-S-AdoMet\\)': 'CPD-7', 'Iron-sulfur \\(3Fe-4S\\)': '3FE-4S', 'Iron-oxo-sulfur \\(4Fe-2O-2S\\)': 'CPD-7',\n",
    "                        'heme': 'Heme-b', 'Molybdate': 'CPD-3', 'heme B': 'Heme-b',\n",
    "                         'L-SELENOCYSTEINE': 'L-SELENOCYSTEINE',\n",
    "                        'Divalent metal cation': 'Any+2'}\n",
    "\n",
    "\n",
    "ACCEPTED_OTHER_FEATURES = {'PYRIDOXAL_PHOSPHATE', 'THIAMINE-PYROPHOSPHATE', 'FMN', 'FAD'}"
   ]
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Connect to api"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "outputs": [],
   "source": [
    "password = input(\"Enter Password: \")"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2023-08-04T23:08:34.497760Z",
     "start_time": "2023-08-04T23:08:30.187464Z"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "outputs": [
    {
     "data": {
      "text/plain": "<Response [200]>"
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "s = requests.Session() # create session\n",
    "# Post login credentials to session:\n",
    "s.post('https://websvc.biocyc.org/credentials/login/', data={'email':'cellulararchitect@protonmail.com', 'password': password})"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2023-08-04T23:08:35.461912Z",
     "start_time": "2023-08-04T23:08:35.194795Z"
    }
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "## api test"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'@ID': 'ECOLI:CHEW-MONOMER',\n",
      " '@detail': 'full',\n",
      " '@frameid': 'CHEW-MONOMER',\n",
      " '@orgid': 'ECOLI',\n",
      " 'citation': [{'Publication': {'@ID': 'ECOLI:PUB-21463637',\n",
      "                               '@detail': 'full',\n",
      "                               '@frameid': 'PUB-21463637',\n",
      "                               '@orgid': 'ECOLI',\n",
      "                               'author': [{'#text': 'Underbakke ES',\n",
      "                                           '@datatype': 'string'},\n",
      "                                          {'#text': 'Zhu Y',\n",
      "                                           '@datatype': 'string'},\n",
      "                                          {'#text': 'Kiessling LL',\n",
      "                                           '@datatype': 'string'}],\n",
      "                               'pubmed-id': {'#text': '21463637',\n",
      "                                             '@datatype': 'string'},\n",
      "                               'source': {'#text': 'J Mol Biol 409(4);483-95',\n",
      "                                          '@datatype': 'string'},\n",
      "                               'title': {'#text': 'Protein footprinting in a '\n",
      "                                                  'complex milieu: identifying '\n",
      "                                                  'the interaction surfaces of '\n",
      "                                                  'the chemotaxis adaptor '\n",
      "                                                  'protein CheW.',\n",
      "                                         '@datatype': 'string'},\n",
      "                               'year': {'#text': '2011',\n",
      "                                        '@datatype': 'integer'}}},\n",
      "              {'Publication': {'@ID': 'ECOLI:PUB-1860813',\n",
      "                               '@detail': 'full',\n",
      "                               '@frameid': 'PUB-1860813',\n",
      "                               '@orgid': 'ECOLI',\n",
      "                               'author': [{'#text': 'Liu JD',\n",
      "                                           '@datatype': 'string'},\n",
      "                                          {'#text': 'Parkinson JS',\n",
      "                                           '@datatype': 'string'}],\n",
      "                               'pubmed-id': {'#text': '1860813',\n",
      "                                             '@datatype': 'string'},\n",
      "                               'source': {'#text': 'J Bacteriol '\n",
      "                                                   '173(16);4941-51',\n",
      "                                          '@datatype': 'string'},\n",
      "                               'title': {'#text': 'Genetic evidence for '\n",
      "                                                  'interaction between the '\n",
      "                                                  'CheW and Tsr proteins '\n",
      "                                                  'during chemoreceptor '\n",
      "                                                  'signaling by Escherichia '\n",
      "                                                  'coli.',\n",
      "                                         '@datatype': 'string'},\n",
      "                               'year': {'#text': '1991',\n",
      "                                        '@datatype': 'integer'}}},\n",
      "              {'Publication': {'@ID': 'ECOLI:PUB-2682657',\n",
      "                               '@detail': 'full',\n",
      "                               '@frameid': 'PUB-2682657',\n",
      "                               '@orgid': 'ECOLI',\n",
      "                               'author': [{'#text': 'Liu JD',\n",
      "                                           '@datatype': 'string'},\n",
      "                                          {'#text': 'Parkinson JS',\n",
      "                                           '@datatype': 'string'}],\n",
      "                               'pubmed-id': {'#text': '2682657',\n",
      "                                             '@datatype': 'string'},\n",
      "                               'source': {'#text': 'Proc Natl Acad Sci U S A '\n",
      "                                                   '86(22);8703-7',\n",
      "                                          '@datatype': 'string'},\n",
      "                               'title': {'#text': 'Role of CheW protein in '\n",
      "                                                  'coupling membrane receptors '\n",
      "                                                  'to the intracellular '\n",
      "                                                  'signaling system of '\n",
      "                                                  'bacterial chemotaxis.',\n",
      "                                         '@datatype': 'string'},\n",
      "                               'year': {'#text': '1989',\n",
      "                                        '@datatype': 'integer'}}},\n",
      "              {'Publication': {'@ID': 'ECOLI:PUB-8456299',\n",
      "                               '@detail': 'full',\n",
      "                               '@frameid': 'PUB-8456299',\n",
      "                               '@orgid': 'ECOLI',\n",
      "                               'author': [{'#text': 'Maddock JR',\n",
      "                                           '@datatype': 'string'},\n",
      "                                          {'#text': 'Shapiro L',\n",
      "                                           '@datatype': 'string'}],\n",
      "                               'pubmed-id': {'#text': '8456299',\n",
      "                                             '@datatype': 'string'},\n",
      "                               'source': {'#text': 'Science 259(5102);1717-23',\n",
      "                                          '@datatype': 'string'},\n",
      "                               'title': {'#text': 'Polar location of the '\n",
      "                                                  'chemoreceptor complex in '\n",
      "                                                  'the Escherichia coli cell.',\n",
      "                                         '@datatype': 'string'},\n",
      "                               'year': {'#text': '1993',\n",
      "                                        '@datatype': 'integer'}}},\n",
      "              {'Publication': {'@ID': 'ECOLI:PUB-23950985',\n",
      "                               '@detail': 'full',\n",
      "                               '@frameid': 'PUB-23950985',\n",
      "                               '@orgid': 'ECOLI',\n",
      "                               'author': [{'#text': 'Cashman DJ',\n",
      "                                           '@datatype': 'string'},\n",
      "                                          {'#text': 'Ortega DR',\n",
      "                                           '@datatype': 'string'},\n",
      "                                          {'#text': 'Zhulin IB',\n",
      "                                           '@datatype': 'string'},\n",
      "                                          {'#text': 'Baudry J',\n",
      "                                           '@datatype': 'string'}],\n",
      "                               'pubmed-id': {'#text': '23950985',\n",
      "                                             '@datatype': 'string'},\n",
      "                               'source': {'#text': 'PLoS One 8(8);e70705',\n",
      "                                          '@datatype': 'string'},\n",
      "                               'title': {'#text': 'Homology modeling of the '\n",
      "                                                  'CheW coupling protein of '\n",
      "                                                  'the chemotaxis signaling '\n",
      "                                                  'complex.',\n",
      "                                         '@datatype': 'string'},\n",
      "                               'year': {'#text': '2013',\n",
      "                                        '@datatype': 'integer'}}},\n",
      "              {'Publication': {'@ID': 'ECOLI:PUB-3510184',\n",
      "                               '@detail': 'full',\n",
      "                               '@frameid': 'PUB-3510184',\n",
      "                               '@orgid': 'ECOLI',\n",
      "                               'author': [{'#text': 'Mutoh N',\n",
      "                                           '@datatype': 'string'},\n",
      "                                          {'#text': 'Simon MI',\n",
      "                                           '@datatype': 'string'}],\n",
      "                               'pubmed-id': {'#text': '3510184',\n",
      "                                             '@datatype': 'string'},\n",
      "                               'source': {'#text': 'J Bacteriol 165(1);161-6',\n",
      "                                          '@datatype': 'string'},\n",
      "                               'title': {'#text': 'Nucleotide sequence '\n",
      "                                                  'corresponding to five '\n",
      "                                                  'chemotaxis genes in '\n",
      "                                                  'Escherichia coli.',\n",
      "                                         '@datatype': 'string'},\n",
      "                               'year': {'#text': '1986',\n",
      "                                        '@datatype': 'integer'}}},\n",
      "              {'Publication': {'@ID': 'ECOLI:PUB-10972797',\n",
      "                               '@detail': 'full',\n",
      "                               '@frameid': 'PUB-10972797',\n",
      "                               '@orgid': 'ECOLI',\n",
      "                               'author': [{'#text': 'Sourjik V',\n",
      "                                           '@datatype': 'string'},\n",
      "                                          {'#text': 'Berg HC',\n",
      "                                           '@datatype': 'string'}],\n",
      "                               'pubmed-id': {'#text': '10972797',\n",
      "                                             '@datatype': 'string'},\n",
      "                               'source': {'#text': 'Mol Microbiol 37(4);740-51',\n",
      "                                          '@datatype': 'string'},\n",
      "                               'title': {'#text': 'Localization of components '\n",
      "                                                  'of the chemotaxis machinery '\n",
      "                                                  'of Escherichia coli using '\n",
      "                                                  'fluorescent protein '\n",
      "                                                  'fusions.',\n",
      "                                         '@datatype': 'string'},\n",
      "                               'year': {'#text': '2000',\n",
      "                                        '@datatype': 'integer'}}},\n",
      "              {'Publication': {'@ID': 'ECOLI:PUB-31925330',\n",
      "                               '@detail': 'full',\n",
      "                               '@frameid': 'PUB-31925330',\n",
      "                               '@orgid': 'ECOLI',\n",
      "                               'author': [{'#text': 'Cassidy CK',\n",
      "                                           '@datatype': 'string'},\n",
      "                                          {'#text': 'Himes BA',\n",
      "                                           '@datatype': 'string'},\n",
      "                                          {'#text': 'Sun D',\n",
      "                                           '@datatype': 'string'},\n",
      "                                          {'#text': 'Ma J',\n",
      "                                           '@datatype': 'string'},\n",
      "                                          {'#text': 'Zhao G',\n",
      "                                           '@datatype': 'string'},\n",
      "                                          {'#text': 'Parkinson JS',\n",
      "                                           '@datatype': 'string'},\n",
      "                                          {'#text': 'Stansfeld PJ',\n",
      "                                           '@datatype': 'string'},\n",
      "                                          {'#text': 'Luthey-Schulten Z',\n",
      "                                           '@datatype': 'string'},\n",
      "                                          {'#text': 'Zhang P',\n",
      "                                           '@datatype': 'string'}],\n",
      "                               'pubmed-id': {'#text': '31925330',\n",
      "                                             '@datatype': 'string'},\n",
      "                               'source': {'#text': 'Commun Biol 3(1);24',\n",
      "                                          '@datatype': 'string'},\n",
      "                               'title': {'#text': 'Structure and dynamics of '\n",
      "                                                  'the E. coli chemotaxis core '\n",
      "                                                  'signaling complex by '\n",
      "                                                  'cryo-electron tomography '\n",
      "                                                  'and molecular simulations.',\n",
      "                                         '@datatype': 'string'},\n",
      "                               'year': {'#text': '2020',\n",
      "                                        '@datatype': 'integer'}}},\n",
      "              {'Publication': {'@ID': 'ECOLI:PUB-35077199',\n",
      "                               '@detail': 'full',\n",
      "                               '@frameid': 'PUB-35077199',\n",
      "                               '@orgid': 'ECOLI',\n",
      "                               'author': [{'#text': 'Pinas GE',\n",
      "                                           '@datatype': 'string'},\n",
      "                                          {'#text': 'DeSantis MD',\n",
      "                                           '@datatype': 'string'},\n",
      "                                          {'#text': 'Cassidy CK',\n",
      "                                           '@datatype': 'string'},\n",
      "                                          {'#text': 'Parkinson JS',\n",
      "                                           '@datatype': 'string'}],\n",
      "                               'doi-id': {'#text': '10.1126/scisignal.abj1737',\n",
      "                                          '@datatype': 'string'},\n",
      "                               'pubmed-id': {'#text': '35077199',\n",
      "                                             '@datatype': 'string'},\n",
      "                               'source': {'#text': 'Sci Signal '\n",
      "                                                   '15(718);eabj1737',\n",
      "                                          '@datatype': 'string'},\n",
      "                               'title': {'#text': 'Hexameric rings of the '\n",
      "                                                  'scaffolding protein CheW '\n",
      "                                                  'enhance response '\n",
      "                                                  'sensitivity and '\n",
      "                                                  'cooperativity in  '\n",
      "                                                  '<i>Escherichia coli</i> '\n",
      "                                                  'chemoreceptor arrays.',\n",
      "                                         '@datatype': 'string'},\n",
      "                               'year': {'#text': '2022',\n",
      "                                        '@datatype': 'integer'}}}],\n",
      " 'comment': {'#text': 'CheW is the coupling protein in the ternary receptor '\n",
      "                      'complexes of two-component signaling pathways '\n",
      "                      '(chemotaxis, thermotaxis and energy taxis) which '\n",
      "                      'respond to environmental cues (both external and '\n",
      "                      'internal) and signal change to the flagellar switch '\n",
      "                      'complex to regulate swimming behavior. \\n'\n",
      "                      '<p>\\n'\n",
      "                      'CheW forms ternary signaling complexes with '\n",
      "                      'chemoreceptor proteins (<a '\n",
      "                      'href=\"/gene?orgid=ECOLI&id=TSR-MONOMER\">Tsr</a>, <a '\n",
      "                      'href=\"/gene?orgid=ECOLI&id=TAR-MONOMER\">Tar</a>, <a '\n",
      "                      'href=\"/gene?orgid=ECOLI&id=TRG-MONOMER\">Trg</a> and <a '\n",
      "                      'href=\"/gene?orgid=ECOLI&id=TAP-MONOMER\">Tap</a>) and '\n",
      "                      'the protein histidine kinase <a '\n",
      "                      'href=\"/gene?orgid=ECOLI&id=PROTEIN-CHEA\">CheA</a>; '\n",
      "                      'ligand binding to the receptor proteins generates a '\n",
      "                      'signal that is transmitted across the inner membrane to '\n",
      "                      'regulate CheA histidine kinase activity and the '\n",
      "                      'subsequent flow of phosphoryl groups to the response '\n",
      "                      'regulators <a '\n",
      "                      'href=\"/gene?orgid=ECOLI&id=CHEY-MONOMER\">CheY</a> - '\n",
      "                      'whose phosphorylation status regulates the direction of '\n",
      "                      'flagellar motor rotation - and <a '\n",
      "                      'href=\"/gene?orgid=ECOLI&id=CHEB-MONOMER\">CheB</a> - a '\n",
      "                      'protein glutamate methylesterase which modulates the '\n",
      "                      'sensory adaption system (see reviews [<a '\n",
      "                      'href=\"http://www.ncbi.nlm.nih.gov/pubmed/9442881\">Falke97</a>, '\n",
      "                      '<a '\n",
      "                      'href=\"http://www.ncbi.nlm.nih.gov/pubmed/25834953\">Parkinson15</a>]). '\n",
      "                      'CheW also assembles into higher-order complexes with '\n",
      "                      'the receptor <a '\n",
      "                      'href=\"/gene?orgid=ECOLI&id=G7595-MONOMER\">Aer</a> and '\n",
      "                      'CheA to mediate the aerotaxis (oxygen-sensing) response '\n",
      "                      '[<a '\n",
      "                      'href=\"http://www.ncbi.nlm.nih.gov/pubmed/27803157\">Samanta16</a>]. '\n",
      "                      'CheW is required for the chemotactic response to PTS '\n",
      "                      'sugars [<a '\n",
      "                      'href=\"http://www.ncbi.nlm.nih.gov/pubmed/7592359\">Rowsell95</a>, '\n",
      "                      '<a '\n",
      "                      'href=\"http://www.ncbi.nlm.nih.gov/pubmed/10198062\">Lux99</a>] '\n",
      "                      '(see also [<a '\n",
      "                      'href=\"http://www.ncbi.nlm.nih.gov/pubmed/22778402\">Neumann12</a>]). \\n'\n",
      "                      '<p>\\n'\n",
      "                      'Purified CheW and CheA form a reversible complex <i>in '\n",
      "                      'vitro</i> with a stoichiometry of two CheW monomers per '\n",
      "                      'CheA dimer [<a '\n",
      "                      'href=\"http://www.ncbi.nlm.nih.gov/pubmed/1992467\">Gegner91</a>] '\n",
      "                      '(see also [<a '\n",
      "                      'href=\"http://www.ncbi.nlm.nih.gov/pubmed/1326408\">Gegner92</a>]). '\n",
      "                      'CheW forms a stable ternary complex with CheA and '\n",
      "                      'receptor protein (Tsr or Tar); autophosphorylation of '\n",
      "                      'CheA is stimulated in the presence of wild-type '\n",
      "                      '(unliganded) receptor (Tar/Tsr) and CheW protein; '\n",
      "                      'addition of ligand (aspartate/serine) inhibits '\n",
      "                      'formation of CheA-phosphate; addition of CheY and CheB '\n",
      "                      'results in the disappearance of CheA-phosphate [<a '\n",
      "                      'href=\"http://www.ncbi.nlm.nih.gov/pubmed/2645576\">Borkovich89</a>, '\n",
      "                      '<a '\n",
      "                      'href=\"http://www.ncbi.nlm.nih.gov/pubmed/2261645\">Borkovich90</a>, '\n",
      "                      '<a '\n",
      "                      'href=\"http://www.ncbi.nlm.nih.gov/pubmed/1326408\">Gegner92</a>]. '\n",
      "                      '<a '\n",
      "                      'href=\"/gene?orgid=ECOLI&id=PROTEIN-CHEA\">CheAL</a>/<a '\n",
      "                      'href=\"/gene?orgid=ECOLI&id=CHEA-SMALL\">CheAS</a>/CheW '\n",
      "                      'complexes can be isolated from wild-type cells; CheW '\n",
      "                      'enhances CheA binding to CheY and increases the '\n",
      "                      'autophosphorylation rate of CheA by increasing its '\n",
      "                      'affinity for ATP [<a '\n",
      "                      'href=\"http://www.ncbi.nlm.nih.gov/pubmed/2068106\">McNally91</a>]. '\n",
      "                      'CheW enhances Tsr methylation <i>in vitro</i>; serine '\n",
      "                      'and CheW promote synergistic increases in Tsr '\n",
      "                      'methylation <i>in vitro</i> [<a '\n",
      "                      'href=\"http://www.ncbi.nlm.nih.gov/pubmed/15766389\">Chalah05</a>].\\n'\n",
      "                      '<p>\\n'\n",
      "                      'Excess levels of CheW disrupt CheA activation and '\n",
      "                      'chemotactic response [<a '\n",
      "                      'href=\"http://www.ncbi.nlm.nih.gov/pubmed/2681160\">Sanders89</a>, '\n",
      "                      '<a '\n",
      "                      'href=\"http://www.ncbi.nlm.nih.gov/pubmed/11923283\">Boukhvalova02</a>] '\n",
      "                      'possibly by disrupting the normal formation of receptor '\n",
      "                      'complexes [<a '\n",
      "                      'href=\"http://www.ncbi.nlm.nih.gov/pubmed/16230637\">Studdert05</a>, '\n",
      "                      '<a '\n",
      "                      'href=\"http://www.ncbi.nlm.nih.gov/pubmed/20487303\">Cardozo10</a>].\\n'\n",
      "                      '<p>\\n'\n",
      "                      'CheW interacts with chemoreceptors in a manner similar '\n",
      "                      'to CheA; CheW and CheA bind to the same region of '\n",
      "                      'chemoreceptors due to structural similarity between '\n",
      "                      'CheW and the regulatory or P5 domain of CheA [<a '\n",
      "                      'href=\"http://www.ncbi.nlm.nih.gov/pubmed/22659323\">Wang12f</a>, '\n",
      "                      '<a '\n",
      "                      'href=\"http://www.ncbi.nlm.nih.gov/pubmed/25060668\">Pedetta14</a>]. '\n",
      "                      'CheW adopts an overall fold similar to <i>Thermotoga '\n",
      "                      'maritima</i> CheW; the two proteins have 29% sequence '\n",
      "                      'identity [<a '\n",
      "                      'href=\"http://www.ncbi.nlm.nih.gov/pubmed/17631272\">Li07d</a>].\\n'\n",
      "                      '<p>\\n'\n",
      "                      'Early work in <i>E. coli</i> K-12 characterized '\n",
      "                      '<i>cheW</i> mutants, all of which were motile, '\n",
      "                      'nonchemotactic, smooth swimmers [<a '\n",
      "                      'href=\"http://www.ncbi.nlm.nih.gov/pubmed/353036\">Parkinson78</a>, '\n",
      "                      '<a '\n",
      "                      'href=\"http://www.ncbi.nlm.nih.gov/pubmed/770453\">Parkinson76</a>, '\n",
      "                      '<a '\n",
      "                      'href=\"http://www.ncbi.nlm.nih.gov/pubmed/324984\">Silverman77a</a>, '\n",
      "                      '<a '\n",
      "                      'href=\"http://www.ncbi.nlm.nih.gov/pubmed/336614\">Matsumura77</a>] '\n",
      "                      '(see also [<a '\n",
      "                      'href=\"http://www.ncbi.nlm.nih.gov/pubmed/3553150\">Wolfe87</a>].\\n'\n",
      "                      '<p>\\n'\n",
      "                      '<i>cheW</i> was found upregulated in two MG1655 '\n",
      "                      'lysogens carrying closely related Stx2a phages O104 and '\n",
      "                      'PA8 [<a '\n",
      "                      'href=\"http://www.ncbi.nlm.nih.gov/pubmed/31208335\">Berger19</a>].\\n'\n",
      "                      '<p>\\n'\n",
      "                      'Reviews: [<a '\n",
      "                      'href=\"http://www.ncbi.nlm.nih.gov/pubmed/30642465\">Huang19</a>]\\n'\n",
      "                      '<p>\\n'\n",
      "                      '\\n'\n",
      "                      '<p>',\n",
      "             '@datatype': 'string'},\n",
      " 'common-name': {'#text': 'chemotaxis protein CheW', '@datatype': 'string'},\n",
      " 'component-of': {'Protein': [{'@frameid': 'TRG-CPLX',\n",
      "                               '@orgid': 'ECOLI',\n",
      "                               '@resource': 'getxml?ECOLI:TRG-CPLX'},\n",
      "                              {'@frameid': 'TAP-CPLX',\n",
      "                               '@orgid': 'ECOLI',\n",
      "                               '@resource': 'getxml?ECOLI:TAP-CPLX'},\n",
      "                              {'@frameid': 'TSR-CPLX',\n",
      "                               '@orgid': 'ECOLI',\n",
      "                               '@resource': 'getxml?ECOLI:TSR-CPLX'},\n",
      "                              {'@frameid': 'TAR-CPLX',\n",
      "                               '@orgid': 'ECOLI',\n",
      "                               '@resource': 'getxml?ECOLI:TAR-CPLX'}]},\n",
      " 'credits': {'last-curated': {'Organization': {'@frameid': 'O0-13',\n",
      "                                               '@orgid': 'ECOLI',\n",
      "                                               '@resource': '#ECOLI:O0-13'},\n",
      "                              'Person': {'@frameid': 'amackie',\n",
      "                                         '@orgid': 'ECOLI',\n",
      "                                         '@resource': '#ECOLI:amackie'},\n",
      "                              'date': {'#text': '2018-07-02',\n",
      "                                       '@datatype': 'date'}},\n",
      "             'revised': {'Organization': {'@frameid': 'O0-13',\n",
      "                                          '@orgid': 'ECOLI',\n",
      "                                          '@resource': '#ECOLI:O0-13'},\n",
      "                         'Person': {'@ID': 'ECOLI:amackie',\n",
      "                                    '@detail': 'full',\n",
      "                                    '@frameid': 'amackie',\n",
      "                                    '@orgid': 'ECOLI',\n",
      "                                    'affiliations': {'Organization': {'@ID': 'ECOLI:O0-13',\n",
      "                                                                      '@detail': 'full',\n",
      "                                                                      '@frameid': 'O0-13',\n",
      "                                                                      '@orgid': 'ECOLI',\n",
      "                                                                      'common-name': {'#text': 'Macquarie '\n",
      "                                                                                               'University',\n",
      "                                                                                      '@datatype': 'string'}}},\n",
      "                                    'email': {'#text': 'amanda.mackie@mq.edu.au',\n",
      "                                              '@datatype': 'string'}},\n",
      "                         'date': {'#text': '2018-07-02', '@datatype': 'date'}}},\n",
      " 'dblink': [{'dblink-URL': 'https://alphafold.ebi.ac.uk/entry/P0A964',\n",
      "             'dblink-db': 'ALPHAFOLD',\n",
      "             'dblink-oid': 'P0A964',\n",
      "             'dblink-relationship': 'unification'},\n",
      "            {'dblink-URL': 'http://www.rcsb.org/structure/6S1K',\n",
      "             'dblink-db': 'PDB',\n",
      "             'dblink-oid': '6S1K',\n",
      "             'dblink-relationship': 'Structure'},\n",
      "            {'dblink-URL': 'http://www.ebi.ac.uk/interpro/entry/IPR039315',\n",
      "             'dblink-db': 'INTERPRO',\n",
      "             'dblink-oid': 'IPR039315',\n",
      "             'dblink-relationship': 'IN-FAMILY'},\n",
      "            {'dblink-URL': 'http://www.pantherdb.org/panther/family.do?clsAccession=PTHR22617',\n",
      "             'dblink-db': 'PANTHER',\n",
      "             'dblink-oid': 'PTHR22617',\n",
      "             'dblink-relationship': 'IN-FAMILY'},\n",
      "            {'dblink-URL': 'http://www.ebi.ac.uk/interpro/entry/IPR036061',\n",
      "             'dblink-db': 'INTERPRO',\n",
      "             'dblink-oid': 'IPR036061',\n",
      "             'dblink-relationship': 'IN-FAMILY'},\n",
      "            {'dblink-URL': 'https://www.ebi.ac.uk/interpro/entry/pfam/PF01584',\n",
      "             'dblink-db': 'PFAM',\n",
      "             'dblink-oid': 'PF01584',\n",
      "             'dblink-relationship': 'IN-FAMILY'},\n",
      "            {'dblink-URL': 'http://www.ebi.ac.uk/pride/searchSummary.do?queryTypeSelected=identification%20accession%20number&identificationAccessionNumber=P0A964',\n",
      "             'dblink-db': 'PRIDE',\n",
      "             'dblink-oid': 'P0A964',\n",
      "             'dblink-relationship': 'unification'},\n",
      "            {'dblink-URL': 'https://swissmodel.expasy.org/repository/uniprot/P0A964',\n",
      "             'dblink-db': 'SMR',\n",
      "             'dblink-oid': 'P0A964',\n",
      "             'dblink-relationship': 'unification'},\n",
      "            {'dblink-URL': 'http://dip.doe-mbi.ucla.edu/dip/Browse.cgi?ID=DIP-48236N',\n",
      "             'dblink-db': 'DIP',\n",
      "             'dblink-oid': 'DIP-48236N',\n",
      "             'dblink-relationship': 'unification'},\n",
      "            {'dblink-URL': 'http://www.ebi.ac.uk/interpro/entry/IPR002545',\n",
      "             'dblink-db': 'INTERPRO',\n",
      "             'dblink-oid': 'IPR002545',\n",
      "             'dblink-relationship': 'IN-FAMILY'},\n",
      "            {'dblink-URL': 'http://smart.embl.de/smart/do_annotation.pl?DOMAIN=SM00260',\n",
      "             'dblink-db': 'SMART',\n",
      "             'dblink-oid': 'SM00260',\n",
      "             'dblink-relationship': 'IN-FAMILY'},\n",
      "            {'dblink-URL': 'http://prosite.expasy.org/cgi-bin/prosite/nicedoc.pl?PS50851',\n",
      "             'dblink-db': 'PROSITE',\n",
      "             'dblink-oid': 'PS50851',\n",
      "             'dblink-relationship': 'IN-FAMILY'},\n",
      "            {'dblink-URL': 'http://www.rcsb.org/structure/2HO9',\n",
      "             'dblink-db': 'PDB',\n",
      "             'dblink-oid': '2HO9',\n",
      "             'dblink-relationship': 'Structure'},\n",
      "            {'dblink-URL': 'http://pir.georgetown.edu/cgi-bin/pro/entry_pro?id=PRO_000022280',\n",
      "             'dblink-db': 'PRODB',\n",
      "             'dblink-oid': 'PRO_000022280',\n",
      "             'dblink-relationship': 'unification'},\n",
      "            {'dblink-URL': 'http://ecoliwiki.org/colipedia/index.php/b1887',\n",
      "             'dblink-db': 'ECOLIWIKI',\n",
      "             'dblink-oid': 'b1887',\n",
      "             'dblink-relationship': 'unification'},\n",
      "            {'dblink-URL': 'http://modbase.compbio.ucsf.edu/modbase-cgi/model_search.cgi?searchkw=name&kword=P0A964',\n",
      "             'dblink-db': 'MODBASE',\n",
      "             'dblink-oid': 'P0A964',\n",
      "             'dblink-relationship': 'unification'},\n",
      "            {'dblink-URL': 'https://www.uniprot.org/uniprot/P0A964',\n",
      "             'dblink-db': 'UNIPROT',\n",
      "             'dblink-oid': 'P0A964',\n",
      "             'dblink-relationship': 'unification'},\n",
      "            {'dblink-URL': 'http://www.ncbi.nlm.nih.gov/entrez/viewer.fcgi?val=NP_416401',\n",
      "             'dblink-db': 'REFSEQ',\n",
      "             'dblink-oid': 'NP_416401',\n",
      "             'dblink-relationship': 'unification'}],\n",
      " 'evidence': [{'Evidence-Code': {'@ID': 'ECOLI:EV-EXP-IDA-PURIFIED-PROTEIN',\n",
      "                                 '@class': 'true',\n",
      "                                 '@detail': 'full',\n",
      "                                 '@frameid': 'EV-EXP-IDA-PURIFIED-PROTEIN',\n",
      "                                 '@orgid': 'ECOLI',\n",
      "                                 'comment': {'#text': 'Direct assay of '\n",
      "                                                      'purified protein. '\n",
      "                                                      'Protein purified to '\n",
      "                                                      'homogeneity, and\\n'\n",
      "                                                      'activity measured using '\n",
      "                                                      'an in vitro assay.  '\n",
      "                                                      'Expression host is '\n",
      "                                                      'unspecified.',\n",
      "                                             '@datatype': 'string'},\n",
      "                                 'common-name': {'#text': 'Assay of protein '\n",
      "                                                          'purified  to '\n",
      "                                                          'homogeneity',\n",
      "                                                 '@datatype': 'string'},\n",
      "                                 'parent': {'Evidence-Code': {'@class': 'true',\n",
      "                                                              '@frameid': 'EV-EXP-IDA',\n",
      "                                                              '@orgid': 'ECOLI',\n",
      "                                                              '@resource': 'getxml?ECOLI:EV-EXP-IDA'}}},\n",
      "               'Publication': {'@ID': 'ECOLI:PUB-1326408',\n",
      "                               '@detail': 'full',\n",
      "                               '@frameid': 'PUB-1326408',\n",
      "                               '@orgid': 'ECOLI',\n",
      "                               'author': [{'#text': 'Gegner JA',\n",
      "                                           '@datatype': 'string'},\n",
      "                                          {'#text': 'Graham DR',\n",
      "                                           '@datatype': 'string'},\n",
      "                                          {'#text': 'Roth AF',\n",
      "                                           '@datatype': 'string'},\n",
      "                                          {'#text': 'Dahlquist FW',\n",
      "                                           '@datatype': 'string'}],\n",
      "                               'pubmed-id': {'#text': '1326408',\n",
      "                                             '@datatype': 'string'},\n",
      "                               'source': {'#text': 'Cell 70(6);975-82',\n",
      "                                          '@datatype': 'string'},\n",
      "                               'title': {'#text': 'Assembly of an MCP '\n",
      "                                                  'receptor, CheW, and kinase '\n",
      "                                                  'CheA complex in the '\n",
      "                                                  'bacterial chemotaxis signal '\n",
      "                                                  'transduction pathway.',\n",
      "                                         '@datatype': 'string'},\n",
      "                               'year': {'#text': '1992',\n",
      "                                        '@datatype': 'integer'}}},\n",
      "              {'Evidence-Code': {'@ID': 'ECOLI:EV-EXP-IMP',\n",
      "                                 '@class': 'true',\n",
      "                                 '@detail': 'full',\n",
      "                                 '@frameid': 'EV-EXP-IMP',\n",
      "                                 '@orgid': 'ECOLI',\n",
      "                                 'comment': {'#text': 'IMP inferred from '\n",
      "                                                      'mutant phenotype.\\n'\n",
      "                                                      '\\n'\n",
      "                                                      '<br>\\n'\n",
      "                                                      'The assertion was '\n",
      "                                                      'inferred from a mutant '\n",
      "                                                      'phenotype such as\\n'\n",
      "                                                      '<ul>\\n'\n",
      "                                                      '    <li> Any gene '\n",
      "                                                      'mutation/knockout\\n'\n",
      "                                                      '    <li> '\n",
      "                                                      'Overexpression/ectopic '\n",
      "                                                      'expression of wild-type '\n",
      "                                                      'or mutant genes\\n'\n",
      "                                                      '    <li> Anti-sense '\n",
      "                                                      'experiments\\n'\n",
      "                                                      '    <li> RNA '\n",
      "                                                      'interference '\n",
      "                                                      'experiments\\n'\n",
      "                                                      '    <li> Specific '\n",
      "                                                      'protein inhibitors\\n'\n",
      "                                                      '    <li> '\n",
      "                                                      'Complementation\\n'\n",
      "                                                      '</ul><p '\n",
      "                                                      'class=ecocomment>\\t\\n'\n",
      "                                                      'Comment: Inferences '\n",
      "                                                      'made from examining '\n",
      "                                                      'mutations or\\n'\n",
      "                                                      'abnormal levels of only '\n",
      "                                                      'the product(s) of the '\n",
      "                                                      'gene of interest are '\n",
      "                                                      'covered by code EV-IMP\\n'\n",
      "                                                      '(compare to code '\n",
      "                                                      'EV-IGI). Use this code '\n",
      "                                                      'for experiments that '\n",
      "                                                      'use antibodies \\n'\n",
      "                                                      'or other specific '\n",
      "                                                      'inhibitors of RNA or '\n",
      "                                                      'protein activity, even '\n",
      "                                                      'though no\\n'\n",
      "                                                      'gene may be mutated '\n",
      "                                                      '(the rationale is that '\n",
      "                                                      'EV-IMP is used where '\n",
      "                                                      'an\\n'\n",
      "                                                      'abnormal situation '\n",
      "                                                      'prevails in a cell or '\n",
      "                                                      'organism).',\n",
      "                                             '@datatype': 'string'},\n",
      "                                 'common-name': {'#text': 'Inferred from '\n",
      "                                                          'mutant phenotype',\n",
      "                                                 '@datatype': 'string'},\n",
      "                                 'parent': {'Evidence-Code': {'@class': 'true',\n",
      "                                                              '@frameid': 'EV-EXP',\n",
      "                                                              '@orgid': 'ECOLI',\n",
      "                                                              '@resource': 'getxml?ECOLI:EV-EXP'}}},\n",
      "               'Publication': {'@ID': 'ECOLI:PUB-353036',\n",
      "                               '@detail': 'full',\n",
      "                               '@frameid': 'PUB-353036',\n",
      "                               '@orgid': 'ECOLI',\n",
      "                               'author': {'#text': 'Parkinson JS',\n",
      "                                          '@datatype': 'string'},\n",
      "                               'pubmed-id': {'#text': '353036',\n",
      "                                             '@datatype': 'string'},\n",
      "                               'source': {'#text': 'J Bacteriol 135(1);45-53',\n",
      "                                          '@datatype': 'string'},\n",
      "                               'title': {'#text': 'Complementation analysis '\n",
      "                                                  'and deletion mapping of '\n",
      "                                                  'Escherichia coli mutants '\n",
      "                                                  'defective in chemotaxis.',\n",
      "                                         '@datatype': 'string'},\n",
      "                               'year': {'#text': '1978',\n",
      "                                        '@datatype': 'integer'}}},\n",
      "              {'Evidence-Code': {'@ID': 'ECOLI:EV-EXP-IDA-UNPURIFIED-PROTEIN',\n",
      "                                 '@class': 'true',\n",
      "                                 '@detail': 'full',\n",
      "                                 '@frameid': 'EV-EXP-IDA-UNPURIFIED-PROTEIN',\n",
      "                                 '@orgid': 'ECOLI',\n",
      "                                 'comment': {'#text': 'Direct assay of '\n",
      "                                                      'unpurified protein. '\n",
      "                                                      'Presence of a protein '\n",
      "                                                      'activity is\\n'\n",
      "                                                      'indicated by an assay, '\n",
      "                                                      'but the protein has not '\n",
      "                                                      'been purified.\\n'\n",
      "                                                      'Expression host is '\n",
      "                                                      'unspecified.',\n",
      "                                             '@datatype': 'string'},\n",
      "                                 'common-name': {'#text': 'Assay of unpurified '\n",
      "                                                          'protein',\n",
      "                                                 '@datatype': 'string'},\n",
      "                                 'parent': {'Evidence-Code': {'@class': 'true',\n",
      "                                                              '@frameid': 'EV-EXP-IDA',\n",
      "                                                              '@orgid': 'ECOLI',\n",
      "                                                              '@resource': 'getxml?ECOLI:EV-EXP-IDA'}}},\n",
      "               'Publication': {'@ID': 'ECOLI:PUB-3553150',\n",
      "                               '@detail': 'full',\n",
      "                               '@frameid': 'PUB-3553150',\n",
      "                               '@orgid': 'ECOLI',\n",
      "                               'author': [{'#text': 'Wolfe AJ',\n",
      "                                           '@datatype': 'string'},\n",
      "                                          {'#text': 'Conley MP',\n",
      "                                           '@datatype': 'string'},\n",
      "                                          {'#text': 'Kramer TJ',\n",
      "                                           '@datatype': 'string'},\n",
      "                                          {'#text': 'Berg HC',\n",
      "                                           '@datatype': 'string'}],\n",
      "                               'pubmed-id': {'#text': '3553150',\n",
      "                                             '@datatype': 'string'},\n",
      "                               'source': {'#text': 'J Bacteriol 169(5);1878-85',\n",
      "                                          '@datatype': 'string'},\n",
      "                               'title': {'#text': 'Reconstitution of signaling '\n",
      "                                                  'in bacterial chemotaxis.',\n",
      "                                         '@datatype': 'string'},\n",
      "                               'year': {'#text': '1987',\n",
      "                                        '@datatype': 'integer'}}}],\n",
      " 'gene': {'Gene': {'@frameid': 'EG10149',\n",
      "                   '@orgid': 'ECOLI',\n",
      "                   '@resource': 'getxml?ECOLI:EG10149'}},\n",
      " 'has-feature': [{'Feature': {'@ID': 'ECOLI:FTR0-1600918',\n",
      "                              '@detail': 'full',\n",
      "                              '@frameid': 'FTR0-1600918',\n",
      "                              '@orgid': 'ECOLI',\n",
      "                              'accession-1': {'#text': 'PF01584.22',\n",
      "                                              '@datatype': 'string'},\n",
      "                              'evidence': {'Evidence-Code': {'@ID': 'ECOLI:EV-COMP-AINF',\n",
      "                                                             '@class': 'true',\n",
      "                                                             '@detail': 'full',\n",
      "                                                             '@frameid': 'EV-COMP-AINF',\n",
      "                                                             '@orgid': 'ECOLI',\n",
      "                                                             'comment': {...},\n",
      "                                                             'common-name': {'#text': 'Inferred '\n",
      "                                                                                      'computationally '\n",
      "                                                                                      'without '\n",
      "                                                                                      'human '\n",
      "                                                                                      'oversight',\n",
      "                                                                             '@datatype': 'string'},\n",
      "                                                             'parent': {...}},\n",
      "                                           'Publication': {'@ID': 'ECOLI:PUB-26673716',\n",
      "                                                           '@detail': 'full',\n",
      "                                                           '@frameid': 'PUB-26673716',\n",
      "                                                           '@orgid': 'ECOLI',\n",
      "                                                           'author': [...],\n",
      "                                                           'pubmed-id': {...},\n",
      "                                                           'source': {...},\n",
      "                                                           'title': {...},\n",
      "                                                           'year': {...}}},\n",
      "                              'feature-of': {'Protein': {'@frameid': 'CHEW-MONOMER',\n",
      "                                                         '@orgid': 'ECOLI',\n",
      "                                                         '@resource': '#ECOLI:CHEW-MONOMER'}},\n",
      "                              'left-end-position': {'#text': '18',\n",
      "                                                    '@datatype': 'integer'},\n",
      "                              'parent': {'Feature': {'@class': 'true',\n",
      "                                                     '@frameid': 'Pfam-Features',\n",
      "                                                     '@orgid': 'ECOLI',\n",
      "                                                     '@resource': 'getxml?ECOLI:Pfam-Features'}},\n",
      "                              'right-end-position': {'#text': '154',\n",
      "                                                     '@datatype': 'integer'}}},\n",
      "                 {'Feature': {'@ID': 'ECOLI:FTR0-1583302',\n",
      "                              '@detail': 'full',\n",
      "                              '@frameid': 'FTR0-1583302',\n",
      "                              '@orgid': 'ECOLI',\n",
      "                              'alternate-sequence': {'#text': 'R',\n",
      "                                                     '@datatype': 'string'},\n",
      "                              'comment': {'#text': 'UniProt: In Ref. 5; '\n",
      "                                                   'AAA23575..',\n",
      "                                          '@datatype': 'string'},\n",
      "                              'evidence': [{'Evidence-Code': {...},\n",
      "                                            'Publication': {...}},\n",
      "                                           {'Evidence-Code': {...},\n",
      "                                            'Publication': {...}}],\n",
      "                              'feature-of': {'Protein': {'@frameid': 'CHEW-MONOMER',\n",
      "                                                         '@orgid': 'ECOLI',\n",
      "                                                         '@resource': '#ECOLI:CHEW-MONOMER'}},\n",
      "                              'left-end-position': {'#text': '16',\n",
      "                                                    '@datatype': 'integer'},\n",
      "                              'parent': {'Feature': {'@class': 'true',\n",
      "                                                     '@frameid': 'Sequence-Conflicts',\n",
      "                                                     '@orgid': 'ECOLI',\n",
      "                                                     '@resource': 'getxml?ECOLI:Sequence-Conflicts'}},\n",
      "                              'right-end-position': {'#text': '16',\n",
      "                                                     '@datatype': 'integer'}}},\n",
      "                 {'Feature': {'@ID': 'ECOLI:FTR0-1583301',\n",
      "                              '@detail': 'full',\n",
      "                              '@frameid': 'FTR0-1583301',\n",
      "                              '@orgid': 'ECOLI',\n",
      "                              'comment': {'#text': 'UniProt: CheW-like.',\n",
      "                                          '@datatype': 'string'},\n",
      "                              'evidence': {'Evidence-Code': {'@ID': 'ECOLI:EV-COMP-HINF-ISM',\n",
      "                                                             '@class': 'true',\n",
      "                                                             '@detail': 'full',\n",
      "                                                             '@frameid': 'EV-COMP-HINF-ISM',\n",
      "                                                             '@orgid': 'ECOLI',\n",
      "                                                             'comment': {...},\n",
      "                                                             'common-name': {'#text': 'Inferred '\n",
      "                                                                                      'from '\n",
      "                                                                                      'Sequence '\n",
      "                                                                                      'Model',\n",
      "                                                                             '@datatype': 'string'},\n",
      "                                                             'parent': {...}},\n",
      "                                           'Publication': {'@frameid': 'PUB-UNIPROT-2021-03',\n",
      "                                                           '@orgid': 'ECOLI',\n",
      "                                                           '@resource': '#ECOLI:PUB-UNIPROT-2021-03'}},\n",
      "                              'feature-of': {'Protein': {'@frameid': 'CHEW-MONOMER',\n",
      "                                                         '@orgid': 'ECOLI',\n",
      "                                                         '@resource': '#ECOLI:CHEW-MONOMER'}},\n",
      "                              'left-end-position': {'#text': '16',\n",
      "                                                    '@datatype': 'integer'},\n",
      "                              'parent': {'Feature': {'@class': 'true',\n",
      "                                                     '@frameid': 'Conserved-Regions',\n",
      "                                                     '@orgid': 'ECOLI',\n",
      "                                                     '@resource': 'getxml?ECOLI:Conserved-Regions'}},\n",
      "                              'right-end-position': {'#text': '156',\n",
      "                                                     '@datatype': 'integer'}}},\n",
      "                 {'Feature': {'@ID': 'ECOLI:FTR0-1502710',\n",
      "                              '@detail': 'full',\n",
      "                              '@frameid': 'FTR0-1502710',\n",
      "                              '@orgid': 'ECOLI',\n",
      "                              'comment': {'#text': 'E &rarr; D: large decrease '\n",
      "                                                   'in affinity for Tar '\n",
      "                                                   'combined with a small '\n",
      "                                                   'decrease in affinity for '\n",
      "                                                   'CheA; less effective than '\n",
      "                                                   'wild-type CheW in '\n",
      "                                                   'mediating Tar-CheA '\n",
      "                                                   'coupling (as measured by '\n",
      "                                                   'CheA autokinase activity) '\n",
      "                                                   '<i>in vitro</i>',\n",
      "                                          '@datatype': 'string'},\n",
      "                              'evidence': {'Evidence-Code': {'@ID': 'ECOLI:EV-EXP-IDA',\n",
      "                                                             '@class': 'true',\n",
      "                                                             '@detail': 'full',\n",
      "                                                             '@frameid': 'EV-EXP-IDA',\n",
      "                                                             '@orgid': 'ECOLI',\n",
      "                                                             'comment': {...},\n",
      "                                                             'common-name': {'#text': 'Inferred '\n",
      "                                                                                      'from '\n",
      "                                                                                      'direct '\n",
      "                                                                                      'assay',\n",
      "                                                                             '@datatype': 'string'},\n",
      "                                                             'parent': {...}},\n",
      "                                           'Publication': {'@ID': 'ECOLI:PUB-11964403',\n",
      "                                                           '@detail': 'full',\n",
      "                                                           '@frameid': 'PUB-11964403',\n",
      "                                                           '@orgid': 'ECOLI',\n",
      "                                                           'author': [...],\n",
      "                                                           'pubmed-id': {...},\n",
      "                                                           'source': {...},\n",
      "                                                           'title': {...},\n",
      "                                                           'year': {...}}},\n",
      "                              'feature-of': {'Protein': {'@frameid': 'CHEW-MONOMER',\n",
      "                                                         '@orgid': 'ECOLI',\n",
      "                                                         '@resource': '#ECOLI:CHEW-MONOMER'}},\n",
      "                              'left-end-position': {'#text': '38',\n",
      "                                                    '@datatype': 'integer'},\n",
      "                              'parent': {'Feature': {'@class': 'true',\n",
      "                                                     '@frameid': 'Mutagenesis-Variants',\n",
      "                                                     '@orgid': 'ECOLI',\n",
      "                                                     '@resource': 'getxml?ECOLI:Mutagenesis-Variants'}},\n",
      "                              'right-end-position': {'#text': '38',\n",
      "                                                     '@datatype': 'integer'}}},\n",
      "                 {'Feature': {'@ID': 'ECOLI:FTR0-1502709',\n",
      "                              '@detail': 'full',\n",
      "                              '@frameid': 'FTR0-1502709',\n",
      "                              '@orgid': 'ECOLI',\n",
      "                              'comment': {'#text': 'L &rarr; Q: dimished '\n",
      "                                                   'affinity for CheA but not '\n",
      "                                                   'for Tar',\n",
      "                                          '@datatype': 'string'},\n",
      "                              'evidence': {'Evidence-Code': {'@class': 'true',\n",
      "                                                             '@frameid': 'EV-EXP-IDA',\n",
      "                                                             '@orgid': 'ECOLI',\n",
      "                                                             '@resource': '#ECOLI:EV-EXP-IDA'},\n",
      "                                           'Publication': {'@frameid': 'PUB-11964403',\n",
      "                                                           '@orgid': 'ECOLI',\n",
      "                                                           '@resource': '#ECOLI:PUB-11964403'}},\n",
      "                              'feature-of': {'Protein': {'@frameid': 'CHEW-MONOMER',\n",
      "                                                         '@orgid': 'ECOLI',\n",
      "                                                         '@resource': '#ECOLI:CHEW-MONOMER'}},\n",
      "                              'left-end-position': {'#text': '158',\n",
      "                                                    '@datatype': 'integer'},\n",
      "                              'parent': {'Feature': {'@class': 'true',\n",
      "                                                     '@frameid': 'Mutagenesis-Variants',\n",
      "                                                     '@orgid': 'ECOLI',\n",
      "                                                     '@resource': 'getxml?ECOLI:Mutagenesis-Variants'}},\n",
      "                              'right-end-position': {'#text': '158',\n",
      "                                                     '@datatype': 'integer'}}},\n",
      "                 {'Feature': {'@ID': 'ECOLI:FTR0-1502708',\n",
      "                              '@detail': 'full',\n",
      "                              '@frameid': 'FTR0-1502708',\n",
      "                              '@orgid': 'ECOLI',\n",
      "                              'comment': {'#text': 'M &rarr; V: dimished '\n",
      "                                                   'affinity for CheA but not '\n",
      "                                                   'for Tar',\n",
      "                                          '@datatype': 'string'},\n",
      "                              'evidence': {'Evidence-Code': {'@class': 'true',\n",
      "                                                             '@frameid': 'EV-EXP-IDA',\n",
      "                                                             '@orgid': 'ECOLI',\n",
      "                                                             '@resource': '#ECOLI:EV-EXP-IDA'},\n",
      "                                           'Publication': {'@frameid': 'PUB-11964403',\n",
      "                                                           '@orgid': 'ECOLI',\n",
      "                                                           '@resource': '#ECOLI:PUB-11964403'}},\n",
      "                              'feature-of': {'Protein': {'@frameid': 'CHEW-MONOMER',\n",
      "                                                         '@orgid': 'ECOLI',\n",
      "                                                         '@resource': '#ECOLI:CHEW-MONOMER'}},\n",
      "                              'left-end-position': {'#text': '156',\n",
      "                                                    '@datatype': 'integer'},\n",
      "                              'parent': {'Feature': {'@class': 'true',\n",
      "                                                     '@frameid': 'Mutagenesis-Variants',\n",
      "                                                     '@orgid': 'ECOLI',\n",
      "                                                     '@resource': 'getxml?ECOLI:Mutagenesis-Variants'}},\n",
      "                              'right-end-position': {'#text': '156',\n",
      "                                                     '@datatype': 'integer'}}},\n",
      "                 {'Feature': {'@ID': 'ECOLI:FTR0-1502707',\n",
      "                              '@detail': 'full',\n",
      "                              '@frameid': 'FTR0-1502707',\n",
      "                              '@orgid': 'ECOLI',\n",
      "                              'comment': {'#text': 'I &rarr; N: dimished '\n",
      "                                                   'affinity for CheA but not '\n",
      "                                                   'for Tar',\n",
      "                                          '@datatype': 'string'},\n",
      "                              'evidence': {'Evidence-Code': {'@class': 'true',\n",
      "                                                             '@frameid': 'EV-EXP-IDA',\n",
      "                                                             '@orgid': 'ECOLI',\n",
      "                                                             '@resource': '#ECOLI:EV-EXP-IDA'},\n",
      "                                           'Publication': {'@frameid': 'PUB-11964403',\n",
      "                                                           '@orgid': 'ECOLI',\n",
      "                                                           '@resource': '#ECOLI:PUB-11964403'}},\n",
      "                              'feature-of': {'Protein': {'@frameid': 'CHEW-MONOMER',\n",
      "                                                         '@orgid': 'ECOLI',\n",
      "                                                         '@resource': '#ECOLI:CHEW-MONOMER'}},\n",
      "                              'left-end-position': {'#text': '65',\n",
      "                                                    '@datatype': 'integer'},\n",
      "                              'parent': {'Feature': {'@class': 'true',\n",
      "                                                     '@frameid': 'Mutagenesis-Variants',\n",
      "                                                     '@orgid': 'ECOLI',\n",
      "                                                     '@resource': 'getxml?ECOLI:Mutagenesis-Variants'}},\n",
      "                              'right-end-position': {'#text': '65',\n",
      "                                                     '@datatype': 'integer'}}},\n",
      "                 {'Feature': {'@ID': 'ECOLI:FTR0-1502706',\n",
      "                              '@detail': 'full',\n",
      "                              '@frameid': 'FTR0-1502706',\n",
      "                              '@orgid': 'ECOLI',\n",
      "                              'comment': {'#text': 'K &rarr; I: dimished '\n",
      "                                                   'affinity for CheA but not '\n",
      "                                                   'for Tar; less effective '\n",
      "                                                   'than wild-type CheW in '\n",
      "                                                   'mediating Tar-CheA '\n",
      "                                                   'coupling (as measured by '\n",
      "                                                   'CheA autokinase activity) '\n",
      "                                                   '<i>in vitro</i>',\n",
      "                                          '@datatype': 'string'},\n",
      "                              'evidence': {'Evidence-Code': {'@class': 'true',\n",
      "                                                             '@frameid': 'EV-EXP-IDA',\n",
      "                                                             '@orgid': 'ECOLI',\n",
      "                                                             '@resource': '#ECOLI:EV-EXP-IDA'},\n",
      "                                           'Publication': {'@frameid': 'PUB-11964403',\n",
      "                                                           '@orgid': 'ECOLI',\n",
      "                                                           '@resource': '#ECOLI:PUB-11964403'}},\n",
      "                              'feature-of': {'Protein': {'@frameid': 'CHEW-MONOMER',\n",
      "                                                         '@orgid': 'ECOLI',\n",
      "                                                         '@resource': '#ECOLI:CHEW-MONOMER'}},\n",
      "                              'left-end-position': {'#text': '56',\n",
      "                                                    '@datatype': 'integer'},\n",
      "                              'parent': {'Feature': {'@class': 'true',\n",
      "                                                     '@frameid': 'Mutagenesis-Variants',\n",
      "                                                     '@orgid': 'ECOLI',\n",
      "                                                     '@resource': 'getxml?ECOLI:Mutagenesis-Variants'}},\n",
      "                              'right-end-position': {'#text': '56',\n",
      "                                                     '@datatype': 'integer'}}},\n",
      "                 {'Feature': {'@ID': 'ECOLI:FTR0-1502705',\n",
      "                              '@detail': 'full',\n",
      "                              '@frameid': 'FTR0-1502705',\n",
      "                              '@orgid': 'ECOLI',\n",
      "                              'comment': {'#text': 'T &rarr; A/P: dimished '\n",
      "                                                   'affinity for CheA but not '\n",
      "                                                   'for Tar',\n",
      "                                          '@datatype': 'string'},\n",
      "                              'evidence': {'Evidence-Code': {'@class': 'true',\n",
      "                                                             '@frameid': 'EV-EXP-IDA',\n",
      "                                                             '@orgid': 'ECOLI',\n",
      "                                                             '@resource': '#ECOLI:EV-EXP-IDA'},\n",
      "                                           'Publication': {'@frameid': 'PUB-11964403',\n",
      "                                                           '@orgid': 'ECOLI',\n",
      "                                                           '@resource': '#ECOLI:PUB-11964403'}},\n",
      "                              'feature-of': {'Protein': {'@frameid': 'CHEW-MONOMER',\n",
      "                                                         '@orgid': 'ECOLI',\n",
      "                                                         '@resource': '#ECOLI:CHEW-MONOMER'}},\n",
      "                              'left-end-position': {'#text': '51',\n",
      "                                                    '@datatype': 'integer'},\n",
      "                              'parent': {'Feature': {'@class': 'true',\n",
      "                                                     '@frameid': 'Mutagenesis-Variants',\n",
      "                                                     '@orgid': 'ECOLI',\n",
      "                                                     '@resource': 'getxml?ECOLI:Mutagenesis-Variants'}},\n",
      "                              'right-end-position': {'#text': '51',\n",
      "                                                     '@datatype': 'integer'}}},\n",
      "                 {'Feature': {'@ID': 'ECOLI:FTR0-1502704',\n",
      "                              '@detail': 'full',\n",
      "                              '@frameid': 'FTR0-1502704',\n",
      "                              '@orgid': 'ECOLI',\n",
      "                              'comment': {'#text': 'T &rarr; S/A: dimished '\n",
      "                                                   'affinity for CheA but not '\n",
      "                                                   'for Tar; less effective '\n",
      "                                                   'than wild-type CheW in '\n",
      "                                                   'mediating Tar-CheA '\n",
      "                                                   'coupling (as measured by '\n",
      "                                                   'CheA autokinase activity) '\n",
      "                                                   '<i>in vitro</i>',\n",
      "                                          '@datatype': 'string'},\n",
      "                              'evidence': {'Evidence-Code': {'@class': 'true',\n",
      "                                                             '@frameid': 'EV-EXP-IDA',\n",
      "                                                             '@orgid': 'ECOLI',\n",
      "                                                             '@resource': '#ECOLI:EV-EXP-IDA'},\n",
      "                                           'Publication': {'@frameid': 'PUB-11964403',\n",
      "                                                           '@orgid': 'ECOLI',\n",
      "                                                           '@resource': '#ECOLI:PUB-11964403'}},\n",
      "                              'feature-of': {'Protein': {'@frameid': 'CHEW-MONOMER',\n",
      "                                                         '@orgid': 'ECOLI',\n",
      "                                                         '@resource': '#ECOLI:CHEW-MONOMER'}},\n",
      "                              'left-end-position': {'#text': '46',\n",
      "                                                    '@datatype': 'integer'},\n",
      "                              'parent': {'Feature': {'@class': 'true',\n",
      "                                                     '@frameid': 'Mutagenesis-Variants',\n",
      "                                                     '@orgid': 'ECOLI',\n",
      "                                                     '@resource': 'getxml?ECOLI:Mutagenesis-Variants'}},\n",
      "                              'right-end-position': {'#text': '46',\n",
      "                                                     '@datatype': 'integer'}}},\n",
      "                 {'Feature': {'@ID': 'ECOLI:FTR0-1502703',\n",
      "                              '@detail': 'full',\n",
      "                              '@frameid': 'FTR0-1502703',\n",
      "                              '@orgid': 'ECOLI',\n",
      "                              'comment': {'#text': 'V &rarr; L: dimished '\n",
      "                                                   'affinity for CheA but not '\n",
      "                                                   'for Tar; less effective '\n",
      "                                                   'than wild-type CheW in '\n",
      "                                                   'mediating Tar-CheA '\n",
      "                                                   'coupling (as measured by '\n",
      "                                                   'CheA autokinase activity) '\n",
      "                                                   '<i>in vitro</i>',\n",
      "                                          '@datatype': 'string'},\n",
      "                              'evidence': {'Evidence-Code': {'@class': 'true',\n",
      "                                                             '@frameid': 'EV-EXP-IDA',\n",
      "                                                             '@orgid': 'ECOLI',\n",
      "                                                             '@resource': '#ECOLI:EV-EXP-IDA'},\n",
      "                                           'Publication': {'@frameid': 'PUB-11964403',\n",
      "                                                           '@orgid': 'ECOLI',\n",
      "                                                           '@resource': '#ECOLI:PUB-11964403'}},\n",
      "                              'feature-of': {'Protein': {'@frameid': 'CHEW-MONOMER',\n",
      "                                                         '@orgid': 'ECOLI',\n",
      "                                                         '@resource': '#ECOLI:CHEW-MONOMER'}},\n",
      "                              'left-end-position': {'#text': '45',\n",
      "                                                    '@datatype': 'integer'},\n",
      "                              'parent': {'Feature': {'@class': 'true',\n",
      "                                                     '@frameid': 'Mutagenesis-Variants',\n",
      "                                                     '@orgid': 'ECOLI',\n",
      "                                                     '@resource': 'getxml?ECOLI:Mutagenesis-Variants'}},\n",
      "                              'right-end-position': {'#text': '45',\n",
      "                                                     '@datatype': 'integer'}}},\n",
      "                 {'Feature': {'@ID': 'ECOLI:FTR0-1502702',\n",
      "                              '@detail': 'full',\n",
      "                              '@frameid': 'FTR0-1502702',\n",
      "                              '@orgid': 'ECOLI',\n",
      "                              'comment': {'#text': 'G &rarr; E: dimished '\n",
      "                                                   'support for chemotaxis; '\n",
      "                                                   'unlike the wild-type '\n",
      "                                                   'protein does not inhibit '\n",
      "                                                   'swarming when '\n",
      "                                                   'overexpressed; moderate '\n",
      "                                                   'decreases in affinity for '\n",
      "                                                   'Tar and CheA and defective '\n",
      "                                                   'in the ability to form an '\n",
      "                                                   'activated ternary complex '\n",
      "                                                   '<i>in vitro</i>',\n",
      "                                          '@datatype': 'string'},\n",
      "                              'evidence': {'Evidence-Code': {'@class': 'true',\n",
      "                                                             '@frameid': 'EV-EXP-IDA',\n",
      "                                                             '@orgid': 'ECOLI',\n",
      "                                                             '@resource': '#ECOLI:EV-EXP-IDA'},\n",
      "                                           'Publication': {'@ID': 'ECOLI:PUB-11923283',\n",
      "                                                           '@detail': 'full',\n",
      "                                                           '@frameid': 'PUB-11923283',\n",
      "                                                           '@orgid': 'ECOLI',\n",
      "                                                           'author': [...],\n",
      "                                                           'pubmed-id': {...},\n",
      "                                                           'source': {...},\n",
      "                                                           'title': {...},\n",
      "                                                           'year': {...}}},\n",
      "                              'feature-of': {'Protein': {'@frameid': 'CHEW-MONOMER',\n",
      "                                                         '@orgid': 'ECOLI',\n",
      "                                                         '@resource': '#ECOLI:CHEW-MONOMER'}},\n",
      "                              'left-end-position': {'#text': '133',\n",
      "                                                    '@datatype': 'integer'},\n",
      "                              'parent': {'Feature': {'@class': 'true',\n",
      "                                                     '@frameid': 'Mutagenesis-Variants',\n",
      "                                                     '@orgid': 'ECOLI',\n",
      "                                                     '@resource': 'getxml?ECOLI:Mutagenesis-Variants'}},\n",
      "                              'right-end-position': {'#text': '133',\n",
      "                                                     '@datatype': 'integer'}}},\n",
      "                 {'Feature': {'@ID': 'ECOLI:FTR0-1502701',\n",
      "                              '@detail': 'full',\n",
      "                              '@frameid': 'FTR0-1502701',\n",
      "                              '@orgid': 'ECOLI',\n",
      "                              'comment': {'#text': 'G &rarr; D: diminished '\n",
      "                                                   'support for chemotaxis; '\n",
      "                                                   'unlike the wild type '\n",
      "                                                   'protein does not inhibit '\n",
      "                                                   'swarming when '\n",
      "                                                   'over-expressed; moderate '\n",
      "                                                   'decreases in affinity for '\n",
      "                                                   'Tar and CheA and defective '\n",
      "                                                   'in the ability to form an '\n",
      "                                                   'activated ternary complex '\n",
      "                                                   '<i>in vitro</i>',\n",
      "                                          '@datatype': 'string'},\n",
      "                              'evidence': {'Evidence-Code': {'@class': 'true',\n",
      "                                                             '@frameid': 'EV-EXP-IDA',\n",
      "                                                             '@orgid': 'ECOLI',\n",
      "                                                             '@resource': '#ECOLI:EV-EXP-IDA'},\n",
      "                                           'Publication': {'@frameid': 'PUB-11923283',\n",
      "                                                           '@orgid': 'ECOLI',\n",
      "                                                           '@resource': '#ECOLI:PUB-11923283'}},\n",
      "                              'feature-of': {'Protein': {'@frameid': 'CHEW-MONOMER',\n",
      "                                                         '@orgid': 'ECOLI',\n",
      "                                                         '@resource': '#ECOLI:CHEW-MONOMER'}},\n",
      "                              'left-end-position': {'#text': '41',\n",
      "                                                    '@datatype': 'integer'},\n",
      "                              'parent': {'Feature': {'@class': 'true',\n",
      "                                                     '@frameid': 'Mutagenesis-Variants',\n",
      "                                                     '@orgid': 'ECOLI',\n",
      "                                                     '@resource': 'getxml?ECOLI:Mutagenesis-Variants'}},\n",
      "                              'right-end-position': {'#text': '41',\n",
      "                                                     '@datatype': 'integer'}}},\n",
      "                 {'Feature': {'@ID': 'ECOLI:FTR0-1502700',\n",
      "                              '@detail': 'full',\n",
      "                              '@frameid': 'FTR0-1502700',\n",
      "                              '@orgid': 'ECOLI',\n",
      "                              'comment': {'#text': 'R &rarr; H: loss of '\n",
      "                                                   'function mutation; small '\n",
      "                                                   'decreases in affinity for '\n",
      "                                                   'Tar and CheA and decreased '\n",
      "                                                   'ability to form an '\n",
      "                                                   'activated ternary complex '\n",
      "                                                   '<i>in vitro</i>',\n",
      "                                          '@datatype': 'string'},\n",
      "                              'evidence': {'Evidence-Code': {'@class': 'true',\n",
      "                                                             '@frameid': 'EV-EXP-IDA',\n",
      "                                                             '@orgid': 'ECOLI',\n",
      "                                                             '@resource': '#ECOLI:EV-EXP-IDA'},\n",
      "                                           'Publication': {'@frameid': 'PUB-11923283',\n",
      "                                                           '@orgid': 'ECOLI',\n",
      "                                                           '@resource': '#ECOLI:PUB-11923283'}},\n",
      "                              'feature-of': {'Protein': {'@frameid': 'CHEW-MONOMER',\n",
      "                                                         '@orgid': 'ECOLI',\n",
      "                                                         '@resource': '#ECOLI:CHEW-MONOMER'}},\n",
      "                              'left-end-position': {'#text': '62',\n",
      "                                                    '@datatype': 'integer'},\n",
      "                              'parent': {'Feature': {'@class': 'true',\n",
      "                                                     '@frameid': 'Mutagenesis-Variants',\n",
      "                                                     '@orgid': 'ECOLI',\n",
      "                                                     '@resource': 'getxml?ECOLI:Mutagenesis-Variants'}},\n",
      "                              'right-end-position': {'#text': '62',\n",
      "                                                     '@datatype': 'integer'}}},\n",
      "                 {'Feature': {'@ID': 'ECOLI:FTR0-1502699',\n",
      "                              '@detail': 'full',\n",
      "                              '@frameid': 'FTR0-1502699',\n",
      "                              '@orgid': 'ECOLI',\n",
      "                              'comment': {'#text': 'V &rarr; M: loss of '\n",
      "                                                   'function mutation; unable '\n",
      "                                                   'to bind Tar, diminished '\n",
      "                                                   'affinity for CheA and '\n",
      "                                                   'unable to form an '\n",
      "                                                   'activated ternary complex '\n",
      "                                                   '<i>in vitro</i>',\n",
      "                                          '@datatype': 'string'},\n",
      "                              'evidence': {'Evidence-Code': {'@class': 'true',\n",
      "                                                             '@frameid': 'EV-EXP-IDA',\n",
      "                                                             '@orgid': 'ECOLI',\n",
      "                                                             '@resource': '#ECOLI:EV-EXP-IDA'},\n",
      "                                           'Publication': {'@frameid': 'PUB-11923283',\n",
      "                                                           '@orgid': 'ECOLI',\n",
      "                                                           '@resource': '#ECOLI:PUB-11923283'}},\n",
      "                              'feature-of': {'Protein': {'@frameid': 'CHEW-MONOMER',\n",
      "                                                         '@orgid': 'ECOLI',\n",
      "                                                         '@resource': '#ECOLI:CHEW-MONOMER'}},\n",
      "                              'left-end-position': {'#text': '36',\n",
      "                                                    '@datatype': 'integer'},\n",
      "                              'parent': {'Feature': {'@class': 'true',\n",
      "                                                     '@frameid': 'Mutagenesis-Variants',\n",
      "                                                     '@orgid': 'ECOLI',\n",
      "                                                     '@resource': 'getxml?ECOLI:Mutagenesis-Variants'}},\n",
      "                              'right-end-position': {'#text': '36',\n",
      "                                                     '@datatype': 'integer'}}},\n",
      "                 {'Feature': {'@ID': 'ECOLI:FTR0-1502698',\n",
      "                              '@detail': 'full',\n",
      "                              '@frameid': 'FTR0-1502698',\n",
      "                              '@orgid': 'ECOLI',\n",
      "                              'comment': {'#text': 'G &rarr; D: loss of '\n",
      "                                                   'function mutation; lacks '\n",
      "                                                   'detectable binding to '\n",
      "                                                   'CheA; diminished affinity '\n",
      "                                                   'for Tar and unable to form '\n",
      "                                                   'an activated ternary '\n",
      "                                                   'complex <i>in vitro</i>',\n",
      "                                          '@datatype': 'string'},\n",
      "                              'evidence': {'Evidence-Code': {'@class': 'true',\n",
      "                                                             '@frameid': 'EV-EXP-IDA',\n",
      "                                                             '@orgid': 'ECOLI',\n",
      "                                                             '@resource': '#ECOLI:EV-EXP-IDA'},\n",
      "                                           'Publication': {'@frameid': 'PUB-11923283',\n",
      "                                                           '@orgid': 'ECOLI',\n",
      "                                                           '@resource': '#ECOLI:PUB-11923283'}},\n",
      "                              'feature-of': {'Protein': {'@frameid': 'CHEW-MONOMER',\n",
      "                                                         '@orgid': 'ECOLI',\n",
      "                                                         '@resource': '#ECOLI:CHEW-MONOMER'}},\n",
      "                              'left-end-position': {'#text': '57',\n",
      "                                                    '@datatype': 'integer'},\n",
      "                              'parent': {'Feature': {'@class': 'true',\n",
      "                                                     '@frameid': 'Mutagenesis-Variants',\n",
      "                                                     '@orgid': 'ECOLI',\n",
      "                                                     '@resource': 'getxml?ECOLI:Mutagenesis-Variants'}},\n",
      "                              'right-end-position': {'#text': '57',\n",
      "                                                     '@datatype': 'integer'}}}],\n",
      " 'has-go-term': [{'GO-Term': {'@class': 'true',\n",
      "                              '@frameid': 'GO:0051286',\n",
      "                              '@orgid': 'ECOLI',\n",
      "                              '@resource': 'getxml?ECOLI:GO:0051286'},\n",
      "                  'evidence': {'Evidence-Code': {'@class': 'true',\n",
      "                                                 '@frameid': 'EV-EXP-IDA',\n",
      "                                                 '@orgid': 'ECOLI',\n",
      "                                                 '@resource': '#ECOLI:EV-EXP-IDA'},\n",
      "                               'Publication': {'@ID': 'ECOLI:PUB-19547746',\n",
      "                                               '@detail': 'full',\n",
      "                                               '@frameid': 'PUB-19547746',\n",
      "                                               '@orgid': 'ECOLI',\n",
      "                                               'author': [{...},\n",
      "                                                          {...},\n",
      "                                                          {...},\n",
      "                                                          {...},\n",
      "                                                          {...},\n",
      "                                                          {...},\n",
      "                                                          {...}],\n",
      "                                               'pubmed-id': {'#text': '19547746',\n",
      "                                                             '@datatype': 'string'},\n",
      "                                               'source': {'#text': 'PLoS Biol '\n",
      "                                                                   '7(6);e1000137',\n",
      "                                                          '@datatype': 'string'},\n",
      "                                               'title': {'#text': 'Self-organization '\n",
      "                                                                  'of the '\n",
      "                                                                  'Escherichia '\n",
      "                                                                  'coli '\n",
      "                                                                  'chemotaxis '\n",
      "                                                                  'network '\n",
      "                                                                  'imaged with '\n",
      "                                                                  'super-resolution '\n",
      "                                                                  'light '\n",
      "                                                                  'microscopy.',\n",
      "                                                         '@datatype': 'string'},\n",
      "                                               'year': {'#text': '2009',\n",
      "                                                        '@datatype': 'integer'}},\n",
      "                               'with': {'@datatype': 'string'}}},\n",
      "                 {'GO-Term': {'@class': 'true',\n",
      "                              '@frameid': 'GO:1901875',\n",
      "                              '@orgid': 'ECOLI',\n",
      "                              '@resource': 'getxml?ECOLI:GO:1901875'},\n",
      "                  'evidence': {'Evidence-Code': {'@class': 'true',\n",
      "                                                 '@frameid': 'EV-EXP-IDA',\n",
      "                                                 '@orgid': 'ECOLI',\n",
      "                                                 '@resource': '#ECOLI:EV-EXP-IDA'},\n",
      "                               'Publication': {'@ID': 'ECOLI:PUB-15916610',\n",
      "                                               '@detail': 'full',\n",
      "                                               '@frameid': 'PUB-15916610',\n",
      "                                               '@orgid': 'ECOLI',\n",
      "                                               'author': [{...}, {...}],\n",
      "                                               'pubmed-id': {'#text': '15916610',\n",
      "                                                             '@datatype': 'string'},\n",
      "                                               'source': {'#text': 'Mol '\n",
      "                                                                   'Microbiol '\n",
      "                                                                   '56(6);1617-26',\n",
      "                                                          '@datatype': 'string'},\n",
      "                                               'title': {'#text': 'Adaptational '\n",
      "                                                                  'assistance '\n",
      "                                                                  'in clusters '\n",
      "                                                                  'of '\n",
      "                                                                  'bacterial '\n",
      "                                                                  'chemoreceptors.',\n",
      "                                                         '@datatype': 'string'},\n",
      "                                               'year': {'#text': '2005',\n",
      "                                                        '@datatype': 'integer'}},\n",
      "                               'with': {'@datatype': 'string'}}},\n",
      "                 {'GO-Term': {'@class': 'true',\n",
      "                              '@frameid': 'GO:0019904',\n",
      "                              '@orgid': 'ECOLI',\n",
      "                              '@resource': 'getxml?ECOLI:GO:0019904'},\n",
      "                  'evidence': {'Evidence-Code': {'@ID': 'ECOLI:EV-EXP-IPI',\n",
      "                                                 '@class': 'true',\n",
      "                                                 '@detail': 'full',\n",
      "                                                 '@frameid': 'EV-EXP-IPI',\n",
      "                                                 '@orgid': 'ECOLI',\n",
      "                                                 'comment': {'#text': 'IPI '\n",
      "                                                                      'inferred '\n",
      "                                                                      'from '\n",
      "                                                                      'physical '\n",
      "                                                                      'interaction\\n'\n",
      "                                                                      '\\n'\n",
      "                                                                      '<br>\\n'\n",
      "                                                                      'The '\n",
      "                                                                      'assertion '\n",
      "                                                                      'was '\n",
      "                                                                      'inferred '\n",
      "                                                                      'from a '\n",
      "                                                                      'physical '\n",
      "                                                                      'interaction '\n",
      "                                                                      'such '\n",
      "                                                                      'as\\n'\n",
      "                                                                      '<ul>\\n'\n",
      "                                                                      '    '\n",
      "                                                                      '<li> '\n",
      "                                                                      '2-hybrid '\n",
      "                                                                      'interactions\\n'\n",
      "                                                                      '    '\n",
      "                                                                      '<li> '\n",
      "                                                                      'Co-purification\\n'\n",
      "                                                                      '    '\n",
      "                                                                      '<li> '\n",
      "                                                                      'Co-immunoprecipitation\\n'\n",
      "                                                                      '    '\n",
      "                                                                      '<li> '\n",
      "                                                                      'Ion/protein '\n",
      "                                                                      'binding '\n",
      "                                                                      'experiments\\n'\n",
      "                                                                      '</ul><p '\n",
      "                                                                      'class=ecocomment>\\n'\n",
      "                                                                      'This '\n",
      "                                                                      'code '\n",
      "                                                                      'covers '\n",
      "                                                                      'physical '\n",
      "                                                                      'interactions '\n",
      "                                                                      'between '\n",
      "                                                                      'the '\n",
      "                                                                      'gene '\n",
      "                                                                      'product '\n",
      "                                                                      'of\\n'\n",
      "                                                                      'interest '\n",
      "                                                                      'and '\n",
      "                                                                      'another '\n",
      "                                                                      'molecule '\n",
      "                                                                      '(or '\n",
      "                                                                      'ion, or '\n",
      "                                                                      'complex). '\n",
      "                                                                      'For '\n",
      "                                                                      'functions '\n",
      "                                                                      'such\\n'\n",
      "                                                                      'as '\n",
      "                                                                      'protein '\n",
      "                                                                      'binding '\n",
      "                                                                      'or '\n",
      "                                                                      'nucleic '\n",
      "                                                                      'acid '\n",
      "                                                                      'binding, '\n",
      "                                                                      'a '\n",
      "                                                                      'binding '\n",
      "                                                                      'assay '\n",
      "                                                                      'is\\n'\n",
      "                                                                      'simultaneously '\n",
      "                                                                      'IPI and '\n",
      "                                                                      'IDA; '\n",
      "                                                                      'IDA is '\n",
      "                                                                      'preferred '\n",
      "                                                                      'because '\n",
      "                                                                      'the '\n",
      "                                                                      'assay\\n'\n",
      "                                                                      'directly '\n",
      "                                                                      'detects '\n",
      "                                                                      'the '\n",
      "                                                                      'binding. '\n",
      "                                                                      'For '\n",
      "                                                                      'both '\n",
      "                                                                      'IPI and '\n",
      "                                                                      'IGI, it '\n",
      "                                                                      'would '\n",
      "                                                                      'be '\n",
      "                                                                      'good\\n'\n",
      "                                                                      'practice '\n",
      "                                                                      'to '\n",
      "                                                                      'qualify '\n",
      "                                                                      'them '\n",
      "                                                                      'with '\n",
      "                                                                      'the '\n",
      "                                                                      'gene/protein/ion.',\n",
      "                                                             '@datatype': 'string'},\n",
      "                                                 'common-name': {'#text': 'Inferred '\n",
      "                                                                          'from '\n",
      "                                                                          'physical '\n",
      "                                                                          'interaction',\n",
      "                                                                 '@datatype': 'string'},\n",
      "                                                 'parent': {'Evidence-Code': {'@class': 'true',\n",
      "                                                                              '@frameid': 'EV-EXP',\n",
      "                                                                              '@orgid': 'ECOLI',\n",
      "                                                                              '@resource': 'getxml?ECOLI:EV-EXP'}}},\n",
      "                               'Publication': {'@ID': 'ECOLI:PUB-18931127',\n",
      "                                               '@detail': 'full',\n",
      "                                               '@frameid': 'PUB-18931127',\n",
      "                                               '@orgid': 'ECOLI',\n",
      "                                               'author': [{...}, {...}, {...}],\n",
      "                                               'pubmed-id': {'#text': '18931127',\n",
      "                                                             '@datatype': 'string'},\n",
      "                                               'source': {'#text': 'J '\n",
      "                                                                   'Bacteriol '\n",
      "                                                                   '190(24);8065-74',\n",
      "                                                          '@datatype': 'string'},\n",
      "                                               'title': {'#text': 'Different '\n",
      "                                                                  'signaling '\n",
      "                                                                  'roles of '\n",
      "                                                                  'two '\n",
      "                                                                  'conserved '\n",
      "                                                                  'residues in '\n",
      "                                                                  'the '\n",
      "                                                                  'cytoplasmic '\n",
      "                                                                  'hairpin tip '\n",
      "                                                                  'of Tsr, the '\n",
      "                                                                  'Escherichia '\n",
      "                                                                  'coli serine '\n",
      "                                                                  'chemoreceptor.',\n",
      "                                                         '@datatype': 'string'},\n",
      "                                               'year': {'#text': '2008',\n",
      "                                                        '@datatype': 'integer'}},\n",
      "                               'with': {'#text': 'UniProtKB\\\\:P02942',\n",
      "                                        '@datatype': 'string'}}},\n",
      "                 {'GO-Term': {'@class': 'true',\n",
      "                              '@frameid': 'GO:0005886',\n",
      "                              '@orgid': 'ECOLI',\n",
      "                              '@resource': 'getxml?ECOLI:GO:0005886'},\n",
      "                  'evidence': [{'Evidence-Code': {'@class': 'true',\n",
      "                                                  '@frameid': 'EV-EXP-IDA',\n",
      "                                                  '@orgid': 'ECOLI',\n",
      "                                                  '@resource': '#ECOLI:EV-EXP-IDA'},\n",
      "                                'Publication': {'@frameid': 'PUB-18931127',\n",
      "                                                '@orgid': 'ECOLI',\n",
      "                                                '@resource': '#ECOLI:PUB-18931127'},\n",
      "                                'with': {'@datatype': 'string'}},\n",
      "                               {'Evidence-Code': {'@class': 'true',\n",
      "                                                  '@frameid': 'EV-EXP-IDA',\n",
      "                                                  '@orgid': 'ECOLI',\n",
      "                                                  '@resource': '#ECOLI:EV-EXP-IDA'},\n",
      "                                'Publication': {'@frameid': 'PUB-15916610',\n",
      "                                                '@orgid': 'ECOLI',\n",
      "                                                '@resource': '#ECOLI:PUB-15916610'},\n",
      "                                'with': {'@datatype': 'string'}}]},\n",
      "                 {'GO-Term': {'@class': 'true',\n",
      "                              '@frameid': 'GO:0051649',\n",
      "                              '@orgid': 'ECOLI',\n",
      "                              '@resource': 'getxml?ECOLI:GO:0051649'},\n",
      "                  'evidence': {'Evidence-Code': {'@class': 'true',\n",
      "                                                 '@frameid': 'EV-EXP-IMP',\n",
      "                                                 '@orgid': 'ECOLI',\n",
      "                                                 '@resource': '#ECOLI:EV-EXP-IMP'},\n",
      "                               'Publication': {'@ID': 'ECOLI:PUB-16856941',\n",
      "                                               '@detail': 'full',\n",
      "                                               '@frameid': 'PUB-16856941',\n",
      "                                               '@orgid': 'ECOLI',\n",
      "                                               'author': [{...},\n",
      "                                                          {...},\n",
      "                                                          {...},\n",
      "                                                          {...}],\n",
      "                                               'pubmed-id': {'#text': '16856941',\n",
      "                                                             '@datatype': 'string'},\n",
      "                                               'source': {'#text': 'Mol '\n",
      "                                                                   'Microbiol '\n",
      "                                                                   '61(2);407-17',\n",
      "                                                          '@datatype': 'string'},\n",
      "                                               'title': {'#text': 'Determinants '\n",
      "                                                                  'of '\n",
      "                                                                  'chemoreceptor '\n",
      "                                                                  'cluster '\n",
      "                                                                  'formation '\n",
      "                                                                  'in '\n",
      "                                                                  'Escherichia '\n",
      "                                                                  'coli.',\n",
      "                                                         '@datatype': 'string'},\n",
      "                                               'year': {'#text': '2006',\n",
      "                                                        '@datatype': 'integer'}},\n",
      "                               'with': {'@datatype': 'string'}}},\n",
      "                 {'GO-Term': {'@class': 'true',\n",
      "                              '@frameid': 'GO:0005737',\n",
      "                              '@orgid': 'ECOLI',\n",
      "                              '@resource': 'getxml?ECOLI:GO:0005737'},\n",
      "                  'evidence': [{'Evidence-Code': {'@class': 'true',\n",
      "                                                  '@frameid': 'EV-COMP-AINF',\n",
      "                                                  '@orgid': 'ECOLI',\n",
      "                                                  '@resource': '#ECOLI:EV-COMP-AINF'},\n",
      "                                'Publication': {'@ID': 'ECOLI:PUB-GO_REF-0000043',\n",
      "                                                '@detail': 'full',\n",
      "                                                '@frameid': 'PUB-GO_REF-0000043',\n",
      "                                                '@orgid': 'ECOLI',\n",
      "                                                'author': {...},\n",
      "                                                'title': {...},\n",
      "                                                'url': {...},\n",
      "                                                'year': {...}},\n",
      "                                'with': {'#text': 'SP_KW\\\\:KW-0963',\n",
      "                                         '@datatype': 'string'}},\n",
      "                               {'Evidence-Code': {'@class': 'true',\n",
      "                                                  '@frameid': 'EV-COMP-AINF',\n",
      "                                                  '@orgid': 'ECOLI',\n",
      "                                                  '@resource': '#ECOLI:EV-COMP-AINF'},\n",
      "                                'Publication': {'@ID': 'ECOLI:PUB-GO_REF-0000044',\n",
      "                                                '@detail': 'full',\n",
      "                                                '@frameid': 'PUB-GO_REF-0000044',\n",
      "                                                '@orgid': 'ECOLI',\n",
      "                                                'author': {...},\n",
      "                                                'title': {...},\n",
      "                                                'url': {...},\n",
      "                                                'year': {...}},\n",
      "                                'with': {'#text': 'SP_SL\\\\:SL-0086',\n",
      "                                         '@datatype': 'string'}}]},\n",
      "                 {'GO-Term': {'@class': 'true',\n",
      "                              '@frameid': 'GO:0042803',\n",
      "                              '@orgid': 'ECOLI',\n",
      "                              '@resource': 'getxml?ECOLI:GO:0042803'},\n",
      "                  'evidence': {'Evidence-Code': {'@class': 'true',\n",
      "                                                 '@frameid': 'EV-EXP-IPI',\n",
      "                                                 '@orgid': 'ECOLI',\n",
      "                                                 '@resource': '#ECOLI:EV-EXP-IPI'},\n",
      "                               'Publication': {'@ID': 'ECOLI:PUB-6448631',\n",
      "                                               '@detail': 'full',\n",
      "                                               '@frameid': 'PUB-6448631',\n",
      "                                               '@orgid': 'ECOLI',\n",
      "                                               'author': [{...}, {...}],\n",
      "                                               'pubmed-id': {'#text': '6448631',\n",
      "                                                             '@datatype': 'string'},\n",
      "                                               'source': {'#text': 'Biochemistry '\n",
      "                                                                   '19(20);4633-9',\n",
      "                                                          '@datatype': 'string'},\n",
      "                                               'title': {'#text': 'Chemotaxis '\n",
      "                                                                  'in '\n",
      "                                                                  'Escherichia '\n",
      "                                                                  'coli: '\n",
      "                                                                  'associations '\n",
      "                                                                  'of protein '\n",
      "                                                                  'components.',\n",
      "                                                         '@datatype': 'string'},\n",
      "                                               'year': {'#text': '1980',\n",
      "                                                        '@datatype': 'integer'}},\n",
      "                               'with': {'@datatype': 'string'}}},\n",
      "                 {'GO-Term': {'@class': 'true',\n",
      "                              '@frameid': 'GO:0009454',\n",
      "                              '@orgid': 'ECOLI',\n",
      "                              '@resource': 'getxml?ECOLI:GO:0009454'},\n",
      "                  'evidence': {'Evidence-Code': {'@class': 'true',\n",
      "                                                 '@frameid': 'EV-EXP-IDA',\n",
      "                                                 '@orgid': 'ECOLI',\n",
      "                                                 '@resource': '#ECOLI:EV-EXP-IDA'},\n",
      "                               'Publication': {'@ID': 'ECOLI:PUB-27803157',\n",
      "                                               '@detail': 'full',\n",
      "                                               '@frameid': 'PUB-27803157',\n",
      "                                               '@orgid': 'ECOLI',\n",
      "                                               'author': [{...},\n",
      "                                                          {...},\n",
      "                                                          {...},\n",
      "                                                          {...},\n",
      "                                                          {...}],\n",
      "                                               'pubmed-id': {'#text': '27803157',\n",
      "                                                             '@datatype': 'string'},\n",
      "                                               'source': {'#text': 'J Biol '\n",
      "                                                                   'Chem '\n",
      "                                                                   '291(50);25809-25814',\n",
      "                                                          '@datatype': 'string'},\n",
      "                                               'title': {'#text': 'Bacterial '\n",
      "                                                                  'Energy '\n",
      "                                                                  'Sensor Aer '\n",
      "                                                                  'Modulates '\n",
      "                                                                  'the '\n",
      "                                                                  'Activity of '\n",
      "                                                                  'the '\n",
      "                                                                  'Chemotaxis '\n",
      "                                                                  'Kinase CheA '\n",
      "                                                                  'Based on '\n",
      "                                                                  'the Redox '\n",
      "                                                                  'State of '\n",
      "                                                                  'the Flavin '\n",
      "                                                                  'Cofactor.',\n",
      "                                                         '@datatype': 'string'},\n",
      "                                               'year': {'#text': '2016',\n",
      "                                                        '@datatype': 'integer'}},\n",
      "                               'with': {'@datatype': 'string'}}},\n",
      "                 {'GO-Term': {'@class': 'true',\n",
      "                              '@frameid': 'GO:0098561',\n",
      "                              '@orgid': 'ECOLI',\n",
      "                              '@resource': 'getxml?ECOLI:GO:0098561'},\n",
      "                  'evidence': [{'Evidence-Code': {'@class': 'true',\n",
      "                                                  '@frameid': 'EV-EXP-IDA',\n",
      "                                                  '@orgid': 'ECOLI',\n",
      "                                                  '@resource': '#ECOLI:EV-EXP-IDA'},\n",
      "                                'Publication': {'@ID': 'ECOLI:PUB-21606342',\n",
      "                                                '@detail': 'full',\n",
      "                                                '@frameid': 'PUB-21606342',\n",
      "                                                '@orgid': 'ECOLI',\n",
      "                                                'author': [...],\n",
      "                                                'pubmed-id': {...},\n",
      "                                                'source': {...},\n",
      "                                                'title': {...},\n",
      "                                                'year': {...}},\n",
      "                                'with': {'@datatype': 'string'}},\n",
      "                               {'Evidence-Code': {'@class': 'true',\n",
      "                                                  '@frameid': 'EV-EXP-IDA',\n",
      "                                                  '@orgid': 'ECOLI',\n",
      "                                                  '@resource': '#ECOLI:EV-EXP-IDA'},\n",
      "                                'Publication': {'@frameid': 'PUB-18931127',\n",
      "                                                '@orgid': 'ECOLI',\n",
      "                                                '@resource': '#ECOLI:PUB-18931127'},\n",
      "                                'with': {'@datatype': 'string'}},\n",
      "                               {'Evidence-Code': {'@class': 'true',\n",
      "                                                  '@frameid': 'EV-EXP-IDA',\n",
      "                                                  '@orgid': 'ECOLI',\n",
      "                                                  '@resource': '#ECOLI:EV-EXP-IDA'},\n",
      "                                'Publication': {'@frameid': 'PUB-15916610',\n",
      "                                                '@orgid': 'ECOLI',\n",
      "                                                '@resource': '#ECOLI:PUB-15916610'},\n",
      "                                'with': {'@datatype': 'string'}}]},\n",
      "                 {'GO-Term': {'@class': 'true',\n",
      "                              '@frameid': 'GO:0005829',\n",
      "                              '@orgid': 'ECOLI',\n",
      "                              '@resource': 'getxml?ECOLI:GO:0005829'},\n",
      "                  'evidence': [{'Evidence-Code': {'@ID': 'ECOLI:EV-COMP-IBA',\n",
      "                                                  '@class': 'true',\n",
      "                                                  '@detail': 'full',\n",
      "                                                  '@frameid': 'EV-COMP-IBA',\n",
      "                                                  '@orgid': 'ECOLI',\n",
      "                                                  'comment': {...},\n",
      "                                                  'common-name': {...},\n",
      "                                                  'parent': {...}},\n",
      "                                'Publication': {'@ID': 'ECOLI:PUB-21873635',\n",
      "                                                '@detail': 'full',\n",
      "                                                '@frameid': 'PUB-21873635',\n",
      "                                                '@orgid': 'ECOLI',\n",
      "                                                'author': [...],\n",
      "                                                'pubmed-id': {...},\n",
      "                                                'source': {...},\n",
      "                                                'title': {...},\n",
      "                                                'year': {...}},\n",
      "                                'with': {'#text': 'PANTHER\\\\:PTN000504256|UniProtKB\\\\:P0A964',\n",
      "                                         '@datatype': 'string'}},\n",
      "                               {'Evidence-Code': {'@class': 'true',\n",
      "                                                  '@frameid': 'EV-EXP-IDA-UNPURIFIED-PROTEIN',\n",
      "                                                  '@orgid': 'ECOLI',\n",
      "                                                  '@resource': '#ECOLI:EV-EXP-IDA-UNPURIFIED-PROTEIN'},\n",
      "                                'Publication': {'@ID': 'ECOLI:PUB-334749',\n",
      "                                                '@detail': 'full',\n",
      "                                                '@frameid': 'PUB-334749',\n",
      "                                                '@orgid': 'ECOLI',\n",
      "                                                'author': [...],\n",
      "                                                'pubmed-id': {...},\n",
      "                                                'source': {...},\n",
      "                                                'title': {...},\n",
      "                                                'year': {...}},\n",
      "                                'with': {'@datatype': 'string'}}]},\n",
      "                 {'GO-Term': {'@class': 'true',\n",
      "                              '@frameid': 'GO:0005515',\n",
      "                              '@orgid': 'ECOLI',\n",
      "                              '@resource': 'getxml?ECOLI:GO:0005515'},\n",
      "                  'evidence': [{'Evidence-Code': {'@class': 'true',\n",
      "                                                  '@frameid': 'EV-EXP-IPI',\n",
      "                                                  '@orgid': 'ECOLI',\n",
      "                                                  '@resource': '#ECOLI:EV-EXP-IPI'},\n",
      "                                'Publication': {'@ID': 'ECOLI:PUB-1992467',\n",
      "                                                '@detail': 'full',\n",
      "                                                '@frameid': 'PUB-1992467',\n",
      "                                                '@orgid': 'ECOLI',\n",
      "                                                'author': [...],\n",
      "                                                'pubmed-id': {...},\n",
      "                                                'source': {...},\n",
      "                                                'title': {...},\n",
      "                                                'year': {...}},\n",
      "                                'with': {'#text': 'UniProtKB\\\\:P07363',\n",
      "                                         '@datatype': 'string'}},\n",
      "                               {'Evidence-Code': {'@class': 'true',\n",
      "                                                  '@frameid': 'EV-EXP-IPI',\n",
      "                                                  '@orgid': 'ECOLI',\n",
      "                                                  '@resource': '#ECOLI:EV-EXP-IPI'},\n",
      "                                'Publication': {'@frameid': 'PUB-11923283',\n",
      "                                                '@orgid': 'ECOLI',\n",
      "                                                '@resource': '#ECOLI:PUB-11923283'},\n",
      "                                'with': {'#text': 'UniProtKB\\\\:P07017|UniProtKB\\\\:P07363',\n",
      "                                         '@datatype': 'string'}},\n",
      "                               {'Evidence-Code': {'@class': 'true',\n",
      "                                                  '@frameid': 'EV-EXP-IPI',\n",
      "                                                  '@orgid': 'ECOLI',\n",
      "                                                  '@resource': '#ECOLI:EV-EXP-IPI'},\n",
      "                                'Publication': {'@ID': 'ECOLI:PUB-25060668',\n",
      "                                                '@detail': 'full',\n",
      "                                                '@frameid': 'PUB-25060668',\n",
      "                                                '@orgid': 'ECOLI',\n",
      "                                                'author': [...],\n",
      "                                                'pubmed-id': {...},\n",
      "                                                'source': {...},\n",
      "                                                'title': {...},\n",
      "                                                'year': {...}},\n",
      "                                'with': {'#text': 'UniProtKB\\\\:P02942',\n",
      "                                         '@datatype': 'string'}},\n",
      "                               {'Evidence-Code': {'@class': 'true',\n",
      "                                                  '@frameid': 'EV-EXP-IPI',\n",
      "                                                  '@orgid': 'ECOLI',\n",
      "                                                  '@resource': '#ECOLI:EV-EXP-IPI'},\n",
      "                                'Publication': {'@frameid': 'PUB-6448631',\n",
      "                                                '@orgid': 'ECOLI',\n",
      "                                                '@resource': '#ECOLI:PUB-6448631'},\n",
      "                                'with': {'#text': 'UniProtKB\\\\:P0A9H9',\n",
      "                                         '@datatype': 'string'}},\n",
      "                               {'Evidence-Code': {'@class': 'true',\n",
      "                                                  '@frameid': 'EV-EXP-IPI',\n",
      "                                                  '@orgid': 'ECOLI',\n",
      "                                                  '@resource': '#ECOLI:EV-EXP-IPI'},\n",
      "                                'Publication': {'@ID': 'ECOLI:PUB-16606699',\n",
      "                                                '@detail': 'full',\n",
      "                                                '@frameid': 'PUB-16606699',\n",
      "                                                '@orgid': 'ECOLI',\n",
      "                                                'author': [...],\n",
      "                                                'pubmed-id': {...},\n",
      "                                                'source': {...},\n",
      "                                                'title': {...},\n",
      "                                                'year': {...}},\n",
      "                                'with': {'#text': 'UniProtKB\\\\:P07017',\n",
      "                                         '@datatype': 'string'}},\n",
      "                               {'Evidence-Code': {'@class': 'true',\n",
      "                                                  '@frameid': 'EV-EXP-IPI',\n",
      "                                                  '@orgid': 'ECOLI',\n",
      "                                                  '@resource': '#ECOLI:EV-EXP-IPI'},\n",
      "                                'Publication': {'@ID': 'ECOLI:PUB-19834901',\n",
      "                                                '@detail': 'full',\n",
      "                                                '@frameid': 'PUB-19834901',\n",
      "                                                '@orgid': 'ECOLI',\n",
      "                                                'author': [...],\n",
      "                                                'pubmed-id': {...},\n",
      "                                                'source': {...},\n",
      "                                                'title': {...},\n",
      "                                                'year': {...}},\n",
      "                                'with': {'#text': 'UniProtKB\\\\:P07017',\n",
      "                                         '@datatype': 'string'}},\n",
      "                               {'Evidence-Code': {'@class': 'true',\n",
      "                                                  '@frameid': 'EV-EXP-IPI',\n",
      "                                                  '@orgid': 'ECOLI',\n",
      "                                                  '@resource': '#ECOLI:EV-EXP-IPI'},\n",
      "                                'Publication': {'@frameid': 'PUB-21606342',\n",
      "                                                '@orgid': 'ECOLI',\n",
      "                                                '@resource': '#ECOLI:PUB-21606342'},\n",
      "                                'with': {'#text': 'UniProtKB\\\\:P07017',\n",
      "                                         '@datatype': 'string'}}]},\n",
      "                 {'GO-Term': {'@class': 'true',\n",
      "                              '@frameid': 'GO:0007165',\n",
      "                              '@orgid': 'ECOLI',\n",
      "                              '@resource': 'getxml?ECOLI:GO:0007165'},\n",
      "                  'evidence': [{'Evidence-Code': {'@class': 'true',\n",
      "                                                  '@frameid': 'EV-COMP-AINF',\n",
      "                                                  '@orgid': 'ECOLI',\n",
      "                                                  '@resource': '#ECOLI:EV-COMP-AINF'},\n",
      "                                'Publication': {'@ID': 'ECOLI:PUB-GO_REF-0000002',\n",
      "                                                '@detail': 'full',\n",
      "                                                '@frameid': 'PUB-GO_REF-0000002',\n",
      "                                                '@orgid': 'ECOLI',\n",
      "                                                'author': [...],\n",
      "                                                'title': {...},\n",
      "                                                'url': {...},\n",
      "                                                'year': {...}},\n",
      "                                'with': {'#text': 'InterPro\\\\:IPR002545|InterPro\\\\:IPR036061',\n",
      "                                         '@datatype': 'string'}},\n",
      "                               {'Evidence-Code': {'@class': 'true',\n",
      "                                                  '@frameid': 'EV-EXP-IDA-UNPURIFIED-PROTEIN',\n",
      "                                                  '@orgid': 'ECOLI',\n",
      "                                                  '@resource': '#ECOLI:EV-EXP-IDA-UNPURIFIED-PROTEIN'},\n",
      "                                'Publication': {'@frameid': 'PUB-3553150',\n",
      "                                                '@orgid': 'ECOLI',\n",
      "                                                '@resource': '#ECOLI:PUB-3553150'},\n",
      "                                'with': {'@datatype': 'string'}},\n",
      "                               {'Evidence-Code': {'@class': 'true',\n",
      "                                                  '@frameid': 'EV-EXP-IDA-PURIFIED-PROTEIN',\n",
      "                                                  '@orgid': 'ECOLI',\n",
      "                                                  '@resource': '#ECOLI:EV-EXP-IDA-PURIFIED-PROTEIN'},\n",
      "                                'Publication': {'@frameid': 'PUB-1326408',\n",
      "                                                '@orgid': 'ECOLI',\n",
      "                                                '@resource': '#ECOLI:PUB-1326408'},\n",
      "                                'with': {'@datatype': 'string'}}]},\n",
      "                 {'GO-Term': {'@class': 'true',\n",
      "                              '@frameid': 'GO:0006935',\n",
      "                              '@orgid': 'ECOLI',\n",
      "                              '@resource': 'getxml?ECOLI:GO:0006935'},\n",
      "                  'evidence': [{'Evidence-Code': {'@class': 'true',\n",
      "                                                  '@frameid': 'EV-COMP-AINF',\n",
      "                                                  '@orgid': 'ECOLI',\n",
      "                                                  '@resource': '#ECOLI:EV-COMP-AINF'},\n",
      "                                'Publication': {'@frameid': 'PUB-GO_REF-0000002',\n",
      "                                                '@orgid': 'ECOLI',\n",
      "                                                '@resource': '#ECOLI:PUB-GO_REF-0000002'},\n",
      "                                'with': {'#text': 'InterPro\\\\:IPR002545|InterPro\\\\:IPR036061|InterPro\\\\:IPR039315',\n",
      "                                         '@datatype': 'string'}},\n",
      "                               {'Evidence-Code': {'@class': 'true',\n",
      "                                                  '@frameid': 'EV-COMP-AINF',\n",
      "                                                  '@orgid': 'ECOLI',\n",
      "                                                  '@resource': '#ECOLI:EV-COMP-AINF'},\n",
      "                                'Publication': {'@frameid': 'PUB-GO_REF-0000043',\n",
      "                                                '@orgid': 'ECOLI',\n",
      "                                                '@resource': '#ECOLI:PUB-GO_REF-0000043'},\n",
      "                                'with': {'#text': 'SP_KW\\\\:KW-0145',\n",
      "                                         '@datatype': 'string'}},\n",
      "                               {'Evidence-Code': {'@class': 'true',\n",
      "                                                  '@frameid': 'EV-EXP-IMP',\n",
      "                                                  '@orgid': 'ECOLI',\n",
      "                                                  '@resource': '#ECOLI:EV-EXP-IMP'},\n",
      "                                'Publication': {'@frameid': 'PUB-353036',\n",
      "                                                '@orgid': 'ECOLI',\n",
      "                                                '@resource': '#ECOLI:PUB-353036'},\n",
      "                                'with': {'@datatype': 'string'}},\n",
      "                               {'Evidence-Code': {'@class': 'true',\n",
      "                                                  '@frameid': 'EV-COMP-IBA',\n",
      "                                                  '@orgid': 'ECOLI',\n",
      "                                                  '@resource': '#ECOLI:EV-COMP-IBA'},\n",
      "                                'Publication': {'@frameid': 'PUB-21873635',\n",
      "                                                '@orgid': 'ECOLI',\n",
      "                                                '@resource': '#ECOLI:PUB-21873635'},\n",
      "                                'with': {'#text': 'PANTHER\\\\:PTN000504256|UniProtKB\\\\:P0A964|UniProtKB\\\\:P39802',\n",
      "                                         '@datatype': 'string'}}]}],\n",
      " 'location': [{'cco': {'@frameid': 'CCI-PM-BAC-NEG-GN',\n",
      "                       '@orgid': 'ECOLI',\n",
      "                       '@resource': 'getxml?ECOLI:CCI-PM-BAC-NEG-GN'}},\n",
      "              {'cco': {'@frameid': 'CCI-CYTOSOL-GN',\n",
      "                       '@orgid': 'ECOLI',\n",
      "                       '@resource': 'getxml?ECOLI:CCI-CYTOSOL-GN'}}],\n",
      " 'molecular-weight-seq': {'#text': '18.084',\n",
      "                          '@datatype': 'float',\n",
      "                          '@units': 'kD'},\n",
      " 'parent': {'Protein': {'@class': 'true',\n",
      "                        '@frameid': 'Polypeptides',\n",
      "                        '@orgid': 'ECOLI',\n",
      "                        '@resource': 'getxml?ECOLI:Polypeptides'}},\n",
      " 'pi': {'#text': '4.54', '@datatype': 'float'},\n",
      " 'species': {'Organism': {'@ID': 'ECOLI:ECOLI',\n",
      "                          '@detail': 'full',\n",
      "                          '@frameid': 'ECOLI',\n",
      "                          '@orgid': 'ECOLI',\n",
      "                          'citation': {'Publication': {'@ID': 'ECOLI:PUB-34394059',\n",
      "                                                       '@detail': 'full',\n",
      "                                                       '@frameid': 'PUB-34394059',\n",
      "                                                       '@orgid': 'ECOLI',\n",
      "                                                       'author': [{...},\n",
      "                                                                  {...},\n",
      "                                                                  {...},\n",
      "                                                                  {...},\n",
      "                                                                  {...},\n",
      "                                                                  {...},\n",
      "                                                                  {...},\n",
      "                                                                  {...},\n",
      "                                                                  {...},\n",
      "                                                                  {...},\n",
      "                                                                  {...},\n",
      "                                                                  {...},\n",
      "                                                                  {...},\n",
      "                                                                  {...},\n",
      "                                                                  {...},\n",
      "                                                                  {...},\n",
      "                                                                  {...},\n",
      "                                                                  {...},\n",
      "                                                                  {...}],\n",
      "                                                       'doi-id': {'#text': '10.3389/fmicb.2021.711077',\n",
      "                                                                  '@datatype': 'string'},\n",
      "                                                       'pubmed-id': {'#text': '34394059',\n",
      "                                                                     '@datatype': 'string'},\n",
      "                                                       'source': {'#text': 'Front '\n",
      "                                                                           'Microbiol '\n",
      "                                                                           '12;711077',\n",
      "                                                                  '@datatype': 'string'},\n",
      "                                                       'title': {'#text': 'The '\n",
      "                                                                          'EcoCyc '\n",
      "                                                                          'Database '\n",
      "                                                                          'in '\n",
      "                                                                          '2021.',\n",
      "                                                                 '@datatype': 'string'},\n",
      "                                                       'year': {'#text': '2021',\n",
      "                                                                '@datatype': 'integer'}}},\n",
      "                          'comment': {'#text': 'EcoCyc is a model-organism '\n",
      "                                               'database for <i>Escherichia '\n",
      "                                               'coli</i> K-12 MG1655. For more '\n",
      "                                               'information, see URL '\n",
      "                                               'EcoCyc.org.\\n'\n",
      "                                               '<p>\\n'\n",
      "                                               'The <i>E. coli</i> genome '\n",
      "                                               'sequence used in EcoCyc has '\n",
      "                                               'been updated from version 2 to '\n",
      "                                               'version 3, GenBank accession '\n",
      "                                               'number U00096.3, for release '\n",
      "                                               '20.0 of EcoCyc. Despite the '\n",
      "                                               'involvement of EcoCyc staff in '\n",
      "                                               'ongoing updates to the U00096 '\n",
      "                                               'record, some annotation '\n",
      "                                               'differences may be found '\n",
      "                                               'between U00096 and EcoCyc, '\n",
      "                                               'such as due to recent updates '\n",
      "                                               'to EcoCyc.',\n",
      "                                      '@datatype': 'string'},\n",
      "                          'common-name': {'#text': 'Escherichia coli',\n",
      "                                          '@datatype': 'string'},\n",
      "                          'dblink': [{'dblink-URL': 'http://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?mode=Info&id=511145',\n",
      "                                      'dblink-db': 'NCBI-TAXONOMY-DB',\n",
      "                                      'dblink-oid': '511145',\n",
      "                                      'dblink-relationship': 'unification'},\n",
      "                                     {'dblink-URL': 'http://genomesonline.org/projects?id=Gp0072557',\n",
      "                                      'dblink-db': 'GOLD',\n",
      "                                      'dblink-oid': 'Gp0072557',\n",
      "                                      'dblink-relationship': 'unification'}],\n",
      "                          'genome': {'Genetic-Element': {'@frameid': 'COLI-K12',\n",
      "                                                         '@orgid': 'ECOLI',\n",
      "                                                         '@resource': 'getxml?ECOLI:COLI-K12'}},\n",
      "                          'parent': {'Organism': {'@class': 'true',\n",
      "                                                  '@frameid': 'TAX-511145',\n",
      "                                                  '@orgid': 'ECOLI',\n",
      "                                                  '@resource': 'getxml?ECOLI:TAX-511145'}},\n",
      "                          'pgdb-author': [{'Person': {'@ID': 'ECOLI:pkarp',\n",
      "                                                      '@detail': 'full',\n",
      "                                                      '@frameid': 'pkarp',\n",
      "                                                      '@orgid': 'ECOLI',\n",
      "                                                      'affiliations': {...},\n",
      "                                                      'common-name': {...},\n",
      "                                                      'middle-name': {...}}},\n",
      "                                          {'Person': {'@ID': 'ECOLI:kothari',\n",
      "                                                      '@detail': 'full',\n",
      "                                                      '@frameid': 'kothari',\n",
      "                                                      '@orgid': 'ECOLI',\n",
      "                                                      'affiliations': {...},\n",
      "                                                      'common-name': {...},\n",
      "                                                      'email': {...}}},\n",
      "                                          {'Person': {'@ID': 'ECOLI:paley',\n",
      "                                                      '@detail': 'full',\n",
      "                                                      '@frameid': 'paley',\n",
      "                                                      '@orgid': 'ECOLI',\n",
      "                                                      'affiliations': {...},\n",
      "                                                      'common-name': {...}}},\n",
      "                                          {'Person': {'@ID': 'ECOLI:kr',\n",
      "                                                      '@detail': 'full',\n",
      "                                                      '@frameid': 'kr',\n",
      "                                                      '@orgid': 'ECOLI',\n",
      "                                                      'affiliations': {...},\n",
      "                                                      'common-name': {...},\n",
      "                                                      'email': {...}}},\n",
      "                                          {'Person': {'@ID': 'ECOLI:ipaulsen',\n",
      "                                                      '@detail': 'full',\n",
      "                                                      '@frameid': 'ipaulsen',\n",
      "                                                      '@orgid': 'ECOLI',\n",
      "                                                      'affiliations': {...},\n",
      "                                                      'common-name': {...}}},\n",
      "                                          {'Person': {'@frameid': 'amackie',\n",
      "                                                      '@orgid': 'ECOLI',\n",
      "                                                      '@resource': '#ECOLI:amackie'}},\n",
      "                                          {'Person': {'@ID': 'ECOLI:moore',\n",
      "                                                      '@detail': 'full',\n",
      "                                                      '@frameid': 'moore',\n",
      "                                                      '@orgid': 'ECOLI',\n",
      "                                                      'affiliations': {...},\n",
      "                                                      'email': {...}}},\n",
      "                                          {'Person': {'@ID': 'ECOLI:collado',\n",
      "                                                      '@detail': 'full',\n",
      "                                                      '@frameid': 'collado',\n",
      "                                                      '@orgid': 'ECOLI',\n",
      "                                                      'affiliations': {...},\n",
      "                                                      'email': {...}}},\n",
      "                                          {'Person': {'@ID': 'ECOLI:cesar',\n",
      "                                                      '@detail': 'full',\n",
      "                                                      '@frameid': 'cesar',\n",
      "                                                      '@orgid': 'ECOLI',\n",
      "                                                      'affiliations': {...},\n",
      "                                                      'email': {...}}},\n",
      "                                          {'Person': {'@ID': 'ECOLI:sgama',\n",
      "                                                      '@detail': 'full',\n",
      "                                                      '@frameid': 'sgama',\n",
      "                                                      '@orgid': 'ECOLI',\n",
      "                                                      'affiliations': {...},\n",
      "                                                      'comment': {...},\n",
      "                                                      'common-name': {...}}},\n",
      "                                          {'Person': {'@ID': 'ECOLI:santos-zavaleta',\n",
      "                                                      '@detail': 'full',\n",
      "                                                      '@frameid': 'santos-zavaleta',\n",
      "                                                      '@orgid': 'ECOLI',\n",
      "                                                      'affiliations': {...},\n",
      "                                                      'common-name': {...}}},\n",
      "                                          {'Person': {'@ID': 'ECOLI:tierrafria',\n",
      "                                                      '@detail': 'full',\n",
      "                                                      '@frameid': 'tierrafria',\n",
      "                                                      '@orgid': 'ECOLI',\n",
      "                                                      'affiliations': {...},\n",
      "                                                      'email': {...},\n",
      "                                                      'middle-name': {...}}},\n",
      "                                          {'Person': {'@ID': 'ECOLI:figueroa',\n",
      "                                                      '@detail': 'full',\n",
      "                                                      '@frameid': 'figueroa',\n",
      "                                                      '@orgid': 'ECOLI',\n",
      "                                                      'affiliations': {...},\n",
      "                                                      'email': {...},\n",
      "                                                      'middle-name': {...}}}],\n",
      "                          'pgdb-copyright': {'#text': 'Copyright SRI '\n",
      "                                                      'International '\n",
      "                                                      '1999-2023, Marine '\n",
      "                                                      'Biological Laboratory '\n",
      "                                                      '1998-2001, DoubleTwist '\n",
      "                                                      'Inc 1998-1999.  All '\n",
      "                                                      'Rights Reserved.',\n",
      "                                             '@datatype': 'string'},\n",
      "                          'pgdb-footer-citation': {'#text': '<a '\n",
      "                                                            'href=\"https://www.frontiersin.org/articles/10.3389/fmicb.2021.711077/full\">Frontiers '\n",
      "                                                            'in Microbiology '\n",
      "                                                            '2021</a>',\n",
      "                                                   '@datatype': 'string'},\n",
      "                          'pgdb-name': {'#text': 'EcoCyc',\n",
      "                                        '@datatype': 'string'},\n",
      "                          'pgdb-tier': {'#text': '1', '@datatype': 'integer'},\n",
      "                          'rank': 'strain',\n",
      "                          'strain-name': {'#text': 'K-12 substr. MG1655',\n",
      "                                          '@datatype': 'string'},\n",
      "                          'synonym': {'#text': 'Escherichia coli K-12 substr. '\n",
      "                                               'MG1655',\n",
      "                                      '@datatype': 'string'}}},\n",
      " 'synonym': [{'#text': 'purine-binding chemotaxis protein',\n",
      "              '@datatype': 'string'},\n",
      "             {'#text': 'CheW', '@datatype': 'string'}]}\n"
     ]
    }
   ],
   "source": [
    "# example entry\n",
    "# entity = 'PWY0-1356'\n",
    "entity = 'CHEW-MONOMER'\n",
    "req_str = f'https://websvc.biocyc.org/getxml?id=ECOLI:{entity}&detail=high'\n",
    "\n",
    "r = s.get(req_str)\n",
    "if r.status_code != 200:\n",
    "    print(entity, r.status_code)\n",
    "\n",
    "o = xmltodict.parse(r.content)\n",
    "pp.pprint(o['ptools-xml']['Protein'])"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2023-08-04T23:08:36.995988Z",
     "start_time": "2023-08-04T23:08:36.213299Z"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'@ptools-version': '27.0',\n",
      " '@xml:base': 'http://BioCyc.org/apixml?fn=get-class-all-instances%26id=ECOLI:Fermentation%26detail=NONE',\n",
      " 'Pathway': [{'@frameid': 'PWY-5480',\n",
      "              '@orgid': 'ECOLI',\n",
      "              '@resource': 'getxml?ECOLI:PWY-5480'},\n",
      "             {'@frameid': 'PWY0-1312',\n",
      "              '@orgid': 'ECOLI',\n",
      "              '@resource': 'getxml?ECOLI:PWY0-1312'},\n",
      "             {'@frameid': 'PWY-5485',\n",
      "              '@orgid': 'ECOLI',\n",
      "              '@resource': 'getxml?ECOLI:PWY-5485'},\n",
      "             {'@frameid': 'FERMENTATION-PWY',\n",
      "              '@orgid': 'ECOLI',\n",
      "              '@resource': 'getxml?ECOLI:FERMENTATION-PWY'},\n",
      "             {'@frameid': 'PWY-5437',\n",
      "              '@orgid': 'ECOLI',\n",
      "              '@resource': 'getxml?ECOLI:PWY-5437'},\n",
      "             {'@frameid': 'PWY-8274',\n",
      "              '@orgid': 'ECOLI',\n",
      "              '@resource': 'getxml?ECOLI:PWY-8274'}],\n",
      " 'metadata': {'num_results': '6',\n",
      "              'query': 'fn=get-class-all-instances&id=ECOLI:Fermentation&detail=NONE',\n",
      "              'service_name': 'apixml',\n",
      "              'url': 'http://BioCyc.org/'}}\n"
     ]
    }
   ],
   "source": [
    "entity = 'Fermentation'\n",
    "\n",
    "fn = 'get-class-all-instances'\n",
    "req_str = f'https://websvc.biocyc.org/apixml?fn={fn}&id=ECOLI:{entity}&detail=none'\n",
    "\n",
    "r = s.get(req_str)\n",
    "if r.status_code != 200:\n",
    "    print(entity, r.status_code)\n",
    "\n",
    "o = xmltodict.parse(r.content)\n",
    "pp.pprint(o['ptools-xml'])"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2023-08-04T23:08:37.047885Z",
     "start_time": "2023-08-04T23:08:36.996885Z"
    }
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "# Getting raw data"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Fetch EcoCyc pathway tree"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "outputs": [],
   "source": [
    "def recursive_pathway_tree(current_node, parent_node, node_dict, level):\n",
    "\n",
    "\n",
    "    if current_node in node_dict.keys():\n",
    "\n",
    "        if level < node_dict[current_node]['level']:\n",
    "            node_dict[current_node]['level'] = level\n",
    "\n",
    "        node_dict[current_node]['parents'].append(parent_node)\n",
    "\n",
    "        return\n",
    "\n",
    "    else:\n",
    "\n",
    "        req_str = f'https://websvc.biocyc.org/getxml?id=ECOLI:{current_node}&detail=high'\n",
    "\n",
    "        node_dict[current_node] = {'parents': [], 'children': [], 'level': level}\n",
    "\n",
    "        if parent_node is not None:\n",
    "            node_dict[current_node]['parents'].append(parent_node)\n",
    "\n",
    "        r = s.get(req_str)\n",
    "        if r.status_code != 200:\n",
    "            print(current_node, r.status_code)\n",
    "            return\n",
    "\n",
    "        o = xmltodict.parse(r.content)\n",
    "\n",
    "\n",
    "        subclasses = o['ptools-xml']['Pathway']['subclass'] if 'subclass' in o['ptools-xml']['Pathway'] else []\n",
    "        if type(subclasses) is dict:\n",
    "            subclasses = [subclasses]\n",
    "\n",
    "        instances = o['ptools-xml']['Pathway']['instance'] if 'instance' in o['ptools-xml']['Pathway'] else []\n",
    "        if type(instances) is dict:\n",
    "            instances = [instances]\n",
    "\n",
    "        pathways = subclasses + instances\n",
    "\n",
    "        # print(f'{current_node}: {len(subclasses)}, {len(instances)}')\n",
    "\n",
    "        for pathway in pathways:\n",
    "            pathway_id = pathway['Pathway']['@frameid']\n",
    "\n",
    "            node_dict[current_node]['children'].append(pathway_id)\n",
    "            recursive_pathway_tree(pathway_id, current_node, node_dict, level+1)\n",
    "\n",
    "\n",
    "    return\n",
    "\n",
    "\n",
    "def get_pathway_ith_level_parents(cur_pathway_idx, pathway_matrix, name_list, level_vector, level=2, parent_dict=None):\n",
    "\n",
    "    if parent_dict is None:\n",
    "        parent_dict = {}\n",
    "\n",
    "    cur_pathway_level = level_vector[cur_pathway_idx]\n",
    "\n",
    "    if cur_pathway_level == level:\n",
    "        parent_dict[name_list[cur_pathway_idx]] = cur_pathway_level\n",
    "\n",
    "    parent_slice = pathway_matrix[:, cur_pathway_idx]\n",
    "    parent_idxs = np.where(parent_slice != 0)[0]\n",
    "\n",
    "\n",
    "    for idx in parent_idxs:\n",
    "\n",
    "        _ = get_pathway_ith_level_parents(idx, pathway_matrix, name_list, level_vector, level, parent_dict)\n",
    "\n",
    "    return parent_dict\n",
    "\n"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2023-08-04T23:08:38.471259Z",
     "start_time": "2023-08-04T23:08:38.468449Z"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ARG+POLYAMINE-SYN 404\n"
     ]
    }
   ],
   "source": [
    "entity = 'Pathways'\n",
    "\n",
    "fn = 'get-class-direct-subs'\n",
    "req_str = f'https://websvc.biocyc.org/apixml?fn={fn}&id=ECOLI:{entity}&detail=none'\n",
    "\n",
    "r = s.get(req_str)\n",
    "o = xmltodict.parse(r.content)['ptools-xml']['Pathway']\n",
    "top_level_pathways = [pathway['@frameid'] for pathway in o]\n",
    "\n",
    "pathway_node_dict = {}\n",
    "\n",
    "for top_node in top_level_pathways:\n",
    "\n",
    "    recursive_pathway_tree(top_node, None, pathway_node_dict, level=1)"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2023-08-04T23:10:18.574484Z",
     "start_time": "2023-08-04T23:08:38.987784Z"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "outputs": [
    {
     "data": {
      "text/plain": "                                   index  \\\n0                     Signaling-Pathways   \n1                              PWY0-1559   \n2                              PWY0-1495   \n3                              PWY0-1518   \n4                              PWY0-1490   \n...                                  ...   \n1159                  GLUTATHIONESYN-PWY   \n1160             Butanediol-Biosynthesis   \n1161                       CYCLOPEPTIDES   \n1162  6-HM-Dihydropterin-PP-Biosynthesis   \n1163                            PWY-6147   \n\n                                   parents  \\\n0                                       []   \n1                     [Signaling-Pathways]   \n2                     [Signaling-Pathways]   \n3                     [Signaling-Pathways]   \n4                     [Signaling-Pathways]   \n...                                    ...   \n1159                          [Reductants]   \n1160                  [Other-biosynthesis]   \n1161                  [Other-biosynthesis]   \n1162                  [Other-biosynthesis]   \n1163  [6-HM-Dihydropterin-PP-Biosynthesis]   \n\n                                               children level  \n0     [PWY0-1559, PWY0-1495, PWY0-1518, PWY0-1490, P...     1  \n1                                                    []     2  \n2                                                    []     2  \n3                                                    []     2  \n4                                                    []     2  \n...                                                 ...   ...  \n1159                                                 []     4  \n1160                                                 []     3  \n1161                                                 []     3  \n1162                                         [PWY-6147]     3  \n1163                                                 []     4  \n\n[1164 rows x 4 columns]",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>index</th>\n      <th>parents</th>\n      <th>children</th>\n      <th>level</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>Signaling-Pathways</td>\n      <td>[]</td>\n      <td>[PWY0-1559, PWY0-1495, PWY0-1518, PWY0-1490, P...</td>\n      <td>1</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>PWY0-1559</td>\n      <td>[Signaling-Pathways]</td>\n      <td>[]</td>\n      <td>2</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>PWY0-1495</td>\n      <td>[Signaling-Pathways]</td>\n      <td>[]</td>\n      <td>2</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>PWY0-1518</td>\n      <td>[Signaling-Pathways]</td>\n      <td>[]</td>\n      <td>2</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>PWY0-1490</td>\n      <td>[Signaling-Pathways]</td>\n      <td>[]</td>\n      <td>2</td>\n    </tr>\n    <tr>\n      <th>...</th>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n    </tr>\n    <tr>\n      <th>1159</th>\n      <td>GLUTATHIONESYN-PWY</td>\n      <td>[Reductants]</td>\n      <td>[]</td>\n      <td>4</td>\n    </tr>\n    <tr>\n      <th>1160</th>\n      <td>Butanediol-Biosynthesis</td>\n      <td>[Other-biosynthesis]</td>\n      <td>[]</td>\n      <td>3</td>\n    </tr>\n    <tr>\n      <th>1161</th>\n      <td>CYCLOPEPTIDES</td>\n      <td>[Other-biosynthesis]</td>\n      <td>[]</td>\n      <td>3</td>\n    </tr>\n    <tr>\n      <th>1162</th>\n      <td>6-HM-Dihydropterin-PP-Biosynthesis</td>\n      <td>[Other-biosynthesis]</td>\n      <td>[PWY-6147]</td>\n      <td>3</td>\n    </tr>\n    <tr>\n      <th>1163</th>\n      <td>PWY-6147</td>\n      <td>[6-HM-Dihydropterin-PP-Biosynthesis]</td>\n      <td>[]</td>\n      <td>4</td>\n    </tr>\n  </tbody>\n</table>\n<p>1164 rows  4 columns</p>\n</div>"
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pathway_df = pd.DataFrame(pathway_node_dict).T.reset_index(names='id')\n",
    "pathway_df"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2023-08-04T23:10:18.619718Z",
     "start_time": "2023-08-04T23:10:18.589321Z"
    }
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Fetch protein monomer annotations"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "outputs": [
    {
     "data": {
      "text/plain": "                                            id  \\\n0      1-ACYLGLYCEROL-3-P-ACYLTRANSFER-MONOMER   \n1                                1-PFK-MONOMER   \n2             2-DEHYDROPANTOATE-REDUCT-MONOMER   \n3                 2-ISOPROPYLMALATESYN-MONOMER   \n4     2-OCTAPRENYL-METHOXY-BENZOQ-METH-MONOMER   \n...                                        ...   \n4415                              YTFR-MONOMER   \n4416                              YTFT-MONOMER   \n4417                              ZNUA-MONOMER   \n4418                              ZNUB-MONOMER   \n4419                              ZNUC-MONOMER   \n\n                                            common_name enzyme_reaction  \\\n0     1-acylglycerol-3-phosphate <i>O</i>-acyltransf...               0   \n1                                 1-phosphofructokinase               0   \n2                         2-dehydropantoate 2-reductase               0   \n3                            2-isopropylmalate synthase               0   \n4     bifunctional 2-octaprenyl-6-methoxy-1,4-benzoq...               0   \n...                                                 ...             ...   \n4415  galactofuranose ABC transporter putative ATP b...               0   \n4416  galactofuranose ABC transporter putative membr...               0   \n4417  Zn<sup>2+</sup> ABC transporter periplasmic bi...               0   \n4418   Zn<sup>2+</sup> ABC transporter membrane subunit               0   \n4419  Zn<sup>2+</sup> ABC transporter ATP binding su...               0   \n\n     cofactors metal_features other_features direct_annotations  \n0            0              0              0                  0  \n1            0              0              0                  0  \n2            0              0              0                  0  \n3            0              0              0                  0  \n4            0              0              0                  0  \n...        ...            ...            ...                ...  \n4415         0              0              0                  0  \n4416         0              0              0                  0  \n4417         0              0              0                  0  \n4418         0              0              0                  0  \n4419         0              0              0                  0  \n\n[4420 rows x 7 columns]",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>id</th>\n      <th>common_name</th>\n      <th>enzyme_reaction</th>\n      <th>cofactors</th>\n      <th>metal_features</th>\n      <th>other_features</th>\n      <th>direct_annotations</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>1-ACYLGLYCEROL-3-P-ACYLTRANSFER-MONOMER</td>\n      <td>1-acylglycerol-3-phosphate &lt;i&gt;O&lt;/i&gt;-acyltransf...</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>1-PFK-MONOMER</td>\n      <td>1-phosphofructokinase</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>2-DEHYDROPANTOATE-REDUCT-MONOMER</td>\n      <td>2-dehydropantoate 2-reductase</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>2-ISOPROPYLMALATESYN-MONOMER</td>\n      <td>2-isopropylmalate synthase</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>2-OCTAPRENYL-METHOXY-BENZOQ-METH-MONOMER</td>\n      <td>bifunctional 2-octaprenyl-6-methoxy-1,4-benzoq...</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n    </tr>\n    <tr>\n      <th>...</th>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n    </tr>\n    <tr>\n      <th>4415</th>\n      <td>YTFR-MONOMER</td>\n      <td>galactofuranose ABC transporter putative ATP b...</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n    </tr>\n    <tr>\n      <th>4416</th>\n      <td>YTFT-MONOMER</td>\n      <td>galactofuranose ABC transporter putative membr...</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n    </tr>\n    <tr>\n      <th>4417</th>\n      <td>ZNUA-MONOMER</td>\n      <td>Zn&lt;sup&gt;2+&lt;/sup&gt; ABC transporter periplasmic bi...</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n    </tr>\n    <tr>\n      <th>4418</th>\n      <td>ZNUB-MONOMER</td>\n      <td>Zn&lt;sup&gt;2+&lt;/sup&gt; ABC transporter membrane subunit</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n    </tr>\n    <tr>\n      <th>4419</th>\n      <td>ZNUC-MONOMER</td>\n      <td>Zn&lt;sup&gt;2+&lt;/sup&gt; ABC transporter ATP binding su...</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n    </tr>\n  </tbody>\n</table>\n<p>4420 rows  7 columns</p>\n</div>"
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get a set of all monomers with an associated uniprot id\n",
    "proteins_df = pd.read_csv('reconstruction/ecoli/flat/proteins.tsv', sep='\\t').loc[:, [\"id\", \"common_name\"]]\n",
    "\n",
    "for column in [\"enzyme_reaction\", \"cofactors\", \"metal_features\", \"other_features\", \"direct_annotations\"]:\n",
    "    proteins_df[column] = 0\n",
    "    proteins_df[column] = proteins_df[column].astype(object)\n",
    "\n",
    "proteins_df"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2023-08-04T23:10:18.710809Z",
     "start_time": "2023-08-04T23:10:18.620612Z"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "100\n",
      "200\n",
      "300\n",
      "400\n",
      "500\n",
      "600\n",
      "700\n",
      "800\n",
      "900\n",
      "1000\n",
      "1100\n",
      "EG11708-MONOMER 404\n",
      "1200\n",
      "1300\n",
      "1400\n",
      "1500\n",
      "1600\n",
      "1700\n",
      "found selenocysteine\n",
      "found selenocysteine\n",
      "found selenocysteine\n",
      "1800\n",
      "1900\n",
      "2000\n",
      "2100\n",
      "2200\n",
      "2300\n",
      "2400\n",
      "2500\n",
      "2600\n",
      "2700\n",
      "2800\n",
      "2900\n",
      "3000\n",
      "3100\n",
      "3200\n",
      "3300\n",
      "3400\n",
      "3500\n",
      "3600\n",
      "3700\n",
      "3800\n",
      "3900\n",
      "4000\n",
      "4100\n",
      "4200\n",
      "4300\n",
      "4400\n"
     ]
    }
   ],
   "source": [
    "for i in range(len(proteins_df.index)):\n",
    "    if i % 100 == 0:\n",
    "        print(i)\n",
    "\n",
    "    protein = proteins_df.loc[i, 'id']\n",
    "    proteins_df.at[i, 'other_features'] = set()\n",
    "    proteins_df.at[i, 'metal_features'] = set()\n",
    "    proteins_df.at[i, 'enzyme_reaction'] = set()\n",
    "    proteins_df.at[i, 'cofactors'] = set()\n",
    "    proteins_df.at[i, 'direct_annotations'] = set()\n",
    "\n",
    "    req_str = f'https://websvc.biocyc.org/getxml?id=ECOLI:{protein}&detail=high'\n",
    "\n",
    "\n",
    "\n",
    "    r = s.get(req_str)\n",
    "    if r.status_code != 200:\n",
    "        print(protein, r.status_code)\n",
    "        continue\n",
    "\n",
    "    o = xmltodict.parse(r.content)['ptools-xml']\n",
    "\n",
    "\n",
    "\n",
    "    metal_set = set()\n",
    "    other_feature_set = set()\n",
    "    if 'Protein' in o and 'has-feature' in o['Protein']:\n",
    "        features = o['Protein']['has-feature']\n",
    "\n",
    "        if type(features) is dict:\n",
    "            features = [features]\n",
    "\n",
    "        for feature in features:\n",
    "            if 'parent' not in feature['Feature']:\n",
    "                continue\n",
    "\n",
    "            category = feature['Feature']['parent']['Feature']['@frameid']\n",
    "            if category == 'Metal-Binding-Sites' and 'comment' in feature['Feature']:\n",
    "\n",
    "                # Detect match to any of the allowed metal names and allowed cofactor names and add to list\n",
    "                comment = feature['Feature']['comment']['#text']\n",
    "                metal_set.add(comment)\n",
    "\n",
    "                proteins_df.at[i, 'metal_features'] = list(metal_set)\n",
    "\n",
    "            if category == 'Nucleotide-Phosphate-Binding-Regions' and 'attached-group' in feature['Feature'] and 'Compound' in feature['Feature']['attached-group']:\n",
    "                attached_group = feature['Feature']['attached-group']['Compound']['@frameid']\n",
    "                other_feature_set.add(attached_group)\n",
    "\n",
    "            if category == 'N6-pyridoxal-phosphate-Lys-Modifications':\n",
    "                other_feature_set.add('PYRIDOXAL_PHOSPHATE')\n",
    "\n",
    "            if category == 'Protein-Segments' and 'comment' in feature['Feature'] and 'Thiamine' in feature['Feature']['comment']['#text']:\n",
    "                other_feature_set.add('THIAMINE-PYROPHOSPHATE')\n",
    "\n",
    "            if category == 'Selenocysteine-sites':\n",
    "                print('found selenocysteine')\n",
    "                metal_set.add('L-SELENOCYSTEINE')\n",
    "\n",
    "            proteins_df.at[i, 'metal_features'] = list(metal_set)\n",
    "            proteins_df.at[i, 'other_features'] = list(other_feature_set)\n",
    "\n",
    "        if 'catalyzes' in o['Protein']:\n",
    "            oc = o['Protein']['catalyzes']['Enzymatic-Reaction']\n",
    "\n",
    "            if type(oc) is dict:\n",
    "                oc = [oc]\n",
    "\n",
    "            cofactor_set = set()\n",
    "            enz_rxn_set = set()\n",
    "\n",
    "            for enzrxn in oc:\n",
    "                enz_id = enzrxn['@frameid']\n",
    "\n",
    "                enz_rxn_set.add(enz_id)\n",
    "\n",
    "                enz_req_str = f'https://websvc.biocyc.org/getxml?id=ECOLI:{enz_id}&detail=high'\n",
    "\n",
    "                rz = s.get(enz_req_str)\n",
    "                oe = xmltodict.parse(rz.content)['ptools-xml']['Enzymatic-Reaction']\n",
    "\n",
    "                if \"cofactor\" in oe:\n",
    "                    oe = oe['cofactor']\n",
    "\n",
    "                    if type(oe) is dict:\n",
    "                        oe = [oe]\n",
    "\n",
    "                    for cofactor in oe:\n",
    "                        cof = cofactor['Compound']['@frameid']\n",
    "                        cofactor_set.add(cof)\n",
    "\n",
    "            proteins_df.at[i, 'enzyme_reaction'] = enz_rxn_set\n",
    "            proteins_df.at[i, 'cofactors'] = cofactor_set\n",
    "\n",
    "    # get gene pathway annotation\n",
    "    if 'Protein' in o and 'gene' in o['Protein']:\n",
    "\n",
    "        gene = o['Protein']['gene']['Gene']['@frameid']\n",
    "        gene_req_str = f'https://websvc.biocyc.org/apixml?fn=pathways-of-gene&id=ECOLI:{gene}&detail=none'\n",
    "\n",
    "        rg = s.get(gene_req_str)\n",
    "        og = xmltodict.parse(rg.content)['ptools-xml']\n",
    "\n",
    "        if 'Pathway' in og:\n",
    "            pathways = og['Pathway']\n",
    "\n",
    "            if type(pathways) is dict:\n",
    "                pathways = [pathways]\n",
    "\n",
    "            for pathway in pathways:\n",
    "                pathway_id = pathway['@frameid']\n",
    "                proteins_df.at[i, 'direct_annotations'].add(pathway_id)\n",
    "\n",
    "\n"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2023-08-04T23:34:08.944039Z",
     "start_time": "2023-08-04T23:10:18.663418Z"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "outputs": [
    {
     "data": {
      "text/plain": "                                            id  \\\n0      1-ACYLGLYCEROL-3-P-ACYLTRANSFER-MONOMER   \n1                                1-PFK-MONOMER   \n2             2-DEHYDROPANTOATE-REDUCT-MONOMER   \n3                 2-ISOPROPYLMALATESYN-MONOMER   \n4     2-OCTAPRENYL-METHOXY-BENZOQ-METH-MONOMER   \n...                                        ...   \n4415                              YTFR-MONOMER   \n4416                              YTFT-MONOMER   \n4417                              ZNUA-MONOMER   \n4418                              ZNUB-MONOMER   \n4419                              ZNUC-MONOMER   \n\n                                            common_name  \\\n0     1-acylglycerol-3-phosphate <i>O</i>-acyltransf...   \n1                                 1-phosphofructokinase   \n2                         2-dehydropantoate 2-reductase   \n3                            2-isopropylmalate synthase   \n4     bifunctional 2-octaprenyl-6-methoxy-1,4-benzoq...   \n...                                                 ...   \n4415  galactofuranose ABC transporter putative ATP b...   \n4416  galactofuranose ABC transporter putative membr...   \n4417  Zn<sup>2+</sup> ABC transporter periplasmic bi...   \n4418   Zn<sup>2+</sup> ABC transporter membrane subunit   \n4419  Zn<sup>2+</sup> ABC transporter ATP binding su...   \n\n                                        enzyme_reaction cofactors  \\\n0     {ENZRXN0-7991, ENZRXN0-8629, ENZRXN0-7995, ENZ...        {}   \n1                                                    {}        {}   \n2                     {2-DEHYDROPANTOATE-REDUCT-ENZRXN}        {}   \n3                                        {ENZRXN0-6250}        {}   \n4     {2-OCTAPRENYL-METHOXY-BENZOQ-METH-ENZRXN, ADOM...        {}   \n...                                                 ...       ...   \n4415                                                 {}        {}   \n4416                                                 {}        {}   \n4417                                                 {}        {}   \n4418                                                 {}        {}   \n4419                                                 {}        {}   \n\n                                         metal_features other_features  \\\n0                                                    []             []   \n1                                                    []          [ATP]   \n2                                                    []         [NADP]   \n3                                                    []             []   \n4                                                    []             []   \n...                                                 ...            ...   \n4415                                                 []          [ATP]   \n4416                                                 []             []   \n4417  [zinc-coordinating residues, implicated in a s...             []   \n4418                                                 []             []   \n4419                                                 []          [ATP]   \n\n                  direct_annotations  \n0              {PWY0-1319, PWY-5667}  \n1                        {PWY0-1314}  \n2                        {PANTO-PWY}  \n3                       {LEUSYN-PWY}  \n4     {PWY-6708, MENAQUINONESYN-PWY}  \n...                              ...  \n4415                              {}  \n4416                              {}  \n4417                              {}  \n4418                              {}  \n4419                              {}  \n\n[4420 rows x 7 columns]",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>id</th>\n      <th>common_name</th>\n      <th>enzyme_reaction</th>\n      <th>cofactors</th>\n      <th>metal_features</th>\n      <th>other_features</th>\n      <th>direct_annotations</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>1-ACYLGLYCEROL-3-P-ACYLTRANSFER-MONOMER</td>\n      <td>1-acylglycerol-3-phosphate &lt;i&gt;O&lt;/i&gt;-acyltransf...</td>\n      <td>{ENZRXN0-7991, ENZRXN0-8629, ENZRXN0-7995, ENZ...</td>\n      <td>{}</td>\n      <td>[]</td>\n      <td>[]</td>\n      <td>{PWY0-1319, PWY-5667}</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>1-PFK-MONOMER</td>\n      <td>1-phosphofructokinase</td>\n      <td>{}</td>\n      <td>{}</td>\n      <td>[]</td>\n      <td>[ATP]</td>\n      <td>{PWY0-1314}</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>2-DEHYDROPANTOATE-REDUCT-MONOMER</td>\n      <td>2-dehydropantoate 2-reductase</td>\n      <td>{2-DEHYDROPANTOATE-REDUCT-ENZRXN}</td>\n      <td>{}</td>\n      <td>[]</td>\n      <td>[NADP]</td>\n      <td>{PANTO-PWY}</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>2-ISOPROPYLMALATESYN-MONOMER</td>\n      <td>2-isopropylmalate synthase</td>\n      <td>{ENZRXN0-6250}</td>\n      <td>{}</td>\n      <td>[]</td>\n      <td>[]</td>\n      <td>{LEUSYN-PWY}</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>2-OCTAPRENYL-METHOXY-BENZOQ-METH-MONOMER</td>\n      <td>bifunctional 2-octaprenyl-6-methoxy-1,4-benzoq...</td>\n      <td>{2-OCTAPRENYL-METHOXY-BENZOQ-METH-ENZRXN, ADOM...</td>\n      <td>{}</td>\n      <td>[]</td>\n      <td>[]</td>\n      <td>{PWY-6708, MENAQUINONESYN-PWY}</td>\n    </tr>\n    <tr>\n      <th>...</th>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n    </tr>\n    <tr>\n      <th>4415</th>\n      <td>YTFR-MONOMER</td>\n      <td>galactofuranose ABC transporter putative ATP b...</td>\n      <td>{}</td>\n      <td>{}</td>\n      <td>[]</td>\n      <td>[ATP]</td>\n      <td>{}</td>\n    </tr>\n    <tr>\n      <th>4416</th>\n      <td>YTFT-MONOMER</td>\n      <td>galactofuranose ABC transporter putative membr...</td>\n      <td>{}</td>\n      <td>{}</td>\n      <td>[]</td>\n      <td>[]</td>\n      <td>{}</td>\n    </tr>\n    <tr>\n      <th>4417</th>\n      <td>ZNUA-MONOMER</td>\n      <td>Zn&lt;sup&gt;2+&lt;/sup&gt; ABC transporter periplasmic bi...</td>\n      <td>{}</td>\n      <td>{}</td>\n      <td>[zinc-coordinating residues, implicated in a s...</td>\n      <td>[]</td>\n      <td>{}</td>\n    </tr>\n    <tr>\n      <th>4418</th>\n      <td>ZNUB-MONOMER</td>\n      <td>Zn&lt;sup&gt;2+&lt;/sup&gt; ABC transporter membrane subunit</td>\n      <td>{}</td>\n      <td>{}</td>\n      <td>[]</td>\n      <td>[]</td>\n      <td>{}</td>\n    </tr>\n    <tr>\n      <th>4419</th>\n      <td>ZNUC-MONOMER</td>\n      <td>Zn&lt;sup&gt;2+&lt;/sup&gt; ABC transporter ATP binding su...</td>\n      <td>{}</td>\n      <td>{}</td>\n      <td>[]</td>\n      <td>[ATP]</td>\n      <td>{}</td>\n    </tr>\n  </tbody>\n</table>\n<p>4420 rows  7 columns</p>\n</div>"
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "proteins_df"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2023-08-04T23:34:08.976924Z",
     "start_time": "2023-08-04T23:34:08.963176Z"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "# proteins_df.to_parquet('notebooks/fbagd/data/raw_protein_features.parquet', index=False)"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "start_time": "2023-08-04T19:11:28.365547Z"
    }
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Reload protein feature table"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "outputs": [
    {
     "data": {
      "text/plain": "                                            id  \\\n0      1-ACYLGLYCEROL-3-P-ACYLTRANSFER-MONOMER   \n1                                1-PFK-MONOMER   \n2             2-DEHYDROPANTOATE-REDUCT-MONOMER   \n3                 2-ISOPROPYLMALATESYN-MONOMER   \n4     2-OCTAPRENYL-METHOXY-BENZOQ-METH-MONOMER   \n...                                        ...   \n4415                              YTFR-MONOMER   \n4416                              YTFT-MONOMER   \n4417                              ZNUA-MONOMER   \n4418                              ZNUB-MONOMER   \n4419                              ZNUC-MONOMER   \n\n                                            common_name  \\\n0     1-acylglycerol-3-phosphate <i>O</i>-acyltransf...   \n1                                 1-phosphofructokinase   \n2                         2-dehydropantoate 2-reductase   \n3                            2-isopropylmalate synthase   \n4     bifunctional 2-octaprenyl-6-methoxy-1,4-benzoq...   \n...                                                 ...   \n4415  galactofuranose ABC transporter putative ATP b...   \n4416  galactofuranose ABC transporter putative membr...   \n4417  Zn<sup>2+</sup> ABC transporter periplasmic bi...   \n4418   Zn<sup>2+</sup> ABC transporter membrane subunit   \n4419  Zn<sup>2+</sup> ABC transporter ATP binding su...   \n\n                                         metal_features other_features  \\\n0                                                    []             []   \n1                                                    []          [ATP]   \n2                                                    []         [NADP]   \n3                                                    []             []   \n4                                                    []             []   \n...                                                 ...            ...   \n4415                                                 []          [ATP]   \n4416                                                 []             []   \n4417  [zinc-coordinating residues, implicated in a s...             []   \n4418                                                 []             []   \n4419                                                 []          [ATP]   \n\n                                        enzyme_reaction cofactors  \\\n0     {ENZRXN0-7991, ENZRXN0-8629, ENZRXN0-7995, ENZ...        {}   \n1                                                    {}        {}   \n2                     {2-DEHYDROPANTOATE-REDUCT-ENZRXN}        {}   \n3                                        {ENZRXN0-6250}        {}   \n4     {2-OCTAPRENYL-METHOXY-BENZOQ-METH-ENZRXN, ADOM...        {}   \n...                                                 ...       ...   \n4415                                                 {}        {}   \n4416                                                 {}        {}   \n4417                                                 {}        {}   \n4418                                                 {}        {}   \n4419                                                 {}        {}   \n\n                  direct_annotations  \n0              {PWY0-1319, PWY-5667}  \n1                        {PWY0-1314}  \n2                        {PANTO-PWY}  \n3                       {LEUSYN-PWY}  \n4     {PWY-6708, MENAQUINONESYN-PWY}  \n...                              ...  \n4415                              {}  \n4416                              {}  \n4417                              {}  \n4418                              {}  \n4419                              {}  \n\n[4420 rows x 7 columns]",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>id</th>\n      <th>common_name</th>\n      <th>metal_features</th>\n      <th>other_features</th>\n      <th>enzyme_reaction</th>\n      <th>cofactors</th>\n      <th>direct_annotations</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>1-ACYLGLYCEROL-3-P-ACYLTRANSFER-MONOMER</td>\n      <td>1-acylglycerol-3-phosphate &lt;i&gt;O&lt;/i&gt;-acyltransf...</td>\n      <td>[]</td>\n      <td>[]</td>\n      <td>{ENZRXN0-7991, ENZRXN0-8629, ENZRXN0-7995, ENZ...</td>\n      <td>{}</td>\n      <td>{PWY0-1319, PWY-5667}</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>1-PFK-MONOMER</td>\n      <td>1-phosphofructokinase</td>\n      <td>[]</td>\n      <td>[ATP]</td>\n      <td>{}</td>\n      <td>{}</td>\n      <td>{PWY0-1314}</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>2-DEHYDROPANTOATE-REDUCT-MONOMER</td>\n      <td>2-dehydropantoate 2-reductase</td>\n      <td>[]</td>\n      <td>[NADP]</td>\n      <td>{2-DEHYDROPANTOATE-REDUCT-ENZRXN}</td>\n      <td>{}</td>\n      <td>{PANTO-PWY}</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>2-ISOPROPYLMALATESYN-MONOMER</td>\n      <td>2-isopropylmalate synthase</td>\n      <td>[]</td>\n      <td>[]</td>\n      <td>{ENZRXN0-6250}</td>\n      <td>{}</td>\n      <td>{LEUSYN-PWY}</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>2-OCTAPRENYL-METHOXY-BENZOQ-METH-MONOMER</td>\n      <td>bifunctional 2-octaprenyl-6-methoxy-1,4-benzoq...</td>\n      <td>[]</td>\n      <td>[]</td>\n      <td>{2-OCTAPRENYL-METHOXY-BENZOQ-METH-ENZRXN, ADOM...</td>\n      <td>{}</td>\n      <td>{PWY-6708, MENAQUINONESYN-PWY}</td>\n    </tr>\n    <tr>\n      <th>...</th>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n    </tr>\n    <tr>\n      <th>4415</th>\n      <td>YTFR-MONOMER</td>\n      <td>galactofuranose ABC transporter putative ATP b...</td>\n      <td>[]</td>\n      <td>[ATP]</td>\n      <td>{}</td>\n      <td>{}</td>\n      <td>{}</td>\n    </tr>\n    <tr>\n      <th>4416</th>\n      <td>YTFT-MONOMER</td>\n      <td>galactofuranose ABC transporter putative membr...</td>\n      <td>[]</td>\n      <td>[]</td>\n      <td>{}</td>\n      <td>{}</td>\n      <td>{}</td>\n    </tr>\n    <tr>\n      <th>4417</th>\n      <td>ZNUA-MONOMER</td>\n      <td>Zn&lt;sup&gt;2+&lt;/sup&gt; ABC transporter periplasmic bi...</td>\n      <td>[zinc-coordinating residues, implicated in a s...</td>\n      <td>[]</td>\n      <td>{}</td>\n      <td>{}</td>\n      <td>{}</td>\n    </tr>\n    <tr>\n      <th>4418</th>\n      <td>ZNUB-MONOMER</td>\n      <td>Zn&lt;sup&gt;2+&lt;/sup&gt; ABC transporter membrane subunit</td>\n      <td>[]</td>\n      <td>[]</td>\n      <td>{}</td>\n      <td>{}</td>\n      <td>{}</td>\n    </tr>\n    <tr>\n      <th>4419</th>\n      <td>ZNUC-MONOMER</td>\n      <td>Zn&lt;sup&gt;2+&lt;/sup&gt; ABC transporter ATP binding su...</td>\n      <td>[]</td>\n      <td>[ATP]</td>\n      <td>{}</td>\n      <td>{}</td>\n      <td>{}</td>\n    </tr>\n  </tbody>\n</table>\n<p>4420 rows  7 columns</p>\n</div>"
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# proteins_df = pd.read_parquet('notebooks/fbagd/data/raw_protein_features.parquet')\n",
    "# proteins_df['metal_features'] = proteins_df['metal_features'].apply(ast.literal_eval)\n",
    "# # for rows of proteins with where other_features is set, convert from string to set with literal_eval\n",
    "# proteins_df.loc[proteins_df['other_features'].str.startswith('['), 'other_features'] = \\\n",
    "#     proteins_df.loc[proteins_df['other_features'].str.startswith('['), 'other_features'].apply(ast.literal_eval)\n",
    "\n",
    "filter_protein_df = proteins_df.copy().loc[:, ['id', 'common_name', 'metal_features', 'other_features', 'enzyme_reaction', 'cofactors', 'direct_annotations']]\n",
    "filter_protein_df"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2023-08-04T23:34:09.017412Z",
     "start_time": "2023-08-04T23:34:08.968084Z"
    }
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Annotate complexation with EcoCyc data"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "outputs": [
    {
     "data": {
      "text/plain": "                                    id  \\\n0                                1-PFK   \n1                2OXOGLUTARATEDEH-CPLX   \n2          3-ISOPROPYLMALDEHYDROG-CPLX   \n3              3-ISOPROPYLMALISOM-CPLX   \n4     3-METHYL-2-OXOBUT-OHCH3XFER-CPLX   \n...                                ...   \n1063                        CPLX0-8053   \n1064                        CPLX0-8253   \n1065                          SRP-CPLX   \n1066                     CPLX0-7796APO   \n1067                         ARCB-CPLX   \n\n                                          stoichiometry  \n0                     {'1-PFK': 1, '1-PFK-MONOMER': -2}  \n1     {'2OXOGLUTARATEDEH-CPLX': 1, 'E1O': -1, 'E2O':...  \n2     {'3-ISOPROPYLMALDEHYDROG-CPLX': 1, '3-ISOPROPY...  \n3     {'3-ISOPROPYLMALISOM-CPLX': 1, 'LEUC-MONOMER':...  \n4     {'3-METHYL-2-OXOBUT-OHCH3XFER-CPLX': 1, '3-CH3...  \n...                                                 ...  \n1063           {'CPLX0-8053': 1, 'EG10942-MONOMER': -1}  \n1064  {'CPLX0-8253': 1, 'CSRC-RNA': -1, 'EG11447-MON...  \n1065  {'SRP-CPLX': 1, 'EG10300-MONOMER': -1, 'FFS-RN...  \n1066                {'CPLX0-7796APO': 1, 'PD04032': -2}  \n1067               {'ARCB-CPLX': 1, 'ARCB-MONOMER': -2}  \n\n[1068 rows x 2 columns]",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>id</th>\n      <th>stoichiometry</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>1-PFK</td>\n      <td>{'1-PFK': 1, '1-PFK-MONOMER': -2}</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>2OXOGLUTARATEDEH-CPLX</td>\n      <td>{'2OXOGLUTARATEDEH-CPLX': 1, 'E1O': -1, 'E2O':...</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>3-ISOPROPYLMALDEHYDROG-CPLX</td>\n      <td>{'3-ISOPROPYLMALDEHYDROG-CPLX': 1, '3-ISOPROPY...</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>3-ISOPROPYLMALISOM-CPLX</td>\n      <td>{'3-ISOPROPYLMALISOM-CPLX': 1, 'LEUC-MONOMER':...</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>3-METHYL-2-OXOBUT-OHCH3XFER-CPLX</td>\n      <td>{'3-METHYL-2-OXOBUT-OHCH3XFER-CPLX': 1, '3-CH3...</td>\n    </tr>\n    <tr>\n      <th>...</th>\n      <td>...</td>\n      <td>...</td>\n    </tr>\n    <tr>\n      <th>1063</th>\n      <td>CPLX0-8053</td>\n      <td>{'CPLX0-8053': 1, 'EG10942-MONOMER': -1}</td>\n    </tr>\n    <tr>\n      <th>1064</th>\n      <td>CPLX0-8253</td>\n      <td>{'CPLX0-8253': 1, 'CSRC-RNA': -1, 'EG11447-MON...</td>\n    </tr>\n    <tr>\n      <th>1065</th>\n      <td>SRP-CPLX</td>\n      <td>{'SRP-CPLX': 1, 'EG10300-MONOMER': -1, 'FFS-RN...</td>\n    </tr>\n    <tr>\n      <th>1066</th>\n      <td>CPLX0-7796APO</td>\n      <td>{'CPLX0-7796APO': 1, 'PD04032': -2}</td>\n    </tr>\n    <tr>\n      <th>1067</th>\n      <td>ARCB-CPLX</td>\n      <td>{'ARCB-CPLX': 1, 'ARCB-MONOMER': -2}</td>\n    </tr>\n  </tbody>\n</table>\n<p>1068 rows  2 columns</p>\n</div>"
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "complex_df = pd.read_csv('reconstruction/ecoli/flat/complexation_reactions.tsv', sep='\\t').loc[:, ['id', 'stoichiometry']]\n",
    "\n",
    "\n",
    "removed_complexes = pd.read_csv('reconstruction/ecoli/flat/complexation_reactions_removed.tsv', sep='\\t')\n",
    "\n",
    "# remove rows where id starts with '#'\n",
    "complex_df = complex_df[~complex_df['id'].str.startswith('#')].reset_index(drop=True)\n",
    "\n",
    "# remove rows of complex_df where id matches an id in removed_complexes\n",
    "complex_df = complex_df[~complex_df['id'].isin(removed_complexes['id'])].reset_index(drop=True)\n",
    "complex_df.stoichiometry = complex_df.stoichiometry.astype(object)\n",
    "\n",
    "\n",
    "for i, stoich in enumerate(complex_df.loc[:, 'stoichiometry']):\n",
    "\n",
    "    if type(stoich) is str and stoich[0] == '{':\n",
    "        stoich = stoich.replace('null', '-1')\n",
    "        stoich = ast.literal_eval(stoich)\n",
    "\n",
    "        complex_df.at[i, 'stoichiometry'] = stoich\n",
    "\n",
    "    else:\n",
    "        complex_df.at[i, 'stoichiometry'] = {}\n",
    "\n",
    "\n",
    "# for each row, find dict entry with positive value\n",
    "for i in range(len(complex_df.index)):\n",
    "\n",
    "    stoich = complex_df.loc[i, 'stoichiometry']\n",
    "\n",
    "    for k,v in stoich.items():\n",
    "        if v > 0:\n",
    "            complex_df.at[i, 'id'] = k\n",
    "\n",
    "complex_df"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2023-08-04T23:34:09.126366Z",
     "start_time": "2023-08-04T23:34:08.991112Z"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "100\n",
      "200\n",
      "CPLX0-2423 404\n",
      "300\n",
      "CPLX0-3976 404\n",
      "400\n",
      "500\n",
      "600\n",
      "700\n",
      "800\n",
      "900\n",
      "RECFOR-CPLX 404\n",
      "1000\n",
      "CPLX0-7796APO 404\n"
     ]
    }
   ],
   "source": [
    "complex_df[\"cofactors\"] = 0\n",
    "complex_df[\"cofactors\"] = complex_df[\"cofactors\"].astype(object)\n",
    "\n",
    "complex_df[\"enzyme_reaction\"] = 0\n",
    "complex_df[\"enzyme_reaction\"] = complex_df[\"enzyme_reaction\"].astype(object)\n",
    "\n",
    "for i in range(len(complex_df.index)):\n",
    "\n",
    "    if i % 100 == 0:\n",
    "        print(i)\n",
    "\n",
    "    complex = complex_df.loc[i, 'id']\n",
    "\n",
    "    req_str = f'https://websvc.biocyc.org/getxml?id=ECOLI:{complex}&detail=low'\n",
    "\n",
    "    r = s.get(req_str)\n",
    "    if r.status_code != 200:\n",
    "        print(complex, r.status_code)\n",
    "        complex_df.at[i, 'enzyme_reaction'] = set()\n",
    "        complex_df.at[i, 'cofactors'] = set()\n",
    "        continue\n",
    "\n",
    "    o = xmltodict.parse(r.content)['ptools-xml']['Protein']\n",
    "\n",
    "\n",
    "    # if enzyme\n",
    "    if 'catalyzes' in o:\n",
    "        o = o['catalyzes']['Enzymatic-Reaction']\n",
    "\n",
    "        if type(o) is dict:\n",
    "            o = [o]\n",
    "\n",
    "        cofactor_set = set()\n",
    "        enz_rxn_set = set()\n",
    "\n",
    "        for enzrxn in o:\n",
    "            enz_id = enzrxn['@frameid']\n",
    "\n",
    "            enz_rxn_set.add(enz_id)\n",
    "\n",
    "            enz_req_str = f'https://websvc.biocyc.org/getxml?id=ECOLI:{enz_id}&detail=high'\n",
    "\n",
    "            rz = s.get(enz_req_str)\n",
    "            oe = xmltodict.parse(rz.content)['ptools-xml']['Enzymatic-Reaction']\n",
    "\n",
    "            if \"cofactor\" in oe:\n",
    "                oe = oe['cofactor']\n",
    "\n",
    "                if type(oe) is dict:\n",
    "                    oe = [oe]\n",
    "\n",
    "                for cofactor in oe:\n",
    "                    cof = cofactor['Compound']['@frameid']\n",
    "                    cofactor_set.add(cof)\n",
    "\n",
    "        complex_df.at[i, 'enzyme_reaction'] = enz_rxn_set\n",
    "        complex_df.at[i, 'cofactors'] = cofactor_set\n",
    "\n",
    "    else:\n",
    "        complex_df.at[i, 'enzyme_reaction'] = set()\n",
    "        complex_df.at[i, 'cofactors'] = set()"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2023-08-04T23:37:24.438882Z",
     "start_time": "2023-08-04T23:34:09.072541Z"
    }
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Get cofactor elemental composition and more"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Any+2 404\n"
     ]
    }
   ],
   "source": [
    "# get set of all cofactors\n",
    "all_metal_cofactors = set(ALLOWED_METAL_NAMES.values())\n",
    "all_other_cofactors = ACCEPTED_OTHER_FEATURES\n",
    "\n",
    "filter_cofactor_df = pd.DataFrame(columns=['id', 'elemental_composition'])\n",
    "filter_cofactor_df['id'] = list(all_metal_cofactors | all_other_cofactors)\n",
    "\n",
    "filter_cofactor_df['elemental_composition'] = 0\n",
    "filter_cofactor_df['elemental_composition'] = filter_cofactor_df['elemental_composition'].astype(object)\n",
    "\n",
    "# for each cofactor, get elemental composition\n",
    "for i in range(len(filter_cofactor_df.index)):\n",
    "\n",
    "    compound = filter_cofactor_df.loc[i, 'id']\n",
    "    atom_dict = {}\n",
    "\n",
    "    url_name = compound.replace('+', '%2b')\n",
    "\n",
    "    req_str = f'https://websvc.biocyc.org/getxml?id=ECOLI:{url_name}&detail=full'\n",
    "\n",
    "    r = s.get(req_str)\n",
    "    if r.status_code != 200:\n",
    "        print(compound, r.status_code)\n",
    "        filter_cofactor_df.at[i, 'elemental_composition'] = {}\n",
    "        continue\n",
    "\n",
    "    o = xmltodict.parse(r.content)['ptools-xml']\n",
    "\n",
    "    atoms = o['Compound']['cml']['molecule']['atomArray']['atom']\n",
    "    if type(atoms) is dict:\n",
    "        atoms = [atoms]\n",
    "\n",
    "    for atom in atoms:\n",
    "        element = atom['@elementType']\n",
    "\n",
    "        # either add new element or add to existing element\n",
    "        if element not in atom_dict:\n",
    "            atom_dict[element] = 1\n",
    "        else:\n",
    "            atom_dict[element] += 1\n",
    "\n",
    "    filter_cofactor_df.at[i, 'elemental_composition'] = atom_dict"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2023-08-05T00:05:34.464366Z",
     "start_time": "2023-08-05T00:05:31.413484Z"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "outputs": [
    {
     "data": {
      "text/plain": "                        id                              elemental_composition\n0                    CPD-3                                  {'MO': 1, 'O': 4}\n1                      NA+                                          {'NA': 1}\n2                     MN+2                                          {'MN': 1}\n3                     FE+2                                          {'FE': 1}\n4                      FMN                  {'C': 17, 'N': 4, 'O': 9, 'P': 1}\n5                    CPD-7                                  {'FE': 4, 'S': 4}\n6                     CU+2                                          {'CU': 1}\n7      PYRIDOXAL_PHOSPHATE                   {'C': 8, 'N': 1, 'O': 6, 'P': 1}\n8                     NI+2                                          {'NI': 1}\n9                   3FE-4S                                  {'FE': 3, 'S': 4}\n10                      K+                                           {'K': 1}\n11                     FAD                 {'C': 27, 'N': 9, 'O': 15, 'P': 2}\n12                    CO+2                                          {'CO': 1}\n13                   CPD-6                          {'FE': 2, 'S': 6, 'R': 4}\n14                    ZN+2                                          {'ZN': 1}\n15                  Heme-b                 {'C': 34, 'N': 4, 'O': 4, 'FE': 1}\n16                    CA+2                                          {'CA': 1}\n17  THIAMINE-PYROPHOSPHATE          {'C': 12, 'N': 4, 'O': 7, 'P': 2, 'S': 1}\n18                CPD-8123  {'C': 10, 'N': 5, 'O': 8, 'P': 1, 'S': 2, 'MO'...\n19                    MG+2                                          {'MG': 1}\n20                   Any+2                                                 {}\n21        L-SELENOCYSTEINE                  {'C': 3, 'N': 1, 'O': 2, 'SE': 1}",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>id</th>\n      <th>elemental_composition</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>CPD-3</td>\n      <td>{'MO': 1, 'O': 4}</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>NA+</td>\n      <td>{'NA': 1}</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>MN+2</td>\n      <td>{'MN': 1}</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>FE+2</td>\n      <td>{'FE': 1}</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>FMN</td>\n      <td>{'C': 17, 'N': 4, 'O': 9, 'P': 1}</td>\n    </tr>\n    <tr>\n      <th>5</th>\n      <td>CPD-7</td>\n      <td>{'FE': 4, 'S': 4}</td>\n    </tr>\n    <tr>\n      <th>6</th>\n      <td>CU+2</td>\n      <td>{'CU': 1}</td>\n    </tr>\n    <tr>\n      <th>7</th>\n      <td>PYRIDOXAL_PHOSPHATE</td>\n      <td>{'C': 8, 'N': 1, 'O': 6, 'P': 1}</td>\n    </tr>\n    <tr>\n      <th>8</th>\n      <td>NI+2</td>\n      <td>{'NI': 1}</td>\n    </tr>\n    <tr>\n      <th>9</th>\n      <td>3FE-4S</td>\n      <td>{'FE': 3, 'S': 4}</td>\n    </tr>\n    <tr>\n      <th>10</th>\n      <td>K+</td>\n      <td>{'K': 1}</td>\n    </tr>\n    <tr>\n      <th>11</th>\n      <td>FAD</td>\n      <td>{'C': 27, 'N': 9, 'O': 15, 'P': 2}</td>\n    </tr>\n    <tr>\n      <th>12</th>\n      <td>CO+2</td>\n      <td>{'CO': 1}</td>\n    </tr>\n    <tr>\n      <th>13</th>\n      <td>CPD-6</td>\n      <td>{'FE': 2, 'S': 6, 'R': 4}</td>\n    </tr>\n    <tr>\n      <th>14</th>\n      <td>ZN+2</td>\n      <td>{'ZN': 1}</td>\n    </tr>\n    <tr>\n      <th>15</th>\n      <td>Heme-b</td>\n      <td>{'C': 34, 'N': 4, 'O': 4, 'FE': 1}</td>\n    </tr>\n    <tr>\n      <th>16</th>\n      <td>CA+2</td>\n      <td>{'CA': 1}</td>\n    </tr>\n    <tr>\n      <th>17</th>\n      <td>THIAMINE-PYROPHOSPHATE</td>\n      <td>{'C': 12, 'N': 4, 'O': 7, 'P': 2, 'S': 1}</td>\n    </tr>\n    <tr>\n      <th>18</th>\n      <td>CPD-8123</td>\n      <td>{'C': 10, 'N': 5, 'O': 8, 'P': 1, 'S': 2, 'MO'...</td>\n    </tr>\n    <tr>\n      <th>19</th>\n      <td>MG+2</td>\n      <td>{'MG': 1}</td>\n    </tr>\n    <tr>\n      <th>20</th>\n      <td>Any+2</td>\n      <td>{}</td>\n    </tr>\n    <tr>\n      <th>21</th>\n      <td>L-SELENOCYSTEINE</td>\n      <td>{'C': 3, 'N': 1, 'O': 2, 'SE': 1}</td>\n    </tr>\n  </tbody>\n</table>\n</div>"
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filter_cofactor_df"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2023-08-05T00:05:38.917710Z",
     "start_time": "2023-08-05T00:05:38.908849Z"
    }
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Save tables to avoid having to re-download"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "outputs": [],
   "source": [
    "# save complex_df to csv in a way that preserves dicts\n",
    "complex_df.to_csv('notebooks/cofactors/data/raw_complexes.csv', index=False)\n",
    "proteins_df.to_csv('notebooks/cofactors/data/raw_proteins.csv', index=False)\n",
    "filter_cofactor_df.to_csv('notebooks/cofactors/data/raw_cofactors.csv', index=False)\n",
    "pathway_df.to_csv('notebooks/cofactors/data/raw_pathways.csv', index=False)"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2023-08-05T17:51:08.690066Z",
     "start_time": "2023-08-05T17:51:08.661920Z"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "outputs": [
    {
     "data": {
      "text/plain": "                                    id  \\\n0                                1-PFK   \n1                2OXOGLUTARATEDEH-CPLX   \n2          3-ISOPROPYLMALDEHYDROG-CPLX   \n3              3-ISOPROPYLMALISOM-CPLX   \n4     3-METHYL-2-OXOBUT-OHCH3XFER-CPLX   \n...                                ...   \n1063                        CPLX0-8053   \n1064                        CPLX0-8253   \n1065                          SRP-CPLX   \n1066                     CPLX0-7796APO   \n1067                         ARCB-CPLX   \n\n                                          stoichiometry  \\\n0                     {'1-PFK': 1, '1-PFK-MONOMER': -2}   \n1     {'2OXOGLUTARATEDEH-CPLX': 1, 'E1O': -1, 'E2O':...   \n2     {'3-ISOPROPYLMALDEHYDROG-CPLX': 1, '3-ISOPROPY...   \n3     {'3-ISOPROPYLMALISOM-CPLX': 1, 'LEUC-MONOMER':...   \n4     {'3-METHYL-2-OXOBUT-OHCH3XFER-CPLX': 1, '3-CH3...   \n...                                                 ...   \n1063           {'CPLX0-8053': 1, 'EG10942-MONOMER': -1}   \n1064  {'CPLX0-8253': 1, 'CSRC-RNA': -1, 'EG11447-MON...   \n1065  {'SRP-CPLX': 1, 'EG10300-MONOMER': -1, 'FFS-RN...   \n1066                {'CPLX0-7796APO': 1, 'PD04032': -2}   \n1067               {'ARCB-CPLX': 1, 'ARCB-MONOMER': -2}   \n\n                                             cofactors  \\\n0                                               {MG+2}   \n1     {LIPOIC-ACID, THIAMINE-PYROPHOSPHATE, MG+2, FAD}   \n2                                         {MG+2, MN+2}   \n3                                              {CPD-7}   \n4                                               {MG+2}   \n...                                                ...   \n1063                                                {}   \n1064                                                {}   \n1065                                                {}   \n1066                                                {}   \n1067                                                {}   \n\n                           enzyme_reaction  \n0                  {1PFRUCTPHOSPHN-ENZRXN}  \n1                {2OXOGLUTARATEDEH-ENZRXN}  \n2          {3-ISOPROPYLMALDEHYDROG-ENZRXN}  \n3              {3-ISOPROPYLMALISOM-ENZRXN}  \n4     {3-METHYL-2-OXOBUT-OHCH3XFER-ENZRXN}  \n...                                    ...  \n1063                                    {}  \n1064                                    {}  \n1065                                    {}  \n1066                                    {}  \n1067                        {ENZRXN0-8568}  \n\n[1068 rows x 4 columns]",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>id</th>\n      <th>stoichiometry</th>\n      <th>cofactors</th>\n      <th>enzyme_reaction</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>1-PFK</td>\n      <td>{'1-PFK': 1, '1-PFK-MONOMER': -2}</td>\n      <td>{MG+2}</td>\n      <td>{1PFRUCTPHOSPHN-ENZRXN}</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>2OXOGLUTARATEDEH-CPLX</td>\n      <td>{'2OXOGLUTARATEDEH-CPLX': 1, 'E1O': -1, 'E2O':...</td>\n      <td>{LIPOIC-ACID, THIAMINE-PYROPHOSPHATE, MG+2, FAD}</td>\n      <td>{2OXOGLUTARATEDEH-ENZRXN}</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>3-ISOPROPYLMALDEHYDROG-CPLX</td>\n      <td>{'3-ISOPROPYLMALDEHYDROG-CPLX': 1, '3-ISOPROPY...</td>\n      <td>{MG+2, MN+2}</td>\n      <td>{3-ISOPROPYLMALDEHYDROG-ENZRXN}</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>3-ISOPROPYLMALISOM-CPLX</td>\n      <td>{'3-ISOPROPYLMALISOM-CPLX': 1, 'LEUC-MONOMER':...</td>\n      <td>{CPD-7}</td>\n      <td>{3-ISOPROPYLMALISOM-ENZRXN}</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>3-METHYL-2-OXOBUT-OHCH3XFER-CPLX</td>\n      <td>{'3-METHYL-2-OXOBUT-OHCH3XFER-CPLX': 1, '3-CH3...</td>\n      <td>{MG+2}</td>\n      <td>{3-METHYL-2-OXOBUT-OHCH3XFER-ENZRXN}</td>\n    </tr>\n    <tr>\n      <th>...</th>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n    </tr>\n    <tr>\n      <th>1063</th>\n      <td>CPLX0-8053</td>\n      <td>{'CPLX0-8053': 1, 'EG10942-MONOMER': -1}</td>\n      <td>{}</td>\n      <td>{}</td>\n    </tr>\n    <tr>\n      <th>1064</th>\n      <td>CPLX0-8253</td>\n      <td>{'CPLX0-8253': 1, 'CSRC-RNA': -1, 'EG11447-MON...</td>\n      <td>{}</td>\n      <td>{}</td>\n    </tr>\n    <tr>\n      <th>1065</th>\n      <td>SRP-CPLX</td>\n      <td>{'SRP-CPLX': 1, 'EG10300-MONOMER': -1, 'FFS-RN...</td>\n      <td>{}</td>\n      <td>{}</td>\n    </tr>\n    <tr>\n      <th>1066</th>\n      <td>CPLX0-7796APO</td>\n      <td>{'CPLX0-7796APO': 1, 'PD04032': -2}</td>\n      <td>{}</td>\n      <td>{}</td>\n    </tr>\n    <tr>\n      <th>1067</th>\n      <td>ARCB-CPLX</td>\n      <td>{'ARCB-CPLX': 1, 'ARCB-MONOMER': -2}</td>\n      <td>{}</td>\n      <td>{ENZRXN0-8568}</td>\n    </tr>\n  </tbody>\n</table>\n<p>1068 rows  4 columns</p>\n</div>"
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "complex_df"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2023-08-05T00:08:22.145342Z",
     "start_time": "2023-08-05T00:08:22.138225Z"
    }
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Reload data"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "outputs": [],
   "source": [
    "parsed_complex_df = pd.read_csv('notebooks/cofactors/data/raw_complexes.csv', index_col=False)\n",
    "\n",
    "# read stoichiometry, cofactors and enzyme_reaction as literal sets\n",
    "for column in ['stoichiometry', 'cofactors', 'enzyme_reaction']:\n",
    "    parsed_complex_df[column] = parsed_complex_df[column].apply(ast.literal_eval)\n",
    "\n",
    "parsed_protein_df = pd.read_csv('notebooks/cofactors/data/raw_proteins.csv', index_col=False)\n",
    "\n",
    "for column in ['cofactors', 'enzyme_reaction', 'metal_features', 'other_features', 'direct_annotations']:\n",
    "    parsed_protein_df[column] = parsed_protein_df[column].apply(ast.literal_eval)\n",
    "\n",
    "\n",
    "parsed_cofactor_df = pd.read_csv('notebooks/cofactors/data/raw_cofactors.csv', index_col=False)\n",
    "\n",
    "for column in ['elemental_composition']:\n",
    "    parsed_cofactor_df[column] = parsed_cofactor_df[column].apply(ast.literal_eval)\n",
    "\n",
    "parsed_pathway_df = pd.read_csv('notebooks/cofactors/data/raw_pathways.csv', index_col=False)\n",
    "\n",
    "for column in ['parents', 'children']:\n",
    "    parsed_pathway_df[column] = parsed_pathway_df[column].apply(ast.literal_eval)"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2023-08-05T19:24:44.427399Z",
     "start_time": "2023-08-05T19:24:44.263802Z"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "outputs": [
    {
     "data": {
      "text/plain": "                                   index  \\\n0                     Signaling-Pathways   \n1                              PWY0-1559   \n2                              PWY0-1495   \n3                              PWY0-1518   \n4                              PWY0-1490   \n...                                  ...   \n1159                  GLUTATHIONESYN-PWY   \n1160             Butanediol-Biosynthesis   \n1161                       CYCLOPEPTIDES   \n1162  6-HM-Dihydropterin-PP-Biosynthesis   \n1163                            PWY-6147   \n\n                                   parents  \\\n0                                       []   \n1                     [Signaling-Pathways]   \n2                     [Signaling-Pathways]   \n3                     [Signaling-Pathways]   \n4                     [Signaling-Pathways]   \n...                                    ...   \n1159                          [Reductants]   \n1160                  [Other-biosynthesis]   \n1161                  [Other-biosynthesis]   \n1162                  [Other-biosynthesis]   \n1163  [6-HM-Dihydropterin-PP-Biosynthesis]   \n\n                                               children  level  \n0     [PWY0-1559, PWY0-1495, PWY0-1518, PWY0-1490, P...      1  \n1                                                    []      2  \n2                                                    []      2  \n3                                                    []      2  \n4                                                    []      2  \n...                                                 ...    ...  \n1159                                                 []      4  \n1160                                                 []      3  \n1161                                                 []      3  \n1162                                         [PWY-6147]      3  \n1163                                                 []      4  \n\n[1164 rows x 4 columns]",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>index</th>\n      <th>parents</th>\n      <th>children</th>\n      <th>level</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>Signaling-Pathways</td>\n      <td>[]</td>\n      <td>[PWY0-1559, PWY0-1495, PWY0-1518, PWY0-1490, P...</td>\n      <td>1</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>PWY0-1559</td>\n      <td>[Signaling-Pathways]</td>\n      <td>[]</td>\n      <td>2</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>PWY0-1495</td>\n      <td>[Signaling-Pathways]</td>\n      <td>[]</td>\n      <td>2</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>PWY0-1518</td>\n      <td>[Signaling-Pathways]</td>\n      <td>[]</td>\n      <td>2</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>PWY0-1490</td>\n      <td>[Signaling-Pathways]</td>\n      <td>[]</td>\n      <td>2</td>\n    </tr>\n    <tr>\n      <th>...</th>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n    </tr>\n    <tr>\n      <th>1159</th>\n      <td>GLUTATHIONESYN-PWY</td>\n      <td>[Reductants]</td>\n      <td>[]</td>\n      <td>4</td>\n    </tr>\n    <tr>\n      <th>1160</th>\n      <td>Butanediol-Biosynthesis</td>\n      <td>[Other-biosynthesis]</td>\n      <td>[]</td>\n      <td>3</td>\n    </tr>\n    <tr>\n      <th>1161</th>\n      <td>CYCLOPEPTIDES</td>\n      <td>[Other-biosynthesis]</td>\n      <td>[]</td>\n      <td>3</td>\n    </tr>\n    <tr>\n      <th>1162</th>\n      <td>6-HM-Dihydropterin-PP-Biosynthesis</td>\n      <td>[Other-biosynthesis]</td>\n      <td>[PWY-6147]</td>\n      <td>3</td>\n    </tr>\n    <tr>\n      <th>1163</th>\n      <td>PWY-6147</td>\n      <td>[6-HM-Dihydropterin-PP-Biosynthesis]</td>\n      <td>[]</td>\n      <td>4</td>\n    </tr>\n  </tbody>\n</table>\n<p>1164 rows  4 columns</p>\n</div>"
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "parsed_pathway_df"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2023-08-05T19:24:44.472809Z",
     "start_time": "2023-08-05T19:24:44.452888Z"
    }
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "# Data processing into final tables\n",
    "## Processing of pathway data into matrices"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "outputs": [],
   "source": [
    "name_list = list(pathway_node_dict.keys())\n",
    "\n",
    "pathway_matrix = np.zeros((len(name_list), len(name_list)), dtype=np.int64)\n",
    "\n",
    "level_vector = np.zeros(len(name_list), dtype=np.int64)\n",
    "\n",
    "for i in range(len(name_list)):\n",
    "\n",
    "    cur_pathway = name_list[i]\n",
    "    level_vector[i] = pathway_node_dict[cur_pathway]['level']\n",
    "\n",
    "    pathway_parents = pathway_node_dict[cur_pathway]['parents']\n",
    "    pathway_children = pathway_node_dict[cur_pathway]['children']\n",
    "\n",
    "    for parent in pathway_parents:\n",
    "        j = name_list.index(parent)\n",
    "        pathway_matrix[j, i] = 1\n",
    "\n",
    "    for child in pathway_children:\n",
    "        j = name_list.index(child)\n",
    "        pathway_matrix[i, j] = 1"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "outputs": [],
   "source": [
    "array_name_list = np.array(name_list)"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "outputs": [
    {
     "data": {
      "text/plain": "{'Noncarbon-Nutrients': 2, 'Electron-Transfer': 2, 'Respiration': 2}"
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pathway = 'PWY0-1321'\n",
    "pathway_idx = name_list.index(pathway)\n",
    "\n",
    "get_pathway_ith_level_parents(pathway_idx, pathway_matrix, name_list, level_vector, level=2)"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Create new column for monomer component stoichiometry"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "test_complex_df = pd.read_parquet('notebooks/fbagd/data/raw_complexes.parquet')\n",
    "test_complex_df\n",
    "\n",
    "# convert cofactors and enzyme_reaction to str\n",
    "for col in ['stoichiometry', 'enzyme_reaction', 'cofactors']:\n",
    "    test_complex_df[col] = test_complex_df[col].astype(str)\n",
    "\n",
    "    # interpret as str literal\n",
    "    test_complex_df[col] = test_complex_df[col].apply(lambda x: ast.literal_eval(x))\n",
    "\n",
    "test_complex_df\n",
    "\n",
    "complex_ids = complex_df['id'].tolist()\n",
    "monomer_names = filter_protein_df['id'].tolist()"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "start_time": "2023-08-04T19:11:28.377623Z"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "def recursive_component_tree(current_component_name, complex_table, protein_table,\n",
    "                             current_multiplier=1, component_list=None, parent=None, return_cofactors=False):\n",
    "    \"\"\"\n",
    "    Recursively find all downstream components of a given complex.\n",
    "    \"\"\"\n",
    "\n",
    "    complex_names = complex_table['id'].tolist()\n",
    "    monomer_names = protein_table['id'].tolist()\n",
    "\n",
    "\n",
    "    my_children = {}\n",
    "\n",
    "    if component_list is None:\n",
    "        component_list = []\n",
    "\n",
    "\n",
    "    if current_component_name in complex_names:\n",
    "\n",
    "\n",
    "        cplx_idx = complex_table.index[complex_table['id'] == current_component_name][0]\n",
    "        stoichiometry = complex_table.at[cplx_idx, 'stoichiometry']\n",
    "\n",
    "        direct_children = {k: abs(v) for k, v in stoichiometry.items() if v < 0}\n",
    "\n",
    "        for component_name, coefficient in stoichiometry.items():\n",
    "\n",
    "            if coefficient < 0 and component_name != current_component_name:\n",
    "\n",
    "                child_multiplier = abs(coefficient * current_multiplier)\n",
    "\n",
    "                new_child = recursive_component_tree(component_name, complex_table, protein_table,\n",
    "                                                     child_multiplier, component_list, current_component_name, return_cofactors)\n",
    "\n",
    "                my_children = my_children | new_child\n",
    "\n",
    "\n",
    "            elif coefficient > 0 and component_name == current_component_name:\n",
    "                continue\n",
    "\n",
    "            else:\n",
    "                raise ValueError(f\"key {component_name} and value {coefficient} for complex {component_name} not processed properly.\")\n",
    "\n",
    "        component_list.append({'name': current_component_name,'parent': parent, 'children': direct_children,\n",
    "                               'multiplier': int(current_multiplier), })\n",
    "\n",
    "\n",
    "    elif current_component_name in monomer_names:\n",
    "\n",
    "        # TODO check if enzrxn\n",
    "        if return_cofactors:\n",
    "            protein_idx = protein_table.index[protein_table['id'] == current_component_name][0]\n",
    "\n",
    "            protein_metals = protein_table.at[protein_idx, 'metal_features_processed']\n",
    "            protein_other = protein_table.at[protein_idx, 'other_features_processed']\n",
    "\n",
    "            table_cofactors = protein_metals | protein_other\n",
    "\n",
    "            if len(table_cofactors) > 0:\n",
    "                # TODO Add apo protein to component list\n",
    "                my_children = {}\n",
    "\n",
    "                for cofactor, cofactor_coefficient in table_cofactors.items():\n",
    "                    if table_cofactors[cofactor] !=  None:\n",
    "                        my_children[cofactor] = cofactor_coefficient\n",
    "                        component_list.append({'parent': current_component_name,\n",
    "                                               'name': cofactor,\n",
    "                                               'multiplier': abs(int(current_multiplier * cofactor_coefficient)),\n",
    "                                               'children': None})\n",
    "\n",
    "            component_list.append({'parent': parent, 'name': current_component_name, 'multiplier': current_multiplier, 'children': my_children})\n",
    "\n",
    "        else:\n",
    "            my_children = None\n",
    "            component_list.append({'parent': parent, 'name': current_component_name, 'multiplier': current_multiplier, 'children': None})\n",
    "\n",
    "\n",
    "\n",
    "    else:\n",
    "        print(f\"component {current_component_name} not found in complex or protein tables\")\n",
    "\n",
    "        return {}\n",
    "\n",
    "\n",
    "    if parent is None:\n",
    "        return {current_component_name: my_children}, component_list\n",
    "    else:\n",
    "        return {current_component_name: my_children}\n"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "start_time": "2023-08-04T19:11:28.378728Z"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "complex_tree_structure, nodes = recursive_component_tree('CPLX0-8167', complex_df, filter_protein_df)\n",
    "pp.pprint(nodes)"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "complex_df['monomer_component_stoichiometry'] = 0\n",
    "complex_df['monomer_component_stoichiometry'] = complex_df['monomer_component_stoichiometry'].astype(object)\n",
    "\n",
    "for i in range(len(complex_df.index)):\n",
    "    complex_name = complex_df.loc[i, 'id']\n",
    "    complex_tree_structure, nodes = recursive_component_tree(complex_name, complex_df, filter_protein_df)\n",
    "\n",
    "    monomer_components = {node['name']: node['multiplier'] for node in nodes if node['children'] is None}\n",
    "\n",
    "    complex_df.at[i, 'monomer_component_stoichiometry'] = monomer_components"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "filter_complex_df = complex_df.loc[:, [\"id\", \"monomer_component_stoichiometry\", \"cofactors\"]]\n",
    "filter_complex_df"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "# filter_complex_df.to_parquet('notebooks/fbagd/data/processed_complexes.parquet', index=False)"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "# save as parquet\n",
    "# filter_cofactor_df.to_parquet('notebooks/fbagd/data/processed_cofactors.parquet', index=False)"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "Seems neglible. Let's just use the protein features."
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Process raw EcoCyc annotations into standard EcoCyc names"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "# proteins_df = pd.read_parquet('notebooks/fbagd/data/raw_protein_features.parquet')\n",
    "# proteins_df['metal_features'] = proteins_df['metal_features'].apply(ast.literal_eval)\n",
    "# # for rows of proteins with where other_features is set, convert from string to set with literal_eval\n",
    "# proteins_df.loc[proteins_df['other_features'].str.startswith('['), 'other_features'] = \\\n",
    "#     proteins_df.loc[proteins_df['other_features'].str.startswith('['), 'other_features'].apply(ast.literal_eval)\n",
    "\n",
    "filter_protein_df = proteins_df.copy().loc[:, ['id', 'common_name', 'metal_features', 'other_features', 'enzyme_reaction', 'cofactors', 'direct_annotations']]\n",
    "filter_protein_df"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "\n",
    "# remove all \\ characters from keys in ALLOWED_METAL_NAMES\n",
    "NON_REGEX_METAL = {key.replace('\\\\', ''): value for key, value in ALLOWED_METAL_NAMES.items()}\n",
    "\n",
    "filter_protein_df['metal_features_processed'] = 0\n",
    "filter_protein_df['metal_features_processed'] = filter_protein_df['metal_features_processed'].astype(object)\n",
    "\n",
    "metal_pattern = '|'.join(ALLOWED_METAL_NAMES.keys())\n",
    "metal_regex = re.compile(f'(({metal_pattern})(\\s\\d[\\.,;]|[\\.,;]|\\s\\())')\n",
    "\n",
    "\n",
    "for i in range(len(filter_protein_df.index)):\n",
    "\n",
    "    metal_binding = filter_protein_df.loc[i, 'metal_features']\n",
    "\n",
    "    metal_count_dict = {}\n",
    "    existing_matches = set()\n",
    "\n",
    "    for feature in metal_binding:\n",
    "        matches = metal_regex.search(feature)\n",
    "        if matches:\n",
    "            metal = matches.group(0)[:-1]\n",
    "\n",
    "            # eliminate duplicates\n",
    "            if metal not in existing_matches:\n",
    "\n",
    "                existing_matches.add(metal)\n",
    "\n",
    "                if 'heme' in feature:\n",
    "                    metal = metal.replace('Iron', 'heme')\n",
    "\n",
    "                # check if last char of metal is a number, then crop\n",
    "                if metal[-1].isdigit():\n",
    "                    metal = metal[:-2]\n",
    "\n",
    "                metal = metal.strip()\n",
    "\n",
    "                # replace metal name with allowed metal name\n",
    "                metal = NON_REGEX_METAL[metal]\n",
    "\n",
    "                if metal in metal_count_dict:\n",
    "                    metal_count_dict[metal] += 1\n",
    "                else:\n",
    "                    metal_count_dict[metal] = 1\n",
    "\n",
    "        else:\n",
    "            print(f'No match for {feature} in {filter_protein_df.loc[i, \"id\"]}')\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    filter_protein_df.at[i, 'metal_features_processed'] = metal_count_dict"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "filter_protein_df = filter_protein_df.drop(columns=['metal_features'])\n",
    "filter_protein_df"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "filter_protein_df['other_features_processed'] = 0\n",
    "filter_protein_df['other_features_processed'] = filter_protein_df['other_features_processed'].astype(object)\n",
    "\n",
    "for i in range(len(filter_protein_df.index)):\n",
    "\n",
    "    other_features = filter_protein_df.loc[i, 'other_features']\n",
    "\n",
    "    other_feature_count_dict = {}\n",
    "    existing_matches = set()\n",
    "\n",
    "    for feature in other_features:\n",
    "\n",
    "        # eliminate duplicates\n",
    "        if feature not in existing_matches:\n",
    "\n",
    "            existing_matches.add(feature)\n",
    "\n",
    "            if feature in ACCEPTED_OTHER_FEATURES:\n",
    "                if feature in other_feature_count_dict:\n",
    "                    other_feature_count_dict[feature] += 1\n",
    "                else:\n",
    "                    other_feature_count_dict[feature] = 1\n",
    "\n",
    "    filter_protein_df.at[i, 'other_features_processed'] = other_feature_count_dict"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "filter_protein_df = filter_protein_df.drop(columns=['other_features'])\n",
    "\n",
    "filter_protein_df"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Get monomer composition of complexes"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "test_complex_df = pd.read_parquet('notebooks/fbagd/data/raw_complexes.parquet')\n",
    "test_complex_df\n",
    "\n",
    "# convert cofactors and enzyme_reaction to str\n",
    "for col in ['stoichiometry', 'enzyme_reaction', 'cofactors']:\n",
    "    test_complex_df[col] = test_complex_df[col].astype(str)\n",
    "\n",
    "    # interpret as str literal\n",
    "    test_complex_df[col] = test_complex_df[col].apply(lambda x: ast.literal_eval(x))\n",
    "\n",
    "test_complex_df\n",
    "\n",
    "complex_ids = complex_df['id'].tolist()\n",
    "monomer_names = filter_protein_df['id'].tolist()"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "def recursive_component_tree(current_component_name, complex_table, protein_table,\n",
    "                             current_multiplier=1, component_list=None, parent=None, return_cofactors=False):\n",
    "    \"\"\"\n",
    "    Recursively find all downstream components of a given complex.\n",
    "    \"\"\"\n",
    "\n",
    "    complex_names = complex_table['id'].tolist()\n",
    "    monomer_names = protein_table['id'].tolist()\n",
    "\n",
    "\n",
    "    my_children = {}\n",
    "\n",
    "    if component_list is None:\n",
    "        component_list = []\n",
    "\n",
    "\n",
    "    if current_component_name in complex_names:\n",
    "\n",
    "\n",
    "        cplx_idx = complex_table.index[complex_table['id'] == current_component_name][0]\n",
    "        stoichiometry = complex_table.at[cplx_idx, 'stoichiometry']\n",
    "\n",
    "        direct_children = {k: abs(v) for k, v in stoichiometry.items() if v < 0}\n",
    "\n",
    "        for component_name, coefficient in stoichiometry.items():\n",
    "\n",
    "            if coefficient < 0 and component_name != current_component_name:\n",
    "\n",
    "                child_multiplier = abs(coefficient * current_multiplier)\n",
    "\n",
    "                new_child = recursive_component_tree(component_name, complex_table, protein_table,\n",
    "                                                     child_multiplier, component_list, current_component_name, return_cofactors)\n",
    "\n",
    "                my_children = my_children | new_child\n",
    "\n",
    "\n",
    "            elif coefficient > 0 and component_name == current_component_name:\n",
    "                continue\n",
    "\n",
    "            else:\n",
    "                raise ValueError(f\"key {component_name} and value {coefficient} for complex {component_name} not processed properly.\")\n",
    "\n",
    "        component_list.append({'name': current_component_name,'parent': parent, 'children': direct_children,\n",
    "                               'multiplier': int(current_multiplier), })\n",
    "\n",
    "\n",
    "    elif current_component_name in monomer_names:\n",
    "\n",
    "        # TODO check if enzrxn\n",
    "        if return_cofactors:\n",
    "            protein_idx = protein_table.index[protein_table['id'] == current_component_name][0]\n",
    "\n",
    "            protein_metals = protein_table.at[protein_idx, 'metal_features_processed']\n",
    "            protein_other = protein_table.at[protein_idx, 'other_features_processed']\n",
    "\n",
    "            table_cofactors = protein_metals | protein_other\n",
    "\n",
    "            if len(table_cofactors) > 0:\n",
    "                # TODO Add apo protein to component list\n",
    "                my_children = {}\n",
    "\n",
    "                for cofactor, cofactor_coefficient in table_cofactors.items():\n",
    "                    if table_cofactors[cofactor] !=  None:\n",
    "                        my_children[cofactor] = cofactor_coefficient\n",
    "                        component_list.append({'parent': current_component_name,\n",
    "                                               'name': cofactor,\n",
    "                                               'multiplier': abs(int(current_multiplier * cofactor_coefficient)),\n",
    "                                               'children': None})\n",
    "\n",
    "            component_list.append({'parent': parent, 'name': current_component_name, 'multiplier': current_multiplier, 'children': my_children})\n",
    "\n",
    "        else:\n",
    "            my_children = None\n",
    "            component_list.append({'parent': parent, 'name': current_component_name, 'multiplier': current_multiplier, 'children': None})\n",
    "\n",
    "\n",
    "\n",
    "    else:\n",
    "        print(f\"component {current_component_name} not found in complex or protein tables\")\n",
    "\n",
    "        return {}\n",
    "\n",
    "\n",
    "    if parent is None:\n",
    "        return {current_component_name: my_children}, component_list\n",
    "    else:\n",
    "        return {current_component_name: my_children}\n"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "complex_tree_structure, nodes = recursive_component_tree('CPLX0-8167', complex_df, filter_protein_df)\n",
    "pp.pprint(nodes)"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "complex_df['monomer_component_stoichiometry'] = 0\n",
    "complex_df['monomer_component_stoichiometry'] = complex_df['monomer_component_stoichiometry'].astype(object)\n",
    "\n",
    "for i in range(len(complex_df.index)):\n",
    "    complex_name = complex_df.loc[i, 'id']\n",
    "    complex_tree_structure, nodes = recursive_component_tree(complex_name, complex_df, filter_protein_df)\n",
    "\n",
    "    monomer_components = {node['name']: node['multiplier'] for node in nodes if node['children'] is None}\n",
    "\n",
    "    complex_df.at[i, 'monomer_component_stoichiometry'] = monomer_components"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "filter_complex_df = complex_df.loc[:, [\"id\", \"monomer_component_stoichiometry\", \"cofactors\"]]\n",
    "filter_complex_df"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Create tree matrix (also for Julia)"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "# save names\n",
    "complex_ids = list(filter_complex_df['id'])\n",
    "protein_ids = list(filter_protein_df['id'])\n",
    "cofactor_ids = list(filter_cofactor_df['id'])\n",
    "\n",
    "\n",
    "name_idx = complex_ids + protein_ids + cofactor_ids\n",
    "tree_matrix = np.zeros([len(complex_ids) + len(protein_ids) + len(cofactor_ids), len(complex_ids) + len(protein_ids) + len(cofactor_ids)], dtype=np.int64)\n",
    "\n",
    "for i in range(len(filter_complex_df)):\n",
    "    name = filter_complex_df.at[i, 'id']\n",
    "    tree_structure, nodes = recursive_component_tree(name, complex_df, filter_protein_df, return_cofactors=True)\n",
    "\n",
    "    for node in nodes:\n",
    "        node_name = node['name']\n",
    "        node_children = node['children']\n",
    "\n",
    "        if node_children != None:\n",
    "            for child_name, child_coefficient in node_children.items():\n",
    "                if child_name in name_idx:\n",
    "                        tree_matrix[name_idx.index(node_name), name_idx.index(child_name)] = child_coefficient"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "# Create matrices to get cofactor counts"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "filter_protein_df"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "# create protein name to index mapping\n",
    "protein_name_to_index = {}\n",
    "for i in range(len(filter_protein_df.index)):\n",
    "    protein_name_to_index[filter_protein_df.at[i, 'id']] = i\n",
    "\n",
    "# C matrix: complexes x proteins\n",
    "C = np.zeros((len(filter_complex_df.index), len(filter_protein_df.index)))\n",
    "\n",
    "for i in range(len(filter_complex_df.index)):\n",
    "\n",
    "    complex_components = filter_complex_df.loc[i, 'monomer_component_stoichiometry']\n",
    "\n",
    "    # TODO consider cofactors\n",
    "    # complex_cofactors = filter_complex_df.loc[i, 'cofactors']\n",
    "\n",
    "\n",
    "    for component_name, component_count in complex_components.items():\n",
    "        if component_count is not None:             # side effect of parquet\n",
    "            # get index of component in filter_protein_df\n",
    "            component_index = protein_name_to_index[component_name]\n",
    "\n",
    "            if filter_complex_df.at[i, 'id'] == 'APORNAP-CPLX':\n",
    "                print(f'component_name: {component_name}, component_count: {component_count}, component_index: {component_index}')\n",
    "\n",
    "            C[i, component_index] = component_count\n",
    "\n",
    "# append an identity matrix to C\n",
    "C = np.concatenate((C, np.identity(len(filter_protein_df.index))), axis=0)\n",
    "\n",
    "C_names = list(filter_complex_df['id']) + list(filter_protein_df['id'])"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "filter_complex_df.at[99, 'id']"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "# create cofactor name to index mapping\n",
    "cofactor_name_to_index = {}\n",
    "for i in range(len(filter_cofactor_df.index)):\n",
    "    cofactor_name_to_index[filter_cofactor_df.at[i, 'id']] = i\n",
    "\n",
    "cofactor_ids = list(filter_cofactor_df['id'])\n",
    "\n",
    "# P matrix: proteins x cofactors\n",
    "P = np.zeros((len(filter_protein_df.index), len(filter_cofactor_df.index)))\n",
    "\n",
    "for i in range(len(filter_protein_df.index)):\n",
    "    protein_metals = filter_protein_df.loc[i, 'metal_features_processed']\n",
    "    protein_other = filter_protein_df.loc[i, 'other_features_processed']\n",
    "\n",
    "    for metal, count in protein_metals.items():\n",
    "        if count is not None:             # side effect of parquet\n",
    "            cofactor_index = cofactor_name_to_index[metal]\n",
    "            P[i, cofactor_index] = count\n",
    "\n",
    "    for other, count in protein_other.items():\n",
    "        if count is not None:             # side effect of parquet\n",
    "            cofactor_index = cofactor_name_to_index[other]\n",
    "            P[i, cofactor_index] = count\n"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "P"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "# create list of unique elements\n",
    "unique_elements = set()\n",
    "for i in range(len(filter_cofactor_df.index)):\n",
    "    cofactor = filter_cofactor_df.at[i, 'elemental_composition']\n",
    "    unique_elements.update(cofactor.keys())\n",
    "\n",
    "unique_elements = list(unique_elements)\n",
    "\n",
    "# create E matrix: cofactors x elements\n",
    "E = np.zeros((len(filter_cofactor_df.index), len(unique_elements)))\n",
    "\n",
    "for i in range(len(filter_cofactor_df.index)):\n",
    "    cofactor = filter_cofactor_df.at[i, 'elemental_composition']\n",
    "\n",
    "    for element, count in cofactor.items():\n",
    "        if count is not None:             # side effect of parquet\n",
    "            element_index = unique_elements.index(element)\n",
    "            E[i, element_index] = count\n",
    "\n",
    "\n",
    "element_ids = unique_elements"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "unique_elements"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "C_to_E = C @ P @ E"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "C_names[1577]"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "# Now ... add the counts >:o"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "time = '50'\n",
    "date = '2023-06-09'\n",
    "experiment = 'fba-redux'\n",
    "entry = f'{experiment}_{time}_{date}'\n",
    "folder = f'out/fbagd/{entry}/'"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "output = np.load(folder + 'output.npy',allow_pickle='TRUE').item()\n",
    "# output = np.load(r\"out/geneRxnVerifData/output_glc.npy\", allow_pickle=True, encoding='ASCII').tolist()\n",
    "output = output['agents']['0']\n",
    "fba = output['listeners']['fba_results']\n",
    "mass = output['listeners']['mass']\n",
    "bulk = pd.DataFrame(output['bulk'])"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "bulk"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "f = open(folder + 'agent_steps.pkl', 'rb')\n",
    "agent = dill.load(f)\n",
    "f.close()\n",
    "\n",
    "metabolism = agent['ecoli-metabolism-redux']\n",
    "stoichiometry = metabolism.stoichiometry\n"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "initial_state = json.load(open('data/wcecoli_t0.json'))\n",
    "\n",
    "bulk_ids = [item[0] for item in initial_state['bulk']]\n",
    "\n",
    "bulk.columns = bulk_ids"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "ecocyc_to_wcm_map = {}\n",
    "\n",
    "# combined complex and protein names\n",
    "complex_protein_names = list(filter_protein_df['id']) + list(filter_complex_df['id'])\n",
    "\n",
    "for name in complex_protein_names:\n",
    "\n",
    "    # find complex name in bulk_ids\n",
    "    found = False\n",
    "\n",
    "    try:\n",
    "        idx = bulk_ids.index(name+'[c]')\n",
    "        ecocyc_to_wcm_map[name] = name+'[c]'\n",
    "        found = True\n",
    "        # print(f'found {complex_name} at {idx}')\n",
    "\n",
    "    except ValueError:\n",
    "        # delete key\n",
    "        found = False\n",
    "\n",
    "\n",
    "    if found == False:\n",
    "\n",
    "        for id in bulk_ids:\n",
    "            if name+'[' in id and id.startswith(name) and bulk.loc[:, id].sum() > 0:\n",
    "                #print(f'found {name} in {id} with nonzero count')\n",
    "                ecocyc_to_wcm_map[name] = id\n",
    "                found = True\n",
    "                break           # ensures preferring nonzero counts\n",
    "\n",
    "            elif name+'[' in id and id.startswith(name):\n",
    "                # print(f'found {name} in {id} with zero count')\n",
    "                ecocyc_to_wcm_map[name] = id\n",
    "                found = True\n",
    "\n",
    "    if found == False:\n",
    "        ecocyc_to_wcm_map[name] = '--TRANS-ACENAPHTHENE-12-DIOL[j]' # should be none\n",
    "        print(f'could not find {name}')\n"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "complex_wcm_names = [ecocyc_to_wcm_map[name] for name in C_names]\n",
    "\n",
    "counts = bulk.loc[0, complex_wcm_names]"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "# Finally ... add the counts >:o"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "factored_cofactor_elements = np.array(counts).reshape(-1,1) * C @ P @ E"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "factored_cofactor_elements"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "{unique_elements[i]: factored_cofactor_elements.sum(axis=0)[i] for i in range(len(unique_elements))}"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "cidx = C_names.index('APORNAP-CPLX')\n",
    "{unique_elements[i]: factored_cofactor_elements[cidx,i] for i in range(len(unique_elements)) if factored_cofactor_elements[cidx,i] > 0}"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "# Saving outputs to files compatible with Julia"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "complex_ids = list(filter_complex_df['id'])\n",
    "protein_ids = list(filter_protein_df['id'])\n",
    "cofactor_ids = list(filter_cofactor_df['id'])\n",
    "\n",
    "\n",
    "# save C, P and E to julia-compatible formats\n",
    "np.savetxt('notebooks/fbagd/data/C_matrix.csv', C.astype(np.int64), delimiter=',', fmt='%i')\n",
    "np.savetxt('notebooks/fbagd/data/P_matrix.csv', P.astype(np.int64), delimiter=',', fmt='%i')\n",
    "np.savetxt('notebooks/fbagd/data/E_matrix.csv', E.astype(np.int64), delimiter=',', fmt='%i')\n",
    "\n",
    "\n",
    "\n",
    "# write all names to single file with each list on a new line\n",
    "with open('notebooks/fbagd/data/complex_ids.txt', 'w') as f:\n",
    "    f.write('\\n'.join(complex_ids))\n",
    "with open('notebooks/fbagd/data/protein_ids.txt', 'w') as f:\n",
    "    f.write('\\n'.join(protein_ids))\n",
    "with open('notebooks/fbagd/data/cofactor_ids.txt', 'w') as f:\n",
    "    f.write('\\n'.join(cofactor_ids))\n",
    "with open('notebooks/fbagd/data/element_ids.txt', 'w') as f:\n",
    "    f.write('\\n'.join(element_ids))\n",
    "\n",
    "# write all names to single file with each list on a new line\n",
    "with open('notebooks/fbagd/data/complex_ids.txt', 'w') as f:\n",
    "    f.write('\\n'.join(complex_ids))\n",
    "with open('notebooks/fbagd/data/protein_ids.txt', 'w') as f:\n",
    "    f.write('\\n'.join(protein_ids))\n",
    "with open('notebooks/fbagd/data/cofactor_ids.txt', 'w') as f:\n",
    "    f.write('\\n'.join(cofactor_ids))\n",
    "\n",
    "\n",
    "# save counts of proteins and complexes\n",
    "np.savetxt('notebooks/fbagd/data/counts.csv', np.array(counts, dtype=np.int64), delimiter=',', fmt='%i')\n",
    "\n",
    "# save tree_matrix\n",
    "np.savetxt('notebooks/fbagd/data/tree_matrix.csv', tree_matrix, delimiter=',', fmt='%i')\n",
    "\n"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "filter_complex_df"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "for name in protein_ids:\n",
    "    if 'CPLX' in name:\n",
    "        print(name)"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "counts[ecocyc_to_wcm_map[\"NAP-CPLX\"]]"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "ecocyc_to_wcm_map[\"NAP-CPLX\"]"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "\"^\"+'NAP-CPLX'+\"[\" in 'NAP-CPLX['"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "protein_ids[1]"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "## go annotation for proteins"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "\n",
    "# if 'Protein' in o and 'has-go-term' in o['Protein']:\n",
    "#     go_terms = o['Protein']['has-go-term']\n",
    "#\n",
    "#     annotation_set = set()\n",
    "#\n",
    "#     if type(go_terms) is dict:\n",
    "#         go_terms = [go_terms]\n",
    "#\n",
    "#     for go_term in go_terms:\n",
    "#         # example entry\n",
    "#         entity = go_term['GO-Term']['@frameid']\n",
    "#         entity = entity.replace(':', '%3A')\n",
    "#         req_str = f'https://www.ebi.ac.uk/QuickGO/services/ontology/go/terms/{entity}'\n",
    "#\n",
    "#         r = s.get(req_str)\n",
    "#         if r.status_code != 200:\n",
    "#             print(entity, r.status_code)\n",
    "#\n",
    "#         o = json.loads(r.content)\n",
    "#         if len(o['results']) != 1:\n",
    "#             print('error', entity, len(o['results']))\n",
    "#             continue\n",
    "#         elif o['results'][0]['aspect'] == 'biological_process':\n",
    "#             annotation_set.add(o['results'][0]['name'])\n",
    "#\n",
    "#             o['ptools-xml']['Protein']['gene']['Gene']['@frameid']\n",
    "#\n",
    "#     proteins_df.at[i, 'direct_annotations'] = annotation_set"
   ],
   "metadata": {
    "collapsed": false
   }
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
