

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../../../../">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>reconstruction.ecoli.dataclasses.state.internal_state &mdash; Vivarium E. coli 1.0.0 documentation</title>
      <link rel="stylesheet" type="text/css" href="../../../../../_static/pygments.css?v=b86133f3" />
      <link rel="stylesheet" type="text/css" href="../../../../../_static/css/theme.css?v=e59714d7" />

  
      <script src="../../../../../_static/jquery.js?v=5d32c60e"></script>
      <script src="../../../../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../../../../../_static/documentation_options.js?v=8d563738"></script>
      <script src="../../../../../_static/doctools.js?v=9bcbadda"></script>
      <script src="../../../../../_static/sphinx_highlight.js?v=dc90522c"></script>
      <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../../../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../../../genindex.html" />
    <link rel="search" title="Search" href="../../../../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../../../index.html" class="icon icon-home">
            Vivarium E. coli
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../../../../stores.html">Stores</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../processes.html">Processes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../composites.html">Composites</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../experiments.html">Experiments</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../workflows.html">Workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../output.html">Output</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../tutorial.html">Tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../docs.html">Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../hpc.html">HPC Clusters</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../gcloud.html">Google Cloud</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../ci.html">Continuous Integration</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../pycharm.html">PyCharm Setup</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../diffs.html">Model Differences From wcEcoli</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../reference/api_ref.html">API Reference</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../../../index.html">Vivarium E. coli</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../../../index.html">Module code</a></li>
      <li class="breadcrumb-item active">reconstruction.ecoli.dataclasses.state.internal_state</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for reconstruction.ecoli.dataclasses.state.internal_state</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">SimulationData state associated data</span>

<span class="sd">&quot;&quot;&quot;</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">wholecell.utils</span><span class="w"> </span><span class="kn">import</span> <span class="n">units</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">reconstruction.ecoli.dataclasses.state.bulkMolecules</span><span class="w"> </span><span class="kn">import</span> <span class="n">BulkMolecules</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">reconstruction.ecoli.dataclasses.state.uniqueMolecules</span><span class="w"> </span><span class="kn">import</span> <span class="n">UniqueMolecules</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>


<div class="viewcode-block" id="InternalState">
<a class="viewcode-back" href="../../../../../reference/api/validation/validation.ecoli.validation_data.html#reconstruction.ecoli.dataclasses.state.internal_state.InternalState">[docs]</a>
<span class="k">class</span><span class="w"> </span><span class="nc">InternalState</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Internal State&quot;&quot;&quot;</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">raw_data</span><span class="p">,</span> <span class="n">sim_data</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">bulk_molecules</span> <span class="o">=</span> <span class="n">BulkMolecules</span><span class="p">(</span><span class="n">raw_data</span><span class="p">,</span> <span class="n">sim_data</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">unique_molecule</span> <span class="o">=</span> <span class="n">UniqueMolecules</span><span class="p">(</span><span class="n">raw_data</span><span class="p">,</span> <span class="n">sim_data</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_build_bulk_molecules</span><span class="p">(</span><span class="n">sim_data</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_build_unique_molecules</span><span class="p">(</span><span class="n">sim_data</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_build_compartments</span><span class="p">(</span><span class="n">raw_data</span><span class="p">,</span> <span class="n">sim_data</span><span class="p">)</span>

<div class="viewcode-block" id="InternalState._build_bulk_molecules">
<a class="viewcode-back" href="../../../../../reference/api/validation/validation.ecoli.validation_data.html#reconstruction.ecoli.dataclasses.state.internal_state.InternalState._build_bulk_molecules">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">_build_bulk_molecules</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">sim_data</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Add data (IDs and mass) for all classes of bulk molecules.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">all_bulk_molecule_ids</span> <span class="o">=</span> <span class="n">sim_data</span><span class="o">.</span><span class="n">getter</span><span class="o">.</span><span class="n">get_all_valid_molecules</span><span class="p">()</span>

        <span class="n">all_bulk_molecule_ids_with_compartments</span><span class="p">,</span> <span class="n">all_bulk_molecule_masses</span> <span class="o">=</span> <span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_build_bulk_molecule_specs</span><span class="p">(</span><span class="n">sim_data</span><span class="p">,</span> <span class="n">all_bulk_molecule_ids</span><span class="p">)</span>
        <span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">bulk_molecules</span><span class="o">.</span><span class="n">add_to_bulk_state</span><span class="p">(</span>
            <span class="n">all_bulk_molecule_ids_with_compartments</span><span class="p">,</span> <span class="n">all_bulk_molecule_masses</span>
        <span class="p">)</span>
        <span class="n">sim_data</span><span class="o">.</span><span class="n">molecule_groups</span><span class="o">.</span><span class="n">bulk_molecules_binomial_division</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span>
            <span class="n">all_bulk_molecule_ids_with_compartments</span>
        <span class="p">)</span></div>


<div class="viewcode-block" id="InternalState._build_bulk_molecule_specs">
<a class="viewcode-back" href="../../../../../reference/api/validation/validation.ecoli.validation_data.html#reconstruction.ecoli.dataclasses.state.internal_state.InternalState._build_bulk_molecule_specs">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">_build_bulk_molecule_specs</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">sim_data</span><span class="p">,</span> <span class="n">molecule_ids</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Builds a list of molecule IDs with compartment tags and a corresponding</span>
<span class="sd">        array of molecular masses to add to the bulk state.</span>
<span class="sd">        Args:</span>
<span class="sd">                molecule_ids (List[str]): List of molecule IDs w/o compartment tags</span>
<span class="sd">        Returns:</span>
<span class="sd">                molecule_ids_with_compartments (List[str]): List of molecule IDs</span>
<span class="sd">                        with compartment tags</span>
<span class="sd">                masses (np.ndarray): Array of molecular masses divided into</span>
<span class="sd">                        submasses</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">molecule_ids_with_compartments</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">masses</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="c1"># Loop through each molecule species and associated compartments</span>
        <span class="k">for</span> <span class="n">molecule_id</span> <span class="ow">in</span> <span class="n">molecule_ids</span><span class="p">:</span>
            <span class="n">mw</span> <span class="o">=</span> <span class="n">sim_data</span><span class="o">.</span><span class="n">getter</span><span class="o">.</span><span class="n">get_submass_array</span><span class="p">(</span><span class="n">molecule_id</span><span class="p">)</span><span class="o">.</span><span class="n">asNumber</span><span class="p">(</span>
                <span class="n">units</span><span class="o">.</span><span class="n">g</span> <span class="o">/</span> <span class="n">units</span><span class="o">.</span><span class="n">mol</span>
            <span class="p">)</span>

            <span class="k">for</span> <span class="n">loc</span> <span class="ow">in</span> <span class="n">sim_data</span><span class="o">.</span><span class="n">getter</span><span class="o">.</span><span class="n">get_compartment</span><span class="p">(</span><span class="n">molecule_id</span><span class="p">):</span>
                <span class="n">molecule_ids_with_compartments</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2">[</span><span class="si">{}</span><span class="s2">]&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">molecule_id</span><span class="p">,</span> <span class="n">loc</span><span class="p">))</span>
                <span class="n">masses</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">mw</span><span class="p">)</span>

        <span class="n">masses</span> <span class="o">=</span> <span class="p">(</span><span class="n">units</span><span class="o">.</span><span class="n">g</span> <span class="o">/</span> <span class="n">units</span><span class="o">.</span><span class="n">mol</span><span class="p">)</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">masses</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">molecule_ids_with_compartments</span><span class="p">,</span> <span class="n">masses</span></div>


<div class="viewcode-block" id="InternalState._build_unique_molecules">
<a class="viewcode-back" href="../../../../../reference/api/validation/validation.ecoli.validation_data.html#reconstruction.ecoli.dataclasses.state.internal_state.InternalState._build_unique_molecules">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">_build_unique_molecules</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">sim_data</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Add data (name, mass, and attribute data structure) for all classes of</span>
<span class="sd">        unique molecules.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Set up dictionary for quick indexing of bulk molecule masses</span>
        <span class="n">bulk_molecule_id_to_mass</span> <span class="o">=</span> <span class="p">{</span>
            <span class="n">molecule_id</span><span class="p">:</span> <span class="n">mass</span>
            <span class="k">for</span> <span class="p">(</span><span class="n">molecule_id</span><span class="p">,</span> <span class="n">mass</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">bulk_molecules</span><span class="o">.</span><span class="n">bulk_data</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">],</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">bulk_molecules</span><span class="o">.</span><span class="n">bulk_data</span><span class="p">[</span><span class="s2">&quot;mass&quot;</span><span class="p">],</span>
            <span class="p">)</span>
        <span class="p">}</span>

        <span class="c1"># Add active RNA polymerase</span>
        <span class="c1"># The attributes of active RNA polymerases are given as:</span>
        <span class="c1"># - domain_index (32-bit int): Domain index of the chromosome domain</span>
        <span class="c1"># that the RNAP is bound to. This value is used to split the RNAPs at</span>
        <span class="c1"># cell division.</span>
        <span class="c1"># - coordinates (64-bit int): Location of the RNAP on the chromosome,</span>
        <span class="c1"># in base pairs from origin.</span>
        <span class="c1"># - is_forward (bool): True if RNAP is moving in the positive direction</span>
        <span class="c1"># of the genomic coordinates, False if RNAP is moving in the negative</span>
        <span class="c1"># direction. This is determined by the orientation of the TU that the</span>
        <span class="c1"># RNAP is transcribing.</span>
        <span class="n">RNAP_mass</span> <span class="o">=</span> <span class="n">bulk_molecule_id_to_mass</span><span class="p">[</span><span class="n">sim_data</span><span class="o">.</span><span class="n">molecule_ids</span><span class="o">.</span><span class="n">full_RNAP</span><span class="p">]</span>
        <span class="n">RNAP_attributes</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;domain_index&quot;</span><span class="p">:</span> <span class="s2">&quot;i4&quot;</span><span class="p">,</span>
            <span class="s2">&quot;coordinates&quot;</span><span class="p">:</span> <span class="s2">&quot;i8&quot;</span><span class="p">,</span>
            <span class="s2">&quot;is_forward&quot;</span><span class="p">:</span> <span class="s2">&quot;?&quot;</span><span class="p">,</span>
        <span class="p">}</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">unique_molecule</span><span class="o">.</span><span class="n">add_to_unique_state</span><span class="p">(</span>
            <span class="s2">&quot;active_RNAP&quot;</span><span class="p">,</span> <span class="n">RNAP_attributes</span><span class="p">,</span> <span class="n">RNAP_mass</span>
        <span class="p">)</span>

        <span class="c1"># RNAPs are divided based on the index of the chromosome domain they</span>
        <span class="c1"># are bound to</span>
        <span class="n">sim_data</span><span class="o">.</span><span class="n">molecule_groups</span><span class="o">.</span><span class="n">unique_molecules_domain_index_division</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
            <span class="s2">&quot;active_RNAP&quot;</span>
        <span class="p">)</span>

        <span class="c1"># Add RNAs</span>
        <span class="c1"># This molecule represents all RNAs that should be represented as</span>
        <span class="c1"># unique molecules. These RNAs include all partially transcribed RNAs</span>
        <span class="c1"># and fully transcribed mRNAs. The attributes of RNAs are given as:</span>
        <span class="c1"># - TU_index (64-bit int): Index of the transcription unit that the</span>
        <span class="c1"># RNA molecule is representing. Determines the sequence and the length</span>
        <span class="c1"># of the fully elongated RNA.</span>
        <span class="c1"># - transcript_length (64-bit int): Current length of the RNA.</span>
        <span class="c1"># - is_mRNA (bool): True if RNA represents an mRNA molecule.</span>
        <span class="c1"># - is_full_transcript (bool): True if RNA is fully transcribed and</span>
        <span class="c1"># released from the RNA polymerase, False if RNA is being actively</span>
        <span class="c1"># transcribed. This cannot be True if &#39;is_mRNA&#39; is False, since all</span>
        <span class="c1"># non-mRNAs are represented as bulk molecules when they are fully</span>
        <span class="c1"># transcribed.</span>
        <span class="c1"># - can_translate (bool): True if the 5&#39; end of the mRNA molecule is</span>
        <span class="c1"># available for translation. This flag is used only for mRNA molecules.</span>
        <span class="c1"># - RNAP_index (64-bit int): Unique index of the RNA polymerase that is</span>
        <span class="c1"># synthesizing the RNA or have synthesized the mRNA. For fully</span>
        <span class="c1"># transcribed mRNAs that are added at initialization, this attribute</span>
        <span class="c1"># is set to -1.</span>
        <span class="n">RNA_mass</span> <span class="o">=</span> <span class="p">(</span><span class="n">units</span><span class="o">.</span><span class="n">g</span> <span class="o">/</span> <span class="n">units</span><span class="o">.</span><span class="n">mol</span><span class="p">)</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros_like</span><span class="p">(</span><span class="n">RNAP_mass</span><span class="p">)</span>
        <span class="n">RNA_attributes</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;TU_index&quot;</span><span class="p">:</span> <span class="s2">&quot;i8&quot;</span><span class="p">,</span>
            <span class="s2">&quot;transcript_length&quot;</span><span class="p">:</span> <span class="s2">&quot;i8&quot;</span><span class="p">,</span>
            <span class="s2">&quot;is_mRNA&quot;</span><span class="p">:</span> <span class="s2">&quot;?&quot;</span><span class="p">,</span>
            <span class="s2">&quot;is_full_transcript&quot;</span><span class="p">:</span> <span class="s2">&quot;?&quot;</span><span class="p">,</span>
            <span class="s2">&quot;can_translate&quot;</span><span class="p">:</span> <span class="s2">&quot;?&quot;</span><span class="p">,</span>
            <span class="s2">&quot;RNAP_index&quot;</span><span class="p">:</span> <span class="s2">&quot;i8&quot;</span><span class="p">,</span>
        <span class="p">}</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">unique_molecule</span><span class="o">.</span><span class="n">add_to_unique_state</span><span class="p">(</span><span class="s2">&quot;RNA&quot;</span><span class="p">,</span> <span class="n">RNA_attributes</span><span class="p">,</span> <span class="n">RNA_mass</span><span class="p">)</span>

        <span class="c1"># Fully transcribed mRNAs are divided binomially, partial transcripts</span>
        <span class="c1"># are divided based on which chromosome domains their associated RNAPs</span>
        <span class="c1"># are bound to</span>
        <span class="n">sim_data</span><span class="o">.</span><span class="n">molecule_groups</span><span class="o">.</span><span class="n">unique_molecules_RNA_division</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;RNA&quot;</span><span class="p">)</span>

        <span class="c1"># Add active ribosomes</span>
        <span class="c1"># The attributes of active ribosomes are given as:</span>
        <span class="c1"># - protein_index (64-bit int): Index of the protein monomer that the</span>
        <span class="c1"># ribosome is translating</span>
        <span class="c1"># - peptide_length (64-bit int): Current length of the polypeptide that</span>
        <span class="c1"># the ribosome is translating, in number of amino acids</span>
        <span class="c1"># - mRNA_index (64-bit int): Unique index of the mRNA that the ribosome</span>
        <span class="c1"># is bound to</span>
        <span class="c1"># - pos_on_mRNA (64-bit int): Location of the ribosome on the bound</span>
        <span class="c1"># mRNA, in number of bases from the transcription start site</span>
        <span class="n">ribosome_30S_mass</span> <span class="o">=</span> <span class="n">bulk_molecule_id_to_mass</span><span class="p">[</span>
            <span class="n">sim_data</span><span class="o">.</span><span class="n">molecule_ids</span><span class="o">.</span><span class="n">s30_full_complex</span>
        <span class="p">]</span>
        <span class="n">ribosome_50S_mass</span> <span class="o">=</span> <span class="n">bulk_molecule_id_to_mass</span><span class="p">[</span>
            <span class="n">sim_data</span><span class="o">.</span><span class="n">molecule_ids</span><span class="o">.</span><span class="n">s50_full_complex</span>
        <span class="p">]</span>
        <span class="n">ribosome_mass</span> <span class="o">=</span> <span class="n">ribosome_30S_mass</span> <span class="o">+</span> <span class="n">ribosome_50S_mass</span>
        <span class="n">ribosome_attributes</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;protein_index&quot;</span><span class="p">:</span> <span class="s2">&quot;i8&quot;</span><span class="p">,</span>
            <span class="s2">&quot;peptide_length&quot;</span><span class="p">:</span> <span class="s2">&quot;i8&quot;</span><span class="p">,</span>
            <span class="s2">&quot;mRNA_index&quot;</span><span class="p">:</span> <span class="s2">&quot;i8&quot;</span><span class="p">,</span>
            <span class="s2">&quot;pos_on_mRNA&quot;</span><span class="p">:</span> <span class="s2">&quot;i8&quot;</span><span class="p">,</span>
        <span class="p">}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">unique_molecule</span><span class="o">.</span><span class="n">add_to_unique_state</span><span class="p">(</span>
            <span class="s2">&quot;active_ribosome&quot;</span><span class="p">,</span> <span class="n">ribosome_attributes</span><span class="p">,</span> <span class="n">ribosome_mass</span>
        <span class="p">)</span>

        <span class="c1"># Active ribosomes are divided such that they always follow the mRNA</span>
        <span class="c1"># molecule they are bound to</span>
        <span class="n">sim_data</span><span class="o">.</span><span class="n">molecule_groups</span><span class="o">.</span><span class="n">unique_molecules_active_ribosome_division</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
            <span class="s2">&quot;active_ribosome&quot;</span>
        <span class="p">)</span>

        <span class="c1"># Add full chromosomes</span>
        <span class="c1"># One full chromosome molecule is added when chromosome replication is</span>
        <span class="c1"># complete, and sets cell division to happen after a length of time</span>
        <span class="c1"># specified by the D period (if D_PERIOD_DIVISION is set to True).</span>
        <span class="c1"># The &#39;has_triggered_division&#39; attribute is initially set to False, and</span>
        <span class="c1"># is reset to True when division_time was reached and the cell has</span>
        <span class="c1"># divided. The &#39;domain_index&#39; keeps track of the index of the oldest</span>
        <span class="c1"># chromosome domain that is part of the full chromosome.</span>
        <span class="n">full_chromosome_mass</span> <span class="o">=</span> <span class="p">(</span><span class="n">units</span><span class="o">.</span><span class="n">g</span> <span class="o">/</span> <span class="n">units</span><span class="o">.</span><span class="n">mol</span><span class="p">)</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros_like</span><span class="p">(</span><span class="n">RNAP_mass</span><span class="p">)</span>
        <span class="n">full_chromosome_mass</span><span class="p">[</span><span class="n">sim_data</span><span class="o">.</span><span class="n">submass_name_to_index</span><span class="p">[</span><span class="s2">&quot;DNA&quot;</span><span class="p">]]</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">sim_data</span><span class="o">.</span><span class="n">getter</span><span class="o">.</span><span class="n">get_mass</span><span class="p">(</span><span class="n">sim_data</span><span class="o">.</span><span class="n">molecule_ids</span><span class="o">.</span><span class="n">full_chromosome</span><span class="p">)</span>
        <span class="p">)</span>
        <span class="n">full_chromosome_attributes</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;division_time&quot;</span><span class="p">:</span> <span class="s2">&quot;f8&quot;</span><span class="p">,</span>
            <span class="s2">&quot;has_triggered_division&quot;</span><span class="p">:</span> <span class="s2">&quot;?&quot;</span><span class="p">,</span>
            <span class="s2">&quot;domain_index&quot;</span><span class="p">:</span> <span class="s2">&quot;i4&quot;</span><span class="p">,</span>
        <span class="p">}</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">unique_molecule</span><span class="o">.</span><span class="n">add_to_unique_state</span><span class="p">(</span>
            <span class="s2">&quot;full_chromosome&quot;</span><span class="p">,</span> <span class="n">full_chromosome_attributes</span><span class="p">,</span> <span class="n">full_chromosome_mass</span>
        <span class="p">)</span>

        <span class="c1"># Full chromosomes are divided based on their domain index</span>
        <span class="n">sim_data</span><span class="o">.</span><span class="n">molecule_groups</span><span class="o">.</span><span class="n">unique_molecules_domain_index_division</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
            <span class="s2">&quot;full_chromosome&quot;</span>
        <span class="p">)</span>

        <span class="c1"># Add chromosome domains</span>
        <span class="c1"># Chromosome domains are zero-mass molecules that accounts for the</span>
        <span class="c1"># structures of replicating chromosomes. Each replication initiation</span>
        <span class="c1"># event creates two new chromosome domains that are given a unique</span>
        <span class="c1"># integer &#39;domain_index&#39;. These two new domains are child domains of</span>
        <span class="c1"># the original domain that the origin belonged to.</span>
        <span class="n">chromosome_domain_mass</span> <span class="o">=</span> <span class="p">(</span><span class="n">units</span><span class="o">.</span><span class="n">g</span> <span class="o">/</span> <span class="n">units</span><span class="o">.</span><span class="n">mol</span><span class="p">)</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros_like</span><span class="p">(</span><span class="n">RNAP_mass</span><span class="p">)</span>
        <span class="n">chromosome_domain_attributes</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;domain_index&quot;</span><span class="p">:</span> <span class="s2">&quot;i4&quot;</span><span class="p">,</span>
            <span class="s2">&quot;child_domains&quot;</span><span class="p">:</span> <span class="p">(</span><span class="s2">&quot;i4&quot;</span><span class="p">,</span> <span class="mi">2</span><span class="p">),</span>
        <span class="p">}</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">unique_molecule</span><span class="o">.</span><span class="n">add_to_unique_state</span><span class="p">(</span>
            <span class="s2">&quot;chromosome_domain&quot;</span><span class="p">,</span> <span class="n">chromosome_domain_attributes</span><span class="p">,</span> <span class="n">chromosome_domain_mass</span>
        <span class="p">)</span>

        <span class="c1"># Chromosome domains are divided based on their domain index</span>
        <span class="n">sim_data</span><span class="o">.</span><span class="n">molecule_groups</span><span class="o">.</span><span class="n">unique_molecules_domain_index_division</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
            <span class="s2">&quot;chromosome_domain&quot;</span>
        <span class="p">)</span>

        <span class="c1"># Add active replisomes</span>
        <span class="c1"># Note that the replisome does not functionally replicate the</span>
        <span class="c1"># chromosome, but instead keeps track of the mass associated with</span>
        <span class="c1"># essential subunits of the replisome complex (if the mechanistic</span>
        <span class="c1"># replisome option is turned on). The list of essential subunits and</span>
        <span class="c1"># their stoichiometry were taken from Reyes-Lamothe et al., 2010.</span>
        <span class="n">replisome_mass</span> <span class="o">=</span> <span class="p">(</span><span class="n">units</span><span class="o">.</span><span class="n">g</span> <span class="o">/</span> <span class="n">units</span><span class="o">.</span><span class="n">mol</span><span class="p">)</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros_like</span><span class="p">(</span><span class="n">RNAP_mass</span><span class="p">)</span>
        <span class="n">replisome_attributes</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;domain_index&quot;</span><span class="p">:</span> <span class="s2">&quot;i4&quot;</span><span class="p">,</span>
            <span class="s2">&quot;right_replichore&quot;</span><span class="p">:</span> <span class="s2">&quot;?&quot;</span><span class="p">,</span>
            <span class="s2">&quot;coordinates&quot;</span><span class="p">:</span> <span class="s2">&quot;i8&quot;</span><span class="p">,</span>
        <span class="p">}</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">unique_molecule</span><span class="o">.</span><span class="n">add_to_unique_state</span><span class="p">(</span>
            <span class="s2">&quot;active_replisome&quot;</span><span class="p">,</span> <span class="n">replisome_attributes</span><span class="p">,</span> <span class="n">replisome_mass</span>
        <span class="p">)</span>

        <span class="c1"># Active replisomes are divided based on their domain index</span>
        <span class="n">sim_data</span><span class="o">.</span><span class="n">molecule_groups</span><span class="o">.</span><span class="n">unique_molecules_domain_index_division</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
            <span class="s2">&quot;active_replisome&quot;</span>
        <span class="p">)</span>

        <span class="c1"># Add origins of replication</span>
        <span class="c1"># Note that origins are conceptual molecules and have zero mass. The</span>
        <span class="c1"># chromosomeIndexes of oriC&#39;s determine the chromosomeIndexes of the</span>
        <span class="c1"># new partial chromosomes and replisomes initiated on the same oriC.</span>
        <span class="n">originMass</span> <span class="o">=</span> <span class="p">(</span><span class="n">units</span><span class="o">.</span><span class="n">g</span> <span class="o">/</span> <span class="n">units</span><span class="o">.</span><span class="n">mol</span><span class="p">)</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros_like</span><span class="p">(</span><span class="n">RNAP_mass</span><span class="p">)</span>
        <span class="n">originAttributes</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;domain_index&quot;</span><span class="p">:</span> <span class="s2">&quot;i4&quot;</span><span class="p">,</span>
        <span class="p">}</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">unique_molecule</span><span class="o">.</span><span class="n">add_to_unique_state</span><span class="p">(</span><span class="s2">&quot;oriC&quot;</span><span class="p">,</span> <span class="n">originAttributes</span><span class="p">,</span> <span class="n">originMass</span><span class="p">)</span>

        <span class="c1"># oriC&#39;s are divided based on their domain index</span>
        <span class="n">sim_data</span><span class="o">.</span><span class="n">molecule_groups</span><span class="o">.</span><span class="n">unique_molecules_domain_index_division</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;oriC&quot;</span><span class="p">)</span>

        <span class="c1"># Add promoters</span>
        <span class="c1"># Promoters are sequences on the DNA where RNA polymerases bind to and</span>
        <span class="c1"># initiate transcription. They can also be bound to transcription</span>
        <span class="c1"># factors(TFs), if the transcription unit associated with the promoter</span>
        <span class="c1"># is regulated by TFs. The promoter itself has zero mass but can hold</span>
        <span class="c1"># the mass of the transcription factor that it is bound to. Its</span>
        <span class="c1"># attributes are given as:</span>
        <span class="c1"># - TU_index (64-bit int): Index of the transcription unit that</span>
        <span class="c1"># the promoter is associated with. This determines which TFs the</span>
        <span class="c1"># promoter can bind to, and how the transcription probability is</span>
        <span class="c1"># calculated.</span>
        <span class="c1"># - coordinates (64-bit int): Location of the promoter on the</span>
        <span class="c1"># chromosome, in base pairs from origin. This value does not change</span>
        <span class="c1"># after the molecule is initialized.</span>
        <span class="c1"># - domain_index (32-bit int): Domain index of the chromosome domain</span>
        <span class="c1"># that the promoter belongs to. This value is used to split the</span>
        <span class="c1"># promoters at cell division.</span>
        <span class="c1"># - bound_TF (boolean array of length n_tf): A boolean array that</span>
        <span class="c1"># shows which TFs the promoter is bound to. Note that one promoter can</span>
        <span class="c1"># bind to multiple TFs.</span>
        <span class="n">n_tf</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">sim_data</span><span class="o">.</span><span class="n">process</span><span class="o">.</span><span class="n">transcription_regulation</span><span class="o">.</span><span class="n">tf_ids</span><span class="p">)</span>

        <span class="n">promoter_mass</span> <span class="o">=</span> <span class="p">(</span><span class="n">units</span><span class="o">.</span><span class="n">g</span> <span class="o">/</span> <span class="n">units</span><span class="o">.</span><span class="n">mol</span><span class="p">)</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros_like</span><span class="p">(</span><span class="n">RNAP_mass</span><span class="p">)</span>
        <span class="n">promoter_attributes</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;TU_index&quot;</span><span class="p">:</span> <span class="s2">&quot;i8&quot;</span><span class="p">,</span>
            <span class="s2">&quot;coordinates&quot;</span><span class="p">:</span> <span class="s2">&quot;i8&quot;</span><span class="p">,</span>
            <span class="s2">&quot;domain_index&quot;</span><span class="p">:</span> <span class="s2">&quot;i4&quot;</span><span class="p">,</span>
            <span class="s2">&quot;bound_TF&quot;</span><span class="p">:</span> <span class="p">(</span><span class="s2">&quot;?&quot;</span><span class="p">,</span> <span class="n">n_tf</span><span class="p">),</span>
        <span class="p">}</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">unique_molecule</span><span class="o">.</span><span class="n">add_to_unique_state</span><span class="p">(</span>
            <span class="s2">&quot;promoter&quot;</span><span class="p">,</span> <span class="n">promoter_attributes</span><span class="p">,</span> <span class="n">promoter_mass</span>
        <span class="p">)</span>

        <span class="c1"># Promoters are divided based on their domain index</span>
        <span class="n">sim_data</span><span class="o">.</span><span class="n">molecule_groups</span><span class="o">.</span><span class="n">unique_molecules_domain_index_division</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
            <span class="s2">&quot;promoter&quot;</span>
        <span class="p">)</span>

        <span class="c1"># Add genes</span>
        <span class="c1"># Genes are sequences on the DNA that encode for particular cistrons.</span>
        <span class="c1"># Its attributes are given as:</span>
        <span class="c1"># - cistron_index (64-bit int): Index of the cistron that the gene</span>
        <span class="c1"># encodes for.</span>
        <span class="c1"># - coordinates (64-bit int): Location of the gene on the chromosome, in</span>
        <span class="c1"># base pairs from origin. This value does not change after the molecule</span>
        <span class="c1"># is initialized.</span>
        <span class="c1"># - domain_index (32-bit int): Domain index of the chromosome domain</span>
        <span class="c1"># that the gene belongs to. This value is used to split the genes at</span>
        <span class="c1"># cell division.</span>
        <span class="n">gene_mass</span> <span class="o">=</span> <span class="p">(</span><span class="n">units</span><span class="o">.</span><span class="n">g</span> <span class="o">/</span> <span class="n">units</span><span class="o">.</span><span class="n">mol</span><span class="p">)</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros_like</span><span class="p">(</span><span class="n">RNAP_mass</span><span class="p">)</span>
        <span class="n">gene_attributes</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;cistron_index&quot;</span><span class="p">:</span> <span class="s2">&quot;i8&quot;</span><span class="p">,</span>
            <span class="s2">&quot;coordinates&quot;</span><span class="p">:</span> <span class="s2">&quot;i8&quot;</span><span class="p">,</span>
            <span class="s2">&quot;domain_index&quot;</span><span class="p">:</span> <span class="s2">&quot;i4&quot;</span><span class="p">,</span>
        <span class="p">}</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">unique_molecule</span><span class="o">.</span><span class="n">add_to_unique_state</span><span class="p">(</span><span class="s2">&quot;gene&quot;</span><span class="p">,</span> <span class="n">gene_attributes</span><span class="p">,</span> <span class="n">gene_mass</span><span class="p">)</span>

        <span class="c1"># Genes are divided based on their domain index</span>
        <span class="n">sim_data</span><span class="o">.</span><span class="n">molecule_groups</span><span class="o">.</span><span class="n">unique_molecules_domain_index_division</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;gene&quot;</span><span class="p">)</span>

        <span class="c1"># Add chromosomal segments</span>
        <span class="c1"># Chromosomal segments are segments of DNA that are topologically</span>
        <span class="c1"># constrained, such that the changes in the linking number of the</span>
        <span class="c1"># segment do not spread beyond its boundaries. For a relaxed DNA double</span>
        <span class="c1"># helix, the linking number is proportional to the length of the</span>
        <span class="c1"># chromosomal segment. Currently only active RNA polymerases and</span>
        <span class="c1"># replisomes serve as segment boundaries. Its attributes are given as:</span>
        <span class="c1"># - boundary_molecule_indexes (pair of 64-bit ints): Unique indexes</span>
        <span class="c1"># of molecules at the boundaries of the chromosomal segment</span>
        <span class="c1"># - boundary_coordinates (pair of 64-bit ints): Coordinates of the</span>
        <span class="c1"># boundaries of the chromosomal segment</span>
        <span class="c1"># - domain_index (32-bit int): Domain index of the chromosome domain</span>
        <span class="c1"># that the segment belongs to.</span>
        <span class="c1"># - linking_number (64-bit float): Linking number of the chromosomal</span>
        <span class="c1"># segment.</span>
        <span class="n">chromosomal_segment_mass</span> <span class="o">=</span> <span class="p">(</span><span class="n">units</span><span class="o">.</span><span class="n">g</span> <span class="o">/</span> <span class="n">units</span><span class="o">.</span><span class="n">mol</span><span class="p">)</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros_like</span><span class="p">(</span><span class="n">RNAP_mass</span><span class="p">)</span>
        <span class="n">chromosomal_segment_attributes</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;boundary_molecule_indexes&quot;</span><span class="p">:</span> <span class="p">(</span><span class="s2">&quot;i8&quot;</span><span class="p">,</span> <span class="mi">2</span><span class="p">),</span>
            <span class="s2">&quot;boundary_coordinates&quot;</span><span class="p">:</span> <span class="p">(</span><span class="s2">&quot;i8&quot;</span><span class="p">,</span> <span class="mi">2</span><span class="p">),</span>
            <span class="s2">&quot;domain_index&quot;</span><span class="p">:</span> <span class="s2">&quot;i4&quot;</span><span class="p">,</span>
            <span class="s2">&quot;linking_number&quot;</span><span class="p">:</span> <span class="s2">&quot;f8&quot;</span><span class="p">,</span>
        <span class="p">}</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">unique_molecule</span><span class="o">.</span><span class="n">add_to_unique_state</span><span class="p">(</span>
            <span class="s2">&quot;chromosomal_segment&quot;</span><span class="p">,</span>
            <span class="n">chromosomal_segment_attributes</span><span class="p">,</span>
            <span class="n">chromosomal_segment_mass</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="c1"># Chromosomal segments are divided based on their domain indexes, but</span>
        <span class="c1"># division occurs after all chromosome-bound molecules are divided to</span>
        <span class="c1"># properly reset the boundary_molecule_indexes attribute</span>
        <span class="n">sim_data</span><span class="o">.</span><span class="n">molecule_groups</span><span class="o">.</span><span class="n">unique_molecules_chromosomal_segment_division</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
            <span class="s2">&quot;chromosomal_segment&quot;</span>
        <span class="p">)</span>

        <span class="c1"># Add DnaA boxes</span>
        <span class="c1"># DnaA boxes are 9-base sequence motifs on the DNA that bind to the</span>
        <span class="c1"># protein DnaA. Except for DnaA boxes close to the origin, these boxes</span>
        <span class="c1"># serve no functional role in replication initiation, but can</span>
        <span class="c1"># effectively titrate away free DnaA molecules and control its</span>
        <span class="c1"># concentration. The molecule itself has zero mass but it can hold the</span>
        <span class="c1"># mass of the DnaA protein that it is bound to. Its attributes are</span>
        <span class="c1"># given as:</span>
        <span class="c1"># - coordinates (64-bit int): Location of the middle base (5th base) of</span>
        <span class="c1"># the DnaA box, in base pairs from origin. This value does not change</span>
        <span class="c1"># after the molecule is initialized.</span>
        <span class="c1"># - domain_index (32-bit int): Domain index of the chromosome domain</span>
        <span class="c1"># that the DnaA box belongs to. This value is used to allocate DnaA</span>
        <span class="c1"># boxes to the two daughter cells at cell division.</span>
        <span class="c1"># - DnaA_bound (boolean): True if bound to a DnaA protein, False if not</span>
        <span class="n">DnaA_box_mass</span> <span class="o">=</span> <span class="p">(</span><span class="n">units</span><span class="o">.</span><span class="n">g</span> <span class="o">/</span> <span class="n">units</span><span class="o">.</span><span class="n">mol</span><span class="p">)</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros_like</span><span class="p">(</span><span class="n">RNAP_mass</span><span class="p">)</span>
        <span class="n">DnaA_box_attributes</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;coordinates&quot;</span><span class="p">:</span> <span class="s2">&quot;i8&quot;</span><span class="p">,</span>
            <span class="s2">&quot;domain_index&quot;</span><span class="p">:</span> <span class="s2">&quot;i4&quot;</span><span class="p">,</span>
            <span class="s2">&quot;DnaA_bound&quot;</span><span class="p">:</span> <span class="s2">&quot;?&quot;</span><span class="p">,</span>
        <span class="p">}</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">unique_molecule</span><span class="o">.</span><span class="n">add_to_unique_state</span><span class="p">(</span>
            <span class="s2">&quot;DnaA_box&quot;</span><span class="p">,</span> <span class="n">DnaA_box_attributes</span><span class="p">,</span> <span class="n">DnaA_box_mass</span>
        <span class="p">)</span>

        <span class="c1"># DnaA boxes are divided based on their domain index</span>
        <span class="n">sim_data</span><span class="o">.</span><span class="n">molecule_groups</span><span class="o">.</span><span class="n">unique_molecules_domain_index_division</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
            <span class="s2">&quot;DnaA_box&quot;</span>
        <span class="p">)</span></div>


<div class="viewcode-block" id="InternalState._build_compartments">
<a class="viewcode-back" href="../../../../../reference/api/validation/validation.ecoli.validation_data.html#reconstruction.ecoli.dataclasses.state.internal_state.InternalState._build_compartments">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">_build_compartments</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">raw_data</span><span class="p">,</span> <span class="n">sim_data</span><span class="p">):</span>
        <span class="n">_</span> <span class="o">=</span> <span class="n">sim_data</span>
        <span class="n">compartmentData</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">empty</span><span class="p">(</span>
            <span class="nb">len</span><span class="p">(</span><span class="n">raw_data</span><span class="o">.</span><span class="n">compartments</span><span class="p">),</span>
            <span class="n">dtype</span><span class="o">=</span><span class="p">[(</span><span class="s2">&quot;id&quot;</span><span class="p">,</span> <span class="s2">&quot;U20&quot;</span><span class="p">),</span> <span class="p">(</span><span class="s2">&quot;compartmentAbbreviation&quot;</span><span class="p">,</span> <span class="s2">&quot;U1&quot;</span><span class="p">)],</span>
        <span class="p">)</span>

        <span class="n">compartmentData</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">raw_data</span><span class="o">.</span><span class="n">compartments</span><span class="p">]</span>
        <span class="n">compartmentData</span><span class="p">[</span><span class="s2">&quot;compartmentAbbreviation&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span>
            <span class="n">x</span><span class="p">[</span><span class="s2">&quot;abbrev&quot;</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">raw_data</span><span class="o">.</span><span class="n">compartments</span>
        <span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">compartments</span> <span class="o">=</span> <span class="n">compartmentData</span></div>
</div>

</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2021-2026, The Vivarium E. coli Authors.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>