

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../../../">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>ecoli.analysis.multivariant.new_gene_translation_efficiency_heatmaps &mdash; Vivarium E. coli 1.0.0 documentation</title>
      <link rel="stylesheet" type="text/css" href="../../../../_static/pygments.css?v=b86133f3" />
      <link rel="stylesheet" type="text/css" href="../../../../_static/css/theme.css?v=e59714d7" />

  
      <script src="../../../../_static/jquery.js?v=5d32c60e"></script>
      <script src="../../../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../../../../_static/documentation_options.js?v=8d563738"></script>
      <script src="../../../../_static/doctools.js?v=9bcbadda"></script>
      <script src="../../../../_static/sphinx_highlight.js?v=dc90522c"></script>
      <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../../genindex.html" />
    <link rel="search" title="Search" href="../../../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../../index.html" class="icon icon-home">
            Vivarium E. coli
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../../../stores.html">Stores</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../processes.html">Processes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../composites.html">Composites</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../experiments.html">Experiments</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../workflows.html">Workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../output.html">Output</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../tutorial.html">Tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../docs.html">Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../hpc.html">HPC Clusters</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../gcloud.html">Google Cloud</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../ci.html">Continuous Integration</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../pycharm.html">PyCharm Setup</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../reference/api_ref.html">API Reference</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../../index.html">Vivarium E. coli</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../../index.html">Module code</a></li>
      <li class="breadcrumb-item active">ecoli.analysis.multivariant.new_gene_translation_efficiency_heatmaps</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for ecoli.analysis.multivariant.new_gene_translation_efficiency_heatmaps</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Plot one value per index via heatmap for</span>
<span class="sd">new_gene_expression_and_translation_efficiency variant.</span>

<span class="sd">Possible Plots:</span>

<span class="sd">- Percent of sims that successfully reached a given generation number</span>
<span class="sd">- Average doubling time</span>
<span class="sd">- Average cell volume, mass, dry cell mass, mRNA mass, protein mass</span>
<span class="sd">- Average translation efficiency, weighted by cistron count</span>
<span class="sd">- Average mRNA count, monomer count, mRNA mass fraction, protein mass fraction,</span>
<span class="sd">  RNAP portion, and ribosome portion for a capacity gene to measure burden on</span>
<span class="sd">  overall host expression</span>
<span class="sd">- Average new gene copy number</span>
<span class="sd">- Average new gene mRNA count</span>
<span class="sd">- Average new gene mRNA mass fraction</span>
<span class="sd">- Average new gene mRNA counts fraction</span>
<span class="sd">- Average new gene NTP mass fraction</span>
<span class="sd">- Average new gene protein count</span>
<span class="sd">- Average new gene protein mass fraction</span>
<span class="sd">- Average new gene protein counts fraction</span>
<span class="sd">- Average new gene initialization rate for RNAP and ribosomes</span>
<span class="sd">- Average new gene initialization probabilities for RNAP and ribosomes</span>
<span class="sd">- Average count and portion of new gene ribosome initialization events per time</span>
<span class="sd">  step</span>
<span class="sd">- Average number and proportion of RNAP on new genes at a given time step</span>
<span class="sd">- Average number and proportion of ribosomes on new gene mRNAs at a given time</span>
<span class="sd">  step</span>
<span class="sd">- Average number and proportion of RNAP making rRNAs at a given time step</span>
<span class="sd">- Average number and proportion of RNAP and ribosomes making RNAP subunits at</span>
<span class="sd">  a given time step</span>
<span class="sd">- Average number and proportion of RNAP and ribosomes making ribosomal proteins</span>
<span class="sd">  at a given time step</span>
<span class="sd">- Average fraction of time new gene is overcrowded by RNAP and Ribosomes</span>
<span class="sd">- Average overcrowding probability ratio for new gene RNA synthesis and</span>
<span class="sd">  polypeptide initiation</span>
<span class="sd">- Average max_p probabilities for RNA synthesis and polypeptide initiation</span>
<span class="sd">- Average number of overcrowded genes for RNAP and Ribosomes</span>
<span class="sd">- Average number of total, active, and free ribosomes</span>
<span class="sd">- Average number of ribosomes initialized at each time step</span>
<span class="sd">- Average number of total active, and free RNA polymerases</span>
<span class="sd">- Average ppGpp concentration</span>
<span class="sd">- Average rate of glucose consumption</span>
<span class="sd">- Average new gene monomer yields - per hour and per fg of glucose</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">itertools</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">duckdb</span><span class="w"> </span><span class="kn">import</span> <span class="n">DuckDBPyConnection</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy.typing</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">npt</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">matplotlib</span><span class="w"> </span><span class="kn">import</span> <span class="n">pyplot</span> <span class="k">as</span> <span class="n">plt</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">math</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">polars</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pl</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">unum.units</span><span class="w"> </span><span class="kn">import</span> <span class="n">fg</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">typing</span><span class="w"> </span><span class="kn">import</span> <span class="n">Any</span><span class="p">,</span> <span class="n">Callable</span><span class="p">,</span> <span class="n">cast</span><span class="p">,</span> <span class="n">Optional</span><span class="p">,</span> <span class="n">TYPE_CHECKING</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">tqdm</span><span class="w"> </span><span class="kn">import</span> <span class="n">tqdm</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">ecoli.variants.new_gene_internal_shift</span><span class="w"> </span><span class="kn">import</span> <span class="n">get_new_gene_ids_and_indices</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">ecoli.library.parquet_emitter</span><span class="w"> </span><span class="kn">import</span> <span class="p">(</span>
    <span class="n">field_metadata</span><span class="p">,</span>
    <span class="n">ndlist_to_ndarray</span><span class="p">,</span>
    <span class="n">open_arbitrary_sim_data</span><span class="p">,</span>
    <span class="n">read_stacked_columns</span><span class="p">,</span>
<span class="p">)</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">ecoli.library.schema</span><span class="w"> </span><span class="kn">import</span> <span class="n">bulk_name_to_idx</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">wholecell.utils.plotting_tools</span><span class="w"> </span><span class="kn">import</span> <span class="n">export_figure</span><span class="p">,</span> <span class="n">heatmap</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">wholecell.utils</span><span class="w"> </span><span class="kn">import</span> <span class="n">units</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">pickle</span>

<span class="k">if</span> <span class="n">TYPE_CHECKING</span><span class="p">:</span>
    <span class="kn">from</span><span class="w"> </span><span class="nn">reconstruction.ecoli.fit_sim_data_1</span><span class="w"> </span><span class="kn">import</span> <span class="n">SimulationDataEcoli</span>

<span class="n">FONT_SIZE</span> <span class="o">=</span> <span class="mi">9</span>

<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Dashboard Flag</span>

<span class="sd">- 0: Separate Only (Each plot is its own file)</span>
<span class="sd">- 1: Dashboard Only (One file with all plots)</span>
<span class="sd">- 2: Both Dashboard and Separate</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="n">DASHBOARD_FLAG</span> <span class="o">=</span> <span class="mi">2</span>

<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Standard Deviations Flag</span>

<span class="sd">- True: Plot an additional copy of all plots with standard deviation displayed</span>
<span class="sd">  insted of the average</span>
<span class="sd">- False: Plot no additional plots</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="n">STD_DEV_FLAG</span> <span class="o">=</span> <span class="kc">True</span>

<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Count number of sims that reach this generation (remember index 7 </span>
<span class="sd">corresponds to generation 8)</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="n">COUNT_INDEX</span> <span class="o">=</span> <span class="mi">32</span>
<span class="c1"># COUNT_INDEX = 2 ### TODO: revert back after developing plot locally</span>

<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Plot data from generations [MIN_CELL_INDEX, MAX_CELL_INDEX)</span>
<span class="sd">Note that early generations may not be representative of dynamics </span>
<span class="sd">due to how they are initialized</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="n">MIN_CELL_INDEX</span> <span class="o">=</span> <span class="mi">16</span>
<span class="c1"># # MIN_CELL_INDEX = 1 ### TODO: revert back after developing plot locally</span>
<span class="c1"># MIN_CELL_INDEX = 0</span>
<span class="n">MAX_CELL_INDEX</span> <span class="o">=</span> <span class="mi">33</span>

<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Specify which subset of heatmaps should be made</span>
<span class="sd">Completed_gens heatmap is always made, because it is used to</span>
<span class="sd">create the other heatmaps, and should not be included here.</span>
<span class="sd">The order listed here will be the order of the heatmaps in the</span>
<span class="sd">dashboard plot.</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="n">HEATMAPS_TO_MAKE_LIST</span> <span class="o">=</span> <span class="p">[</span>
    <span class="s2">&quot;completed_gens_heatmap&quot;</span><span class="p">,</span>
    <span class="s2">&quot;doubling_times_heatmap&quot;</span><span class="p">,</span>
    <span class="s2">&quot;cell_mass_heatmap&quot;</span><span class="p">,</span>
    <span class="s2">&quot;cell_dry_mass_heatmap&quot;</span><span class="p">,</span>
    <span class="s2">&quot;cell_volume_heatmap&quot;</span><span class="p">,</span>
    <span class="s2">&quot;ppgpp_concentration_heatmap&quot;</span><span class="p">,</span>
    <span class="s2">&quot;rnap_crowding_heatmap&quot;</span><span class="p">,</span>
    <span class="s2">&quot;ribosome_crowding_heatmap&quot;</span><span class="p">,</span>
    <span class="s2">&quot;cell_mRNA_mass_heatmap&quot;</span><span class="p">,</span>
    <span class="s2">&quot;cell_protein_mass_heatmap&quot;</span><span class="p">,</span>
    <span class="s2">&quot;rnap_counts_heatmap&quot;</span><span class="p">,</span>
    <span class="s2">&quot;ribosome_counts_heatmap&quot;</span><span class="p">,</span>
    <span class="s2">&quot;new_gene_mRNA_counts_heatmap&quot;</span><span class="p">,</span>
    <span class="s2">&quot;new_gene_monomer_counts_heatmap&quot;</span><span class="p">,</span>
    <span class="s2">&quot;new_gene_copy_number_heatmap&quot;</span><span class="p">,</span>
    <span class="s2">&quot;new_gene_rnap_init_rate_heatmap&quot;</span><span class="p">,</span>
    <span class="s2">&quot;new_gene_ribosome_init_rate_heatmap&quot;</span><span class="p">,</span>
    <span class="s2">&quot;new_gene_mRNA_mass_fraction_heatmap&quot;</span><span class="p">,</span>
    <span class="s2">&quot;new_gene_monomer_mass_fraction_heatmap&quot;</span><span class="p">,</span>
    <span class="s2">&quot;new_gene_rnap_time_overcrowded_heatmap&quot;</span><span class="p">,</span>
    <span class="s2">&quot;new_gene_ribosome_time_overcrowded_heatmap&quot;</span><span class="p">,</span>
    <span class="s2">&quot;new_gene_mRNA_counts_fraction_heatmap&quot;</span><span class="p">,</span>
    <span class="s2">&quot;new_gene_monomer_counts_fraction_heatmap&quot;</span><span class="p">,</span>
    <span class="s2">&quot;active_rnap_counts_heatmap&quot;</span><span class="p">,</span>
    <span class="s2">&quot;active_ribosome_counts_heatmap&quot;</span><span class="p">,</span>
    <span class="s2">&quot;new_gene_rnap_counts_heatmap&quot;</span><span class="p">,</span>
    <span class="s2">&quot;new_gene_rnap_portion_heatmap&quot;</span><span class="p">,</span>
    <span class="s2">&quot;rrna_rnap_counts_heatmap&quot;</span><span class="p">,</span>
    <span class="s2">&quot;rrna_rnap_portion_heatmap&quot;</span><span class="p">,</span>
    <span class="s2">&quot;rnap_subunit_rnap_counts_heatmap&quot;</span><span class="p">,</span>
    <span class="s2">&quot;rnap_subunit_rnap_portion_heatmap&quot;</span><span class="p">,</span>
    <span class="s2">&quot;rnap_subunit_ribosome_counts_heatmap&quot;</span><span class="p">,</span>
    <span class="s2">&quot;rnap_subunit_ribosome_portion_heatmap&quot;</span><span class="p">,</span>
    <span class="s2">&quot;ribosomal_protein_rnap_counts_heatmap&quot;</span><span class="p">,</span>
    <span class="s2">&quot;ribosomal_protein_rnap_portion_heatmap&quot;</span><span class="p">,</span>
    <span class="s2">&quot;ribosomal_protein_ribosome_counts_heatmap&quot;</span><span class="p">,</span>
    <span class="s2">&quot;ribosomal_protein_ribosome_portion_heatmap&quot;</span><span class="p">,</span>
    <span class="s2">&quot;new_gene_ribosome_counts_heatmap&quot;</span><span class="p">,</span>
    <span class="s2">&quot;new_gene_ribosome_portion_heatmap&quot;</span><span class="p">,</span>
    <span class="s2">&quot;weighted_avg_translation_efficiency_heatmap&quot;</span><span class="p">,</span>
    <span class="s2">&quot;protein_init_prob_max_p_heatmap&quot;</span><span class="p">,</span>
    <span class="s2">&quot;new_gene_protein_init_prob_max_p_target_ratio_heatmap&quot;</span><span class="p">,</span>
    <span class="s2">&quot;new_gene_target_protein_init_prob_heatmap&quot;</span><span class="p">,</span>
    <span class="s2">&quot;new_gene_actual_protein_init_prob_heatmap&quot;</span><span class="p">,</span>
    <span class="s2">&quot;rna_synth_prob_max_p_heatmap&quot;</span><span class="p">,</span>
    <span class="s2">&quot;new_gene_rna_synth_prob_max_p_target_ratio_heatmap&quot;</span><span class="p">,</span>
    <span class="s2">&quot;new_gene_target_rna_synth_prob_heatmap&quot;</span><span class="p">,</span>
    <span class="s2">&quot;new_gene_actual_rna_synth_prob_heatmap&quot;</span><span class="p">,</span>
    <span class="s2">&quot;capacity_gene_mRNA_counts_heatmap&quot;</span><span class="p">,</span>
    <span class="s2">&quot;capacity_gene_monomer_counts_heatmap&quot;</span><span class="p">,</span>
    <span class="s2">&quot;capacity_gene_rnap_portion_heatmap&quot;</span><span class="p">,</span>
    <span class="s2">&quot;capacity_gene_ribosome_portion_heatmap&quot;</span><span class="p">,</span>
    <span class="s2">&quot;capacity_gene_mRNA_mass_fraction_heatmap&quot;</span><span class="p">,</span>
    <span class="s2">&quot;capacity_gene_monomer_mass_fraction_heatmap&quot;</span><span class="p">,</span>
    <span class="s2">&quot;capacity_gene_mRNA_counts_fraction_heatmap&quot;</span><span class="p">,</span>
    <span class="s2">&quot;capacity_gene_monomer_counts_fraction_heatmap&quot;</span><span class="p">,</span>
    <span class="s2">&quot;free_rnap_counts_heatmap&quot;</span><span class="p">,</span>
    <span class="s2">&quot;free_ribosome_counts_heatmap&quot;</span><span class="p">,</span>
    <span class="s2">&quot;rnap_ribosome_counts_ratio_heatmap&quot;</span><span class="p">,</span>
    <span class="s2">&quot;ribosome_init_events_heatmap&quot;</span><span class="p">,</span>
    <span class="s2">&quot;new_gene_ribosome_init_events_heatmap&quot;</span><span class="p">,</span>
    <span class="s2">&quot;new_gene_ribosome_init_events_portion_heatmap&quot;</span><span class="p">,</span>
    <span class="s2">&quot;new_gene_yield_per_glucose&quot;</span><span class="p">,</span>
    <span class="s2">&quot;new_gene_yield_per_hour&quot;</span><span class="p">,</span>
    <span class="s2">&quot;glucose_consumption_rate&quot;</span><span class="p">,</span>
    <span class="s2">&quot;new_gene_mRNA_NTP_fraction_heatmap&quot;</span><span class="p">,</span>
<span class="p">]</span>

<span class="n">capacity_gene_monomer_ids</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;EG10544-MONOMER[m]&quot;</span><span class="p">]</span>
<span class="c1"># capacity_gene_monomer_id = [&quot;EG11036-MONOMER[c]&quot;]</span>


<div class="viewcode-block" id="get_mean_and_std_matrices">
<a class="viewcode-back" href="../../../../reference/api/ecoli/ecoli.analysis.multivariant.new_gene_translation_efficiency_heatmaps.html#ecoli.analysis.multivariant.new_gene_translation_efficiency_heatmaps.get_mean_and_std_matrices">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">get_mean_and_std_matrices</span><span class="p">(</span>
    <span class="n">conn</span><span class="p">:</span> <span class="n">DuckDBPyConnection</span><span class="p">,</span>
    <span class="n">variant_mapping</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">tuple</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">]],</span>
    <span class="n">variant_matrix_shape</span><span class="p">:</span> <span class="nb">tuple</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">],</span>
    <span class="n">history_sql</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">columns</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
    <span class="n">remove_first</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">func</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Callable</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">order_results</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">success_sql</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">custom_sql</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">post_func</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Callable</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">num_digits_rounding</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">default_value</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Any</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">tuple</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Reads one or more columns and calculates mean and std. dev. for each</span>
<span class="sd">    variant. If no custom SQL query is provided, this defaults to averaging</span>
<span class="sd">    per cell, then calculating the averages and standard deviations of all</span>
<span class="sd">    cells per variant.</span>

<span class="sd">    Args:</span>
<span class="sd">        conn: DuckDB connection</span>
<span class="sd">        variant_mapping: Mapping of variant IDs to row and column in matrix</span>
<span class="sd">            of new gene translation efficiency and expression factor variants</span>
<span class="sd">        variant_matrix_shape: Number of rows and columns in variant matrix</span>
<span class="sd">        history_sql: SQL subquery from :py:func:`ecoli.library.parquet_emitter.dataset_sql`</span>
<span class="sd">        columns, remove_first, func, order_results, success_sql: See</span>
<span class="sd">            :py:func:`ecoli.library.parquet_emitter.read_stacked_columns`</span>
<span class="sd">        custom_sql: SQL string containing a placeholder with name ``subquery``</span>
<span class="sd">            where the result of read_stacked_columns will be placed. Final query</span>
<span class="sd">            result must only have two columns in order: ``variant`` and a value</span>
<span class="sd">            for each variant. If not provided, defaults to average of averages</span>
<span class="sd">        post_func: Function that is called on Polars DataFrame resulting from query.</span>
<span class="sd">            Should return a Polars DataFrame with exactly three columns: ``variant``</span>
<span class="sd">            for the variant IDs, ``mean`` for some mean aggregate value (can be</span>
<span class="sd">            N-D list column), and ``std`` for some standard deviation aggregate.</span>
<span class="sd">        num_digits_rounding: Number of decimal places to round to</span>
<span class="sd">        default_value: Default value to put in output variant matrices if</span>
<span class="sd">            variant ID not included in query result (e.g. if variant failed in</span>
<span class="sd">            first generation and had no completed sims)</span>
<span class="sd">        new_gene_NTP_fraction: Set to True for NTP fraction heatmap so query output</span>
<span class="sd">            is properly handled</span>

<span class="sd">    Returns:</span>
<span class="sd">        Tuple of Numpy matrices with first two dimensions ``variant_matrix_shape``.</span>
<span class="sd">        Each cell in first matrix has the mean for that variant. Each cell in the</span>
<span class="sd">        second matrix has the std. dev. for that variant. These values can be Numpy</span>
<span class="sd">        arrays instead of scalar values (e.g. when calculating aggregates for many</span>
<span class="sd">        genes at once), in which case the matrices have shapes</span>
<span class="sd">        ``variant_matrix_shape + (num_genes,)``</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">subquery</span> <span class="o">=</span> <span class="n">read_stacked_columns</span><span class="p">(</span>
        <span class="n">history_sql</span><span class="o">=</span><span class="n">history_sql</span><span class="p">,</span>
        <span class="n">columns</span><span class="o">=</span><span class="n">columns</span><span class="p">,</span>
        <span class="n">remove_first</span><span class="o">=</span><span class="n">remove_first</span><span class="p">,</span>
        <span class="n">func</span><span class="o">=</span><span class="n">func</span><span class="p">,</span>
        <span class="n">order_results</span><span class="o">=</span><span class="n">order_results</span><span class="p">,</span>
        <span class="n">success_sql</span><span class="o">=</span><span class="n">success_sql</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="k">if</span> <span class="n">custom_sql</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">columns</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span>
                <span class="s2">&quot;Must provide custom SQL expression to handle multiple columns at once.&quot;</span>
            <span class="p">)</span>
        <span class="n">custom_sql</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">            WITH avg_per_cell AS (</span>
<span class="s2">                SELECT avg(</span><span class="si">{</span><span class="n">columns</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2">) AS avg_col, experiment_id, variant</span>
<span class="s2">                FROM (</span><span class="se">{{</span><span class="s2">subquery</span><span class="se">}}</span><span class="s2">)</span>
<span class="s2">                GROUP BY experiment_id, variant, lineage_seed, generation, agent_id</span>
<span class="s2">            )</span>
<span class="s2">            SELECT variant, avg(avg_col) AS mean, stddev(avg_col) AS std</span>
<span class="s2">            FROM avg_per_cell</span>
<span class="s2">            GROUP BY experiment_id, variant</span>
<span class="s2">            &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">post_func</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">data</span> <span class="o">=</span> <span class="n">conn</span><span class="o">.</span><span class="n">sql</span><span class="p">(</span><span class="n">custom_sql</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">subquery</span><span class="o">=</span><span class="n">subquery</span><span class="p">))</span><span class="o">.</span><span class="n">pl</span><span class="p">()</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">data</span> <span class="o">=</span> <span class="n">conn</span><span class="o">.</span><span class="n">sql</span><span class="p">(</span><span class="n">custom_sql</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">subquery</span><span class="o">=</span><span class="n">subquery</span><span class="p">))</span><span class="o">.</span><span class="n">pl</span><span class="p">()</span>
        <span class="n">data</span> <span class="o">=</span> <span class="n">post_func</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
    <span class="k">if</span> <span class="nb">set</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span> <span class="o">!=</span> <span class="p">{</span><span class="s2">&quot;variant&quot;</span><span class="p">,</span> <span class="s2">&quot;mean&quot;</span><span class="p">,</span> <span class="s2">&quot;std&quot;</span><span class="p">}:</span>
        <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span>
            <span class="s2">&quot;post_func should return a Polars DataFrame with &quot;</span>
            <span class="s2">&quot;exactly three columns named `variant`, `mean`, and `std`&quot;</span>
        <span class="p">)</span>
    <span class="n">data</span> <span class="o">=</span> <span class="p">[(</span><span class="n">i</span><span class="p">[</span><span class="s2">&quot;variant&quot;</span><span class="p">],</span> <span class="n">i</span><span class="p">[</span><span class="s2">&quot;mean&quot;</span><span class="p">],</span> <span class="n">i</span><span class="p">[</span><span class="s2">&quot;std&quot;</span><span class="p">])</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">data</span><span class="o">.</span><span class="n">rows</span><span class="p">(</span><span class="n">named</span><span class="o">=</span><span class="kc">True</span><span class="p">)]</span>  <span class="c1"># type: ignore[assignment]</span>
    <span class="n">mean_matrix</span> <span class="o">=</span> <span class="p">[</span>
        <span class="p">[</span><span class="n">default_value</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">variant_matrix_shape</span><span class="p">[</span><span class="mi">1</span><span class="p">])]</span>
        <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">variant_matrix_shape</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
    <span class="p">]</span>
    <span class="n">std_matrix</span> <span class="o">=</span> <span class="p">[</span>
        <span class="p">[</span><span class="n">default_value</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">variant_matrix_shape</span><span class="p">[</span><span class="mi">1</span><span class="p">])]</span>
        <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">variant_matrix_shape</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
    <span class="p">]</span>
    <span class="k">for</span> <span class="n">variant</span><span class="p">,</span> <span class="n">mean</span><span class="p">,</span> <span class="n">std</span> <span class="ow">in</span> <span class="n">data</span><span class="p">:</span>
        <span class="n">variant_row</span><span class="p">,</span> <span class="n">variant_col</span> <span class="o">=</span> <span class="n">variant_mapping</span><span class="p">[</span><span class="n">variant</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">mean</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">num_digits_rounding</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">mean</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="n">mean</span><span class="p">,</span> <span class="n">num_digits_rounding</span><span class="p">)</span>
            <span class="n">mean_matrix</span><span class="p">[</span><span class="n">variant_row</span><span class="p">][</span><span class="n">variant_col</span><span class="p">]</span> <span class="o">=</span> <span class="n">mean</span>
        <span class="k">if</span> <span class="n">std</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">num_digits_rounding</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">std</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="n">std</span><span class="p">,</span> <span class="n">num_digits_rounding</span><span class="p">)</span>
            <span class="n">std_matrix</span><span class="p">[</span><span class="n">variant_row</span><span class="p">][</span><span class="n">variant_col</span><span class="p">]</span> <span class="o">=</span> <span class="n">std</span>
    <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">mean_matrix</span><span class="p">),</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">std_matrix</span><span class="p">)</span></div>



<div class="viewcode-block" id="get_mRNA_ids_from_monomer_ids">
<a class="viewcode-back" href="../../../../reference/api/ecoli/ecoli.analysis.multivariant.new_gene_translation_efficiency_heatmaps.html#ecoli.analysis.multivariant.new_gene_translation_efficiency_heatmaps.get_mRNA_ids_from_monomer_ids">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">get_mRNA_ids_from_monomer_ids</span><span class="p">(</span>
    <span class="n">sim_data</span><span class="p">:</span> <span class="s2">&quot;SimulationDataEcoli&quot;</span><span class="p">,</span> <span class="n">target_monomer_ids</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Map monomer IDs back to the mRNA IDs that they were translated from.</span>

<span class="sd">    Args:</span>
<span class="sd">        target_monomer_ids: IDs of the monomers to map to mRNA IDs</span>

<span class="sd">    Returns:</span>
<span class="sd">        List of mRNA ID lists, one for each monomer ID</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Map protein ids to cistron ids</span>
    <span class="n">monomer_ids</span> <span class="o">=</span> <span class="n">sim_data</span><span class="o">.</span><span class="n">process</span><span class="o">.</span><span class="n">translation</span><span class="o">.</span><span class="n">monomer_data</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]</span>
    <span class="n">cistron_ids</span> <span class="o">=</span> <span class="n">sim_data</span><span class="o">.</span><span class="n">process</span><span class="o">.</span><span class="n">translation</span><span class="o">.</span><span class="n">monomer_data</span><span class="p">[</span><span class="s2">&quot;cistron_id&quot;</span><span class="p">]</span>
    <span class="n">monomer_to_cistron_id_dict</span> <span class="o">=</span> <span class="p">{</span>
        <span class="n">monomer_id</span><span class="p">:</span> <span class="n">cistron_id</span>
        <span class="k">for</span> <span class="n">monomer_id</span><span class="p">,</span> <span class="n">cistron_id</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">monomer_ids</span><span class="p">,</span> <span class="n">cistron_ids</span><span class="p">)</span>
    <span class="p">}</span>
    <span class="n">target_cistron_ids</span> <span class="o">=</span> <span class="p">[</span>
        <span class="n">monomer_to_cistron_id_dict</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">monomer_id</span><span class="p">)</span> <span class="k">for</span> <span class="n">monomer_id</span> <span class="ow">in</span> <span class="n">target_monomer_ids</span>
    <span class="p">]</span>
    <span class="n">RNA_ids</span> <span class="o">=</span> <span class="n">sim_data</span><span class="o">.</span><span class="n">process</span><span class="o">.</span><span class="n">transcription</span><span class="o">.</span><span class="n">rna_data</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]</span>
    <span class="n">target_RNA_ids</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="c1"># Map cistron ids to RNA ids</span>
    <span class="k">for</span> <span class="n">RNAP_cistron_id</span> <span class="ow">in</span> <span class="n">target_cistron_ids</span><span class="p">:</span>
        <span class="n">target_RNA_idx</span> <span class="o">=</span> <span class="n">sim_data</span><span class="o">.</span><span class="n">process</span><span class="o">.</span><span class="n">transcription</span><span class="o">.</span><span class="n">cistron_id_to_rna_indexes</span><span class="p">(</span>
            <span class="n">RNAP_cistron_id</span>
        <span class="p">)</span>
        <span class="n">target_RNA_ids</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">RNA_ids</span><span class="p">[</span><span class="n">target_RNA_idx</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">())</span>
    <span class="k">return</span> <span class="n">target_RNA_ids</span></div>



<div class="viewcode-block" id="get_indexes">
<a class="viewcode-back" href="../../../../reference/api/ecoli/ecoli.analysis.multivariant.new_gene_translation_efficiency_heatmaps.html#ecoli.analysis.multivariant.new_gene_translation_efficiency_heatmaps.get_indexes">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">get_indexes</span><span class="p">(</span>
    <span class="n">conn</span><span class="p">:</span> <span class="n">DuckDBPyConnection</span><span class="p">,</span>
    <span class="n">config_sql</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">index_type</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">ids</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">|</span> <span class="nb">list</span><span class="p">[</span><span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]],</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="nb">int</span> <span class="o">|</span> <span class="kc">None</span><span class="p">]</span> <span class="o">|</span> <span class="nb">list</span><span class="p">[</span><span class="nb">list</span><span class="p">[</span><span class="nb">int</span> <span class="o">|</span> <span class="kc">None</span><span class="p">]]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Retrieve DuckDB indices of a given type for a set of IDs. Note that</span>
<span class="sd">    DuckDB lists are 1-indexed.</span>

<span class="sd">    Args:</span>
<span class="sd">        conn: DuckDB database connection</span>
<span class="sd">        config_sql: DuckDB SQL query for sim config data (see</span>
<span class="sd">            :py:func:`~ecoli.library.parquet_emitter.dataset_sql`)</span>
<span class="sd">        index_type: Type of indices to return (one of ``cistron``,</span>
<span class="sd">            ``RNA``, ``mRNA``, or ``monomer``)</span>
<span class="sd">        ids: List of IDs to get indices for (must be monomer IDs</span>
<span class="sd">            if ``index_type`` is ``monomer``, else mRNA IDs)</span>

<span class="sd">    Returns:</span>
<span class="sd">        List of requested indexes</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">index_type</span> <span class="o">==</span> <span class="s2">&quot;cistron&quot;</span><span class="p">:</span>
        <span class="c1"># Extract cistron indexes for each new gene</span>
        <span class="n">cistron_idx_dict</span> <span class="o">=</span> <span class="p">{</span>
            <span class="n">cis</span><span class="p">:</span> <span class="n">i</span> <span class="o">+</span> <span class="mi">1</span>
            <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">cis</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span>
                <span class="n">field_metadata</span><span class="p">(</span>
                    <span class="n">conn</span><span class="p">,</span> <span class="n">config_sql</span><span class="p">,</span> <span class="s2">&quot;listeners__rnap_data__rna_init_event_per_cistron&quot;</span>
                <span class="p">)</span>
            <span class="p">)</span>
        <span class="p">}</span>
        <span class="k">return</span> <span class="p">[</span><span class="n">cistron_idx_dict</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">cistron</span><span class="p">)</span> <span class="k">for</span> <span class="n">cistron</span> <span class="ow">in</span> <span class="n">ids</span><span class="p">]</span>
    <span class="k">elif</span> <span class="n">index_type</span> <span class="o">==</span> <span class="s2">&quot;RNA&quot;</span><span class="p">:</span>
        <span class="c1"># Extract RNA indexes for each new gene</span>
        <span class="n">RNA_idx_dict</span> <span class="o">=</span> <span class="p">{</span>
            <span class="n">rna</span><span class="p">:</span> <span class="n">i</span> <span class="o">+</span> <span class="mi">1</span>
            <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">rna</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span>
                <span class="n">field_metadata</span><span class="p">(</span>
                    <span class="n">conn</span><span class="p">,</span> <span class="n">config_sql</span><span class="p">,</span> <span class="s2">&quot;listeners__rna_synth_prob__target_rna_synth_prob&quot;</span>
                <span class="p">)</span>
            <span class="p">)</span>
        <span class="p">}</span>
        <span class="k">return</span> <span class="p">[[</span><span class="n">RNA_idx_dict</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">rna_id</span><span class="p">)</span> <span class="k">for</span> <span class="n">rna_id</span> <span class="ow">in</span> <span class="n">rna_ids</span><span class="p">]</span> <span class="k">for</span> <span class="n">rna_ids</span> <span class="ow">in</span> <span class="n">ids</span><span class="p">]</span>
    <span class="k">elif</span> <span class="n">index_type</span> <span class="o">==</span> <span class="s2">&quot;mRNA&quot;</span><span class="p">:</span>
        <span class="c1"># Extract mRNA indexes for each new gene</span>
        <span class="n">mRNA_idx_dict</span> <span class="o">=</span> <span class="p">{</span>
            <span class="n">rna</span><span class="p">:</span> <span class="n">i</span> <span class="o">+</span> <span class="mi">1</span>
            <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">rna</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span>
                <span class="n">field_metadata</span><span class="p">(</span><span class="n">conn</span><span class="p">,</span> <span class="n">config_sql</span><span class="p">,</span> <span class="s2">&quot;listeners__rna_counts__mRNA_counts&quot;</span><span class="p">)</span>
            <span class="p">)</span>
        <span class="p">}</span>
        <span class="k">return</span> <span class="p">[[</span><span class="n">mRNA_idx_dict</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">rna_id</span><span class="p">)</span> <span class="k">for</span> <span class="n">rna_id</span> <span class="ow">in</span> <span class="n">rna_ids</span><span class="p">]</span> <span class="k">for</span> <span class="n">rna_ids</span> <span class="ow">in</span> <span class="n">ids</span><span class="p">]</span>
    <span class="k">elif</span> <span class="n">index_type</span> <span class="o">==</span> <span class="s2">&quot;monomer&quot;</span><span class="p">:</span>
        <span class="c1"># Extract protein indexes for each new gene</span>
        <span class="n">monomer_idx_dict</span> <span class="o">=</span> <span class="p">{</span>
            <span class="n">monomer</span><span class="p">:</span> <span class="n">i</span> <span class="o">+</span> <span class="mi">1</span>
            <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">monomer</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span>
                <span class="n">field_metadata</span><span class="p">(</span><span class="n">conn</span><span class="p">,</span> <span class="n">config_sql</span><span class="p">,</span> <span class="s2">&quot;listeners__monomer_counts&quot;</span><span class="p">)</span>
            <span class="p">)</span>
        <span class="p">}</span>
        <span class="k">return</span> <span class="p">[</span><span class="n">monomer_idx_dict</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">monomer_id</span><span class="p">)</span> <span class="k">for</span> <span class="n">monomer_id</span> <span class="ow">in</span> <span class="n">ids</span><span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span>
            <span class="s2">&quot;Index type &quot;</span> <span class="o">+</span> <span class="n">index_type</span> <span class="o">+</span> <span class="s2">&quot; has no instructions for data extraction.&quot;</span>
        <span class="p">)</span></div>



<span class="n">GENE_COUNTS_SQL</span> <span class="o">=</span> <span class="s2">&quot;&quot;&quot;</span>
<span class="s2">    WITH unnested_counts AS (</span>
<span class="s2">        SELECT unnest(gene_counts) AS gene_counts,</span>
<span class="s2">            generate_subscripts(gene_counts, 1)</span>
<span class="s2">            AS gene_idx, experiment_id, variant, lineage_seed, generation,</span>
<span class="s2">            agent_id</span>
<span class="s2">        FROM (</span><span class="si">{subquery}</span><span class="s2">)</span>
<span class="s2">    ),</span>
<span class="s2">    avg_per_cell AS (</span>
<span class="s2">        SELECT avg(gene_counts) AS avg_count,</span>
<span class="s2">            experiment_id, variant, gene_idx</span>
<span class="s2">        FROM unnested_counts</span>
<span class="s2">        GROUP BY experiment_id, variant, lineage_seed,</span>
<span class="s2">            generation, agent_id, gene_idx</span>
<span class="s2">    ),</span>
<span class="s2">    avg_per_variant AS (</span>
<span class="s2">        SELECT log10(avg(avg_count) + 1) AS avg_count,</span>
<span class="s2">            log10(stddev(avg_count) + 1) AS std_count,</span>
<span class="s2">            experiment_id, variant, gene_idx</span>
<span class="s2">        FROM avg_per_cell</span>
<span class="s2">        GROUP BY experiment_id, variant, gene_idx</span>
<span class="s2">    )</span>
<span class="s2">    SELECT variant, list(avg_count ORDER BY gene_idx) AS mean,</span>
<span class="s2">        list(std_count ORDER BY gene_idx) AS std,</span>
<span class="s2">    FROM avg_per_variant</span>
<span class="s2">    GROUP BY experiment_id, variant</span>
<span class="s2">    &quot;&quot;&quot;</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Generic SQL query for calculating average of a 1D-array column</span>
<span class="sd">per cell, aggregates that per variant into ``log10(mean + 1)`` and</span>
<span class="sd">``log10(std + 1)`` columns.</span>
<span class="sd">&quot;&quot;&quot;</span>


<div class="viewcode-block" id="get_gene_mass_prod_func">
<a class="viewcode-back" href="../../../../reference/api/ecoli/ecoli.analysis.multivariant.new_gene_translation_efficiency_heatmaps.html#ecoli.analysis.multivariant.new_gene_translation_efficiency_heatmaps.get_gene_mass_prod_func">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">get_gene_mass_prod_func</span><span class="p">(</span>
    <span class="n">sim_data</span><span class="p">:</span> <span class="s2">&quot;SimulationDataEcoli&quot;</span><span class="p">,</span>
    <span class="n">index_type</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">gene_ids</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">|</span> <span class="nb">list</span><span class="p">[</span><span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]],</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Callable</span><span class="p">[[</span><span class="n">pl</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">],</span> <span class="n">pl</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Create a function to be passed as the ``post_func`` argument to</span>
<span class="sd">    :py:func:`~.get_mean_and_std_matrices` which multiplies the</span>
<span class="sd">    average and standard deviation 1D array columns by the mass of</span>
<span class="sd">    the gene ID for each element.</span>

<span class="sd">    Args:</span>
<span class="sd">        sim_data: Simulation data</span>
<span class="sd">        index_type: Either ``mRNA`` or ``monomer``. If ``mRNA``,</span>
<span class="sd">            ``gene_ids`` is list of lists of mRNA IDs, where inner</span>
<span class="sd">            lists correspond to mRNAs for each gene. Therefore,</span>
<span class="sd">            we sum the masses for the mRNAs of each inner list</span>
<span class="sd">            and multiply the input mean and std by this sum per gene.</span>
<span class="sd">        gene_ids: IDs of genes in the order they appear in the</span>
<span class="sd">            1D arrays of the query result</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Get mass for gene</span>
    <span class="k">if</span> <span class="n">index_type</span> <span class="o">==</span> <span class="s2">&quot;monomer&quot;</span><span class="p">:</span>
        <span class="n">gene_masses</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span>
            <span class="p">[</span>
                <span class="p">(</span>
                    <span class="n">sim_data</span><span class="o">.</span><span class="n">getter</span><span class="o">.</span><span class="n">get_mass</span><span class="p">(</span><span class="n">gene_id</span><span class="p">)</span> <span class="o">/</span> <span class="n">sim_data</span><span class="o">.</span><span class="n">constants</span><span class="o">.</span><span class="n">n_avogadro</span>
                <span class="p">)</span><span class="o">.</span><span class="n">asNumber</span><span class="p">(</span><span class="n">fg</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">gene_id</span> <span class="ow">in</span> <span class="n">gene_ids</span>
            <span class="p">]</span>
        <span class="p">)</span>
    <span class="k">elif</span> <span class="n">index_type</span> <span class="o">==</span> <span class="s2">&quot;mRNA&quot;</span><span class="p">:</span>
        <span class="n">gene_masses</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span>
            <span class="p">[</span>
                <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span>
                    <span class="p">[</span>
                        <span class="p">(</span>
                            <span class="n">sim_data</span><span class="o">.</span><span class="n">getter</span><span class="o">.</span><span class="n">get_mass</span><span class="p">(</span><span class="n">gene_id</span><span class="p">)</span>
                            <span class="o">/</span> <span class="n">sim_data</span><span class="o">.</span><span class="n">constants</span><span class="o">.</span><span class="n">n_avogadro</span>
                        <span class="p">)</span><span class="o">.</span><span class="n">asNumber</span><span class="p">(</span><span class="n">fg</span><span class="p">)</span>
                        <span class="k">for</span> <span class="n">gene_id</span> <span class="ow">in</span> <span class="n">one_gene_ids</span>
                    <span class="p">]</span>
                <span class="p">)</span>
                <span class="k">for</span> <span class="n">one_gene_ids</span> <span class="ow">in</span> <span class="n">gene_ids</span>
            <span class="p">]</span>
        <span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="s2">&quot;index_type must be monomer or mRNA.&quot;</span><span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">gene_mass_prod</span><span class="p">(</span><span class="n">variant_agg</span><span class="p">):</span>
        <span class="n">avg_arr</span> <span class="o">=</span> <span class="n">ndlist_to_ndarray</span><span class="p">(</span><span class="n">variant_agg</span><span class="p">[</span><span class="s2">&quot;mean&quot;</span><span class="p">])</span>
        <span class="n">std_arr</span> <span class="o">=</span> <span class="n">ndlist_to_ndarray</span><span class="p">(</span><span class="n">variant_agg</span><span class="p">[</span><span class="s2">&quot;std&quot;</span><span class="p">])</span>
        <span class="k">return</span> <span class="n">pl</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
            <span class="p">{</span>
                <span class="s2">&quot;variant&quot;</span><span class="p">:</span> <span class="n">variant_agg</span><span class="p">[</span><span class="s2">&quot;variant&quot;</span><span class="p">],</span>
                <span class="s2">&quot;mean&quot;</span><span class="p">:</span> <span class="n">pl</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">avg_arr</span> <span class="o">*</span> <span class="n">gene_masses</span><span class="p">),</span>
                <span class="s2">&quot;std&quot;</span><span class="p">:</span> <span class="n">pl</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">std_arr</span> <span class="o">*</span> <span class="n">gene_masses</span><span class="p">),</span>
            <span class="p">}</span>
        <span class="p">)</span>

    <span class="k">return</span> <span class="n">gene_mass_prod</span></div>



<div class="viewcode-block" id="get_gene_count_fraction_sql">
<a class="viewcode-back" href="../../../../reference/api/ecoli/ecoli.analysis.multivariant.new_gene_translation_efficiency_heatmaps.html#ecoli.analysis.multivariant.new_gene_translation_efficiency_heatmaps.get_gene_count_fraction_sql">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">get_gene_count_fraction_sql</span><span class="p">(</span>
    <span class="n">gene_indices</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">|</span> <span class="nb">list</span><span class="p">[</span><span class="nb">list</span><span class="p">[</span><span class="nb">int</span><span class="p">]],</span> <span class="n">column</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">index_type</span><span class="p">:</span> <span class="nb">str</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Construct generic SQL query that gets the average per cell of a select</span>
<span class="sd">    set of indices from a 1D list column divided by the total of all elements</span>
<span class="sd">    per row of that list column, and aggregates those ratios per variant</span>
<span class="sd">    into mean and std columns.</span>

<span class="sd">    Args:</span>
<span class="sd">        gene_indices: Indices to extract from 1D list column to get ratios for</span>
<span class="sd">        column: Name of 1D list column</span>
<span class="sd">        index_type: Can either be ``monomer`` or ``mRNA``. For ``monomer``,</span>
<span class="sd">            function works exactly as described above. For ``mRNA``,</span>
<span class="sd">            ``gene_indices`` will be a list of lists of mRNA indices. This is</span>
<span class="sd">            because one gene can have to multiple mRNAs (transcription units).</span>
<span class="sd">            Therefore, we sum the elements corresponding to each gene before</span>
<span class="sd">            proceeding (see :py:func:`~.get_rnas_combined_as_genes_projection`).</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">index_type</span> <span class="o">==</span> <span class="s2">&quot;monomer&quot;</span><span class="p">:</span>
        <span class="n">list_to_unnest</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;list_select(</span><span class="si">{</span><span class="n">column</span><span class="si">}</span><span class="s2">, </span><span class="si">{</span><span class="n">gene_indices</span><span class="si">}</span><span class="s2">)&quot;</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">list_to_unnest</span> <span class="o">=</span> <span class="p">(</span>
            <span class="s2">&quot;[&quot;</span>
            <span class="o">+</span> <span class="s2">&quot;, &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
                <span class="p">[</span>
                    <span class="sa">f</span><span class="s2">&quot;list_sum(list_select(</span><span class="si">{</span><span class="n">column</span><span class="si">}</span><span class="s2">, </span><span class="si">{</span><span class="n">idx_for_one_gene</span><span class="si">}</span><span class="s2">))&quot;</span>
                    <span class="k">for</span> <span class="n">idx_for_one_gene</span> <span class="ow">in</span> <span class="n">gene_indices</span>
                <span class="p">]</span>
            <span class="p">)</span>
            <span class="o">+</span> <span class="s2">&quot;]&quot;</span>
        <span class="p">)</span>
    <span class="k">return</span> <span class="sa">f</span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">        WITH list_counts AS (</span>
<span class="s2">            SELECT </span><span class="si">{</span><span class="n">list_to_unnest</span><span class="si">}</span><span class="s2"> AS selected_counts, list_sum(</span><span class="si">{</span><span class="n">column</span><span class="si">}</span><span class="s2">)</span>
<span class="s2">                AS total_counts, experiment_id, variant, lineage_seed,</span>
<span class="s2">                generation, agent_id</span>
<span class="s2">            FROM (</span><span class="se">{{</span><span class="s2">subquery</span><span class="se">}}</span><span class="s2">)</span>
<span class="s2">        ),</span>
<span class="s2">        unnested_fracs AS (</span>
<span class="s2">            SELECT unnest(selected_counts) / total_counts AS gene_fracs,</span>
<span class="s2">                generate_subscripts(selected_counts, 1) AS gene_idx,</span>
<span class="s2">                experiment_id, variant, lineage_seed, generation, agent_id</span>
<span class="s2">            FROM list_counts</span>
<span class="s2">        ),</span>
<span class="s2">        avg_per_cell AS (</span>
<span class="s2">            SELECT avg(gene_fracs) AS avg_frac,</span>
<span class="s2">                experiment_id, variant, gene_idx</span>
<span class="s2">            FROM unnested_fracs</span>
<span class="s2">            GROUP BY experiment_id, variant, lineage_seed,</span>
<span class="s2">                generation, agent_id, gene_idx</span>
<span class="s2">        ),</span>
<span class="s2">        avg_per_variant AS (</span>
<span class="s2">            SELECT experiment_id, variant, avg(avg_frac)</span>
<span class="s2">                AS avg_frac, stddev(avg_frac) AS std_frac, gene_idx</span>
<span class="s2">            FROM avg_per_cell</span>
<span class="s2">            GROUP BY experiment_id, variant, gene_idx</span>
<span class="s2">        )</span>
<span class="s2">        SELECT variant, list(avg_frac ORDER BY gene_idx) AS mean,</span>
<span class="s2">            list(std_frac ORDER BY gene_idx) AS std,</span>
<span class="s2">        FROM avg_per_variant</span>
<span class="s2">        GROUP BY experiment_id, variant</span>
<span class="s2">        &quot;&quot;&quot;</span></div>



<div class="viewcode-block" id="get_new_gene_mRNA_NTP_fraction_sql">
<a class="viewcode-back" href="../../../../reference/api/ecoli/ecoli.analysis.multivariant.new_gene_translation_efficiency_heatmaps.html#ecoli.analysis.multivariant.new_gene_translation_efficiency_heatmaps.get_new_gene_mRNA_NTP_fraction_sql">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">get_new_gene_mRNA_NTP_fraction_sql</span><span class="p">(</span>
    <span class="n">sim_data</span><span class="p">:</span> <span class="s2">&quot;SimulationDataEcoli&quot;</span><span class="p">,</span>
    <span class="n">new_gene_mRNA_idx</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">list</span><span class="p">[</span><span class="nb">int</span><span class="p">]],</span>
    <span class="n">ntp_ids</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Construct SQL query that gets, for each NTP, the fraction used by</span>
<span class="sd">    the mRNAs of each new gene, averages that per cell, and aggregate</span>
<span class="sd">    those fractions per variant into mean and std columns where each</span>
<span class="sd">    row is a 2D list with shape ``(# NTPs, # new genes)``.</span>

<span class="sd">    Args:</span>
<span class="sd">        sim_data: Simulation data</span>
<span class="sd">        new_gene_mRNA_idx: List of lists of mRNA indices for each new gene</span>
<span class="sd">        ntp_ids: IDs for NTPs in same order that they appear in</span>
<span class="sd">            ``sim_data.process.transcription.rna_data[&quot;counts_ACGU&quot;]``</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Determine number of NTPs per new gene mRNA and for all mRNAs</span>
    <span class="n">all_gene_mRNA_ACGU</span> <span class="o">=</span> <span class="n">sim_data</span><span class="o">.</span><span class="n">process</span><span class="o">.</span><span class="n">transcription</span><span class="o">.</span><span class="n">rna_data</span><span class="p">[</span>
        <span class="s2">&quot;counts_ACGU&quot;</span>
    <span class="p">]</span><span class="o">.</span><span class="n">asNumber</span><span class="p">()[</span><span class="n">sim_data</span><span class="o">.</span><span class="n">process</span><span class="o">.</span><span class="n">transcription</span><span class="o">.</span><span class="n">rna_data</span><span class="p">[</span><span class="s2">&quot;is_mRNA&quot;</span><span class="p">]]</span>

    <span class="c1"># Strip location tags from NTP IDs so they can be used in SQL</span>
    <span class="c1"># identifiers without quoting</span>
    <span class="n">ntp_ids</span> <span class="o">=</span> <span class="p">[</span><span class="n">ntp</span><span class="p">[:</span><span class="o">-</span><span class="mi">3</span><span class="p">]</span> <span class="k">for</span> <span class="n">ntp</span> <span class="ow">in</span> <span class="n">ntp_ids</span><span class="p">]</span>

    <span class="c1"># DuckDB list comprehension to calculate # of each NTP used by each mRNA</span>
    <span class="n">ntp_projections</span> <span class="o">=</span> <span class="s2">&quot;, &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
        <span class="sa">f</span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">        [count_ntp[1] * count_ntp[2] for count_ntp in</span>
<span class="s2">            list_zip(listeners__rna_counts__mRNA_counts,</span>
<span class="s2">                </span><span class="si">{</span><span class="n">all_gene_mRNA_ACGU</span><span class="p">[:,</span><span class="w"> </span><span class="n">ntp_idx</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span><span class="si">}</span><span class="s2">)] AS count_</span><span class="si">{</span><span class="n">ntp_id</span><span class="si">}</span>
<span class="s2">        &quot;&quot;&quot;</span>
        <span class="k">for</span> <span class="n">ntp_idx</span><span class="p">,</span> <span class="n">ntp_id</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">ntp_ids</span><span class="p">)</span>
    <span class="p">)</span>
    <span class="c1"># For each NTP, calculate fraction used in mRNAs for each new gene</span>
    <span class="n">ntp_frac_projections</span> <span class="o">=</span> <span class="s2">&quot;, &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
        <span class="s2">&quot;[&quot;</span>
        <span class="o">+</span> <span class="s2">&quot;, &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;list_sum(list_select(count_</span><span class="si">{</span><span class="n">ntp_id</span><span class="si">}</span><span class="s2">, </span><span class="si">{</span><span class="n">one_gene_idx</span><span class="si">}</span><span class="s2">)) / &quot;</span>
            <span class="sa">f</span><span class="s2">&quot;list_sum(count_</span><span class="si">{</span><span class="n">ntp_id</span><span class="si">}</span><span class="s2">)&quot;</span>
            <span class="k">for</span> <span class="n">one_gene_idx</span> <span class="ow">in</span> <span class="n">new_gene_mRNA_idx</span>
        <span class="p">)</span>
        <span class="o">+</span> <span class="sa">f</span><span class="s2">&quot;] AS frac_</span><span class="si">{</span><span class="n">ntp_id</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="k">for</span> <span class="n">ntp_id</span> <span class="ow">in</span> <span class="n">ntp_ids</span>
    <span class="p">)</span>
    <span class="c1"># Unnest average NTP fraction per mRNA</span>
    <span class="n">unnested_frac_projections</span> <span class="o">=</span> <span class="p">(</span>
        <span class="s2">&quot;, &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
            <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;unnest(frac_</span><span class="si">{</span><span class="n">ntp_id</span><span class="si">}</span><span class="s2">) AS unnested_</span><span class="si">{</span><span class="n">ntp_id</span><span class="si">}</span><span class="s2">_frac&quot;</span> <span class="k">for</span> <span class="n">ntp_id</span> <span class="ow">in</span> <span class="n">ntp_ids</span><span class="p">]</span>
        <span class="p">)</span>
        <span class="o">+</span> <span class="sa">f</span><span class="s2">&quot;, generate_subscripts(frac_</span><span class="si">{</span><span class="n">ntp_ids</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2">, 1) AS gene_idx&quot;</span>
    <span class="p">)</span>
    <span class="c1"># Average NTP fraction per new gene first per cell, then per variant</span>
    <span class="n">cell_avg_frac_projections</span> <span class="o">=</span> <span class="s2">&quot;, &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
        <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;avg(unnested_</span><span class="si">{</span><span class="n">ntp_id</span><span class="si">}</span><span class="s2">_frac) AS avg_</span><span class="si">{</span><span class="n">ntp_id</span><span class="si">}</span><span class="s2">_frac&quot;</span> <span class="k">for</span> <span class="n">ntp_id</span> <span class="ow">in</span> <span class="n">ntp_ids</span><span class="p">]</span>
    <span class="p">)</span>
    <span class="n">variant_avg_frac_projections</span> <span class="o">=</span> <span class="s2">&quot;, &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
        <span class="p">[</span>
            <span class="sa">f</span><span class="s2">&quot;avg(avg_</span><span class="si">{</span><span class="n">ntp_id</span><span class="si">}</span><span class="s2">_frac) AS avg_</span><span class="si">{</span><span class="n">ntp_id</span><span class="si">}</span><span class="s2">_frac, &quot;</span>
            <span class="sa">f</span><span class="s2">&quot;stddev(avg_</span><span class="si">{</span><span class="n">ntp_id</span><span class="si">}</span><span class="s2">_frac) AS std_</span><span class="si">{</span><span class="n">ntp_id</span><span class="si">}</span><span class="s2">_frac&quot;</span>
            <span class="k">for</span> <span class="n">ntp_id</span> <span class="ow">in</span> <span class="n">ntp_ids</span>
        <span class="p">]</span>
    <span class="p">)</span>
    <span class="c1"># Compile new gene average NTP fractions into 2D list columns</span>
    <span class="c1"># with shape (# of NTPs, # of new genes)</span>
    <span class="n">list_avg_frac_projections</span> <span class="o">=</span> <span class="p">(</span>
        <span class="s2">&quot;[&quot;</span>
        <span class="o">+</span> <span class="s2">&quot;, &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
            <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;list(avg_</span><span class="si">{</span><span class="n">ntp_id</span><span class="si">}</span><span class="s2">_frac ORDER BY gene_idx)&quot;</span> <span class="k">for</span> <span class="n">ntp_id</span> <span class="ow">in</span> <span class="n">ntp_ids</span><span class="p">]</span>
        <span class="p">)</span>
        <span class="o">+</span> <span class="s2">&quot;] AS mean&quot;</span>
    <span class="p">)</span>
    <span class="n">list_std_frac_projections</span> <span class="o">=</span> <span class="p">(</span>
        <span class="s2">&quot;[&quot;</span>
        <span class="o">+</span> <span class="s2">&quot;, &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
            <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;list(std_</span><span class="si">{</span><span class="n">ntp_id</span><span class="si">}</span><span class="s2">_frac ORDER BY gene_idx)&quot;</span> <span class="k">for</span> <span class="n">ntp_id</span> <span class="ow">in</span> <span class="n">ntp_ids</span><span class="p">]</span>
        <span class="p">)</span>
        <span class="o">+</span> <span class="s2">&quot;] AS std&quot;</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="sa">f</span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">        WITH ntp_counts AS (</span>
<span class="s2">            SELECT </span><span class="si">{</span><span class="n">ntp_projections</span><span class="si">}</span><span class="s2">, experiment_id, variant, lineage_seed,</span>
<span class="s2">                generation, agent_id</span>
<span class="s2">            FROM (</span><span class="se">{{</span><span class="s2">subquery</span><span class="se">}}</span><span class="s2">)</span>
<span class="s2">        ),</span>
<span class="s2">        ntp_frac AS (</span>
<span class="s2">            SELECT </span><span class="si">{</span><span class="n">ntp_frac_projections</span><span class="si">}</span><span class="s2">, experiment_id, variant, lineage_seed,</span>
<span class="s2">                generation, agent_id</span>
<span class="s2">            FROM ntp_counts</span>
<span class="s2">        ),</span>
<span class="s2">        unnested_frac AS (</span>
<span class="s2">            SELECT </span><span class="si">{</span><span class="n">unnested_frac_projections</span><span class="si">}</span><span class="s2">, experiment_id, variant, lineage_seed,</span>
<span class="s2">                generation, agent_id</span>
<span class="s2">            FROM ntp_frac</span>
<span class="s2">        ),</span>
<span class="s2">        avg_per_cell AS (</span>
<span class="s2">            SELECT </span><span class="si">{</span><span class="n">cell_avg_frac_projections</span><span class="si">}</span><span class="s2">,</span>
<span class="s2">                experiment_id, variant, lineage_seed,</span>
<span class="s2">                generation, agent_id, gene_idx</span>
<span class="s2">            FROM unnested_frac</span>
<span class="s2">            GROUP BY experiment_id, variant, lineage_seed,</span>
<span class="s2">                generation, agent_id, gene_idx</span>
<span class="s2">        ),</span>
<span class="s2">        avg_per_variant AS (</span>
<span class="s2">            SELECT experiment_id, variant, gene_idx, </span><span class="si">{</span><span class="n">variant_avg_frac_projections</span><span class="si">}</span>
<span class="s2">            FROM avg_per_cell</span>
<span class="s2">            GROUP BY experiment_id, variant, gene_idx</span>
<span class="s2">        )</span>
<span class="s2">        SELECT variant, </span><span class="si">{</span><span class="n">list_avg_frac_projections</span><span class="si">}</span><span class="s2">, </span><span class="si">{</span><span class="n">list_std_frac_projections</span><span class="si">}</span>
<span class="s2">        FROM avg_per_variant</span>
<span class="s2">        GROUP BY experiment_id, variant</span>
<span class="s2">        &quot;&quot;&quot;</span></div>



<div class="viewcode-block" id="avg_ratio_of_1d_arrays_sql">
<a class="viewcode-back" href="../../../../reference/api/ecoli/ecoli.analysis.multivariant.new_gene_translation_efficiency_heatmaps.html#ecoli.analysis.multivariant.new_gene_translation_efficiency_heatmaps.avg_ratio_of_1d_arrays_sql">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">avg_ratio_of_1d_arrays_sql</span><span class="p">(</span><span class="n">numerator</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">denominator</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Create generic SQL query that calculates the average per cell of each</span>
<span class="sd">    element in two 1D list columns divided elementwise and aggregates those</span>
<span class="sd">    ratios per variant into mean and std columns.</span>

<span class="sd">    .. note::</span>
<span class="sd">        Time steps with 0 in the denominator are assigned a ratio of 0.</span>

<span class="sd">    Args:</span>
<span class="sd">        numerator: Name of 1D list column that will be numerator in ratio</span>
<span class="sd">        denominator: Name of 1D list column that will be denominator in ratio</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="sa">f</span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">        WITH unnested_data AS (</span>
<span class="s2">            SELECT unnest(</span><span class="si">{</span><span class="n">denominator</span><span class="si">}</span><span class="s2">) AS denominator,</span>
<span class="s2">                unnest(</span><span class="si">{</span><span class="n">numerator</span><span class="si">}</span><span class="s2">) AS numerator,</span>
<span class="s2">                generate_subscripts(</span><span class="si">{</span><span class="n">numerator</span><span class="si">}</span><span class="s2">, 1) AS list_idx,</span>
<span class="s2">                experiment_id, variant, lineage_seed, generation, agent_id</span>
<span class="s2">            FROM (</span><span class="se">{{</span><span class="s2">subquery</span><span class="se">}}</span><span class="s2">)</span>
<span class="s2">        ),</span>
<span class="s2">        ratio_avg_per_cell AS (</span>
<span class="s2">            SELECT avg(</span>
<span class="s2">                CASE</span>
<span class="s2">                    WHEN denominator = 0 THEN 0</span>
<span class="s2">                    ELSE numerator / denominator</span>
<span class="s2">                END</span>
<span class="s2">            ) AS ratio_avg, experiment_id, variant, list_idx</span>
<span class="s2">            FROM unnested_data</span>
<span class="s2">            GROUP BY experiment_id, variant, lineage_seed, generation, agent_id, list_idx</span>
<span class="s2">        ),</span>
<span class="s2">        ratio_avg_per_variant AS (</span>
<span class="s2">            SELECT avg(ratio_avg) AS ratio_avg, stddev(ratio_avg) AS ratio_std,</span>
<span class="s2">                list_idx, experiment_id, variant</span>
<span class="s2">            FROM ratio_avg_per_cell</span>
<span class="s2">            GROUP BY experiment_id, variant, list_idx</span>
<span class="s2">        )</span>
<span class="s2">        SELECT variant, list(ratio_avg ORDER BY list_idx) AS mean,</span>
<span class="s2">            list(ratio_std ORDER BY list_idx) AS std,</span>
<span class="s2">        FROM ratio_avg_per_variant</span>
<span class="s2">        GROUP BY experiment_id, variant</span>
<span class="s2">        &quot;&quot;&quot;</span></div>



<div class="viewcode-block" id="avg_1d_array_sql">
<a class="viewcode-back" href="../../../../reference/api/ecoli/ecoli.analysis.multivariant.new_gene_translation_efficiency_heatmaps.html#ecoli.analysis.multivariant.new_gene_translation_efficiency_heatmaps.avg_1d_array_sql">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">avg_1d_array_sql</span><span class="p">(</span><span class="n">column</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Create generic SQL query that calculates the average per cell of</span>
<span class="sd">    each element in a 1D array column and aggregates that per variant</span>
<span class="sd">    into mean and std columns.</span>

<span class="sd">    Args:</span>
<span class="sd">        column: Name of 1D list column to aggregate</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="sa">f</span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">        WITH unnested_counts AS (</span>
<span class="s2">            SELECT unnest(</span><span class="si">{</span><span class="n">column</span><span class="si">}</span><span class="s2">) AS array_col,</span>
<span class="s2">                generate_subscripts(</span><span class="si">{</span><span class="n">column</span><span class="si">}</span><span class="s2">, 1) AS array_idx,</span>
<span class="s2">                experiment_id, variant, lineage_seed, generation, agent_id</span>
<span class="s2">            FROM (</span><span class="se">{{</span><span class="s2">subquery</span><span class="se">}}</span><span class="s2">)</span>
<span class="s2">        ),</span>
<span class="s2">        avg_per_cell AS (</span>
<span class="s2">            SELECT avg(array_col) AS avg_array_col,</span>
<span class="s2">                experiment_id, variant, lineage_seed,</span>
<span class="s2">                generation, agent_id, array_idx</span>
<span class="s2">            FROM unnested_counts</span>
<span class="s2">            GROUP BY experiment_id, variant, lineage_seed,</span>
<span class="s2">                generation, agent_id, array_idx</span>
<span class="s2">        ),</span>
<span class="s2">        avg_per_variant AS (</span>
<span class="s2">            SELECT avg(avg_array_col) AS avg_array_col,</span>
<span class="s2">                stddev(avg_array_col) AS std_array_col,</span>
<span class="s2">                experiment_id, variant, array_idx</span>
<span class="s2">            FROM avg_per_cell</span>
<span class="s2">            GROUP BY experiment_id, variant, array_idx</span>
<span class="s2">        )</span>
<span class="s2">        SELECT variant, list(avg_array_col ORDER BY array_idx) AS mean,</span>
<span class="s2">            list(std_array_col ORDER BY array_idx) AS std</span>
<span class="s2">        FROM avg_per_variant</span>
<span class="s2">        GROUP BY experiment_id, variant</span>
<span class="s2">        &quot;&quot;&quot;</span></div>



<div class="viewcode-block" id="avg_sum_1d_array_sql">
<a class="viewcode-back" href="../../../../reference/api/ecoli/ecoli.analysis.multivariant.new_gene_translation_efficiency_heatmaps.html#ecoli.analysis.multivariant.new_gene_translation_efficiency_heatmaps.avg_sum_1d_array_sql">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">avg_sum_1d_array_sql</span><span class="p">(</span><span class="n">column</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Create generic SQL query that calculates the average per cell of</span>
<span class="sd">    the sum of elements in a 1D array column and</span>
<span class="sd">    aggregates that per variant into mean and std columns.</span>

<span class="sd">    Args:</span>
<span class="sd">        column: Name of 1D list column to aggregate</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="sa">f</span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">        WITH avg_per_cell AS (</span>
<span class="s2">            SELECT avg(list_sum(</span><span class="si">{</span><span class="n">column</span><span class="si">}</span><span class="s2">)) AS avg_sum,</span>
<span class="s2">                experiment_id, variant, lineage_seed,</span>
<span class="s2">                generation, agent_id</span>
<span class="s2">            FROM (</span><span class="se">{{</span><span class="s2">subquery</span><span class="se">}}</span><span class="s2">)</span>
<span class="s2">            GROUP BY experiment_id, variant, lineage_seed,</span>
<span class="s2">                generation, agent_id</span>
<span class="s2">        )</span>
<span class="s2">        SELECT variant, avg(avg_sum) AS mean,</span>
<span class="s2">            stddev(avg_sum) AS std</span>
<span class="s2">        FROM avg_per_cell</span>
<span class="s2">        GROUP BY experiment_id, variant</span>
<span class="s2">        &quot;&quot;&quot;</span></div>



<div class="viewcode-block" id="avg_1d_array_over_scalar_sql">
<a class="viewcode-back" href="../../../../reference/api/ecoli/ecoli.analysis.multivariant.new_gene_translation_efficiency_heatmaps.html#ecoli.analysis.multivariant.new_gene_translation_efficiency_heatmaps.avg_1d_array_over_scalar_sql">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">avg_1d_array_over_scalar_sql</span><span class="p">(</span><span class="n">array_column</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">scalar_column</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Create generic SQL query that calculates the average per cell of</span>
<span class="sd">    each element in a 1D array column divided by a scalar column, and</span>
<span class="sd">    aggregates those ratios per variant into mean and std columns.</span>

<span class="sd">    .. note::</span>
<span class="sd">        Time steps with 0 in the scalar column are assigned a ratio of 0.</span>

<span class="sd">    Args:</span>
<span class="sd">        array_column: Name of 1D list column to aggregate</span>
<span class="sd">        scalar_column: Name of scalar column to divide ``array_column``</span>
<span class="sd">            cell averages by</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="sa">f</span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">        WITH unnested_counts AS (</span>
<span class="s2">            SELECT unnest(</span><span class="si">{</span><span class="n">array_column</span><span class="si">}</span><span class="s2">) AS array_col,</span>
<span class="s2">                generate_subscripts(</span><span class="si">{</span><span class="n">array_column</span><span class="si">}</span><span class="s2">, 1) AS array_idx,</span>
<span class="s2">                </span><span class="si">{</span><span class="n">scalar_column</span><span class="si">}</span><span class="s2"> AS scalar_col,</span>
<span class="s2">                experiment_id, variant, lineage_seed, generation, agent_id</span>
<span class="s2">            FROM (</span><span class="se">{{</span><span class="s2">subquery</span><span class="se">}}</span><span class="s2">)</span>
<span class="s2">        ),</span>
<span class="s2">        avg_per_cell AS (</span>
<span class="s2">            SELECT avg(</span>
<span class="s2">                CASE</span>
<span class="s2">                    WHEN scalar_col = 0 THEN 0</span>
<span class="s2">                    ELSE array_col / scalar_col</span>
<span class="s2">                END) AS avg_ratio,</span>
<span class="s2">                experiment_id, variant, lineage_seed,</span>
<span class="s2">                generation, agent_id, array_idx</span>
<span class="s2">            FROM unnested_counts</span>
<span class="s2">            GROUP BY experiment_id, variant, lineage_seed,</span>
<span class="s2">                generation, agent_id, array_idx</span>
<span class="s2">        ),</span>
<span class="s2">        avg_per_variant AS (</span>
<span class="s2">            SELECT avg(avg_ratio) AS avg_ratio, stddev(avg_ratio) AS std_ratio,</span>
<span class="s2">                experiment_id, variant, array_idx</span>
<span class="s2">            FROM avg_per_cell</span>
<span class="s2">            GROUP BY experiment_id, variant, array_idx</span>
<span class="s2">        )</span>
<span class="s2">        SELECT variant, list(avg_ratio ORDER BY array_idx) AS mean,</span>
<span class="s2">            list(std_ratio ORDER BY array_idx) AS std</span>
<span class="s2">        FROM avg_per_variant</span>
<span class="s2">        GROUP BY experiment_id, variant</span>
<span class="s2">        &quot;&quot;&quot;</span></div>



<div class="viewcode-block" id="avg_sum_1d_array_over_scalar_sql">
<a class="viewcode-back" href="../../../../reference/api/ecoli/ecoli.analysis.multivariant.new_gene_translation_efficiency_heatmaps.html#ecoli.analysis.multivariant.new_gene_translation_efficiency_heatmaps.avg_sum_1d_array_over_scalar_sql">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">avg_sum_1d_array_over_scalar_sql</span><span class="p">(</span><span class="n">array_column</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">scalar_column</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Create generic SQL query that calculates the average per cell of</span>
<span class="sd">    the sum of elements in a 1D array column divided by a scalar column, and</span>
<span class="sd">    aggregates those ratios per variant as mean and std columns.</span>

<span class="sd">    .. note::</span>
<span class="sd">        Time steps with 0 in the scalar column are assigned a ratio of 0.</span>

<span class="sd">    Args:</span>
<span class="sd">        array_column: Name of 1D list column to aggregate</span>
<span class="sd">        scalar_column: Name of scalar column to divide ``array_column``</span>
<span class="sd">            cell averages by</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="sa">f</span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">        WITH avg_per_cell AS (</span>
<span class="s2">            SELECT avg(</span>
<span class="s2">                CASE</span>
<span class="s2">                    WHEN </span><span class="si">{</span><span class="n">scalar_column</span><span class="si">}</span><span class="s2"> = 0 THEN 0</span>
<span class="s2">                    ELSE list_sum(</span><span class="si">{</span><span class="n">array_column</span><span class="si">}</span><span class="s2">) / </span><span class="si">{</span><span class="n">scalar_column</span><span class="si">}</span>
<span class="s2">                END) AS avg_ratio,</span>
<span class="s2">                experiment_id, variant, lineage_seed,</span>
<span class="s2">                generation, agent_id</span>
<span class="s2">            FROM (</span><span class="se">{{</span><span class="s2">subquery</span><span class="se">}}</span><span class="s2">)</span>
<span class="s2">            GROUP BY experiment_id, variant, lineage_seed,</span>
<span class="s2">                generation, agent_id</span>
<span class="s2">        )</span>
<span class="s2">        SELECT variant, avg(avg_ratio) AS mean, stddev(avg_ratio) AS std</span>
<span class="s2">        FROM avg_per_cell</span>
<span class="s2">        GROUP BY experiment_id, variant</span>
<span class="s2">        &quot;&quot;&quot;</span></div>



<div class="viewcode-block" id="get_rnap_counts_projection">
<a class="viewcode-back" href="../../../../reference/api/ecoli/ecoli.analysis.multivariant.new_gene_translation_efficiency_heatmaps.html#ecoli.analysis.multivariant.new_gene_translation_efficiency_heatmaps.get_rnap_counts_projection">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">get_rnap_counts_projection</span><span class="p">(</span>
    <span class="n">sim_data</span><span class="p">:</span> <span class="s2">&quot;SimulationDataEcoli&quot;</span><span class="p">,</span> <span class="n">bulk_ids</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Return SQL projection to selectively read bulk inactive RNAP count.</span>

<span class="sd">    Args:</span>
<span class="sd">        sim_data: Simulation data</span>
<span class="sd">        bulk_ids: List of all bulk IDs in order</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">rnap_idx</span> <span class="o">=</span> <span class="n">bulk_name_to_idx</span><span class="p">(</span><span class="n">sim_data</span><span class="o">.</span><span class="n">molecule_ids</span><span class="o">.</span><span class="n">full_RNAP</span><span class="p">,</span> <span class="n">bulk_ids</span><span class="p">)</span>
    <span class="k">return</span> <span class="sa">f</span><span class="s2">&quot;bulk[</span><span class="si">{</span><span class="n">rnap_idx</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="mi">1</span><span class="si">}</span><span class="s2">] AS bulk&quot;</span></div>



<div class="viewcode-block" id="get_ribosome_counts_projection">
<a class="viewcode-back" href="../../../../reference/api/ecoli/ecoli.analysis.multivariant.new_gene_translation_efficiency_heatmaps.html#ecoli.analysis.multivariant.new_gene_translation_efficiency_heatmaps.get_ribosome_counts_projection">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">get_ribosome_counts_projection</span><span class="p">(</span>
    <span class="n">sim_data</span><span class="p">:</span> <span class="s2">&quot;SimulationDataEcoli&quot;</span><span class="p">,</span> <span class="n">bulk_ids</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Return SQL projection to selectively read bulk inactive ribosome count</span>
<span class="sd">    (defined as minimum of free 30S and 50S subunits at any given moment)</span>

<span class="sd">    Args:</span>
<span class="sd">        sim_data: Simulation data</span>
<span class="sd">        bulk_ids: List of all bulk IDs in order</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">ribosome_idx</span> <span class="o">=</span> <span class="n">cast</span><span class="p">(</span>
        <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
        <span class="n">bulk_name_to_idx</span><span class="p">(</span>
            <span class="p">[</span>
                <span class="n">sim_data</span><span class="o">.</span><span class="n">molecule_ids</span><span class="o">.</span><span class="n">s30_full_complex</span><span class="p">,</span>
                <span class="n">sim_data</span><span class="o">.</span><span class="n">molecule_ids</span><span class="o">.</span><span class="n">s50_full_complex</span><span class="p">,</span>
            <span class="p">],</span>
            <span class="n">bulk_ids</span><span class="p">,</span>
        <span class="p">),</span>
    <span class="p">)</span>
    <span class="k">return</span> <span class="sa">f</span><span class="s2">&quot;least(bulk[</span><span class="si">{</span><span class="n">ribosome_idx</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="mi">1</span><span class="si">}</span><span class="s2">], bulk[</span><span class="si">{</span><span class="n">ribosome_idx</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="mi">1</span><span class="si">}</span><span class="s2">]) AS bulk&quot;</span></div>



<div class="viewcode-block" id="get_overcrowding_sql">
<a class="viewcode-back" href="../../../../reference/api/ecoli/ecoli.analysis.multivariant.new_gene_translation_efficiency_heatmaps.html#ecoli.analysis.multivariant.new_gene_translation_efficiency_heatmaps.get_overcrowding_sql">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">get_overcrowding_sql</span><span class="p">(</span><span class="n">target_col</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">actual_col</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Create generic SQL query that calculates for average number of genes</span>
<span class="sd">    that are overcrowded per time step for each cell, then aggregates that</span>
<span class="sd">    per variant into mean and std columns.</span>

<span class="sd">    At every time step, if the element in ``target_col`` is greater than the</span>
<span class="sd">    corresponding element in ``actual_col``, we say that the gene for that</span>
<span class="sd">    element is overcrowded. We average the number of overcrowded genes over</span>
<span class="sd">    all the time steps for each cell. Then, we average the per-cell averages</span>
<span class="sd">    over all cells in each variant.</span>

<span class="sd">    Args:</span>
<span class="sd">        target_col: Name of 1D list column with target values</span>
<span class="sd">        actual_col: Name of 1D list column with actual values.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="sa">f</span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">        WITH avg_per_cell AS (</span>
<span class="s2">            SELECT avg(list_sum([(i[1] &gt; i[2])::UTINYINT</span>
<span class="s2">                for i in list_zip(</span><span class="si">{</span><span class="n">target_col</span><span class="si">}</span><span class="s2">, </span><span class="si">{</span><span class="n">actual_col</span><span class="si">}</span><span class="s2">)])) AS overcrowded,</span>
<span class="s2">                experiment_id, variant, lineage_seed, generation, agent_id</span>
<span class="s2">            FROM (</span><span class="se">{{</span><span class="s2">subquery</span><span class="se">}}</span><span class="s2">)</span>
<span class="s2">            GROUP BY experiment_id, variant, lineage_seed, generation, agent_id</span>
<span class="s2">        )</span>
<span class="s2">        SELECT variant, avg(overcrowded) AS mean, stddev(overcrowded) AS std,</span>
<span class="s2">        FROM avg_per_cell</span>
<span class="s2">        GROUP BY experiment_id, variant</span>
<span class="s2">        &quot;&quot;&quot;</span></div>



<div class="viewcode-block" id="get_rnas_combined_as_genes_projection">
<a class="viewcode-back" href="../../../../reference/api/ecoli/ecoli.analysis.multivariant.new_gene_translation_efficiency_heatmaps.html#ecoli.analysis.multivariant.new_gene_translation_efficiency_heatmaps.get_rnas_combined_as_genes_projection">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">get_rnas_combined_as_genes_projection</span><span class="p">(</span>
    <span class="n">column</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">rna_idx</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">list</span><span class="p">[</span><span class="nb">int</span><span class="p">]],</span> <span class="n">name</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">cast_type</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Create generic SQL projection that evaluates to a list column</span>
<span class="sd">    where each element is the sum of a subset of elements from the</span>
<span class="sd">    original list column. This is mainly used to sum up all RNA</span>
<span class="sd">    data that corresponds to a single gene / cistron / monomer.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">cast_type</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">cast_type</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;::</span><span class="si">{</span><span class="n">cast_type</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">cast_type</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
    <span class="n">sum_projections</span> <span class="o">=</span> <span class="s2">&quot;, &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
        <span class="p">[</span>
            <span class="sa">f</span><span class="s2">&quot;list_sum(list_select(</span><span class="si">{</span><span class="n">column</span><span class="si">}</span><span class="s2">, </span><span class="si">{</span><span class="n">rna_idx_for_one_gene</span><span class="si">}</span><span class="s2">)</span><span class="si">{</span><span class="n">cast_type</span><span class="si">}</span><span class="s2">)&quot;</span>
            <span class="k">for</span> <span class="n">rna_idx_for_one_gene</span> <span class="ow">in</span> <span class="n">rna_idx</span>
        <span class="p">]</span>
    <span class="p">)</span>
    <span class="k">return</span> <span class="sa">f</span><span class="s2">&quot;[</span><span class="si">{</span><span class="n">sum_projections</span><span class="si">}</span><span class="s2">] AS </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span></div>



<div class="viewcode-block" id="get_variant_mask">
<a class="viewcode-back" href="../../../../reference/api/ecoli/ecoli.analysis.multivariant.new_gene_translation_efficiency_heatmaps.html#ecoli.analysis.multivariant.new_gene_translation_efficiency_heatmaps.get_variant_mask">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">get_variant_mask</span><span class="p">(</span>
    <span class="n">conn</span><span class="p">:</span> <span class="n">DuckDBPyConnection</span><span class="p">,</span>
    <span class="n">config_sql</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">variant_to_row_col</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">tuple</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">]],</span>
    <span class="n">variant_matrix_shape</span><span class="p">:</span> <span class="nb">tuple</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">],</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">bool_</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Get a boolean matrix where the rows represent the different</span>
<span class="sd">    translation efficiencies and the columns represent the different</span>
<span class="sd">    expression factors that were used to create variants. The matrix</span>
<span class="sd">    is True for each combination that was actually simulated and</span>
<span class="sd">    False otherwise.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">variants</span> <span class="o">=</span> <span class="n">conn</span><span class="o">.</span><span class="n">sql</span><span class="p">(</span>
        <span class="sa">f</span><span class="s2">&quot;SELECT DISTINCT variant AS variant FROM (</span><span class="si">{</span><span class="n">config_sql</span><span class="si">}</span><span class="s2">)&quot;</span>
    <span class="p">)</span><span class="o">.</span><span class="n">fetchnumpy</span><span class="p">()[</span><span class="s2">&quot;variant&quot;</span><span class="p">]</span>
    <span class="n">variant_mask</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">variant_matrix_shape</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">bool_</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">variant</span> <span class="ow">in</span> <span class="n">variants</span><span class="p">:</span>
        <span class="n">variant_mask</span><span class="p">[</span><span class="o">*</span><span class="n">variant_to_row_col</span><span class="p">[</span><span class="n">variant</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">True</span>
    <span class="k">return</span> <span class="n">variant_mask</span></div>



<div class="viewcode-block" id="plot_heatmaps">
<a class="viewcode-back" href="../../../../reference/api/ecoli/ecoli.analysis.multivariant.new_gene_translation_efficiency_heatmaps.html#ecoli.analysis.multivariant.new_gene_translation_efficiency_heatmaps.plot_heatmaps">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">plot_heatmaps</span><span class="p">(</span>
    <span class="n">heatmap_data</span><span class="p">,</span>
    <span class="n">heatmap_details</span><span class="p">,</span>
    <span class="n">new_gene_cistron_ids</span><span class="p">,</span>
    <span class="n">ntp_ids</span><span class="p">,</span>
    <span class="n">capacity_gene_common_names</span><span class="p">,</span>
    <span class="n">total_heatmaps_to_make</span><span class="p">,</span>
    <span class="n">is_dashboard</span><span class="p">,</span>
    <span class="n">variant_mask</span><span class="p">,</span>
    <span class="n">heatmap_x_label</span><span class="p">,</span>
    <span class="n">heatmap_y_label</span><span class="p">,</span>
    <span class="n">new_gene_expression_factors</span><span class="p">,</span>
    <span class="n">new_gene_translation_efficiency_values</span><span class="p">,</span>
    <span class="n">summary_statistic</span><span class="p">,</span>
    <span class="n">figsize_x</span><span class="p">,</span>
    <span class="n">figsize_y</span><span class="p">,</span>
    <span class="n">plotOutDir</span><span class="p">,</span>
    <span class="n">plot_suffix</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Plots all heatmaps in order given by HEATMAPS_TO_MAKE_LIST.</span>

<span class="sd">    Args:</span>
<span class="sd">        is_dashboard: Boolean flag for whether we are creating a dashboard</span>
<span class="sd">            of heatmaps or a number of individual heatmaps</span>
<span class="sd">        variant_mask: np.array of dimension</span>
<span class="sd">            (len(new_gene_translation_efficiency_values),</span>
<span class="sd">            len(new_gene_expression_factors)) with entries set to True if</span>
<span class="sd">            variant was run, False otherwise.</span>
<span class="sd">        heatmap_x_label: Label for x axis of heatmap</span>
<span class="sd">        heatmap_y_label: Label for y axis of heatmap</span>
<span class="sd">        new_gene_expression_factors: New gene expression factors used in</span>
<span class="sd">            these variants</span>
<span class="sd">        new_gene_translation_efficiency_values: New gene translation</span>
<span class="sd">            efficiency values used in these variants</span>
<span class="sd">        summary_statistic: Specifies whether average (mean) or</span>
<span class="sd">            standard deviation (std_dev) should be displayed on the</span>
<span class="sd">            heatmaps</span>
<span class="sd">        figsize_x: Horizontal size of each heatmap</span>
<span class="sd">        figsize_y: Vertical size of each heatmap</span>
<span class="sd">        plotOutDir: Output directory for plots</span>
<span class="sd">        plot_suffix: Suffix to add to plot file names, usually specifying</span>
<span class="sd">            which generations were plotted</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">summary_statistic</span> <span class="o">==</span> <span class="s2">&quot;std_dev&quot;</span><span class="p">:</span>
        <span class="n">plot_suffix</span> <span class="o">=</span> <span class="n">plot_suffix</span> <span class="o">+</span> <span class="s2">&quot;_std_dev&quot;</span>
    <span class="k">elif</span> <span class="n">summary_statistic</span> <span class="o">!=</span> <span class="s2">&quot;mean&quot;</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span>
            <span class="s2">&quot;mean and std_dev are the only currently supported summary statistics&quot;</span>
        <span class="p">)</span>

    <span class="k">if</span> <span class="n">is_dashboard</span><span class="p">:</span>
        <span class="c1"># Determine dashboard layout</span>
        <span class="k">if</span> <span class="n">total_heatmaps_to_make</span> <span class="o">&gt;</span> <span class="mi">3</span><span class="p">:</span>
            <span class="n">dashboard_ncols</span> <span class="o">=</span> <span class="mi">4</span>
            <span class="n">dashboard_nrows</span> <span class="o">=</span> <span class="n">math</span><span class="o">.</span><span class="n">ceil</span><span class="p">((</span><span class="n">total_heatmaps_to_make</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="o">/</span> <span class="n">dashboard_ncols</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">dashboard_ncols</span> <span class="o">=</span> <span class="n">total_heatmaps_to_make</span> <span class="o">+</span> <span class="mi">1</span>
            <span class="n">dashboard_nrows</span> <span class="o">=</span> <span class="mi">1</span>
        <span class="n">fig</span><span class="p">,</span> <span class="n">axs</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span>
            <span class="n">nrows</span><span class="o">=</span><span class="n">dashboard_nrows</span><span class="p">,</span>
            <span class="n">ncols</span><span class="o">=</span><span class="n">dashboard_ncols</span><span class="p">,</span>
            <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="n">figsize_y</span> <span class="o">*</span> <span class="n">dashboard_ncols</span><span class="p">,</span> <span class="n">figsize_x</span> <span class="o">*</span> <span class="n">dashboard_nrows</span><span class="p">),</span>
            <span class="n">layout</span><span class="o">=</span><span class="s2">&quot;constrained&quot;</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="k">if</span> <span class="n">dashboard_nrows</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
            <span class="n">axs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="n">axs</span><span class="p">,</span> <span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">dashboard_ncols</span><span class="p">))</span>
        <span class="n">row_ax</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">col_ax</span> <span class="o">=</span> <span class="mi">0</span>

    <span class="k">for</span> <span class="n">h</span> <span class="ow">in</span> <span class="n">HEATMAPS_TO_MAKE_LIST</span><span class="p">:</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">heatmap_details</span><span class="p">[</span><span class="n">h</span><span class="p">][</span><span class="s2">&quot;is_nonstandard_plot&quot;</span><span class="p">]:</span>
            <span class="n">stop_index</span> <span class="o">=</span> <span class="mi">1</span>
            <span class="n">title_addition</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
            <span class="n">filename_addition</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
            <span class="k">if</span> <span class="n">heatmap_details</span><span class="p">[</span><span class="n">h</span><span class="p">][</span><span class="s2">&quot;is_new_gene_heatmap&quot;</span><span class="p">]:</span>
                <span class="n">stop_index</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">new_gene_cistron_ids</span><span class="p">)</span>
            <span class="k">elif</span> <span class="n">heatmap_details</span><span class="p">[</span><span class="n">h</span><span class="p">][</span><span class="s2">&quot;is_capacity_gene_heatmap&quot;</span><span class="p">]:</span>
                <span class="n">stop_index</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">capacity_gene_common_names</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">stop_index</span><span class="p">):</span>
                <span class="k">if</span> <span class="n">heatmap_details</span><span class="p">[</span><span class="n">h</span><span class="p">][</span><span class="s2">&quot;is_new_gene_heatmap&quot;</span><span class="p">]:</span>
                    <span class="n">title_addition</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;: </span><span class="si">{</span><span class="n">new_gene_cistron_ids</span><span class="p">[</span><span class="n">i</span><span class="p">][:</span><span class="o">-</span><span class="mi">4</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span>
                    <span class="n">filename_addition</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;_</span><span class="si">{</span><span class="n">new_gene_cistron_ids</span><span class="p">[</span><span class="n">i</span><span class="p">][:</span><span class="o">-</span><span class="mi">4</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span>
                    <span class="n">curr_heatmap_data</span> <span class="o">=</span> <span class="n">heatmap_data</span><span class="p">[</span><span class="n">h</span><span class="p">][</span><span class="n">summary_statistic</span><span class="p">][:,</span> <span class="p">:,</span> <span class="n">i</span><span class="p">]</span>
                <span class="k">elif</span> <span class="n">heatmap_details</span><span class="p">[</span><span class="n">h</span><span class="p">][</span><span class="s2">&quot;is_capacity_gene_heatmap&quot;</span><span class="p">]:</span>
                    <span class="n">title_addition</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;: </span><span class="si">{</span><span class="n">capacity_gene_common_names</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span>
                    <span class="n">filename_addition</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;_</span><span class="si">{</span><span class="n">capacity_gene_common_names</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span>
                    <span class="n">curr_heatmap_data</span> <span class="o">=</span> <span class="n">heatmap_data</span><span class="p">[</span><span class="n">h</span><span class="p">][</span><span class="n">summary_statistic</span><span class="p">][:,</span> <span class="p">:,</span> <span class="n">i</span><span class="p">]</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">curr_heatmap_data</span> <span class="o">=</span> <span class="n">heatmap_data</span><span class="p">[</span><span class="n">h</span><span class="p">][</span><span class="n">summary_statistic</span><span class="p">]</span>
                <span class="n">title</span> <span class="o">=</span> <span class="n">heatmap_details</span><span class="p">[</span><span class="n">h</span><span class="p">][</span><span class="s2">&quot;plot_title&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="n">title_addition</span>
                <span class="k">if</span> <span class="n">summary_statistic</span> <span class="o">==</span> <span class="s2">&quot;std_dev&quot;</span><span class="p">:</span>
                    <span class="n">title</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;Std Dev: </span><span class="si">{</span><span class="n">title</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="k">if</span> <span class="n">is_dashboard</span><span class="p">:</span>
                    <span class="n">curr_ax</span> <span class="o">=</span> <span class="n">axs</span><span class="p">[</span><span class="n">row_ax</span><span class="p">][</span><span class="n">col_ax</span><span class="p">]</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">fig</span><span class="p">,</span> <span class="n">curr_ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="n">figsize_x</span><span class="p">,</span> <span class="n">figsize_y</span><span class="p">))</span>
                <span class="n">heatmap</span><span class="p">(</span>
                    <span class="n">curr_ax</span><span class="p">,</span>
                    <span class="n">variant_mask</span><span class="p">,</span>
                    <span class="n">curr_heatmap_data</span><span class="p">,</span>
                    <span class="n">heatmap_data</span><span class="p">[</span><span class="s2">&quot;completed_gens_heatmap&quot;</span><span class="p">][</span><span class="s2">&quot;mean&quot;</span><span class="p">],</span>
                    <span class="n">new_gene_expression_factors</span><span class="p">,</span>
                    <span class="n">new_gene_translation_efficiency_values</span><span class="p">,</span>
                    <span class="n">heatmap_x_label</span><span class="p">,</span>
                    <span class="n">heatmap_y_label</span><span class="p">,</span>
                    <span class="n">title</span><span class="p">,</span>
                    <span class="n">heatmap_details</span><span class="p">[</span><span class="n">h</span><span class="p">][</span><span class="s2">&quot;box_text_size&quot;</span><span class="p">],</span>
                <span class="p">)</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">is_dashboard</span><span class="p">:</span>
                    <span class="n">fig</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
                    <span class="n">export_figure</span><span class="p">(</span><span class="n">plt</span><span class="p">,</span> <span class="n">plotOutDir</span><span class="p">,</span> <span class="n">h</span> <span class="o">+</span> <span class="n">filename_addition</span> <span class="o">+</span> <span class="n">plot_suffix</span><span class="p">)</span>
                    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">col_ax</span> <span class="o">+=</span> <span class="mi">1</span>
                    <span class="k">if</span> <span class="n">col_ax</span> <span class="o">==</span> <span class="n">dashboard_ncols</span><span class="p">:</span>
                        <span class="n">col_ax</span> <span class="o">=</span> <span class="mi">0</span>
                        <span class="n">row_ax</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="k">elif</span> <span class="n">h</span> <span class="o">==</span> <span class="s2">&quot;new_gene_mRNA_NTP_fraction_heatmap&quot;</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">cistron_idx</span><span class="p">,</span> <span class="n">cistron_id</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">new_gene_cistron_ids</span><span class="p">):</span>
                <span class="k">for</span> <span class="n">ntp_idx</span><span class="p">,</span> <span class="n">ntp_id</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">ntp_ids</span><span class="p">):</span>
                    <span class="n">title</span> <span class="o">=</span> <span class="p">(</span>
                        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">heatmap_details</span><span class="p">[</span><span class="n">h</span><span class="p">][</span><span class="s1">&#39;plot_title&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2"> </span><span class="si">{</span><span class="n">ntp_id</span><span class="p">[:</span><span class="o">-</span><span class="mi">3</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span>
                        <span class="sa">f</span><span class="s2">&quot; Fraction: </span><span class="si">{</span><span class="n">cistron_id</span><span class="p">[:</span><span class="o">-</span><span class="mi">4</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span>
                    <span class="p">)</span>
                    <span class="k">if</span> <span class="n">summary_statistic</span> <span class="o">==</span> <span class="s2">&quot;std_dev&quot;</span><span class="p">:</span>
                        <span class="n">title</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;Std Dev: </span><span class="si">{</span><span class="n">title</span><span class="si">}</span><span class="s2">&quot;</span>
                    <span class="k">if</span> <span class="n">is_dashboard</span><span class="p">:</span>
                        <span class="n">curr_ax</span> <span class="o">=</span> <span class="n">axs</span><span class="p">[</span><span class="n">row_ax</span><span class="p">][</span><span class="n">col_ax</span><span class="p">]</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">fig</span><span class="p">,</span> <span class="n">curr_ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span>
                            <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="n">figsize_x</span><span class="p">,</span> <span class="n">figsize_y</span><span class="p">)</span>
                        <span class="p">)</span>
                    <span class="n">heatmap</span><span class="p">(</span>
                        <span class="n">curr_ax</span><span class="p">,</span>
                        <span class="n">variant_mask</span><span class="p">,</span>
                        <span class="n">heatmap_data</span><span class="p">[</span><span class="n">h</span><span class="p">][</span><span class="n">summary_statistic</span><span class="p">][:,</span> <span class="p">:,</span> <span class="n">ntp_idx</span><span class="p">,</span> <span class="n">cistron_idx</span><span class="p">],</span>
                        <span class="n">heatmap_data</span><span class="p">[</span><span class="s2">&quot;completed_gens_heatmap&quot;</span><span class="p">][</span><span class="s2">&quot;mean&quot;</span><span class="p">],</span>
                        <span class="n">new_gene_expression_factors</span><span class="p">,</span>
                        <span class="n">new_gene_translation_efficiency_values</span><span class="p">,</span>
                        <span class="n">heatmap_x_label</span><span class="p">,</span>
                        <span class="n">heatmap_y_label</span><span class="p">,</span>
                        <span class="n">title</span><span class="p">,</span>
                        <span class="n">heatmap_details</span><span class="p">[</span><span class="n">h</span><span class="p">][</span><span class="s2">&quot;box_text_size&quot;</span><span class="p">],</span>
                    <span class="p">)</span>
                    <span class="k">if</span> <span class="ow">not</span> <span class="n">is_dashboard</span><span class="p">:</span>
                        <span class="n">fig</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
                        <span class="n">export_figure</span><span class="p">(</span>
                            <span class="n">plt</span><span class="p">,</span>
                            <span class="n">plotOutDir</span><span class="p">,</span>
                            <span class="sa">f</span><span class="s2">&quot;new_gene_mRNA_</span><span class="si">{</span><span class="n">ntp_id</span><span class="p">[:</span><span class="o">-</span><span class="mi">3</span><span class="p">]</span><span class="si">}</span><span class="s2">_fraction_heatmap&quot;</span>
                            <span class="o">+</span> <span class="n">filename_addition</span>
                            <span class="o">+</span> <span class="n">plot_suffix</span><span class="p">,</span>
                        <span class="p">)</span>
                        <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">col_ax</span> <span class="o">+=</span> <span class="mi">1</span>
                        <span class="k">if</span> <span class="n">col_ax</span> <span class="o">==</span> <span class="n">dashboard_ncols</span><span class="p">:</span>
                            <span class="n">col_ax</span> <span class="o">=</span> <span class="mi">0</span>
                            <span class="n">row_ax</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Heatmap </span><span class="si">{</span><span class="n">h</span><span class="si">}</span><span class="s2"> is neither a standard plot nor a&quot;</span>
                <span class="sa">f</span><span class="s2">&quot; nonstandard plot that has specific instructions for&quot;</span>
                <span class="sa">f</span><span class="s2">&quot; plotting.&quot;</span>
            <span class="p">)</span>
    <span class="k">if</span> <span class="n">is_dashboard</span><span class="p">:</span>
        <span class="n">fig</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
        <span class="n">export_figure</span><span class="p">(</span>
            <span class="n">plt</span><span class="p">,</span> <span class="n">plotOutDir</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;00_new_gene_exp_trl_eff_dashboard</span><span class="si">{</span><span class="n">plot_suffix</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="p">)</span>  <span class="c1">## TODO: Revert back after running new plots on Sherlock sims</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">(</span><span class="s2">&quot;all&quot;</span><span class="p">)</span></div>



<div class="viewcode-block" id="plot">
<a class="viewcode-back" href="../../../../reference/api/ecoli/ecoli.analysis.multivariant.new_gene_translation_efficiency_heatmaps.html#ecoli.analysis.multivariant.new_gene_translation_efficiency_heatmaps.plot">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">plot</span><span class="p">(</span>
    <span class="n">params</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">],</span>
    <span class="n">conn</span><span class="p">:</span> <span class="n">DuckDBPyConnection</span><span class="p">,</span>
    <span class="n">history_sql</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">config_sql</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">success_sql</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">sim_data_dict</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">str</span><span class="p">]],</span>
    <span class="n">validation_data_paths</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
    <span class="n">outdir</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">variant_metadata</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="n">Any</span><span class="p">]],</span>
    <span class="n">variant_names</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">],</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Create either a single multi-heatmap plot or 1+ separate heatmaps of data</span>
<span class="sd">    for a grid of new gene variant simulations with varying expression and</span>
<span class="sd">    translation efficiencies.</span>

<span class="sd">    Params (override corresponding hard-coded global variables):</span>
<span class="sd">        font_size, dashboard_flag, std_dev_flag, count_index, min_cell_index,</span>
<span class="sd">        max_cell_index</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Extract plot parameters</span>
    <span class="n">min_cell_index</span> <span class="o">=</span> <span class="n">params</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;min_cell_index&quot;</span><span class="p">,</span> <span class="n">MIN_CELL_INDEX</span><span class="p">)</span>
    <span class="n">max_cell_index</span> <span class="o">=</span> <span class="n">params</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;max_cell_index&quot;</span><span class="p">,</span> <span class="n">MAX_CELL_INDEX</span><span class="p">)</span>
    <span class="n">dashboard_flag</span> <span class="o">=</span> <span class="n">params</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;dashboard_flag&quot;</span><span class="p">,</span> <span class="n">DASHBOARD_FLAG</span><span class="p">)</span>
    <span class="n">std_dev_flag</span> <span class="o">=</span> <span class="n">params</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;std_dev_flag&quot;</span><span class="p">,</span> <span class="n">STD_DEV_FLAG</span><span class="p">)</span>
    <span class="n">count_index</span> <span class="o">=</span> <span class="n">params</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;count_index&quot;</span><span class="p">,</span> <span class="n">COUNT_INDEX</span><span class="p">)</span>

    <span class="c1"># Filter to specified generation range</span>
    <span class="n">history_sql</span> <span class="o">=</span> <span class="p">(</span>
        <span class="sa">f</span><span class="s2">&quot;FROM (</span><span class="si">{</span><span class="n">history_sql</span><span class="si">}</span><span class="s2">) WHERE generation &gt;= </span><span class="si">{</span><span class="n">min_cell_index</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="sa">f</span><span class="s2">&quot; AND generation &lt; </span><span class="si">{</span><span class="n">max_cell_index</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="p">)</span>
    <span class="n">config_sql</span> <span class="o">=</span> <span class="p">(</span>
        <span class="sa">f</span><span class="s2">&quot;FROM (</span><span class="si">{</span><span class="n">config_sql</span><span class="si">}</span><span class="s2">) WHERE generation &gt;= </span><span class="si">{</span><span class="n">min_cell_index</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="sa">f</span><span class="s2">&quot; AND generation &lt; </span><span class="si">{</span><span class="n">max_cell_index</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="p">)</span>
    <span class="c1"># Define baseline variant (ID = 0) as 0 new gene expr. and trans. eff.</span>
    <span class="n">experiment_id</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="nb">iter</span><span class="p">(</span><span class="n">variant_metadata</span><span class="o">.</span><span class="n">keys</span><span class="p">()))</span>
    <span class="n">variant_metadata</span> <span class="o">=</span> <span class="n">variant_metadata</span><span class="p">[</span><span class="n">experiment_id</span><span class="p">]</span>
    <span class="n">variant_metadata</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;exp_trl_eff&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;exp&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;trl_eff&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">}}</span>

    <span class="k">with</span> <span class="n">open_arbitrary_sim_data</span><span class="p">(</span><span class="n">sim_data_dict</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="n">sim_data</span> <span class="o">=</span> <span class="n">pickle</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>

    <span class="c1"># Determine new gene cistron and monomer ids</span>
    <span class="p">(</span><span class="n">new_gene_cistron_ids</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">new_gene_monomer_ids</span><span class="p">,</span> <span class="n">_</span><span class="p">)</span> <span class="o">=</span> <span class="n">get_new_gene_ids_and_indices</span><span class="p">(</span>
        <span class="n">sim_data</span>
    <span class="p">)</span>

    <span class="c1"># Assuming we ran a workflow with `n` new gene expression factors</span>
    <span class="c1"># and `m` new gene translation efficiency values, create an `n * m`</span>
    <span class="c1"># grid sorted in ascending order along both axes and calculate</span>
    <span class="c1"># mapping from variant numbers to row and column indices in grid</span>
    <span class="k">assert</span> <span class="s2">&quot;exp_trl_eff&quot;</span> <span class="ow">in</span> <span class="nb">next</span><span class="p">(</span><span class="nb">iter</span><span class="p">(</span><span class="n">variant_metadata</span><span class="o">.</span><span class="n">values</span><span class="p">())),</span> <span class="p">(</span>
        <span class="s2">&quot;This plot is intended to be run on simulations where the&quot;</span>
        <span class="s2">&quot; new gene expression-translation efficiency variant was &quot;</span>
        <span class="s2">&quot;enabled, but no parameters for this variant were found.&quot;</span>
    <span class="p">)</span>
    <span class="n">new_gene_expression_factors</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span>
        <span class="nb">set</span><span class="p">(</span>
            <span class="n">variant_params</span><span class="p">[</span><span class="s2">&quot;exp_trl_eff&quot;</span><span class="p">][</span><span class="s2">&quot;exp&quot;</span><span class="p">]</span>
            <span class="k">for</span> <span class="n">variant_params</span> <span class="ow">in</span> <span class="n">variant_metadata</span><span class="o">.</span><span class="n">values</span><span class="p">()</span>
        <span class="p">)</span>
    <span class="p">)</span>
    <span class="n">new_gene_translation_efficiency_values</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span>
        <span class="nb">set</span><span class="p">(</span>
            <span class="n">variant_params</span><span class="p">[</span><span class="s2">&quot;exp_trl_eff&quot;</span><span class="p">][</span><span class="s2">&quot;trl_eff&quot;</span><span class="p">]</span>
            <span class="k">for</span> <span class="n">variant_params</span> <span class="ow">in</span> <span class="n">variant_metadata</span><span class="o">.</span><span class="n">values</span><span class="p">()</span>
        <span class="p">)</span>
    <span class="p">)</span>
    <span class="n">variant_matrix_shape</span> <span class="o">=</span> <span class="p">(</span>
        <span class="nb">len</span><span class="p">(</span><span class="n">new_gene_translation_efficiency_values</span><span class="p">),</span>
        <span class="nb">len</span><span class="p">(</span><span class="n">new_gene_expression_factors</span><span class="p">),</span>
    <span class="p">)</span>
    <span class="n">variant_to_row_col</span> <span class="o">=</span> <span class="p">{</span>
        <span class="n">variant</span><span class="p">:</span> <span class="p">(</span>
            <span class="n">new_gene_translation_efficiency_values</span><span class="o">.</span><span class="n">index</span><span class="p">(</span>
                <span class="n">variant_params</span><span class="p">[</span><span class="s2">&quot;exp_trl_eff&quot;</span><span class="p">][</span><span class="s2">&quot;trl_eff&quot;</span><span class="p">]</span>
            <span class="p">),</span>
            <span class="n">new_gene_expression_factors</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">variant_params</span><span class="p">[</span><span class="s2">&quot;exp_trl_eff&quot;</span><span class="p">][</span><span class="s2">&quot;exp&quot;</span><span class="p">]),</span>
        <span class="p">)</span>
        <span class="k">for</span> <span class="n">variant</span><span class="p">,</span> <span class="n">variant_params</span> <span class="ow">in</span> <span class="n">variant_metadata</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
    <span class="p">}</span>
    <span class="n">variant_mask</span> <span class="o">=</span> <span class="n">get_variant_mask</span><span class="p">(</span>
        <span class="n">conn</span><span class="p">,</span> <span class="n">config_sql</span><span class="p">,</span> <span class="n">variant_to_row_col</span><span class="p">,</span> <span class="n">variant_matrix_shape</span>
    <span class="p">)</span>

    <span class="n">bulk_ids</span> <span class="o">=</span> <span class="n">field_metadata</span><span class="p">(</span><span class="n">conn</span><span class="p">,</span> <span class="n">config_sql</span><span class="p">,</span> <span class="s2">&quot;bulk&quot;</span><span class="p">)</span>
    <span class="n">ntp_ids</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">sim_data</span><span class="o">.</span><span class="n">ntp_code_to_id_ordered</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>
    <span class="c1"># Get indices for data extraction</span>
    <span class="n">rnap_subunit_mRNA_ids</span> <span class="o">=</span> <span class="n">get_mRNA_ids_from_monomer_ids</span><span class="p">(</span>
        <span class="n">sim_data</span><span class="p">,</span> <span class="n">sim_data</span><span class="o">.</span><span class="n">molecule_groups</span><span class="o">.</span><span class="n">RNAP_subunits</span>
    <span class="p">)</span>
    <span class="n">rnap_subunit_mRNA_indexes</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span>
        <span class="nb">set</span><span class="p">(</span>
            <span class="n">itertools</span><span class="o">.</span><span class="n">chain</span><span class="o">.</span><span class="n">from_iterable</span><span class="p">(</span>
                <span class="n">cast</span><span class="p">(</span>
                    <span class="nb">list</span><span class="p">[</span><span class="nb">list</span><span class="p">[</span><span class="nb">int</span><span class="p">]],</span>
                    <span class="n">get_indexes</span><span class="p">(</span><span class="n">conn</span><span class="p">,</span> <span class="n">config_sql</span><span class="p">,</span> <span class="s2">&quot;mRNA&quot;</span><span class="p">,</span> <span class="n">rnap_subunit_mRNA_ids</span><span class="p">),</span>
                <span class="p">)</span>
            <span class="p">)</span>
        <span class="p">)</span>
    <span class="p">)</span>
    <span class="n">rnap_subunit_monomer_indexes</span> <span class="o">=</span> <span class="n">get_indexes</span><span class="p">(</span>
        <span class="n">conn</span><span class="p">,</span> <span class="n">config_sql</span><span class="p">,</span> <span class="s2">&quot;monomer&quot;</span><span class="p">,</span> <span class="n">sim_data</span><span class="o">.</span><span class="n">molecule_groups</span><span class="o">.</span><span class="n">RNAP_subunits</span>
    <span class="p">)</span>
    <span class="n">ribosomal_mRNA_ids</span> <span class="o">=</span> <span class="n">get_mRNA_ids_from_monomer_ids</span><span class="p">(</span>
        <span class="n">sim_data</span><span class="p">,</span> <span class="n">sim_data</span><span class="o">.</span><span class="n">molecule_groups</span><span class="o">.</span><span class="n">ribosomal_proteins</span>
    <span class="p">)</span>
    <span class="n">ribosomal_mRNA_indexes</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span>
        <span class="nb">set</span><span class="p">(</span>
            <span class="n">itertools</span><span class="o">.</span><span class="n">chain</span><span class="o">.</span><span class="n">from_iterable</span><span class="p">(</span>
                <span class="n">cast</span><span class="p">(</span>
                    <span class="nb">list</span><span class="p">[</span><span class="nb">list</span><span class="p">[</span><span class="nb">int</span><span class="p">]],</span>
                    <span class="n">get_indexes</span><span class="p">(</span><span class="n">conn</span><span class="p">,</span> <span class="n">config_sql</span><span class="p">,</span> <span class="s2">&quot;mRNA&quot;</span><span class="p">,</span> <span class="n">ribosomal_mRNA_ids</span><span class="p">),</span>
                <span class="p">)</span>
            <span class="p">)</span>
        <span class="p">)</span>
    <span class="p">)</span>
    <span class="n">ribosomal_monomer_indexes</span> <span class="o">=</span> <span class="n">get_indexes</span><span class="p">(</span>
        <span class="n">conn</span><span class="p">,</span> <span class="n">config_sql</span><span class="p">,</span> <span class="s2">&quot;monomer&quot;</span><span class="p">,</span> <span class="n">sim_data</span><span class="o">.</span><span class="n">molecule_groups</span><span class="o">.</span><span class="n">ribosomal_proteins</span>
    <span class="p">)</span>
    <span class="n">cistron_ids</span> <span class="o">=</span> <span class="n">field_metadata</span><span class="p">(</span>
        <span class="n">conn</span><span class="p">,</span> <span class="n">config_sql</span><span class="p">,</span> <span class="s2">&quot;listeners__rna_counts__full_mRNA_cistron_counts&quot;</span>
    <span class="p">)</span>
    <span class="n">capacity_gene_mRNA_ids</span> <span class="o">=</span> <span class="n">get_mRNA_ids_from_monomer_ids</span><span class="p">(</span>
        <span class="n">sim_data</span><span class="p">,</span> <span class="n">capacity_gene_monomer_ids</span>
    <span class="p">)</span>
    <span class="n">capacity_gene_mRNA_indexes</span> <span class="o">=</span> <span class="n">cast</span><span class="p">(</span>
        <span class="nb">list</span><span class="p">[</span><span class="nb">list</span><span class="p">[</span><span class="nb">int</span><span class="p">]],</span> <span class="n">get_indexes</span><span class="p">(</span><span class="n">conn</span><span class="p">,</span> <span class="n">config_sql</span><span class="p">,</span> <span class="s2">&quot;mRNA&quot;</span><span class="p">,</span> <span class="n">capacity_gene_mRNA_ids</span><span class="p">)</span>
    <span class="p">)</span>
    <span class="n">capacity_gene_monomer_indexes</span> <span class="o">=</span> <span class="n">cast</span><span class="p">(</span>
        <span class="nb">list</span><span class="p">[</span><span class="nb">int</span><span class="p">],</span> <span class="n">get_indexes</span><span class="p">(</span><span class="n">conn</span><span class="p">,</span> <span class="n">config_sql</span><span class="p">,</span> <span class="s2">&quot;monomer&quot;</span><span class="p">,</span> <span class="n">capacity_gene_monomer_ids</span><span class="p">)</span>
    <span class="p">)</span>
    <span class="n">capacity_gene_common_names</span> <span class="o">=</span> <span class="p">[</span>
        <span class="n">sim_data</span><span class="o">.</span><span class="n">common_names</span><span class="o">.</span><span class="n">get_common_name</span><span class="p">(</span><span class="n">monomer_id</span><span class="p">[:</span><span class="o">-</span><span class="mi">3</span><span class="p">])</span>
        <span class="k">for</span> <span class="n">monomer_id</span> <span class="ow">in</span> <span class="n">capacity_gene_monomer_ids</span>
    <span class="p">]</span>
    <span class="c1"># Convert cistron IDs to RNA IDs</span>
    <span class="n">RNA_ids</span> <span class="o">=</span> <span class="n">sim_data</span><span class="o">.</span><span class="n">process</span><span class="o">.</span><span class="n">transcription</span><span class="o">.</span><span class="n">rna_data</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]</span>
    <span class="n">new_gene_mRNA_ids</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">cistron_id</span> <span class="ow">in</span> <span class="n">new_gene_cistron_ids</span><span class="p">:</span>
        <span class="n">target_RNA_idx</span> <span class="o">=</span> <span class="n">sim_data</span><span class="o">.</span><span class="n">process</span><span class="o">.</span><span class="n">transcription</span><span class="o">.</span><span class="n">cistron_id_to_rna_indexes</span><span class="p">(</span>
            <span class="n">cistron_id</span>
        <span class="p">)</span>
        <span class="n">new_gene_mRNA_ids</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">RNA_ids</span><span class="p">[</span><span class="n">target_RNA_idx</span><span class="p">])</span>
    <span class="n">new_gene_mRNA_indexes</span> <span class="o">=</span> <span class="n">cast</span><span class="p">(</span>
        <span class="nb">list</span><span class="p">[</span><span class="nb">list</span><span class="p">[</span><span class="nb">int</span><span class="p">]],</span> <span class="n">get_indexes</span><span class="p">(</span><span class="n">conn</span><span class="p">,</span> <span class="n">config_sql</span><span class="p">,</span> <span class="s2">&quot;mRNA&quot;</span><span class="p">,</span> <span class="n">new_gene_mRNA_ids</span><span class="p">)</span>
    <span class="p">)</span>
    <span class="n">new_gene_monomer_indexes</span> <span class="o">=</span> <span class="n">cast</span><span class="p">(</span>
        <span class="nb">list</span><span class="p">[</span><span class="nb">int</span><span class="p">],</span> <span class="n">get_indexes</span><span class="p">(</span><span class="n">conn</span><span class="p">,</span> <span class="n">config_sql</span><span class="p">,</span> <span class="s2">&quot;monomer&quot;</span><span class="p">,</span> <span class="n">new_gene_monomer_ids</span><span class="p">)</span>
    <span class="p">)</span>
    <span class="n">new_gene_RNA_indexes</span> <span class="o">=</span> <span class="n">cast</span><span class="p">(</span>
        <span class="nb">list</span><span class="p">[</span><span class="nb">list</span><span class="p">[</span><span class="nb">int</span><span class="p">]],</span> <span class="n">get_indexes</span><span class="p">(</span><span class="n">conn</span><span class="p">,</span> <span class="n">config_sql</span><span class="p">,</span> <span class="s2">&quot;RNA&quot;</span><span class="p">,</span> <span class="n">new_gene_mRNA_ids</span><span class="p">)</span>
    <span class="p">)</span>
    <span class="n">new_gene_cistron_indexes</span> <span class="o">=</span> <span class="n">cast</span><span class="p">(</span>
        <span class="nb">list</span><span class="p">[</span><span class="nb">int</span><span class="p">],</span> <span class="n">get_indexes</span><span class="p">(</span><span class="n">conn</span><span class="p">,</span> <span class="n">config_sql</span><span class="p">,</span> <span class="s2">&quot;cistron&quot;</span><span class="p">,</span> <span class="n">new_gene_cistron_ids</span><span class="p">)</span>
    <span class="p">)</span>

    <span class="c1"># Determine glucose index in exchange fluxes</span>
    <span class="n">external_molecule_ids</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span>
        <span class="n">field_metadata</span><span class="p">(</span>
            <span class="n">conn</span><span class="p">,</span> <span class="n">config_sql</span><span class="p">,</span> <span class="s2">&quot;listeners__fba_results__external_exchange_fluxes&quot;</span>
        <span class="p">)</span>
    <span class="p">)</span>
    <span class="k">if</span> <span class="s2">&quot;GLC[p]&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">external_molecule_ids</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;This plot only runs when glucose is the carbon source.&quot;</span><span class="p">)</span>
        <span class="k">return</span>
    <span class="n">glucose_idx</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">external_molecule_ids</span> <span class="o">==</span> <span class="s2">&quot;GLC[p]&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="mi">1</span>
    <span class="n">flux_scaling_factor</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">sim_data</span><span class="o">.</span><span class="n">getter</span><span class="o">.</span><span class="n">get_mass</span><span class="p">(</span><span class="s2">&quot;GLC[p]&quot;</span><span class="p">)</span>
        <span class="o">*</span> <span class="p">(</span><span class="n">units</span><span class="o">.</span><span class="n">mmol</span> <span class="o">/</span> <span class="n">units</span><span class="o">.</span><span class="n">g</span> <span class="o">/</span> <span class="n">units</span><span class="o">.</span><span class="n">h</span><span class="p">)</span>
        <span class="o">*</span> <span class="n">units</span><span class="o">.</span><span class="n">fg</span>  <span class="c1"># Flux * dry mass units</span>
    <span class="p">)</span><span class="o">.</span><span class="n">asNumber</span><span class="p">()</span>

    <span class="c1"># Get normalized translation efficiency for all mRNAs</span>
    <span class="n">trl_effs</span> <span class="o">=</span> <span class="n">sim_data</span><span class="o">.</span><span class="n">process</span><span class="o">.</span><span class="n">translation</span><span class="o">.</span><span class="n">translation_efficiencies_by_monomer</span>
    <span class="n">trl_eff_ids</span> <span class="o">=</span> <span class="n">sim_data</span><span class="o">.</span><span class="n">process</span><span class="o">.</span><span class="n">translation</span><span class="o">.</span><span class="n">monomer_data</span><span class="p">[</span><span class="s2">&quot;cistron_id&quot;</span><span class="p">]</span>
    <span class="n">mRNA_cistron_idx_dict</span> <span class="o">=</span> <span class="p">{</span><span class="n">rna</span><span class="p">:</span> <span class="n">i</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">rna</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">cistron_ids</span><span class="p">)}</span>
    <span class="n">trl_eff_id_mapping</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">mRNA_cistron_idx_dict</span><span class="p">[</span><span class="nb">id</span><span class="p">]</span> <span class="k">for</span> <span class="nb">id</span> <span class="ow">in</span> <span class="n">trl_eff_ids</span><span class="p">])</span>
    <span class="n">ordered_trl_effs</span> <span class="o">=</span> <span class="n">trl_effs</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="n">trl_eff_id_mapping</span><span class="p">)]</span>

<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Details needed to create all possible heatmaps</span>
<span class="sd">        key (string): Heatmap identifier, will also be used in file name if</span>
<span class="sd">            plots are saved separately</span>
<span class="sd">        is_new_gene_heatmap (bool): If True, one heatmap will be made </span>
<span class="sd">            for each new gene</span>
<span class="sd">        is_nonstandard_plotting (bool): False if only one plot (or one plot</span>
<span class="sd">            per new gene) needs to be made. True in all other cases.</span>
<span class="sd">        data_table (string): Table to get data from.</span>
<span class="sd">        data_column (string): Column in table to get data from.</span>
<span class="sd">        default_value (int): Value to use in heatmap if no data is </span>
<span class="sd">            extracted for this parameter combination.</span>
<span class="sd">        remove_first (bool): If True, removes the first column of data </span>
<span class="sd">            from each cell (which might be set to a default value in </span>
<span class="sd">            some cases)</span>
<span class="sd">        num_digits_rounding (int): Specifies how to round the number </span>
<span class="sd">            displayed in each sequare of the heatmap</span>
<span class="sd">        box_text_size (string): Specifies font size of number displayed </span>
<span class="sd">            in each square of the heatmap</span>
<span class="sd">        plot_title (string): Title of heatmap to display</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Specify unique fields and non-default values here</span>
    <span class="n">heatmap_details</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">dict</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;completed_gens_heatmap&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;columns&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;time&quot;</span><span class="p">],</span>
            <span class="s2">&quot;custom_sql&quot;</span><span class="p">:</span> <span class="sa">f</span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">                WITH max_gen_per_seed AS (</span>
<span class="s2">                    SELECT max(generation) AS max_generation, experiment_id, variant</span>
<span class="s2">                    FROM (</span><span class="se">{{</span><span class="s2">subquery</span><span class="se">}}</span><span class="s2">)</span>
<span class="s2">                    GROUP BY experiment_id, variant, lineage_seed</span>
<span class="s2">                )</span>
<span class="s2">                -- Boolean values must be explicitly cast to numeric for aggregation</span>
<span class="s2">                SELECT variant, avg((max_generation = </span><span class="si">{</span><span class="n">count_index</span><span class="si">}</span><span class="s2">)::BIGINT) AS mean,</span>
<span class="s2">                    stddev((max_generation = </span><span class="si">{</span><span class="n">count_index</span><span class="si">}</span><span class="s2">)::BIGINT) AS std</span>
<span class="s2">                FROM max_gen_per_seed</span>
<span class="s2">                GROUP BY experiment_id, variant</span>
<span class="s2">                &quot;&quot;&quot;</span><span class="p">,</span>
            <span class="s2">&quot;plot_title&quot;</span><span class="p">:</span> <span class="sa">f</span><span class="s2">&quot;Percentage of Sims That Reached Generation </span><span class="si">{</span><span class="n">count_index</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span>
        <span class="p">},</span>
        <span class="s2">&quot;doubling_times_heatmap&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;columns&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;time&quot;</span><span class="p">],</span>
            <span class="s2">&quot;custom_sql&quot;</span><span class="p">:</span> <span class="s2">&quot;&quot;&quot;</span>
<span class="s2">                WITH doubling_times AS (</span>
<span class="s2">                    SELECT (max(time) - min(time)) / 60 AS doubling_time,</span>
<span class="s2">                        experiment_id, variant, lineage_seed, generation, agent_id</span>
<span class="s2">                    FROM (</span><span class="si">{subquery}</span><span class="s2">)</span>
<span class="s2">                    GROUP BY experiment_id, variant, lineage_seed, generation, agent_id</span>
<span class="s2">                )</span>
<span class="s2">                SELECT variant, avg(doubling_time) AS mean, stddev(doubling_time) AS std</span>
<span class="s2">                FROM doubling_times</span>
<span class="s2">                GROUP BY experiment_id, variant</span>
<span class="s2">                &quot;&quot;&quot;</span><span class="p">,</span>
            <span class="s2">&quot;plot_title&quot;</span><span class="p">:</span> <span class="s2">&quot;Doubling Time (minutes)&quot;</span><span class="p">,</span>
        <span class="p">},</span>
        <span class="s2">&quot;cell_volume_heatmap&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;columns&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;listeners__mass__volume&quot;</span><span class="p">],</span>
            <span class="s2">&quot;plot_title&quot;</span><span class="p">:</span> <span class="s2">&quot;Cell Volume (fL)&quot;</span><span class="p">,</span>
        <span class="p">},</span>
        <span class="s2">&quot;cell_mass_heatmap&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;columns&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;listeners__mass__cell_mass&quot;</span><span class="p">],</span>
            <span class="s2">&quot;plot_title&quot;</span><span class="p">:</span> <span class="s2">&quot;Cell Mass (fg)&quot;</span><span class="p">,</span>
        <span class="p">},</span>
        <span class="s2">&quot;cell_dry_mass_heatmap&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;columns&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;listeners__mass__dry_mass&quot;</span><span class="p">],</span>
            <span class="s2">&quot;plot_title&quot;</span><span class="p">:</span> <span class="s2">&quot;Dry Cell Mass (fg)&quot;</span><span class="p">,</span>
        <span class="p">},</span>
        <span class="s2">&quot;cell_mRNA_mass_heatmap&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;columns&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;listeners__mass__mRna_mass&quot;</span><span class="p">],</span>
            <span class="s2">&quot;plot_title&quot;</span><span class="p">:</span> <span class="s2">&quot;Total mRNA Mass (fg)&quot;</span><span class="p">,</span>
        <span class="p">},</span>
        <span class="s2">&quot;cell_protein_mass_heatmap&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;columns&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;listeners__mass__protein_mass&quot;</span><span class="p">],</span>
            <span class="s2">&quot;plot_title&quot;</span><span class="p">:</span> <span class="s2">&quot;Total Protein Mass (fg)&quot;</span><span class="p">,</span>
            <span class="s2">&quot;box_text_size&quot;</span><span class="p">:</span> <span class="s2">&quot;x-small&quot;</span><span class="p">,</span>
            <span class="s2">&quot;num_digits_rounding&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
        <span class="p">},</span>
        <span class="s2">&quot;ppgpp_concentration_heatmap&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;columns&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;listeners__growth_limits__ppgpp_conc&quot;</span><span class="p">],</span>
            <span class="s2">&quot;plot_title&quot;</span><span class="p">:</span> <span class="s2">&quot;ppGpp Concentration (uM)&quot;</span><span class="p">,</span>
            <span class="s2">&quot;remove_first&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
            <span class="s2">&quot;num_digits_rounding&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span>
        <span class="p">},</span>
        <span class="s2">&quot;ribosome_init_events_heatmap&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;columns&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;listeners__ribosome_data__did_initialize&quot;</span><span class="p">],</span>
            <span class="s2">&quot;plot_title&quot;</span><span class="p">:</span> <span class="s2">&quot;Ribosome Init Events Per Time Step&quot;</span><span class="p">,</span>
            <span class="s2">&quot;num_digits_rounding&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
            <span class="s2">&quot;box_text_size&quot;</span><span class="p">:</span> <span class="s2">&quot;x-small&quot;</span><span class="p">,</span>
        <span class="p">},</span>
        <span class="s2">&quot;rnap_counts_heatmap&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;plot_title&quot;</span><span class="p">:</span> <span class="s2">&quot;RNA Polymerase (RNAP) Counts&quot;</span><span class="p">,</span>
            <span class="s2">&quot;num_digits_rounding&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
            <span class="s2">&quot;box_text_size&quot;</span><span class="p">:</span> <span class="s2">&quot;x-small&quot;</span><span class="p">,</span>
            <span class="s2">&quot;columns&quot;</span><span class="p">:</span> <span class="p">[</span>
                <span class="n">get_rnap_counts_projection</span><span class="p">(</span><span class="n">sim_data</span><span class="p">,</span> <span class="n">bulk_ids</span><span class="p">),</span>
                <span class="s2">&quot;listeners__unique_molecule_counts__active_RNAP&quot;</span><span class="p">,</span>
            <span class="p">],</span>
            <span class="s2">&quot;custom_sql&quot;</span><span class="p">:</span> <span class="s2">&quot;&quot;&quot;</span>
<span class="s2">                WITH total_counts AS (</span>
<span class="s2">                    SELECT avg(bulk +</span>
<span class="s2">                        listeners__unique_molecule_counts__active_RNAP) AS rnap_counts,</span>
<span class="s2">                        experiment_id, variant</span>
<span class="s2">                    FROM (</span><span class="si">{subquery}</span><span class="s2">)</span>
<span class="s2">                    GROUP BY experiment_id, variant, lineage_seed, generation, agent_id</span>
<span class="s2">                )</span>
<span class="s2">                SELECT variant, avg(rnap_counts) AS mean, stddev(rnap_counts) AS std</span>
<span class="s2">                FROM total_counts</span>
<span class="s2">                GROUP BY experiment_id, variant</span>
<span class="s2">                &quot;&quot;&quot;</span><span class="p">,</span>
        <span class="p">},</span>
        <span class="s2">&quot;ribosome_counts_heatmap&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;plot_title&quot;</span><span class="p">:</span> <span class="s2">&quot;Active Ribosome Counts&quot;</span><span class="p">,</span>
            <span class="s2">&quot;num_digits_rounding&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
            <span class="s2">&quot;box_text_size&quot;</span><span class="p">:</span> <span class="s2">&quot;x-small&quot;</span><span class="p">,</span>
            <span class="s2">&quot;columns&quot;</span><span class="p">:</span> <span class="p">[</span>
                <span class="n">get_ribosome_counts_projection</span><span class="p">(</span><span class="n">sim_data</span><span class="p">,</span> <span class="n">bulk_ids</span><span class="p">),</span>
                <span class="s2">&quot;listeners__unique_molecule_counts__active_ribosome&quot;</span><span class="p">,</span>
            <span class="p">],</span>
            <span class="s2">&quot;custom_sql&quot;</span><span class="p">:</span> <span class="s2">&quot;&quot;&quot;</span>
<span class="s2">                WITH total_counts AS (</span>
<span class="s2">                    SELECT avg(bulk +</span>
<span class="s2">                        listeners__unique_molecule_counts__active_ribosome)</span>
<span class="s2">                        AS ribosome_counts,</span>
<span class="s2">                        experiment_id, variant</span>
<span class="s2">                    FROM (</span><span class="si">{subquery}</span><span class="s2">)</span>
<span class="s2">                    GROUP BY experiment_id, variant, lineage_seed, generation, agent_id</span>
<span class="s2">                )</span>
<span class="s2">                SELECT variant, avg(ribosome_counts) AS mean,</span>
<span class="s2">                    stddev(ribosome_counts) AS std</span>
<span class="s2">                FROM total_counts</span>
<span class="s2">                GROUP BY experiment_id, variant</span>
<span class="s2">                &quot;&quot;&quot;</span><span class="p">,</span>
        <span class="p">},</span>
        <span class="s2">&quot;active_ribosome_counts_heatmap&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;columns&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;listeners__unique_molecule_counts__active_ribosome&quot;</span><span class="p">],</span>
            <span class="s2">&quot;plot_title&quot;</span><span class="p">:</span> <span class="s2">&quot;Active Ribosome Counts&quot;</span><span class="p">,</span>
            <span class="s2">&quot;num_digits_rounding&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
            <span class="s2">&quot;box_text_size&quot;</span><span class="p">:</span> <span class="s2">&quot;x-small&quot;</span><span class="p">,</span>
        <span class="p">},</span>
        <span class="s2">&quot;active_rnap_counts_heatmap&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;columns&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;listeners__unique_molecule_counts__active_RNAP&quot;</span><span class="p">],</span>
            <span class="s2">&quot;plot_title&quot;</span><span class="p">:</span> <span class="s2">&quot;Active RNA Polymerase (RNAP) Counts&quot;</span><span class="p">,</span>
            <span class="s2">&quot;num_digits_rounding&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
            <span class="s2">&quot;box_text_size&quot;</span><span class="p">:</span> <span class="s2">&quot;x-small&quot;</span><span class="p">,</span>
        <span class="p">},</span>
        <span class="s2">&quot;free_ribosome_counts_heatmap&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;plot_title&quot;</span><span class="p">:</span> <span class="s2">&quot;Free Ribosome Counts&quot;</span><span class="p">,</span>
            <span class="s2">&quot;num_digits_rounding&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
            <span class="s2">&quot;box_text_size&quot;</span><span class="p">:</span> <span class="s2">&quot;x-small&quot;</span><span class="p">,</span>
            <span class="s2">&quot;columns&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">get_ribosome_counts_projection</span><span class="p">(</span><span class="n">sim_data</span><span class="p">,</span> <span class="n">bulk_ids</span><span class="p">)],</span>
            <span class="s2">&quot;custom_sql&quot;</span><span class="p">:</span> <span class="s2">&quot;&quot;&quot;</span>
<span class="s2">                WITH avg_per_cell AS (</span>
<span class="s2">                    SELECT avg(bulk) AS avg_col, experiment_id, variant</span>
<span class="s2">                    FROM (</span><span class="si">{subquery}</span><span class="s2">)</span>
<span class="s2">                    GROUP BY experiment_id, variant, lineage_seed, generation, agent_id</span>
<span class="s2">                )</span>
<span class="s2">                SELECT variant, avg(avg_col) AS mean, stddev(avg_col) AS std</span>
<span class="s2">                FROM avg_per_cell</span>
<span class="s2">                GROUP BY experiment_id, variant</span>
<span class="s2">                &quot;&quot;&quot;</span><span class="p">,</span>
        <span class="p">},</span>
        <span class="s2">&quot;free_rnap_counts_heatmap&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;plot_title&quot;</span><span class="p">:</span> <span class="s2">&quot;Free RNA Polymerase (RNAP) Counts&quot;</span><span class="p">,</span>
            <span class="s2">&quot;num_digits_rounding&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
            <span class="s2">&quot;box_text_size&quot;</span><span class="p">:</span> <span class="s2">&quot;x-small&quot;</span><span class="p">,</span>
            <span class="s2">&quot;columns&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">get_rnap_counts_projection</span><span class="p">(</span><span class="n">sim_data</span><span class="p">,</span> <span class="n">bulk_ids</span><span class="p">)],</span>
            <span class="s2">&quot;custom_sql&quot;</span><span class="p">:</span> <span class="s2">&quot;&quot;&quot;</span>
<span class="s2">                WITH avg_per_cell AS (</span>
<span class="s2">                    SELECT avg(bulk) AS avg_col, experiment_id, variant</span>
<span class="s2">                    FROM (</span><span class="si">{subquery}</span><span class="s2">)</span>
<span class="s2">                    GROUP BY experiment_id, variant, lineage_seed, generation, agent_id</span>
<span class="s2">                )</span>
<span class="s2">                SELECT variant, avg(avg_col) AS mean, stddev(avg_col) AS std</span>
<span class="s2">                FROM avg_per_cell</span>
<span class="s2">                GROUP BY experiment_id, variant</span>
<span class="s2">                &quot;&quot;&quot;</span><span class="p">,</span>
        <span class="p">},</span>
        <span class="s2">&quot;rnap_ribosome_counts_ratio_heatmap&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;plot_title&quot;</span><span class="p">:</span> <span class="s2">&quot;RNAP Counts / Ribosome Counts&quot;</span><span class="p">,</span>
            <span class="s2">&quot;num_digits_rounding&quot;</span><span class="p">:</span> <span class="mi">4</span><span class="p">,</span>
            <span class="s2">&quot;box_text_size&quot;</span><span class="p">:</span> <span class="s2">&quot;x-small&quot;</span><span class="p">,</span>
            <span class="s2">&quot;columns&quot;</span><span class="p">:</span> <span class="p">[</span>
                <span class="n">get_rnap_counts_projection</span><span class="p">(</span><span class="n">sim_data</span><span class="p">,</span> <span class="n">bulk_ids</span><span class="p">)</span>
                <span class="o">+</span> <span class="s2">&quot;_rnap, &quot;</span>
                <span class="o">+</span> <span class="n">get_ribosome_counts_projection</span><span class="p">(</span><span class="n">sim_data</span><span class="p">,</span> <span class="n">bulk_ids</span><span class="p">)</span>
                <span class="o">+</span> <span class="s2">&quot;_ribosome&quot;</span><span class="p">,</span>
                <span class="s2">&quot;listeners__unique_molecule_counts__active_RNAP&quot;</span><span class="p">,</span>
                <span class="s2">&quot;listeners__unique_molecule_counts__active_ribosome&quot;</span><span class="p">,</span>
            <span class="p">],</span>
            <span class="s2">&quot;custom_sql&quot;</span><span class="p">:</span> <span class="s2">&quot;&quot;&quot;</span>
<span class="s2">                WITH total_counts AS (</span>
<span class="s2">                    SELECT avg(bulk_ribosome +</span>
<span class="s2">                        listeners__unique_molecule_counts__active_ribosome)</span>
<span class="s2">                        AS ribosome_counts,</span>
<span class="s2">                        avg(bulk_rnap +</span>
<span class="s2">                        listeners__unique_molecule_counts__active_RNAP) AS rnap_counts,</span>
<span class="s2">                        experiment_id, variant</span>
<span class="s2">                    FROM (</span><span class="si">{subquery}</span><span class="s2">)</span>
<span class="s2">                    GROUP BY experiment_id, variant, lineage_seed, generation, agent_id</span>
<span class="s2">                )</span>
<span class="s2">                SELECT variant, avg(rnap_counts / ribosome_counts) AS mean,</span>
<span class="s2">                    stddev(rnap_counts / ribosome_counts) AS std,</span>
<span class="s2">                FROM total_counts</span>
<span class="s2">                GROUP BY experiment_id, variant</span>
<span class="s2">                &quot;&quot;&quot;</span><span class="p">,</span>
        <span class="p">},</span>
        <span class="s2">&quot;rnap_crowding_heatmap&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;columns&quot;</span><span class="p">:</span> <span class="p">[</span>
                <span class="s2">&quot;listeners__rna_synth_prob__target_rna_synth_prob&quot;</span><span class="p">,</span>
                <span class="s2">&quot;listeners__rna_synth_prob__actual_rna_synth_prob&quot;</span><span class="p">,</span>
            <span class="p">],</span>
            <span class="s2">&quot;plot_title&quot;</span><span class="p">:</span> <span class="s2">&quot;RNAP Crowding: # of TUs&quot;</span><span class="p">,</span>
            <span class="s2">&quot;num_digits_rounding&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
            <span class="s2">&quot;box_text_size&quot;</span><span class="p">:</span> <span class="s2">&quot;x-small&quot;</span><span class="p">,</span>
            <span class="s2">&quot;custom_sql&quot;</span><span class="p">:</span> <span class="n">get_overcrowding_sql</span><span class="p">(</span>
                <span class="s2">&quot;listeners__rna_synth_prob__target_rna_synth_prob&quot;</span><span class="p">,</span>
                <span class="s2">&quot;listeners__rna_synth_prob__actual_rna_synth_prob&quot;</span><span class="p">,</span>
            <span class="p">),</span>
        <span class="p">},</span>
        <span class="s2">&quot;ribosome_crowding_heatmap&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;columns&quot;</span><span class="p">:</span> <span class="p">[</span>
                <span class="s2">&quot;listeners__ribosome_data__target_prob_translation_per_transcript&quot;</span><span class="p">,</span>
                <span class="s2">&quot;listeners__ribosome_data__actual_prob_translation_per_transcript&quot;</span><span class="p">,</span>
            <span class="p">],</span>
            <span class="s2">&quot;plot_title&quot;</span><span class="p">:</span> <span class="s2">&quot;Ribosome Crowding: # of Monomers&quot;</span><span class="p">,</span>
            <span class="s2">&quot;num_digits_rounding&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
            <span class="s2">&quot;box_text_size&quot;</span><span class="p">:</span> <span class="s2">&quot;x-small&quot;</span><span class="p">,</span>
            <span class="s2">&quot;custom_sql&quot;</span><span class="p">:</span> <span class="n">get_overcrowding_sql</span><span class="p">(</span>
                <span class="s2">&quot;listeners__ribosome_data__target_prob_translation_per_transcript&quot;</span><span class="p">,</span>
                <span class="s2">&quot;listeners__ribosome_data__actual_prob_translation_per_transcript&quot;</span><span class="p">,</span>
            <span class="p">),</span>
        <span class="p">},</span>
        <span class="s2">&quot;rna_synth_prob_max_p_heatmap&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;columns&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;listeners__rna_synth_prob__max_p&quot;</span><span class="p">],</span>
            <span class="s2">&quot;plot_title&quot;</span><span class="p">:</span> <span class="s2">&quot;RNA Synth Max Prob&quot;</span><span class="p">,</span>
            <span class="s2">&quot;num_digits_rounding&quot;</span><span class="p">:</span> <span class="mi">4</span><span class="p">,</span>
        <span class="p">},</span>
        <span class="s2">&quot;protein_init_prob_max_p_heatmap&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;columns&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;listeners__ribosome_data__max_p&quot;</span><span class="p">],</span>
            <span class="s2">&quot;plot_title&quot;</span><span class="p">:</span> <span class="s2">&quot;Protein Init Max Prob&quot;</span><span class="p">,</span>
            <span class="s2">&quot;num_digits_rounding&quot;</span><span class="p">:</span> <span class="mi">4</span><span class="p">,</span>
        <span class="p">},</span>
        <span class="s2">&quot;weighted_avg_translation_efficiency_heatmap&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;columns&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;listeners__rna_counts__full_mRNA_cistron_counts&quot;</span><span class="p">],</span>
            <span class="s2">&quot;plot_title&quot;</span><span class="p">:</span> <span class="s2">&quot;Translation Efficiency (Weighted Average)&quot;</span><span class="p">,</span>
            <span class="s2">&quot;num_digits_rounding&quot;</span><span class="p">:</span> <span class="mi">3</span><span class="p">,</span>
            <span class="s2">&quot;custom_sql&quot;</span><span class="p">:</span> <span class="sa">f</span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">                WITH unnested_counts AS (</span>
<span class="s2">                    SELECT unnest(listeners__rna_counts__full_mRNA_cistron_counts) AS array_col,</span>
<span class="s2">                        generate_subscripts(listeners__rna_counts__full_mRNA_cistron_counts, 1) AS array_idx,</span>
<span class="s2">                        unnest(</span><span class="si">{</span><span class="n">ordered_trl_effs</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span><span class="si">}</span><span class="s2">) AS trl_eff,</span>
<span class="s2">                        experiment_id, variant, lineage_seed, generation, agent_id</span>
<span class="s2">                    FROM (</span><span class="se">{{</span><span class="s2">subquery</span><span class="se">}}</span><span class="s2">)</span>
<span class="s2">                ),</span>
<span class="s2">                -- Materialize in memory so we can left join per-gene averages with</span>
<span class="s2">                -- sum of all gene averages without re-reading data</span>
<span class="s2">                avg_per_cell AS MATERIALIZED (</span>
<span class="s2">                    SELECT avg(array_col) AS avg_array_col,</span>
<span class="s2">                        experiment_id, variant, lineage_seed,</span>
<span class="s2">                        generation, agent_id, array_idx, any_value(trl_eff) AS trl_eff</span>
<span class="s2">                    FROM unnested_counts</span>
<span class="s2">                    GROUP BY experiment_id, variant, lineage_seed,</span>
<span class="s2">                        generation, agent_id, array_idx</span>
<span class="s2">                ),</span>
<span class="s2">                total_avg_per_cell AS (</span>
<span class="s2">                    SELECT sum(avg_array_col) AS sum_avg_array_col,</span>
<span class="s2">                        experiment_id, variant, lineage_seed,</span>
<span class="s2">                        generation, agent_id</span>
<span class="s2">                    FROM avg_per_cell</span>
<span class="s2">                    GROUP BY experiment_id, variant, lineage_seed,</span>
<span class="s2">                        generation, agent_id</span>
<span class="s2">                ),</span>
<span class="s2">                weighted_avg_per_cell AS (</span>
<span class="s2">                    SELECT sum(avg_array_col / sum_avg_array_col * trl_eff) AS</span>
<span class="s2">                        weighted_avg_array_col, experiment_id, variant</span>
<span class="s2">                    FROM avg_per_cell</span>
<span class="s2">                    LEFT JOIN total_avg_per_cell USING (experiment_id,</span>
<span class="s2">                        variant, lineage_seed, generation, agent_id)</span>
<span class="s2">                    GROUP BY experiment_id, variant, lineage_seed,</span>
<span class="s2">                        generation, agent_id</span>
<span class="s2">                )</span>
<span class="s2">                SELECT variant, avg(weighted_avg_array_col) AS mean,</span>
<span class="s2">                    stddev(weighted_avg_array_col) AS std</span>
<span class="s2">                FROM weighted_avg_per_cell</span>
<span class="s2">                GROUP BY experiment_id, variant</span>
<span class="s2">                &quot;&quot;&quot;</span><span class="p">,</span>
        <span class="p">},</span>
        <span class="s2">&quot;capacity_gene_mRNA_counts_heatmap&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;plot_title&quot;</span><span class="p">:</span> <span class="s2">&quot;Log(Capacity Gene mRNA Counts+1): &quot;</span><span class="p">,</span>
            <span class="s2">&quot;columns&quot;</span><span class="p">:</span> <span class="p">[</span>
                <span class="n">get_rnas_combined_as_genes_projection</span><span class="p">(</span>
                    <span class="s2">&quot;listeners__rna_counts__mRNA_counts&quot;</span><span class="p">,</span>
                    <span class="n">capacity_gene_mRNA_indexes</span><span class="p">,</span>
                    <span class="s2">&quot;gene_counts&quot;</span><span class="p">,</span>
                <span class="p">)</span>
            <span class="p">],</span>
            <span class="s2">&quot;custom_sql&quot;</span><span class="p">:</span> <span class="n">GENE_COUNTS_SQL</span><span class="p">,</span>
        <span class="p">},</span>
        <span class="s2">&quot;capacity_gene_monomer_counts_heatmap&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;plot_title&quot;</span><span class="p">:</span> <span class="s2">&quot;Log(Capacity Gene Protein Counts+1): &quot;</span><span class="p">,</span>
            <span class="s2">&quot;columns&quot;</span><span class="p">:</span> <span class="p">[</span>
                <span class="s2">&quot;list_select(listeners__monomer_counts, &quot;</span>
                <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">capacity_gene_monomer_indexes</span><span class="si">}</span><span class="s2">) AS gene_counts&quot;</span>
            <span class="p">],</span>
            <span class="s2">&quot;custom_sql&quot;</span><span class="p">:</span> <span class="n">GENE_COUNTS_SQL</span><span class="p">,</span>
        <span class="p">},</span>
        <span class="s2">&quot;capacity_gene_mRNA_mass_fraction_heatmap&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;plot_title&quot;</span><span class="p">:</span> <span class="s2">&quot;Capacity Gene mRNA Mass Fraction: &quot;</span><span class="p">,</span>
            <span class="s2">&quot;num_digits_rounding&quot;</span><span class="p">:</span> <span class="mi">3</span><span class="p">,</span>
            <span class="s2">&quot;columns&quot;</span><span class="p">:</span> <span class="p">[</span>
                <span class="n">get_rnas_combined_as_genes_projection</span><span class="p">(</span>
                    <span class="s2">&quot;listeners__rna_counts__mRNA_counts&quot;</span><span class="p">,</span>
                    <span class="n">capacity_gene_mRNA_indexes</span><span class="p">,</span>
                    <span class="s2">&quot;gene_counts&quot;</span><span class="p">,</span>
                <span class="p">),</span>
                <span class="s2">&quot;listeners__mass__mRna_mass AS mass&quot;</span><span class="p">,</span>
            <span class="p">],</span>
            <span class="s2">&quot;custom_sql&quot;</span><span class="p">:</span> <span class="n">avg_1d_array_over_scalar_sql</span><span class="p">(</span><span class="s2">&quot;gene_counts&quot;</span><span class="p">,</span> <span class="s2">&quot;mass&quot;</span><span class="p">),</span>
            <span class="s2">&quot;post_func&quot;</span><span class="p">:</span> <span class="n">get_gene_mass_prod_func</span><span class="p">(</span>
                <span class="n">sim_data</span><span class="p">,</span> <span class="s2">&quot;mRNA&quot;</span><span class="p">,</span> <span class="n">capacity_gene_mRNA_ids</span>
            <span class="p">),</span>
        <span class="p">},</span>
        <span class="s2">&quot;capacity_gene_monomer_mass_fraction_heatmap&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;plot_title&quot;</span><span class="p">:</span> <span class="s2">&quot;Capacity Gene Protein Mass Fraction: &quot;</span><span class="p">,</span>
            <span class="s2">&quot;num_digits_rounding&quot;</span><span class="p">:</span> <span class="mi">3</span><span class="p">,</span>
            <span class="s2">&quot;columns&quot;</span><span class="p">:</span> <span class="p">[</span>
                <span class="s2">&quot;list_select(listeners__monomer_counts, &quot;</span>
                <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">capacity_gene_monomer_indexes</span><span class="si">}</span><span class="s2">) AS gene_counts&quot;</span><span class="p">,</span>
                <span class="s2">&quot;listeners__mass__protein_mass AS mass&quot;</span><span class="p">,</span>
            <span class="p">],</span>
            <span class="s2">&quot;custom_sql&quot;</span><span class="p">:</span> <span class="n">avg_1d_array_over_scalar_sql</span><span class="p">(</span><span class="s2">&quot;gene_counts&quot;</span><span class="p">,</span> <span class="s2">&quot;mass&quot;</span><span class="p">),</span>
            <span class="s2">&quot;post_func&quot;</span><span class="p">:</span> <span class="n">get_gene_mass_prod_func</span><span class="p">(</span>
                <span class="n">sim_data</span><span class="p">,</span> <span class="s2">&quot;monomer&quot;</span><span class="p">,</span> <span class="n">capacity_gene_monomer_ids</span>
            <span class="p">),</span>
        <span class="p">},</span>
        <span class="s2">&quot;capacity_gene_mRNA_counts_fraction_heatmap&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;columns&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;listeners__rna_counts__mRNA_counts&quot;</span><span class="p">],</span>
            <span class="s2">&quot;plot_title&quot;</span><span class="p">:</span> <span class="s2">&quot;Capacity Gene mRNA Counts Fraction: &quot;</span><span class="p">,</span>
            <span class="s2">&quot;num_digits_rounding&quot;</span><span class="p">:</span> <span class="mi">3</span><span class="p">,</span>
            <span class="s2">&quot;custom_sql&quot;</span><span class="p">:</span> <span class="n">get_gene_count_fraction_sql</span><span class="p">(</span>
                <span class="n">capacity_gene_mRNA_indexes</span><span class="p">,</span>
                <span class="s2">&quot;listeners__rna_counts__mRNA_counts&quot;</span><span class="p">,</span>
                <span class="s2">&quot;mRNA&quot;</span><span class="p">,</span>
            <span class="p">),</span>
        <span class="p">},</span>
        <span class="s2">&quot;capacity_gene_monomer_counts_fraction_heatmap&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;columns&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;listeners__monomer_counts&quot;</span><span class="p">],</span>
            <span class="s2">&quot;plot_title&quot;</span><span class="p">:</span> <span class="s2">&quot;Capacity Gene Protein Counts Fraction: &quot;</span><span class="p">,</span>
            <span class="s2">&quot;num_digits_rounding&quot;</span><span class="p">:</span> <span class="mi">3</span><span class="p">,</span>
            <span class="s2">&quot;custom_sql&quot;</span><span class="p">:</span> <span class="n">get_gene_count_fraction_sql</span><span class="p">(</span>
                <span class="n">capacity_gene_monomer_indexes</span><span class="p">,</span> <span class="s2">&quot;listeners__monomer_counts&quot;</span><span class="p">,</span> <span class="s2">&quot;monomer&quot;</span>
            <span class="p">),</span>
        <span class="p">},</span>
        <span class="s2">&quot;new_gene_copy_number_heatmap&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;plot_title&quot;</span><span class="p">:</span> <span class="s2">&quot;New Gene Copy Number&quot;</span><span class="p">,</span>
            <span class="s2">&quot;columns&quot;</span><span class="p">:</span> <span class="p">[</span>
                <span class="s2">&quot;list_select(listeners__rna_synth_prob__gene_copy_number, &quot;</span>
                <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">new_gene_cistron_indexes</span><span class="si">}</span><span class="s2">) AS gene_counts&quot;</span>
            <span class="p">],</span>
            <span class="s2">&quot;num_digits_rounding&quot;</span><span class="p">:</span> <span class="mi">3</span><span class="p">,</span>
            <span class="s2">&quot;custom_sql&quot;</span><span class="p">:</span> <span class="n">avg_1d_array_sql</span><span class="p">(</span><span class="s2">&quot;gene_counts&quot;</span><span class="p">),</span>
        <span class="p">},</span>
        <span class="s2">&quot;new_gene_mRNA_counts_heatmap&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;plot_title&quot;</span><span class="p">:</span> <span class="s2">&quot;Log(New Gene mRNA Counts+1)&quot;</span><span class="p">,</span>
            <span class="s2">&quot;columns&quot;</span><span class="p">:</span> <span class="p">[</span>
                <span class="n">get_rnas_combined_as_genes_projection</span><span class="p">(</span>
                    <span class="s2">&quot;listeners__rna_counts__mRNA_counts&quot;</span><span class="p">,</span>
                    <span class="n">new_gene_mRNA_indexes</span><span class="p">,</span>
                    <span class="s2">&quot;gene_counts&quot;</span><span class="p">,</span>
                <span class="p">)</span>
            <span class="p">],</span>
            <span class="s2">&quot;custom_sql&quot;</span><span class="p">:</span> <span class="n">GENE_COUNTS_SQL</span><span class="p">,</span>
        <span class="p">},</span>
        <span class="s2">&quot;new_gene_monomer_counts_heatmap&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;plot_title&quot;</span><span class="p">:</span> <span class="s2">&quot;Log(New Gene Protein Counts+1)&quot;</span><span class="p">,</span>
            <span class="s2">&quot;columns&quot;</span><span class="p">:</span> <span class="p">[</span>
                <span class="s2">&quot;list_select(listeners__monomer_counts, &quot;</span>
                <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">new_gene_monomer_indexes</span><span class="si">}</span><span class="s2">) AS gene_counts&quot;</span>
            <span class="p">],</span>
            <span class="s2">&quot;custom_sql&quot;</span><span class="p">:</span> <span class="n">GENE_COUNTS_SQL</span><span class="p">,</span>
        <span class="p">},</span>
        <span class="s2">&quot;new_gene_mRNA_mass_fraction_heatmap&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;plot_title&quot;</span><span class="p">:</span> <span class="s2">&quot;New Gene mRNA Mass Fraction&quot;</span><span class="p">,</span>
            <span class="s2">&quot;columns&quot;</span><span class="p">:</span> <span class="p">[</span>
                <span class="n">get_rnas_combined_as_genes_projection</span><span class="p">(</span>
                    <span class="s2">&quot;listeners__rna_counts__mRNA_counts&quot;</span><span class="p">,</span>
                    <span class="n">new_gene_mRNA_indexes</span><span class="p">,</span>
                    <span class="s2">&quot;gene_counts&quot;</span><span class="p">,</span>
                <span class="p">),</span>
                <span class="s2">&quot;listeners__mass__mRna_mass AS mass&quot;</span><span class="p">,</span>
            <span class="p">],</span>
            <span class="s2">&quot;custom_sql&quot;</span><span class="p">:</span> <span class="n">avg_1d_array_over_scalar_sql</span><span class="p">(</span><span class="s2">&quot;gene_counts&quot;</span><span class="p">,</span> <span class="s2">&quot;mass&quot;</span><span class="p">),</span>
            <span class="s2">&quot;post_func&quot;</span><span class="p">:</span> <span class="n">get_gene_mass_prod_func</span><span class="p">(</span><span class="n">sim_data</span><span class="p">,</span> <span class="s2">&quot;mRNA&quot;</span><span class="p">,</span> <span class="n">new_gene_mRNA_ids</span><span class="p">),</span>
        <span class="p">},</span>
        <span class="s2">&quot;new_gene_mRNA_counts_fraction_heatmap&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;columns&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;listeners__rna_counts__mRNA_counts&quot;</span><span class="p">],</span>
            <span class="s2">&quot;plot_title&quot;</span><span class="p">:</span> <span class="s2">&quot;New Gene mRNA Counts Fraction&quot;</span><span class="p">,</span>
            <span class="s2">&quot;custom_sql&quot;</span><span class="p">:</span> <span class="n">get_gene_count_fraction_sql</span><span class="p">(</span>
                <span class="n">new_gene_mRNA_indexes</span><span class="p">,</span> <span class="s2">&quot;listeners__rna_counts__mRNA_counts&quot;</span><span class="p">,</span> <span class="s2">&quot;mRNA&quot;</span>
            <span class="p">),</span>
        <span class="p">},</span>
        <span class="s2">&quot;new_gene_mRNA_NTP_fraction_heatmap&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;columns&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;listeners__rna_counts__mRNA_counts&quot;</span><span class="p">],</span>
            <span class="s2">&quot;plot_title&quot;</span><span class="p">:</span> <span class="s2">&quot;New Gene&quot;</span><span class="p">,</span>
            <span class="s2">&quot;num_digits_rounding&quot;</span><span class="p">:</span> <span class="mi">4</span><span class="p">,</span>
            <span class="s2">&quot;box_text_size&quot;</span><span class="p">:</span> <span class="s2">&quot;x-small&quot;</span><span class="p">,</span>
            <span class="s2">&quot;custom_sql&quot;</span><span class="p">:</span> <span class="n">get_new_gene_mRNA_NTP_fraction_sql</span><span class="p">(</span>
                <span class="n">sim_data</span><span class="p">,</span> <span class="n">new_gene_mRNA_indexes</span><span class="p">,</span> <span class="n">ntp_ids</span>
            <span class="p">),</span>
            <span class="s2">&quot;is_nonstandard_plot&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
        <span class="p">},</span>
        <span class="s2">&quot;new_gene_monomer_mass_fraction_heatmap&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;plot_title&quot;</span><span class="p">:</span> <span class="s2">&quot;New Gene Protein Mass Fraction&quot;</span><span class="p">,</span>
            <span class="s2">&quot;columns&quot;</span><span class="p">:</span> <span class="p">[</span>
                <span class="s2">&quot;list_select(listeners__monomer_counts, &quot;</span>
                <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">new_gene_monomer_indexes</span><span class="si">}</span><span class="s2">) AS gene_counts&quot;</span><span class="p">,</span>
                <span class="s2">&quot;listeners__mass__protein_mass AS mass&quot;</span><span class="p">,</span>
            <span class="p">],</span>
            <span class="s2">&quot;custom_sql&quot;</span><span class="p">:</span> <span class="n">avg_1d_array_over_scalar_sql</span><span class="p">(</span><span class="s2">&quot;gene_counts&quot;</span><span class="p">,</span> <span class="s2">&quot;mass&quot;</span><span class="p">),</span>
            <span class="s2">&quot;post_func&quot;</span><span class="p">:</span> <span class="n">get_gene_mass_prod_func</span><span class="p">(</span>
                <span class="n">sim_data</span><span class="p">,</span> <span class="s2">&quot;monomer&quot;</span><span class="p">,</span> <span class="n">new_gene_monomer_ids</span>
            <span class="p">),</span>
        <span class="p">},</span>
        <span class="s2">&quot;new_gene_monomer_counts_fraction_heatmap&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;columns&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;listeners__monomer_counts&quot;</span><span class="p">],</span>
            <span class="s2">&quot;plot_title&quot;</span><span class="p">:</span> <span class="s2">&quot;New Gene Protein Counts Fraction&quot;</span><span class="p">,</span>
            <span class="s2">&quot;custom_sql&quot;</span><span class="p">:</span> <span class="n">get_gene_count_fraction_sql</span><span class="p">(</span>
                <span class="n">new_gene_monomer_indexes</span><span class="p">,</span> <span class="s2">&quot;listeners__monomer_counts&quot;</span><span class="p">,</span> <span class="s2">&quot;monomer&quot;</span>
            <span class="p">),</span>
        <span class="p">},</span>
        <span class="s2">&quot;new_gene_rnap_init_rate_heatmap&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;plot_title&quot;</span><span class="p">:</span> <span class="s2">&quot;New Gene RNAP Initialization Rate&quot;</span><span class="p">,</span>
            <span class="s2">&quot;columns&quot;</span><span class="p">:</span> <span class="p">[</span>
                <span class="s2">&quot;list_select(listeners__rna_synth_prob__gene_copy_number, &quot;</span>
                <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">new_gene_cistron_indexes</span><span class="si">}</span><span class="s2">) AS gene_copy_number&quot;</span><span class="p">,</span>
                <span class="s2">&quot;list_select(listeners__rnap_data__rna_init_event_per_cistron, &quot;</span>
                <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">new_gene_cistron_indexes</span><span class="si">}</span><span class="s2">) AS rna_init_event_per_cistron&quot;</span><span class="p">,</span>
            <span class="p">],</span>
            <span class="s2">&quot;custom_sql&quot;</span><span class="p">:</span> <span class="n">avg_ratio_of_1d_arrays_sql</span><span class="p">(</span>
                <span class="s2">&quot;rna_init_event_per_cistron&quot;</span><span class="p">,</span> <span class="s2">&quot;gene_copy_number&quot;</span>
            <span class="p">),</span>
        <span class="p">},</span>
        <span class="s2">&quot;new_gene_ribosome_init_rate_heatmap&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;plot_title&quot;</span><span class="p">:</span> <span class="s2">&quot;New Gene Ribosome Initalization Rate&quot;</span><span class="p">,</span>
            <span class="s2">&quot;columns&quot;</span><span class="p">:</span> <span class="p">[</span>
                <span class="n">get_rnas_combined_as_genes_projection</span><span class="p">(</span>
                    <span class="s2">&quot;listeners__rna_counts__mRNA_counts&quot;</span><span class="p">,</span>
                    <span class="n">new_gene_mRNA_indexes</span><span class="p">,</span>
                    <span class="s2">&quot;mRNA_counts&quot;</span><span class="p">,</span>
                <span class="p">),</span>
                <span class="s2">&quot;list_select(listeners__ribosome_data__ribosome_init_event_per_monomer, &quot;</span>
                <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">new_gene_monomer_indexes</span><span class="si">}</span><span class="s2">) AS ribosome_init_event_per_monomer&quot;</span><span class="p">,</span>
            <span class="p">],</span>
            <span class="s2">&quot;custom_sql&quot;</span><span class="p">:</span> <span class="n">avg_ratio_of_1d_arrays_sql</span><span class="p">(</span>
                <span class="s2">&quot;ribosome_init_event_per_monomer&quot;</span><span class="p">,</span> <span class="s2">&quot;mRNA_counts&quot;</span>
            <span class="p">),</span>
        <span class="p">},</span>
        <span class="s2">&quot;new_gene_rnap_time_overcrowded_heatmap&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;plot_title&quot;</span><span class="p">:</span> <span class="s2">&quot;Fraction of Time RNAP Overcrowded New Gene&quot;</span><span class="p">,</span>
            <span class="c1"># Need to explicitly cast boolean list to numeric list for aggregation</span>
            <span class="s2">&quot;columns&quot;</span><span class="p">:</span> <span class="p">[</span>
                <span class="n">get_rnas_combined_as_genes_projection</span><span class="p">(</span>
                    <span class="s2">&quot;listeners__rna_synth_prob__tu_is_overcrowded&quot;</span><span class="p">,</span>
                    <span class="n">new_gene_RNA_indexes</span><span class="p">,</span>
                    <span class="s2">&quot;overcrowded&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;UTINYINT[]&quot;</span><span class="p">,</span>
                <span class="p">)</span>
            <span class="p">],</span>
            <span class="s2">&quot;custom_sql&quot;</span><span class="p">:</span> <span class="n">avg_1d_array_sql</span><span class="p">(</span><span class="s2">&quot;overcrowded&quot;</span><span class="p">),</span>
        <span class="p">},</span>
        <span class="s2">&quot;new_gene_ribosome_time_overcrowded_heatmap&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;plot_title&quot;</span><span class="p">:</span> <span class="s2">&quot;Fraction of Time Ribosome Overcrowded New Gene&quot;</span><span class="p">,</span>
            <span class="c1"># Need to explicitly cast boolean list to numeric list for aggregation</span>
            <span class="s2">&quot;columns&quot;</span><span class="p">:</span> <span class="p">[</span>
                <span class="s2">&quot;list_select(listeners__ribosome_data__mRNA_is_overcrowded, &quot;</span>
                <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">new_gene_monomer_indexes</span><span class="si">}</span><span class="s2">)::UTINYINT[] AS overcrowded&quot;</span>
            <span class="p">],</span>
            <span class="s2">&quot;custom_sql&quot;</span><span class="p">:</span> <span class="n">avg_1d_array_sql</span><span class="p">(</span><span class="s2">&quot;overcrowded&quot;</span><span class="p">),</span>
        <span class="p">},</span>
        <span class="s2">&quot;new_gene_actual_protein_init_prob_heatmap&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;plot_title&quot;</span><span class="p">:</span> <span class="s2">&quot;New Gene Actual Protein Init Prob&quot;</span><span class="p">,</span>
            <span class="s2">&quot;num_digits_rounding&quot;</span><span class="p">:</span> <span class="mi">4</span><span class="p">,</span>
            <span class="s2">&quot;columns&quot;</span><span class="p">:</span> <span class="p">[</span>
                <span class="s2">&quot;list_select(listeners__ribosome_data__actual_prob_translation_per_transcript, &quot;</span>
                <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">new_gene_monomer_indexes</span><span class="si">}</span><span class="s2">) AS init_prob&quot;</span>
            <span class="p">],</span>
            <span class="s2">&quot;custom_sql&quot;</span><span class="p">:</span> <span class="n">avg_1d_array_sql</span><span class="p">(</span><span class="s2">&quot;init_prob&quot;</span><span class="p">),</span>
        <span class="p">},</span>
        <span class="s2">&quot;new_gene_target_protein_init_prob_heatmap&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;plot_title&quot;</span><span class="p">:</span> <span class="s2">&quot;New Gene Target Protein Init Prob&quot;</span><span class="p">,</span>
            <span class="s2">&quot;num_digits_rounding&quot;</span><span class="p">:</span> <span class="mi">4</span><span class="p">,</span>
            <span class="s2">&quot;columns&quot;</span><span class="p">:</span> <span class="p">[</span>
                <span class="s2">&quot;list_select(listeners__ribosome_data__target_prob_translation_per_transcript, &quot;</span>
                <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">new_gene_monomer_indexes</span><span class="si">}</span><span class="s2">) AS init_prob&quot;</span>
            <span class="p">],</span>
            <span class="s2">&quot;custom_sql&quot;</span><span class="p">:</span> <span class="n">avg_1d_array_sql</span><span class="p">(</span><span class="s2">&quot;init_prob&quot;</span><span class="p">),</span>
        <span class="p">},</span>
        <span class="s2">&quot;new_gene_protein_init_prob_max_p_target_ratio_heatmap&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;plot_title&quot;</span><span class="p">:</span> <span class="s2">&quot;New Gene Protein Max Prob / Target Prob Ratio&quot;</span><span class="p">,</span>
            <span class="s2">&quot;num_digits_rounding&quot;</span><span class="p">:</span> <span class="mi">4</span><span class="p">,</span>
            <span class="s2">&quot;columns&quot;</span><span class="p">:</span> <span class="p">[</span>
                <span class="s2">&quot;list_select(listeners__ribosome_data__target_prob_translation_per_transcript, &quot;</span>
                <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">new_gene_monomer_indexes</span><span class="si">}</span><span class="s2">) AS target_prob&quot;</span><span class="p">,</span>
                <span class="s2">&quot;list_select(listeners__ribosome_data__max_p_per_protein, &quot;</span>
                <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">new_gene_monomer_indexes</span><span class="si">}</span><span class="s2">) AS max_p&quot;</span><span class="p">,</span>
            <span class="p">],</span>
            <span class="s2">&quot;custom_sql&quot;</span><span class="p">:</span> <span class="n">avg_ratio_of_1d_arrays_sql</span><span class="p">(</span><span class="s2">&quot;target_prob&quot;</span><span class="p">,</span> <span class="s2">&quot;max_p&quot;</span><span class="p">),</span>
        <span class="p">},</span>
        <span class="s2">&quot;new_gene_rna_synth_prob_max_p_target_ratio_heatmap&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;plot_title&quot;</span><span class="p">:</span> <span class="s2">&quot;New Gene Protein Max Prob / Target Prob Ratio&quot;</span><span class="p">,</span>
            <span class="s2">&quot;num_digits_rounding&quot;</span><span class="p">:</span> <span class="mi">4</span><span class="p">,</span>
            <span class="s2">&quot;columns&quot;</span><span class="p">:</span> <span class="p">[</span>
                <span class="n">get_rnas_combined_as_genes_projection</span><span class="p">(</span>
                    <span class="s2">&quot;listeners__rna_synth_prob__target_rna_synth_prob&quot;</span><span class="p">,</span>
                    <span class="n">new_gene_RNA_indexes</span><span class="p">,</span>
                    <span class="s2">&quot;target_prob&quot;</span><span class="p">,</span>
                <span class="p">),</span>
                <span class="s2">&quot;listeners__rna_synth_prob__max_p AS max_p&quot;</span><span class="p">,</span>
            <span class="p">],</span>
            <span class="s2">&quot;custom_sql&quot;</span><span class="p">:</span> <span class="n">avg_1d_array_over_scalar_sql</span><span class="p">(</span><span class="s2">&quot;target_prob&quot;</span><span class="p">,</span> <span class="s2">&quot;max_p&quot;</span><span class="p">),</span>
        <span class="p">},</span>
        <span class="s2">&quot;new_gene_ribosome_init_events_heatmap&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;plot_title&quot;</span><span class="p">:</span> <span class="s2">&quot;New Gene Ribosome Init Events Per Time Step&quot;</span><span class="p">,</span>
            <span class="s2">&quot;num_digits_rounding&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
            <span class="s2">&quot;box_test_size&quot;</span><span class="p">:</span> <span class="s2">&quot;x-small&quot;</span><span class="p">,</span>
            <span class="s2">&quot;columns&quot;</span><span class="p">:</span> <span class="p">[</span>
                <span class="s2">&quot;list_select(listeners__ribosome_data__ribosome_init_event_per_monomer, &quot;</span>
                <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">new_gene_monomer_indexes</span><span class="si">}</span><span class="s2">) AS init_events&quot;</span>
            <span class="p">],</span>
            <span class="s2">&quot;custom_sql&quot;</span><span class="p">:</span> <span class="n">avg_1d_array_sql</span><span class="p">(</span><span class="s2">&quot;init_events&quot;</span><span class="p">),</span>
        <span class="p">},</span>
        <span class="s2">&quot;new_gene_ribosome_init_events_portion_heatmap&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;plot_title&quot;</span><span class="p">:</span> <span class="s2">&quot;New Gene Portion of Initiated Ribosomes&quot;</span><span class="p">,</span>
            <span class="s2">&quot;num_digits_rounding&quot;</span><span class="p">:</span> <span class="mi">4</span><span class="p">,</span>
            <span class="s2">&quot;columns&quot;</span><span class="p">:</span> <span class="p">[</span>
                <span class="s2">&quot;list_select(listeners__ribosome_data__ribosome_init_event_per_monomer, &quot;</span>
                <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">new_gene_monomer_indexes</span><span class="si">}</span><span class="s2">) AS init_events&quot;</span><span class="p">,</span>
                <span class="s2">&quot;listeners__ribosome_data__did_initialize AS did_initialize&quot;</span><span class="p">,</span>
            <span class="p">],</span>
            <span class="s2">&quot;custom_sql&quot;</span><span class="p">:</span> <span class="n">avg_1d_array_over_scalar_sql</span><span class="p">(</span><span class="s2">&quot;init_events&quot;</span><span class="p">,</span> <span class="s2">&quot;did_initialize&quot;</span><span class="p">),</span>
        <span class="p">},</span>
        <span class="s2">&quot;new_gene_actual_rna_synth_prob_heatmap&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;plot_title&quot;</span><span class="p">:</span> <span class="s2">&quot;New Gene Actual RNA Synth Prob&quot;</span><span class="p">,</span>
            <span class="s2">&quot;num_digits_rounding&quot;</span><span class="p">:</span> <span class="mi">4</span><span class="p">,</span>
            <span class="s2">&quot;columns&quot;</span><span class="p">:</span> <span class="p">[</span>
                <span class="n">get_rnas_combined_as_genes_projection</span><span class="p">(</span>
                    <span class="s2">&quot;listeners__rna_synth_prob__actual_rna_synth_prob&quot;</span><span class="p">,</span>
                    <span class="n">new_gene_RNA_indexes</span><span class="p">,</span>
                    <span class="s2">&quot;synth_probs&quot;</span><span class="p">,</span>
                <span class="p">)</span>
            <span class="p">],</span>
            <span class="s2">&quot;custom_sql&quot;</span><span class="p">:</span> <span class="n">avg_1d_array_sql</span><span class="p">(</span><span class="s2">&quot;synth_probs&quot;</span><span class="p">),</span>
        <span class="p">},</span>
        <span class="s2">&quot;new_gene_target_rna_synth_prob_heatmap&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;plot_title&quot;</span><span class="p">:</span> <span class="s2">&quot;New Gene Target RNA Synth Prob&quot;</span><span class="p">,</span>
            <span class="s2">&quot;num_digits_rounding&quot;</span><span class="p">:</span> <span class="mi">4</span><span class="p">,</span>
            <span class="s2">&quot;columns&quot;</span><span class="p">:</span> <span class="p">[</span>
                <span class="n">get_rnas_combined_as_genes_projection</span><span class="p">(</span>
                    <span class="s2">&quot;listeners__rna_synth_prob__target_rna_synth_prob&quot;</span><span class="p">,</span>
                    <span class="n">new_gene_RNA_indexes</span><span class="p">,</span>
                    <span class="s2">&quot;synth_probs&quot;</span><span class="p">,</span>
                <span class="p">)</span>
            <span class="p">],</span>
            <span class="s2">&quot;custom_sql&quot;</span><span class="p">:</span> <span class="n">avg_1d_array_sql</span><span class="p">(</span><span class="s2">&quot;synth_probs&quot;</span><span class="p">),</span>
        <span class="p">},</span>
        <span class="s2">&quot;new_gene_rnap_counts_heatmap&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;plot_title&quot;</span><span class="p">:</span> <span class="s2">&quot;New Gene RNAP Counts&quot;</span><span class="p">,</span>
            <span class="s2">&quot;num_digits_rounding&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
            <span class="s2">&quot;box_text_size&quot;</span><span class="p">:</span> <span class="s2">&quot;x-small&quot;</span><span class="p">,</span>
            <span class="s2">&quot;columns&quot;</span><span class="p">:</span> <span class="p">[</span>
                <span class="n">get_rnas_combined_as_genes_projection</span><span class="p">(</span>
                    <span class="s2">&quot;listeners__rna_counts__partial_mRNA_counts&quot;</span><span class="p">,</span>
                    <span class="n">new_gene_mRNA_indexes</span><span class="p">,</span>
                    <span class="s2">&quot;rnap_counts&quot;</span><span class="p">,</span>
                <span class="p">)</span>
            <span class="p">],</span>
            <span class="s2">&quot;custom_sql&quot;</span><span class="p">:</span> <span class="n">avg_1d_array_sql</span><span class="p">(</span><span class="s2">&quot;rnap_counts&quot;</span><span class="p">),</span>
        <span class="p">},</span>
        <span class="s2">&quot;new_gene_rnap_portion_heatmap&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;plot_title&quot;</span><span class="p">:</span> <span class="s2">&quot;New Gene RNAP Portion&quot;</span><span class="p">,</span>
            <span class="s2">&quot;num_digits_rounding&quot;</span><span class="p">:</span> <span class="mi">3</span><span class="p">,</span>
            <span class="s2">&quot;columns&quot;</span><span class="p">:</span> <span class="p">[</span>
                <span class="n">get_rnas_combined_as_genes_projection</span><span class="p">(</span>
                    <span class="s2">&quot;listeners__rna_counts__partial_mRNA_counts&quot;</span><span class="p">,</span>
                    <span class="n">new_gene_mRNA_indexes</span><span class="p">,</span>
                    <span class="s2">&quot;rnap_counts&quot;</span><span class="p">,</span>
                <span class="p">),</span>
                <span class="s2">&quot;listeners__unique_molecule_counts__active_RNAP AS active_counts&quot;</span><span class="p">,</span>
            <span class="p">],</span>
            <span class="s2">&quot;custom_sql&quot;</span><span class="p">:</span> <span class="n">avg_1d_array_over_scalar_sql</span><span class="p">(</span><span class="s2">&quot;rnap_counts&quot;</span><span class="p">,</span> <span class="s2">&quot;active_counts&quot;</span><span class="p">),</span>
        <span class="p">},</span>
        <span class="s2">&quot;rrna_rnap_counts_heatmap&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;plot_title&quot;</span><span class="p">:</span> <span class="s2">&quot;rRNA RNAP Counts&quot;</span><span class="p">,</span>
            <span class="s2">&quot;num_digits_rounding&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
            <span class="s2">&quot;columns&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;listeners__rna_counts__partial_rRNA_counts AS rnap_counts&quot;</span><span class="p">],</span>
            <span class="s2">&quot;custom_sql&quot;</span><span class="p">:</span> <span class="n">avg_sum_1d_array_sql</span><span class="p">(</span><span class="s2">&quot;rnap_counts&quot;</span><span class="p">),</span>
        <span class="p">},</span>
        <span class="s2">&quot;rrna_rnap_portion_heatmap&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;plot_title&quot;</span><span class="p">:</span> <span class="s2">&quot;rRNA RNAP Portion&quot;</span><span class="p">,</span>
            <span class="s2">&quot;num_digits_rounding&quot;</span><span class="p">:</span> <span class="mi">3</span><span class="p">,</span>
            <span class="s2">&quot;columns&quot;</span><span class="p">:</span> <span class="p">[</span>
                <span class="s2">&quot;listeners__rna_counts__partial_rRNA_counts AS rnap_counts&quot;</span><span class="p">,</span>
                <span class="s2">&quot;listeners__unique_molecule_counts__active_RNAP AS active_counts&quot;</span><span class="p">,</span>
            <span class="p">],</span>
            <span class="s2">&quot;custom_sql&quot;</span><span class="p">:</span> <span class="n">avg_sum_1d_array_over_scalar_sql</span><span class="p">(</span>
                <span class="s2">&quot;rnap_counts&quot;</span><span class="p">,</span> <span class="s2">&quot;active_counts&quot;</span>
            <span class="p">),</span>
        <span class="p">},</span>
        <span class="s2">&quot;rnap_subunit_rnap_counts_heatmap&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;plot_title&quot;</span><span class="p">:</span> <span class="s2">&quot;RNAP Subunit RNAP Counts&quot;</span><span class="p">,</span>
            <span class="s2">&quot;num_digits_rounding&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
            <span class="s2">&quot;columns&quot;</span><span class="p">:</span> <span class="p">[</span>
                <span class="s2">&quot;list_select(listeners__rna_counts__partial_mRNA_counts, &quot;</span>
                <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">rnap_subunit_mRNA_indexes</span><span class="si">}</span><span class="s2">) AS rnap_counts&quot;</span>
            <span class="p">],</span>
            <span class="s2">&quot;custom_sql&quot;</span><span class="p">:</span> <span class="n">avg_sum_1d_array_sql</span><span class="p">(</span><span class="s2">&quot;rnap_counts&quot;</span><span class="p">),</span>
        <span class="p">},</span>
        <span class="s2">&quot;rnap_subunit_rnap_portion_heatmap&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;plot_title&quot;</span><span class="p">:</span> <span class="s2">&quot;RNAP Subunit RNAP Portion&quot;</span><span class="p">,</span>
            <span class="s2">&quot;num_digits_rounding&quot;</span><span class="p">:</span> <span class="mi">3</span><span class="p">,</span>
            <span class="s2">&quot;columns&quot;</span><span class="p">:</span> <span class="p">[</span>
                <span class="s2">&quot;list_select(listeners__rna_counts__partial_mRNA_counts, &quot;</span>
                <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">rnap_subunit_mRNA_indexes</span><span class="si">}</span><span class="s2">) AS rnap_counts&quot;</span><span class="p">,</span>
                <span class="s2">&quot;listeners__unique_molecule_counts__active_RNAP AS active_counts&quot;</span><span class="p">,</span>
            <span class="p">],</span>
            <span class="s2">&quot;custom_sql&quot;</span><span class="p">:</span> <span class="n">avg_sum_1d_array_over_scalar_sql</span><span class="p">(</span>
                <span class="s2">&quot;rnap_counts&quot;</span><span class="p">,</span> <span class="s2">&quot;active_counts&quot;</span>
            <span class="p">),</span>
        <span class="p">},</span>
        <span class="s2">&quot;rnap_subunit_ribosome_counts_heatmap&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;plot_title&quot;</span><span class="p">:</span> <span class="s2">&quot;RNAP Subunit Ribosome Counts&quot;</span><span class="p">,</span>
            <span class="s2">&quot;num_digits_rounding&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
            <span class="s2">&quot;box_text_size&quot;</span><span class="p">:</span> <span class="s2">&quot;x-small&quot;</span><span class="p">,</span>
            <span class="s2">&quot;columns&quot;</span><span class="p">:</span> <span class="p">[</span>
                <span class="s2">&quot;list_select(listeners__ribosome_data__n_ribosomes_per_transcript, &quot;</span>
                <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">rnap_subunit_monomer_indexes</span><span class="si">}</span><span class="s2">) AS ribosome_counts&quot;</span>
            <span class="p">],</span>
            <span class="s2">&quot;custom_sql&quot;</span><span class="p">:</span> <span class="n">avg_sum_1d_array_sql</span><span class="p">(</span><span class="s2">&quot;ribosome_counts&quot;</span><span class="p">),</span>
        <span class="p">},</span>
        <span class="s2">&quot;rnap_subunit_ribosome_portion_heatmap&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;plot_title&quot;</span><span class="p">:</span> <span class="s2">&quot;RNAP Subunit Ribosome Portion&quot;</span><span class="p">,</span>
            <span class="s2">&quot;num_digits_rounding&quot;</span><span class="p">:</span> <span class="mi">3</span><span class="p">,</span>
            <span class="s2">&quot;columns&quot;</span><span class="p">:</span> <span class="p">[</span>
                <span class="s2">&quot;list_select(listeners__ribosome_data__n_ribosomes_per_transcript, &quot;</span>
                <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">rnap_subunit_monomer_indexes</span><span class="si">}</span><span class="s2">) AS ribosome_counts&quot;</span><span class="p">,</span>
                <span class="s2">&quot;listeners__unique_molecule_counts__active_ribosome AS active_counts&quot;</span><span class="p">,</span>
            <span class="p">],</span>
            <span class="s2">&quot;custom_sql&quot;</span><span class="p">:</span> <span class="n">avg_sum_1d_array_over_scalar_sql</span><span class="p">(</span>
                <span class="s2">&quot;ribosome_counts&quot;</span><span class="p">,</span> <span class="s2">&quot;active_counts&quot;</span>
            <span class="p">),</span>
        <span class="p">},</span>
        <span class="s2">&quot;ribosomal_protein_rnap_counts_heatmap&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;plot_title&quot;</span><span class="p">:</span> <span class="s2">&quot;Ribosomal Protein RNAP Counts&quot;</span><span class="p">,</span>
            <span class="s2">&quot;num_digits_rounding&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
            <span class="s2">&quot;box_text_size&quot;</span><span class="p">:</span> <span class="s2">&quot;x-small&quot;</span><span class="p">,</span>
            <span class="s2">&quot;columns&quot;</span><span class="p">:</span> <span class="p">[</span>
                <span class="s2">&quot;list_select(listeners__rna_counts__partial_mRNA_counts, &quot;</span>
                <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ribosomal_mRNA_indexes</span><span class="si">}</span><span class="s2">) AS rnap_counts&quot;</span>
            <span class="p">],</span>
            <span class="s2">&quot;custom_sql&quot;</span><span class="p">:</span> <span class="n">avg_sum_1d_array_sql</span><span class="p">(</span><span class="s2">&quot;rnap_counts&quot;</span><span class="p">),</span>
        <span class="p">},</span>
        <span class="s2">&quot;ribosomal_protein_rnap_portion_heatmap&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;plot_title&quot;</span><span class="p">:</span> <span class="s2">&quot;Ribosomal Protein RNAP Portion&quot;</span><span class="p">,</span>
            <span class="s2">&quot;num_digits_rounding&quot;</span><span class="p">:</span> <span class="mi">3</span><span class="p">,</span>
            <span class="s2">&quot;columns&quot;</span><span class="p">:</span> <span class="p">[</span>
                <span class="s2">&quot;list_select(listeners__rna_counts__partial_mRNA_counts, &quot;</span>
                <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ribosomal_mRNA_indexes</span><span class="si">}</span><span class="s2">) AS rnap_counts&quot;</span><span class="p">,</span>
                <span class="s2">&quot;listeners__unique_molecule_counts__active_RNAP AS active_counts&quot;</span><span class="p">,</span>
            <span class="p">],</span>
            <span class="s2">&quot;custom_sql&quot;</span><span class="p">:</span> <span class="n">avg_sum_1d_array_over_scalar_sql</span><span class="p">(</span>
                <span class="s2">&quot;rnap_counts&quot;</span><span class="p">,</span> <span class="s2">&quot;active_counts&quot;</span>
            <span class="p">),</span>
        <span class="p">},</span>
        <span class="s2">&quot;ribosomal_protein_ribosome_counts_heatmap&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;plot_title&quot;</span><span class="p">:</span> <span class="s2">&quot;Ribosomal Protein Ribosome Counts&quot;</span><span class="p">,</span>
            <span class="s2">&quot;num_digits_rounding&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
            <span class="s2">&quot;box_text_size&quot;</span><span class="p">:</span> <span class="s2">&quot;x-small&quot;</span><span class="p">,</span>
            <span class="s2">&quot;columns&quot;</span><span class="p">:</span> <span class="p">[</span>
                <span class="s2">&quot;list_select(listeners__ribosome_data__n_ribosomes_per_transcript, &quot;</span>
                <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ribosomal_monomer_indexes</span><span class="si">}</span><span class="s2">) AS ribosome_counts&quot;</span>
            <span class="p">],</span>
            <span class="s2">&quot;custom_sql&quot;</span><span class="p">:</span> <span class="n">avg_sum_1d_array_sql</span><span class="p">(</span><span class="s2">&quot;ribosome_counts&quot;</span><span class="p">),</span>
        <span class="p">},</span>
        <span class="s2">&quot;ribosomal_protein_ribosome_portion_heatmap&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;plot_title&quot;</span><span class="p">:</span> <span class="s2">&quot;Ribosomal Protein Ribosome Portion&quot;</span><span class="p">,</span>
            <span class="s2">&quot;num_digits_rounding&quot;</span><span class="p">:</span> <span class="mi">3</span><span class="p">,</span>
            <span class="s2">&quot;columns&quot;</span><span class="p">:</span> <span class="p">[</span>
                <span class="s2">&quot;list_select(listeners__ribosome_data__n_ribosomes_per_transcript, &quot;</span>
                <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ribosomal_monomer_indexes</span><span class="si">}</span><span class="s2">) AS ribosome_counts&quot;</span><span class="p">,</span>
                <span class="s2">&quot;listeners__unique_molecule_counts__active_ribosome AS active_counts&quot;</span><span class="p">,</span>
            <span class="p">],</span>
            <span class="s2">&quot;custom_sql&quot;</span><span class="p">:</span> <span class="n">avg_sum_1d_array_over_scalar_sql</span><span class="p">(</span>
                <span class="s2">&quot;ribosome_counts&quot;</span><span class="p">,</span> <span class="s2">&quot;active_counts&quot;</span>
            <span class="p">),</span>
        <span class="p">},</span>
        <span class="s2">&quot;new_gene_ribosome_counts_heatmap&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;plot_title&quot;</span><span class="p">:</span> <span class="s2">&quot;New Gene Ribosome Counts&quot;</span><span class="p">,</span>
            <span class="s2">&quot;num_digits_rounding&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
            <span class="s2">&quot;box_text_size&quot;</span><span class="p">:</span> <span class="s2">&quot;x-small&quot;</span><span class="p">,</span>
            <span class="s2">&quot;columns&quot;</span><span class="p">:</span> <span class="p">[</span>
                <span class="s2">&quot;list_select(listeners__ribosome_data__n_ribosomes_per_transcript, &quot;</span>
                <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">new_gene_monomer_indexes</span><span class="si">}</span><span class="s2">) AS ribosome_counts&quot;</span>
            <span class="p">],</span>
            <span class="s2">&quot;custom_sql&quot;</span><span class="p">:</span> <span class="n">avg_1d_array_sql</span><span class="p">(</span><span class="s2">&quot;ribosome_counts&quot;</span><span class="p">),</span>
        <span class="p">},</span>
        <span class="s2">&quot;new_gene_ribosome_portion_heatmap&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;plot_title&quot;</span><span class="p">:</span> <span class="s2">&quot;New Gene Ribosome Portion&quot;</span><span class="p">,</span>
            <span class="s2">&quot;num_digits_rounding&quot;</span><span class="p">:</span> <span class="mi">3</span><span class="p">,</span>
            <span class="s2">&quot;columns&quot;</span><span class="p">:</span> <span class="p">[</span>
                <span class="s2">&quot;list_select(listeners__ribosome_data__n_ribosomes_per_transcript, &quot;</span>
                <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">new_gene_monomer_indexes</span><span class="si">}</span><span class="s2">) AS ribosome_counts&quot;</span><span class="p">,</span>
                <span class="s2">&quot;listeners__unique_molecule_counts__active_ribosome AS active_counts&quot;</span><span class="p">,</span>
            <span class="p">],</span>
            <span class="s2">&quot;custom_sql&quot;</span><span class="p">:</span> <span class="n">avg_1d_array_over_scalar_sql</span><span class="p">(</span>
                <span class="s2">&quot;ribosome_counts&quot;</span><span class="p">,</span> <span class="s2">&quot;active_counts&quot;</span>
            <span class="p">),</span>
        <span class="p">},</span>
        <span class="s2">&quot;capacity_gene_rnap_portion_heatmap&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;plot_title&quot;</span><span class="p">:</span> <span class="s2">&quot;Capacity Gene RNAP Portion: &quot;</span><span class="p">,</span>
            <span class="s2">&quot;num_digits_rounding&quot;</span><span class="p">:</span> <span class="mi">4</span><span class="p">,</span>
            <span class="s2">&quot;columns&quot;</span><span class="p">:</span> <span class="p">[</span>
                <span class="n">get_rnas_combined_as_genes_projection</span><span class="p">(</span>
                    <span class="s2">&quot;listeners__rna_counts__partial_mRNA_counts&quot;</span><span class="p">,</span>
                    <span class="n">capacity_gene_mRNA_indexes</span><span class="p">,</span>
                    <span class="s2">&quot;rnap_counts&quot;</span><span class="p">,</span>
                <span class="p">),</span>
                <span class="s2">&quot;listeners__unique_molecule_counts__active_RNAP AS active_counts&quot;</span><span class="p">,</span>
            <span class="p">],</span>
            <span class="s2">&quot;custom_sql&quot;</span><span class="p">:</span> <span class="n">avg_1d_array_over_scalar_sql</span><span class="p">(</span><span class="s2">&quot;rnap_counts&quot;</span><span class="p">,</span> <span class="s2">&quot;active_counts&quot;</span><span class="p">),</span>
        <span class="p">},</span>
        <span class="s2">&quot;capacity_gene_ribosome_portion_heatmap&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;plot_title&quot;</span><span class="p">:</span> <span class="s2">&quot;Capacity Gene Ribosome Portion: &quot;</span><span class="p">,</span>
            <span class="s2">&quot;num_digits_rounding&quot;</span><span class="p">:</span> <span class="mi">4</span><span class="p">,</span>
            <span class="s2">&quot;columns&quot;</span><span class="p">:</span> <span class="p">[</span>
                <span class="s2">&quot;list_select(listeners__ribosome_data__n_ribosomes_per_transcript, &quot;</span>
                <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">capacity_gene_monomer_indexes</span><span class="si">}</span><span class="s2">) AS ribosome_counts&quot;</span><span class="p">,</span>
                <span class="s2">&quot;listeners__unique_molecule_counts__active_ribosome AS active_counts&quot;</span><span class="p">,</span>
            <span class="p">],</span>
            <span class="s2">&quot;custom_sql&quot;</span><span class="p">:</span> <span class="n">avg_1d_array_over_scalar_sql</span><span class="p">(</span>
                <span class="s2">&quot;ribosome_counts&quot;</span><span class="p">,</span> <span class="s2">&quot;active_counts&quot;</span>
            <span class="p">),</span>
        <span class="p">},</span>
        <span class="s2">&quot;glucose_consumption_rate&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;plot_title&quot;</span><span class="p">:</span> <span class="s2">&quot;Average Glucose Consumption Rate (fg/hr)&quot;</span><span class="p">,</span>
            <span class="s2">&quot;num_digits_rounding&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span>
            <span class="s2">&quot;columns&quot;</span><span class="p">:</span> <span class="p">[</span>
                <span class="s2">&quot;-listeners__fba_results__external_exchange_fluxes[&quot;</span>
                <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">glucose_idx</span><span class="si">}</span><span class="s2">] AS glucose_flux&quot;</span><span class="p">,</span>
                <span class="s2">&quot;listeners__mass__dry_mass AS dry_mass&quot;</span><span class="p">,</span>
            <span class="p">],</span>
            <span class="s2">&quot;remove_first&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
            <span class="s2">&quot;custom_sql&quot;</span><span class="p">:</span> <span class="sa">f</span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">                WITH avg_per_cell AS (</span>
<span class="s2">                    SELECT avg(glucose_flux * dry_mass) * </span><span class="si">{</span><span class="n">flux_scaling_factor</span><span class="si">}</span>
<span class="s2">                        AS avg_fg_glc_per_hr, experiment_id, variant</span>
<span class="s2">                    FROM (</span><span class="se">{{</span><span class="s2">subquery</span><span class="se">}}</span><span class="s2">)</span>
<span class="s2">                    GROUP BY experiment_id, variant, lineage_seed, generation, agent_id</span>
<span class="s2">                )</span>
<span class="s2">                SELECT variant, avg(avg_fg_glc_per_hr) AS mean,</span>
<span class="s2">                    stddev(avg_fg_glc_per_hr) AS std</span>
<span class="s2">                FROM avg_per_cell</span>
<span class="s2">                GROUP BY experiment_id, variant</span>
<span class="s2">                &quot;&quot;&quot;</span><span class="p">,</span>
        <span class="p">},</span>
        <span class="s2">&quot;new_gene_yield_per_glucose&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;plot_title&quot;</span><span class="p">:</span> <span class="s2">&quot;New Gene fg Protein Yield per fg Glucose&quot;</span><span class="p">,</span>
            <span class="s2">&quot;num_digits_rounding&quot;</span><span class="p">:</span> <span class="mi">3</span><span class="p">,</span>
            <span class="s2">&quot;columns&quot;</span><span class="p">:</span> <span class="p">[</span>
                <span class="s2">&quot;-listeners__fba_results__external_exchange_fluxes[&quot;</span>
                <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">glucose_idx</span><span class="si">}</span><span class="s2">] AS glucose_flux&quot;</span><span class="p">,</span>
                <span class="s2">&quot;listeners__mass__dry_mass AS dry_mass&quot;</span><span class="p">,</span>
                <span class="s2">&quot;time&quot;</span><span class="p">,</span>
                <span class="s2">&quot;list_select(listeners__monomer_counts, &quot;</span>
                <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">new_gene_monomer_indexes</span><span class="si">}</span><span class="s2">) AS monomer_counts&quot;</span><span class="p">,</span>
            <span class="p">],</span>
            <span class="s2">&quot;remove_first&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
            <span class="s2">&quot;custom_sql&quot;</span><span class="p">:</span> <span class="sa">f</span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">                WITH unnested_counts AS (</span>
<span class="s2">                    SELECT unnest(monomer_counts) AS monomer_counts,</span>
<span class="s2">                        generate_subscripts(monomer_counts, 1) AS monomer_idx,</span>
<span class="s2">                        glucose_flux, dry_mass, time, experiment_id, variant,</span>
<span class="s2">                        lineage_seed, generation, agent_id</span>
<span class="s2">                    FROM (</span><span class="se">{{</span><span class="s2">subquery</span><span class="se">}}</span><span class="s2">)</span>
<span class="s2">                ),</span>
<span class="s2">                avg_per_cell AS (</span>
<span class="s2">                    SELECT avg(glucose_flux * dry_mass) * </span><span class="si">{</span><span class="n">flux_scaling_factor</span><span class="si">}</span>
<span class="s2">                        AS avg_fg_glc_per_hr, (max(time) - min(time)) / 60</span>
<span class="s2">                        AS doubling_time, last(monomer_counts::BIGINT ORDER BY time) -</span>
<span class="s2">                        first(monomer_counts::BIGINT ORDER BY time) AS monomer_delta,</span>
<span class="s2">                        experiment_id, variant, monomer_idx</span>
<span class="s2">                    FROM unnested_counts</span>
<span class="s2">                    GROUP BY experiment_id, variant, lineage_seed, generation,</span>
<span class="s2">                        agent_id, monomer_idx</span>
<span class="s2">                ),</span>
<span class="s2">                avg_per_variant AS (</span>
<span class="s2">                    SELECT experiment_id, variant, avg(monomer_delta / (</span>
<span class="s2">                        avg_fg_glc_per_hr * doubling_time))</span>
<span class="s2">                        AS avg_monomer_per_fg_glc, monomer_idx,</span>
<span class="s2">                        stddev(monomer_delta / (avg_fg_glc_per_hr *</span>
<span class="s2">                        doubling_time)) AS std_monomer_per_fg_glc</span>
<span class="s2">                    FROM avg_per_cell</span>
<span class="s2">                    GROUP BY experiment_id, variant, monomer_idx</span>
<span class="s2">                )</span>
<span class="s2">                SELECT variant, list(avg_monomer_per_fg_glc ORDER BY monomer_idx)</span>
<span class="s2">                    AS mean, list(std_monomer_per_fg_glc ORDER BY monomer_idx) AS std</span>
<span class="s2">                FROM avg_per_variant</span>
<span class="s2">                GROUP BY experiment_id, variant</span>
<span class="s2">                &quot;&quot;&quot;</span><span class="p">,</span>
            <span class="s2">&quot;post_func&quot;</span><span class="p">:</span> <span class="n">get_gene_mass_prod_func</span><span class="p">(</span>
                <span class="n">sim_data</span><span class="p">,</span> <span class="s2">&quot;monomer&quot;</span><span class="p">,</span> <span class="n">new_gene_monomer_ids</span>
            <span class="p">),</span>
        <span class="p">},</span>
        <span class="s2">&quot;new_gene_yield_per_hour&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;plot_title&quot;</span><span class="p">:</span> <span class="s2">&quot;New Gene fg Protein Yield per Hour&quot;</span><span class="p">,</span>
            <span class="s2">&quot;num_digits_rounding&quot;</span><span class="p">:</span> <span class="mi">2</span><span class="p">,</span>
            <span class="s2">&quot;columns&quot;</span><span class="p">:</span> <span class="p">[</span>
                <span class="s2">&quot;time&quot;</span><span class="p">,</span>
                <span class="s2">&quot;list_select(listeners__monomer_counts, &quot;</span>
                <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">new_gene_monomer_indexes</span><span class="si">}</span><span class="s2">) AS monomer_counts&quot;</span><span class="p">,</span>
            <span class="p">],</span>
            <span class="s2">&quot;remove_first&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
            <span class="s2">&quot;custom_sql&quot;</span><span class="p">:</span> <span class="s2">&quot;&quot;&quot;</span>
<span class="s2">                WITH unnested_counts AS (</span>
<span class="s2">                    SELECT unnest(monomer_counts) AS monomer_counts, time,</span>
<span class="s2">                        generate_subscripts(monomer_counts, 1) AS monomer_idx,</span>
<span class="s2">                        experiment_id, variant, lineage_seed, generation, agent_id</span>
<span class="s2">                    FROM (</span><span class="si">{subquery}</span><span class="s2">)</span>
<span class="s2">                ),</span>
<span class="s2">                avg_per_cell AS (</span>
<span class="s2">                    SELECT (max(time) - min(time)) / 60 AS doubling_time,</span>
<span class="s2">                        last(monomer_counts::BIGINT ORDER BY time) - first(</span>
<span class="s2">                            monomer_counts::BIGINT ORDER BY time) AS monomer_delta,</span>
<span class="s2">                        experiment_id, variant, monomer_idx</span>
<span class="s2">                    FROM unnested_counts</span>
<span class="s2">                    GROUP BY experiment_id, variant, lineage_seed, generation,</span>
<span class="s2">                        agent_id, monomer_idx</span>
<span class="s2">                ),</span>
<span class="s2">                avg_per_variant AS (</span>
<span class="s2">                    SELECT experiment_id, variant, avg(monomer_delta /</span>
<span class="s2">                        doubling_time) AS avg_monomer_per_hr, monomer_idx,</span>
<span class="s2">                        stddev(monomer_delta / doubling_time) AS std_monomer_per_hr</span>
<span class="s2">                    FROM avg_per_cell</span>
<span class="s2">                    GROUP BY experiment_id, variant, monomer_idx</span>
<span class="s2">                )</span>
<span class="s2">                SELECT variant, list(avg_monomer_per_hr ORDER BY monomer_idx)</span>
<span class="s2">                    AS mean, list(std_monomer_per_hr ORDER BY monomer_idx) AS std</span>
<span class="s2">                FROM avg_per_variant</span>
<span class="s2">                GROUP BY experiment_id, variant</span>
<span class="s2">                &quot;&quot;&quot;</span><span class="p">,</span>
            <span class="s2">&quot;post_func&quot;</span><span class="p">:</span> <span class="n">get_gene_mass_prod_func</span><span class="p">(</span>
                <span class="n">sim_data</span><span class="p">,</span> <span class="s2">&quot;monomer&quot;</span><span class="p">,</span> <span class="n">new_gene_monomer_ids</span>
            <span class="p">),</span>
        <span class="p">},</span>
    <span class="p">}</span>

    <span class="c1"># Check validity of requested heatmaps and fill in default values where needed</span>
    <span class="n">heatmaps_to_make</span> <span class="o">=</span> <span class="n">HEATMAPS_TO_MAKE_LIST</span>
    <span class="n">total_heatmaps_to_make</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">for</span> <span class="n">h</span> <span class="ow">in</span> <span class="n">heatmaps_to_make</span><span class="p">:</span>
        <span class="k">assert</span> <span class="n">h</span> <span class="ow">in</span> <span class="n">heatmap_details</span><span class="p">,</span> <span class="s2">&quot;Heatmap &quot;</span> <span class="o">+</span> <span class="n">h</span> <span class="o">+</span> <span class="s2">&quot; is not an option&quot;</span>
        <span class="n">heatmap_details</span><span class="p">[</span><span class="n">h</span><span class="p">]</span><span class="o">.</span><span class="n">setdefault</span><span class="p">(</span><span class="s2">&quot;is_nonstandard_plot&quot;</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>
        <span class="n">heatmap_details</span><span class="p">[</span><span class="n">h</span><span class="p">]</span><span class="o">.</span><span class="n">setdefault</span><span class="p">(</span><span class="s2">&quot;remove_first&quot;</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>
        <span class="n">heatmap_details</span><span class="p">[</span><span class="n">h</span><span class="p">]</span><span class="o">.</span><span class="n">setdefault</span><span class="p">(</span><span class="s2">&quot;num_digits_rounding&quot;</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>
        <span class="n">heatmap_details</span><span class="p">[</span><span class="n">h</span><span class="p">]</span><span class="o">.</span><span class="n">setdefault</span><span class="p">(</span><span class="s2">&quot;box_text_size&quot;</span><span class="p">,</span> <span class="s2">&quot;medium&quot;</span><span class="p">)</span>
        <span class="n">heatmap_details</span><span class="p">[</span><span class="n">h</span><span class="p">]</span><span class="o">.</span><span class="n">setdefault</span><span class="p">(</span><span class="s2">&quot;is_new_gene_heatmap&quot;</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>
        <span class="n">heatmap_details</span><span class="p">[</span><span class="n">h</span><span class="p">]</span><span class="o">.</span><span class="n">setdefault</span><span class="p">(</span><span class="s2">&quot;is_capacity_gene_heatmap&quot;</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>
        <span class="n">heatmap_details</span><span class="p">[</span><span class="n">h</span><span class="p">]</span><span class="o">.</span><span class="n">setdefault</span><span class="p">(</span><span class="s2">&quot;default_value&quot;</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">)</span>
        <span class="n">heatmap_details</span><span class="p">[</span><span class="n">h</span><span class="p">]</span><span class="o">.</span><span class="n">setdefault</span><span class="p">(</span><span class="s2">&quot;order_results&quot;</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>
        <span class="n">heatmap_details</span><span class="p">[</span><span class="n">h</span><span class="p">]</span><span class="o">.</span><span class="n">setdefault</span><span class="p">(</span><span class="s2">&quot;sucess_sql&quot;</span><span class="p">,</span> <span class="n">success_sql</span><span class="p">)</span>
        <span class="n">heatmap_details</span><span class="p">[</span><span class="n">h</span><span class="p">]</span><span class="o">.</span><span class="n">setdefault</span><span class="p">(</span><span class="s2">&quot;custom_sql&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
        <span class="n">heatmap_details</span><span class="p">[</span><span class="n">h</span><span class="p">]</span><span class="o">.</span><span class="n">setdefault</span><span class="p">(</span><span class="s2">&quot;post_func&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">h</span> <span class="o">==</span> <span class="s2">&quot;new_gene_mRNA_NTP_fraction_heatmap&quot;</span><span class="p">:</span>
            <span class="n">total_heatmaps_to_make</span> <span class="o">+=</span> <span class="nb">len</span><span class="p">(</span><span class="n">ntp_ids</span><span class="p">)</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">new_gene_cistron_ids</span><span class="p">)</span>
            <span class="n">heatmap_details</span><span class="p">[</span><span class="n">h</span><span class="p">][</span><span class="s2">&quot;default_value&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">full</span><span class="p">(</span>
                <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">ntp_ids</span><span class="p">),</span> <span class="nb">len</span><span class="p">(</span><span class="n">new_gene_cistron_ids</span><span class="p">)),</span> <span class="o">-</span><span class="mi">1</span>
            <span class="p">)</span>
        <span class="k">elif</span> <span class="n">h</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;new_gene_&quot;</span><span class="p">):</span>
            <span class="n">total_heatmaps_to_make</span> <span class="o">+=</span> <span class="nb">len</span><span class="p">(</span><span class="n">new_gene_cistron_ids</span><span class="p">)</span>
            <span class="n">heatmap_details</span><span class="p">[</span><span class="n">h</span><span class="p">][</span><span class="s2">&quot;is_new_gene_heatmap&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="kc">True</span>
            <span class="n">heatmap_details</span><span class="p">[</span><span class="n">h</span><span class="p">][</span><span class="s2">&quot;default_value&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">full</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">new_gene_mRNA_ids</span><span class="p">),</span> <span class="o">-</span><span class="mi">1</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">h</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;capacity_gene_&quot;</span><span class="p">):</span>
            <span class="n">total_heatmaps_to_make</span> <span class="o">+=</span> <span class="nb">len</span><span class="p">(</span><span class="n">capacity_gene_monomer_ids</span><span class="p">)</span>
            <span class="n">heatmap_details</span><span class="p">[</span><span class="n">h</span><span class="p">][</span><span class="s2">&quot;is_capacity_gene_heatmap&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="kc">True</span>
            <span class="n">heatmap_details</span><span class="p">[</span><span class="n">h</span><span class="p">][</span><span class="s2">&quot;default_value&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">full</span><span class="p">(</span>
                <span class="nb">len</span><span class="p">(</span><span class="n">capacity_gene_monomer_ids</span><span class="p">),</span> <span class="o">-</span><span class="mi">1</span>
            <span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">total_heatmaps_to_make</span> <span class="o">+=</span> <span class="mi">1</span>

    <span class="c1"># Data extraction</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;---Data Extraction---&quot;</span><span class="p">)</span>
    <span class="n">heatmap_data</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">h</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="n">heatmaps_to_make</span><span class="p">):</span>
        <span class="n">h_details</span> <span class="o">=</span> <span class="n">heatmap_details</span><span class="p">[</span><span class="n">h</span><span class="p">]</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">h</span><span class="p">)</span>
        <span class="n">mean_matrix</span><span class="p">,</span> <span class="n">std_matrix</span> <span class="o">=</span> <span class="n">get_mean_and_std_matrices</span><span class="p">(</span>
            <span class="n">conn</span><span class="p">,</span>
            <span class="n">variant_to_row_col</span><span class="p">,</span>
            <span class="n">variant_matrix_shape</span><span class="p">,</span>
            <span class="n">history_sql</span><span class="p">,</span>
            <span class="n">h_details</span><span class="p">[</span><span class="s2">&quot;columns&quot;</span><span class="p">],</span>
            <span class="n">h_details</span><span class="p">[</span><span class="s2">&quot;remove_first&quot;</span><span class="p">],</span>
            <span class="kc">None</span><span class="p">,</span>
            <span class="n">h_details</span><span class="p">[</span><span class="s2">&quot;order_results&quot;</span><span class="p">],</span>
            <span class="n">h_details</span><span class="p">[</span><span class="s2">&quot;sucess_sql&quot;</span><span class="p">],</span>
            <span class="n">h_details</span><span class="p">[</span><span class="s2">&quot;custom_sql&quot;</span><span class="p">],</span>
            <span class="n">h_details</span><span class="p">[</span><span class="s2">&quot;post_func&quot;</span><span class="p">],</span>
            <span class="n">h_details</span><span class="p">[</span><span class="s2">&quot;num_digits_rounding&quot;</span><span class="p">],</span>
            <span class="n">h_details</span><span class="p">[</span><span class="s2">&quot;default_value&quot;</span><span class="p">],</span>
        <span class="p">)</span>
        <span class="n">heatmap_data</span><span class="p">[</span><span class="n">h</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;mean&quot;</span><span class="p">:</span> <span class="n">mean_matrix</span><span class="p">,</span> <span class="s2">&quot;std_dev&quot;</span><span class="p">:</span> <span class="n">std_matrix</span><span class="p">}</span>

    <span class="c1"># Plotting</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;---Plotting---&quot;</span><span class="p">)</span>
    <span class="n">plot_suffix</span> <span class="o">=</span> <span class="s2">&quot;_gens_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">MIN_CELL_INDEX</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;_through_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">MAX_CELL_INDEX</span><span class="p">)</span>
    <span class="n">heatmap_x_label</span> <span class="o">=</span> <span class="s2">&quot;Expression Variant&quot;</span>
    <span class="n">heatmap_y_label</span> <span class="o">=</span> <span class="s2">&quot;Translation Efficiency Value&quot;</span>
    <span class="n">figsize_x</span> <span class="o">=</span> <span class="mi">2</span> <span class="o">+</span> <span class="mi">2</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">new_gene_expression_factors</span><span class="p">)</span> <span class="o">/</span> <span class="mi">3</span>
    <span class="n">figsize_y</span> <span class="o">=</span> <span class="mi">2</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">new_gene_translation_efficiency_values</span><span class="p">)</span> <span class="o">/</span> <span class="mi">2</span>

    <span class="c1"># Figure out whether to create dashboard / individual plots and</span>
    <span class="c1"># whether to make std. dev. plots in addition to mean plots</span>
    <span class="n">summary_statistics</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;mean&quot;</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">dashboard_flag</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">is_dashboards</span> <span class="o">=</span> <span class="p">[</span><span class="kc">False</span><span class="p">]</span>
    <span class="k">elif</span> <span class="n">dashboard_flag</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
        <span class="n">is_dashboards</span> <span class="o">=</span> <span class="p">[</span><span class="kc">True</span><span class="p">]</span>
    <span class="k">elif</span> <span class="n">dashboard_flag</span> <span class="o">==</span> <span class="mi">2</span><span class="p">:</span>
        <span class="n">is_dashboards</span> <span class="o">=</span> <span class="p">[</span><span class="kc">True</span><span class="p">,</span> <span class="kc">False</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">std_dev_flag</span><span class="p">:</span>
        <span class="n">summary_statistics</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;std_dev&quot;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">is_dashboard</span> <span class="ow">in</span> <span class="n">is_dashboards</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">summary_statistic</span> <span class="ow">in</span> <span class="n">summary_statistics</span><span class="p">:</span>
            <span class="n">plot_heatmaps</span><span class="p">(</span>
                <span class="n">heatmap_data</span><span class="p">,</span>
                <span class="n">heatmap_details</span><span class="p">,</span>
                <span class="n">new_gene_cistron_ids</span><span class="p">,</span>
                <span class="n">ntp_ids</span><span class="p">,</span>
                <span class="n">capacity_gene_common_names</span><span class="p">,</span>
                <span class="n">total_heatmaps_to_make</span><span class="p">,</span>
                <span class="n">is_dashboard</span><span class="p">,</span>
                <span class="n">variant_mask</span><span class="p">,</span>
                <span class="n">heatmap_x_label</span><span class="p">,</span>
                <span class="n">heatmap_y_label</span><span class="p">,</span>
                <span class="n">new_gene_expression_factors</span><span class="p">,</span>
                <span class="n">new_gene_translation_efficiency_values</span><span class="p">,</span>
                <span class="n">summary_statistic</span><span class="p">,</span>
                <span class="n">figsize_x</span><span class="p">,</span>
                <span class="n">figsize_y</span><span class="p">,</span>
                <span class="n">outdir</span><span class="p">,</span>
                <span class="n">plot_suffix</span><span class="p">,</span>
            <span class="p">)</span></div>

</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2021-2025, The Vivarium E. coli Authors.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>