

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>ecoli.library.schema &mdash; Vivarium E. coli 1.0.0 documentation</title>
      <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=b86133f3" />
      <link rel="stylesheet" type="text/css" href="../../../_static/css/theme.css?v=e59714d7" />

  
      <script src="../../../_static/jquery.js?v=5d32c60e"></script>
      <script src="../../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../../../_static/documentation_options.js?v=8d563738"></script>
      <script src="../../../_static/doctools.js?v=9bcbadda"></script>
      <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
      <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="ecoli.library.serialize" href="ecoli.library.serialize.html" />
    <link rel="prev" title="ecoli.library.pymunk_multibody" href="ecoli.library.pymunk_multibody.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../index.html" class="icon icon-home">
            Vivarium E. coli
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../../stores.html">Stores</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../processes.html">Processes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../composites.html">Composites</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../experiments.html">Experiments</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../workflows.html">Workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../output.html">Output</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../tutorial.html">Tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../docs.html">Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../hpc.html">HPC Clusters</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../gcloud.html">Google Cloud</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../ci.html">Continuous Integration</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../pycharm.html">PyCharm Setup</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../../api_ref.html">API Reference</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="ecoli.html"><code class="docutils literal notranslate"><span class="pre">ecoli</span></code></a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="ecoli.analysis.html"><code class="docutils literal notranslate"><span class="pre">ecoli.analysis</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="ecoli.composites.html"><code class="docutils literal notranslate"><span class="pre">ecoli.composites</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="ecoli.experiments.html"><code class="docutils literal notranslate"><span class="pre">ecoli.experiments</span></code></a></li>
<li class="toctree-l3 current"><a class="reference internal" href="ecoli.library.html"><code class="docutils literal notranslate"><span class="pre">ecoli.library</span></code></a><ul class="current">
<li class="toctree-l4"><a class="reference internal" href="ecoli.library.cell_wall.html"><code class="docutils literal notranslate"><span class="pre">ecoli.library.cell_wall</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.library.create_timeline.html"><code class="docutils literal notranslate"><span class="pre">ecoli.library.create_timeline</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.library.data_predicates.html"><code class="docutils literal notranslate"><span class="pre">ecoli.library.data_predicates</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.library.initial_conditions.html"><code class="docutils literal notranslate"><span class="pre">ecoli.library.initial_conditions</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.library.json_state.html"><code class="docutils literal notranslate"><span class="pre">ecoli.library.json_state</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.library.kinetic_rate_laws.html"><code class="docutils literal notranslate"><span class="pre">ecoli.library.kinetic_rate_laws</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.library.lattice_utils.html"><code class="docutils literal notranslate"><span class="pre">ecoli.library.lattice_utils</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.library.logging_tools.html"><code class="docutils literal notranslate"><span class="pre">ecoli.library.logging_tools</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.library.parameters.html"><code class="docutils literal notranslate"><span class="pre">ecoli.library.parameters</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.library.parquet_emitter.html"><code class="docutils literal notranslate"><span class="pre">ecoli.library.parquet_emitter</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.library.pymunk_multibody.html"><code class="docutils literal notranslate"><span class="pre">ecoli.library.pymunk_multibody</span></code></a></li>
<li class="toctree-l4 current"><a class="current reference internal" href="#"><code class="docutils literal notranslate"><span class="pre">ecoli.library.schema</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.library.serialize.html"><code class="docutils literal notranslate"><span class="pre">ecoli.library.serialize</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.library.sim_data.html"><code class="docutils literal notranslate"><span class="pre">ecoli.library.sim_data</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.library.test_parquet_emitter.html"><code class="docutils literal notranslate"><span class="pre">ecoli.library.test_parquet_emitter</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.library.units.html"><code class="docutils literal notranslate"><span class="pre">ecoli.library.units</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.library.updaters.html"><code class="docutils literal notranslate"><span class="pre">ecoli.library.updaters</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="ecoli.processes.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="ecoli.variants.html"><code class="docutils literal notranslate"><span class="pre">ecoli.variants</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../reconstruction/reconstruction.html"><code class="docutils literal notranslate"><span class="pre">reconstruction</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="../runscripts/runscripts.html"><code class="docutils literal notranslate"><span class="pre">runscripts</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="../validation/validation.html"><code class="docutils literal notranslate"><span class="pre">validation</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="../wholecell/wholecell.html"><code class="docutils literal notranslate"><span class="pre">wholecell</span></code></a></li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">Vivarium E. coli</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../api_ref.html">API Reference</a></li>
          <li class="breadcrumb-item"><a href="ecoli.html"><code class="docutils literal notranslate"><span class="pre">ecoli</span></code></a></li>
          <li class="breadcrumb-item"><a href="ecoli.library.html"><code class="docutils literal notranslate"><span class="pre">ecoli.library</span></code></a></li>
      <li class="breadcrumb-item active"><code class="docutils literal notranslate"><span class="pre">ecoli.library.schema</span></code></li>
      <li class="wy-breadcrumbs-aside">
            <a href="../../../_sources/reference/api/ecoli/ecoli.library.schema.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="module-ecoli.library.schema">
<span id="ecoli-library-schema"></span><h1><code class="docutils literal notranslate"><span class="pre">ecoli.library.schema</span></code><a class="headerlink" href="#module-ecoli.library.schema" title="Link to this heading"></a></h1>
<section id="simulation-helper-functions">
<h2>Simulation Helper Functions<a class="headerlink" href="#simulation-helper-functions" title="Link to this heading"></a></h2>
<p>This is a collection of helper functions used thoughout our code base.</p>
</section>
<dl class="py class">
<dt class="sig sig-object py" id="ecoli.library.schema.MetadataArray">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">ecoli.library.schema.</span></span><span class="sig-name descname"><span class="pre">MetadataArray</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input_array</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metadata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ecoli/library/schema.html#MetadataArray"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ecoli.library.schema.MetadataArray" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></p>
<p>Subclass of Numpy array that allows for metadata to be stored with the array.
Currently used to store next unique molecule index for unique molecule arrays.</p>
</dd></dl>

<dl class="py data">
<dt class="sig sig-object py" id="ecoli.library.schema.UNIQUE_DIVIDERS">
<span class="sig-prename descclassname"><span class="pre">ecoli.library.schema.</span></span><span class="sig-name descname"><span class="pre">UNIQUE_DIVIDERS</span></span><a class="headerlink" href="#ecoli.library.schema.UNIQUE_DIVIDERS" title="Link to this definition"></a></dt>
<dd><p>A mapping of unique molecules to the names of their divider functions ars they are registered
in the <code class="docutils literal notranslate"><span class="pre">divider_registry</span></code> in <code class="docutils literal notranslate"><span class="pre">ecoli/__init__.py</span></code></p>
<dl class="field-list simple">
</dl>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="ecoli.library.schema.UniqueNumpyUpdater">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">ecoli.library.schema.</span></span><span class="sig-name descname"><span class="pre">UniqueNumpyUpdater</span></span><a class="reference internal" href="../../../_modules/ecoli/library/schema.html#UniqueNumpyUpdater"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ecoli.library.schema.UniqueNumpyUpdater" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.13)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>Updates that set attributes of currently active unique molecules
must be applied before any updates that add or delete molecules. If this
is not enforced, in a single timestep, an update might delete a molecule
and allow a subsequent update to add a new molecule in the same row. Then,
an update that intends to modify an attribute of the original molecule
in that row will actually corrupt the data for the new molecule.</p>
<p>To fix this, this unique molecule updater is a bound method with access
to instance attributes that allow it to accumulate updates until given
the signal to apply the accumulated updates in the proper order. The
signal to apply these updates is given by a special process
(<a class="reference internal" href="ecoli.processes.unique_update.html#ecoli.processes.unique_update.UniqueUpdate" title="ecoli.processes.unique_update.UniqueUpdate"><code class="xref py py-class docutils literal notranslate"><span class="pre">UniqueUpdate</span></code></a>) that is
automatically added to the simulation by
<a class="reference internal" href="ecoli.composites.ecoli_master.html#ecoli.composites.ecoli_master.Ecoli.generate_processes_and_steps" title="ecoli.composites.ecoli_master.Ecoli.generate_processes_and_steps"><code class="xref py py-meth docutils literal notranslate"><span class="pre">generate_processes_and_steps()</span></code></a>.</p>
<p>Sets up instance attributes to accumulate updates.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="ecoli.library.schema.UniqueNumpyUpdater.add_updates">
<span class="sig-name descname"><span class="pre">add_updates</span></span><a class="headerlink" href="#ecoli.library.schema.UniqueNumpyUpdater.add_updates" title="Link to this definition"></a></dt>
<dd><p>List of updates that add unique molecules</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="ecoli.library.schema.UniqueNumpyUpdater.set_updates">
<span class="sig-name descname"><span class="pre">set_updates</span></span><a class="headerlink" href="#ecoli.library.schema.UniqueNumpyUpdater.set_updates" title="Link to this definition"></a></dt>
<dd><p>List of updates that modify existing unique molecules</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="ecoli.library.schema.UniqueNumpyUpdater.delete_updates">
<span class="sig-name descname"><span class="pre">delete_updates</span></span><a class="headerlink" href="#ecoli.library.schema.UniqueNumpyUpdater.delete_updates" title="Link to this definition"></a></dt>
<dd><p>List of updates that delete unique molecules</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ecoli.library.schema.UniqueNumpyUpdater.updater">
<span class="sig-name descname"><span class="pre">updater</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">current</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">update</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ecoli/library/schema.html#UniqueNumpyUpdater.updater"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ecoli.library.schema.UniqueNumpyUpdater.updater" title="Link to this definition"></a></dt>
<dd><p>Accumulates updates in instance attributes until given signal to
apply all updates in the following order: <code class="docutils literal notranslate"><span class="pre">set</span></code>, <code class="docutils literal notranslate"><span class="pre">add</span></code>, <code class="docutils literal notranslate"><span class="pre">delete</span></code></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>current</strong> (<a class="reference internal" href="#ecoli.library.schema.MetadataArray" title="ecoli.library.schema.MetadataArray"><em>MetadataArray</em></a>) – Structured Numpy array for a given unique molecule</p></li>
<li><p><strong>update</strong> (<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Dict" title="(in Python v3.13)"><em>Dict</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Any" title="(in Python v3.13)"><em>Any</em></a><em>]</em>) – <p>Dictionary of updates to apply that can contain any
combination of the following keys:</p>
<ul>
<li><dl class="simple">
<dt><code class="docutils literal notranslate"><span class="pre">set</span></code>: Dictionary or list of dictionaries</dt><dd><p>Each key is an attribute of the given unique molecule
and each value is an array. Each array contains
the new attribute values for all active unique
molecules in a givne timestep. Can have multiple such
dictionaries in a list to apply multiple <code class="docutils literal notranslate"><span class="pre">set</span></code> updates.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><code class="docutils literal notranslate"><span class="pre">add</span></code>: Dictionary or list of dictionaries</dt><dd><p>Each key is an attribute of the given unique moleucle
and each value is an array. The nth element of
each array is the value for the corresponding
attribute for the nth unique molecule to be added. If not
provided, unique indices for the <code class="docutils literal notranslate"><span class="pre">unique_index</span></code> attribute
are automatically generated for each new molecule. If
you need to reference the unique indices of new molecules in
the same process and time step in which you generated them,
you MUST use <a class="reference internal" href="#ecoli.library.schema.create_unique_indices" title="ecoli.library.schema.create_unique_indices"><code class="xref py py-func docutils literal notranslate"><span class="pre">create_unique_indices()</span></code></a>
to generate the indices and supply them under the <code class="docutils literal notranslate"><span class="pre">unique_index</span></code>
key of your <code class="docutils literal notranslate"><span class="pre">add</span></code> update. Can have multiple such
dictionaries in a list to apply multiple <code class="docutils literal notranslate"><span class="pre">add</span></code> updates.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><code class="docutils literal notranslate"><span class="pre">delete</span></code>: List or 1D Numpy array of integers, or list of those</dt><dd><p>List of <strong>active</strong> molecule indices to delete. Note that
<code class="docutils literal notranslate"><span class="pre">current</span></code> may have rows that are marked as inactive, so
deleting the 10th active molecule may not equate to
deleting the value in the 10th row of <code class="docutils literal notranslate"><span class="pre">current</span></code>. Can have
multiple such lists in a list to apply multiple <code class="docutils literal notranslate"><span class="pre">delete</span></code> updates.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><code class="docutils literal notranslate"><span class="pre">update</span></code>: Boolean</dt><dd><p>Special key that should only be included in the update of
<a class="reference internal" href="ecoli.processes.unique_update.html#ecoli.processes.unique_update.UniqueUpdate" title="ecoli.processes.unique_update.UniqueUpdate"><code class="xref py py-class docutils literal notranslate"><span class="pre">UniqueUpdate</span></code></a>.
Tells updater to apply all cached updates at the
end of an “execution layer” (see <a class="reference internal" href="../../../stores.html#partitioning"><span class="std std-ref">Partitioning</span></a>).</p>
</dd>
</dl>
</li>
</ul>
</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Updated unique molecule structured Numpy array.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#ecoli.library.schema.MetadataArray" title="ecoli.library.schema.MetadataArray"><em>MetadataArray</em></a></p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ecoli.library.schema.array_from">
<span class="sig-prename descclassname"><span class="pre">ecoli.library.schema.</span></span><span class="sig-name descname"><span class="pre">array_from</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">d</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ecoli/library/schema.html#array_from"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ecoli.library.schema.array_from" title="Link to this definition"></a></dt>
<dd><p>Makes a Numpy array from dictionary values.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>d</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)"><em>dict</em></a>) – Dictionary whose values are to be converted</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Array of all values in d.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ecoli.library.schema.attrs">
<span class="sig-prename descclassname"><span class="pre">ecoli.library.schema.</span></span><span class="sig-name descname"><span class="pre">attrs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">states</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">attributes</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ecoli/library/schema.html#attrs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ecoli.library.schema.attrs" title="Link to this definition"></a></dt>
<dd><p>Helper function to pull out arrays for unique molecule attributes</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>states</strong> (<a class="reference internal" href="#ecoli.library.schema.MetadataArray" title="ecoli.library.schema.MetadataArray"><em>MetadataArray</em></a>) – Structured Numpy array for all unique molecules of a given
type (e.g. RNA, active RNAP, etc.)</p></li>
<li><p><strong>attributes</strong> (<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.List" title="(in Python v3.13)"><em>List</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>]</em>) – List of field names (attributes) whose data should be
retrieved for all active unique molecules in <code class="docutils literal notranslate"><span class="pre">states</span></code></p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>List of arrays, one for each attribute. nth entry in each array
corresponds to the value of that attribute for the nth active
unique molecule in <code class="docutils literal notranslate"><span class="pre">states</span></code></p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.List" title="(in Python v3.13)"><em>List</em></a>[<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ecoli.library.schema.bulk_name_to_idx">
<span class="sig-prename descclassname"><span class="pre">ecoli.library.schema.</span></span><span class="sig-name descname"><span class="pre">bulk_name_to_idx</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">names</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bulk_names</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ecoli/library/schema.html#bulk_name_to_idx"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ecoli.library.schema.bulk_name_to_idx" title="Link to this definition"></a></dt>
<dd><p>Primarily used to retrieve indices for groups of bulk molecules (e.g. NTPs)
in the first run of a process and cache for future runs</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>names</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em> | </em><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.List" title="(in Python v3.13)"><em>List</em></a><em> | </em><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a>) – List or array of things to find. Can also be single string.</p></li>
<li><p><strong>bulk_names</strong> (<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.List" title="(in Python v3.13)"><em>List</em></a><em> | </em><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a>) – List of array of things to search</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Index or indices such that <code class="docutils literal notranslate"><span class="pre">bulk_names[indices]</span> <span class="pre">==</span> <span class="pre">names</span></code></p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)">int</a> | <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ecoli.library.schema.bulk_numpy_updater">
<span class="sig-prename descclassname"><span class="pre">ecoli.library.schema.</span></span><span class="sig-name descname"><span class="pre">bulk_numpy_updater</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">current</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">update</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ecoli/library/schema.html#bulk_numpy_updater"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ecoli.library.schema.bulk_numpy_updater" title="Link to this definition"></a></dt>
<dd><p>Updater function for bulk molecule structured array.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>current</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a>) – Bulk molecule structured array</p></li>
<li><p><strong>update</strong> (<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.List" title="(in Python v3.13)"><em>List</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Tuple" title="(in Python v3.13)"><em>Tuple</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a><em> | </em><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a><em> | </em><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a><em>]</em><em>]</em>) – List of tuples <code class="docutils literal notranslate"><span class="pre">(mol_idx,</span> <span class="pre">add_val)</span></code>, where
<code class="docutils literal notranslate"><span class="pre">mol_idx</span></code> is the index (or array of indices) for
the molecule(s) to be updated and <code class="docutils literal notranslate"><span class="pre">add_val</span></code> is the
count (or array of counts) to be added to the current
count(s) for the specified molecule(s).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Updated bulk molecule structured array</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ecoli.library.schema.counts">
<span class="sig-prename descclassname"><span class="pre">ecoli.library.schema.</span></span><span class="sig-name descname"><span class="pre">counts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">states</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">idx</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ecoli/library/schema.html#counts"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ecoli.library.schema.counts" title="Link to this definition"></a></dt>
<dd><p>Helper function to pull out counts at given indices.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>states</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a>) – Either a Numpy structured array with a <cite>‘count’</cite> field or a 1D
Numpy array of counts.</p></li>
<li><p><strong>idx</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a><em> | </em><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a>) – Indices for the counts of interest.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Counts of molecules at specified indices (copy so can be safely mutated)</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ecoli.library.schema.create_unique_indices">
<span class="sig-prename descclassname"><span class="pre">ecoli.library.schema.</span></span><span class="sig-name descname"><span class="pre">create_unique_indices</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n_indexes</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unique_molecules</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ecoli/library/schema.html#create_unique_indices"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ecoli.library.schema.create_unique_indices" title="Link to this definition"></a></dt>
<dd><p>We strongly recommend letting
<a class="reference internal" href="#ecoli.library.schema.UniqueNumpyUpdater.updater" title="ecoli.library.schema.UniqueNumpyUpdater.updater"><code class="xref py py-meth docutils literal notranslate"><span class="pre">UniqueNumpyUpdater.updater()</span></code></a> generate unique indices
for new unique molecules. If that is not possible, this function
can be used to generate unique indices that should not conflict
with any existing unique indices.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>n_indexes</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a>) – Number of indexes to generate.</p></li>
<li><p><strong>unique_molecules</strong> (<a class="reference internal" href="#ecoli.library.schema.MetadataArray" title="ecoli.library.schema.MetadataArray"><em>MetadataArray</em></a>) – Structured Numpy array of unique molecules.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>List of unique indexes for new unique molecules.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ecoli.library.schema.divide_RNAs_by_domain">
<span class="sig-prename descclassname"><span class="pre">ecoli.library.schema.</span></span><span class="sig-name descname"><span class="pre">divide_RNAs_by_domain</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">values</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">state</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ecoli/library/schema.html#divide_RNAs_by_domain"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ecoli.library.schema.divide_RNAs_by_domain" title="Link to this definition"></a></dt>
<dd><p>Divider function for RNA unique molecules. Ensures that incomplete
transcripts are divided in accordance with how active RNAPs are
divided (which themselves depend on how chromosome domains are divided).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>values</strong> (<a class="reference internal" href="#ecoli.library.schema.MetadataArray" title="ecoli.library.schema.MetadataArray"><em>MetadataArray</em></a>) – Structured Numpy array of RNA unique molecule state</p></li>
<li><p><strong>state</strong> (<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Dict" title="(in Python v3.13)"><em>Dict</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Any" title="(in Python v3.13)"><em>Any</em></a><em>]</em>) – View of relevant unique molecule states according to the
topology under the <code class="docutils literal notranslate"><span class="pre">RNAs</span></code> key in
<a class="reference internal" href="#ecoli.library.schema.UNIQUE_DIVIDERS" title="ecoli.library.schema.UNIQUE_DIVIDERS"><code class="xref py py-data docutils literal notranslate"><span class="pre">ecoli.library.schema.UNIQUE_DIVIDERS</span></code></a>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>List of two structured Numpy arrays, each containing the RNA
unique molecule state of a daughter cell.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)">tuple</a>[<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a>, <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ecoli.library.schema.divide_binomial">
<span class="sig-prename descclassname"><span class="pre">ecoli.library.schema.</span></span><span class="sig-name descname"><span class="pre">divide_binomial</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">state</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ecoli/library/schema.html#divide_binomial"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ecoli.library.schema.divide_binomial" title="Link to this definition"></a></dt>
<dd><p>Binomial Divider</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>state</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a>) – The value to divide.</p></li>
<li><p><strong>config</strong> – Must contain a <code class="docutils literal notranslate"><span class="pre">seed</span></code> key with an integer seed. This
seed will be added to <code class="docutils literal notranslate"><span class="pre">int(state)</span></code> to seed a random number
generator used to calculate the binomial.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The divided values.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)">tuple</a>[<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)">int</a>, <a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)">int</a>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ecoli.library.schema.divide_bulk">
<span class="sig-prename descclassname"><span class="pre">ecoli.library.schema.</span></span><span class="sig-name descname"><span class="pre">divide_bulk</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">state</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ecoli/library/schema.html#divide_bulk"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ecoli.library.schema.divide_bulk" title="Link to this definition"></a></dt>
<dd><p>Divider function for bulk molecules. Automatically added to bulk
molecule ports schemas by <a class="reference internal" href="#ecoli.library.schema.numpy_schema" title="ecoli.library.schema.numpy_schema"><code class="xref py py-func docutils literal notranslate"><span class="pre">ecoli.library.schema.numpy_schema()</span></code></a>
when <code class="docutils literal notranslate"><span class="pre">name</span> <span class="pre">==</span> <span class="pre">'bulk'</span></code>. Uses binomial distribution with <code class="docutils literal notranslate"><span class="pre">p=0.5</span></code> to
randomly partition counts.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>state</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a>) – Structured Numpy array of bulk molecule data.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>List of two structured Numpy arrays, each representing the bulk
molecule state of a daughter cell.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)">tuple</a>[<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a>, <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ecoli.library.schema.divide_by_domain">
<span class="sig-prename descclassname"><span class="pre">ecoli.library.schema.</span></span><span class="sig-name descname"><span class="pre">divide_by_domain</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">values</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">state</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ecoli/library/schema.html#divide_by_domain"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ecoli.library.schema.divide_by_domain" title="Link to this definition"></a></dt>
<dd><p>Divider function for unique molecules that are attached to the
chromsome. Ensures that these molecules are divided in accordance
with the way that chromosome domains are divided.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>values</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a>) – Structured Numpy array of unique molecule state</p></li>
<li><p><strong>state</strong> (<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Dict" title="(in Python v3.13)"><em>Dict</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Any" title="(in Python v3.13)"><em>Any</em></a><em>]</em>) – View of <code class="docutils literal notranslate"><span class="pre">full_chromosome</span></code> and <code class="docutils literal notranslate"><span class="pre">chromosome_domain</span></code>
state as configured under any of the unique molecules with
a divider of <cite>by_domain</cite> in
<a class="reference internal" href="#ecoli.library.schema.UNIQUE_DIVIDERS" title="ecoli.library.schema.UNIQUE_DIVIDERS"><code class="xref py py-data docutils literal notranslate"><span class="pre">ecoli.library.schema.UNIQUE_DIVIDERS</span></code></a>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>List of two structured Numpy arrays, each containing the
unique molecule state of a daughter cell.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)">tuple</a>[<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a>, <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ecoli.library.schema.divide_domains">
<span class="sig-prename descclassname"><span class="pre">ecoli.library.schema.</span></span><span class="sig-name descname"><span class="pre">divide_domains</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">state</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ecoli/library/schema.html#divide_domains"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ecoli.library.schema.divide_domains" title="Link to this definition"></a></dt>
<dd><p>Divider function for chromosome domains. Ensures that all chromosome
domains associated with a full chromosome go to the same daughter cell
that the full chromosome does.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>state</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)"><em>dict</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>, </em><a class="reference internal" href="#ecoli.library.schema.MetadataArray" title="ecoli.library.schema.MetadataArray"><em>MetadataArray</em></a><em>]</em>) – Structured Numpy array of chromosome domain unique molecule
state.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>List of two structured Numpy arrays, each containing the chromosome
domain unique molecule state for a daughter cell.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)">dict</a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)">str</a>, <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ecoli.library.schema.divide_ribosomes_by_RNA">
<span class="sig-prename descclassname"><span class="pre">ecoli.library.schema.</span></span><span class="sig-name descname"><span class="pre">divide_ribosomes_by_RNA</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">values</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">state</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ecoli/library/schema.html#divide_ribosomes_by_RNA"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ecoli.library.schema.divide_ribosomes_by_RNA" title="Link to this definition"></a></dt>
<dd><p>Divider function for active ribosome unique molecules. Automatically
added to ports schema by <a class="reference internal" href="#ecoli.library.schema.numpy_schema" title="ecoli.library.schema.numpy_schema"><code class="xref py py-func docutils literal notranslate"><span class="pre">ecoli.library.schema.numpy_schema()</span></code></a> when
<code class="docutils literal notranslate"><span class="pre">name</span> <span class="pre">==</span> <span class="pre">'active_ribosome'</span></code>. Ensures that ribosomes are divided the same
way that their associated mRNAs are.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>values</strong> (<a class="reference internal" href="#ecoli.library.schema.MetadataArray" title="ecoli.library.schema.MetadataArray"><em>MetadataArray</em></a>) – Structured Numpy array of active ribosome unique molecule state</p></li>
<li><p><strong>state</strong> (<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Dict" title="(in Python v3.13)"><em>Dict</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Any" title="(in Python v3.13)"><em>Any</em></a><em>]</em>) – View into relevant unique molecule states according to the
topology defined under the <code class="docutils literal notranslate"><span class="pre">active_ribosome</span></code> key in
<a class="reference internal" href="#ecoli.library.schema.UNIQUE_DIVIDERS" title="ecoli.library.schema.UNIQUE_DIVIDERS"><code class="xref py py-data docutils literal notranslate"><span class="pre">ecoli.library.schema.UNIQUE_DIVIDERS</span></code></a>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>List of two structured Numpy arrays, each containing the active
ribosome unique molecule state of a daughter cell</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)">tuple</a>[<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a>, <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ecoli.library.schema.divide_set_none">
<span class="sig-prename descclassname"><span class="pre">ecoli.library.schema.</span></span><span class="sig-name descname"><span class="pre">divide_set_none</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">values</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ecoli/library/schema.html#divide_set_none"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ecoli.library.schema.divide_set_none" title="Link to this definition"></a></dt>
<dd><p>Divider function that sets both daughter cell states to <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ecoli.library.schema.empty_dict_divider">
<span class="sig-prename descclassname"><span class="pre">ecoli.library.schema.</span></span><span class="sig-name descname"><span class="pre">empty_dict_divider</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">values</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ecoli/library/schema.html#empty_dict_divider"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ecoli.library.schema.empty_dict_divider" title="Link to this definition"></a></dt>
<dd><p>Divider function that sets both daughter cell states to empty dicts.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ecoli.library.schema.flatten">
<span class="sig-prename descclassname"><span class="pre">ecoli.library.schema.</span></span><span class="sig-name descname"><span class="pre">flatten</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">nested_list</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ecoli/library/schema.html#flatten"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ecoli.library.schema.flatten" title="Link to this definition"></a></dt>
<dd><p>Flattens a nested list into a single list.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>l</strong> – Nested list to flatten.</p></li>
<li><p><strong>nested_list</strong> (<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.List" title="(in Python v3.13)"><em>List</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.List" title="(in Python v3.13)"><em>List</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Any" title="(in Python v3.13)"><em>Any</em></a><em>]</em><em>]</em>)</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.List" title="(in Python v3.13)"><em>List</em></a>[<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Any" title="(in Python v3.13)"><em>Any</em></a>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ecoli.library.schema.follow_domain_tree">
<span class="sig-prename descclassname"><span class="pre">ecoli.library.schema.</span></span><span class="sig-name descname"><span class="pre">follow_domain_tree</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">domain</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">domain_index</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">child_domains</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">place_holder</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ecoli/library/schema.html#follow_domain_tree"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ecoli.library.schema.follow_domain_tree" title="Link to this definition"></a></dt>
<dd><p>Recursive function that returns all the descendents of a single node in
the domain tree, including itself.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>domain</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a>) – Domain index to find all descendents for</p></li>
<li><p><strong>domain_index</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a>) – Array of all domain indices</p></li>
<li><p><strong>child_domains</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a>) – Array of child domains for each index in <code class="docutils literal notranslate"><span class="pre">domain_index</span></code></p></li>
<li><p><strong>place_holder</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a>) – Placeholder domain index (e.g. used in <code class="docutils literal notranslate"><span class="pre">child_domains</span></code>
for domain indices that do not have child domains)</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.List" title="(in Python v3.13)"><em>List</em></a>[<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)">int</a>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="ecoli.library.schema.get_bulk_counts">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">ecoli.library.schema.</span></span><span class="sig-name descname"><span class="pre">get_bulk_counts</span></span><a class="reference internal" href="../../../_modules/ecoli/library/schema.html#get_bulk_counts"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ecoli.library.schema.get_bulk_counts" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference external" href="https://vivarium-core.readthedocs.io/en/latest/reference/api/vivarium.core.registry.html#vivarium.core.registry.Serializer" title="(in Vivarium Core vv1.5.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Serializer</span></code></a></p>
<p>Serializer for bulk molecules that saves counts without IDs or masses.</p>
<dl class="py method">
<dt class="sig sig-object py" id="ecoli.library.schema.get_bulk_counts.serialize">
<span class="sig-name descname"><span class="pre">serialize</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ecoli/library/schema.html#get_bulk_counts.serialize"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ecoli.library.schema.get_bulk_counts.serialize" title="Link to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>bulk</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a>) – Numpy structured array with a <cite>count</cite> field</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Contiguous (required by orjson) array of bulk molecule counts</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a></p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ecoli.library.schema.get_descendent_domains">
<span class="sig-prename descclassname"><span class="pre">ecoli.library.schema.</span></span><span class="sig-name descname"><span class="pre">get_descendent_domains</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">root_domains</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">domain_index</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">child_domains</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">place_holder</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ecoli/library/schema.html#get_descendent_domains"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ecoli.library.schema.get_descendent_domains" title="Link to this definition"></a></dt>
<dd><p>Returns an array of domain indexes that are descendents of the indexes
listed in root_domains, including the indexes in root_domains themselves.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>root_domains</strong> – List of domains to get descendents of</p></li>
<li><p><strong>domain_index</strong> – Array of all domain indices for chromosome domains</p></li>
<li><p><strong>child_domains</strong> – Array of child domains for each index in <code class="docutils literal notranslate"><span class="pre">domain_index</span></code></p></li>
<li><p><strong>place_holder</strong> – Placeholder domain index (e.g. used in <code class="docutils literal notranslate"><span class="pre">child_domains</span></code>
for domain indices that do not have any child domains)</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ecoli.library.schema.get_free_indices">
<span class="sig-prename descclassname"><span class="pre">ecoli.library.schema.</span></span><span class="sig-name descname"><span class="pre">get_free_indices</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">result</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_objects</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ecoli/library/schema.html#get_free_indices"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ecoli.library.schema.get_free_indices" title="Link to this definition"></a></dt>
<dd><p>Find inactive rows for new molecules and expand array if needed</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>result</strong> (<a class="reference internal" href="#ecoli.library.schema.MetadataArray" title="ecoli.library.schema.MetadataArray"><em>MetadataArray</em></a>) – Structured Numpy array for all unique molecules of a given
type (e.g. RNA, active RNAP, etc.)</p></li>
<li><p><strong>n_objects</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a>) – Number of new unique molecules to be added</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A tuple <code class="docutils literal notranslate"><span class="pre">(result,</span> <span class="pre">free_idx)</span></code>. <code class="docutils literal notranslate"><span class="pre">result</span></code> is the same as the
input argument unless <code class="docutils literal notranslate"><span class="pre">n_objects</span></code> is greater than the number
of inactive rows in <code class="docutils literal notranslate"><span class="pre">result</span></code>. In this case, <code class="docutils literal notranslate"><span class="pre">result</span></code> is
grown by at least 10% by concatenating new rows (all zeros).
<code class="docutils literal notranslate"><span class="pre">free_idx</span></code> is an array of size <code class="docutils literal notranslate"><span class="pre">n_objects</span></code> that contains
the indices of rows in <code class="docutils literal notranslate"><span class="pre">result</span></code> that are inactive (<code class="docutils literal notranslate"><span class="pre">_entryState</span></code>
field is 0).</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Tuple" title="(in Python v3.13)"><em>Tuple</em></a>[<a class="reference internal" href="#ecoli.library.schema.MetadataArray" title="ecoli.library.schema.MetadataArray"><em>MetadataArray</em></a>, <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="ecoli.library.schema.get_unique_fields">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">ecoli.library.schema.</span></span><span class="sig-name descname"><span class="pre">get_unique_fields</span></span><a class="reference internal" href="../../../_modules/ecoli/library/schema.html#get_unique_fields"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ecoli.library.schema.get_unique_fields" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference external" href="https://vivarium-core.readthedocs.io/en/latest/reference/api/vivarium.core.registry.html#vivarium.core.registry.Serializer" title="(in Vivarium Core vv1.5.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Serializer</span></code></a></p>
<p>Serializer for unique molecules.</p>
<dl class="py method">
<dt class="sig sig-object py" id="ecoli.library.schema.get_unique_fields.serialize">
<span class="sig-name descname"><span class="pre">serialize</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ecoli/library/schema.html#get_unique_fields.serialize"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ecoli.library.schema.get_unique_fields.serialize" title="Link to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>unique</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a>) – Numpy structured array of attributes for one unique molecule</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>List of contiguous (required by orjson) arrays, one for each attribute</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.13)">list</a>[<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a>]</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ecoli.library.schema.listener_schema">
<span class="sig-prename descclassname"><span class="pre">ecoli.library.schema.</span></span><span class="sig-name descname"><span class="pre">listener_schema</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">elements</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ecoli/library/schema.html#listener_schema"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ecoli.library.schema.listener_schema" title="Link to this definition"></a></dt>
<dd><p>Helper function that can be used in <code class="docutils literal notranslate"><span class="pre">ports_schema</span></code> to create generic
schema for a collection of listeners.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>elements</strong> (<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Dict" title="(in Python v3.13)"><em>Dict</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Any" title="(in Python v3.13)"><em>Any</em></a><em>]</em>) – Dictionary where keys are listener names and values are the
defaults for each listener. Alternatively, if the value is a
tuple, assume that the first element is the default and the second
is metadata that will be emitted at the beginning of a simulation
(see <a class="reference internal" href="ecoli.experiments.ecoli_master_sim.html#ecoli.experiments.ecoli_master_sim.EcoliSim.output_metadata" title="ecoli.experiments.ecoli_master_sim.EcoliSim.output_metadata"><code class="xref py py-meth docutils literal notranslate"><span class="pre">output_metadata()</span></code></a>).
This metadata can then be retrieved later to aid in interpreting
listener values (see <a class="reference internal" href="ecoli.library.parquet_emitter.html#ecoli.library.parquet_emitter.field_metadata" title="ecoli.library.parquet_emitter.field_metadata"><code class="xref py py-func docutils literal notranslate"><span class="pre">field_metadata()</span></code></a>).
As an example, this metadata might be an array of molecule names
for a listener whose emits are arrays of counts, where the nth
molecule name in the metadata corresponds to the nth value in the
counts that are emitted.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Ports schemas for all listeners in <code class="docutils literal notranslate"><span class="pre">elements</span></code>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Dict" title="(in Python v3.13)"><em>Dict</em></a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)">str</a>, <a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Dict" title="(in Python v3.13)"><em>Dict</em></a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)">str</a>, <a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Any" title="(in Python v3.13)"><em>Any</em></a>]]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ecoli.library.schema.not_a_process">
<span class="sig-prename descclassname"><span class="pre">ecoli.library.schema.</span></span><span class="sig-name descname"><span class="pre">not_a_process</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">value</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ecoli/library/schema.html#not_a_process"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ecoli.library.schema.not_a_process" title="Link to this definition"></a></dt>
<dd><p>Returns <code class="docutils literal notranslate"><span class="pre">True</span></code> if not a <a class="reference external" href="https://vivarium-core.readthedocs.io/en/latest/reference/api/vivarium.core.process.html#vivarium.core.process.Process" title="(in Vivarium Core vv1.5.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">vivarium.core.process.Process</span></code></a> instance.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ecoli.library.schema.numpy_schema">
<span class="sig-prename descclassname"><span class="pre">ecoli.library.schema.</span></span><span class="sig-name descname"><span class="pre">numpy_schema</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">emit</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ecoli/library/schema.html#numpy_schema"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ecoli.library.schema.numpy_schema" title="Link to this definition"></a></dt>
<dd><p>Helper function used in ports schemas for bulk and unique molecules</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>name</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a>) – <code class="docutils literal notranslate"><span class="pre">bulk</span></code> for bulk molecules or one of the keys in <a class="reference internal" href="#ecoli.library.schema.UNIQUE_DIVIDERS" title="ecoli.library.schema.UNIQUE_DIVIDERS"><code class="xref py py-data docutils literal notranslate"><span class="pre">UNIQUE_DIVIDERS</span></code></a>
for unique molecules</p></li>
<li><p><strong>emit</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a>) – <code class="docutils literal notranslate"><span class="pre">True</span></code> if should be emitted (default)</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Fully configured ports schema for molecules of type <code class="docutils literal notranslate"><span class="pre">name</span></code></p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Dict" title="(in Python v3.13)"><em>Dict</em></a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)">str</a>, <a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Any" title="(in Python v3.13)"><em>Any</em></a>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ecoli.library.schema.remove_properties">
<span class="sig-prename descclassname"><span class="pre">ecoli.library.schema.</span></span><span class="sig-name descname"><span class="pre">remove_properties</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">schema</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">properties</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ecoli/library/schema.html#remove_properties"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ecoli.library.schema.remove_properties" title="Link to this definition"></a></dt>
<dd><p>Helper function to recursively remove certain properties from a
ports schema.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>schema</strong> (<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Dict" title="(in Python v3.13)"><em>Dict</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Any" title="(in Python v3.13)"><em>Any</em></a><em>]</em>) – Ports schema to remove properties from</p></li>
<li><p><strong>properties</strong> (<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.List" title="(in Python v3.13)"><em>List</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>]</em>) – List of properties to remove</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Ports schema with all properties in <code class="docutils literal notranslate"><span class="pre">properties</span></code> recursively removed.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Dict" title="(in Python v3.13)"><em>Dict</em></a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)">str</a>, <a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Any" title="(in Python v3.13)"><em>Any</em></a>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ecoli.library.schema.zero_listener">
<span class="sig-prename descclassname"><span class="pre">ecoli.library.schema.</span></span><span class="sig-name descname"><span class="pre">zero_listener</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">listener</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ecoli/library/schema.html#zero_listener"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ecoli.library.schema.zero_listener" title="Link to this definition"></a></dt>
<dd><p>Takes a listener dictionary and creates a zeroed version of it.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>listener</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)"><em>dict</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Any" title="(in Python v3.13)"><em>Any</em></a><em>]</em>)</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)">dict</a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)">str</a>, <a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Any" title="(in Python v3.13)"><em>Any</em></a>]</p>
</dd>
</dl>
</dd></dl>

</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="ecoli.library.pymunk_multibody.html" class="btn btn-neutral float-left" title="ecoli.library.pymunk_multibody" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="ecoli.library.serialize.html" class="btn btn-neutral float-right" title="ecoli.library.serialize" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2021-2025, The Vivarium E. coli Authors.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>