

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>ecoli.processes.polypeptide_elongation &mdash; Vivarium E. coli 1.0.0 documentation</title>
      <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=b86133f3" />
      <link rel="stylesheet" type="text/css" href="../../../_static/css/theme.css?v=e59714d7" />

  
      <script src="../../../_static/jquery.js?v=5d32c60e"></script>
      <script src="../../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../../../_static/documentation_options.js?v=8d563738"></script>
      <script src="../../../_static/doctools.js?v=9bcbadda"></script>
      <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
      <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="ecoli.processes.polypeptide_initiation" href="ecoli.processes.polypeptide_initiation.html" />
    <link rel="prev" title="ecoli.processes.partition" href="ecoli.processes.partition.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../index.html" class="icon icon-home">
            Vivarium E. coli
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../../stores.html">Stores</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../processes.html">Processes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../composites.html">Composites</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../experiments.html">Experiments</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../workflows.html">Workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../output.html">Output</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../tutorial.html">Tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../docs.html">Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../hpc.html">HPC Clusters</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../gcloud.html">Google Cloud</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../ci.html">Continuous Integration</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../pycharm.html">PyCharm Setup</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../../api_ref.html">API Reference</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="ecoli.html"><code class="docutils literal notranslate"><span class="pre">ecoli</span></code></a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="ecoli.analysis.html"><code class="docutils literal notranslate"><span class="pre">ecoli.analysis</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="ecoli.composites.html"><code class="docutils literal notranslate"><span class="pre">ecoli.composites</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="ecoli.experiments.html"><code class="docutils literal notranslate"><span class="pre">ecoli.experiments</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="ecoli.library.html"><code class="docutils literal notranslate"><span class="pre">ecoli.library</span></code></a></li>
<li class="toctree-l3 current"><a class="reference internal" href="ecoli.processes.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes</span></code></a><ul class="current">
<li class="toctree-l4"><a class="reference internal" href="ecoli.processes.chemotaxis.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.chemotaxis</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.processes.environment.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.environment</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.processes.listeners.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.listeners</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.processes.membrane.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.membrane</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.processes.spatiality.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.spatiality</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.processes.stubs.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.stubs</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.processes.allocator.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.allocator</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.processes.bulk_timeline.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.bulk_timeline</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.processes.cell_division.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.cell_division</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.processes.chemostat.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.chemostat</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.processes.chromosome_replication.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.chromosome_replication</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.processes.chromosome_structure.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.chromosome_structure</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.processes.complexation.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.complexation</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.processes.concentrations_deriver.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.concentrations_deriver</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.processes.division_detector.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.division_detector</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.processes.engine_process.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.engine_process</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.processes.enzyme_kinetics.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.enzyme_kinetics</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.processes.equilibrium.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.equilibrium</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.processes.global_clock.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.global_clock</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.processes.metabolism.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.metabolism</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.processes.metabolism_redux.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.metabolism_redux</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.processes.metabolism_redux_classic.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.metabolism_redux_classic</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.processes.partition.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.partition</span></code></a></li>
<li class="toctree-l4 current"><a class="current reference internal" href="#"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.polypeptide_elongation</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.processes.polypeptide_initiation.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.polypeptide_initiation</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.processes.protein_degradation.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.protein_degradation</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.processes.registries.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.registries</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.processes.rna_degradation.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.rna_degradation</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.processes.rna_interference.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.rna_interference</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.processes.rna_maturation.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.rna_maturation</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.processes.shape.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.shape</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.processes.tf_binding.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.tf_binding</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.processes.tf_unbinding.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.tf_unbinding</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.processes.transcript_elongation.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.transcript_elongation</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.processes.transcript_initiation.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.transcript_initiation</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.processes.two_component_system.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.two_component_system</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.processes.unique_update.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.unique_update</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="ecoli.variants.html"><code class="docutils literal notranslate"><span class="pre">ecoli.variants</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../reconstruction/reconstruction.html"><code class="docutils literal notranslate"><span class="pre">reconstruction</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="../runscripts/runscripts.html"><code class="docutils literal notranslate"><span class="pre">runscripts</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="../validation/validation.html"><code class="docutils literal notranslate"><span class="pre">validation</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="../wholecell/wholecell.html"><code class="docutils literal notranslate"><span class="pre">wholecell</span></code></a></li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">Vivarium E. coli</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../api_ref.html">API Reference</a></li>
          <li class="breadcrumb-item"><a href="ecoli.html"><code class="docutils literal notranslate"><span class="pre">ecoli</span></code></a></li>
          <li class="breadcrumb-item"><a href="ecoli.processes.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes</span></code></a></li>
      <li class="breadcrumb-item active"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.polypeptide_elongation</span></code></li>
      <li class="wy-breadcrumbs-aside">
            <a href="../../../_sources/reference/api/ecoli/ecoli.processes.polypeptide_elongation.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="module-ecoli.processes.polypeptide_elongation">
<span id="ecoli-processes-polypeptide-elongation"></span><h1><code class="docutils literal notranslate"><span class="pre">ecoli.processes.polypeptide_elongation</span></code><a class="headerlink" href="#module-ecoli.processes.polypeptide_elongation" title="Link to this heading"></a></h1>
<section id="polypeptide-elongation">
<h2>Polypeptide Elongation<a class="headerlink" href="#polypeptide-elongation" title="Link to this heading"></a></h2>
<p>This process models the polymerization of amino acids into polypeptides
by ribosomes using an mRNA transcript as a template. Elongation terminates
once a ribosome has reached the end of an mRNA transcript. Polymerization
occurs across all ribosomes simultaneously and resources are allocated to
maximize the progress of all ribosomes within the limits of the maximum ribosome
elongation rate, available amino acids and GTP, and the length of the transcript.</p>
</section>
<dl class="py class">
<dt class="sig sig-object py" id="ecoli.processes.polypeptide_elongation.BaseElongationModel">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">ecoli.processes.polypeptide_elongation.</span></span><span class="sig-name descname"><span class="pre">BaseElongationModel</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">parameters</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">process</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ecoli/processes/polypeptide_elongation.html#BaseElongationModel"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ecoli.processes.polypeptide_elongation.BaseElongationModel" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.13)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>Base Model: Request amino acids according to upcoming sequence, assuming
max ribosome elongation.</p>
<dl class="py method">
<dt class="sig sig-object py" id="ecoli.processes.polypeptide_elongation.BaseElongationModel.amino_acid_counts">
<span class="sig-name descname"><span class="pre">amino_acid_counts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">aasInSequences</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ecoli/processes/polypeptide_elongation.html#BaseElongationModel.amino_acid_counts"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ecoli.processes.polypeptide_elongation.BaseElongationModel.amino_acid_counts" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ecoli.processes.polypeptide_elongation.BaseElongationModel.elongation_rate">
<span class="sig-name descname"><span class="pre">elongation_rate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">states</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ecoli/processes/polypeptide_elongation.html#BaseElongationModel.elongation_rate"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ecoli.processes.polypeptide_elongation.BaseElongationModel.elongation_rate" title="Link to this definition"></a></dt>
<dd><p>Sets ribosome elongation rate accordint to the media; returns
max value of 22 amino acids/second.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ecoli.processes.polypeptide_elongation.BaseElongationModel.evolve">
<span class="sig-name descname"><span class="pre">evolve</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">states</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">total_aa_counts</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">aas_used</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">next_amino_acid_count</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nElongations</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nInitialized</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ecoli/processes/polypeptide_elongation.html#BaseElongationModel.evolve"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ecoli.processes.polypeptide_elongation.BaseElongationModel.evolve" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ecoli.processes.polypeptide_elongation.BaseElongationModel.final_amino_acids">
<span class="sig-name descname"><span class="pre">final_amino_acids</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">total_aa_counts</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">charged_trna_counts</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ecoli/processes/polypeptide_elongation.html#BaseElongationModel.final_amino_acids"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ecoli.processes.polypeptide_elongation.BaseElongationModel.final_amino_acids" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ecoli.processes.polypeptide_elongation.BaseElongationModel.request">
<span class="sig-name descname"><span class="pre">request</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">states</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">aasInSequences</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ecoli/processes/polypeptide_elongation.html#BaseElongationModel.request"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ecoli.processes.polypeptide_elongation.BaseElongationModel.request" title="Link to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>states</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)"><em>dict</em></a>)</p></li>
<li><p><strong>aasInSequences</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)"><em>tuple</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a><em>, </em><em>...</em><em>]</em><em>, </em><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.dtype.html#numpy.dtype" title="(in NumPy v2.3)"><em>dtype</em></a><em>[</em><em>int64</em><em>]</em><em>]</em>)</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)">tuple</a>[<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)">tuple</a>[<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)">int</a>, …], <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.dtype.html#numpy.dtype" title="(in NumPy v2.3)"><em>dtype</em></a>[<em>float64</em>]], <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)">tuple</a>[<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)">int</a>, …], <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.dtype.html#numpy.dtype" title="(in NumPy v2.3)"><em>dtype</em></a>[<em>float64</em>]], <a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)">dict</a>]</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py data">
<dt class="sig sig-object py" id="ecoli.processes.polypeptide_elongation.MICROMOLAR_UNITS">
<span class="sig-prename descclassname"><span class="pre">ecoli.processes.polypeptide_elongation.</span></span><span class="sig-name descname"><span class="pre">MICROMOLAR_UNITS</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">1.0</span> <span class="pre">[umol/L]</span></em><a class="headerlink" href="#ecoli.processes.polypeptide_elongation.MICROMOLAR_UNITS" title="Link to this definition"></a></dt>
<dd><p>Units used for all concentrations.</p>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="ecoli.processes.polypeptide_elongation.PolypeptideElongation">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">ecoli.processes.polypeptide_elongation.</span></span><span class="sig-name descname"><span class="pre">PolypeptideElongation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">parameters</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ecoli/processes/polypeptide_elongation.html#PolypeptideElongation"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ecoli.processes.polypeptide_elongation.PolypeptideElongation" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="ecoli.processes.partition.html#ecoli.processes.partition.PartitionedProcess" title="ecoli.processes.partition.PartitionedProcess"><code class="xref py py-class docutils literal notranslate"><span class="pre">PartitionedProcess</span></code></a></p>
<p>Polypeptide Elongation PartitionedProcess</p>
<dl class="simple">
<dt>defaults:</dt><dd><p>proteinIds: array length n of protein names</p>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="ecoli.processes.polypeptide_elongation.PolypeptideElongation.calculate_request">
<span class="sig-name descname"><span class="pre">calculate_request</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">timestep</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">states</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ecoli/processes/polypeptide_elongation.html#PolypeptideElongation.calculate_request"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ecoli.processes.polypeptide_elongation.PolypeptideElongation.calculate_request" title="Link to this definition"></a></dt>
<dd><p>Set ribosome elongation rate based on simulation medium environment and elongation rate factor
which is used to create single-cell variability in growth rate
The maximum number of amino acids that can be elongated in a single timestep is set to 22
intentionally as the minimum number of padding values on the protein sequence matrix is set to 22.
If timesteps longer than 1.0s are used, this feature will lead to errors in the effective ribosome
elongation rate.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="ecoli.processes.polypeptide_elongation.PolypeptideElongation.defaults">
<span class="sig-name descname"><span class="pre">defaults</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">Dict</span><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><span class="pre">str</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">Any</span><span class="p"><span class="pre">]</span></span></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'KD_RelA':</span> <span class="pre">0.26,</span> <span class="pre">'KI_SpoT':</span> <span class="pre">20.0,</span> <span class="pre">'KMaa':</span> <span class="pre">100.0,</span> <span class="pre">'KMtf':</span> <span class="pre">1.0,</span> <span class="pre">'aaNames':</span> <span class="pre">['L-ALPHA-ALANINE[c]',</span> <span class="pre">'ARG[c]',</span> <span class="pre">'ASN[c]',</span> <span class="pre">'L-ASPARTATE[c]',</span> <span class="pre">'CYS[c]',</span> <span class="pre">'GLT[c]',</span> <span class="pre">'GLN[c]',</span> <span class="pre">'GLY[c]',</span> <span class="pre">'HIS[c]',</span> <span class="pre">'ILE[c]',</span> <span class="pre">'LEU[c]',</span> <span class="pre">'LYS[c]',</span> <span class="pre">'MET[c]',</span> <span class="pre">'PHE[c]',</span> <span class="pre">'PRO[c]',</span> <span class="pre">'SER[c]',</span> <span class="pre">'THR[c]',</span> <span class="pre">'TRP[c]',</span> <span class="pre">'TYR[c]',</span> <span class="pre">'L-SELENOCYSTEINE[c]',</span> <span class="pre">'VAL[c]'],</span> <span class="pre">'aaWeightsIncorporated':</span> <span class="pre">array([],</span> <span class="pre">dtype=float64),</span> <span class="pre">'aa_enzymes':</span> <span class="pre">[],</span> <span class="pre">'aa_exchange_names':</span> <span class="pre">['L-ALPHA-ALANINE[c]',</span> <span class="pre">'ARG[c]',</span> <span class="pre">'ASN[c]',</span> <span class="pre">'L-ASPARTATE[c]',</span> <span class="pre">'CYS[c]',</span> <span class="pre">'GLT[c]',</span> <span class="pre">'GLN[c]',</span> <span class="pre">'GLY[c]',</span> <span class="pre">'HIS[c]',</span> <span class="pre">'ILE[c]',</span> <span class="pre">'LEU[c]',</span> <span class="pre">'LYS[c]',</span> <span class="pre">'MET[c]',</span> <span class="pre">'PHE[c]',</span> <span class="pre">'PRO[c]',</span> <span class="pre">'SER[c]',</span> <span class="pre">'THR[c]',</span> <span class="pre">'TRP[c]',</span> <span class="pre">'TYR[c]',</span> <span class="pre">'L-SELENOCYSTEINE[c]',</span> <span class="pre">'VAL[c]'],</span> <span class="pre">'aa_exporters':</span> <span class="pre">[],</span> <span class="pre">'aa_from_synthetase':</span> <span class="pre">array([],</span> <span class="pre">shape=(1,</span> <span class="pre">0),</span> <span class="pre">dtype=float64),</span> <span class="pre">'aa_from_trna':</span> <span class="pre">array([0.,</span> <span class="pre">0.,</span> <span class="pre">0.,</span> <span class="pre">0.,</span> <span class="pre">0.,</span> <span class="pre">0.,</span> <span class="pre">0.,</span> <span class="pre">0.,</span> <span class="pre">0.,</span> <span class="pre">0.,</span> <span class="pre">0.,</span> <span class="pre">0.,</span> <span class="pre">0.,</span> <span class="pre">0.,</span> <span class="pre">0.,</span> <span class="pre">0.,</span> <span class="pre">0.,</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">0.,</span> <span class="pre">0.,</span> <span class="pre">0.,</span> <span class="pre">0.]),</span> <span class="pre">'aa_importers':</span> <span class="pre">[],</span> <span class="pre">'aa_supply_in_charging':</span> <span class="pre">False,</span> <span class="pre">'aa_supply_scaling':</span> <span class="pre">&lt;function</span> <span class="pre">PolypeptideElongation.&lt;lambda&gt;&gt;,</span> <span class="pre">'amino_acid_export':</span> <span class="pre">None,</span> <span class="pre">'amino_acid_import':</span> <span class="pre">None,</span> <span class="pre">'amino_acid_synthesis':</span> <span class="pre">None,</span> <span class="pre">'amino_acids':</span> <span class="pre">['L-ALPHA-ALANINE[c]',</span> <span class="pre">'ARG[c]',</span> <span class="pre">'ASN[c]',</span> <span class="pre">'L-ASPARTATE[c]',</span> <span class="pre">'CYS[c]',</span> <span class="pre">'GLT[c]',</span> <span class="pre">'GLN[c]',</span> <span class="pre">'GLY[c]',</span> <span class="pre">'HIS[c]',</span> <span class="pre">'ILE[c]',</span> <span class="pre">'LEU[c]',</span> <span class="pre">'LYS[c]',</span> <span class="pre">'MET[c]',</span> <span class="pre">'PHE[c]',</span> <span class="pre">'PRO[c]',</span> <span class="pre">'SER[c]',</span> <span class="pre">'THR[c]',</span> <span class="pre">'TRP[c]',</span> <span class="pre">'TYR[c]',</span> <span class="pre">'L-SELENOCYSTEINE[c]',</span> <span class="pre">'VAL[c]'],</span> <span class="pre">'basal_elongation_rate':</span> <span class="pre">22.0,</span> <span class="pre">'cellDensity':</span> <span class="pre">1100.0</span> <span class="pre">[g/L],</span> <span class="pre">'charged_trna_names':</span> <span class="pre">[],</span> <span class="pre">'charging_molecule_names':</span> <span class="pre">[],</span> <span class="pre">'charging_stoich_matrix':</span> <span class="pre">array([],</span> <span class="pre">shape=(1,</span> <span class="pre">0),</span> <span class="pre">dtype=float64),</span> <span class="pre">'degradation_index':</span> <span class="pre">1,</span> <span class="pre">'disable_ppgpp_elongation_inhibition':</span> <span class="pre">False,</span> <span class="pre">'elong_rate_by_ppgpp':</span> <span class="pre">0,</span> <span class="pre">'elongation_max':</span> <span class="pre">22.0</span> <span class="pre">[amino_acid/s],</span> <span class="pre">'emit_unique':</span> <span class="pre">False,</span> <span class="pre">'endWeight':</span> <span class="pre">array([2.99146113e-08]),</span> <span class="pre">'get_pathway_enzyme_counts_per_aa':</span> <span class="pre">None,</span> <span class="pre">'gtpPerElongation':</span> <span class="pre">4.2,</span> <span class="pre">'import_constraint_threshold':</span> <span class="pre">0,</span> <span class="pre">'import_threshold':</span> <span class="pre">1e-05,</span> <span class="pre">'kS':</span> <span class="pre">100.0,</span> <span class="pre">'k_RelA':</span> <span class="pre">75.0,</span> <span class="pre">'k_SpoT_deg':</span> <span class="pre">0.23,</span> <span class="pre">'k_SpoT_syn':</span> <span class="pre">2.6,</span> <span class="pre">'krta':</span> <span class="pre">1.0,</span> <span class="pre">'krtf':</span> <span class="pre">500.0,</span> <span class="pre">'make_elongation_rates':</span> <span class="pre">&lt;function</span> <span class="pre">PolypeptideElongation.&lt;lambda&gt;&gt;,</span> <span class="pre">'mechanistic_aa_transport':</span> <span class="pre">False,</span> <span class="pre">'mechanistic_supply':</span> <span class="pre">False,</span> <span class="pre">'mechanistic_translation_supply':</span> <span class="pre">False,</span> <span class="pre">'n_avogadro':</span> <span class="pre">6.02214076e+23</span> <span class="pre">[1/mol],</span> <span class="pre">'next_aa_pad':</span> <span class="pre">1,</span> <span class="pre">'ppgpp':</span> <span class="pre">'ppGpp',</span> <span class="pre">'ppgpp_degradation_reaction':</span> <span class="pre">'PPGPPSYN-RXN',</span> <span class="pre">'ppgpp_reaction_metabolites':</span> <span class="pre">[],</span> <span class="pre">'ppgpp_reaction_names':</span> <span class="pre">[],</span> <span class="pre">'ppgpp_reaction_stoich':</span> <span class="pre">array([],</span> <span class="pre">shape=(1,</span> <span class="pre">0),</span> <span class="pre">dtype=float64),</span> <span class="pre">'ppgpp_regulation':</span> <span class="pre">False,</span> <span class="pre">'ppgpp_synthesis_reaction':</span> <span class="pre">'GDPPYPHOSKIN-RXN',</span> <span class="pre">'proteinIds':</span> <span class="pre">array([],</span> <span class="pre">dtype=float64),</span> <span class="pre">'proteinLengths':</span> <span class="pre">array([],</span> <span class="pre">dtype=float64),</span> <span class="pre">'proteinSequences':</span> <span class="pre">array([],</span> <span class="pre">shape=(1,</span> <span class="pre">0),</span> <span class="pre">dtype=float64),</span> <span class="pre">'proton':</span> <span class="pre">'PROTON',</span> <span class="pre">'rela':</span> <span class="pre">'RELA',</span> <span class="pre">'ribosome30S':</span> <span class="pre">'ribosome30S',</span> <span class="pre">'ribosome50S':</span> <span class="pre">'ribosome50S',</span> <span class="pre">'ribosomeElongationRate':</span> <span class="pre">17.388824902723737,</span> <span class="pre">'ribosomeElongationRateDict':</span> <span class="pre">{'minimal':</span> <span class="pre">17.388824902723737</span> <span class="pre">[amino_acid/s]},</span> <span class="pre">'seed':</span> <span class="pre">0,</span> <span class="pre">'spot':</span> <span class="pre">'SPOT',</span> <span class="pre">'synthesis_index':</span> <span class="pre">0,</span> <span class="pre">'synthetase_names':</span> <span class="pre">[],</span> <span class="pre">'time_step':</span> <span class="pre">1,</span> <span class="pre">'translation_aa_supply':</span> <span class="pre">{'minimal':</span> <span class="pre">array([],</span> <span class="pre">dtype=float64)},</span> <span class="pre">'translation_supply':</span> <span class="pre">False,</span> <span class="pre">'trna_charging':</span> <span class="pre">False,</span> <span class="pre">'uncharged_trna_names':</span> <span class="pre">array([],</span> <span class="pre">dtype=float64),</span> <span class="pre">'unit_conversion':</span> <span class="pre">0,</span> <span class="pre">'variable_elongation':</span> <span class="pre">False,</span> <span class="pre">'water':</span> <span class="pre">'H2O'}</span></em><a class="headerlink" href="#ecoli.processes.polypeptide_elongation.PolypeptideElongation.defaults" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ecoli.processes.polypeptide_elongation.PolypeptideElongation.evolve_state">
<span class="sig-name descname"><span class="pre">evolve_state</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">timestep</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">states</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ecoli/processes/polypeptide_elongation.html#PolypeptideElongation.evolve_state"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ecoli.processes.polypeptide_elongation.PolypeptideElongation.evolve_state" title="Link to this definition"></a></dt>
<dd><p>Set ribosome elongation rate based on simulation medium environment and elongation rate factor
which is used to create single-cell variability in growth rate
The maximum number of amino acids that can be elongated in a single timestep is set to 22
intentionally as the minimum number of padding values on the protein sequence matrix is set to 22.
If timesteps longer than 1.0s are used, this feature will lead to errors in the effective ribosome
elongation rate.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="ecoli.processes.polypeptide_elongation.PolypeptideElongation.name">
<span class="sig-name descname"><span class="pre">name</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">'ecoli-polypeptide-elongation'</span></em><a class="headerlink" href="#ecoli.processes.polypeptide_elongation.PolypeptideElongation.name" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ecoli.processes.polypeptide_elongation.PolypeptideElongation.ports_schema">
<span class="sig-name descname"><span class="pre">ports_schema</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ecoli/processes/polypeptide_elongation.html#PolypeptideElongation.ports_schema"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ecoli.processes.polypeptide_elongation.PolypeptideElongation.ports_schema" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="ecoli.processes.polypeptide_elongation.PolypeptideElongation.topology">
<span class="sig-name descname"><span class="pre">topology</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'active_ribosome':</span> <span class="pre">('unique',</span> <span class="pre">'active_ribosome'),</span> <span class="pre">'boundary':</span> <span class="pre">('boundary',),</span> <span class="pre">'bulk':</span> <span class="pre">('bulk',),</span> <span class="pre">'bulk_total':</span> <span class="pre">('bulk',),</span> <span class="pre">'environment':</span> <span class="pre">('environment',),</span> <span class="pre">'listeners':</span> <span class="pre">('listeners',),</span> <span class="pre">'polypeptide_elongation':</span> <span class="pre">('process_state',</span> <span class="pre">'polypeptide_elongation'),</span> <span class="pre">'timestep':</span> <span class="pre">('timestep',)}</span></em><a class="headerlink" href="#ecoli.processes.polypeptide_elongation.PolypeptideElongation.topology" title="Link to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py data">
<dt class="sig sig-object py" id="ecoli.processes.polypeptide_elongation.REMOVED_FROM_CHARGING">
<span class="sig-prename descclassname"><span class="pre">ecoli.processes.polypeptide_elongation.</span></span><span class="sig-name descname"><span class="pre">REMOVED_FROM_CHARGING</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'L-SELENOCYSTEINE[c]'}</span></em><a class="headerlink" href="#ecoli.processes.polypeptide_elongation.REMOVED_FROM_CHARGING" title="Link to this definition"></a></dt>
<dd><p>Amino acids to remove from charging when running with
<code class="docutils literal notranslate"><span class="pre">steady_state_trna_charging</span></code></p>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="ecoli.processes.polypeptide_elongation.SteadyStateElongationModel">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">ecoli.processes.polypeptide_elongation.</span></span><span class="sig-name descname"><span class="pre">SteadyStateElongationModel</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">parameters</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">process</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ecoli/processes/polypeptide_elongation.html#SteadyStateElongationModel"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ecoli.processes.polypeptide_elongation.SteadyStateElongationModel" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#ecoli.processes.polypeptide_elongation.TranslationSupplyElongationModel" title="ecoli.processes.polypeptide_elongation.TranslationSupplyElongationModel"><code class="xref py py-class docutils literal notranslate"><span class="pre">TranslationSupplyElongationModel</span></code></a></p>
<p>Steady State Charging Model: Requests amino acids based on the
Michaelis-Menten competitive inhibition model.</p>
<dl class="py method">
<dt class="sig sig-object py" id="ecoli.processes.polypeptide_elongation.SteadyStateElongationModel.distribution_from_aa">
<span class="sig-name descname"><span class="pre">distribution_from_aa</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n_aa</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_trna</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">limited</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ecoli/processes/polypeptide_elongation.html#SteadyStateElongationModel.distribution_from_aa"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ecoli.processes.polypeptide_elongation.SteadyStateElongationModel.distribution_from_aa" title="Link to this definition"></a></dt>
<dd><p>Distributes counts of amino acids to tRNAs that are associated with
each amino acid. Uses self.process.aa_from_trna mapping to distribute
from amino acids to tRNA based on the fraction that each tRNA species
makes up for all tRNA species that code for the same amino acid.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>n_aa</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)"><em>tuple</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a><em>, </em><em>...</em><em>]</em><em>, </em><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.dtype.html#numpy.dtype" title="(in NumPy v2.3)"><em>dtype</em></a><em>[</em><em>int64</em><em>]</em><em>]</em>) – counts of each amino acid to distribute to each tRNA</p></li>
<li><p><strong>n_trna</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)"><em>tuple</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a><em>, </em><em>...</em><em>]</em><em>, </em><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.dtype.html#numpy.dtype" title="(in NumPy v2.3)"><em>dtype</em></a><em>[</em><em>int64</em><em>]</em><em>]</em>) – counts of each tRNA to determine the distribution</p></li>
<li><p><strong>limited</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a>) – optional, if True, limits the amino acids
distributed to each tRNA to the number of tRNA that are
available (n_trna)</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Distributed counts for each tRNA</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)">tuple</a>[<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)">int</a>, …], <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.dtype.html#numpy.dtype" title="(in NumPy v2.3)"><em>dtype</em></a>[<em>int64</em>]]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ecoli.processes.polypeptide_elongation.SteadyStateElongationModel.elongation_rate">
<span class="sig-name descname"><span class="pre">elongation_rate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">states</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ecoli/processes/polypeptide_elongation.html#SteadyStateElongationModel.elongation_rate"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ecoli.processes.polypeptide_elongation.SteadyStateElongationModel.elongation_rate" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ecoli.processes.polypeptide_elongation.SteadyStateElongationModel.evolve">
<span class="sig-name descname"><span class="pre">evolve</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">states</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">total_aa_counts</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">aas_used</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">next_amino_acid_count</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nElongations</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nInitialized</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ecoli/processes/polypeptide_elongation.html#SteadyStateElongationModel.evolve"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ecoli.processes.polypeptide_elongation.SteadyStateElongationModel.evolve" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ecoli.processes.polypeptide_elongation.SteadyStateElongationModel.final_amino_acids">
<span class="sig-name descname"><span class="pre">final_amino_acids</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">total_aa_counts</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">charged_trna_counts</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ecoli/processes/polypeptide_elongation.html#SteadyStateElongationModel.final_amino_acids"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ecoli.processes.polypeptide_elongation.SteadyStateElongationModel.final_amino_acids" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ecoli.processes.polypeptide_elongation.SteadyStateElongationModel.request">
<span class="sig-name descname"><span class="pre">request</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">states</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">aasInSequences</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ecoli/processes/polypeptide_elongation.html#SteadyStateElongationModel.request"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ecoli.processes.polypeptide_elongation.SteadyStateElongationModel.request" title="Link to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>states</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)"><em>dict</em></a>)</p></li>
<li><p><strong>aasInSequences</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)"><em>tuple</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a><em>, </em><em>...</em><em>]</em><em>, </em><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.dtype.html#numpy.dtype" title="(in NumPy v2.3)"><em>dtype</em></a><em>[</em><em>int64</em><em>]</em><em>]</em>)</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)">tuple</a>[<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)">tuple</a>[<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)">int</a>, …], <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.dtype.html#numpy.dtype" title="(in NumPy v2.3)"><em>dtype</em></a>[<em>float64</em>]], <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)">tuple</a>[<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)">int</a>, …], <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.dtype.html#numpy.dtype" title="(in NumPy v2.3)"><em>dtype</em></a>[<em>float64</em>]], <a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)">dict</a>]</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="ecoli.processes.polypeptide_elongation.TranslationSupplyElongationModel">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">ecoli.processes.polypeptide_elongation.</span></span><span class="sig-name descname"><span class="pre">TranslationSupplyElongationModel</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">parameters</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">process</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ecoli/processes/polypeptide_elongation.html#TranslationSupplyElongationModel"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ecoli.processes.polypeptide_elongation.TranslationSupplyElongationModel" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#ecoli.processes.polypeptide_elongation.BaseElongationModel" title="ecoli.processes.polypeptide_elongation.BaseElongationModel"><code class="xref py py-class docutils literal notranslate"><span class="pre">BaseElongationModel</span></code></a></p>
<p>Translation Supply Model: Requests minimum of 1) upcoming amino acid
sequence assuming max ribosome elongation (ie. Base Model) and 2)
estimation based on doubling the proteome in one cell cycle (does not
use ribosome elongation, computed in Parca).</p>
<dl class="py method">
<dt class="sig sig-object py" id="ecoli.processes.polypeptide_elongation.TranslationSupplyElongationModel.amino_acid_counts">
<span class="sig-name descname"><span class="pre">amino_acid_counts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">aasInSequences</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ecoli/processes/polypeptide_elongation.html#TranslationSupplyElongationModel.amino_acid_counts"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ecoli.processes.polypeptide_elongation.TranslationSupplyElongationModel.amino_acid_counts" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ecoli.processes.polypeptide_elongation.TranslationSupplyElongationModel.elongation_rate">
<span class="sig-name descname"><span class="pre">elongation_rate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">states</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ecoli/processes/polypeptide_elongation.html#TranslationSupplyElongationModel.elongation_rate"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ecoli.processes.polypeptide_elongation.TranslationSupplyElongationModel.elongation_rate" title="Link to this definition"></a></dt>
<dd><p>Sets ribosome elongation rate accordint to the media; returns
max value of 22 amino acids/second.</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ecoli.processes.polypeptide_elongation.calculate_trna_charging">
<span class="sig-prename descclassname"><span class="pre">ecoli.processes.polypeptide_elongation.</span></span><span class="sig-name descname"><span class="pre">calculate_trna_charging</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">synthetase_conc</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">uncharged_trna_conc</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">charged_trna_conc</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">aa_conc</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ribosome_conc</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">f</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">params</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">supply</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">time_limit</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">limit_v_rib</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_disabled_aas</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ecoli/processes/polypeptide_elongation.html#calculate_trna_charging"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ecoli.processes.polypeptide_elongation.calculate_trna_charging" title="Link to this definition"></a></dt>
<dd><p>Calculates the steady state value of tRNA based on charging and
incorporation through polypeptide elongation. The fraction of
charged/uncharged is also used to determine how quickly the
ribosome is elongating. All concentrations are given in units of
<a class="reference internal" href="#ecoli.processes.polypeptide_elongation.MICROMOLAR_UNITS" title="ecoli.processes.polypeptide_elongation.MICROMOLAR_UNITS"><code class="xref py py-data docutils literal notranslate"><span class="pre">MICROMOLAR_UNITS</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>synthetase_conc</strong> (<em>Unum</em>) – concentration of synthetases associated with
each amino acid</p></li>
<li><p><strong>uncharged_trna_conc</strong> (<em>Unum</em>) – concentration of uncharged tRNA associated
with each amino acid</p></li>
<li><p><strong>charged_trna_conc</strong> (<em>Unum</em>) – concentration of charged tRNA associated with
each amino acid</p></li>
<li><p><strong>aa_conc</strong> (<em>Unum</em>) – concentration of each amino acid</p></li>
<li><p><strong>ribosome_conc</strong> (<em>Unum</em>) – concentration of active ribosomes</p></li>
<li><p><strong>f</strong> (<em>Unum</em>) – fraction of each amino acid to be incorporated to total amino
acids incorporated</p></li>
<li><p><strong>params</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)"><em>dict</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Any" title="(in Python v3.13)"><em>Any</em></a><em>]</em>) – parameters used in charging equations</p></li>
<li><p><strong>supply</strong> (<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Callable" title="(in Python v3.13)"><em>Callable</em></a><em> | </em><em>None</em>) – function to get the rate of amino acid supply (synthesis
and import) based on amino acid concentrations. If None, amino
acid concentrations remain constant during charging</p></li>
<li><p><strong>time_limit</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.13)"><em>float</em></a>) – time limit to reach steady state</p></li>
<li><p><strong>limit_v_rib</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a>) – if True, v_rib is limited to the number of amino acids
that are available</p></li>
<li><p><strong>use_disabled_aas</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a>) – if False, amino acids in
<a class="reference internal" href="#ecoli.processes.polypeptide_elongation.REMOVED_FROM_CHARGING" title="ecoli.processes.polypeptide_elongation.REMOVED_FROM_CHARGING"><code class="xref py py-data docutils literal notranslate"><span class="pre">REMOVED_FROM_CHARGING</span></code></a>
are excluded from charging</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><p>5-element tuple containing</p>
<ul class="simple">
<li><p><strong>new_fraction_charged</strong>: fraction of total tRNA that is charged for each
amino acid species</p></li>
<li><p><strong>v_rib</strong>: ribosomal elongation rate in units of uM/s</p></li>
<li><p><strong>total_synthesis</strong>: the total amount of amino acids synthesized during charging
in units of MICROMOLAR_UNITS. Will be zeros if supply function is not given.</p></li>
<li><p><strong>total_import</strong>: the total amount of amino acids imported during charging
in units of MICROMOLAR_UNITS. Will be zeros if supply function is not given.</p></li>
<li><p><strong>total_export</strong>: the total amount of amino acids exported during charging
in units of MICROMOLAR_UNITS. Will be zeros if supply function is not given.</p></li>
</ul>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)">tuple</a>[<em>Unum</em>, <a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.13)">float</a>, <em>Unum</em>, <em>Unum</em>, <em>Unum</em>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ecoli.processes.polypeptide_elongation.dcdt_jit">
<span class="sig-prename descclassname"><span class="pre">ecoli.processes.polypeptide_elongation.</span></span><span class="sig-name descname"><span class="pre">dcdt_jit</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">t</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">c</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_aas_masked</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_aas</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mask</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">kS</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">synthetase_conc</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">KMaa</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">KMtf</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">f</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">krta</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">krtf</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_elong_rate</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ribosome_conc</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">limit_v_rib</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">aa_rate_limit</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">v_rib_max</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ecoli/processes/polypeptide_elongation.html#dcdt_jit"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ecoli.processes.polypeptide_elongation.dcdt_jit" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ecoli.processes.polypeptide_elongation.get_charging_supply_function">
<span class="sig-prename descclassname"><span class="pre">ecoli.processes.polypeptide_elongation.</span></span><span class="sig-name descname"><span class="pre">get_charging_supply_function</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">supply_in_charging</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mechanistic_supply</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mechanistic_aa_transport</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">amino_acid_synthesis</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">amino_acid_import</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">amino_acid_export</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">aa_supply_scaling</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">counts_to_molar</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">aa_supply</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fwd_enzyme_counts</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rev_enzyme_counts</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dry_mass</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">importer_counts</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exporter_counts</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">aa_in_media</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ecoli/processes/polypeptide_elongation.html#get_charging_supply_function"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ecoli.processes.polypeptide_elongation.get_charging_supply_function" title="Link to this definition"></a></dt>
<dd><p>Get a function mapping internal amino acid concentrations to the amount of
amino acid supply expected.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>supply_in_charging</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a>) – True if using aa_supply_in_charging option</p></li>
<li><p><strong>mechanistic_supply</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a>) – True if using mechanistic_translation_supply option</p></li>
<li><p><strong>mechanistic_aa_transport</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a>) – True if using mechanistic_aa_transport option</p></li>
<li><p><strong>amino_acid_synthesis</strong> (<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Callable" title="(in Python v3.13)"><em>Callable</em></a>) – function to provide rates of synthesis for amino
acids based on the internal state</p></li>
<li><p><strong>amino_acid_import</strong> (<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Callable" title="(in Python v3.13)"><em>Callable</em></a>) – function to provide import rates for amino
acids based on the internal and external state</p></li>
<li><p><strong>amino_acid_export</strong> (<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Callable" title="(in Python v3.13)"><em>Callable</em></a>) – function to provide export rates for amino
acids based on the internal state</p></li>
<li><p><strong>aa_supply_scaling</strong> (<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Callable" title="(in Python v3.13)"><em>Callable</em></a>) – function to scale the amino acid supply based
on the internal state</p></li>
<li><p><strong>counts_to_molar</strong> (<em>Unum</em>) – conversion factor for counts to molar
(<a class="reference internal" href="#ecoli.processes.polypeptide_elongation.MICROMOLAR_UNITS" title="ecoli.processes.polypeptide_elongation.MICROMOLAR_UNITS"><code class="xref py py-data docutils literal notranslate"><span class="pre">MICROMOLAR_UNITS</span></code></a>)</p></li>
<li><p><strong>aa_supply</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)"><em>tuple</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a><em>, </em><em>...</em><em>]</em><em>, </em><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.dtype.html#numpy.dtype" title="(in NumPy v2.3)"><em>dtype</em></a><em>[</em><em>float64</em><em>]</em><em>]</em>) – rate of amino acid supply expected</p></li>
<li><p><strong>fwd_enzyme_counts</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)"><em>tuple</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a><em>, </em><em>...</em><em>]</em><em>, </em><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.dtype.html#numpy.dtype" title="(in NumPy v2.3)"><em>dtype</em></a><em>[</em><em>int64</em><em>]</em><em>]</em>) – enzyme counts in forward reactions for each amino acid</p></li>
<li><p><strong>rev_enzyme_counts</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)"><em>tuple</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a><em>, </em><em>...</em><em>]</em><em>, </em><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.dtype.html#numpy.dtype" title="(in NumPy v2.3)"><em>dtype</em></a><em>[</em><em>int64</em><em>]</em><em>]</em>) – enzyme counts in loss reactions for each amino acid</p></li>
<li><p><strong>dry_mass</strong> (<em>Unum</em>) – dry mass of the cell with mass units</p></li>
<li><p><strong>importer_counts</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)"><em>tuple</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a><em>, </em><em>...</em><em>]</em><em>, </em><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.dtype.html#numpy.dtype" title="(in NumPy v2.3)"><em>dtype</em></a><em>[</em><em>int64</em><em>]</em><em>]</em>) – counts for amino acid importers</p></li>
<li><p><strong>exporter_counts</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)"><em>tuple</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a><em>, </em><em>...</em><em>]</em><em>, </em><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.dtype.html#numpy.dtype" title="(in NumPy v2.3)"><em>dtype</em></a><em>[</em><em>int64</em><em>]</em><em>]</em>) – counts for amino acid exporters</p></li>
<li><p><strong>aa_in_media</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)"><em>tuple</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a><em>, </em><em>...</em><em>]</em><em>, </em><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.dtype.html#numpy.dtype" title="(in NumPy v2.3)"><em>dtype</em></a><em>[</em><a class="reference external" href="https://numpy.org/doc/stable/reference/arrays.scalars.html#numpy.bool" title="(in NumPy v2.3)"><em>bool</em></a><em>]</em><em>]</em>) – True for each amino acid that is present in the media</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Function that provides the amount of supply (synthesis, import, export)
for each amino acid based on the internal state of the cell</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Callable" title="(in Python v3.13)"><em>Callable</em></a>[[<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)">tuple</a>[<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)">int</a>, …], <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.dtype.html#numpy.dtype" title="(in NumPy v2.3)"><em>dtype</em></a>[<em>float64</em>]]], <a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Tuple" title="(in Python v3.13)"><em>Tuple</em></a>[<em>Unum</em>, <em>Unum</em>, <em>Unum</em>]] | None</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ecoli.processes.polypeptide_elongation.main">
<span class="sig-prename descclassname"><span class="pre">ecoli.processes.polypeptide_elongation.</span></span><span class="sig-name descname"><span class="pre">main</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ecoli/processes/polypeptide_elongation.html#main"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ecoli.processes.polypeptide_elongation.main" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ecoli.processes.polypeptide_elongation.ppgpp_metabolite_changes">
<span class="sig-prename descclassname"><span class="pre">ecoli.processes.polypeptide_elongation.</span></span><span class="sig-name descname"><span class="pre">ppgpp_metabolite_changes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">uncharged_trna_conc</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">charged_trna_conc</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ribosome_conc</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">f</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rela_conc</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">spot_conc</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ppgpp_conc</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">counts_to_molar</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">v_rib</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">charging_params</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ppgpp_params</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">time_step</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">request</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">limits</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">random_state</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ecoli/processes/polypeptide_elongation.html#ppgpp_metabolite_changes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ecoli.processes.polypeptide_elongation.ppgpp_metabolite_changes" title="Link to this definition"></a></dt>
<dd><p>Calculates the changes in metabolite counts based on ppGpp synthesis and
degradation reactions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>uncharged_trna_conc</strong> (<em>Unum</em>) – concentration (<a class="reference internal" href="#ecoli.processes.polypeptide_elongation.MICROMOLAR_UNITS" title="ecoli.processes.polypeptide_elongation.MICROMOLAR_UNITS"><code class="xref py py-data docutils literal notranslate"><span class="pre">MICROMOLAR_UNITS</span></code></a>)
of uncharged tRNA associated with each amino acid</p></li>
<li><p><strong>charged_trna_conc</strong> (<em>Unum</em>) – concentration (<a class="reference internal" href="#ecoli.processes.polypeptide_elongation.MICROMOLAR_UNITS" title="ecoli.processes.polypeptide_elongation.MICROMOLAR_UNITS"><code class="xref py py-data docutils literal notranslate"><span class="pre">MICROMOLAR_UNITS</span></code></a>)
of charged tRNA associated with each amino acid</p></li>
<li><p><strong>ribosome_conc</strong> (<em>Unum</em>) – concentration (<a class="reference internal" href="#ecoli.processes.polypeptide_elongation.MICROMOLAR_UNITS" title="ecoli.processes.polypeptide_elongation.MICROMOLAR_UNITS"><code class="xref py py-data docutils literal notranslate"><span class="pre">MICROMOLAR_UNITS</span></code></a>)
of active ribosomes</p></li>
<li><p><strong>f</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)"><em>tuple</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a><em>, </em><em>...</em><em>]</em><em>, </em><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.dtype.html#numpy.dtype" title="(in NumPy v2.3)"><em>dtype</em></a><em>[</em><em>float64</em><em>]</em><em>]</em>) – fraction of each amino acid to be incorporated
to total amino acids incorporated</p></li>
<li><p><strong>rela_conc</strong> (<em>Unum</em>) – concentration (<a class="reference internal" href="#ecoli.processes.polypeptide_elongation.MICROMOLAR_UNITS" title="ecoli.processes.polypeptide_elongation.MICROMOLAR_UNITS"><code class="xref py py-data docutils literal notranslate"><span class="pre">MICROMOLAR_UNITS</span></code></a>) of RelA</p></li>
<li><p><strong>spot_conc</strong> (<em>Unum</em>) – concentration (<a class="reference internal" href="#ecoli.processes.polypeptide_elongation.MICROMOLAR_UNITS" title="ecoli.processes.polypeptide_elongation.MICROMOLAR_UNITS"><code class="xref py py-data docutils literal notranslate"><span class="pre">MICROMOLAR_UNITS</span></code></a>) of SpoT</p></li>
<li><p><strong>ppgpp_conc</strong> (<em>Unum</em>) – concentration (<a class="reference internal" href="#ecoli.processes.polypeptide_elongation.MICROMOLAR_UNITS" title="ecoli.processes.polypeptide_elongation.MICROMOLAR_UNITS"><code class="xref py py-data docutils literal notranslate"><span class="pre">MICROMOLAR_UNITS</span></code></a>) of ppGpp</p></li>
<li><p><strong>counts_to_molar</strong> (<em>Unum</em>) – conversion factor
from counts to molarity (<a class="reference internal" href="#ecoli.processes.polypeptide_elongation.MICROMOLAR_UNITS" title="ecoli.processes.polypeptide_elongation.MICROMOLAR_UNITS"><code class="xref py py-data docutils literal notranslate"><span class="pre">MICROMOLAR_UNITS</span></code></a>)</p></li>
<li><p><strong>v_rib</strong> (<em>Unum</em>) – rate of amino acid incorporation at the ribosome (units of uM/s)</p></li>
<li><p><strong>charging_params</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)"><em>dict</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Any" title="(in Python v3.13)"><em>Any</em></a><em>]</em>) – parameters used in charging equations</p></li>
<li><p><strong>ppgpp_params</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)"><em>dict</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Any" title="(in Python v3.13)"><em>Any</em></a><em>]</em>) – parameters used in ppGpp reactions</p></li>
<li><p><strong>time_step</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.13)"><em>float</em></a>) – length of the current time step</p></li>
<li><p><strong>request</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a>) – if True, only considers reactant stoichiometry,
otherwise considers reactants and products. For use in
calculateRequest. GDP appears as both a reactant and product
and the request can be off the actual use if not handled in this
manner.</p></li>
<li><p><strong>limits</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)"><em>tuple</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a><em>, </em><em>...</em><em>]</em><em>, </em><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.dtype.html#numpy.dtype" title="(in NumPy v2.3)"><em>dtype</em></a><em>[</em><em>float64</em><em>]</em><em>] </em><em>| </em><em>None</em>) – counts of molecules that are available to prevent
negative total counts as a result of delta_metabolites.
If None, no limits are placed on molecule changes.</p></li>
<li><p><strong>random_state</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/random/legacy.html#numpy.random.RandomState" title="(in NumPy v2.3)"><em>RandomState</em></a><em> | </em><em>None</em>) – random state for the process</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><p>7-element tuple containing</p>
<ul class="simple">
<li><p><strong>delta_metabolites</strong>: the change in counts of each metabolite
involved in ppGpp reactions</p></li>
<li><p><strong>n_syn_reactions</strong>: the number of ppGpp synthesis reactions</p></li>
<li><p><strong>n_deg_reactions</strong>: the number of ppGpp degradation reactions</p></li>
<li><p><strong>v_rela_syn</strong>: rate of synthesis from RelA per amino
acid tRNA species</p></li>
<li><p><strong>v_spot_syn</strong>: rate of synthesis from SpoT</p></li>
<li><p><strong>v_deg</strong>: rate of degradation from SpoT</p></li>
<li><p><strong>v_deg_inhibited</strong>: rate of degradation from SpoT per
amino acid tRNA species</p></li>
</ul>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)">tuple</a>[<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)">tuple</a>[<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)">int</a>, …], <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.dtype.html#numpy.dtype" title="(in NumPy v2.3)"><em>dtype</em></a>[<em>int64</em>]], <a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)">int</a>, <a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)">int</a>, <em>Unum</em>, <em>Unum</em>, <em>Unum</em>, <em>Unum</em>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ecoli.processes.polypeptide_elongation.run_plot">
<span class="sig-prename descclassname"><span class="pre">ecoli.processes.polypeptide_elongation.</span></span><span class="sig-name descname"><span class="pre">run_plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">config</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ecoli/processes/polypeptide_elongation.html#run_plot"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ecoli.processes.polypeptide_elongation.run_plot" title="Link to this definition"></a></dt>
<dd></dd></dl>

</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="ecoli.processes.partition.html" class="btn btn-neutral float-left" title="ecoli.processes.partition" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="ecoli.processes.polypeptide_initiation.html" class="btn btn-neutral float-right" title="ecoli.processes.polypeptide_initiation" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2021-2025, The Vivarium E. coli Authors.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>