

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>ecoli.processes.metabolism &mdash; Vivarium E. coli 1.0.0 documentation</title>
      <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=b86133f3" />
      <link rel="stylesheet" type="text/css" href="../../../_static/css/theme.css?v=e59714d7" />

  
      <script src="../../../_static/jquery.js?v=5d32c60e"></script>
      <script src="../../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../../../_static/documentation_options.js?v=8d563738"></script>
      <script src="../../../_static/doctools.js?v=9bcbadda"></script>
      <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
      <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="ecoli.processes.metabolism_redux" href="ecoli.processes.metabolism_redux.html" />
    <link rel="prev" title="ecoli.processes.global_clock" href="ecoli.processes.global_clock.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../index.html" class="icon icon-home">
            Vivarium E. coli
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../../stores.html">Stores</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../processes.html">Processes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../composites.html">Composites</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../experiments.html">Experiments</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../workflows.html">Workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../output.html">Output</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../tutorial.html">Tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../docs.html">Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../hpc.html">HPC Clusters</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../gcloud.html">Google Cloud</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../ci.html">Continuous Integration</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../pycharm.html">PyCharm Setup</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../../api_ref.html">API Reference</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="ecoli.html"><code class="docutils literal notranslate"><span class="pre">ecoli</span></code></a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="ecoli.analysis.html"><code class="docutils literal notranslate"><span class="pre">ecoli.analysis</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="ecoli.composites.html"><code class="docutils literal notranslate"><span class="pre">ecoli.composites</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="ecoli.experiments.html"><code class="docutils literal notranslate"><span class="pre">ecoli.experiments</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="ecoli.library.html"><code class="docutils literal notranslate"><span class="pre">ecoli.library</span></code></a></li>
<li class="toctree-l3 current"><a class="reference internal" href="ecoli.processes.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes</span></code></a><ul class="current">
<li class="toctree-l4"><a class="reference internal" href="ecoli.processes.chemotaxis.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.chemotaxis</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.processes.environment.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.environment</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.processes.listeners.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.listeners</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.processes.membrane.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.membrane</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.processes.spatiality.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.spatiality</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.processes.stubs.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.stubs</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.processes.allocator.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.allocator</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.processes.bulk_timeline.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.bulk_timeline</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.processes.cell_division.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.cell_division</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.processes.chemostat.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.chemostat</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.processes.chromosome_replication.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.chromosome_replication</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.processes.chromosome_structure.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.chromosome_structure</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.processes.complexation.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.complexation</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.processes.concentrations_deriver.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.concentrations_deriver</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.processes.division_detector.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.division_detector</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.processes.engine_process.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.engine_process</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.processes.enzyme_kinetics.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.enzyme_kinetics</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.processes.equilibrium.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.equilibrium</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.processes.global_clock.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.global_clock</span></code></a></li>
<li class="toctree-l4 current"><a class="current reference internal" href="#"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.metabolism</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.processes.metabolism_redux.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.metabolism_redux</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.processes.metabolism_redux_classic.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.metabolism_redux_classic</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.processes.partition.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.partition</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.processes.polypeptide_elongation.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.polypeptide_elongation</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.processes.polypeptide_initiation.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.polypeptide_initiation</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.processes.protein_degradation.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.protein_degradation</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.processes.registries.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.registries</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.processes.rna_degradation.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.rna_degradation</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.processes.rna_interference.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.rna_interference</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.processes.rna_maturation.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.rna_maturation</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.processes.shape.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.shape</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.processes.tf_binding.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.tf_binding</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.processes.tf_unbinding.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.tf_unbinding</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.processes.transcript_elongation.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.transcript_elongation</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.processes.transcript_initiation.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.transcript_initiation</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.processes.two_component_system.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.two_component_system</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="ecoli.processes.unique_update.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.unique_update</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="ecoli.variants.html"><code class="docutils literal notranslate"><span class="pre">ecoli.variants</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../reconstruction/reconstruction.html"><code class="docutils literal notranslate"><span class="pre">reconstruction</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="../runscripts/runscripts.html"><code class="docutils literal notranslate"><span class="pre">runscripts</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="../validation/validation.html"><code class="docutils literal notranslate"><span class="pre">validation</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="../wholecell/wholecell.html"><code class="docutils literal notranslate"><span class="pre">wholecell</span></code></a></li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">Vivarium E. coli</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../api_ref.html">API Reference</a></li>
          <li class="breadcrumb-item"><a href="ecoli.html"><code class="docutils literal notranslate"><span class="pre">ecoli</span></code></a></li>
          <li class="breadcrumb-item"><a href="ecoli.processes.html"><code class="docutils literal notranslate"><span class="pre">ecoli.processes</span></code></a></li>
      <li class="breadcrumb-item active"><code class="docutils literal notranslate"><span class="pre">ecoli.processes.metabolism</span></code></li>
      <li class="wy-breadcrumbs-aside">
            <a href="../../../_sources/reference/api/ecoli/ecoli.processes.metabolism.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="module-ecoli.processes.metabolism">
<span id="ecoli-processes-metabolism"></span><h1><code class="docutils literal notranslate"><span class="pre">ecoli.processes.metabolism</span></code><a class="headerlink" href="#module-ecoli.processes.metabolism" title="Link to this heading"></a></h1>
<section id="metabolism">
<h2>Metabolism<a class="headerlink" href="#metabolism" title="Link to this heading"></a></h2>
<p>Encodes molecular simulation of microbial metabolism using flux-balance analysis.</p>
<p>This process demonstrates how metabolites are taken up from the environment
and converted into other metabolites for use in other processes.</p>
<p>NOTE:
- In wcEcoli, metabolism only runs after all other processes have completed
and internal states have been updated (deriver-like, no partitioning necessary)</p>
</section>
<dl class="py class">
<dt class="sig sig-object py" id="ecoli.processes.metabolism.FluxBalanceAnalysisModel">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">ecoli.processes.metabolism.</span></span><span class="sig-name descname"><span class="pre">FluxBalanceAnalysisModel</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">parameters</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">timeline</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">include_ppgpp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ecoli/processes/metabolism.html#FluxBalanceAnalysisModel"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ecoli.processes.metabolism.FluxBalanceAnalysisModel" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.13)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>Metabolism model that solves an FBA problem with modular_fba.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>parameters</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)"><em>dict</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Any" title="(in Python v3.13)"><em>Any</em></a><em>]</em>) – parameters from simulation data</p></li>
<li><p><strong>timeline</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)"><em>tuple</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)"><em>tuple</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>]</em><em>, </em><em>...</em><em>]</em>) – timeline for nutrient changes during simulation
(time of change, media ID), by default [(0.0, ‘minimal’)]</p></li>
<li><p><strong>include_ppgpp</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a>) – if True, ppGpp is included as a concentration target</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="ecoli.processes.metabolism.FluxBalanceAnalysisModel.set_molecule_levels">
<span class="sig-name descname"><span class="pre">set_molecule_levels</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">metabolite_counts</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">counts_to_molar</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">coefficient</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">current_media_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unconstrained</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">constrained</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">conc_updates</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">aa_uptake_package</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ecoli/processes/metabolism.html#FluxBalanceAnalysisModel.set_molecule_levels"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ecoli.processes.metabolism.FluxBalanceAnalysisModel.set_molecule_levels" title="Link to this definition"></a></dt>
<dd><p>Set internal and external molecule levels available to the FBA solver.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>metabolite_counts</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)"><em>tuple</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a><em>, </em><em>...</em><em>]</em><em>, </em><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.dtype.html#numpy.dtype" title="(in NumPy v2.3)"><em>dtype</em></a><em>[</em><em>int64</em><em>]</em><em>]</em>) – counts for each metabolite with a concentration target</p></li>
<li><p><strong>counts_to_molar</strong> (<em>Unum</em>) – conversion from counts to molar (counts/volume)</p></li>
<li><p><strong>coefficient</strong> (<em>Unum</em>) – coefficient to convert from mmol/g DCW/hr to mM basis
(mass*time/volume)</p></li>
<li><p><strong>current_media_id</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a>) – ID of current media</p></li>
<li><p><strong>unconstrained</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#set" title="(in Python v3.13)"><em>set</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>]</em>) – molecules that have unconstrained import</p></li>
<li><p><strong>constrained</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#set" title="(in Python v3.13)"><em>set</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>]</em>) – molecules (keys) and their limited max uptake rates
(mol / mass / time)</p></li>
<li><p><strong>conc_updates</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)"><em>dict</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>, </em><em>Unum</em><em>]</em>) – updates to concentrations targets for molecules (mmol/L)</p></li>
<li><p><strong>aa_uptake_package</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)"><em>tuple</em></a><em>[</em><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)"><em>tuple</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a><em>, </em><em>...</em><em>]</em><em>, </em><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.dtype.html#numpy.dtype" title="(in NumPy v2.3)"><em>dtype</em></a><em>[</em><em>float64</em><em>]</em><em>]</em><em>, </em><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)"><em>tuple</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a><em>, </em><em>...</em><em>]</em><em>, </em><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.dtype.html#numpy.dtype" title="(in NumPy v2.3)"><em>dtype</em></a><em>[</em><a class="reference external" href="https://numpy.org/doc/stable/reference/arrays.scalars.html#numpy.str_" title="(in NumPy v2.3)"><em>str_</em></a><em>]</em><em>]</em><em>, </em><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a><em>] </em><em>| </em><em>None</em>) – (uptake rates, amino acid names, force levels),
determines whether to set hard uptake rates</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ecoli.processes.metabolism.FluxBalanceAnalysisModel.set_reaction_bounds">
<span class="sig-name descname"><span class="pre">set_reaction_bounds</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">catalyst_counts</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">counts_to_molar</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">coefficient</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gtp_to_hydrolyze</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ecoli/processes/metabolism.html#FluxBalanceAnalysisModel.set_reaction_bounds"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ecoli.processes.metabolism.FluxBalanceAnalysisModel.set_reaction_bounds" title="Link to this definition"></a></dt>
<dd><p>Set reaction bounds for constrained reactions in the FBA object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>catalyst_counts</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)"><em>tuple</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a><em>, </em><em>...</em><em>]</em><em>, </em><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.dtype.html#numpy.dtype" title="(in NumPy v2.3)"><em>dtype</em></a><em>[</em><em>int64</em><em>]</em><em>]</em>) – counts of enzyme catalysts</p></li>
<li><p><strong>counts_to_molar</strong> (<em>Unum</em>) – conversion from counts to molar (counts/volume)</p></li>
<li><p><strong>coefficient</strong> (<em>Unum</em>) – coefficient to convert from mmol/g DCW/hr to mM basis
(mass*time/volume)</p></li>
<li><p><strong>gtp_to_hydrolyze</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.13)"><em>float</em></a>) – number of GTP molecules to hydrolyze to
account for consumption in translation</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ecoli.processes.metabolism.FluxBalanceAnalysisModel.set_reaction_targets">
<span class="sig-name descname"><span class="pre">set_reaction_targets</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">kinetic_enzyme_counts</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">kinetic_substrate_counts</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">counts_to_molar</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">time_step</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ecoli/processes/metabolism.html#FluxBalanceAnalysisModel.set_reaction_targets"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ecoli.processes.metabolism.FluxBalanceAnalysisModel.set_reaction_targets" title="Link to this definition"></a></dt>
<dd><p>Set reaction targets for constrained reactions in the FBA object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>kinetic_enzyme_counts</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)"><em>tuple</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a><em>, </em><em>...</em><em>]</em><em>, </em><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.dtype.html#numpy.dtype" title="(in NumPy v2.3)"><em>dtype</em></a><em>[</em><em>int64</em><em>]</em><em>]</em>) – counts of enzymes used in kinetic constraints</p></li>
<li><p><strong>kinetic_substrate_counts</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)"><em>tuple</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a><em>, </em><em>...</em><em>]</em><em>, </em><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.dtype.html#numpy.dtype" title="(in NumPy v2.3)"><em>dtype</em></a><em>[</em><em>int64</em><em>]</em><em>]</em>) – counts of substrates used in kinetic
constraints</p></li>
<li><p><strong>counts_to_molar</strong> (<em>Unum</em>) – conversion from counts to molar (counts/volume)</p></li>
<li><p><strong>time_step</strong> (<em>Unum</em>) – current time step (time)</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><p>3-element tuple containing</p>
<ul class="simple">
<li><p><strong>mean_targets</strong>: mean target for each constrained reaction</p></li>
<li><p><strong>upper_targets</strong>: upper target limit for each constrained reaction</p></li>
<li><p><strong>lower_targets</strong>: lower target limit for each constrained reaction</p></li>
</ul>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)">tuple</a>[<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)">tuple</a>[<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)">int</a>, …], <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.dtype.html#numpy.dtype" title="(in NumPy v2.3)"><em>dtype</em></a>[<em>float64</em>]], <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)">tuple</a>[<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)">int</a>, …], <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.dtype.html#numpy.dtype" title="(in NumPy v2.3)"><em>dtype</em></a>[<em>float64</em>]], <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)">tuple</a>[<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)">int</a>, …], <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.dtype.html#numpy.dtype" title="(in NumPy v2.3)"><em>dtype</em></a>[<em>float64</em>]]]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ecoli.processes.metabolism.FluxBalanceAnalysisModel.update_external_molecule_levels">
<span class="sig-name descname"><span class="pre">update_external_molecule_levels</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">objective</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metabolite_concentrations</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">external_molecule_levels</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ecoli/processes/metabolism.html#FluxBalanceAnalysisModel.update_external_molecule_levels"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ecoli.processes.metabolism.FluxBalanceAnalysisModel.update_external_molecule_levels" title="Link to this definition"></a></dt>
<dd><p>Limit amino acid uptake to what is needed to meet concentration
objective to prevent use as carbon source, otherwise could be used
as an infinite nutrient source.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>objective</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)"><em>dict</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>, </em><em>Unum</em><em>]</em>) – homeostatic objective for internal
molecules (molecule ID: concentration in counts/volume units)</p></li>
<li><p><strong>metabolite_concentrations</strong> (<em>Unum</em>) – concentration for each
molecule in metabolite_names</p></li>
<li><p><strong>external_molecule_levels</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)"><em>tuple</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a><em>, </em><em>...</em><em>]</em><em>, </em><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.dtype.html#numpy.dtype" title="(in NumPy v2.3)"><em>dtype</em></a><em>[</em><em>float64</em><em>]</em><em>]</em>) – current limits on
external molecule availability</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Updated limits on external molecule availability</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)">tuple</a>[<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)">int</a>, …], <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.dtype.html#numpy.dtype" title="(in NumPy v2.3)"><em>dtype</em></a>[<em>float64</em>]]</p>
</dd>
</dl>
<dl class="simple">
<dt>TODO(wcEcoli):</dt><dd><p>determine rate of uptake so that some amino acid uptake can
be used as a carbon/nitrogen source</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="ecoli.processes.metabolism.Metabolism">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">ecoli.processes.metabolism.</span></span><span class="sig-name descname"><span class="pre">Metabolism</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">parameters</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ecoli/processes/metabolism.html#Metabolism"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ecoli.processes.metabolism.Metabolism" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference external" href="https://vivarium-core.readthedocs.io/en/latest/reference/api/vivarium.core.process.html#vivarium.core.process.Step" title="(in Vivarium Core vv1.5.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Step</span></code></a></p>
<p>Metabolism Process</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="ecoli.processes.metabolism.Metabolism.defaults">
<span class="sig-name descname"><span class="pre">defaults</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">Dict</span><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><span class="pre">str</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">Any</span><span class="p"><span class="pre">]</span></span></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'aa_exchange_names':</span> <span class="pre">[],</span> <span class="pre">'aa_names':</span> <span class="pre">[],</span> <span class="pre">'aa_targets_not_updated':</span> <span class="pre">{},</span> <span class="pre">'amino_acid_ids':</span> <span class="pre">{},</span> <span class="pre">'avogadro':</span> <span class="pre">6.02214076e+23</span> <span class="pre">[1/mol],</span> <span class="pre">'base_reaction_ids':</span> <span class="pre">[],</span> <span class="pre">'cell_density':</span> <span class="pre">1100.0</span> <span class="pre">[g/L],</span> <span class="pre">'cell_dry_mass_fraction':</span> <span class="pre">0.3,</span> <span class="pre">'current_timeline':</span> <span class="pre">None,</span> <span class="pre">'dark_atp':</span> <span class="pre">33.565052868380675</span> <span class="pre">[mmol/g],</span> <span class="pre">'doubling_time':</span> <span class="pre">44.0</span> <span class="pre">[min],</span> <span class="pre">'exchange_data_from_media':</span> <span class="pre">&lt;function</span> <span class="pre">Metabolism.&lt;lambda&gt;&gt;,</span> <span class="pre">'exchange_molecules':</span> <span class="pre">[],</span> <span class="pre">'fba_reaction_ids_to_base_reaction_ids':</span> <span class="pre">[],</span> <span class="pre">'get_biomass_as_concentrations':</span> <span class="pre">&lt;function</span> <span class="pre">Metabolism.&lt;lambda&gt;&gt;,</span> <span class="pre">'get_import_constraints':</span> <span class="pre">&lt;function</span> <span class="pre">Metabolism.&lt;lambda&gt;&gt;,</span> <span class="pre">'get_masses':</span> <span class="pre">&lt;function</span> <span class="pre">Metabolism.&lt;lambda&gt;&gt;,</span> <span class="pre">'get_ppGpp_conc':</span> <span class="pre">&lt;function</span> <span class="pre">Metabolism.&lt;lambda&gt;&gt;,</span> <span class="pre">'import_constraint_threshold':</span> <span class="pre">0,</span> <span class="pre">'imports':</span> <span class="pre">{},</span> <span class="pre">'include_ppgpp':</span> <span class="pre">False,</span> <span class="pre">'linked_metabolites':</span> <span class="pre">None,</span> <span class="pre">'mechanistic_aa_transport':</span> <span class="pre">False,</span> <span class="pre">'media_id':</span> <span class="pre">'minimal',</span> <span class="pre">'metabolism':</span> <span class="pre">{},</span> <span class="pre">'ngam':</span> <span class="pre">8.39</span> <span class="pre">[mmol/g.h],</span> <span class="pre">'nutrientToDoublingTime':</span> <span class="pre">{},</span> <span class="pre">'ppgpp_id':</span> <span class="pre">'ppgpp',</span> <span class="pre">'reduce_murein_objective':</span> <span class="pre">False,</span> <span class="pre">'removed_aa_uptake':</span> <span class="pre">[],</span> <span class="pre">'seed':</span> <span class="pre">0,</span> <span class="pre">'time_step':</span> <span class="pre">1,</span> <span class="pre">'use_trna_charging':</span> <span class="pre">False}</span></em><a class="headerlink" href="#ecoli.processes.metabolism.Metabolism.defaults" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="ecoli.processes.metabolism.Metabolism.name">
<span class="sig-name descname"><span class="pre">name</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">'ecoli-metabolism'</span></em><a class="headerlink" href="#ecoli.processes.metabolism.Metabolism.name" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ecoli.processes.metabolism.Metabolism.next_update">
<span class="sig-name descname"><span class="pre">next_update</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">timestep</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">states</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ecoli/processes/metabolism.html#Metabolism.next_update"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ecoli.processes.metabolism.Metabolism.next_update" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ecoli.processes.metabolism.Metabolism.ports_schema">
<span class="sig-name descname"><span class="pre">ports_schema</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ecoli/processes/metabolism.html#Metabolism.ports_schema"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ecoli.processes.metabolism.Metabolism.ports_schema" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="ecoli.processes.metabolism.Metabolism.topology">
<span class="sig-name descname"><span class="pre">topology</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'boundary':</span> <span class="pre">('boundary',),</span> <span class="pre">'bulk':</span> <span class="pre">('bulk',),</span> <span class="pre">'bulk_total':</span> <span class="pre">('bulk',),</span> <span class="pre">'environment':</span> <span class="pre">{'_path':</span> <span class="pre">('environment',),</span> <span class="pre">'exchange':</span> <span class="pre">('exchange',)},</span> <span class="pre">'global_time':</span> <span class="pre">('global_time',),</span> <span class="pre">'listeners':</span> <span class="pre">('listeners',),</span> <span class="pre">'next_update_time':</span> <span class="pre">('next_update_time',</span> <span class="pre">'metabolism'),</span> <span class="pre">'polypeptide_elongation':</span> <span class="pre">('process_state',</span> <span class="pre">'polypeptide_elongation'),</span> <span class="pre">'timestep':</span> <span class="pre">('timestep',)}</span></em><a class="headerlink" href="#ecoli.processes.metabolism.Metabolism.topology" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ecoli.processes.metabolism.Metabolism.update_amino_acid_targets">
<span class="sig-name descname"><span class="pre">update_amino_acid_targets</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">counts_to_molar</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">count_diff</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">amino_acid_counts</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ecoli/processes/metabolism.html#Metabolism.update_amino_acid_targets"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ecoli.processes.metabolism.Metabolism.update_amino_acid_targets" title="Link to this definition"></a></dt>
<dd><p>Finds new amino acid concentration targets based on difference in
supply and number of amino acids used in polypeptide_elongation.
Skips updates to molecules defined in self.aa_targets_not_updated:
- L-SELENOCYSTEINE: rare AA that led to high variability when updated</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>counts_to_molar</strong> (<em>Unum</em>) – conversion from counts to molar</p></li>
<li><p><strong>count_diff</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)"><em>dict</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.13)"><em>float</em></a><em>]</em>)</p></li>
<li><p><strong>amino_acid_counts</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)"><em>dict</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.13)"><em>float</em></a><em>]</em>)</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><code class="docutils literal notranslate"><span class="pre">{AA</span> <span class="pre">name</span> <span class="pre">(str):</span> <span class="pre">new</span> <span class="pre">target</span> <span class="pre">AA</span> <span class="pre">conc</span> <span class="pre">(float</span> <span class="pre">with</span> <span class="pre">mol/volume</span> <span class="pre">units)}</span></code></p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)">dict</a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)">str</a>, <em>Unum</em>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ecoli.processes.metabolism.Metabolism.update_condition">
<span class="sig-name descname"><span class="pre">update_condition</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">timestep</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">states</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ecoli/processes/metabolism.html#Metabolism.update_condition"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ecoli.processes.metabolism.Metabolism.update_condition" title="Link to this definition"></a></dt>
<dd><p>See <a class="reference internal" href="ecoli.processes.partition.html#ecoli.processes.partition.Requester.update_condition" title="ecoli.processes.partition.Requester.update_condition"><code class="xref py py-meth docutils literal notranslate"><span class="pre">update_condition()</span></code></a>.</p>
</dd></dl>

</dd></dl>

</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="ecoli.processes.global_clock.html" class="btn btn-neutral float-left" title="ecoli.processes.global_clock" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="ecoli.processes.metabolism_redux.html" class="btn btn-neutral float-right" title="ecoli.processes.metabolism_redux" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2021-2025, The Vivarium E. coli Authors.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>