

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>reconstruction.ecoli.fit_sim_data_1 &mdash; Vivarium E. coli 1.0.0 documentation</title>
      <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=b86133f3" />
      <link rel="stylesheet" type="text/css" href="../../../_static/css/theme.css?v=e59714d7" />

  
      <script src="../../../_static/jquery.js?v=5d32c60e"></script>
      <script src="../../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../../../_static/documentation_options.js?v=8d563738"></script>
      <script src="../../../_static/doctools.js?v=9bcbadda"></script>
      <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
      <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="reconstruction.ecoli.knowledge_base_raw" href="reconstruction.ecoli.knowledge_base_raw.html" />
    <link rel="prev" title="reconstruction.ecoli.scripts.update_biocyc_files" href="reconstruction.ecoli.scripts.update_biocyc_files.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../index.html" class="icon icon-home">
            Vivarium E. coli
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../../stores.html">Stores</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../processes.html">Processes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../composites.html">Composites</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../experiments.html">Experiments</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../workflows.html">Workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../output.html">Output</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../tutorial.html">Tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../docs.html">Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../hpc.html">HPC Clusters</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../gcloud.html">Google Cloud</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../ci.html">Continuous Integration</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../pycharm.html">PyCharm Setup</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../../api_ref.html">API Reference</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../ecoli/ecoli.html"><code class="docutils literal notranslate"><span class="pre">ecoli</span></code></a></li>
<li class="toctree-l2 current"><a class="reference internal" href="reconstruction.html"><code class="docutils literal notranslate"><span class="pre">reconstruction</span></code></a><ul class="current">
<li class="toctree-l3 current"><a class="reference internal" href="reconstruction.ecoli.html"><code class="docutils literal notranslate"><span class="pre">reconstruction.ecoli</span></code></a><ul class="current">
<li class="toctree-l4"><a class="reference internal" href="reconstruction.ecoli.compendium.html"><code class="docutils literal notranslate"><span class="pre">reconstruction.ecoli.compendium</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="reconstruction.ecoli.dataclasses.html"><code class="docutils literal notranslate"><span class="pre">reconstruction.ecoli.dataclasses</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="reconstruction.ecoli.scripts.html"><code class="docutils literal notranslate"><span class="pre">reconstruction.ecoli.scripts</span></code></a></li>
<li class="toctree-l4 current"><a class="current reference internal" href="#"><code class="docutils literal notranslate"><span class="pre">reconstruction.ecoli.fit_sim_data_1</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="reconstruction.ecoli.knowledge_base_raw.html"><code class="docutils literal notranslate"><span class="pre">reconstruction.ecoli.knowledge_base_raw</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="reconstruction.ecoli.simulation_data.html"><code class="docutils literal notranslate"><span class="pre">reconstruction.ecoli.simulation_data</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="reconstruction.tests.html"><code class="docutils literal notranslate"><span class="pre">reconstruction.tests</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="reconstruction.spreadsheets.html"><code class="docutils literal notranslate"><span class="pre">reconstruction.spreadsheets</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../runscripts/runscripts.html"><code class="docutils literal notranslate"><span class="pre">runscripts</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="../validation/validation.html"><code class="docutils literal notranslate"><span class="pre">validation</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="../wholecell/wholecell.html"><code class="docutils literal notranslate"><span class="pre">wholecell</span></code></a></li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">Vivarium E. coli</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../api_ref.html">API Reference</a></li>
          <li class="breadcrumb-item"><a href="reconstruction.html"><code class="docutils literal notranslate"><span class="pre">reconstruction</span></code></a></li>
          <li class="breadcrumb-item"><a href="reconstruction.ecoli.html"><code class="docutils literal notranslate"><span class="pre">reconstruction.ecoli</span></code></a></li>
      <li class="breadcrumb-item active"><code class="docutils literal notranslate"><span class="pre">reconstruction.ecoli.fit_sim_data_1</span></code></li>
      <li class="wy-breadcrumbs-aside">
            <a href="../../../_sources/reference/api/reconstruction/reconstruction.ecoli.fit_sim_data_1.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="module-reconstruction.ecoli.fit_sim_data_1">
<span id="reconstruction-ecoli-fit-sim-data-1"></span><h1><code class="docutils literal notranslate"><span class="pre">reconstruction.ecoli.fit_sim_data_1</span></code><a class="headerlink" href="#module-reconstruction.ecoli.fit_sim_data_1" title="Link to this heading"></a></h1>
<p>The parca, aka parameter calculator.</p>
<p>TODO: establish a controlled language for function behaviors (i.e. create* set* fit*)
TODO: functionalize so that values are not both set and returned from some methods</p>
<dl class="py function">
<dt class="sig sig-object py" id="reconstruction.ecoli.fit_sim_data_1.adjust_promoters">
<span class="sig-prename descclassname"><span class="pre">reconstruction.ecoli.fit_sim_data_1.</span></span><span class="sig-name descname"><span class="pre">adjust_promoters</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_specs</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/fit_sim_data_1.html#adjust_promoters"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.fit_sim_data_1.adjust_promoters" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="reconstruction.ecoli.fit_sim_data_1.apply_updates">
<span class="sig-prename descclassname"><span class="pre">reconstruction.ecoli.fit_sim_data_1.</span></span><span class="sig-name descname"><span class="pre">apply_updates</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">func</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">labels</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dest</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cpus</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/fit_sim_data_1.html#apply_updates"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.fit_sim_data_1.apply_updates" title="Link to this definition"></a></dt>
<dd><p>Use multiprocessing (if cpus &gt; 1) to apply args to a function to get
dictionary updates for a destination dictionary.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>func</strong> (<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Callable" title="(in Python v3.13)"><em>Callable</em></a><em>[</em><em>[</em><em>...</em><em>]</em><em>, </em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)"><em>dict</em></a><em>]</em>) – function to call with args</p></li>
<li><p><strong>args</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.13)"><em>list</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)"><em>tuple</em></a><em>]</em>) – list of args to apply to func</p></li>
<li><p><strong>labels</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.13)"><em>list</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>]</em>) – label for each set of args for exception information</p></li>
<li><p><strong>dest</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)"><em>dict</em></a>) – destination dictionary that will be updated with results
from each function call</p></li>
<li><p><strong>cpus</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a>) – number of cpus to use</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="reconstruction.ecoli.fit_sim_data_1.basal_specs">
<span class="sig-prename descclassname"><span class="pre">reconstruction.ecoli.fit_sim_data_1.</span></span><span class="sig-name descname"><span class="pre">basal_specs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_specs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">disable_ribosome_capacity_fitting</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">disable_rnapoly_capacity_fitting</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">variable_elongation_transcription</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">variable_elongation_translation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/fit_sim_data_1.html#basal_specs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.fit_sim_data_1.basal_specs" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="reconstruction.ecoli.fit_sim_data_1.buildBasalCellSpecifications">
<span class="sig-prename descclassname"><span class="pre">reconstruction.ecoli.fit_sim_data_1.</span></span><span class="sig-name descname"><span class="pre">buildBasalCellSpecifications</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">variable_elongation_transcription</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">variable_elongation_translation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">disable_ribosome_capacity_fitting</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">disable_rnapoly_capacity_fitting</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/fit_sim_data_1.html#buildBasalCellSpecifications"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.fit_sim_data_1.buildBasalCellSpecifications" title="Link to this definition"></a></dt>
<dd><p>Creates cell specifications for the basal condition by fitting expression.
Relies on expressionConverge() to set the expression and update masses.</p>
<section id="inputs">
<h2>Inputs<a class="headerlink" href="#inputs" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>disable_ribosome_capacity_fitting (bool) - if True, ribosome expression</p></li>
</ul>
<p>is not fit
- disable_rnapoly_capacity_fitting (bool) - if True, RNA polymerase
expression is not fit</p>
</section>
<section id="requires">
<h2>Requires<a class="headerlink" href="#requires" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>Metabolite concentrations based on ‘minimal’ nutrients</p></li>
<li><p>‘basal’ RNA expression</p></li>
<li><p>‘basal’ doubling time</p></li>
</ul>
</section>
<section id="modifies">
<h2>Modifies<a class="headerlink" href="#modifies" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>Average mass values of the cell</p></li>
<li><p>cistron expression</p></li>
<li><p>RNA expression and synthesis probabilities</p></li>
</ul>
<dl class="field-list">
<dt class="field-odd">returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>- dict {‘basal’</strong> –</p>
<dl class="simple">
<dt>‘concDict’ {metabolite_name (str): concentration (float with units)} -</dt><dd><p>dictionary of concentrations for each metabolite with a concentration</p>
</dd>
<dt>‘fit_cistron_expression’ (array of floats) - hypothetical expression for</dt><dd><p>each RNA cistron post-fit, total normalized to 1, if all
transcription units were monocistronic</p>
</dd>
</dl>
<p>‘expression’ (array of floats) - expression for each RNA, total normalized to 1
‘doubling_time’ (float with units) - cell doubling time
‘synthProb’ (array of floats) - synthesis probability for each RNA,</p>
<blockquote>
<div><p>total normalized to 1</p>
</div></blockquote>
<p>‘avgCellDryMassInit’ (float with units) - average initial cell dry mass
‘fitAvgSolubleTargetMolMass’ (float with units) - the adjusted dry mass</p>
<blockquote>
<div><p>of the soluble fraction of a cell</p>
</div></blockquote>
<ul class="simple">
<li><dl class="simple">
<dt>bulkContainer (np.ndarray object) - Two columns: ‘id’ for name and ‘count’</dt><dd><p>for expected counts based on expression of all bulk molecules</p>
</dd>
</dl>
</li>
</ul>
</dd>
<dt class="field-even">rtype<span class="colon">:</span></dt>
<dd class="field-even"><p>dict} with the following keys in the dict from key ‘basal’:</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<ul class="simple">
<li><p>TODO - sets sim_data attributes and returns values - change to only return values</p></li>
</ul>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="reconstruction.ecoli.fit_sim_data_1.buildCombinedConditionCellSpecifications">
<span class="sig-prename descclassname"><span class="pre">reconstruction.ecoli.fit_sim_data_1.</span></span><span class="sig-name descname"><span class="pre">buildCombinedConditionCellSpecifications</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_specs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">variable_elongation_transcription</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">variable_elongation_translation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">disable_ribosome_capacity_fitting</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">disable_rnapoly_capacity_fitting</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/fit_sim_data_1.html#buildCombinedConditionCellSpecifications"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.fit_sim_data_1.buildCombinedConditionCellSpecifications" title="Link to this definition"></a></dt>
<dd><p>Creates cell specifications for sets of transcription factors being active.
These sets include conditions like ‘with_aa’ or ‘no_oxygen’ where multiple
transcription factors will be active at the same time.</p>
<section id="id1">
<h2>Inputs<a class="headerlink" href="#id1" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>cell_specs {condition (str): dict} - information about each individual</p></li>
</ul>
<p>transcription factor condition
- disable_ribosome_capacity_fitting (bool) - if True, ribosome expression
is not fit
- disable_rnapoly_capacity_fitting (bool) - if True, RNA polymerase
expression is not fit</p>
</section>
<section id="id2">
<h2>Requires<a class="headerlink" href="#id2" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>Metabolite concentrations based on nutrients for the condition</p></li>
<li><p>Adjusted ‘basal’ RNA expression</p></li>
<li><p>Doubling time for the combined condition</p></li>
<li><p>Fold changes in expression for each gene given the TF</p></li>
</ul>
</section>
<section id="id3">
<h2>Modifies<a class="headerlink" href="#id3" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>cell_specs dictionary for each combined condition</p></li>
<li><p>RNA expression and synthesis probabilities for each combined condition</p></li>
</ul>
<p class="rubric">Notes</p>
<ul class="simple">
<li><p>TODO - determine how to handle fold changes when multiple TFs change the</p></li>
</ul>
<p>same gene because multiplying both fold changes together might not be
appropriate</p>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="reconstruction.ecoli.fit_sim_data_1.buildTfConditionCellSpecifications">
<span class="sig-prename descclassname"><span class="pre">reconstruction.ecoli.fit_sim_data_1.</span></span><span class="sig-name descname"><span class="pre">buildTfConditionCellSpecifications</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tf</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">variable_elongation_transcription</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">variable_elongation_translation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">disable_ribosome_capacity_fitting</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">disable_rnapoly_capacity_fitting</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/fit_sim_data_1.html#buildTfConditionCellSpecifications"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.fit_sim_data_1.buildTfConditionCellSpecifications" title="Link to this definition"></a></dt>
<dd><p>Creates cell specifications for a given transcription factor by
fitting expression. Will set for the active and inactive TF condition.
Relies on expressionConverge() to set the expression and masses.
Uses fold change data relative to the ‘basal’ condition to determine
expression for a given TF.</p>
<section id="id4">
<h2>Inputs<a class="headerlink" href="#id4" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>tf (str) - label for the transcription factor to fit (eg. ‘CPLX-125’)</p></li>
<li><p>disable_ribosome_capacity_fitting (bool) - if True, ribosome expression</p></li>
</ul>
<p>is not fit
- disable_rnapoly_capacity_fitting (bool) - if True, RNA polymerase
expression is not fit</p>
</section>
<section id="id5">
<h2>Requires<a class="headerlink" href="#id5" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>Metabolite concentrations based on nutrients for the TF</p></li>
<li><p>Adjusted ‘basal’ cistron expression</p></li>
<li><p>Doubling time for the TF</p></li>
<li><p>Fold changes in expression for each gene given the TF</p></li>
</ul>
<dl class="field-list">
<dt class="field-odd">returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>- dict {tf + ‘__active’/’__inactive’</strong> –</p>
<dl class="simple">
<dt>‘concDict’ {metabolite_name (str): concentration (float with units)} -</dt><dd><p>dictionary of concentrations for each metabolite with a concentration</p>
</dd>
</dl>
<p>‘expression’ (array of floats) - expression for each RNA, total normalized to 1
‘doubling_time’ (float with units) - cell doubling time
‘synthProb’ (array of floats) - synthesis probability for each RNA,</p>
<blockquote>
<div><p>total normalized to 1</p>
</div></blockquote>
<dl class="simple">
<dt>‘cistron_expression’ (array of floats) - hypothetical expression for</dt><dd><p>each RNA cistron, calculated from basal cistron expression levels
and fold change data</p>
</dd>
<dt>‘fit_cistron_expression’ (array of floats) - hypothetical expression for</dt><dd><p>each RNA cistron post-fit, total normalized to 1, if all
transcription units were monocistronic</p>
</dd>
</dl>
<p>‘avgCellDryMassInit’ (float with units) - average initial cell dry mass
‘fitAvgSolubleTargetMolMass’ (float with units) - the adjusted dry mass</p>
<blockquote>
<div><p>of the soluble fraction of a cell</p>
</div></blockquote>
<ul class="simple">
<li><dl class="simple">
<dt>bulkContainer (np.ndarray object) - Two columns: ‘id’ for name and ‘count’</dt><dd><p>for expected counts based on expression of all bulk molecules</p>
</dd>
</dl>
</li>
</ul>
</dd>
<dt class="field-even">rtype<span class="colon">:</span></dt>
<dd class="field-even"><p>dict} with the following keys in each dict:</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="reconstruction.ecoli.fit_sim_data_1.calculateBulkDistributions">
<span class="sig-prename descclassname"><span class="pre">reconstruction.ecoli.fit_sim_data_1.</span></span><span class="sig-name descname"><span class="pre">calculateBulkDistributions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">expression</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">concDict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">avgCellDryMassInit</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">doubling_time</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/fit_sim_data_1.html#calculateBulkDistributions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.fit_sim_data_1.calculateBulkDistributions" title="Link to this definition"></a></dt>
<dd><p>Finds a distribution of copy numbers for macromolecules. While RNA and protein
expression can be approximated using well-described statistical     distributions,
complexes require absolute copy numbers. To get these distributions, this
function instantiates many cells with a reduced set of molecules, forms complexes,
and iterates through equilibrium and two-component system processes until
metabolite counts reach a steady-state. It then computes the resulting
statistical distributions.</p>
<section id="id6">
<h2>Requires<a class="headerlink" href="#id6" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>N_SEEDS (int) - the number of instantiated cells</p></li>
</ul>
</section>
<section id="id7">
<h2>Inputs<a class="headerlink" href="#id7" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>expression (array of floats) - expression for each RNA, normalized to 1</p></li>
<li><p>concDict {metabolite (str): concentration (float with units of mol/volume)} -</p></li>
</ul>
<p>dictionary for concentrations of each metabolite with location tag
- avgCellDryMassInit (float with units of mass) - initial dry cell mass
- doubling_time (float with units of time) - doubling time for condition</p>
<dl class="field-list simple">
<dt class="field-odd">returns<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>- bulkAverageContainer (np.ndarray object) - Two columns</strong> (<em>‘id’ for name and ‘count’</em>) – for the mean of the counts of all bulk molecules</p></li>
<li><p><strong>- bulkDeviationContainer (np.ndarray object) - Two columns</strong> (<em>‘id’ for name and ‘count’</em>) – for the standard deviation of the counts of all bulk molecules</p></li>
<li><p><strong>- proteinMonomerAverageContainer (np.ndarray object) - Two columns</strong> (<em>‘id’ for name and ‘count’</em>) – for the mean of the counts of all protein monomers</p></li>
<li><p><strong>- proteinMonomerDeviationContainer (np.ndarray object) - Two columns</strong> (<em>‘id’ for name and ‘count’</em>) – for the standard deviation of the counts of all protein monomers</p></li>
</ul>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="reconstruction.ecoli.fit_sim_data_1.calculateMinPolymerizingEnzymeByProductDistribution">
<span class="sig-prename descclassname"><span class="pre">reconstruction.ecoli.fit_sim_data_1.</span></span><span class="sig-name descname"><span class="pre">calculateMinPolymerizingEnzymeByProductDistribution</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">productLengths</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">elongationRates</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">netLossRate</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">productCounts</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/fit_sim_data_1.html#calculateMinPolymerizingEnzymeByProductDistribution"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.fit_sim_data_1.calculateMinPolymerizingEnzymeByProductDistribution" title="Link to this definition"></a></dt>
<dd><p>Compute the number of ribosomes required to maintain steady state.</p>
<p>dP/dt = production rate - loss rate
dP/dt = e_r * (1/L) * R - (k_loss * P)</p>
<p>At steady state: dP/dt = 0
R = sum over i ((L_i / e_r) * k_loss_i * P_i)</p>
<p>Multiplying both sides by volume gives an equation in terms of counts.</p>
<p>P = protein concentration
e_r = polypeptide elongation rate per ribosome
L = protein length
R = ribosome concentration
k_loss = net protein loss rate
i = ith protein</p>
<section id="id8">
<h2>Inputs<a class="headerlink" href="#id8" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>productLengths (array of ints with units of amino_acids) - L, protein lengths</p></li>
<li><p>elongationRates (array of ints with units of amino_acid/time) e_r, polypeptide elongation rate</p></li>
<li><p>netLossRate (array of floats with units of 1/time) - k_loss, protein loss rate</p></li>
<li><p>productCounts (array of floats) - P, protein counts</p></li>
</ul>
<dl class="field-list simple">
<dt class="field-odd">returns<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><em>- float with dimensionless units for the number of ribosomes required to</em></p></li>
<li><p><em>maintain steady state</em></p></li>
</ul>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="reconstruction.ecoli.fit_sim_data_1.calculateMinPolymerizingEnzymeByProductDistributionRNA">
<span class="sig-prename descclassname"><span class="pre">reconstruction.ecoli.fit_sim_data_1.</span></span><span class="sig-name descname"><span class="pre">calculateMinPolymerizingEnzymeByProductDistributionRNA</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">productLengths</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">elongationRates</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">netLossRate</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/fit_sim_data_1.html#calculateMinPolymerizingEnzymeByProductDistributionRNA"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.fit_sim_data_1.calculateMinPolymerizingEnzymeByProductDistributionRNA" title="Link to this definition"></a></dt>
<dd><p>Compute the number of RNA polymerases required to maintain steady state of mRNA.</p>
<p>dR/dt = production rate - loss rate
dR/dt = e_r * (1/L) * RNAp - k_loss</p>
<p>At steady state: dR/dt = 0
RNAp = sum over i ((L_i / e_r) * k_loss_i)</p>
<p>Multiplying both sides by volume gives an equation in terms of counts.</p>
<p>R = mRNA transcript concentration
e_r = transcript elongation rate per RNAp
L = transcript length
RNAp = RNAp concentration
k_loss = net transcript loss rate (unit: concentration / time)
i = ith transcript</p>
<section id="id9">
<h2>Inputs<a class="headerlink" href="#id9" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>productLengths (array of ints with units of nucleotides) - L, transcript lengths</p></li>
<li><p>elongationRates (array of ints with units of nucleotide/time) - e_r, transcript elongation rate</p></li>
<li><p>netLossRate (array of floats with units of 1/time) - k_loss, transcript loss rate</p></li>
</ul>
<dl class="field-list simple">
<dt class="field-odd">returns<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><em>- float with dimensionless units for the number of RNA polymerases required to</em></p></li>
<li><p><em>maintain steady state</em></p></li>
</ul>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="reconstruction.ecoli.fit_sim_data_1.calculatePromoterBoundProbability">
<span class="sig-prename descclassname"><span class="pre">reconstruction.ecoli.fit_sim_data_1.</span></span><span class="sig-name descname"><span class="pre">calculatePromoterBoundProbability</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_specs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/fit_sim_data_1.html#calculatePromoterBoundProbability"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.fit_sim_data_1.calculatePromoterBoundProbability" title="Link to this definition"></a></dt>
<dd><p>Calculate the probability that a transcription factor is bound to its
associated promoter for all simulated growth conditions. The bulk
average concentrations calculated for TFs and their ligands are used to
compute the probabilities based on the type (0CS, 1CS, 2CS) of the TF.</p>
<section id="id10">
<h2>Requires<a class="headerlink" href="#id10" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>Bulk average counts of transcription factors and associated ligands</p></li>
</ul>
<p>for each condition (in cell_specs)</p>
<dl class="field-list simple">
<dt class="field-odd">returns<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>- pPromoterBound</strong> (<em>Probability that a transcription factor is bound to</em>)</p></li>
<li><p><em>its promoter, per growth condition and TF. Each probability is indexed by</em></p></li>
<li><p><em>pPromoterBound[condition][TF].</em></p></li>
</ul>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="reconstruction.ecoli.fit_sim_data_1.calculateRnapRecruitment">
<span class="sig-prename descclassname"><span class="pre">reconstruction.ecoli.fit_sim_data_1.</span></span><span class="sig-name descname"><span class="pre">calculateRnapRecruitment</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_specs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/fit_sim_data_1.html#calculateRnapRecruitment"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.fit_sim_data_1.calculateRnapRecruitment" title="Link to this definition"></a></dt>
<dd><p>Constructs the basal_prob vector and delta_prob matrix from values of r.
The basal_prob vector holds the basal transcription probabilities of each
transcription unit. The delta_prob matrix holds the differences in
transcription probabilities when transcription factors bind to the
promoters of each transcription unit. Both values are stored in sim_data.</p>
<section id="id11">
<h2>Requires<a class="headerlink" href="#id11" title="Link to this heading"></a></h2>
<ul>
<li><dl>
<dt>cell_specs[‘basal’]:</dt><dd><ul class="simple">
<li><p>[‘r_vector’]: Fit parameters on how the recruitment of a TF affects the expression</p></li>
</ul>
<p>of a gene. High (positive) values of r indicate that the TF binding
increases the probability that the gene is expressed.
- [‘r_columns’]: mapping of column name to index in r</p>
</dd>
</dl>
</li>
</ul>
</section>
<section id="id12">
<h2>Modifies<a class="headerlink" href="#id12" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>Rescales values in basal_prob such that all values are positive</p></li>
<li><p>Adds basal_prob and delta_prob arrays to sim_data</p></li>
</ul>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="reconstruction.ecoli.fit_sim_data_1.calculateTranslationSupply">
<span class="sig-prename descclassname"><span class="pre">reconstruction.ecoli.fit_sim_data_1.</span></span><span class="sig-name descname"><span class="pre">calculateTranslationSupply</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">doubling_time</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bulkContainer</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">avgCellDryMassInit</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/fit_sim_data_1.html#calculateTranslationSupply"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.fit_sim_data_1.calculateTranslationSupply" title="Link to this definition"></a></dt>
<dd><p>Returns the supply rates of all amino acids to translation given the desired
doubling time. This creates a limit on the polypeptide elongation process,
and thus on growth. The amino acid supply rate is found by calculating the
concentration of amino acids per gram dry cell weight and multiplying by the
loss to dilution given doubling time.</p>
<section id="id13">
<h2>Inputs<a class="headerlink" href="#id13" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>doubling_time (float with units of time) - measured doubling times given the condition</p></li>
<li><dl class="simple">
<dt>bulkContainer (np.ndarray object) - Two columns: ‘id’ for name and ‘count’</dt><dd><p>for count of all bulk molecules</p>
</dd>
</dl>
</li>
<li><p>avgCellDryMassInit (float with units of mass) - the average initial cell dry mass</p></li>
</ul>
<p class="rubric">Notes</p>
<ul class="simple">
<li><p>The supply of amino acids should not be based on a desired doubling time,</p></li>
</ul>
<p>but should come from a more mechanistic basis. This would allow simulations
of environmental shifts in which the doubling time is unknown.</p>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="reconstruction.ecoli.fit_sim_data_1.crc32">
<span class="sig-prename descclassname"><span class="pre">reconstruction.ecoli.fit_sim_data_1.</span></span><span class="sig-name descname"><span class="pre">crc32</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">arrays</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">initial</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/fit_sim_data_1.html#crc32"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.fit_sim_data_1.crc32" title="Link to this definition"></a></dt>
<dd><p>Return a CRC32 checksum of the given ndarrays.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>arrays</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a>)</p></li>
<li><p><strong>initial</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a>)</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="reconstruction.ecoli.fit_sim_data_1.createBulkContainer">
<span class="sig-prename descclassname"><span class="pre">reconstruction.ecoli.fit_sim_data_1.</span></span><span class="sig-name descname"><span class="pre">createBulkContainer</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">expression</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">doubling_time</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/fit_sim_data_1.html#createBulkContainer"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.fit_sim_data_1.createBulkContainer" title="Link to this definition"></a></dt>
<dd><p>Creates a container that tracks the counts of all bulk molecules. Relies on
totalCountIdDistributionRNA and totalCountIdDistributionProtein to set the
counts and IDs of all RNAs and proteins.</p>
<section id="id14">
<h2>Inputs<a class="headerlink" href="#id14" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>expression (array of floats) - relative frequency distribution of RNA expression</p></li>
<li><p>doubling_time (float with units of time) - measured doubling time given the condition</p></li>
</ul>
<dl class="field-list simple">
<dt class="field-odd">returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>- bulkContainer (np.ndarray object) - Two columns</strong> – for count of all bulk molecules</p>
</dd>
<dt class="field-even">rtype<span class="colon">:</span></dt>
<dd class="field-even"><p>‘id’ for name and ‘count’</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="reconstruction.ecoli.fit_sim_data_1.expressionConverge">
<span class="sig-prename descclassname"><span class="pre">reconstruction.ecoli.fit_sim_data_1.</span></span><span class="sig-name descname"><span class="pre">expressionConverge</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">expression</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">concDict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">doubling_time</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Km</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">conditionKey</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">variable_elongation_transcription</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">variable_elongation_translation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">disable_ribosome_capacity_fitting</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">disable_rnapoly_capacity_fitting</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/fit_sim_data_1.html#expressionConverge"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.fit_sim_data_1.expressionConverge" title="Link to this definition"></a></dt>
<dd><p>Iteratively fits synthesis probabilities for RNA. Calculates initial
expression based on gene expression data and makes adjustments to match
physiological constraints for ribosome and RNAP counts. Relies on
fitExpression() to converge</p>
<section id="id15">
<h2>Inputs<a class="headerlink" href="#id15" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>expression (array of floats) - expression for each RNA, normalized to 1</p></li>
<li><p>concDict {metabolite (str): concentration (float with units of mol/volume)} -</p></li>
</ul>
<p>dictionary for concentrations of each metabolite with location tag
- doubling_time (float with units of time) - doubling time
- Km (array of floats with units of mol/volume) - Km for each RNA associated
with RNases
- disable_ribosome_capacity_fitting (bool) - if True, ribosome expression
is not fit
- disable_rnapoly_capacity_fitting (bool) - if True, RNA polymerase
expression is not fit</p>
</section>
<section id="id16">
<h2>Requires<a class="headerlink" href="#id16" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>MAX_FITTING_ITERATIONS (int) - number of iterations to adjust expression</p></li>
</ul>
<p>before an exception is raised
- FITNESS_THRESHOLD (float) - acceptable change from one iteration to break
the fitting loop</p>
<dl class="field-list simple">
<dt class="field-odd">returns<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><em>- expression (array of floats) - adjusted expression for each RNA,</em></p></li>
<li><p><em>normalized to 1</em></p></li>
<li><p><em>- synthProb (array of floats) - synthesis probability for each RNA which</em></p></li>
<li><p><em>accounts for expression and degradation rate, normalized to 1</em></p></li>
<li><p><em>- avgCellDryMassInit (float with units of mass) - expected initial dry cell mass</em></p></li>
<li><p><em>- fitAvgSolubleTargetMolMass (float with units of mass) - the adjusted dry mass</em></p></li>
<li><p><em>of the soluble fraction of a cell</em></p></li>
<li><p><strong>- bulkContainer (np.ndarray object) - Two columns</strong> (<em>‘id’ for name and ‘count’</em>) – for expected counts based on expression of all bulk molecules</p></li>
</ul>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="reconstruction.ecoli.fit_sim_data_1.expressionFromConditionAndFoldChange">
<span class="sig-prename descclassname"><span class="pre">reconstruction.ecoli.fit_sim_data_1.</span></span><span class="sig-name descname"><span class="pre">expressionFromConditionAndFoldChange</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">transcription</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">condPerturbations</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tfFCs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/fit_sim_data_1.html#expressionFromConditionAndFoldChange"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.fit_sim_data_1.expressionFromConditionAndFoldChange" title="Link to this definition"></a></dt>
<dd><p>Adjusts expression of RNA based on fold changes from basal for a given
condition. Since fold changes are reported for individual RNA cistrons, the
changes are applied to the basal expression levels of each cistron and the
resulting vector is mapped back to RNA expression through nonnegative least
squares. For genotype perturbations, the expression of all RNAs that include
the given cistron are set to the given value.</p>
<section id="id17">
<h2>Inputs<a class="headerlink" href="#id17" title="Link to this heading"></a></h2>
<ul class="simple">
<li><dl class="simple">
<dt>transcription: Instance of the Transcription class from</dt><dd><p>reconstruction.ecoli.dataclasses.process.transcription</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>condPerturbations {cistron ID (str): fold change (float)} -</dt><dd><p>dictionary of fold changes for cistrons based on the given condition</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>tfFCs {cistron ID (str): fold change (float)} -</dt><dd><p>dictionary of fold changes for cistrons based on transcription factors
in the given condition</p>
</dd>
</dl>
</li>
</ul>
<dl class="field-list simple">
<dt class="field-odd">returns<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><em>- expression (array of floats) - adjusted expression for each RNA,</em></p></li>
<li><p><em>normalized to 1</em></p></li>
</ul>
</dd>
</dl>
<p class="rubric">Notes</p>
<ul class="simple">
<li><p>TODO (Travis) - Might not properly handle if an RNA is adjusted from both a</p></li>
</ul>
<p>perturbation and a transcription factor, currently RNA self regulation is not
included in tfFCs</p>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="reconstruction.ecoli.fit_sim_data_1.final_adjustments">
<span class="sig-prename descclassname"><span class="pre">reconstruction.ecoli.fit_sim_data_1.</span></span><span class="sig-name descname"><span class="pre">final_adjustments</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_specs</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/fit_sim_data_1.html#final_adjustments"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.fit_sim_data_1.final_adjustments" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="reconstruction.ecoli.fit_sim_data_1.fitCondition">
<span class="sig-prename descclassname"><span class="pre">reconstruction.ecoli.fit_sim_data_1.</span></span><span class="sig-name descname"><span class="pre">fitCondition</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">spec</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">condition</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/fit_sim_data_1.html#fitCondition"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.fit_sim_data_1.fitCondition" title="Link to this definition"></a></dt>
<dd><p>Takes a given condition and returns the predicted bulk average, bulk deviation,
protein monomer average, protein monomer deviation, and amino acid supply to
translation. This relies on calculateBulkDistributions and calculateTranslationSupply.</p>
<section id="id18">
<h2>Inputs<a class="headerlink" href="#id18" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>condition (str) - condition to fit (eg ‘CPLX0-7705__active’)</p></li>
<li><p>spec {property (str): property values} - cell specifications for the given condition.</p></li>
</ul>
<p>This function uses the specs “expression”, “concDict”, “avgCellDryMassInit”,
and “doubling_time”</p>
<dl class="field-list">
<dt class="field-odd">returns<span class="colon">:</span></dt>
<dd class="field-odd"><ul>
<li><p><strong>- A dictionary {condition (str)</strong> (<em>spec (dict)} with the updated spec dictionary</em>)</p></li>
<li><p><em>with the following values updated</em> –</p>
<ul class="simple">
<li><dl class="simple">
<dt>bulkAverageContainer (np.ndarray object) - Two columns: ‘id’ for name and ‘count’</dt><dd><p>for the mean of the counts of all bulk molecules</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>bulkDeviationContainer (np.ndarray object) - Two columns: ‘id’ for name and ‘count’</dt><dd><p>for the standard deviation of the counts of all bulk molecules</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>proteinMonomerAverageContainer (np.ndarray object) - Two columns: ‘id’ for name and ‘count’</dt><dd><p>for the mean of the counts of all protein monomers</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>proteinMonomerDeviationContainer (np.ndarray object) - Two columns: ‘id’ for name and ‘count’</dt><dd><p>for the standard deviation of the counts of all protein monomers</p>
</dd>
</dl>
</li>
<li><p>translation_aa_supply (array with units of mol/(mass.time)) - the supply rates</p></li>
</ul>
<p>for each amino acid to translation</p>
</li>
</ul>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="reconstruction.ecoli.fit_sim_data_1.fitExpression">
<span class="sig-prename descclassname"><span class="pre">reconstruction.ecoli.fit_sim_data_1.</span></span><span class="sig-name descname"><span class="pre">fitExpression</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bulkContainer</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">doubling_time</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">avgCellDryMassInit</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Km</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/fit_sim_data_1.html#fitExpression"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.fit_sim_data_1.fitExpression" title="Link to this definition"></a></dt>
<dd><p>Determines expression and synthesis probabilities for RNA molecules to fit
protein levels and RNA degradation rates. Assumes a steady state analysis
where the RNA synthesis probability will be the same as the degradation rate.
If no Km is given, then RNA degradation is assumed to be linear otherwise
degradation is calculated based on saturation with RNases.</p>
<section id="id19">
<h2>Inputs<a class="headerlink" href="#id19" title="Link to this heading"></a></h2>
<ul class="simple">
<li><dl class="simple">
<dt>bulkContainer (np.ndarray object) - Two columns: ‘id’ for name and ‘count’</dt><dd><p>for expected count based on expression of all bulk molecules</p>
</dd>
</dl>
</li>
<li><p>doubling_time (float with units of time) - doubling time</p></li>
<li><p>avgCellDryMassInit (float with units of mass) - expected initial dry cell mass</p></li>
<li><p>Km (array of floats with units of mol/volume) - Km for each RNA associated</p></li>
</ul>
<p>with RNases</p>
</section>
<section id="id20">
<h2>Modifies<a class="headerlink" href="#id20" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>bulkContainer counts of RNA and proteins</p></li>
</ul>
<dl class="field-list simple">
<dt class="field-odd">returns<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><em>- expression (array of floats) - adjusted expression for each RNA,</em></p></li>
<li><p><em>normalized to 1</em></p></li>
<li><p><em>- synth_prob (array of floats) - synthesis probability for each RNA which</em></p></li>
<li><p><em>accounts for expression and degradation rate, normalized to 1</em></p></li>
<li><p><em>- fit_cistron_expression (array of floats) - target expression levels of</em></p></li>
<li><p><em>each cistron (gene) used to calculate RNA expression levels</em></p></li>
<li><p><em>- cistron_expression_res (array of floats) - the residuals of the NNLS</em></p></li>
<li><p><em>problem solved to calculate RNA expression levels</em></p></li>
</ul>
</dd>
</dl>
<p class="rubric">Notes</p>
<ul class="simple">
<li><p>TODO - sets bulkContainer counts and returns values - change to only return values</p></li>
</ul>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="reconstruction.ecoli.fit_sim_data_1.fitLigandConcentrations">
<span class="sig-prename descclassname"><span class="pre">reconstruction.ecoli.fit_sim_data_1.</span></span><span class="sig-name descname"><span class="pre">fitLigandConcentrations</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_specs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/fit_sim_data_1.html#fitLigandConcentrations"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.fit_sim_data_1.fitLigandConcentrations" title="Link to this definition"></a></dt>
<dd><p>Using the fit values of pPromoterBound, updates the set concentrations of
ligand metabolites and the kd’s of the ligand-TF binding reactions.</p>
<section id="id21">
<h2>Requires<a class="headerlink" href="#id21" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>Fitted pPromoterBound: probabilities that a TF will bind to its promoter,</p></li>
</ul>
<p>fit by function fitPromoterBoundProbability().</p>
</section>
<section id="id22">
<h2>Inputs<a class="headerlink" href="#id22" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>cell_specs {condition (str): dict} - information about each condition</p></li>
</ul>
</section>
<section id="id23">
<h2>Modifies<a class="headerlink" href="#id23" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>Set concentrations of metabolites that are ligands in 1CS</p></li>
<li><p>kd’s of equilibrium reactions in 1CS</p></li>
</ul>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="reconstruction.ecoli.fit_sim_data_1.fitMaintenanceCosts">
<span class="sig-prename descclassname"><span class="pre">reconstruction.ecoli.fit_sim_data_1.</span></span><span class="sig-name descname"><span class="pre">fitMaintenanceCosts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bulkContainer</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/fit_sim_data_1.html#fitMaintenanceCosts"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.fit_sim_data_1.fitMaintenanceCosts" title="Link to this definition"></a></dt>
<dd><p>Fits the growth-associated maintenance (GAM) cost associated with metabolism.</p>
<p>The energetic costs associated with growth have been estimated utilizing flux-balance analysis
and are used with FBA to obtain accurate growth predictions.  In the whole-cell model, some of
these costs are explicitly associated with the energetic costs of translation, a biomass
assembly process.  Consequently we must estimate the amount of energy utilized by translation
per unit of biomass (i.e. dry mass) produced, and subtract that quantity from reported GAM to
acquire the modified GAM that we use in the metabolic submodel.</p>
<section id="id24">
<h2>Requires<a class="headerlink" href="#id24" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>amino acid counts associated with protein monomers</p></li>
<li><p>average initial dry mass</p></li>
<li><p>energetic (GTP) cost of translation (per amino acid polymerized)</p></li>
<li><p>observed growth-associated maintenance (GAM)</p></li>
</ul>
<p>In dimensions of ATP or ATP equivalents consumed per biomass</p>
</section>
<section id="id25">
<h2>Modifies<a class="headerlink" href="#id25" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>the “dark” ATP, i.e. the modified GAM</p></li>
</ul>
<p class="rubric">Notes</p>
<p>As more non-metabolic submodels account for energetic costs, this function should be extended
to subtract those costs off the observed GAM.</p>
<p>There also exists, in contrast, non-growth-associated-maintenance (NGAM), which is relative to
total biomass rather than the biomass accumulation rate.  As the name would imply, this
accounts for the energetic costs of maintaining the existing biomass.  It is also accounted for
in the metabolic submodel.</p>
<p>TODO (John): Rewrite as a true function.</p>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="reconstruction.ecoli.fit_sim_data_1.fitPromoterBoundProbability">
<span class="sig-prename descclassname"><span class="pre">reconstruction.ecoli.fit_sim_data_1.</span></span><span class="sig-name descname"><span class="pre">fitPromoterBoundProbability</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_specs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/fit_sim_data_1.html#fitPromoterBoundProbability"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.fit_sim_data_1.fitPromoterBoundProbability" title="Link to this definition"></a></dt>
<dd><p>Calculates the probabilities (P) that each transcription factor will bind
to its target RNA. This function initially calculates these probabilities
from the bulk average counts of the TFs and ligands calculated from
previous steps. Then, values of parameters alpha and r in the equation
below are fit such that the computed RNA synthesis probabilities converge
to the measured RNA synthesis probabilities.</p>
<p>v_{synth, j} = alpha_j + sum_{i} P_{T,i}*r_{ij}</p>
<p>Due to constraints applied in the optimization, both v and P need to
be shifted from their initial values.</p>
<section id="id26">
<h2>Requires<a class="headerlink" href="#id26" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>Bulk average counts of transcription factors and associated ligands</p></li>
</ul>
<p>for each condition (in cell_specs)</p>
</section>
<section id="id27">
<h2>Inputs<a class="headerlink" href="#id27" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>cell_specs {condition (str): dict} - information about each condition</p></li>
</ul>
</section>
<section id="id28">
<h2>Modifies<a class="headerlink" href="#id28" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>Probabilities of TFs binding to their promoters</p></li>
<li><p>RNA synthesis probabilities</p></li>
<li><p>cell_specs[‘basal’][‘r_vector’]: Fit parameters on how the recruitment of</p></li>
</ul>
<p>a TF affects the expression of a gene. High (positive) values of r indicate
that the TF binding increases the probability that the gene is expressed.
- cell_specs[‘basal’][‘r_columns’]: mapping of column name to index in r</p>
<p class="rubric">Notes</p>
<p>See supplementary materials on transcription regulation for details on
the parameters being fit.</p>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="reconstruction.ecoli.fit_sim_data_1.fitSimData_1">
<span class="sig-prename descclassname"><span class="pre">reconstruction.ecoli.fit_sim_data_1.</span></span><span class="sig-name descname"><span class="pre">fitSimData_1</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">raw_data</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/fit_sim_data_1.html#fitSimData_1"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.fit_sim_data_1.fitSimData_1" title="Link to this definition"></a></dt>
<dd><p>Fits parameters necessary for the simulation based on the knowledge base</p>
<dl>
<dt>Inputs:</dt><dd><dl class="simple">
<dt>raw_data (KnowledgeBaseEcoli) - knowledge base consisting of the</dt><dd><p>necessary raw data</p>
</dd>
</dl>
<p>cpus (int) - number of processes to use (if &gt; 1, use multiprocessing)
debug (bool) - if True, fit only one arbitrarily-chosen transcription</p>
<blockquote>
<div><p>factor in order to speed up a debug cycle (should not be used for
an actual simulation)</p>
</div></blockquote>
<dl class="simple">
<dt>save_intermediates (bool) - if True, save the state (sim_data and cell_specs)</dt><dd><p>to disk in intermediates_directory after each Parca step</p>
</dd>
<dt>intermediates_directory (str) - path to the directory to save intermediate</dt><dd><p>sim_data and cell_specs files to</p>
</dd>
<dt>load_intermediate (str) - the function name of the Parca step to load</dt><dd><p>sim_data and cell_specs from; functions prior to and including this
will be skipped but all following functions will run</p>
</dd>
<dt>variable_elongation_transcription (bool) - enable variable elongation</dt><dd><p>for transcription</p>
</dd>
<dt>variable_elongation_translation (bool) - enable variable elongation for</dt><dd><p>translation</p>
</dd>
<dt>disable_ribosome_capacity_fitting (bool) - if True, ribosome expression</dt><dd><p>is not fit to protein synthesis demands</p>
</dd>
<dt>disable_rnapoly_capacity_fitting (bool) - if True, RNA polymerase</dt><dd><p>expression is not fit to protein synthesis demands</p>
</dd>
<dt>cache_dir (str) - path to the directory to save cached data for</dt><dd><p>affinities of RNAs binding to endoRNases</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="reconstruction.ecoli.fit_sim_data_1.fit_condition">
<span class="sig-prename descclassname"><span class="pre">reconstruction.ecoli.fit_sim_data_1.</span></span><span class="sig-name descname"><span class="pre">fit_condition</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_specs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cpus</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/fit_sim_data_1.html#fit_condition"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.fit_sim_data_1.fit_condition" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="reconstruction.ecoli.fit_sim_data_1.initialize">
<span class="sig-prename descclassname"><span class="pre">reconstruction.ecoli.fit_sim_data_1.</span></span><span class="sig-name descname"><span class="pre">initialize</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_specs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">raw_data</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/fit_sim_data_1.html#initialize"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.fit_sim_data_1.initialize" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="reconstruction.ecoli.fit_sim_data_1.input_adjustments">
<span class="sig-prename descclassname"><span class="pre">reconstruction.ecoli.fit_sim_data_1.</span></span><span class="sig-name descname"><span class="pre">input_adjustments</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_specs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">debug</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/fit_sim_data_1.html#input_adjustments"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.fit_sim_data_1.input_adjustments" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="reconstruction.ecoli.fit_sim_data_1.mRNADistributionFromProtein">
<span class="sig-prename descclassname"><span class="pre">reconstruction.ecoli.fit_sim_data_1.</span></span><span class="sig-name descname"><span class="pre">mRNADistributionFromProtein</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">distribution_protein</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">translation_efficiencies</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">netLossRate</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/fit_sim_data_1.html#mRNADistributionFromProtein"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.fit_sim_data_1.mRNADistributionFromProtein" title="Link to this definition"></a></dt>
<dd><p>dP_i / dt = k * M_i * e_i - P_i * Loss_i</p>
<p>At steady state:
M_i = Loss_i * P_i / (k * e_i)</p>
<p>Fraction of protein for ith gene is defined as:
f_i = P_i / P_total</p>
<p>Substituting in:
M_i = Loss_i * f_i * P_total / (k * e_i)</p>
<p>Normalizing M_i by summing over all i cancels out k and P_total
assuming a constant translation rate.</p>
<section id="id29">
<h2>Inputs<a class="headerlink" href="#id29" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>distribution_protein (array of floats) - distribution for each protein,</p></li>
</ul>
<p>normalized to 1
- translation_efficiencies (array of floats) - translational efficiency for each mRNA,
normalized to 1
- netLossRate (array of floats with units of 1/time) - rate of loss for each protein</p>
<dl class="field-list simple">
<dt class="field-odd">rtype<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p>array of floats for the distribution of each mRNA, normalized to 1</p></li>
</ul>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="reconstruction.ecoli.fit_sim_data_1.netLossRateFromDilutionAndDegradationProtein">
<span class="sig-prename descclassname"><span class="pre">reconstruction.ecoli.fit_sim_data_1.</span></span><span class="sig-name descname"><span class="pre">netLossRateFromDilutionAndDegradationProtein</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">doublingTime</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">degradationRates</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/fit_sim_data_1.html#netLossRateFromDilutionAndDegradationProtein"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.fit_sim_data_1.netLossRateFromDilutionAndDegradationProtein" title="Link to this definition"></a></dt>
<dd><p>Compute total loss rate (summed contributions of degradation and dilution).</p>
<section id="id30">
<h2>Inputs<a class="headerlink" href="#id30" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>doublingTime (float with units of time) - doubling time of the cell</p></li>
<li><p>degradationRates (array of floats with units of 1/time) - protein degradation rate</p></li>
</ul>
<dl class="field-list simple">
<dt class="field-odd">rtype<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p>array of floats with units of 1/time for the total loss rate for each protein</p></li>
</ul>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="reconstruction.ecoli.fit_sim_data_1.netLossRateFromDilutionAndDegradationRNA">
<span class="sig-prename descclassname"><span class="pre">reconstruction.ecoli.fit_sim_data_1.</span></span><span class="sig-name descname"><span class="pre">netLossRateFromDilutionAndDegradationRNA</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">doublingTime</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">totalEndoRnaseCountsCapacity</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Km</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rnaConc</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">countsToMolar</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/fit_sim_data_1.html#netLossRateFromDilutionAndDegradationRNA"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.fit_sim_data_1.netLossRateFromDilutionAndDegradationRNA" title="Link to this definition"></a></dt>
<dd><p>Compute total loss rate (summed impact of degradation and dilution).
Returns the loss rate in units of (counts/time) in preparation for use in
the steady state analysis in fitExpression() and
setRNAPCountsConstrainedByPhysiology()
(see calculateMinPolymerizingEnzymeByProductDistributionRNA()).</p>
<p>Derived from steady state analysis of Michaelis-Menten enzyme kinetics with
competitive inhibition: for a given RNA, all other RNAs compete for RNase.</p>
<p>V_i = k_cat * [ES_i]
v_i = k_cat * [E]0 * ([S_i]/Km_i) / (1 + sum over j genes([S_j] / Km_j))</p>
<section id="id31">
<h2>Inputs<a class="headerlink" href="#id31" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>doublingTime (float with units of time) - doubling time of the cell</p></li>
<li><p>totalEndoRnaseCountsCapacity (float with units of 1/time) total kinetic</p></li>
</ul>
<p>capacity of all RNases in the cell
- Km (array of floats with units of mol/volume) - Michaelis-Menten constant
for each RNA
- rnaConc (array of floats with units of mol/volume) - concentration for each RNA
- countsToMolar (float with units of mol/volume) - conversion between counts and molar</p>
<dl class="field-list simple">
<dt class="field-odd">rtype<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p>array of floats with units of 1/time for the total loss rate for each RNA</p></li>
</ul>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="reconstruction.ecoli.fit_sim_data_1.netLossRateFromDilutionAndDegradationRNALinear">
<span class="sig-prename descclassname"><span class="pre">reconstruction.ecoli.fit_sim_data_1.</span></span><span class="sig-name descname"><span class="pre">netLossRateFromDilutionAndDegradationRNALinear</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">doublingTime</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">degradationRates</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rnaCounts</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/fit_sim_data_1.html#netLossRateFromDilutionAndDegradationRNALinear"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.fit_sim_data_1.netLossRateFromDilutionAndDegradationRNALinear" title="Link to this definition"></a></dt>
<dd><p>Compute total loss rate (summed contributions of degradation and dilution).
Returns the loss rate in units of (counts/time) in preparation for use in
the steady state analysis in fitExpression() and
setRNAPCountsConstrainedByPhysiology()
(see calculateMinPolymerizingEnzymeByProductDistributionRNA()).</p>
<section id="id32">
<h2>Requires<a class="headerlink" href="#id32" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>doublingTime (float with units of time) - doubling time of the cell</p></li>
<li><p>degradationRates (array of floats with units of 1/time) - degradation rate</p></li>
</ul>
<p>for each RNA
- rnaCounts (array of floats) - counts for each RNA</p>
<dl class="field-list simple">
<dt class="field-odd">rtype<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p>array of floats with units of 1/time for the total loss rate for each RNA</p></li>
</ul>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="reconstruction.ecoli.fit_sim_data_1.promoter_binding">
<span class="sig-prename descclassname"><span class="pre">reconstruction.ecoli.fit_sim_data_1.</span></span><span class="sig-name descname"><span class="pre">promoter_binding</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_specs</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/fit_sim_data_1.html#promoter_binding"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.fit_sim_data_1.promoter_binding" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="reconstruction.ecoli.fit_sim_data_1.proteinDistributionFrommRNA">
<span class="sig-prename descclassname"><span class="pre">reconstruction.ecoli.fit_sim_data_1.</span></span><span class="sig-name descname"><span class="pre">proteinDistributionFrommRNA</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">distribution_mRNA</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">translation_efficiencies</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">netLossRate</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/fit_sim_data_1.html#proteinDistributionFrommRNA"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.fit_sim_data_1.proteinDistributionFrommRNA" title="Link to this definition"></a></dt>
<dd><p>dP_i / dt = k * M_i * e_i - P_i * Loss_i</p>
<p>At steady state:
P_i = k * M_i * e_i / Loss_i</p>
<p>Fraction of mRNA for ith gene is defined as:
f_i = M_i / M_total</p>
<p>Substituting in:
P_i = k * f_i * e_i * M_total / Loss_i</p>
<p>Normalizing P_i by summing over all i cancels out k and M_total
assuming constant translation rate.</p>
<section id="id33">
<h2>Inputs<a class="headerlink" href="#id33" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>distribution_mRNA (array of floats) - distribution for each mRNA,</p></li>
</ul>
<p>normalized to 1
- translation_efficiencies (array of floats) - translational efficiency for each mRNA,
normalized to 1
- netLossRate (array of floats with units of 1/time) - rate of loss for each protein</p>
<dl class="field-list simple">
<dt class="field-odd">rtype<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p>array of floats for the distribution of each protein, normalized to 1</p></li>
</ul>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="reconstruction.ecoli.fit_sim_data_1.rescaleMassForSolubleMetabolites">
<span class="sig-prename descclassname"><span class="pre">reconstruction.ecoli.fit_sim_data_1.</span></span><span class="sig-name descname"><span class="pre">rescaleMassForSolubleMetabolites</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bulkMolCntr</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">concDict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">doubling_time</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/fit_sim_data_1.html#rescaleMassForSolubleMetabolites"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.fit_sim_data_1.rescaleMassForSolubleMetabolites" title="Link to this definition"></a></dt>
<dd><p>Adjust the cell’s mass to accomodate target small molecule concentrations.</p>
<section id="id34">
<h2>Inputs<a class="headerlink" href="#id34" title="Link to this heading"></a></h2>
<ul class="simple">
<li><dl class="simple">
<dt>bulkMolCntr (np.ndarray object) - Two columns: ‘id’ for name and ‘count’</dt><dd><p>for count of all bulk molecules</p>
</dd>
</dl>
</li>
<li><p>concDict (dict) - a dictionary of metabolite ID (string) : concentration (unit’d number, dimensions of concentration) pairs</p></li>
<li><p>doubling_time (float with units of time) - measured doubling times given the condition</p></li>
</ul>
</section>
<section id="id35">
<h2>Requires<a class="headerlink" href="#id35" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>Cell mass fraction data at a given doubling time.</p></li>
<li><p>Average cell density.</p></li>
<li><p>The conversion factor for transforming from the size of an average cell to the size of a cell
immediately following division.</p></li>
<li><p>Avogadro’s number.</p></li>
<li><p>Concentrations of small molecules (including both dry mass components and water).</p></li>
</ul>
</section>
<section id="id36">
<h2>Modifies<a class="headerlink" href="#id36" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>Adds small molecule counts to bulkMolCntr.</p></li>
</ul>
<dl class="field-list simple">
<dt class="field-odd">returns<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><em>- newAvgCellDryMassInit, the adjusted dry mass of a cell immediately following division.</em></p></li>
<li><p><em>- fitAvgSolubleTargetMolMass, the adjusted dry mass of the soluble fraction of a cell</em></p></li>
</ul>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="reconstruction.ecoli.fit_sim_data_1.save_state">
<span class="sig-prename descclassname"><span class="pre">reconstruction.ecoli.fit_sim_data_1.</span></span><span class="sig-name descname"><span class="pre">save_state</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">func</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/fit_sim_data_1.html#save_state"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.fit_sim_data_1.save_state" title="Link to this definition"></a></dt>
<dd><p>Wrapper for functions called in fitSimData_1() to allow saving and loading
of sim_data and cell_specs at different points in the parameter calculation
pipeline.  This is useful for development in order to skip time intensive
steps that are not required to recalculate in order to work with the desired
stage of parameter calculation.</p>
<dl class="simple">
<dt>This wrapper expects arguments in the kwargs passed into a wrapped function:</dt><dd><dl class="simple">
<dt>save_intermediates (bool): if True, the state (sim_data and cell_specs)</dt><dd><p>will be saved to disk in intermediates_directory</p>
</dd>
<dt>intermediates_directory (str): path to the directory to save intermediate</dt><dd><p>sim_data and cell_specs files to</p>
</dd>
<dt>load_intermediate (str): the name of the function to load sim_data and</dt><dd><p>cell_specs from, functions prior to and including this will be
skipped but all following functions will run</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="reconstruction.ecoli.fit_sim_data_1.setInitialRnaExpression">
<span class="sig-prename descclassname"><span class="pre">reconstruction.ecoli.fit_sim_data_1.</span></span><span class="sig-name descname"><span class="pre">setInitialRnaExpression</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">expression</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">doubling_time</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/fit_sim_data_1.html#setInitialRnaExpression"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.fit_sim_data_1.setInitialRnaExpression" title="Link to this definition"></a></dt>
<dd><p>Creates a container that with the initial count and ID of each RNA,
calculated based on the mass fraction, molecular weight, and expression
distribution of each RNA. For rRNA the counts are set based on mass, while
for tRNA and mRNA the counts are set based on mass and relative abundance.
Relies on the math function totalCountFromMassesAndRatios.</p>
<section id="id37">
<h2>Requires<a class="headerlink" href="#id37" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>Needs information from the knowledge base about the mass fraction,</p></li>
</ul>
<p>molecular weight, and distribution of each RNA species.</p>
</section>
<section id="id38">
<h2>Inputs<a class="headerlink" href="#id38" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>expression (array of floats) - expression for each RNA, normalized to 1</p></li>
<li><p>doubling_time (float with units of time) - doubling time for condition</p></li>
</ul>
<dl class="field-list simple">
<dt class="field-odd">returns<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><em>- expression (array of floats) - contains the adjusted RNA expression,</em></p></li>
<li><p><em>normalized to 1</em></p></li>
</ul>
</dd>
</dl>
<p class="rubric">Notes</p>
<ul class="simple">
<li><p>Now rnaData[“synthProb”] does not match “expression”</p></li>
</ul>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="reconstruction.ecoli.fit_sim_data_1.setKmCooperativeEndoRNonLinearRNAdecay">
<span class="sig-prename descclassname"><span class="pre">reconstruction.ecoli.fit_sim_data_1.</span></span><span class="sig-name descname"><span class="pre">setKmCooperativeEndoRNonLinearRNAdecay</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bulkContainer</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cache_dir</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/fit_sim_data_1.html#setKmCooperativeEndoRNonLinearRNAdecay"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.fit_sim_data_1.setKmCooperativeEndoRNonLinearRNAdecay" title="Link to this definition"></a></dt>
<dd><p>Fits the affinities (Michaelis-Menten constants) for RNAs binding to
endoRNAses.</p>
<p>EndoRNAses perform the first step of RNA decay by cleaving a whole RNA
somewhere inside its extent.  This results in RNA fragments, which are then
digested into monomers by exoRNAses. To model endoRNAse activity, we need to
determine an affinity (Michaelis-Menten constant) for each RNA that is
consistent with experimentally observed half-lives.  The Michaelis-Menten
constants must be determined simultaneously, as the RNAs must compete for
the active site of the endoRNAse. (See the RnaDegradation Process class for
more information about the dynamical model.) The parameters are estimated
using a root solver (scipy.optimize.fsolve).  (See the
sim_data.process.rna_decay.kmLossFunction method for more information about
the optimization problem.)</p>
<section id="id39">
<h2>Requires<a class="headerlink" href="#id39" title="Link to this heading"></a></h2>
<ul class="simple">
<li><dl class="simple">
<dt>cell density, dry mass fraction, and average initial dry mass</dt><dd><p>Used to calculate the cell volume, which in turn is used to calculate
concentrations.</p>
</dd>
</dl>
</li>
<li><p>observed RNA degradation rates (half-lives)</p></li>
<li><p>endoRNAse counts</p></li>
<li><p>endoRNAse catalytic rate constants</p></li>
<li><p>RNA counts</p></li>
<li><p>boolean options that enable sensitivity analyses (see Notes below)</p></li>
</ul>
</section>
<section id="id40">
<h2>Modifies<a class="headerlink" href="#id40" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>Michaelis-Menten constants for first-order decay (initially set to zeros)</p></li>
<li><dl class="simple">
<dt>Several optimization-related values</dt><dd><p>Sensitivity analyses (optional, see Notes below)
Terminal values for optimization-related functions</p>
</dd>
</dl>
</li>
</ul>
<dl class="field-list simple">
<dt class="field-odd">rtype<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p>enoRNAse Km values, in units of M</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>If certain options are set, a sensitivity analysis will be performed using a
range of metaparameters. Outputs will be cached and utilized instead of
running the optimization if possible.
The function that generates the optimization functions is defined under
sim_data but has no dependency on sim_data, and therefore could be moved
here or elsewhere. (TODO)</p>
<p>TODO (John): Refactor as a pure function.
TODO (John): Why is this function called ‘cooperative’?  It seems to instead</p>
<blockquote>
<div><p>assume and model competitive binding.</p>
</div></blockquote>
<dl class="simple">
<dt>TODO (John): Determine what part (if any) of the ‘linear’ parameter fitting</dt><dd><p>should be retained.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="reconstruction.ecoli.fit_sim_data_1.setProteinDegRates">
<span class="sig-prename descclassname"><span class="pre">reconstruction.ecoli.fit_sim_data_1.</span></span><span class="sig-name descname"><span class="pre">setProteinDegRates</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/fit_sim_data_1.html#setProteinDegRates"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.fit_sim_data_1.setProteinDegRates" title="Link to this definition"></a></dt>
<dd><p>This function’s goal is to set the degradation rates for a subset of proteins.
It first gathers the index of the proteins it wants to modify, then changes the degradation
rates of those proteins. These adjustments were made so that the simulation could run.</p>
<section id="id41">
<h2>Requires<a class="headerlink" href="#id41" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>For each protein that needs to be modified it take in an adjustment factor.</p></li>
</ul>
</section>
<section id="id42">
<h2>Modifies<a class="headerlink" href="#id42" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>This function modifies the protein degradation rates for the chosen proteins in sim_data.</p></li>
</ul>
<p>It takes their current degradation rate and multiplies them by the factor specified in adjustments.</p>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="reconstruction.ecoli.fit_sim_data_1.setRNADegRates">
<span class="sig-prename descclassname"><span class="pre">reconstruction.ecoli.fit_sim_data_1.</span></span><span class="sig-name descname"><span class="pre">setRNADegRates</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/fit_sim_data_1.html#setRNADegRates"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.fit_sim_data_1.setRNADegRates" title="Link to this definition"></a></dt>
<dd><p>This function’s goal is to adjust the degradation rates for a subset of
metabolic RNA’s. It first gathers the index of the RNA’s it wants to modify,
then changes the degradation rates of those RNAs. If the specified ID is
that of an RNA cistron, the degradation rates of all RNA molecules
containing the cistron are adjusted. (Note: since RNA concentrations are
assumed to be in equilibrium, increasing the degradation rate increases the
synthesis rates of these RNAs)</p>
<section id="id43">
<h2>Requires<a class="headerlink" href="#id43" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>For each RNA that needs to be modified, it takes in an adjustment factor</p></li>
</ul>
</section>
<section id="id44">
<h2>Modifies<a class="headerlink" href="#id44" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>This function modifies the RNA degradation rates for the chosen RNAs in</p></li>
</ul>
<p>sim_data. It takes their current degradation rate and multiplies them by the
factor specified in adjustments.</p>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="reconstruction.ecoli.fit_sim_data_1.setRNAExpression">
<span class="sig-prename descclassname"><span class="pre">reconstruction.ecoli.fit_sim_data_1.</span></span><span class="sig-name descname"><span class="pre">setRNAExpression</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/fit_sim_data_1.html#setRNAExpression"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.fit_sim_data_1.setRNAExpression" title="Link to this definition"></a></dt>
<dd><p>This function’s goal is to set expression levels for a subset of RNAs.
It first gathers the index of the RNA’s it wants to modify, then changes
the expression levels of those RNAs, within sim_data, based on the
specified adjustment factor. If the specified ID is an RNA cistron, the
expression levels of all RNA molecules containing the cistron are adjusted.</p>
<section id="id45">
<h2>Requires<a class="headerlink" href="#id45" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>For each RNA that needs to be modified, it takes in an adjustment factor.</p></li>
</ul>
</section>
<section id="id46">
<h2>Modifies<a class="headerlink" href="#id46" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>This function modifies the basal RNA expression levels set in sim_data,</p></li>
</ul>
<p>for the chosen RNAs. It takes their current basal expression and multiplies
them by the factor specified in adjustments.
- After updating the basal expression levels for the given RNAs, the
function normalizes all the basal expression levels.</p>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="reconstruction.ecoli.fit_sim_data_1.setRNAPCountsConstrainedByPhysiology">
<span class="sig-prename descclassname"><span class="pre">reconstruction.ecoli.fit_sim_data_1.</span></span><span class="sig-name descname"><span class="pre">setRNAPCountsConstrainedByPhysiology</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bulkContainer</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">doubling_time</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">avgCellDryMassInit</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">variable_elongation_transcription</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Km</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/fit_sim_data_1.html#setRNAPCountsConstrainedByPhysiology"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.fit_sim_data_1.setRNAPCountsConstrainedByPhysiology" title="Link to this definition"></a></dt>
<dd><p>Set counts of RNA polymerase based on two constraints:
(1) Number of RNAP subunits required to maintain steady state of mRNAs
(2) Expected RNAP subunit counts based on (mRNA) distribution recorded in</p>
<blockquote>
<div><p>bulkContainer</p>
</div></blockquote>
<section id="id47">
<h2>Inputs<a class="headerlink" href="#id47" title="Link to this heading"></a></h2>
<ul class="simple">
<li><dl class="simple">
<dt>bulkContainer (np.ndarray object) - Two columns: ‘id’ for name and ‘count’</dt><dd><p>for count of all bulk molecules</p>
</dd>
</dl>
</li>
<li><p>doubling_time (float with units of time) - doubling time given the condition</p></li>
<li><p>avgCellDryMassInit (float with units of mass) - expected initial dry cell mass</p></li>
<li><p>Km (array of floats with units of mol/volume) - Km for each RNA associated</p></li>
</ul>
<p>with RNases</p>
</section>
<section id="id48">
<h2>Modifies<a class="headerlink" href="#id48" title="Link to this heading"></a></h2>
<ul class="simple">
<li><dl class="simple">
<dt>bulkContainer (np.ndarray object) - the counts of RNA polymerase</dt><dd><p>subunits are set according to Constraint 1</p>
</dd>
</dl>
</li>
</ul>
<p class="rubric">Notes</p>
<ul class="simple">
<li><p>Constraint 2 is not being used – see final line of this function.</p></li>
</ul>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="reconstruction.ecoli.fit_sim_data_1.setRibosomeCountsConstrainedByPhysiology">
<span class="sig-prename descclassname"><span class="pre">reconstruction.ecoli.fit_sim_data_1.</span></span><span class="sig-name descname"><span class="pre">setRibosomeCountsConstrainedByPhysiology</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bulkContainer</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">doubling_time</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">variable_elongation_translation</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/fit_sim_data_1.html#setRibosomeCountsConstrainedByPhysiology"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.fit_sim_data_1.setRibosomeCountsConstrainedByPhysiology" title="Link to this definition"></a></dt>
<dd><p>Set counts of ribosomal protein subunits based on three constraints:
(1) Expected protein distribution doubles in one cell cycle
(2) Measured rRNA mass fractions
(3) Expected ribosomal protein subunit counts based on RNA expression data</p>
<section id="id49">
<h2>Inputs<a class="headerlink" href="#id49" title="Link to this heading"></a></h2>
<dl class="simple">
<dt>bulkContainer (np.ndarray object) - Two columns: ‘id’ for name and ‘count’</dt><dd><p>for count of all bulk molecules</p>
</dd>
</dl>
<p>doubling_time (float with units of time) - doubling time given the condition
variable_elongation_translation (bool) - whether there is variable elongation for translation</p>
</section>
<section id="id50">
<h2>Modifies<a class="headerlink" href="#id50" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>counts of ribosomal protein subunits in bulkContainer</p></li>
</ul>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="reconstruction.ecoli.fit_sim_data_1.setTranslationEfficiencies">
<span class="sig-prename descclassname"><span class="pre">reconstruction.ecoli.fit_sim_data_1.</span></span><span class="sig-name descname"><span class="pre">setTranslationEfficiencies</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/fit_sim_data_1.html#setTranslationEfficiencies"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.fit_sim_data_1.setTranslationEfficiencies" title="Link to this definition"></a></dt>
<dd><p>This function’s goal is to set translation efficiencies for a subset of metabolic proteins.
It first gathers the index of the proteins it wants to modify, then changes the monomer
translation efficiencies based on the adjustment that is specified.
These adjustments were made so that the simulation could run.</p>
<section id="id51">
<h2>Requires<a class="headerlink" href="#id51" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>For each protein that needs to be modified, it takes in an adjustment factor.</p></li>
</ul>
</section>
<section id="id52">
<h2>Modifies<a class="headerlink" href="#id52" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>This function modifies, for a subset of proteins, their translational efficiencies in sim_data.</p></li>
</ul>
<p>It takes their current efficiency and multiplies them by the factor specified in adjustments.</p>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="reconstruction.ecoli.fit_sim_data_1.set_balanced_translation_efficiencies">
<span class="sig-prename descclassname"><span class="pre">reconstruction.ecoli.fit_sim_data_1.</span></span><span class="sig-name descname"><span class="pre">set_balanced_translation_efficiencies</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/fit_sim_data_1.html#set_balanced_translation_efficiencies"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.fit_sim_data_1.set_balanced_translation_efficiencies" title="Link to this definition"></a></dt>
<dd><p>Sets the translation efficiencies of a group of proteins to be equal to the
mean value of all proteins within the group.</p>
<section id="id53">
<h2>Requires<a class="headerlink" href="#id53" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>List of proteins that should have balanced translation efficiencies.</p></li>
</ul>
</section>
<section id="id54">
<h2>Modifies<a class="headerlink" href="#id54" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>Translation efficiencies of proteins within each specified group.</p></li>
</ul>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="reconstruction.ecoli.fit_sim_data_1.set_conditions">
<span class="sig-prename descclassname"><span class="pre">reconstruction.ecoli.fit_sim_data_1.</span></span><span class="sig-name descname"><span class="pre">set_conditions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_specs</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/fit_sim_data_1.html#set_conditions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.fit_sim_data_1.set_conditions" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="reconstruction.ecoli.fit_sim_data_1.tf_condition_specs">
<span class="sig-prename descclassname"><span class="pre">reconstruction.ecoli.fit_sim_data_1.</span></span><span class="sig-name descname"><span class="pre">tf_condition_specs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_specs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cpus</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">disable_ribosome_capacity_fitting</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">disable_rnapoly_capacity_fitting</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">variable_elongation_transcription</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">variable_elongation_translation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/fit_sim_data_1.html#tf_condition_specs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.fit_sim_data_1.tf_condition_specs" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="reconstruction.ecoli.fit_sim_data_1.totalCountFromMassesAndRatios">
<span class="sig-prename descclassname"><span class="pre">reconstruction.ecoli.fit_sim_data_1.</span></span><span class="sig-name descname"><span class="pre">totalCountFromMassesAndRatios</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">totalMass</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">individualMasses</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">distribution</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/fit_sim_data_1.html#totalCountFromMassesAndRatios"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.fit_sim_data_1.totalCountFromMassesAndRatios" title="Link to this definition"></a></dt>
<dd><p>Function to determine the expected total counts for a group of molecules
in order to achieve a total mass with a given distribution of individual
molecules.</p>
<dl>
<dt>Math:</dt><dd><p>Total mass = dot(mass, count)</p>
<p>Fraction of i:
f = count / Total counts</p>
<p>Substituting:
Total mass = dot(mass, f * Total counts)
Total mass = Total counts * dot(mass, f)</p>
<p>Total counts = Total mass / dot(mass, f)</p>
</dd>
</dl>
<section id="id55">
<h2>Requires<a class="headerlink" href="#id55" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>totalMass (float with mass units): total mass of the group of molecules</p></li>
<li><p>individualMasses (array of floats with mass units): mass for individual</p></li>
</ul>
<p>molecules in the group
- distribution (array of floats): distribution of individual molecules,
normalized to 1</p>
<dl class="field-list simple">
<dt class="field-odd">returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>- counts (float)</strong></p>
</dd>
<dt class="field-even">rtype<span class="colon">:</span></dt>
<dd class="field-even"><p>total counts (does not need to be a whole number)</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="reconstruction.ecoli.fit_sim_data_1.totalCountIdDistributionProtein">
<span class="sig-prename descclassname"><span class="pre">reconstruction.ecoli.fit_sim_data_1.</span></span><span class="sig-name descname"><span class="pre">totalCountIdDistributionProtein</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">expression</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">doubling_time</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/fit_sim_data_1.html#totalCountIdDistributionProtein"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.fit_sim_data_1.totalCountIdDistributionProtein" title="Link to this definition"></a></dt>
<dd><p>Calculates the total counts of proteins from the relative expression of RNA,
individual protein mass, and total protein mass. Relies on the math functions
netLossRateFromDilutionAndDegradationProtein, proteinDistributionFrommRNA,
totalCountFromMassesAndRatios.</p>
<section id="id56">
<h2>Inputs<a class="headerlink" href="#id56" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>expression (array of floats) - relative frequency distribution of RNA expression</p></li>
<li><p>doubling_time (float with units of time) - measured doubling time given the condition</p></li>
</ul>
<dl class="field-list simple">
<dt class="field-odd">returns<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><em>- total_count_protein (float) - total number of proteins</em></p></li>
<li><p><em>- ids_protein (array of str) - name of each protein with location tag</em></p></li>
<li><p><em>- distribution_protein (array of floats) - distribution for each protein,</em></p></li>
<li><p><em>normalized to 1</em></p></li>
</ul>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="reconstruction.ecoli.fit_sim_data_1.totalCountIdDistributionRNA">
<span class="sig-prename descclassname"><span class="pre">reconstruction.ecoli.fit_sim_data_1.</span></span><span class="sig-name descname"><span class="pre">totalCountIdDistributionRNA</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">expression</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">doubling_time</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/fit_sim_data_1.html#totalCountIdDistributionRNA"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.fit_sim_data_1.totalCountIdDistributionRNA" title="Link to this definition"></a></dt>
<dd><p>Calculates the total counts of RNA from their relative expression,
individual mass, and total RNA mass. Relies on the math function
totalCountFromMassesAndRatios.</p>
<section id="id57">
<h2>Inputs<a class="headerlink" href="#id57" title="Link to this heading"></a></h2>
<ul class="simple">
<li><dl class="simple">
<dt>expression (array of floats) - relative frequency distribution of RNA</dt><dd><p>expression</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>doubling_time (float with units of time) - measured doubling time given</dt><dd><p>the condition</p>
</dd>
</dl>
</li>
</ul>
<dl class="field-list simple">
<dt class="field-odd">returns<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><em>- total_count_RNA (float) - total number of RNAs</em></p></li>
<li><p><em>- ids_rnas (array of str) - name of each RNA with location tag</em></p></li>
<li><p><em>- distribution_RNA (array of floats) - distribution for each RNA,</em> – normalized to 1</p></li>
</ul>
</dd>
</dl>
</section>
</dd></dl>

</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="reconstruction.ecoli.scripts.update_biocyc_files.html" class="btn btn-neutral float-left" title="reconstruction.ecoli.scripts.update_biocyc_files" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="reconstruction.ecoli.knowledge_base_raw.html" class="btn btn-neutral float-right" title="reconstruction.ecoli.knowledge_base_raw" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2021-2025, The Vivarium E. coli Authors.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>