

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>reconstruction.ecoli.dataclasses.getter_functions &mdash; Vivarium E. coli 1.0.0 documentation</title>
      <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=b86133f3" />
      <link rel="stylesheet" type="text/css" href="../../../_static/css/theme.css?v=e59714d7" />

  
      <script src="../../../_static/jquery.js?v=5d32c60e"></script>
      <script src="../../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../../../_static/documentation_options.js?v=8d563738"></script>
      <script src="../../../_static/doctools.js?v=9bcbadda"></script>
      <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
      <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="reconstruction.ecoli.dataclasses.growth_rate_dependent_parameters" href="reconstruction.ecoli.dataclasses.growth_rate_dependent_parameters.html" />
    <link rel="prev" title="reconstruction.ecoli.dataclasses.constants" href="reconstruction.ecoli.dataclasses.constants.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../index.html" class="icon icon-home">
            Vivarium E. coli
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../../stores.html">Stores</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../processes.html">Processes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../composites.html">Composites</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../experiments.html">Experiments</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../workflows.html">Workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../output.html">Output</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../tutorial.html">Tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../docs.html">Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../hpc.html">HPC Clusters</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../gcloud.html">Google Cloud</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../ci.html">Continuous Integration</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../pycharm.html">PyCharm Setup</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../diffs.html">Model Differences From wcEcoli</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../../api_ref.html">API Reference</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../ecoli/ecoli.html"><code class="docutils literal notranslate"><span class="pre">ecoli</span></code></a></li>
<li class="toctree-l2 current"><a class="reference internal" href="reconstruction.html"><code class="docutils literal notranslate"><span class="pre">reconstruction</span></code></a><ul class="current">
<li class="toctree-l3 current"><a class="reference internal" href="reconstruction.ecoli.html"><code class="docutils literal notranslate"><span class="pre">reconstruction.ecoli</span></code></a><ul class="current">
<li class="toctree-l4"><a class="reference internal" href="reconstruction.ecoli.compendium.html"><code class="docutils literal notranslate"><span class="pre">reconstruction.ecoli.compendium</span></code></a></li>
<li class="toctree-l4 current"><a class="reference internal" href="reconstruction.ecoli.dataclasses.html"><code class="docutils literal notranslate"><span class="pre">reconstruction.ecoli.dataclasses</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="reconstruction.ecoli.scripts.html"><code class="docutils literal notranslate"><span class="pre">reconstruction.ecoli.scripts</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="reconstruction.ecoli.fit_sim_data_1.html"><code class="docutils literal notranslate"><span class="pre">reconstruction.ecoli.fit_sim_data_1</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="reconstruction.ecoli.knowledge_base_raw.html"><code class="docutils literal notranslate"><span class="pre">reconstruction.ecoli.knowledge_base_raw</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="reconstruction.ecoli.simulation_data.html"><code class="docutils literal notranslate"><span class="pre">reconstruction.ecoli.simulation_data</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="reconstruction.tests.html"><code class="docutils literal notranslate"><span class="pre">reconstruction.tests</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="reconstruction.spreadsheets.html"><code class="docutils literal notranslate"><span class="pre">reconstruction.spreadsheets</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../runscripts/runscripts.html"><code class="docutils literal notranslate"><span class="pre">runscripts</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="../validation/validation.html"><code class="docutils literal notranslate"><span class="pre">validation</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="../wholecell/wholecell.html"><code class="docutils literal notranslate"><span class="pre">wholecell</span></code></a></li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">Vivarium E. coli</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../api_ref.html">API Reference</a></li>
          <li class="breadcrumb-item"><a href="reconstruction.html"><code class="docutils literal notranslate"><span class="pre">reconstruction</span></code></a></li>
          <li class="breadcrumb-item"><a href="reconstruction.ecoli.html"><code class="docutils literal notranslate"><span class="pre">reconstruction.ecoli</span></code></a></li>
          <li class="breadcrumb-item"><a href="reconstruction.ecoli.dataclasses.html"><code class="docutils literal notranslate"><span class="pre">reconstruction.ecoli.dataclasses</span></code></a></li>
      <li class="breadcrumb-item active"><code class="docutils literal notranslate"><span class="pre">reconstruction.ecoli.dataclasses.getter_functions</span></code></li>
      <li class="wy-breadcrumbs-aside">
            <a href="../../../_sources/reference/api/reconstruction/reconstruction.ecoli.dataclasses.getter_functions.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="module-reconstruction.ecoli.dataclasses.getter_functions">
<span id="reconstruction-ecoli-dataclasses-getter-functions"></span><h1><code class="docutils literal notranslate"><span class="pre">reconstruction.ecoli.dataclasses.getter_functions</span></code><a class="headerlink" href="#module-reconstruction.ecoli.dataclasses.getter_functions" title="Link to this heading"></a></h1>
<p>SimulationData getter functions</p>
<dl class="py class">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.getter_functions.GetterFunctions">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">reconstruction.ecoli.dataclasses.getter_functions.</span></span><span class="sig-name descname"><span class="pre">GetterFunctions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">raw_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/getter_functions.html#GetterFunctions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.getter_functions.GetterFunctions" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>getterFunctions</p>
<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.getter_functions.GetterFunctions._build_all_masses">
<span class="sig-name descname"><span class="pre">_build_all_masses</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">raw_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/getter_functions.html#GetterFunctions._build_all_masses"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.getter_functions.GetterFunctions._build_all_masses" title="Link to this definition"></a></dt>
<dd><p>Builds dictionary of molecular weights keyed with the IDs of molecules.
Depending on the type of the molecule, weights are either pulled
directly from raw data, or dynamically calculated from the existing
data.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.getter_functions.GetterFunctions._build_compartments">
<span class="sig-name descname"><span class="pre">_build_compartments</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">raw_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/getter_functions.html#GetterFunctions._build_compartments"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.getter_functions.GetterFunctions._build_compartments" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.getter_functions.GetterFunctions._build_full_chromosome_mass">
<span class="sig-name descname"><span class="pre">_build_full_chromosome_mass</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">raw_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/getter_functions.html#GetterFunctions._build_full_chromosome_mass"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.getter_functions.GetterFunctions._build_full_chromosome_mass" title="Link to this definition"></a></dt>
<dd><p>Calculates the mass of the full chromosome from its sequence and the
weigths of polymerized dNTPs.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.getter_functions.GetterFunctions._build_genomic_coordinates">
<span class="sig-name descname"><span class="pre">_build_genomic_coordinates</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">raw_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/getter_functions.html#GetterFunctions._build_genomic_coordinates"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.getter_functions.GetterFunctions._build_genomic_coordinates" title="Link to this definition"></a></dt>
<dd><p>Builds a dictionary of genomic coordinates of DNA sites. Keys are the
IDs of the sites, and the values are tuples of the left-end coordinate
and the right-end coordinate of the site. Sites whose types are included
in IGNORED_DNA_SITE_TYPES are ignored.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.getter_functions.GetterFunctions._build_modified_protein_masses">
<span class="sig-name descname"><span class="pre">_build_modified_protein_masses</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">raw_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/getter_functions.html#GetterFunctions._build_modified_protein_masses"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.getter_functions.GetterFunctions._build_modified_protein_masses" title="Link to this definition"></a></dt>
<dd><p>Builds dictionary of molecular weights of modified proteins keyed with
the molecule IDs. Molecular weights are calculated from the
stoichiometries of the modification reactions.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.getter_functions.GetterFunctions._build_modified_rna_masses">
<span class="sig-name descname"><span class="pre">_build_modified_rna_masses</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">raw_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/getter_functions.html#GetterFunctions._build_modified_rna_masses"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.getter_functions.GetterFunctions._build_modified_rna_masses" title="Link to this definition"></a></dt>
<dd><p>Builds dictionary of molecular weights of modified RNAs keyed with
the molecule IDs. Molecular weights are calculated from the
stoichiometries of the modification reactions.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.getter_functions.GetterFunctions._build_polymerized_subunit_masses">
<span class="sig-name descname"><span class="pre">_build_polymerized_subunit_masses</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/getter_functions.html#GetterFunctions._build_polymerized_subunit_masses"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.getter_functions.GetterFunctions._build_polymerized_subunit_masses" title="Link to this definition"></a></dt>
<dd><p>Builds dictionary of molecular weights of polymerized subunits (NTPs,
dNTPs, and amino acids) by subtracting the end weights from the weights
of original subunits.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.getter_functions.GetterFunctions._build_protein_complex_compartments">
<span class="sig-name descname"><span class="pre">_build_protein_complex_compartments</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">raw_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">all_compartments_sorted</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/getter_functions.html#GetterFunctions._build_protein_complex_compartments"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.getter_functions.GetterFunctions._build_protein_complex_compartments" title="Link to this definition"></a></dt>
<dd><p>Builds dictionary of compartment tags for protein complexes keyed with
the molecule IDs. Each complex is assigned to the compartment that
contains a subunit of the complex and has the highest complexation
priority. Compartment tags of subunits that are also protein complexes
are determined recursively. Compartments of metabolite subunits are not
considered since metabolites are assigned to all compartments that are
being modeled.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.getter_functions.GetterFunctions._build_protein_complex_masses">
<span class="sig-name descname"><span class="pre">_build_protein_complex_masses</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">raw_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/getter_functions.html#GetterFunctions._build_protein_complex_masses"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.getter_functions.GetterFunctions._build_protein_complex_masses" title="Link to this definition"></a></dt>
<dd><p>Builds dictionary of molecular weights of protein complexes keyed with
the molecule IDs. Molecular weights are calculated from the
stoichiometries of the complexation/equilibrium reactions. For
complexes whose subunits are also complexes, the molecular weights are
calculated recursively.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.getter_functions.GetterFunctions._build_protein_masses">
<span class="sig-name descname"><span class="pre">_build_protein_masses</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">raw_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/getter_functions.html#GetterFunctions._build_protein_masses"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.getter_functions.GetterFunctions._build_protein_masses" title="Link to this definition"></a></dt>
<dd><p>Builds dictionary of molecular weights of protein monomers keyed with
the protein IDs. Molecular weights are calculated from the protein
sequence and the weights of polymerized amino acids.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.getter_functions.GetterFunctions._build_protein_sequences">
<span class="sig-name descname"><span class="pre">_build_protein_sequences</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">raw_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/getter_functions.html#GetterFunctions._build_protein_sequences"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.getter_functions.GetterFunctions._build_protein_sequences" title="Link to this definition"></a></dt>
<dd><p>Builds the amino acid sequences of each protein monomer using sequences
in raw_data.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.getter_functions.GetterFunctions._build_rna_masses">
<span class="sig-name descname"><span class="pre">_build_rna_masses</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">raw_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/getter_functions.html#GetterFunctions._build_rna_masses"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.getter_functions.GetterFunctions._build_rna_masses" title="Link to this definition"></a></dt>
<dd><p>Builds dictionary of molecular weights of RNAs keyed with the RNA IDs.
Molecular weights are calculated from the RNA sequence and the weights
of polymerized NTPs.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.getter_functions.GetterFunctions._build_rna_sequences">
<span class="sig-name descname"><span class="pre">_build_rna_sequences</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">raw_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/getter_functions.html#GetterFunctions._build_rna_sequences"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.getter_functions.GetterFunctions._build_rna_sequences" title="Link to this definition"></a></dt>
<dd><p>Builds nucleotide sequences of each transcription unit using the genome
sequence and the left and right end positions.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.getter_functions.GetterFunctions._build_sequences">
<span class="sig-name descname"><span class="pre">_build_sequences</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">raw_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/getter_functions.html#GetterFunctions._build_sequences"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.getter_functions.GetterFunctions._build_sequences" title="Link to this definition"></a></dt>
<dd><p>Builds sequences of RNAs and proteins.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.getter_functions.GetterFunctions._build_submass_array">
<span class="sig-name descname"><span class="pre">_build_submass_array</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mw</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">submass_name</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/getter_functions.html#GetterFunctions._build_submass_array"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.getter_functions.GetterFunctions._build_submass_array" title="Link to this definition"></a></dt>
<dd><p>Converts a scalar molecular weight value to an array of submasses,
given the name of the submass category that the molecule belongs to.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>mw</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)"><em>float</em></a>)</p></li>
<li><p><strong>submass_name</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>ndarray</em></a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.14)">tuple</a>[<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)">int</a>, …], <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.dtype.html#numpy.dtype" title="(in NumPy v2.4)"><em>dtype</em></a>[<em>float64</em>]]</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.getter_functions.GetterFunctions._compartment_tag">
<span class="sig-name descname"><span class="pre">_compartment_tag</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">re.compile('\\[[a-z]]')</span></em><a class="headerlink" href="#reconstruction.ecoli.dataclasses.getter_functions.GetterFunctions._compartment_tag" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.getter_functions.GetterFunctions.get_all_valid_molecules">
<span class="sig-name descname"><span class="pre">get_all_valid_molecules</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/getter_functions.html#GetterFunctions.get_all_valid_molecules"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.getter_functions.GetterFunctions.get_all_valid_molecules" title="Link to this definition"></a></dt>
<dd><p>Returns a list of all molecule IDs with assigned submass arrays and
compartments.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)">list</a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)">str</a>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.getter_functions.GetterFunctions.get_compartment">
<span class="sig-name descname"><span class="pre">get_compartment</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mol_id</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/getter_functions.html#GetterFunctions.get_compartment"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.getter_functions.GetterFunctions.get_compartment" title="Link to this definition"></a></dt>
<dd><p>Returns the list of one-letter codes for the compartments that the
molecule with the given ID can exist in.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>mol_id</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)">list</a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)">list</a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)">str</a>]]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.getter_functions.GetterFunctions.get_compartment_tag">
<span class="sig-name descname"><span class="pre">get_compartment_tag</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">id_</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/getter_functions.html#GetterFunctions.get_compartment_tag"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.getter_functions.GetterFunctions.get_compartment_tag" title="Link to this definition"></a></dt>
<dd><p>Look up a molecule id and return a compartment suffix tag like ‘[c]’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>id_</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)">str</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.getter_functions.GetterFunctions.get_compartments">
<span class="sig-name descname"><span class="pre">get_compartments</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ids</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/getter_functions.html#GetterFunctions.get_compartments"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.getter_functions.GetterFunctions.get_compartments" title="Link to this definition"></a></dt>
<dd><p>Returns a list of the list of one-letter codes for the compartments
that each of the molecules with the given IDs can exist in.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>ids</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a><em>] </em><em>| </em><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>ndarray</em></a>)</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)">list</a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)">list</a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)">str</a>]]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.getter_functions.GetterFunctions.get_genomic_coordinates">
<span class="sig-name descname"><span class="pre">get_genomic_coordinates</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">site_id</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/getter_functions.html#GetterFunctions.get_genomic_coordinates"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.getter_functions.GetterFunctions.get_genomic_coordinates" title="Link to this definition"></a></dt>
<dd><p>Returns the genomic coordinates of the left and right ends of a DNA site
given the ID of the site.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>site_id</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.14)">tuple</a>[<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)">int</a>, <a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)">int</a>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.getter_functions.GetterFunctions.get_mass">
<span class="sig-name descname"><span class="pre">get_mass</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mol_id</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/getter_functions.html#GetterFunctions.get_mass"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.getter_functions.GetterFunctions.get_mass" title="Link to this definition"></a></dt>
<dd><p>Return the total mass of the molecule with a given ID.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>mol_id</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.getter_functions.GetterFunctions.get_masses">
<span class="sig-name descname"><span class="pre">get_masses</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mol_ids</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/getter_functions.html#GetterFunctions.get_masses"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.getter_functions.GetterFunctions.get_masses" title="Link to this definition"></a></dt>
<dd><p>Return an array of total masses of the molecules with the given IDs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>mol_ids</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a><em> | </em><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>ndarray</em></a>)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.getter_functions.GetterFunctions.get_miscrnas_with_singleton_tus">
<span class="sig-name descname"><span class="pre">get_miscrnas_with_singleton_tus</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/getter_functions.html#GetterFunctions.get_miscrnas_with_singleton_tus"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.getter_functions.GetterFunctions.get_miscrnas_with_singleton_tus" title="Link to this definition"></a></dt>
<dd><p>Returns a list of all miscRNA IDs with corresponding single-gene
transcription units.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)">list</a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)">str</a>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.getter_functions.GetterFunctions.get_sequences">
<span class="sig-name descname"><span class="pre">get_sequences</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ids</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/getter_functions.html#GetterFunctions.get_sequences"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.getter_functions.GetterFunctions.get_sequences" title="Link to this definition"></a></dt>
<dd><p>Return a list of sequences of the molecules with the given IDs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>ids</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a><em>] </em><em>| </em><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>ndarray</em></a>)</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)">list</a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)">str</a>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.getter_functions.GetterFunctions.get_singleton_tu_id">
<span class="sig-name descname"><span class="pre">get_singleton_tu_id</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rna_id</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/getter_functions.html#GetterFunctions.get_singleton_tu_id"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.getter_functions.GetterFunctions.get_singleton_tu_id" title="Link to this definition"></a></dt>
<dd><p>Returns the ID of the single-gene transcription unit corresponding to
the given miscRNA ID, if such a transcription unit exists. This is
necessary to replace some references to miscRNA IDs in complexation or
equilibrium reactions with their corresponding TU IDs, which are the
molecules that are actually transcribed.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>rna_id</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)">str</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.getter_functions.GetterFunctions.get_submass_array">
<span class="sig-name descname"><span class="pre">get_submass_array</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mol_id</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/getter_functions.html#GetterFunctions.get_submass_array"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.getter_functions.GetterFunctions.get_submass_array" title="Link to this definition"></a></dt>
<dd><p>Return the submass array of the molecule with a given ID.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>mol_id</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.getter_functions.GetterFunctions.is_valid_molecule">
<span class="sig-name descname"><span class="pre">is_valid_molecule</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mol_id</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/getter_functions.html#GetterFunctions.is_valid_molecule"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.getter_functions.GetterFunctions.is_valid_molecule" title="Link to this definition"></a></dt>
<dd><p>Returns True if the molecule with the given ID is a valid molecule (has
both a submass array and a compartment tag).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>mol_id</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)">bool</a></p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py exception">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.getter_functions.InvalidProteinComplexError">
<em class="property"><span class="pre">exception</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">reconstruction.ecoli.dataclasses.getter_functions.</span></span><span class="sig-name descname"><span class="pre">InvalidProteinComplexError</span></span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/getter_functions.html#InvalidProteinComplexError"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.getter_functions.InvalidProteinComplexError" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3/library/exceptions.html#Exception" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Exception</span></code></a></p>
</dd></dl>

<dl class="py exception">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.getter_functions.TranscriptionDirectionError">
<em class="property"><span class="pre">exception</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">reconstruction.ecoli.dataclasses.getter_functions.</span></span><span class="sig-name descname"><span class="pre">TranscriptionDirectionError</span></span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/getter_functions.html#TranscriptionDirectionError"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.getter_functions.TranscriptionDirectionError" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3/library/exceptions.html#Exception" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Exception</span></code></a></p>
</dd></dl>

<dl class="py exception">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.getter_functions.UnknownModifiedRnaError">
<em class="property"><span class="pre">exception</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">reconstruction.ecoli.dataclasses.getter_functions.</span></span><span class="sig-name descname"><span class="pre">UnknownModifiedRnaError</span></span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/getter_functions.html#UnknownModifiedRnaError"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.getter_functions.UnknownModifiedRnaError" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3/library/exceptions.html#Exception" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Exception</span></code></a></p>
</dd></dl>

<dl class="py exception">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.getter_functions.UnknownMolecularWeightError">
<em class="property"><span class="pre">exception</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">reconstruction.ecoli.dataclasses.getter_functions.</span></span><span class="sig-name descname"><span class="pre">UnknownMolecularWeightError</span></span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/getter_functions.html#UnknownMolecularWeightError"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.getter_functions.UnknownMolecularWeightError" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3/library/exceptions.html#Exception" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Exception</span></code></a></p>
</dd></dl>

</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="reconstruction.ecoli.dataclasses.constants.html" class="btn btn-neutral float-left" title="reconstruction.ecoli.dataclasses.constants" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="reconstruction.ecoli.dataclasses.growth_rate_dependent_parameters.html" class="btn btn-neutral float-right" title="reconstruction.ecoli.dataclasses.growth_rate_dependent_parameters" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2021-2026, The Vivarium E. coli Authors.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>