

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>reconstruction.ecoli.dataclasses.process.transcription &mdash; Vivarium E. coli 1.0.0 documentation</title>
      <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=b86133f3" />
      <link rel="stylesheet" type="text/css" href="../../../_static/css/theme.css?v=e59714d7" />

  
      <script src="../../../_static/jquery.js?v=5d32c60e"></script>
      <script src="../../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../../../_static/documentation_options.js?v=8d563738"></script>
      <script src="../../../_static/doctools.js?v=9bcbadda"></script>
      <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
      <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="reconstruction.ecoli.dataclasses.process.transcription_regulation" href="reconstruction.ecoli.dataclasses.process.transcription_regulation.html" />
    <link rel="prev" title="reconstruction.ecoli.dataclasses.process.rna_decay" href="reconstruction.ecoli.dataclasses.process.rna_decay.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../index.html" class="icon icon-home">
            Vivarium E. coli
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../../stores.html">Stores</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../processes.html">Processes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../composites.html">Composites</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../experiments.html">Experiments</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../workflows.html">Workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../output.html">Output</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../tutorial.html">Tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../docs.html">Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../hpc.html">HPC Clusters</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../gcloud.html">Google Cloud</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../ci.html">Continuous Integration</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../pycharm.html">PyCharm Setup</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../diffs.html">Model Differences From wcEcoli</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../../api_ref.html">API Reference</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../ecoli/ecoli.html"><code class="docutils literal notranslate"><span class="pre">ecoli</span></code></a></li>
<li class="toctree-l2 current"><a class="reference internal" href="reconstruction.html"><code class="docutils literal notranslate"><span class="pre">reconstruction</span></code></a><ul class="current">
<li class="toctree-l3 current"><a class="reference internal" href="reconstruction.ecoli.html"><code class="docutils literal notranslate"><span class="pre">reconstruction.ecoli</span></code></a><ul class="current">
<li class="toctree-l4"><a class="reference internal" href="reconstruction.ecoli.compendium.html"><code class="docutils literal notranslate"><span class="pre">reconstruction.ecoli.compendium</span></code></a></li>
<li class="toctree-l4 current"><a class="reference internal" href="reconstruction.ecoli.dataclasses.html"><code class="docutils literal notranslate"><span class="pre">reconstruction.ecoli.dataclasses</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="reconstruction.ecoli.scripts.html"><code class="docutils literal notranslate"><span class="pre">reconstruction.ecoli.scripts</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="reconstruction.ecoli.fit_sim_data_1.html"><code class="docutils literal notranslate"><span class="pre">reconstruction.ecoli.fit_sim_data_1</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="reconstruction.ecoli.knowledge_base_raw.html"><code class="docutils literal notranslate"><span class="pre">reconstruction.ecoli.knowledge_base_raw</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="reconstruction.ecoli.simulation_data.html"><code class="docutils literal notranslate"><span class="pre">reconstruction.ecoli.simulation_data</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="reconstruction.tests.html"><code class="docutils literal notranslate"><span class="pre">reconstruction.tests</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="reconstruction.spreadsheets.html"><code class="docutils literal notranslate"><span class="pre">reconstruction.spreadsheets</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../runscripts/runscripts.html"><code class="docutils literal notranslate"><span class="pre">runscripts</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="../validation/validation.html"><code class="docutils literal notranslate"><span class="pre">validation</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="../wholecell/wholecell.html"><code class="docutils literal notranslate"><span class="pre">wholecell</span></code></a></li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">Vivarium E. coli</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../api_ref.html">API Reference</a></li>
          <li class="breadcrumb-item"><a href="reconstruction.html"><code class="docutils literal notranslate"><span class="pre">reconstruction</span></code></a></li>
          <li class="breadcrumb-item"><a href="reconstruction.ecoli.html"><code class="docutils literal notranslate"><span class="pre">reconstruction.ecoli</span></code></a></li>
          <li class="breadcrumb-item"><a href="reconstruction.ecoli.dataclasses.html"><code class="docutils literal notranslate"><span class="pre">reconstruction.ecoli.dataclasses</span></code></a></li>
          <li class="breadcrumb-item"><a href="reconstruction.ecoli.dataclasses.process.html"><code class="docutils literal notranslate"><span class="pre">reconstruction.ecoli.dataclasses.process</span></code></a></li>
      <li class="breadcrumb-item active"><code class="docutils literal notranslate"><span class="pre">reconstruction.ecoli.dataclasses.process.transcription</span></code></li>
      <li class="wy-breadcrumbs-aside">
            <a href="../../../_sources/reference/api/reconstruction/reconstruction.ecoli.dataclasses.process.transcription.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="module-reconstruction.ecoli.dataclasses.process.transcription">
<span id="reconstruction-ecoli-dataclasses-process-transcription"></span><h1><code class="docutils literal notranslate"><span class="pre">reconstruction.ecoli.dataclasses.process.transcription</span></code><a class="headerlink" href="#module-reconstruction.ecoli.dataclasses.process.transcription" title="Link to this heading"></a></h1>
<p>SimulationData for transcription process</p>
<p>TODO: add mapping of tRNA to charged tRNA if allowing more than one modified form of tRNA and separate mappings for tRNA and charged tRNA to AA
TODO: handle ppGpp and DksA-ppGpp regulation separately</p>
<dl class="py class">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.transcription.Transcription">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">reconstruction.ecoli.dataclasses.process.transcription.</span></span><span class="sig-name descname"><span class="pre">Transcription</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">raw_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/process/transcription.html#Transcription"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.transcription.Transcription" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>SimulationData for the transcription process</p>
<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.transcription.Transcription._apply_rnaseq_correction">
<span class="sig-name descname"><span class="pre">_apply_rnaseq_correction</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/process/transcription.html#Transcription._apply_rnaseq_correction"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.transcription.Transcription._apply_rnaseq_correction" title="Link to this definition"></a></dt>
<dd><p>Applies correction to RNAseq data for shorter genes as required when
operon structure is included in the model.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.transcription.Transcription._build_attenuation">
<span class="sig-name descname"><span class="pre">_build_attenuation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">raw_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/process/transcription.html#Transcription._build_attenuation"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.transcription.Transcription._build_attenuation" title="Link to this definition"></a></dt>
<dd><p>Load fold changes related to transcriptional attenuation.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.transcription.Transcription._build_charged_trna">
<span class="sig-name descname"><span class="pre">_build_charged_trna</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">raw_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/process/transcription.html#Transcription._build_charged_trna"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.transcription.Transcription._build_charged_trna" title="Link to this definition"></a></dt>
<dd><p>Loads information and creates data structures necessary for charging of tRNA</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Requires self.rna_data so can’t be built in translation even if some
data structures would be more appropriate there.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.transcription.Transcription._build_cistron_data">
<span class="sig-name descname"><span class="pre">_build_cistron_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">raw_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/process/transcription.html#Transcription._build_cistron_data"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.transcription.Transcription._build_cistron_data" title="Link to this definition"></a></dt>
<dd><p>Build cistron-associated simulation data from raw data. Cistrons are
sections of RNAs that encode for a specific polypeptide. A single RNA
molecule may contain one or more cistrons.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.transcription.Transcription._build_elongation_rates">
<span class="sig-name descname"><span class="pre">_build_elongation_rates</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">raw_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/process/transcription.html#Transcription._build_elongation_rates"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.transcription.Transcription._build_elongation_rates" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.transcription.Transcription._build_mature_rna_data">
<span class="sig-name descname"><span class="pre">_build_mature_rna_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">raw_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/process/transcription.html#Transcription._build_mature_rna_data"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.transcription.Transcription._build_mature_rna_data" title="Link to this definition"></a></dt>
<dd><p>Build mature RNA-associated simulation data from raw data.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.transcription.Transcription._build_new_gene_data">
<span class="sig-name descname"><span class="pre">_build_new_gene_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">raw_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/process/transcription.html#Transcription._build_new_gene_data"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.transcription.Transcription._build_new_gene_data" title="Link to this definition"></a></dt>
<dd><p>Load baseline values for new gene expression in all simulations.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.transcription.Transcription._build_oric_terc_coordinates">
<span class="sig-name descname"><span class="pre">_build_oric_terc_coordinates</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">raw_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/process/transcription.html#Transcription._build_oric_terc_coordinates"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.transcription.Transcription._build_oric_terc_coordinates" title="Link to this definition"></a></dt>
<dd><p>Builds coordinates of oriC and terC that are used when calculating
genomic positions of cistrons and RNAs relative to the origin</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.transcription.Transcription._build_ppgpp_regulation">
<span class="sig-name descname"><span class="pre">_build_ppgpp_regulation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">raw_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/process/transcription.html#Transcription._build_ppgpp_regulation"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.transcription.Transcription._build_ppgpp_regulation" title="Link to this definition"></a></dt>
<dd><p>Determine which genes are regulated by ppGpp and store the fold
change in expression associated with each gene.</p>
<dl>
<dt>Attributes set:</dt><dd><p>ppgpp_regulated_genes (ndarray[str]): cistron ID of regulated genes
ppgpp_fold_changes (ndarray[float]): log2 fold change for each gene</p>
<blockquote>
<div><p>in ppgpp_regulated_genes</p>
</div></blockquote>
<dl class="simple">
<dt>_ppgpp_growth_parameters: parameters for interpolate.splev</dt><dd><p>to estimate growth rate from ppGpp concentration</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.transcription.Transcription._build_rna_data">
<span class="sig-name descname"><span class="pre">_build_rna_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">raw_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/process/transcription.html#Transcription._build_rna_data"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.transcription.Transcription._build_rna_data" title="Link to this definition"></a></dt>
<dd><p>Build RNA-associated simulation data from raw data.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.transcription.Transcription._build_transcription">
<span class="sig-name descname"><span class="pre">_build_transcription</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">raw_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/process/transcription.html#Transcription._build_transcription"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.transcription.Transcription._build_transcription" title="Link to this definition"></a></dt>
<dd><p>Build transcription-associated simulation data from raw data.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.transcription.Transcription._get_relative_coordinates">
<span class="sig-name descname"><span class="pre">_get_relative_coordinates</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">coordinates</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/process/transcription.html#Transcription._get_relative_coordinates"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.transcription.Transcription._get_relative_coordinates" title="Link to this definition"></a></dt>
<dd><p>Returns the genomic coordinates of a given gene coordinate relative
to the origin of replication.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.transcription.Transcription._normalize_ppgpp_expression">
<span class="sig-name descname"><span class="pre">_normalize_ppgpp_expression</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/process/transcription.html#Transcription._normalize_ppgpp_expression"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.transcription.Transcription._normalize_ppgpp_expression" title="Link to this definition"></a></dt>
<dd><p>Normalize both free and ppGpp bound expression values to 1.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.transcription.Transcription._solve_ppgpp_km">
<span class="sig-name descname"><span class="pre">_solve_ppgpp_km</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">raw_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/process/transcription.html#Transcription._solve_ppgpp_km"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.transcription.Transcription._solve_ppgpp_km" title="Link to this definition"></a></dt>
<dd><p>Solves for general expression rates for bound and free RNAP and
a KM for ppGpp to RNAP based on global cellular measurements.
Parameters are solved for at different doubling times using a
gradient descent method to minimize the difference in expression
of stable RNA compared to the measured RNA in a cell.
Assumes a Hill coefficient of 2 for ppGpp binding to RNAP.</p>
<dl class="simple">
<dt>Attributes set:</dt><dd><dl class="simple">
<dt>_fit_ppgpp_fc (float): log2 fold change in stable RNA expression</dt><dd><p>from a fast doubling time to a slow doubling time based on
the rates of bound and free RNAP expression found</p>
</dd>
<dt>_ppgpp_km_squared (float): squared and unitless KM value for</dt><dd><p>to limit computation needed for fraction bound</p>
</dd>
<dt>ppgpp_km (float with mol / volume units): KM for ppGpp binding</dt><dd><p>to RNAP</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.transcription.Transcription.adjust_polymerizing_ppgpp_expression">
<span class="sig-name descname"><span class="pre">adjust_polymerizing_ppgpp_expression</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/process/transcription.html#Transcription.adjust_polymerizing_ppgpp_expression"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.transcription.Transcription.adjust_polymerizing_ppgpp_expression" title="Link to this definition"></a></dt>
<dd><p>Adjust ppGpp expression based on fit for ribosome and RNAP physiological
constraints using least squares fit for 3 conditions with different
growth rates/ppGpp.</p>
<dl class="simple">
<dt>Modifies attributes:</dt><dd><dl class="simple">
<dt>exp_ppgpp (ndarray[float]): expression for each gene when RNAP</dt><dd><p>is bound to ppGpp, adjusted for necessary RNAP and ribosome
expression, normalized to 1</p>
</dd>
<dt>exp_free (ndarray[float]): expression for each gene when RNAP</dt><dd><p>is not bound to ppGpp, adjusted for necessary RNAP and ribosome
expression, normalized to 1</p>
</dd>
</dl>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>See docs/processes/transcription_regulation.pdf for a description
of the math used in this section.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.transcription.Transcription.adjust_ppgpp_expression_for_tfs">
<span class="sig-name descname"><span class="pre">adjust_ppgpp_expression_for_tfs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/process/transcription.html#Transcription.adjust_ppgpp_expression_for_tfs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.transcription.Transcription.adjust_ppgpp_expression_for_tfs" title="Link to this definition"></a></dt>
<dd><p>Adjusts ppGpp regulated expression to get expression with and without
ppGpp regulation to match in basal condition and taking into account
the effect transcription factors will have.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.transcription.Transcription.calculate_attenuation">
<span class="sig-name descname"><span class="pre">calculate_attenuation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_specs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/process/transcription.html#Transcription.calculate_attenuation"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.transcription.Transcription.calculate_attenuation" title="Link to this definition"></a></dt>
<dd><p>Calculate constants for each attenuated gene.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.transcription.Transcription.charging_stoich_matrix">
<span class="sig-name descname"><span class="pre">charging_stoich_matrix</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/process/transcription.html#Transcription.charging_stoich_matrix"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.transcription.Transcription.charging_stoich_matrix" title="Link to this definition"></a></dt>
<dd><p>Creates stoich matrix from i, j, v arrays</p>
<p>Returns 2D array with rows of metabolites for each tRNA charging reaction on the column</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.transcription.Transcription.cistron_id_to_rna_indexes">
<span class="sig-name descname"><span class="pre">cistron_id_to_rna_indexes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cistron_id</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/process/transcription.html#Transcription.cistron_id_to_rna_indexes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.transcription.Transcription.cistron_id_to_rna_indexes" title="Link to this definition"></a></dt>
<dd><p>Returns the indexes of transcription units containing the given RNA
cistron given the ID of the cistron.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.transcription.Transcription.expression_from_ppgpp">
<span class="sig-name descname"><span class="pre">expression_from_ppgpp</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ppgpp</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/process/transcription.html#Transcription.expression_from_ppgpp"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.transcription.Transcription.expression_from_ppgpp" title="Link to this definition"></a></dt>
<dd><p>Calculates the expression of each gene at a given concentration of ppGpp.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>ppgpp</strong> (<em>float with</em><em> or </em><em>without mol / volume units</em>) – concentration of ppGpp,
if unitless, should represent the concentration of PPGPP_CONC_UNITS</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>normalized expression for each gene</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>ndarray[<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)">float</a>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.transcription.Transcription.fit_rna_expression">
<span class="sig-name descname"><span class="pre">fit_rna_expression</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cistron_expression</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/process/transcription.html#Transcription.fit_rna_expression"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.transcription.Transcription.fit_rna_expression" title="Link to this definition"></a></dt>
<dd><p>Calculates the expression of RNA transcription units that best fits the
given expression levels of cistrons using nonnegative least squares.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.transcription.Transcription.fit_trna_expression">
<span class="sig-name descname"><span class="pre">fit_trna_expression</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">tRNA_cistron_expression</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/process/transcription.html#Transcription.fit_trna_expression"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.transcription.Transcription.fit_trna_expression" title="Link to this definition"></a></dt>
<dd><p>Calculates the expression of tRNA transcription units that best fits the
given expression levels of tRNA cistrons using nonnegative least
squares.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.transcription.Transcription.fraction_rnap_bound_ppgpp">
<span class="sig-name descname"><span class="pre">fraction_rnap_bound_ppgpp</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ppgpp</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/process/transcription.html#Transcription.fraction_rnap_bound_ppgpp"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.transcription.Transcription.fraction_rnap_bound_ppgpp" title="Link to this definition"></a></dt>
<dd><p>Calculates the fraction of RNAP expected to be bound to ppGpp
at a given concentration of ppGpp.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>ppgpp</strong> (<em>float with</em><em> or </em><em>without mol / volume units</em>) – concentration of ppGpp,
if unitless, should represent the concentration of PPGPP_CONC_UNITS</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>fraction of RNAP that will be bound to ppGpp</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)">float</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.transcription.Transcription.get_attenuation_stop_probabilities">
<span class="sig-name descname"><span class="pre">get_attenuation_stop_probabilities</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">trna_conc</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/process/transcription.html#Transcription.get_attenuation_stop_probabilities"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.transcription.Transcription.get_attenuation_stop_probabilities" title="Link to this definition"></a></dt>
<dd><p>Calculate the probability of a transcript stopping early due to attenuation.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.transcription.Transcription.get_rna_fractions">
<span class="sig-name descname"><span class="pre">get_rna_fractions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ppgpp</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/process/transcription.html#Transcription.get_rna_fractions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.transcription.Transcription.get_rna_fractions" title="Link to this definition"></a></dt>
<dd><p>Calculates expected RNA subgroup mass fractions based on ppGpp
concentration. If ppGpp expression has not been set yet, uses default
measured fractions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>ppgpp</strong> (<em>float with</em><em> or </em><em>without mol / volume units</em>) – concentration of
ppGpp, if unitless, should represent the concentration of
PPGPP_CONC_UNITS</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>mass fraction for each subgroup mass, values sum
to 1</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.14)">dict</a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)">str</a>, <a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)">float</a>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.transcription.Transcription.get_rnap_active_fraction_from_ppGpp">
<span class="sig-name descname"><span class="pre">get_rnap_active_fraction_from_ppGpp</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ppgpp</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/process/transcription.html#Transcription.get_rnap_active_fraction_from_ppGpp"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.transcription.Transcription.get_rnap_active_fraction_from_ppGpp" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.transcription.Transcription.make_elongation_rates">
<span class="sig-name descname"><span class="pre">make_elongation_rates</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">random</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">base</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">time_step</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">variable_elongation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/process/transcription.html#Transcription.make_elongation_rates"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.transcription.Transcription.make_elongation_rates" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.transcription.Transcription.rna_id_to_cistron_indexes">
<span class="sig-name descname"><span class="pre">rna_id_to_cistron_indexes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rna_id</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/process/transcription.html#Transcription.rna_id_to_cistron_indexes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.transcription.Transcription.rna_id_to_cistron_indexes" title="Link to this definition"></a></dt>
<dd><p>Returns the indexes of cistrons that constitute the given transcription
unit given the ID of the RNA transcription unit.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.transcription.Transcription.set_ppgpp_expression">
<span class="sig-name descname"><span class="pre">set_ppgpp_expression</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/process/transcription.html#Transcription.set_ppgpp_expression"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.transcription.Transcription.set_ppgpp_expression" title="Link to this definition"></a></dt>
<dd><p>Called during the parca to determine expression of each transcription
unit for ppGpp bound and free RNAP.</p>
<dl class="simple">
<dt>Attributes set:</dt><dd><dl class="simple">
<dt>exp_ppgpp (ndarray[float]): expression for each TU when RNAP is</dt><dd><p>bound to ppGpp</p>
</dd>
<dt>exp_free (ndarray[float]): expression for each TU when RNAP is not</dt><dd><p>bound to ppGpp</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.transcription.Transcription.set_ppgpp_kinetics_parameters">
<span class="sig-name descname"><span class="pre">set_ppgpp_kinetics_parameters</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">init_container</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">constants</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/process/transcription.html#Transcription.set_ppgpp_kinetics_parameters"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.transcription.Transcription.set_ppgpp_kinetics_parameters" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.transcription.Transcription.synth_prob_from_ppgpp">
<span class="sig-name descname"><span class="pre">synth_prob_from_ppgpp</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ppgpp</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">copy_number</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">balanced_rRNA_prob</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/process/transcription.html#Transcription.synth_prob_from_ppgpp"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.transcription.Transcription.synth_prob_from_ppgpp" title="Link to this definition"></a></dt>
<dd><p>Calculates the synthesis probability of each gene at a given concentration
of ppGpp.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ppgpp</strong> (<em>float with mol / volume units</em>) – concentration of ppGpp</p></li>
<li><p><strong>copy_number</strong> (<em>Callable</em><em>[</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)"><em>float</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a><em>]</em>) – function that gives the expected copy
number given a doubling time and gene replication coordinate</p></li>
<li><p><strong>balanced_rRNA_prob</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>) – if True, set synthesis probabilities
of rRNA promoters equal to one another</p></li>
</ul>
</dd>
</dl>
<dl class="simple">
<dt>Returns</dt><dd><p>prob (ndarray[float]): normalized synthesis probability for each gene
factor (ndarray[float]): factor to adjust expression to probability for each gene</p>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>copy_number should be sim_data.process.replication.get_average_copy_number
but saving the function handle as a class attribute prevents pickling of sim_data
without additional handling</p>
</div>
</dd></dl>

</dd></dl>

<dl class="py exception">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.transcription.TranscriptionDirectionError">
<em class="property"><span class="pre">exception</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">reconstruction.ecoli.dataclasses.process.transcription.</span></span><span class="sig-name descname"><span class="pre">TranscriptionDirectionError</span></span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/process/transcription.html#TranscriptionDirectionError"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.transcription.TranscriptionDirectionError" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3/library/exceptions.html#Exception" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Exception</span></code></a></p>
</dd></dl>

</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="reconstruction.ecoli.dataclasses.process.rna_decay.html" class="btn btn-neutral float-left" title="reconstruction.ecoli.dataclasses.process.rna_decay" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="reconstruction.ecoli.dataclasses.process.transcription_regulation.html" class="btn btn-neutral float-right" title="reconstruction.ecoli.dataclasses.process.transcription_regulation" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2021-2026, The Vivarium E. coli Authors.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>