

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>reconstruction.ecoli.dataclasses.process.metabolism &mdash; Vivarium E. coli 1.0.0 documentation</title>
      <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=b86133f3" />
      <link rel="stylesheet" type="text/css" href="../../../_static/css/theme.css?v=e59714d7" />

  
      <script src="../../../_static/jquery.js?v=5d32c60e"></script>
      <script src="../../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../../../_static/documentation_options.js?v=8d563738"></script>
      <script src="../../../_static/doctools.js?v=9bcbadda"></script>
      <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
      <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="reconstruction.ecoli.dataclasses.process.process" href="reconstruction.ecoli.dataclasses.process.process.html" />
    <link rel="prev" title="reconstruction.ecoli.dataclasses.process.equilibrium" href="reconstruction.ecoli.dataclasses.process.equilibrium.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../index.html" class="icon icon-home">
            Vivarium E. coli
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../../stores.html">Stores</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../processes.html">Processes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../composites.html">Composites</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../experiments.html">Experiments</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../workflows.html">Workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../output.html">Output</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../tutorial.html">Tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../docs.html">Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../hpc.html">HPC Clusters</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../gcloud.html">Google Cloud</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../ci.html">Continuous Integration</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../pycharm.html">PyCharm Setup</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../../api_ref.html">API Reference</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../ecoli/ecoli.html"><code class="docutils literal notranslate"><span class="pre">ecoli</span></code></a></li>
<li class="toctree-l2 current"><a class="reference internal" href="reconstruction.html"><code class="docutils literal notranslate"><span class="pre">reconstruction</span></code></a><ul class="current">
<li class="toctree-l3 current"><a class="reference internal" href="reconstruction.ecoli.html"><code class="docutils literal notranslate"><span class="pre">reconstruction.ecoli</span></code></a><ul class="current">
<li class="toctree-l4"><a class="reference internal" href="reconstruction.ecoli.compendium.html"><code class="docutils literal notranslate"><span class="pre">reconstruction.ecoli.compendium</span></code></a></li>
<li class="toctree-l4 current"><a class="reference internal" href="reconstruction.ecoli.dataclasses.html"><code class="docutils literal notranslate"><span class="pre">reconstruction.ecoli.dataclasses</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="reconstruction.ecoli.scripts.html"><code class="docutils literal notranslate"><span class="pre">reconstruction.ecoli.scripts</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="reconstruction.ecoli.fit_sim_data_1.html"><code class="docutils literal notranslate"><span class="pre">reconstruction.ecoli.fit_sim_data_1</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="reconstruction.ecoli.knowledge_base_raw.html"><code class="docutils literal notranslate"><span class="pre">reconstruction.ecoli.knowledge_base_raw</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="reconstruction.ecoli.simulation_data.html"><code class="docutils literal notranslate"><span class="pre">reconstruction.ecoli.simulation_data</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="reconstruction.tests.html"><code class="docutils literal notranslate"><span class="pre">reconstruction.tests</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="reconstruction.spreadsheets.html"><code class="docutils literal notranslate"><span class="pre">reconstruction.spreadsheets</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../runscripts/runscripts.html"><code class="docutils literal notranslate"><span class="pre">runscripts</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="../validation/validation.html"><code class="docutils literal notranslate"><span class="pre">validation</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="../wholecell/wholecell.html"><code class="docutils literal notranslate"><span class="pre">wholecell</span></code></a></li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">Vivarium E. coli</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../api_ref.html">API Reference</a></li>
          <li class="breadcrumb-item"><a href="reconstruction.html"><code class="docutils literal notranslate"><span class="pre">reconstruction</span></code></a></li>
          <li class="breadcrumb-item"><a href="reconstruction.ecoli.html"><code class="docutils literal notranslate"><span class="pre">reconstruction.ecoli</span></code></a></li>
          <li class="breadcrumb-item"><a href="reconstruction.ecoli.dataclasses.html"><code class="docutils literal notranslate"><span class="pre">reconstruction.ecoli.dataclasses</span></code></a></li>
          <li class="breadcrumb-item"><a href="reconstruction.ecoli.dataclasses.process.html"><code class="docutils literal notranslate"><span class="pre">reconstruction.ecoli.dataclasses.process</span></code></a></li>
      <li class="breadcrumb-item active"><code class="docutils literal notranslate"><span class="pre">reconstruction.ecoli.dataclasses.process.metabolism</span></code></li>
      <li class="wy-breadcrumbs-aside">
            <a href="../../../_sources/reference/api/reconstruction/reconstruction.ecoli.dataclasses.process.metabolism.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="module-reconstruction.ecoli.dataclasses.process.metabolism">
<span id="reconstruction-ecoli-dataclasses-process-metabolism"></span><h1><code class="docutils literal notranslate"><span class="pre">reconstruction.ecoli.dataclasses.process.metabolism</span></code><a class="headerlink" href="#module-reconstruction.ecoli.dataclasses.process.metabolism" title="Link to this heading">ÔÉÅ</a></h1>
<p>SimulationData for metabolism process</p>
<p>TODO:</p>
<ul class="simple">
<li><p>improved estimate of ILE/LEU abundance or some external data point</p></li>
<li><p>implement L1-norm minimization for AA concentrations</p></li>
<li><p>find concentration for PI[c]</p></li>
<li><p>add (d)NTP byproduct concentrations</p></li>
</ul>
<dl class="py class">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.ConcentrationUpdates">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">reconstruction.ecoli.dataclasses.process.metabolism.</span></span><span class="sig-name descname"><span class="pre">ConcentrationUpdates</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">concDict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">relative_changes</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">equilibriumReactions</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exchange_data_dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">all_metabolite_ids</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">linked_metabolites</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/process/metabolism.html#ConcentrationUpdates"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.ConcentrationUpdates" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.13)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.ConcentrationUpdates._add_molecule_amounts">
<span class="sig-name descname"><span class="pre">_add_molecule_amounts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">equilibriumReactions</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">concDict</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/process/metabolism.html#ConcentrationUpdates._add_molecule_amounts"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.ConcentrationUpdates._add_molecule_amounts" title="Link to this definition">ÔÉÅ</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.ConcentrationUpdates._exchange_flux_present">
<span class="sig-name descname"><span class="pre">_exchange_flux_present</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">exchange_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/process/metabolism.html#ConcentrationUpdates._exchange_flux_present"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.ConcentrationUpdates._exchange_flux_present" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>Caches the presence of exchanges in each media condition based on
exchange_data to set concentrations in concentrations_based_on_nutrients().</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>exchange_data</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)"><em>dict</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Any" title="(in Python v3.13)"><em>Any</em></a><em>]</em>) ‚Äì <p>dictionary of exchange data for all media conditions
with keys:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">{</span><span class="n">importUnconstrainedExchangeMolecules</span> <span class="p">(</span><span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">set</span><span class="p">[</span><span class="nb">str</span><span class="p">]]):</span>
        <span class="n">exchange</span> <span class="n">molecules</span> <span class="p">(</span><span class="k">with</span> <span class="n">location</span> <span class="n">tag</span><span class="p">)</span> <span class="k">for</span> <span class="n">each</span> <span class="n">media</span> <span class="n">key</span>
        <span class="n">that</span> <span class="n">do</span> <span class="ow">not</span> <span class="n">have</span> <span class="n">an</span> <span class="n">upper</span> <span class="n">bound</span> <span class="n">on</span> <span class="n">their</span> <span class="n">flux</span><span class="p">,</span>
<span class="n">importConstrainedExchangeMolecules</span> <span class="p">(</span><span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span>
        <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">float</span> <span class="k">with</span> <span class="n">mol</span><span class="o">/</span><span class="n">mass</span><span class="o">/</span><span class="n">time</span> <span class="n">units</span><span class="p">]]):</span> <span class="n">constrained</span>
        <span class="n">molecules</span> <span class="p">(</span><span class="k">with</span> <span class="n">location</span> <span class="n">tag</span><span class="p">)</span> <span class="k">for</span> <span class="n">each</span> <span class="n">media</span> <span class="n">key</span> <span class="k">with</span> <span class="n">upper</span>
        <span class="n">bound</span> <span class="n">flux</span> <span class="n">constraints</span><span class="p">}</span>
</pre></div>
</div>
</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Sets of molecules IDs (with location tags) that can be imported
for each media ID</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)">dict</a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)">str</a>, <a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#set" title="(in Python v3.13)">set</a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)">str</a>]]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.ConcentrationUpdates.concentrations_based_on_nutrients">
<span class="sig-name descname"><span class="pre">concentrations_based_on_nutrients</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">media_id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">imports</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">conversion_units</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/process/metabolism.html#ConcentrationUpdates.concentrations_based_on_nutrients"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.ConcentrationUpdates.concentrations_based_on_nutrients" title="Link to this definition">ÔÉÅ</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>media_id</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em> | </em><em>None</em>)</p></li>
<li><p><strong>imports</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em> | </em><em>None</em>)</p></li>
<li><p><strong>conversion_units</strong> (<em>Unum</em><em> | </em><em>None</em>)</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)">dict</a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)">str</a>, <a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Any" title="(in Python v3.13)"><em>Any</em></a>]</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py data">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.DRY_MASS_UNITS">
<span class="sig-prename descclassname"><span class="pre">reconstruction.ecoli.dataclasses.process.metabolism.</span></span><span class="sig-name descname"><span class="pre">DRY_MASS_UNITS</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">1</span> <span class="pre">[fg]</span></em><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.DRY_MASS_UNITS" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>Units for dry mass</p>
</dd></dl>

<dl class="py exception">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.InvalidReactionDirectionError">
<em class="property"><span class="pre">exception</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">reconstruction.ecoli.dataclasses.process.metabolism.</span></span><span class="sig-name descname"><span class="pre">InvalidReactionDirectionError</span></span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/process/metabolism.html#InvalidReactionDirectionError"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.InvalidReactionDirectionError" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3/library/exceptions.html#Exception" title="(in Python v3.13)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Exception</span></code></a></p>
</dd></dl>

<dl class="py data">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.K_CAT_UNITS">
<span class="sig-prename descclassname"><span class="pre">reconstruction.ecoli.dataclasses.process.metabolism.</span></span><span class="sig-name descname"><span class="pre">K_CAT_UNITS</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">1.0</span> <span class="pre">[1/s]</span></em><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.K_CAT_UNITS" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>Units for k <sub>cat</sub> values</p>
</dd></dl>

<dl class="py data">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.METABOLITE_CONCENTRATION_UNITS">
<span class="sig-prename descclassname"><span class="pre">reconstruction.ecoli.dataclasses.process.metabolism.</span></span><span class="sig-name descname"><span class="pre">METABOLITE_CONCENTRATION_UNITS</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">1.0</span> <span class="pre">[mol/L]</span></em><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.METABOLITE_CONCENTRATION_UNITS" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>Units for metabolite concentrations</p>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">reconstruction.ecoli.dataclasses.process.metabolism.</span></span><span class="sig-name descname"><span class="pre">Metabolism</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">raw_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/process/metabolism.html#Metabolism"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.13)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>raw_data</strong> (<a class="reference internal" href="reconstruction.ecoli.knowledge_base_raw.html#reconstruction.ecoli.knowledge_base_raw.KnowledgeBaseEcoli" title="reconstruction.ecoli.knowledge_base_raw.KnowledgeBaseEcoli"><em>KnowledgeBaseEcoli</em></a>) ‚Äì Raw data object</p></li>
<li><p><strong>sim_data</strong> (<a class="reference internal" href="reconstruction.ecoli.simulation_data.html#reconstruction.ecoli.simulation_data.SimulationDataEcoli" title="reconstruction.ecoli.simulation_data.SimulationDataEcoli"><em>SimulationDataEcoli</em></a>) ‚Äì Simulation data object</p></li>
</ul>
</dd>
</dl>
<dl class="py attribute">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.solver">
<span class="sig-name descname"><span class="pre">solver</span></span><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.solver" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>solver ID, should match a value in modular_fba.py,
set by <a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._set_solver_values" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._set_solver_values"><code class="xref py py-meth docutils literal notranslate"><span class="pre">_set_solver_values()</span></code></a></p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)">str</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.kinetic_objective_weight">
<span class="sig-name descname"><span class="pre">kinetic_objective_weight</span></span><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.kinetic_objective_weight" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>weighting for the kinetic objective,
1-weighting for the homeostatic objective, set by
<a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._set_solver_values" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._set_solver_values"><code class="xref py py-meth docutils literal notranslate"><span class="pre">_set_solver_values()</span></code></a></p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.13)">float</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.kinetic_objective_weight_in_range">
<span class="sig-name descname"><span class="pre">kinetic_objective_weight_in_range</span></span><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.kinetic_objective_weight_in_range" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>weighting for deviations
from the kinetic target within min and max ranges, set by
<a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._set_solver_values" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._set_solver_values"><code class="xref py py-meth docutils literal notranslate"><span class="pre">_set_solver_values()</span></code></a></p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.13)">float</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.secretion_penalty_coeff">
<span class="sig-name descname"><span class="pre">secretion_penalty_coeff</span></span><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.secretion_penalty_coeff" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>penalty on secretion fluxes, set by
<a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._set_solver_values" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._set_solver_values"><code class="xref py py-meth docutils literal notranslate"><span class="pre">_set_solver_values()</span></code></a></p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.13)">float</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.metabolite_charge">
<span class="sig-name descname"><span class="pre">metabolite_charge</span></span><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.metabolite_charge" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>mapping of metabolite IDs to charge,
set by <a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._add_metabolite_charge" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._add_metabolite_charge"><code class="xref py py-meth docutils literal notranslate"><span class="pre">_add_metabolite_charge()</span></code></a></p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)">dict</a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)">str</a>, <a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)">int</a>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.concentration_updates">
<span class="sig-name descname"><span class="pre">concentration_updates</span></span><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.concentration_updates" title="Link to this definition">ÔÉÅ</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.conc_dict">
<span class="sig-name descname"><span class="pre">conc_dict</span></span><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.conc_dict" title="Link to this definition">ÔÉÅ</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)">dict</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.nutrients_to_internal_conc">
<span class="sig-name descname"><span class="pre">nutrients_to_internal_conc</span></span><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.nutrients_to_internal_conc" title="Link to this definition">ÔÉÅ</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)">dict</a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)">str</a>, <a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)">dict</a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)">str</a>, Unum]]</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.kinetic_constraint_reactions">
<span class="sig-name descname"><span class="pre">kinetic_constraint_reactions</span></span><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.kinetic_constraint_reactions" title="Link to this definition">ÔÉÅ</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.kinetic_constraint_enzymes">
<span class="sig-name descname"><span class="pre">kinetic_constraint_enzymes</span></span><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.kinetic_constraint_enzymes" title="Link to this definition">ÔÉÅ</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.kinetic_constraint_substrates">
<span class="sig-name descname"><span class="pre">kinetic_constraint_substrates</span></span><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.kinetic_constraint_substrates" title="Link to this definition">ÔÉÅ</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._kcats">
<span class="sig-name descname"><span class="pre">_kcats</span></span><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._kcats" title="Link to this definition">ÔÉÅ</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._saturations">
<span class="sig-name descname"><span class="pre">_saturations</span></span><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._saturations" title="Link to this definition">ÔÉÅ</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._enzymes">
<span class="sig-name descname"><span class="pre">_enzymes</span></span><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._enzymes" title="Link to this definition">ÔÉÅ</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.constraint_is_kcat_only">
<span class="sig-name descname"><span class="pre">constraint_is_kcat_only</span></span><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.constraint_is_kcat_only" title="Link to this definition">ÔÉÅ</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._compiled_enzymes">
<span class="sig-name descname"><span class="pre">_compiled_enzymes</span></span><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._compiled_enzymes" title="Link to this definition">ÔÉÅ</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._compiled_saturation">
<span class="sig-name descname"><span class="pre">_compiled_saturation</span></span><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._compiled_saturation" title="Link to this definition">ÔÉÅ</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.reaction_stoich">
<span class="sig-name descname"><span class="pre">reaction_stoich</span></span><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.reaction_stoich" title="Link to this definition">ÔÉÅ</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.maintenance_reaction">
<span class="sig-name descname"><span class="pre">maintenance_reaction</span></span><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.maintenance_reaction" title="Link to this definition">ÔÉÅ</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.reaction_catalysts">
<span class="sig-name descname"><span class="pre">reaction_catalysts</span></span><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.reaction_catalysts" title="Link to this definition">ÔÉÅ</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.catalyst_ids">
<span class="sig-name descname"><span class="pre">catalyst_ids</span></span><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.catalyst_ids" title="Link to this definition">ÔÉÅ</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.reactions_with_catalyst">
<span class="sig-name descname"><span class="pre">reactions_with_catalyst</span></span><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.reactions_with_catalyst" title="Link to this definition">ÔÉÅ</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.catalysis_matrix_I">
<span class="sig-name descname"><span class="pre">catalysis_matrix_I</span></span><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.catalysis_matrix_I" title="Link to this definition">ÔÉÅ</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.catalysis_matrix_J">
<span class="sig-name descname"><span class="pre">catalysis_matrix_J</span></span><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.catalysis_matrix_J" title="Link to this definition">ÔÉÅ</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.catalysis_matrix_V">
<span class="sig-name descname"><span class="pre">catalysis_matrix_V</span></span><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.catalysis_matrix_V" title="Link to this definition">ÔÉÅ</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.use_all_constraints">
<span class="sig-name descname"><span class="pre">use_all_constraints</span></span><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.use_all_constraints" title="Link to this definition">ÔÉÅ</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.constraints_to_disable">
<span class="sig-name descname"><span class="pre">constraints_to_disable</span></span><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.constraints_to_disable" title="Link to this definition">ÔÉÅ</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.base_reaction_ids">
<span class="sig-name descname"><span class="pre">base_reaction_ids</span></span><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.base_reaction_ids" title="Link to this definition">ÔÉÅ</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.reaction_id_to_base_reaction_id">
<span class="sig-name descname"><span class="pre">reaction_id_to_base_reaction_id</span></span><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.reaction_id_to_base_reaction_id" title="Link to this definition">ÔÉÅ</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.amino_acid_export_kms">
<span class="sig-name descname"><span class="pre">amino_acid_export_kms</span></span><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.amino_acid_export_kms" title="Link to this definition">ÔÉÅ</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.transport_reactions">
<span class="sig-name descname"><span class="pre">transport_reactions</span></span><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.transport_reactions" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>transport reaction IDs in the
metabolic network (includes reverse reactions and reactions
with kinetic constraints), set by
<a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._build_transport_reactions" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._build_transport_reactions"><code class="xref py py-meth docutils literal notranslate"><span class="pre">_build_transport_reactions()</span></code></a></p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.13)">list</a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)">str</a>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.aa_synthesis_pathways">
<span class="sig-name descname"><span class="pre">aa_synthesis_pathways</span></span><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.aa_synthesis_pathways" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>data for allosteric
inhibition of amino acid pathways indexed by amino acid ID with
location tag and nested dictionary with the following keys:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">{</span><span class="s1">&#39;enzymes&#39;</span> <span class="p">(</span><span class="nb">str</span><span class="p">):</span> <span class="n">limiting</span><span class="o">/</span><span class="n">regulated</span> <span class="n">enzyme</span> <span class="n">ID</span> <span class="ow">in</span> <span class="n">synthesis</span>
        <span class="n">pathway</span> <span class="k">with</span> <span class="n">location</span> <span class="n">tag</span><span class="p">,</span>
<span class="s1">&#39;kcat_data&#39;</span> <span class="p">(</span><span class="n">units</span><span class="o">.</span><span class="n">Unum</span><span class="p">):</span> <span class="n">kcat</span> <span class="n">associated</span> <span class="k">with</span> <span class="n">enzyme</span>
        <span class="n">reaction</span> <span class="k">with</span> <span class="n">units</span> <span class="n">of</span> <span class="mi">1</span><span class="o">/</span><span class="n">time</span><span class="p">,</span>
<span class="s1">&#39;ki&#39;</span> <span class="p">(</span><span class="n">Tuple</span><span class="p">[</span><span class="n">units</span><span class="o">.</span><span class="n">Unum</span><span class="p">,</span> <span class="n">units</span><span class="o">.</span><span class="n">Unum</span><span class="p">]]):</span> <span class="n">lower</span> <span class="ow">and</span> <span class="n">upper</span>
        <span class="n">limits</span> <span class="n">of</span> <span class="n">KI</span> <span class="n">associated</span> <span class="k">with</span> <span class="n">enzyme</span> <span class="n">reaction</span> <span class="k">with</span> <span class="n">units</span>
        <span class="n">of</span> <span class="n">mol</span><span class="o">/</span><span class="n">volume</span><span class="p">}</span>
</pre></div>
</div>
<p>Set by <a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._build_amino_acid_pathways" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._build_amino_acid_pathways"><code class="xref py py-meth docutils literal notranslate"><span class="pre">_build_amino_acid_pathways()</span></code></a></p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)">dict</a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)">str</a>, <a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)">dict</a>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.KI_aa_synthesis">
<span class="sig-name descname"><span class="pre">KI_aa_synthesis</span></span><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.KI_aa_synthesis" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>KI for each AA for synthesis
portion of supply (in units of
<a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.METABOLITE_CONCENTRATION_UNITS" title="reconstruction.ecoli.dataclasses.process.metabolism.METABOLITE_CONCENTRATION_UNITS"><code class="xref py py-data docutils literal notranslate"><span class="pre">METABOLITE_CONCENTRATION_UNITS</span></code></a>), set by
<a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.set_phenomological_supply_constants" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.set_phenomological_supply_constants"><code class="xref py py-meth docutils literal notranslate"><span class="pre">set_phenomological_supply_constants()</span></code></a></p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)">numpy.ndarray</a>[<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.13)">float</a>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.KM_aa_export">
<span class="sig-name descname"><span class="pre">KM_aa_export</span></span><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.KM_aa_export" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>KM for each AA for export portion
of supply (in units of
<a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.METABOLITE_CONCENTRATION_UNITS" title="reconstruction.ecoli.dataclasses.process.metabolism.METABOLITE_CONCENTRATION_UNITS"><code class="xref py py-data docutils literal notranslate"><span class="pre">METABOLITE_CONCENTRATION_UNITS</span></code></a>), set by
<a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.set_phenomological_supply_constants" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.set_phenomological_supply_constants"><code class="xref py py-meth docutils literal notranslate"><span class="pre">set_phenomological_supply_constants()</span></code></a></p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)">numpy.ndarray</a>[<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.13)">float</a>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.fraction_supply_rate">
<span class="sig-name descname"><span class="pre">fraction_supply_rate</span></span><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.fraction_supply_rate" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>fraction of AA supply that comes from
a base synthesis rate, set by
<a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.set_phenomological_supply_constants" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.set_phenomological_supply_constants"><code class="xref py py-meth docutils literal notranslate"><span class="pre">set_phenomological_supply_constants()</span></code></a></p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.13)">float</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.fraction_import_rate">
<span class="sig-name descname"><span class="pre">fraction_import_rate</span></span><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.fraction_import_rate" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>fraction of AA supply that
comes from AA import if nutrients are present, set by
<a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.set_phenomological_supply_constants" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.set_phenomological_supply_constants"><code class="xref py py-meth docutils literal notranslate"><span class="pre">set_phenomological_supply_constants()</span></code></a></p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)">numpy.ndarray</a>[<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.13)">float</a>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.ppgpp_synthesis_reaction">
<span class="sig-name descname"><span class="pre">ppgpp_synthesis_reaction</span></span><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.ppgpp_synthesis_reaction" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>reaction ID for ppGpp synthesis
(catalyzed by RelA and SpoT), set by <a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._build_ppgpp_reactions" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._build_ppgpp_reactions"><code class="xref py py-meth docutils literal notranslate"><span class="pre">_build_ppgpp_reactions()</span></code></a></p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)">str</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.ppgpp_degradation_reaction">
<span class="sig-name descname"><span class="pre">ppgpp_degradation_reaction</span></span><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.ppgpp_degradation_reaction" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>reaction ID for ppGpp degradation
(catalyzed by SpoT), set by <a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._build_ppgpp_reactions" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._build_ppgpp_reactions"><code class="xref py py-meth docutils literal notranslate"><span class="pre">_build_ppgpp_reactions()</span></code></a></p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)">str</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.ppgpp_reaction_names">
<span class="sig-name descname"><span class="pre">ppgpp_reaction_names</span></span><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.ppgpp_reaction_names" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>names of reaction involved in ppGpp,
set by <a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._build_ppgpp_reactions" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._build_ppgpp_reactions"><code class="xref py py-meth docutils literal notranslate"><span class="pre">_build_ppgpp_reactions()</span></code></a></p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.13)">list</a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)">str</a>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.ppgpp_reaction_metabolites">
<span class="sig-name descname"><span class="pre">ppgpp_reaction_metabolites</span></span><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.ppgpp_reaction_metabolites" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>names of metabolites in
ppGpp reactions, set by <a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._build_ppgpp_reactions" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._build_ppgpp_reactions"><code class="xref py py-meth docutils literal notranslate"><span class="pre">_build_ppgpp_reactions()</span></code></a></p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.13)">list</a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)">str</a>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.ppgpp_reaction_stoich">
<span class="sig-name descname"><span class="pre">ppgpp_reaction_stoich</span></span><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.ppgpp_reaction_stoich" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>2D array with metabolites
on rows and reactions on columns containing the stoichiometric
coefficient, set by <a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._build_ppgpp_reactions" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._build_ppgpp_reactions"><code class="xref py py-meth docutils literal notranslate"><span class="pre">_build_ppgpp_reactions()</span></code></a></p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)">numpy.ndarray</a>[<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)">int</a>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.aa_to_exporters">
<span class="sig-name descname"><span class="pre">aa_to_exporters</span></span><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.aa_to_exporters" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>dictonary that maps aa to
transporters involved in export reactions. Set by
<a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.set_mechanistic_export_constants" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.set_mechanistic_export_constants"><code class="xref py py-meth docutils literal notranslate"><span class="pre">set_mechanistic_export_constants()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)">dict</a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)">str</a>, <a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.13)">list</a>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.aa_to_exporters_matrix">
<span class="sig-name descname"><span class="pre">aa_to_exporters_matrix</span></span><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.aa_to_exporters_matrix" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>correlation matrix.
Columns correspond to exporting enzymes and rows to amino acids.
Set by <a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.set_mechanistic_export_constants" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.set_mechanistic_export_constants"><code class="xref py py-meth docutils literal notranslate"><span class="pre">set_mechanistic_export_constants()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)">numpy.ndarray</a>[<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)">int</a>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.aa_exporter_names">
<span class="sig-name descname"><span class="pre">aa_exporter_names</span></span><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.aa_exporter_names" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>names of all exporters. Set by
<a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.set_mechanistic_export_constants" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.set_mechanistic_export_constants"><code class="xref py py-meth docutils literal notranslate"><span class="pre">set_mechanistic_export_constants()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)">numpy.ndarray</a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)">str</a>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.aa_export_kms">
<span class="sig-name descname"><span class="pre">aa_export_kms</span></span><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.aa_export_kms" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>kms corresponding to generic
transport/enzyme reactions for each AA in concentration units.
Set by <a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.set_mechanistic_export_constants" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.set_mechanistic_export_constants"><code class="xref py py-meth docutils literal notranslate"><span class="pre">set_mechanistic_export_constants()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)">numpy.ndarray</a>[<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.13)">float</a>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.export_kcats_per_aa">
<span class="sig-name descname"><span class="pre">export_kcats_per_aa</span></span><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.export_kcats_per_aa" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>kcats corresponding to
generic export reactions for each AA. Units in counts/second.
Set by <a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.set_mechanistic_export_constants" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.set_mechanistic_export_constants"><code class="xref py py-meth docutils literal notranslate"><span class="pre">set_mechanistic_export_constants()</span></code></a> and
<a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.set_mechanistic_export_constants" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.set_mechanistic_export_constants"><code class="xref py py-meth docutils literal notranslate"><span class="pre">set_mechanistic_export_constants()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)">numpy.ndarray</a>[<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.13)">float</a>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.aa_to_importers">
<span class="sig-name descname"><span class="pre">aa_to_importers</span></span><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.aa_to_importers" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>dictonary that maps aa to
transporters involved in import reactions. Set by
<a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.set_mechanistic_uptake_constants" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.set_mechanistic_uptake_constants"><code class="xref py py-meth docutils literal notranslate"><span class="pre">set_mechanistic_uptake_constants()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)">dict</a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)">str</a>, <a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.13)">list</a>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.aa_to_importers_matrix">
<span class="sig-name descname"><span class="pre">aa_to_importers_matrix</span></span><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.aa_to_importers_matrix" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>correlation matrix.
Columns correspond to importing enzymes and rows to amino acids.
Set by <a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.set_mechanistic_export_constants" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.set_mechanistic_export_constants"><code class="xref py py-meth docutils literal notranslate"><span class="pre">set_mechanistic_export_constants()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)">numpy.ndarray</a>[<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)">int</a>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.aa_importer_names">
<span class="sig-name descname"><span class="pre">aa_importer_names</span></span><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.aa_importer_names" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>names of all importers.
Set by <a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.set_mechanistic_export_constants" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.set_mechanistic_export_constants"><code class="xref py py-meth docutils literal notranslate"><span class="pre">set_mechanistic_export_constants()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)">numpy.ndarray</a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)">str</a>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.import_kcats_per_aa">
<span class="sig-name descname"><span class="pre">import_kcats_per_aa</span></span><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.import_kcats_per_aa" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>kcats corresponding
to generic import reactions for each AA. Units in counts/second.
Set by <a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.set_mechanistic_export_constants" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.set_mechanistic_export_constants"><code class="xref py py-meth docutils literal notranslate"><span class="pre">set_mechanistic_export_constants()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)">numpy.ndarray</a>[<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.13)">float</a>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.aa_enzymes">
<span class="sig-name descname"><span class="pre">aa_enzymes</span></span><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.aa_enzymes" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>enzyme ID with location tag for each
enzyme that can catalyze an amino acid pathway with
<a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.enzyme_to_amino_acid" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.enzyme_to_amino_acid"><code class="xref py py-attr docutils literal notranslate"><span class="pre">enzyme_to_amino_acid</span></code></a> mapping these to each amino acid.
Set by <a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.set_mechanistic_supply_constants" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.set_mechanistic_supply_constants"><code class="xref py py-meth docutils literal notranslate"><span class="pre">set_mechanistic_supply_constants()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)">numpy.ndarray</a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)">str</a>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.aa_kcats_fwd">
<span class="sig-name descname"><span class="pre">aa_kcats_fwd</span></span><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.aa_kcats_fwd" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>forward kcat value for each
synthesis pathway in units of <a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.K_CAT_UNITS" title="reconstruction.ecoli.dataclasses.process.metabolism.K_CAT_UNITS"><code class="xref py py-data docutils literal notranslate"><span class="pre">K_CAT_UNITS</span></code></a>, ordered by
amino acid molecule group. Set by
<a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.set_mechanistic_supply_constants" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.set_mechanistic_supply_constants"><code class="xref py py-meth docutils literal notranslate"><span class="pre">set_mechanistic_supply_constants()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)">numpy.ndarray</a>[<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.13)">float</a>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.aa_kcats_rev">
<span class="sig-name descname"><span class="pre">aa_kcats_rev</span></span><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.aa_kcats_rev" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>reverse kcat value for each
synthesis pathway in units of <a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.K_CAT_UNITS" title="reconstruction.ecoli.dataclasses.process.metabolism.K_CAT_UNITS"><code class="xref py py-data docutils literal notranslate"><span class="pre">K_CAT_UNITS</span></code></a>, ordered by
amino acid molecule group. Set by
<a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.set_mechanistic_supply_constants" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.set_mechanistic_supply_constants"><code class="xref py py-meth docutils literal notranslate"><span class="pre">set_mechanistic_supply_constants()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)">numpy.ndarray</a>[<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.13)">float</a>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.aa_kis">
<span class="sig-name descname"><span class="pre">aa_kis</span></span><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.aa_kis" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>KI value for each synthesis pathway
in units of <a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.METABOLITE_CONCENTRATION_UNITS" title="reconstruction.ecoli.dataclasses.process.metabolism.METABOLITE_CONCENTRATION_UNITS"><code class="xref py py-data docutils literal notranslate"><span class="pre">METABOLITE_CONCENTRATION_UNITS</span></code></a>, ordered
by amino acid molecule group. Will be inf if there is no inhibitory
control. Set by <a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.set_mechanistic_supply_constants" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.set_mechanistic_supply_constants"><code class="xref py py-meth docutils literal notranslate"><span class="pre">set_mechanistic_supply_constants()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)">numpy.ndarray</a>[<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.13)">float</a>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.aa_upstream_kms">
<span class="sig-name descname"><span class="pre">aa_upstream_kms</span></span><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.aa_upstream_kms" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>KM value associated with the
amino acid that feeds into each synthesis pathway in units of
<a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.METABOLITE_CONCENTRATION_UNITS" title="reconstruction.ecoli.dataclasses.process.metabolism.METABOLITE_CONCENTRATION_UNITS"><code class="xref py py-data docutils literal notranslate"><span class="pre">METABOLITE_CONCENTRATION_UNITS</span></code></a>, ordered by amino
acid molecule group. Will be 0 if there is no upstream amino
acid considered. Set by
<a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.set_mechanistic_supply_constants" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.set_mechanistic_supply_constants"><code class="xref py py-meth docutils literal notranslate"><span class="pre">set_mechanistic_supply_constants()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.13)">list</a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.13)">list</a>[<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.13)">float</a>]]</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.aa_reverse_kms">
<span class="sig-name descname"><span class="pre">aa_reverse_kms</span></span><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.aa_reverse_kms" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>KM value associated with the
amino acid in each synthesis pathway in units of
<a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.METABOLITE_CONCENTRATION_UNITS" title="reconstruction.ecoli.dataclasses.process.metabolism.METABOLITE_CONCENTRATION_UNITS"><code class="xref py py-data docutils literal notranslate"><span class="pre">METABOLITE_CONCENTRATION_UNITS</span></code></a>, ordered by amino acid
molecule group. Will be inf if the synthesis pathway is not
reversible. Set by <a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.set_mechanistic_supply_constants" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.set_mechanistic_supply_constants"><code class="xref py py-meth docutils literal notranslate"><span class="pre">set_mechanistic_supply_constants()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)">numpy.ndarray</a>[<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.13)">float</a>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.aa_upstream_mapping">
<span class="sig-name descname"><span class="pre">aa_upstream_mapping</span></span><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.aa_upstream_mapping" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>index of the upstream amino
acid that feeds into each synthesis pathway, ordered by amino
acid molecule group. Set by
<a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.set_mechanistic_supply_constants" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.set_mechanistic_supply_constants"><code class="xref py py-meth docutils literal notranslate"><span class="pre">set_mechanistic_supply_constants()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)">numpy.ndarray</a>[<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)">int</a>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.enzyme_to_amino_acid">
<span class="sig-name descname"><span class="pre">enzyme_to_amino_acid</span></span><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.enzyme_to_amino_acid" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>relationship mapping from
aa_enzymes to amino acids (n enzymes, m amino acids).  Will
contain a 1 if the enzyme associated with the row can catalyze
the pathway for the amino acid associated with the column.
Set by <a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.set_mechanistic_supply_constants" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.set_mechanistic_supply_constants"><code class="xref py py-meth docutils literal notranslate"><span class="pre">set_mechanistic_supply_constants()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)">numpy.ndarray</a>[<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.13)">float</a>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.aa_forward_stoich">
<span class="sig-name descname"><span class="pre">aa_forward_stoich</span></span><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.aa_forward_stoich" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>relationship mapping from
upstream amino acids to downstream amino acids (n upstream,
m downstream).  Will contain a -1 if the amino acid associated
with the row is required for synthesis of the amino acid
associated with the column. Set by
<a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.set_mechanistic_supply_constants" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.set_mechanistic_supply_constants"><code class="xref py py-meth docutils literal notranslate"><span class="pre">set_mechanistic_supply_constants()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)">numpy.ndarray</a>[<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.13)">float</a>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.aa_reverse_stoich">
<span class="sig-name descname"><span class="pre">aa_reverse_stoich</span></span><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.aa_reverse_stoich" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>relationship mapping from
upstream amino acids to downstream amino acids (n downstream,
m upstream).  Will contain a -1 if the amino acid associated
with the row is produced through a reverse reaction from
the amino acid associated with the column. Set by
<a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.set_mechanistic_supply_constants" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.set_mechanistic_supply_constants"><code class="xref py py-meth docutils literal notranslate"><span class="pre">set_mechanistic_supply_constants()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)">numpy.ndarray</a>[<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.13)">float</a>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.aa_import_kis">
<span class="sig-name descname"><span class="pre">aa_import_kis</span></span><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.aa_import_kis" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>inhibition constants for amino
acid import based on the internal amino acid concentration</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)">numpy.ndarray</a>[<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.13)">float</a>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.specific_import_rates">
<span class="sig-name descname"><span class="pre">specific_import_rates</span></span><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.specific_import_rates" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>import rates expected
in rich media conditions for each amino acid normalized by dry
cell mass in units of <a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.K_CAT_UNITS" title="reconstruction.ecoli.dataclasses.process.metabolism.K_CAT_UNITS"><code class="xref py py-data docutils literal notranslate"><span class="pre">K_CAT_UNITS</span></code></a> /
<a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.DRY_MASS_UNITS" title="reconstruction.ecoli.dataclasses.process.metabolism.DRY_MASS_UNITS"><code class="xref py py-data docutils literal notranslate"><span class="pre">DRY_MASS_UNITS</span></code></a>, ordered by amino acid molecule group.
Set by <a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.set_mechanistic_supply_constants" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.set_mechanistic_supply_constants"><code class="xref py py-meth docutils literal notranslate"><span class="pre">set_mechanistic_supply_constants()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)">numpy.ndarray</a>[<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.13)">float</a>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.max_specific_import_rates">
<span class="sig-name descname"><span class="pre">max_specific_import_rates</span></span><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.max_specific_import_rates" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>max import rates
for each amino acid without including internal concentration
inhibition normalized by dry cell mass in units of
<a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.K_CAT_UNITS" title="reconstruction.ecoli.dataclasses.process.metabolism.K_CAT_UNITS"><code class="xref py py-data docutils literal notranslate"><span class="pre">K_CAT_UNITS</span></code></a> / <a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.DRY_MASS_UNITS" title="reconstruction.ecoli.dataclasses.process.metabolism.DRY_MASS_UNITS"><code class="xref py py-data docutils literal notranslate"><span class="pre">DRY_MASS_UNITS</span></code></a>, ordered by
amino acid molecule group. Set by
<a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.set_mechanistic_supply_constants" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.set_mechanistic_supply_constants"><code class="xref py py-meth docutils literal notranslate"><span class="pre">set_mechanistic_supply_constants()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)">numpy.ndarray</a>[<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.13)">float</a>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._add_metabolite_charge">
<span class="sig-name descname"><span class="pre">_add_metabolite_charge</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">raw_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/process/metabolism.html#Metabolism._add_metabolite_charge"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._add_metabolite_charge" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>Compiles all metabolite charges.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>raw_data</strong> (<a class="reference internal" href="reconstruction.ecoli.knowledge_base_raw.html#reconstruction.ecoli.knowledge_base_raw.KnowledgeBaseEcoli" title="reconstruction.ecoli.knowledge_base_raw.KnowledgeBaseEcoli"><em>KnowledgeBaseEcoli</em></a>) ‚Äì Raw data object</p>
</dd>
</dl>
<dl class="simple">
<dt>Attributes set:</dt><dd><ul class="simple">
<li><p><a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.metabolite_charge" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.metabolite_charge"><code class="xref py py-attr docutils literal notranslate"><span class="pre">metabolite_charge</span></code></a></p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._build_amino_acid_pathways">
<span class="sig-name descname"><span class="pre">_build_amino_acid_pathways</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">raw_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/process/metabolism.html#Metabolism._build_amino_acid_pathways"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._build_amino_acid_pathways" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>Creates mapping between enzymes and amino acid pathways with
allosteric inhibition feedback from the amino acid.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>raw_data</strong> (<a class="reference internal" href="reconstruction.ecoli.knowledge_base_raw.html#reconstruction.ecoli.knowledge_base_raw.KnowledgeBaseEcoli" title="reconstruction.ecoli.knowledge_base_raw.KnowledgeBaseEcoli"><em>KnowledgeBaseEcoli</em></a>) ‚Äì Raw data object</p></li>
<li><p><strong>sim_data</strong> (<a class="reference internal" href="reconstruction.ecoli.simulation_data.html#reconstruction.ecoli.simulation_data.SimulationDataEcoli" title="reconstruction.ecoli.simulation_data.SimulationDataEcoli"><em>SimulationDataEcoli</em></a>) ‚Äì Simulation data object</p></li>
</ul>
</dd>
</dl>
<dl class="simple">
<dt>Attributes set:</dt><dd><ul class="simple">
<li><p><a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.aa_synthesis_pathways" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.aa_synthesis_pathways"><code class="xref py py-attr docutils literal notranslate"><span class="pre">aa_synthesis_pathways</span></code></a></p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._build_biomass">
<span class="sig-name descname"><span class="pre">_build_biomass</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">raw_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/process/metabolism.html#Metabolism._build_biomass"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._build_biomass" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>Calculates metabolite concentration targets.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>raw_data</strong> (<a class="reference internal" href="reconstruction.ecoli.knowledge_base_raw.html#reconstruction.ecoli.knowledge_base_raw.KnowledgeBaseEcoli" title="reconstruction.ecoli.knowledge_base_raw.KnowledgeBaseEcoli"><em>KnowledgeBaseEcoli</em></a>) ‚Äì Raw data object</p></li>
<li><p><strong>sim_data</strong> (<a class="reference internal" href="reconstruction.ecoli.simulation_data.html#reconstruction.ecoli.simulation_data.SimulationDataEcoli" title="reconstruction.ecoli.simulation_data.SimulationDataEcoli"><em>SimulationDataEcoli</em></a>) ‚Äì Simulation data object</p></li>
</ul>
</dd>
</dl>
<dl class="simple">
<dt>Attributes set:</dt><dd><ul class="simple">
<li><p><a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.concentration_updates" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.concentration_updates"><code class="xref py py-attr docutils literal notranslate"><span class="pre">concentration_updates</span></code></a></p></li>
<li><p><a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.conc_dict" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.conc_dict"><code class="xref py py-attr docutils literal notranslate"><span class="pre">conc_dict</span></code></a></p></li>
<li><p><a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.nutrients_to_internal_conc" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.nutrients_to_internal_conc"><code class="xref py py-attr docutils literal notranslate"><span class="pre">nutrients_to_internal_conc</span></code></a></p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._build_linked_metabolites">
<span class="sig-name descname"><span class="pre">_build_linked_metabolites</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">raw_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">conc_dict</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/process/metabolism.html#Metabolism._build_linked_metabolites"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._build_linked_metabolites" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>Calculates ratio between linked metabolites to keep it constant
throughout a simulation.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>raw_data</strong> (<a class="reference internal" href="reconstruction.ecoli.knowledge_base_raw.html#reconstruction.ecoli.knowledge_base_raw.KnowledgeBaseEcoli" title="reconstruction.ecoli.knowledge_base_raw.KnowledgeBaseEcoli"><em>KnowledgeBaseEcoli</em></a>) ‚Äì Raw data object</p></li>
<li><p><strong>conc_dict</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)"><em>dict</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>, </em><em>Unum</em><em>]</em>) ‚Äì Mapping of metabolite IDs to homeostatic concentrations
calculated by <a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._build_biomass" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._build_biomass"><code class="xref py py-meth docutils literal notranslate"><span class="pre">_build_biomass()</span></code></a></p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><p>Mapping from a linked metabolite to its lead
metabolite and concentration ratio to be maintained:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">{</span><span class="s1">&#39;lead&#39;</span> <span class="p">(</span><span class="nb">str</span><span class="p">):</span> <span class="n">metabolite</span> <span class="n">to</span> <span class="n">link</span> <span class="n">the</span> <span class="n">concentration</span> <span class="n">to</span><span class="p">,</span>
<span class="s1">&#39;ratio&#39;</span> <span class="p">(</span><span class="nb">float</span><span class="p">):</span> <span class="n">ratio</span> <span class="n">to</span> <span class="n">multiply</span> <span class="n">the</span> <span class="n">lead</span> <span class="n">concentration</span> <span class="n">by</span><span class="p">}</span>
</pre></div>
</div>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)">dict</a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)">str</a>, <a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)">dict</a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)">str</a>, <a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Any" title="(in Python v3.13)"><em>Any</em></a>]]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._build_metabolism">
<span class="sig-name descname"><span class="pre">_build_metabolism</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">raw_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/process/metabolism.html#Metabolism._build_metabolism"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._build_metabolism" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>Build the matrices/vectors for metabolism (FBA)
Reads in and stores reaction and kinetic constraint information</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>raw_data</strong> (<a class="reference internal" href="reconstruction.ecoli.knowledge_base_raw.html#reconstruction.ecoli.knowledge_base_raw.KnowledgeBaseEcoli" title="reconstruction.ecoli.knowledge_base_raw.KnowledgeBaseEcoli"><em>KnowledgeBaseEcoli</em></a>) ‚Äì Raw data object</p></li>
<li><p><strong>sim_data</strong> (<a class="reference internal" href="reconstruction.ecoli.simulation_data.html#reconstruction.ecoli.simulation_data.SimulationDataEcoli" title="reconstruction.ecoli.simulation_data.SimulationDataEcoli"><em>SimulationDataEcoli</em></a>) ‚Äì Simulation data object</p></li>
</ul>
</dd>
</dl>
<dl class="simple">
<dt>Attributes set:</dt><dd><ul class="simple">
<li><p><a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.kinetic_constraint_reactions" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.kinetic_constraint_reactions"><code class="xref py py-attr docutils literal notranslate"><span class="pre">kinetic_constraint_reactions</span></code></a></p></li>
<li><p><a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.kinetic_constraint_enzymes" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.kinetic_constraint_enzymes"><code class="xref py py-attr docutils literal notranslate"><span class="pre">kinetic_constraint_enzymes</span></code></a></p></li>
<li><p><a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.kinetic_constraint_substrates" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.kinetic_constraint_substrates"><code class="xref py py-attr docutils literal notranslate"><span class="pre">kinetic_constraint_substrates</span></code></a></p></li>
<li><p><a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._kcats" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._kcats"><code class="xref py py-attr docutils literal notranslate"><span class="pre">_kcats</span></code></a></p></li>
<li><p><a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._saturations" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._saturations"><code class="xref py py-attr docutils literal notranslate"><span class="pre">_saturations</span></code></a></p></li>
<li><p><a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._enzymes" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._enzymes"><code class="xref py py-attr docutils literal notranslate"><span class="pre">_enzymes</span></code></a></p></li>
<li><p><a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.constraint_is_kcat_only" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.constraint_is_kcat_only"><code class="xref py py-attr docutils literal notranslate"><span class="pre">constraint_is_kcat_only</span></code></a></p></li>
<li><p><a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._compiled_enzymes" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._compiled_enzymes"><code class="xref py py-attr docutils literal notranslate"><span class="pre">_compiled_enzymes</span></code></a></p></li>
<li><p><a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._compiled_saturation" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._compiled_saturation"><code class="xref py py-attr docutils literal notranslate"><span class="pre">_compiled_saturation</span></code></a></p></li>
<li><p><a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.reaction_stoich" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.reaction_stoich"><code class="xref py py-attr docutils literal notranslate"><span class="pre">reaction_stoich</span></code></a></p></li>
<li><p><a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.maintenance_reaction" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.maintenance_reaction"><code class="xref py py-attr docutils literal notranslate"><span class="pre">maintenance_reaction</span></code></a></p></li>
<li><p><a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.reaction_catalysts" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.reaction_catalysts"><code class="xref py py-attr docutils literal notranslate"><span class="pre">reaction_catalysts</span></code></a></p></li>
<li><p><a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.catalyst_ids" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.catalyst_ids"><code class="xref py py-attr docutils literal notranslate"><span class="pre">catalyst_ids</span></code></a></p></li>
<li><p><a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.reactions_with_catalyst" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.reactions_with_catalyst"><code class="xref py py-attr docutils literal notranslate"><span class="pre">reactions_with_catalyst</span></code></a></p></li>
<li><p><a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.catalysis_matrix_I" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.catalysis_matrix_I"><code class="xref py py-attr docutils literal notranslate"><span class="pre">catalysis_matrix_I</span></code></a></p></li>
<li><p><a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.catalysis_matrix_J" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.catalysis_matrix_J"><code class="xref py py-attr docutils literal notranslate"><span class="pre">catalysis_matrix_J</span></code></a></p></li>
<li><p><a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.catalysis_matrix_V" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.catalysis_matrix_V"><code class="xref py py-attr docutils literal notranslate"><span class="pre">catalysis_matrix_V</span></code></a></p></li>
<li><p><a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.use_all_constraints" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.use_all_constraints"><code class="xref py py-attr docutils literal notranslate"><span class="pre">use_all_constraints</span></code></a></p></li>
<li><p><a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.constraints_to_disable" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.constraints_to_disable"><code class="xref py py-attr docutils literal notranslate"><span class="pre">constraints_to_disable</span></code></a></p></li>
<li><p><a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.base_reaction_ids" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.base_reaction_ids"><code class="xref py py-attr docutils literal notranslate"><span class="pre">base_reaction_ids</span></code></a></p></li>
<li><p><a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.reaction_id_to_base_reaction_id" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.reaction_id_to_base_reaction_id"><code class="xref py py-attr docutils literal notranslate"><span class="pre">reaction_id_to_base_reaction_id</span></code></a></p></li>
<li><p><a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.amino_acid_export_kms" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.amino_acid_export_kms"><code class="xref py py-attr docutils literal notranslate"><span class="pre">amino_acid_export_kms</span></code></a></p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._build_ppgpp_reactions">
<span class="sig-name descname"><span class="pre">_build_ppgpp_reactions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">raw_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/process/metabolism.html#Metabolism._build_ppgpp_reactions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._build_ppgpp_reactions" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>Creates structures for ppGpp reactions for use in polypeptide_elongation.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>raw_data</strong> (<a class="reference internal" href="reconstruction.ecoli.knowledge_base_raw.html#reconstruction.ecoli.knowledge_base_raw.KnowledgeBaseEcoli" title="reconstruction.ecoli.knowledge_base_raw.KnowledgeBaseEcoli"><em>KnowledgeBaseEcoli</em></a>) ‚Äì Raw data object</p></li>
<li><p><strong>sim_data</strong> (<a class="reference internal" href="reconstruction.ecoli.simulation_data.html#reconstruction.ecoli.simulation_data.SimulationDataEcoli" title="reconstruction.ecoli.simulation_data.SimulationDataEcoli"><em>SimulationDataEcoli</em></a>) ‚Äì Simulation data object</p></li>
</ul>
</dd>
</dl>
<dl class="simple">
<dt>Attributes set:</dt><dd><ul class="simple">
<li><p><a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.ppgpp_synthesis_reaction" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.ppgpp_synthesis_reaction"><code class="xref py py-attr docutils literal notranslate"><span class="pre">ppgpp_synthesis_reaction</span></code></a></p></li>
<li><p><a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.ppgpp_degradation_reaction" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.ppgpp_degradation_reaction"><code class="xref py py-attr docutils literal notranslate"><span class="pre">ppgpp_degradation_reaction</span></code></a></p></li>
<li><p><a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.ppgpp_reaction_names" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.ppgpp_reaction_names"><code class="xref py py-attr docutils literal notranslate"><span class="pre">ppgpp_reaction_names</span></code></a></p></li>
<li><p><a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.ppgpp_reaction_metabolites" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.ppgpp_reaction_metabolites"><code class="xref py py-attr docutils literal notranslate"><span class="pre">ppgpp_reaction_metabolites</span></code></a></p></li>
<li><p><a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.ppgpp_reaction_stoich" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.ppgpp_reaction_stoich"><code class="xref py py-attr docutils literal notranslate"><span class="pre">ppgpp_reaction_stoich</span></code></a></p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._build_transport_reactions">
<span class="sig-name descname"><span class="pre">_build_transport_reactions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">raw_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/process/metabolism.html#Metabolism._build_transport_reactions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._build_transport_reactions" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>Creates list of transport reactions that are included in the
reaction network.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>raw_data</strong> (<a class="reference internal" href="reconstruction.ecoli.knowledge_base_raw.html#reconstruction.ecoli.knowledge_base_raw.KnowledgeBaseEcoli" title="reconstruction.ecoli.knowledge_base_raw.KnowledgeBaseEcoli"><em>KnowledgeBaseEcoli</em></a>) ‚Äì Raw data object</p></li>
<li><p><strong>sim_data</strong> (<a class="reference internal" href="reconstruction.ecoli.simulation_data.html#reconstruction.ecoli.simulation_data.SimulationDataEcoli" title="reconstruction.ecoli.simulation_data.SimulationDataEcoli"><em>SimulationDataEcoli</em></a>) ‚Äì Simulation data object</p></li>
</ul>
</dd>
</dl>
<dl class="simple">
<dt>Attributes set:</dt><dd><ul class="simple">
<li><p><a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.transport_reactions" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.transport_reactions"><code class="xref py py-attr docutils literal notranslate"><span class="pre">transport_reactions</span></code></a></p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._construct_default_saturation_equation">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">_construct_default_saturation_equation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mets</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">kms</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">kis</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">known_mets</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/process/metabolism.html#Metabolism._construct_default_saturation_equation"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._construct_default_saturation_equation" title="Link to this definition">ÔÉÅ</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>mets</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.13)"><em>list</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>]</em>) ‚Äì metabolite IDs with location tag for KM and KI
parameters ordered to match order of kms then kis</p></li>
<li><p><strong>kms</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.13)"><em>list</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.13)"><em>float</em></a><em>]</em>) ‚Äì KM parameters associated with mets</p></li>
<li><p><strong>kis</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.13)"><em>list</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.13)"><em>float</em></a><em>]</em>) ‚Äì KI parameters associated with mets</p></li>
<li><p><strong>known_mets</strong> (<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Iterable" title="(in Python v3.13)"><em>Iterable</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>]</em>) ‚Äì metabolite IDs with location tag with known
concentrations</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Saturation equation with metabolites to replace delimited
by double quote (e.g. ‚Äúmetabolite‚Äù)</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)">str</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._extract_custom_constraint">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">_extract_custom_constraint</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">constraint</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reactant_tags</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">product_tags</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">known_mets</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/process/metabolism.html#Metabolism._extract_custom_constraint"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._extract_custom_constraint" title="Link to this definition">ÔÉÅ</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>constraint</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)"><em>dict</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Any" title="(in Python v3.13)"><em>Any</em></a><em>]</em>) ‚Äì <p>values defining a kinetic constraint:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">{</span><span class="s1">&#39;customRateEquation&#39;</span> <span class="p">(</span><span class="nb">str</span><span class="p">):</span> <span class="n">mathematical</span> <span class="n">representation</span> <span class="n">of</span>
        <span class="n">rate</span> <span class="p">(</span><span class="n">must</span> <span class="n">contain</span> <span class="s1">&#39;kcat*E&#39;</span><span class="p">),</span>
<span class="s1">&#39;customParameterVariables&#39;</span> <span class="p">(</span><span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]):</span> <span class="n">mapping</span> <span class="n">of</span>
        <span class="n">variable</span> <span class="n">names</span> <span class="ow">in</span> <span class="n">the</span> <span class="n">rate</span> <span class="n">equation</span> <span class="n">to</span> <span class="n">metabolite</span> <span class="n">IDs</span>
        <span class="n">without</span> <span class="n">location</span> <span class="n">tags</span> <span class="p">(</span><span class="n">must</span> <span class="n">contain</span> <span class="s1">&#39;E&#39;</span><span class="p">),</span>
<span class="s1">&#39;customParameterConstants&#39;</span> <span class="p">(</span><span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]):</span> <span class="n">constant</span> <span class="n">strings</span>
        <span class="ow">in</span> <span class="n">the</span> <span class="n">rate</span> <span class="n">equation</span> <span class="n">that</span> <span class="n">correspond</span> <span class="n">to</span> <span class="n">values</span> <span class="p">(</span><span class="n">must</span>
        <span class="n">contain</span> <span class="s1">&#39;kcat&#39;</span><span class="p">),</span>
<span class="s1">&#39;customParameterConstantValues&#39;</span> <span class="p">(</span><span class="nb">list</span><span class="p">[</span><span class="nb">float</span><span class="p">]):</span> <span class="n">values</span> <span class="k">for</span>
        <span class="n">each</span> <span class="n">of</span> <span class="n">the</span> <span class="n">constant</span> <span class="n">strings</span><span class="p">,</span>
<span class="s1">&#39;Temp&#39;</span> <span class="p">(</span><span class="nb">float</span> <span class="ow">or</span> <span class="s1">&#39;&#39;</span><span class="p">):</span> <span class="n">temperature</span> <span class="n">of</span> <span class="n">measurement</span><span class="p">}</span>
</pre></div>
</div>
</p></li>
<li><p><strong>reactant_tags</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)"><em>dict</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>]</em>) ‚Äì mapping of molecule IDs without a location tag to
molecule IDs with a location tag for all reactants</p></li>
<li><p><strong>product_tags</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)"><em>dict</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>]</em>) ‚Äì mapping of molecule IDs without a location tag to
molecule IDs with a location tag for all products</p></li>
<li><p><strong>known_mets</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#set" title="(in Python v3.13)"><em>set</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>]</em>) ‚Äì molecule IDs with a location tag for molecules with
known concentrations</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><p>2-element tuple containing</p>
<blockquote>
<div><ul class="simple">
<li><p>kcats: temperature adjusted kcat value, in units of 1/s</p></li>
<li><p>saturation: saturation equation with metabolites to replace
delimited by double quote (eg. ‚Äúmetabolite‚Äù)</p></li>
</ul>
</div></blockquote>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)">tuple</a>[<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)">tuple</a>[<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)">int</a>, ‚Ä¶], <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.dtype.html#numpy.dtype" title="(in NumPy v2.3)"><em>dtype</em></a>[<em>float64</em>]] | None, <a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.13)">list</a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)">str</a>]]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._is_transport_rxn">
<span class="sig-name descname"><span class="pre">_is_transport_rxn</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">stoich</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/process/metabolism.html#Metabolism._is_transport_rxn"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._is_transport_rxn" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>Determines if the metabolic reaction with a given stoichiometry is a
transport reactions that transports metabolites between different
compartments. A metabolic reaction is considered to be a transport
reaction if the substrate set and the product share the same metabolite
tagged into different compartments.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>stoich</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)"><em>dict</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a><em>]</em>) ‚Äì <p>Stoichiometry of the metabolic reaction:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">{</span>
<span class="n">metabolite</span> <span class="n">ID</span> <span class="p">(</span><span class="nb">str</span><span class="p">):</span> <span class="n">stoichiometric</span> <span class="n">coefficient</span> <span class="p">(</span><span class="nb">int</span><span class="p">)</span>
<span class="p">}</span>
</pre></div>
</div>
</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>True if given stoichiometry is a transport reaction</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)">bool</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._lambdify_constraints">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">_lambdify_constraints</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">constraints</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/process/metabolism.html#Metabolism._lambdify_constraints"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._lambdify_constraints" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>Creates str representations of kinetic terms to be used to create
kinetic constraints that are returned with getKineticConstraints().</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>constraints</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)"><em>dict</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Any" title="(in Python v3.13)"><em>Any</em></a><em>]</em>) ‚Äì <p>valid kinetic constraints for each reaction:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">{</span><span class="n">reaction</span> <span class="n">ID</span><span class="p">:</span> <span class="p">{</span>
        <span class="s1">&#39;enzyme&#39;</span><span class="p">:</span> <span class="n">enzyme</span> <span class="n">catalyst</span> <span class="p">(</span><span class="nb">str</span><span class="p">),</span>
        <span class="s1">&#39;kcat&#39;</span><span class="p">:</span> <span class="n">kcat</span> <span class="n">values</span> <span class="p">(</span><span class="nb">list</span><span class="p">[</span><span class="nb">float</span><span class="p">]),</span>
        <span class="s1">&#39;saturation&#39;</span><span class="p">:</span> <span class="n">saturation</span> <span class="n">equations</span> <span class="p">(</span><span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">])</span>
<span class="p">}}</span>
</pre></div>
</div>
</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><p>7-element tuple containing</p>
<blockquote>
<div><ul class="simple">
<li><p>rxns: sorted reaction IDs for reactions with a kinetic
constraint</p></li>
<li><p>enzymes: sorted enzyme IDs for enzymes that catalyze a
kinetic reaction</p></li>
<li><p>substrates: sorted substrate IDs for substrates that are
needed for kinetic saturation terms</p></li>
<li><p>all_kcats: (n rxns, 3) min, mean and max kcat value for
each reaction</p></li>
<li><p>all_saturations: sympy str representation of a list of
saturation terms (eg. ‚Äò[s[0] / (1 + s[0]), 2 / (2 + s[1])]‚Äô)</p></li>
<li><p>all_enzymes: sympy str representation of enzymes for each
reaction (e.g. ‚Äò[e[0], e[2], e[1]]‚Äô)</p></li>
<li><p>constraint_is_kcat_only: True if reaction only has kcat
values and no saturation terms</p></li>
</ul>
</div></blockquote>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)">tuple</a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.13)">list</a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)">str</a>], <a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.13)">list</a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)">str</a>], <a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.13)">list</a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)">str</a>], <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)">tuple</a>[<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)">int</a>, ‚Ä¶], <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.dtype.html#numpy.dtype" title="(in NumPy v2.3)"><em>dtype</em></a>[<em>float64</em>]], <a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)">str</a>, <a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)">str</a>, <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)">tuple</a>[<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)">int</a>, ‚Ä¶], <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.dtype.html#numpy.dtype" title="(in NumPy v2.3)"><em>dtype</em></a>[<a class="reference external" href="https://numpy.org/doc/stable/reference/arrays.scalars.html#numpy.bool" title="(in NumPy v2.3)"><em>bool</em></a>]]]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._replace_enzyme_reactions">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">_replace_enzyme_reactions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">constraints</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stoich</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rxn_catalysts</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reversible_rxns</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rxn_id_to_compiled_id</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/process/metabolism.html#Metabolism._replace_enzyme_reactions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._replace_enzyme_reactions" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>Modifies reaction IDs in data structures to duplicate reactions with
kinetic constraints and multiple enzymes.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>constraints</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)"><em>dict</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)"><em>tuple</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>]</em><em>, </em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)"><em>dict</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.13)"><em>list</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Any" title="(in Python v3.13)"><em>Any</em></a><em>]</em><em>]</em><em>]</em>) ‚Äì <p>valid kinetic constraints for each reaction/enzyme pair:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">{(</span><span class="n">reaction</span> <span class="n">ID</span><span class="p">,</span> <span class="n">enzyme</span> <span class="k">with</span> <span class="n">location</span> <span class="n">tag</span><span class="p">):</span> <span class="p">{</span>
        <span class="s1">&#39;kcat&#39;</span><span class="p">:</span> <span class="n">kcat</span> <span class="n">values</span> <span class="p">(</span><span class="nb">list</span><span class="p">[</span><span class="nb">float</span><span class="p">]),</span>
        <span class="s1">&#39;saturation&#39;</span><span class="p">:</span> <span class="n">saturation</span> <span class="n">equations</span> <span class="p">(</span><span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">])</span>
<span class="p">}}</span>
</pre></div>
</div>
</p></li>
<li><p><strong>stoich</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)"><em>dict</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)"><em>dict</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a><em>]</em><em>]</em>) ‚Äì <p>stoichiometry of metabolites for each reaction (if None, data
is loaded from raw_data and sim_data):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">{</span><span class="n">reaction</span> <span class="n">ID</span><span class="p">:</span> <span class="p">{</span><span class="n">metabolite</span> <span class="n">ID</span> <span class="k">with</span> <span class="n">location</span> <span class="n">tag</span><span class="p">:</span> <span class="n">stoichiometry</span><span class="p">}}</span>
</pre></div>
</div>
</p></li>
<li><p><strong>rxn_catalysts</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)"><em>dict</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.13)"><em>list</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>]</em><em>]</em>) ‚Äì <p>enzyme catalysts for each reaction with known catalysts,
likely a subset of reactions in stoich (if None, data is loaded
from raw_data and sim_data):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">{</span><span class="n">reaction</span> <span class="n">ID</span><span class="p">:</span> <span class="n">enzyme</span> <span class="n">IDs</span> <span class="k">with</span> <span class="n">location</span> <span class="n">tag</span><span class="p">}</span>
</pre></div>
</div>
</p></li>
<li><p><strong>reversible_rxns</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.13)"><em>list</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>]</em>) ‚Äì reaction IDs for reactions that have a reverse
complement, does not have reverse tag</p></li>
<li><p><strong>rxn_id_to_compiled_id</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)"><em>dict</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>]</em>) ‚Äì mapping from reaction IDs to the IDs of the
original reactions they were derived from</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><p>5-element tuple containing</p>
<blockquote>
<div><ul>
<li><p>new_constraints: valid kinetic constraints for each reaction:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">{</span><span class="n">reaction</span> <span class="n">ID</span><span class="p">:</span> <span class="p">{</span>
        <span class="s1">&#39;enzyme&#39;</span><span class="p">:</span> <span class="n">enzyme</span> <span class="n">catalyst</span> <span class="p">(</span><span class="nb">str</span><span class="p">),</span>
        <span class="s1">&#39;kcat&#39;</span><span class="p">:</span> <span class="n">kcat</span> <span class="n">values</span> <span class="p">(</span><span class="nb">list</span><span class="p">[</span><span class="nb">float</span><span class="p">]),</span>
        <span class="s1">&#39;saturation&#39;</span><span class="p">:</span> <span class="n">saturation</span> <span class="n">equations</span> <span class="p">(</span><span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">])</span>
<span class="p">}}</span>
</pre></div>
</div>
</li>
<li><p>stoich: stoichiometry of metabolites for each reaction with
updated reactions for enzyme catalyzed kinetic reactions:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">{</span><span class="n">reaction</span> <span class="n">ID</span><span class="p">:</span> <span class="p">{</span><span class="n">metabolite</span> <span class="n">ID</span> <span class="k">with</span> <span class="n">location</span> <span class="n">tag</span><span class="p">:</span> <span class="n">stoichiometry</span><span class="p">}}</span>
</pre></div>
</div>
</li>
<li><p>rxn_catalysts: enzyme catalysts for each reaction with known
catalysts, likely a subset of reactions in stoich with
updated reactions for enzyme catalyzed kinetic reactions:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">{</span><span class="n">reaction</span> <span class="n">ID</span><span class="p">:</span> <span class="n">enzyme</span> <span class="n">IDs</span> <span class="k">with</span> <span class="n">location</span> <span class="n">tag</span><span class="p">}</span>
</pre></div>
</div>
</li>
<li><p>reversible_rxns: reaction IDs for reactions that have a reverse
complement with updated reactions for enzyme catalyzed kinetic
reactions, does not have reverse tag</p></li>
<li><p>rxn_id_to_compiled_id: mapping from reaction IDs to the IDs
of the original reactions they were derived from, with updated
reactions for enzyme catalyzed kinetic reactions</p></li>
</ul>
</div></blockquote>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)">tuple</a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)">dict</a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)">str</a>, <a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Any" title="(in Python v3.13)"><em>Any</em></a>], <a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)">dict</a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)">str</a>, <a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)">dict</a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)">str</a>, <a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)">int</a>]], <a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)">dict</a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)">str</a>, <a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.13)">list</a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)">str</a>]], <a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.13)">list</a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)">str</a>], <a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)">dict</a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)">str</a>, <a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)">str</a>]]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._set_solver_values">
<span class="sig-name descname"><span class="pre">_set_solver_values</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">constants</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/process/metabolism.html#Metabolism._set_solver_values"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._set_solver_values" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>Sets values to be used in the FBA solver.</p>
<dl class="simple">
<dt>Attributes set:</dt><dd><ul class="simple">
<li><p><a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.solver" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.solver"><code class="xref py py-attr docutils literal notranslate"><span class="pre">solver</span></code></a></p></li>
<li><p><a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.kinetic_objective_weight" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.kinetic_objective_weight"><code class="xref py py-attr docutils literal notranslate"><span class="pre">kinetic_objective_weight</span></code></a></p></li>
<li><p><a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.secretion_penalty_coeff" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.secretion_penalty_coeff"><code class="xref py py-attr docutils literal notranslate"><span class="pre">secretion_penalty_coeff</span></code></a></p></li>
</ul>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>constants</strong> (<a class="reference internal" href="reconstruction.ecoli.dataclasses.constants.html#reconstruction.ecoli.dataclasses.constants.Constants" title="reconstruction.ecoli.dataclasses.constants.Constants"><em>Constants</em></a>)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.aa_supply_scaling">
<span class="sig-name descname"><span class="pre">aa_supply_scaling</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">aa_conc</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">aa_present</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/process/metabolism.html#Metabolism.aa_supply_scaling"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.aa_supply_scaling" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>Called during polypeptide_elongation process
Determine amino acid supply rate scaling based on current amino acid
concentrations.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>aa_conc</strong> (<em>Unum</em>) ‚Äì internal concentration for each amino acid (ndarray[float])</p></li>
<li><p><strong>aa_present</strong> (<em>Unum</em>) ‚Äì whether each amino acid is in the
external environment or not (ndarray[bool])</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Scaling for the supply of each amino acid with
higher supply rate if &gt;1, lower supply rate if &lt;1</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)">tuple</a>[<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)">int</a>, ‚Ä¶], <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.dtype.html#numpy.dtype" title="(in NumPy v2.3)"><em>dtype</em></a>[<em>float64</em>]]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.amino_acid_export">
<span class="sig-name descname"><span class="pre">amino_acid_export</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">aa_transporters_counts</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">aa_conc</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mechanistic_uptake</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/process/metabolism.html#Metabolism.amino_acid_export"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.amino_acid_export" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>Calculate the rate of amino acid export.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>aa_transporters_counts</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)"><em>tuple</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a><em>, </em><em>...</em><em>]</em><em>, </em><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.dtype.html#numpy.dtype" title="(in NumPy v2.3)"><em>dtype</em></a><em>[</em><em>int64</em><em>]</em><em>]</em>) ‚Äì counts of each transporter</p></li>
<li><p><strong>aa_conc</strong> (<em>Unum</em><em> | </em><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)"><em>tuple</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a><em>, </em><em>...</em><em>]</em><em>, </em><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.dtype.html#numpy.dtype" title="(in NumPy v2.3)"><em>dtype</em></a><em>[</em><em>float64</em><em>]</em><em>]</em>) ‚Äì concentrations of each amino acid with mol/volume units</p></li>
<li><p><strong>mechanistic_uptake</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a>) ‚Äì if true, the uptake is calculated based on transporters</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Rate of export for each amino acid (unitless but
represents counts of amino acid per second)</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)">tuple</a>[<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)">int</a>, ‚Ä¶], <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.dtype.html#numpy.dtype" title="(in NumPy v2.3)"><em>dtype</em></a>[<em>float64</em>]]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.amino_acid_import">
<span class="sig-name descname"><span class="pre">amino_acid_import</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">aa_in_media</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dry_mass</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">internal_aa_conc</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">aa_transporters_counts</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mechanistic_uptake</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/process/metabolism.html#Metabolism.amino_acid_import"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.amino_acid_import" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>Calculate the rate of amino acid uptake.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>aa_in_media</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)"><em>tuple</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a><em>, </em><em>...</em><em>]</em><em>, </em><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.dtype.html#numpy.dtype" title="(in NumPy v2.3)"><em>dtype</em></a><em>[</em><a class="reference external" href="https://numpy.org/doc/stable/reference/arrays.scalars.html#numpy.bool" title="(in NumPy v2.3)"><em>bool</em></a><em>]</em><em>]</em>) ‚Äì bool for each amino acid being present in current media</p></li>
<li><p><strong>dry_mass</strong> (<em>Unum</em>) ‚Äì current dry mass of the cell, with mass units</p></li>
<li><p><strong>internal_aa_conc</strong> (<em>Unum</em><em> | </em><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)"><em>tuple</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a><em>, </em><em>...</em><em>]</em><em>, </em><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.dtype.html#numpy.dtype" title="(in NumPy v2.3)"><em>dtype</em></a><em>[</em><em>float64</em><em>]</em><em>]</em>) ‚Äì internal concentrations of amino acids</p></li>
<li><p><strong>aa_transporters_counts</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)"><em>tuple</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a><em>, </em><em>...</em><em>]</em><em>, </em><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.dtype.html#numpy.dtype" title="(in NumPy v2.3)"><em>dtype</em></a><em>[</em><em>int64</em><em>]</em><em>]</em>) ‚Äì counts of each transporter</p></li>
<li><p><strong>mechanistic_uptake</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a>) ‚Äì if true, the uptake is calculated based on
transporters</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Rate of uptake for each amino acid (unitless but
represents counts of amino acid per second)</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)">tuple</a>[<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)">int</a>, ‚Ä¶], <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.dtype.html#numpy.dtype" title="(in NumPy v2.3)"><em>dtype</em></a>[<em>float64</em>]]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.amino_acid_synthesis">
<span class="sig-name descname"><span class="pre">amino_acid_synthesis</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">counts_per_aa_fwd</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">counts_per_aa_rev</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">aa_conc</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/process/metabolism.html#Metabolism.amino_acid_synthesis"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.amino_acid_synthesis" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>Calculate the net rate of synthesis for amino acid pathways (can be
negative with reverse reactions).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>counts_per_aa_fwd</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)"><em>tuple</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a><em>, </em><em>...</em><em>]</em><em>, </em><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.dtype.html#numpy.dtype" title="(in NumPy v2.3)"><em>dtype</em></a><em>[</em><em>int64</em><em>]</em><em>]</em>) ‚Äì counts for enzymes in forward reactions for
each amino acid</p></li>
<li><p><strong>counts_per_aa_rev</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)"><em>tuple</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a><em>, </em><em>...</em><em>]</em><em>, </em><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.dtype.html#numpy.dtype" title="(in NumPy v2.3)"><em>dtype</em></a><em>[</em><em>int64</em><em>]</em><em>]</em>) ‚Äì counts for enzymes in loss reactions for each
amino acid</p></li>
<li><p><strong>aa_conc</strong> (<em>Unum</em><em> | </em><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)"><em>tuple</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a><em>, </em><em>...</em><em>]</em><em>, </em><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.dtype.html#numpy.dtype" title="(in NumPy v2.3)"><em>dtype</em></a><em>[</em><em>float64</em><em>]</em><em>]</em>) ‚Äì concentrations of each amino acid with mol/volume units</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><p>3-element tuple containing</p>
<blockquote>
<div><ul class="simple">
<li><p>synthesis: net rate of synthesis for each amino acid pathway.
array is unitless but represents counts of amino acid per second</p></li>
<li><p>forward_fraction: saturated fraction for forward reactions</p></li>
<li><p>loss_fraction: saturated fraction for loss reactions</p></li>
</ul>
</div></blockquote>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)">tuple</a>[<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)">tuple</a>[<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)">int</a>, ‚Ä¶], <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.dtype.html#numpy.dtype" title="(in NumPy v2.3)"><em>dtype</em></a>[<em>float64</em>]], <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)">tuple</a>[<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)">int</a>, ‚Ä¶], <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.dtype.html#numpy.dtype" title="(in NumPy v2.3)"><em>dtype</em></a>[<em>float64</em>]], <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)">tuple</a>[<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)">int</a>, ‚Ä¶], <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.dtype.html#numpy.dtype" title="(in NumPy v2.3)"><em>dtype</em></a>[<em>float64</em>]]]</p>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Currently does not match saturation terms used in calc_kcats since
it assumes only a reverse or degradation KM exists for simpler calculations</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.exchange_constraints">
<span class="sig-name descname"><span class="pre">exchange_constraints</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">exchangeIDs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">coefficient</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">targetUnits</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">media_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unconstrained</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">constrained</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">concModificationsBasedOnCondition</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/process/metabolism.html#Metabolism.exchange_constraints"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.exchange_constraints" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>Called during Metabolism process
Returns the homeostatic objective concentrations based on the current nutrients
Returns levels for external molecules available to exchange based on the current nutrients</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.extract_kinetic_constraints">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">extract_kinetic_constraints</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">raw_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stoich</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">catalysts</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">known_metabolites</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/process/metabolism.html#Metabolism.extract_kinetic_constraints"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.extract_kinetic_constraints" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>Load and parse kinetic constraint information from raw_data</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>raw_data</strong> (<a class="reference internal" href="reconstruction.ecoli.knowledge_base_raw.html#reconstruction.ecoli.knowledge_base_raw.KnowledgeBaseEcoli" title="reconstruction.ecoli.knowledge_base_raw.KnowledgeBaseEcoli"><em>KnowledgeBaseEcoli</em></a>) ‚Äì knowledge base data</p></li>
<li><p><strong>sim_data</strong> (<a class="reference internal" href="reconstruction.ecoli.simulation_data.html#reconstruction.ecoli.simulation_data.SimulationDataEcoli" title="reconstruction.ecoli.simulation_data.SimulationDataEcoli"><em>SimulationDataEcoli</em></a>) ‚Äì simulation data</p></li>
<li><p><strong>stoich</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)"><em>dict</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)"><em>dict</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a><em>]</em><em>] </em><em>| </em><em>None</em>) ‚Äì <p>stoichiometry of metabolites for each reaction (if <code class="docutils literal notranslate"><span class="pre">None</span></code>,
data is loaded from <code class="docutils literal notranslate"><span class="pre">raw_data</span></code> and <code class="docutils literal notranslate"><span class="pre">sim_data</span></code>):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">{</span><span class="n">reaction</span> <span class="n">ID</span><span class="p">:</span> <span class="p">{</span><span class="n">metabolite</span> <span class="n">ID</span> <span class="k">with</span> <span class="n">location</span> <span class="n">tag</span><span class="p">:</span> <span class="n">stoichiometry</span><span class="p">}}</span>
</pre></div>
</div>
</p></li>
<li><p><strong>catalysts</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)"><em>dict</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.13)"><em>list</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>]</em><em>] </em><em>| </em><em>None</em>) ‚Äì <p>enzyme catalysts for each reaction with known catalysts,
likely a subset of reactions in <code class="docutils literal notranslate"><span class="pre">stoich</span></code> (if <code class="docutils literal notranslate"><span class="pre">None</span></code>, data
is loaded from <code class="docutils literal notranslate"><span class="pre">raw_data</span></code> and <code class="docutils literal notranslate"><span class="pre">sim_data</span></code>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">{</span><span class="n">reaction</span> <span class="n">ID</span><span class="p">:</span> <span class="n">enzyme</span> <span class="n">IDs</span> <span class="k">with</span> <span class="n">location</span> <span class="n">tag</span><span class="p">}</span>
</pre></div>
</div>
</p></li>
<li><p><strong>known_metabolites</strong> (<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Set" title="(in Python v3.13)"><em>Set</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>] </em><em>| </em><em>None</em>) ‚Äì metabolites with known concentrations</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><p>Valid kinetic constraints for each reaction/enzyme pair:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">{(</span><span class="n">reaction</span> <span class="n">ID</span><span class="p">,</span> <span class="n">enzyme</span> <span class="k">with</span> <span class="n">location</span> <span class="n">tag</span><span class="p">):</span> <span class="p">{</span>
        <span class="s1">&#39;kcat&#39;</span><span class="p">:</span> <span class="n">kcat</span> <span class="n">values</span> <span class="p">(</span><span class="nb">list</span><span class="p">[</span><span class="nb">float</span><span class="p">]),</span>
        <span class="s1">&#39;saturation&#39;</span><span class="p">:</span> <span class="n">saturation</span> <span class="n">equations</span> <span class="p">(</span><span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">])</span>
<span class="p">}}</span>
</pre></div>
</div>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)">dict</a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)">tuple</a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)">str</a>, <a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)">str</a>], <a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)">dict</a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)">str</a>, <a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.13)">list</a>[<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Any" title="(in Python v3.13)"><em>Any</em></a>]]]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.extract_reactions">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">extract_reactions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">raw_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/process/metabolism.html#Metabolism.extract_reactions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.extract_reactions" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>Extracts reaction data from raw_data to build metabolism reaction
network with stoichiometry, reversibility and enzyme catalysts.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>raw_data</strong> (<a class="reference internal" href="reconstruction.ecoli.knowledge_base_raw.html#reconstruction.ecoli.knowledge_base_raw.KnowledgeBaseEcoli" title="reconstruction.ecoli.knowledge_base_raw.KnowledgeBaseEcoli"><em>KnowledgeBaseEcoli</em></a>) ‚Äì knowledge base data</p></li>
<li><p><strong>sim_data</strong> (<a class="reference internal" href="reconstruction.ecoli.simulation_data.html#reconstruction.ecoli.simulation_data.SimulationDataEcoli" title="reconstruction.ecoli.simulation_data.SimulationDataEcoli"><em>SimulationDataEcoli</em></a>) ‚Äì simulation data</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><p>5-element tuple containing</p>
<blockquote>
<div><ul>
<li><p>base_rxn_ids: list of base reaction IDs from which reaction
IDs were derived from</p></li>
<li><p>reaction_stoich: stoichiometry of metabolites for each reaction:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">{</span><span class="n">reaction</span> <span class="n">ID</span><span class="p">:</span> <span class="p">{</span><span class="n">metabolite</span> <span class="n">ID</span> <span class="k">with</span> <span class="n">location</span> <span class="n">tag</span><span class="p">:</span> <span class="n">stoichiometry</span><span class="p">}}</span>
</pre></div>
</div>
</li>
<li><p>reversible_reactions: reaction IDs for reactions that have a
reverse complement, does not have reverse tag</p></li>
<li><p>reaction_catalysts: enzyme catalysts for each reaction with known
catalysts, likely a subset of reactions in stoich:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">{</span><span class="n">reaction</span> <span class="n">ID</span><span class="p">:</span> <span class="n">enzyme</span> <span class="n">IDs</span> <span class="k">with</span> <span class="n">location</span> <span class="n">tag</span><span class="p">}</span>
</pre></div>
</div>
</li>
<li><p>rxn_id_to_base_rxn_id: mapping from reaction IDs to the IDs of
the base reactions they were derived from:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">{</span><span class="n">reaction</span> <span class="n">ID</span><span class="p">:</span> <span class="n">base</span> <span class="n">ID</span><span class="p">}</span>
</pre></div>
</div>
</li>
</ul>
</div></blockquote>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)">tuple</a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.13)">list</a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)">str</a>], <a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)">dict</a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)">str</a>, <a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)">dict</a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)">str</a>, <a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)">int</a>]], <a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.13)">list</a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)">str</a>], <a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)">dict</a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)">str</a>, <a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.13)">list</a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)">str</a>]], <a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)">dict</a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)">str</a>, <a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)">str</a>]]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.get_aa_to_transporters_mapping_data">
<span class="sig-name descname"><span class="pre">get_aa_to_transporters_mapping_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">export</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/process/metabolism.html#Metabolism.get_aa_to_transporters_mapping_data"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.get_aa_to_transporters_mapping_data" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>Creates a dictionary that maps amino acids with their transporters.
Based on this dictionary, it creates a correlation matrix with rows
as AA and columns as transporters.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sim_data</strong> (<a class="reference internal" href="reconstruction.ecoli.simulation_data.html#reconstruction.ecoli.simulation_data.SimulationDataEcoli" title="reconstruction.ecoli.simulation_data.SimulationDataEcoli"><em>SimulationDataEcoli</em></a>) ‚Äì simulation data</p></li>
<li><p><strong>export</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a>) ‚Äì if True, the parameters calculated are for mechanistic
export instead of uptake</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><p>3-element tuple containing</p>
<blockquote>
<div><ul class="simple">
<li><p>aa_to_transporters: dictonary that maps aa to
transporters involved in transport reactions</p></li>
<li><p>aa_to_transporters_matrix: correlation matrix.
Columns correspond to transporter enzymes and rows to
amino acids</p></li>
<li><p>aa_transporters_names: names of all transporters</p></li>
</ul>
</div></blockquote>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)">tuple</a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)">dict</a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)">str</a>, <a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.13)">list</a>], <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)">tuple</a>[<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)">int</a>, ‚Ä¶], <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.dtype.html#numpy.dtype" title="(in NumPy v2.3)"><em>dtype</em></a>[<em>float64</em>]], <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)">tuple</a>[<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)">int</a>, ‚Ä¶], <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.dtype.html#numpy.dtype" title="(in NumPy v2.3)"><em>dtype</em></a>[<a class="reference external" href="https://numpy.org/doc/stable/reference/arrays.scalars.html#numpy.str_" title="(in NumPy v2.3)"><em>str_</em></a>]]]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.get_amino_acid_conc_conversion">
<span class="sig-name descname"><span class="pre">get_amino_acid_conc_conversion</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">conc_units</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/process/metabolism.html#Metabolism.get_amino_acid_conc_conversion"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.get_amino_acid_conc_conversion" title="Link to this definition">ÔÉÅ</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.get_kinetic_constraints">
<span class="sig-name descname"><span class="pre">get_kinetic_constraints</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">enzymes</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">substrates</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/process/metabolism.html#Metabolism.get_kinetic_constraints"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.get_kinetic_constraints" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>Allows for dynamic code generation for kinetic constraint calculation
for use in Metabolism process. Inputs should be unitless but the order
of magnitude should match the kinetics parameters (umol/L/s).</p>
<p>If trying to pickle sim_data object after function has been called,
_compiled_enzymes and _compiled_saturation might not be able to be pickled.
See __getstate__(), __setstate__() comments on PR 111 to address.</p>
<p>Returns np.array of floats of the kinetic constraint target for each
reaction with kinetic parameters</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>enzymes</strong> (<em>Unum</em>) ‚Äì concentrations of enzymes associated with kinetic
constraints (mol / volume units)</p></li>
<li><p><strong>substrates</strong> (<em>Unum</em>) ‚Äì concentrations of substrates associated with kinetic
constraints (mol / volume units)</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Array of dimensions (n reactions, 3) where each row contains the
min, mean and max kinetic constraints for each reaction with kinetic
constraints (mol / volume / time units)</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><em>Unum</em></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.get_pathway_enzyme_counts_per_aa">
<span class="sig-name descname"><span class="pre">get_pathway_enzyme_counts_per_aa</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">enzyme_counts</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/process/metabolism.html#Metabolism.get_pathway_enzyme_counts_per_aa"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.get_pathway_enzyme_counts_per_aa" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>Get the counts of enzymes for forward and reverse reactions in the
amino acid synthesis network based on all of the enzymes used in the
network.  Useful to get the counts to pass to amino_acid_synthesis()
from counts based on self.aa_enzymes.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>enzyme_counts</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a>) ‚Äì counts of all enzymes in the amino acid network</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt>2-element tuple containing</dt><dd><ul class="simple">
<li><p>counts_per_aa_fwd: counts of enzymes for the forward reaction
for each amino acid</p></li>
<li><p>counts_per_aa_rev: counts of enzymes for the reverse reaction
for each amino acid</p></li>
</ul>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Tuple" title="(in Python v3.13)"><em>Tuple</em></a>[<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a>, <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.match_reaction">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">match_reaction</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">stoich</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">catalysts</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rxn_to_match</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">enz</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mets</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">direction</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/process/metabolism.html#Metabolism.match_reaction"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.match_reaction" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>Matches a given reaction (rxn_to_match) to reactions that exist in
stoich given that enz is known to catalyze the reaction and mets are
reactants in the reaction. Can perform a fuzzy reaction match since
rxn_to_match just needs to be part of the actual reaction name to match
specific instances of a reaction.
(eg. rxn_to_match=‚ÄùALCOHOL-DEHYDROG-GENERIC-RXN‚Äù can match
‚ÄúALCOHOL-DEHYDROG-GENERIC-RXN-ETOH/NAD//ACETALD/NADH/PROTON.30.‚Äù).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>stoich</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)"><em>dict</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)"><em>dict</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a><em>]</em><em>]</em>) ‚Äì <p>stoichiometry of metabolites for each reaction:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">{</span><span class="n">reaction</span> <span class="n">ID</span><span class="p">:</span> <span class="p">{</span><span class="n">metabolite</span> <span class="n">ID</span> <span class="k">with</span> <span class="n">location</span> <span class="n">tag</span><span class="p">:</span> <span class="n">stoichiometry</span><span class="p">}}</span>
</pre></div>
</div>
</p></li>
<li><p><strong>catalysts</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)"><em>dict</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.13)"><em>list</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>]</em><em>]</em>) ‚Äì <p>enzyme catalysts for each reaction with known catalysts,
likely a subset of reactions in stoich:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">{</span><span class="n">reaction</span> <span class="n">ID</span><span class="p">:</span> <span class="n">enzyme</span> <span class="n">IDs</span> <span class="k">with</span> <span class="n">location</span> <span class="n">tag</span><span class="p">}</span>
</pre></div>
</div>
</p></li>
<li><p><strong>rxn_to_match</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a>) ‚Äì reaction ID from kinetics to match to existing reactions</p></li>
<li><p><strong>enz</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a>) ‚Äì enzyme ID with location tag</p></li>
<li><p><strong>mets</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.13)"><em>list</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>]</em>) ‚Äì metabolite IDs with no location tag from kinetics</p></li>
<li><p><strong>direction</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em> | </em><em>None</em>) ‚Äì reaction directionality, <code class="docutils literal notranslate"><span class="pre">'forward'</span></code> or <code class="docutils literal notranslate"><span class="pre">'reverse'</span></code>
or <code class="docutils literal notranslate"><span class="pre">None</span></code></p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Matched reaction IDs in stoich</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.13)">list</a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)">str</a>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.set_mechanistic_export_constants">
<span class="sig-name descname"><span class="pre">set_mechanistic_export_constants</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_specs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">basal_container</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/process/metabolism.html#Metabolism.set_mechanistic_export_constants"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.set_mechanistic_export_constants" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>Calls get_aa_to_transporters_mapping_data() for AA export, which calculates
the total amount of export transporter counts per AA. Kcats are calculated using
the same exchange rates as for uptake and transporter counts. Missing KMs are calculated
based on present KMs. This is done by calculating the average factor for
KMs compared to estimated concentration (av_factor = sum(KM / concentration) / n_aa_with_kms).
** KM = av_factor * concentration</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sim_data</strong> (<a class="reference internal" href="reconstruction.ecoli.simulation_data.html#reconstruction.ecoli.simulation_data.SimulationDataEcoli" title="reconstruction.ecoli.simulation_data.SimulationDataEcoli"><em>SimulationDataEcoli</em></a>) ‚Äì simulation data</p></li>
<li><p><strong>cell_specs</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)"><em>dict</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)"><em>dict</em></a><em>]</em>) ‚Äì mapping from condition to calculated cell properties</p></li>
<li><p><strong>basal_container</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a>) ‚Äì average initial bulk molecule counts in the basal
condition (structured Numpy array, see <a class="reference internal" href="../../../stores.html#bulk"><span class="std std-ref">Bulk Molecules</span></a>)</p></li>
</ul>
</dd>
</dl>
<dl class="simple">
<dt>Attributes set:</dt><dd><ul class="simple">
<li><p><a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.aa_to_exporters" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.aa_to_exporters"><code class="xref py py-attr docutils literal notranslate"><span class="pre">aa_to_exporters</span></code></a></p></li>
<li><p><a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.aa_to_exporters_matrix" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.aa_to_exporters_matrix"><code class="xref py py-attr docutils literal notranslate"><span class="pre">aa_to_exporters_matrix</span></code></a></p></li>
<li><p><a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.aa_exporter_names" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.aa_exporter_names"><code class="xref py py-attr docutils literal notranslate"><span class="pre">aa_exporter_names</span></code></a></p></li>
<li><p><a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.aa_export_kms" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.aa_export_kms"><code class="xref py py-attr docutils literal notranslate"><span class="pre">aa_export_kms</span></code></a></p></li>
<li><p><a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.export_kcats_per_aa" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.export_kcats_per_aa"><code class="xref py py-attr docutils literal notranslate"><span class="pre">export_kcats_per_aa</span></code></a></p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.set_mechanistic_supply_constants">
<span class="sig-name descname"><span class="pre">set_mechanistic_supply_constants</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_specs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">basal_container</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">with_aa_container</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/process/metabolism.html#Metabolism.set_mechanistic_supply_constants"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.set_mechanistic_supply_constants" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>Sets constants to determine amino acid supply during translation.  Used
with amino_acid_synthesis() and amino_acid_import() during simulations
but supply can alternatively be determined phenomologically.  This
approach is more detailed and should better respond to environmental
changes and perturbations but has more variability related to gene
expression and regulation.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sim_data</strong> (<a class="reference internal" href="reconstruction.ecoli.simulation_data.html#reconstruction.ecoli.simulation_data.SimulationDataEcoli" title="reconstruction.ecoli.simulation_data.SimulationDataEcoli"><em>SimulationDataEcoli</em></a>) ‚Äì simulation data</p></li>
<li><p><strong>cell_specs</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)"><em>dict</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)"><em>dict</em></a><em>]</em>) ‚Äì mapping from condition to calculated cell properties</p></li>
<li><p><strong>basal_container</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a>) ‚Äì average initial bulk molecule counts in the basal
condition (structured Numpy array, see <a class="reference internal" href="../../../stores.html#bulk"><span class="std std-ref">Bulk Molecules</span></a>)</p></li>
<li><p><strong>with_aa_container</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a>) ‚Äì average initial bulk molecule counts in the
<code class="docutils literal notranslate"><span class="pre">with_aa</span></code> condition</p></li>
</ul>
</dd>
</dl>
<dl class="simple">
<dt>Sets class attributes:</dt><dd><ul class="simple">
<li><p><a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.aa_enzymes" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.aa_enzymes"><code class="xref py py-attr docutils literal notranslate"><span class="pre">aa_enzymes</span></code></a></p></li>
<li><p><a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.aa_kcats_fwd" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.aa_kcats_fwd"><code class="xref py py-attr docutils literal notranslate"><span class="pre">aa_kcats_fwd</span></code></a></p></li>
<li><p><a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.aa_kcats_rev" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.aa_kcats_rev"><code class="xref py py-attr docutils literal notranslate"><span class="pre">aa_kcats_rev</span></code></a></p></li>
<li><p><a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.aa_kis" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.aa_kis"><code class="xref py py-attr docutils literal notranslate"><span class="pre">aa_kis</span></code></a></p></li>
<li><p><a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.aa_upstream_kms" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.aa_upstream_kms"><code class="xref py py-attr docutils literal notranslate"><span class="pre">aa_upstream_kms</span></code></a></p></li>
<li><p><a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.aa_reverse_kms" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.aa_reverse_kms"><code class="xref py py-attr docutils literal notranslate"><span class="pre">aa_reverse_kms</span></code></a></p></li>
<li><p><a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.aa_upstream_mapping" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.aa_upstream_mapping"><code class="xref py py-attr docutils literal notranslate"><span class="pre">aa_upstream_mapping</span></code></a></p></li>
<li><p><a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.enzyme_to_amino_acid" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.enzyme_to_amino_acid"><code class="xref py py-attr docutils literal notranslate"><span class="pre">enzyme_to_amino_acid</span></code></a></p></li>
<li><p><a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.aa_forward_stoich" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.aa_forward_stoich"><code class="xref py py-attr docutils literal notranslate"><span class="pre">aa_forward_stoich</span></code></a></p></li>
<li><p><a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.aa_reverse_stoich" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.aa_reverse_stoich"><code class="xref py py-attr docutils literal notranslate"><span class="pre">aa_reverse_stoich</span></code></a></p></li>
<li><p><a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.aa_import_kis" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.aa_import_kis"><code class="xref py py-attr docutils literal notranslate"><span class="pre">aa_import_kis</span></code></a></p></li>
<li><p><a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.specific_import_rates" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.specific_import_rates"><code class="xref py py-attr docutils literal notranslate"><span class="pre">specific_import_rates</span></code></a></p></li>
<li><p><a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.max_specific_import_rates" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.max_specific_import_rates"><code class="xref py py-attr docutils literal notranslate"><span class="pre">max_specific_import_rates</span></code></a></p></li>
</ul>
</dd>
</dl>
<p>Assumptions:</p>
<blockquote>
<div><ul class="simple">
<li><p>Only one reaction is limiting in an amino acid pathway (typically
the first and one with KI) and the kcat for forward or reverse
directions will apply to all enzymes that can catalyze that step</p></li>
<li><p>kcat for reverse and degradation reactions is the same (each amino
acid only has reverse or degradation at this point but that could
change with modifications to the amino_acid_pathways flat file)</p></li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.set_mechanistic_uptake_constants">
<span class="sig-name descname"><span class="pre">set_mechanistic_uptake_constants</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_specs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">with_aa_container</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/process/metabolism.html#Metabolism.set_mechanistic_uptake_constants"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.set_mechanistic_uptake_constants" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>Based on the matrix calculated in get_aa_to_transporters_mapping_data(),
we calculate the total amount of transporter counts per AA.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sim_data</strong> (<a class="reference internal" href="reconstruction.ecoli.simulation_data.html#reconstruction.ecoli.simulation_data.SimulationDataEcoli" title="reconstruction.ecoli.simulation_data.SimulationDataEcoli"><em>SimulationDataEcoli</em></a>) ‚Äì simulation data</p></li>
<li><p><strong>cell_specs</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)"><em>dict</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)"><em>dict</em></a><em>]</em>) ‚Äì mapping from condition to calculated cell properties</p></li>
<li><p><strong>with_aa_container</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a>) ‚Äì average initial bulk molecule counts in the
<code class="docutils literal notranslate"><span class="pre">with_aa</span></code> condition (structured Numpy array, see <a class="reference internal" href="../../../stores.html#bulk"><span class="std std-ref">Bulk Molecules</span></a>)</p></li>
</ul>
</dd>
</dl>
<dl class="simple">
<dt>Attributes set:</dt><dd><ul class="simple">
<li><p><a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.aa_to_importers" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.aa_to_importers"><code class="xref py py-attr docutils literal notranslate"><span class="pre">aa_to_importers</span></code></a></p></li>
<li><p><a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.aa_to_importers_matrix" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.aa_to_importers_matrix"><code class="xref py py-attr docutils literal notranslate"><span class="pre">aa_to_importers_matrix</span></code></a></p></li>
<li><p><a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.aa_importer_names" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.aa_importer_names"><code class="xref py py-attr docutils literal notranslate"><span class="pre">aa_importer_names</span></code></a></p></li>
<li><p><a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.import_kcats_per_aa" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.import_kcats_per_aa"><code class="xref py py-attr docutils literal notranslate"><span class="pre">import_kcats_per_aa</span></code></a></p></li>
<li><p><a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.export_kcats_per_aa" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.export_kcats_per_aa"><code class="xref py py-attr docutils literal notranslate"><span class="pre">export_kcats_per_aa</span></code></a></p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.set_phenomological_supply_constants">
<span class="sig-name descname"><span class="pre">set_phenomological_supply_constants</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/process/metabolism.html#Metabolism.set_phenomological_supply_constants"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.set_phenomological_supply_constants" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>Sets constants to determine amino acid supply during translation.  Used
with aa_supply_scaling() during simulations but supply can
alternatively be determined mechanistically.  This approach may require
manually adjusting constants (fraction_supply_inhibited and
fraction_supply_exported) but has less variability related to gene
expression and regulation.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>sim_data</strong> (<a class="reference internal" href="reconstruction.ecoli.simulation_data.html#reconstruction.ecoli.simulation_data.SimulationDataEcoli" title="reconstruction.ecoli.simulation_data.SimulationDataEcoli"><em>SimulationDataEcoli</em></a>) ‚Äì simulation data</p>
</dd>
</dl>
<dl class="simple">
<dt>Attributes set:</dt><dd><ul class="simple">
<li><p><a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.KI_aa_synthesis" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.KI_aa_synthesis"><code class="xref py py-attr docutils literal notranslate"><span class="pre">KI_aa_synthesis</span></code></a></p></li>
<li><p><a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.KM_aa_export" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.KM_aa_export"><code class="xref py py-attr docutils literal notranslate"><span class="pre">KM_aa_export</span></code></a></p></li>
<li><p><a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.fraction_supply_rate" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.fraction_supply_rate"><code class="xref py py-attr docutils literal notranslate"><span class="pre">fraction_supply_rate</span></code></a></p></li>
<li><p><a class="reference internal" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.fraction_import_rate" title="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.fraction_import_rate"><code class="xref py py-attr docutils literal notranslate"><span class="pre">fraction_import_rate</span></code></a></p></li>
</ul>
</dd>
<dt>Assumptions:</dt><dd><ul class="simple">
<li><p>Each internal amino acid concentration in ‚Äòminimal_plus_amino_acids‚Äô
media is not lower than in ‚Äòminimal‚Äô media</p></li>
</ul>
</dd>
<dt>TODO (Travis):</dt><dd><p>Better handling of concentration assumption</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.temperature_adjusted_kcat">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">temperature_adjusted_kcat</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">kcat</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">temp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/process/metabolism.html#Metabolism.temperature_adjusted_kcat"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.temperature_adjusted_kcat" title="Link to this definition">ÔÉÅ</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>kcat</strong> (<em>Unum</em>) ‚Äì enzyme turnover number(s) (1 / time)</p></li>
<li><p><strong>temp</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.13)"><em>float</em></a><em> | </em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a>) ‚Äì temperature of measurement, defaults to 25 if ‚Äò‚Äô</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Temperature adjusted kcat values, in units of 1/s</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.3)"><em>ndarray</em></a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)">tuple</a>[<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)">int</a>, ‚Ä¶], <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.dtype.html#numpy.dtype" title="(in NumPy v2.3)"><em>dtype</em></a>[<em>float64</em>]]</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.amino_acid_export_jit">
<span class="sig-prename descclassname"><span class="pre">reconstruction.ecoli.dataclasses.process.metabolism.</span></span><span class="sig-name descname"><span class="pre">amino_acid_export_jit</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">aa_transporters_counts</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">aa_conc</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mechanistic_uptake</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">aa_to_exporters_matrix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">export_kcats_per_aa</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">aa_export_kms</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/process/metabolism.html#amino_acid_export_jit"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.amino_acid_export_jit" title="Link to this definition">ÔÉÅ</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.amino_acid_import_jit">
<span class="sig-prename descclassname"><span class="pre">reconstruction.ecoli.dataclasses.process.metabolism.</span></span><span class="sig-name descname"><span class="pre">amino_acid_import_jit</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">aa_in_media</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dry_mass</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">internal_aa_conc</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">aa_transporters_counts</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mechanistic_uptake</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">aa_import_kis</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">aa_to_importers_matrix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">import_kcats_per_aa</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_specific_import_rates</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/process/metabolism.html#amino_acid_import_jit"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.amino_acid_import_jit" title="Link to this definition">ÔÉÅ</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.amino_acid_synthesis_jit">
<span class="sig-prename descclassname"><span class="pre">reconstruction.ecoli.dataclasses.process.metabolism.</span></span><span class="sig-name descname"><span class="pre">amino_acid_synthesis_jit</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">counts_per_aa_fwd</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">counts_per_aa_rev</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">aa_conc</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">aa_upstream_kms</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">aa_kis</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">aa_reverse_kms</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">aa_degradation_kms</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">aa_forward_stoich</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">aa_kcats_fwd</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">aa_reverse_stoich</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">aa_kcats_rev</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/process/metabolism.html#amino_acid_synthesis_jit"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.amino_acid_synthesis_jit" title="Link to this definition">ÔÉÅ</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.np_apply_along_axis">
<span class="sig-prename descclassname"><span class="pre">reconstruction.ecoli.dataclasses.process.metabolism.</span></span><span class="sig-name descname"><span class="pre">np_apply_along_axis</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">func1d</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">axis</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">arr</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/process/metabolism.html#np_apply_along_axis"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.np_apply_along_axis" title="Link to this definition">ÔÉÅ</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="reconstruction.ecoli.dataclasses.process.metabolism.np_prod">
<span class="sig-prename descclassname"><span class="pre">reconstruction.ecoli.dataclasses.process.metabolism.</span></span><span class="sig-name descname"><span class="pre">np_prod</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">array</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">axis</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/process/metabolism.html#np_prod"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#reconstruction.ecoli.dataclasses.process.metabolism.np_prod" title="Link to this definition">ÔÉÅ</a></dt>
<dd></dd></dl>

</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="reconstruction.ecoli.dataclasses.process.equilibrium.html" class="btn btn-neutral float-left" title="reconstruction.ecoli.dataclasses.process.equilibrium" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="reconstruction.ecoli.dataclasses.process.process.html" class="btn btn-neutral float-right" title="reconstruction.ecoli.dataclasses.process.process" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2021-2025, The Vivarium E. coli Authors.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>