

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>validation.ecoli.validation_data &mdash; Vivarium E. coli 1.0.0 documentation</title>
      <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=b86133f3" />
      <link rel="stylesheet" type="text/css" href="../../../_static/css/theme.css?v=e59714d7" />

  
      <script src="../../../_static/jquery.js?v=5d32c60e"></script>
      <script src="../../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../../../_static/documentation_options.js?v=8d563738"></script>
      <script src="../../../_static/doctools.js?v=9bcbadda"></script>
      <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
      <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="validation.ecoli.validation_data_raw" href="validation.ecoli.validation_data_raw.html" />
    <link rel="prev" title="validation.ecoli.scripts.rnaDegRates" href="validation.ecoli.scripts.rnaDegRates.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../index.html" class="icon icon-home">
            Vivarium E. coli
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../../stores.html">Stores</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../processes.html">Processes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../composites.html">Composites</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../experiments.html">Experiments</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../workflows.html">Workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../output.html">Output</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../tutorial.html">Tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../docs.html">Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../hpc.html">HPC Clusters</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../gcloud.html">Google Cloud</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../ci.html">Continuous Integration</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../pycharm.html">PyCharm Setup</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../diffs.html">Model Differences From wcEcoli</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../../api_ref.html">API Reference</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../ecoli/ecoli.html"><code class="docutils literal notranslate"><span class="pre">ecoli</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="../reconstruction/reconstruction.html"><code class="docutils literal notranslate"><span class="pre">reconstruction</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="../runscripts/runscripts.html"><code class="docutils literal notranslate"><span class="pre">runscripts</span></code></a></li>
<li class="toctree-l2 current"><a class="reference internal" href="validation.html"><code class="docutils literal notranslate"><span class="pre">validation</span></code></a><ul class="current">
<li class="toctree-l3 current"><a class="reference internal" href="validation.ecoli.html"><code class="docutils literal notranslate"><span class="pre">validation.ecoli</span></code></a><ul class="current">
<li class="toctree-l4"><a class="reference internal" href="validation.ecoli.scripts.html"><code class="docutils literal notranslate"><span class="pre">validation.ecoli.scripts</span></code></a></li>
<li class="toctree-l4 current"><a class="current reference internal" href="#"><code class="docutils literal notranslate"><span class="pre">validation.ecoli.validation_data</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="validation.ecoli.validation_data_raw.html"><code class="docutils literal notranslate"><span class="pre">validation.ecoli.validation_data_raw</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../wholecell/wholecell.html"><code class="docutils literal notranslate"><span class="pre">wholecell</span></code></a></li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">Vivarium E. coli</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../api_ref.html">API Reference</a></li>
          <li class="breadcrumb-item"><a href="validation.html"><code class="docutils literal notranslate"><span class="pre">validation</span></code></a></li>
          <li class="breadcrumb-item"><a href="validation.ecoli.html"><code class="docutils literal notranslate"><span class="pre">validation.ecoli</span></code></a></li>
      <li class="breadcrumb-item active"><code class="docutils literal notranslate"><span class="pre">validation.ecoli.validation_data</span></code></li>
      <li class="wy-breadcrumbs-aside">
            <a href="../../../_sources/reference/api/validation/validation.ecoli.validation_data.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="module-validation.ecoli.validation_data">
<span id="validation-ecoli-validation-data"></span><h1><code class="docutils literal notranslate"><span class="pre">validation.ecoli.validation_data</span></code><a class="headerlink" href="#module-validation.ecoli.validation_data" title="Link to this heading"></a></h1>
<p>ValidationData for Ecoli</p>
<p>Raw data processed into forms convienent for validation and analysis</p>
<dl class="py class">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.Constants">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">validation.ecoli.validation_data.</span></span><span class="sig-name descname"><span class="pre">Constants</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">raw_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/constants.html#Constants"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.Constants" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.Constants._build_constants">
<span class="sig-name descname"><span class="pre">_build_constants</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">raw_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/constants.html#Constants._build_constants"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.Constants._build_constants" title="Link to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.EssentialGenes">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">validation.ecoli.validation_data.</span></span><span class="sig-name descname"><span class="pre">EssentialGenes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">validation_data_raw</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/validation/ecoli/validation_data.html#EssentialGenes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.EssentialGenes" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.EssentialGenes._load_essential_genes">
<span class="sig-name descname"><span class="pre">_load_essential_genes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">validation_data_raw</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/validation/ecoli/validation_data.html#EssentialGenes._load_essential_genes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.EssentialGenes._load_essential_genes" title="Link to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.GeneFunctions">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">validation.ecoli.validation_data.</span></span><span class="sig-name descname"><span class="pre">GeneFunctions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">validation_data_raw</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/validation/ecoli/validation_data.html#GeneFunctions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.GeneFunctions" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.GeneFunctions._loadGeneFunctions">
<span class="sig-name descname"><span class="pre">_loadGeneFunctions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">validation_data_raw</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/validation/ecoli/validation_data.html#GeneFunctions._loadGeneFunctions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.GeneFunctions._loadGeneFunctions" title="Link to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.GetterFunctions">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">validation.ecoli.validation_data.</span></span><span class="sig-name descname"><span class="pre">GetterFunctions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">raw_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/getter_functions.html#GetterFunctions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.GetterFunctions" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>getterFunctions</p>
<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.GetterFunctions._build_all_masses">
<span class="sig-name descname"><span class="pre">_build_all_masses</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">raw_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/getter_functions.html#GetterFunctions._build_all_masses"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.GetterFunctions._build_all_masses" title="Link to this definition"></a></dt>
<dd><p>Builds dictionary of molecular weights keyed with the IDs of molecules.
Depending on the type of the molecule, weights are either pulled
directly from raw data, or dynamically calculated from the existing
data.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.GetterFunctions._build_compartments">
<span class="sig-name descname"><span class="pre">_build_compartments</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">raw_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/getter_functions.html#GetterFunctions._build_compartments"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.GetterFunctions._build_compartments" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.GetterFunctions._build_full_chromosome_mass">
<span class="sig-name descname"><span class="pre">_build_full_chromosome_mass</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">raw_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/getter_functions.html#GetterFunctions._build_full_chromosome_mass"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.GetterFunctions._build_full_chromosome_mass" title="Link to this definition"></a></dt>
<dd><p>Calculates the mass of the full chromosome from its sequence and the
weigths of polymerized dNTPs.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.GetterFunctions._build_genomic_coordinates">
<span class="sig-name descname"><span class="pre">_build_genomic_coordinates</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">raw_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/getter_functions.html#GetterFunctions._build_genomic_coordinates"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.GetterFunctions._build_genomic_coordinates" title="Link to this definition"></a></dt>
<dd><p>Builds a dictionary of genomic coordinates of DNA sites. Keys are the
IDs of the sites, and the values are tuples of the left-end coordinate
and the right-end coordinate of the site. Sites whose types are included
in IGNORED_DNA_SITE_TYPES are ignored.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.GetterFunctions._build_modified_protein_masses">
<span class="sig-name descname"><span class="pre">_build_modified_protein_masses</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">raw_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/getter_functions.html#GetterFunctions._build_modified_protein_masses"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.GetterFunctions._build_modified_protein_masses" title="Link to this definition"></a></dt>
<dd><p>Builds dictionary of molecular weights of modified proteins keyed with
the molecule IDs. Molecular weights are calculated from the
stoichiometries of the modification reactions.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.GetterFunctions._build_modified_rna_masses">
<span class="sig-name descname"><span class="pre">_build_modified_rna_masses</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">raw_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/getter_functions.html#GetterFunctions._build_modified_rna_masses"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.GetterFunctions._build_modified_rna_masses" title="Link to this definition"></a></dt>
<dd><p>Builds dictionary of molecular weights of modified RNAs keyed with
the molecule IDs. Molecular weights are calculated from the
stoichiometries of the modification reactions.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.GetterFunctions._build_polymerized_subunit_masses">
<span class="sig-name descname"><span class="pre">_build_polymerized_subunit_masses</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/getter_functions.html#GetterFunctions._build_polymerized_subunit_masses"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.GetterFunctions._build_polymerized_subunit_masses" title="Link to this definition"></a></dt>
<dd><p>Builds dictionary of molecular weights of polymerized subunits (NTPs,
dNTPs, and amino acids) by subtracting the end weights from the weights
of original subunits.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.GetterFunctions._build_protein_complex_compartments">
<span class="sig-name descname"><span class="pre">_build_protein_complex_compartments</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">raw_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">all_compartments_sorted</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/getter_functions.html#GetterFunctions._build_protein_complex_compartments"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.GetterFunctions._build_protein_complex_compartments" title="Link to this definition"></a></dt>
<dd><p>Builds dictionary of compartment tags for protein complexes keyed with
the molecule IDs. Each complex is assigned to the compartment that
contains a subunit of the complex and has the highest complexation
priority. Compartment tags of subunits that are also protein complexes
are determined recursively. Compartments of metabolite subunits are not
considered since metabolites are assigned to all compartments that are
being modeled.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.GetterFunctions._build_protein_complex_masses">
<span class="sig-name descname"><span class="pre">_build_protein_complex_masses</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">raw_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/getter_functions.html#GetterFunctions._build_protein_complex_masses"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.GetterFunctions._build_protein_complex_masses" title="Link to this definition"></a></dt>
<dd><p>Builds dictionary of molecular weights of protein complexes keyed with
the molecule IDs. Molecular weights are calculated from the
stoichiometries of the complexation/equilibrium reactions. For
complexes whose subunits are also complexes, the molecular weights are
calculated recursively.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.GetterFunctions._build_protein_masses">
<span class="sig-name descname"><span class="pre">_build_protein_masses</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">raw_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/getter_functions.html#GetterFunctions._build_protein_masses"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.GetterFunctions._build_protein_masses" title="Link to this definition"></a></dt>
<dd><p>Builds dictionary of molecular weights of protein monomers keyed with
the protein IDs. Molecular weights are calculated from the protein
sequence and the weights of polymerized amino acids.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.GetterFunctions._build_protein_sequences">
<span class="sig-name descname"><span class="pre">_build_protein_sequences</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">raw_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/getter_functions.html#GetterFunctions._build_protein_sequences"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.GetterFunctions._build_protein_sequences" title="Link to this definition"></a></dt>
<dd><p>Builds the amino acid sequences of each protein monomer using sequences
in raw_data.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.GetterFunctions._build_rna_masses">
<span class="sig-name descname"><span class="pre">_build_rna_masses</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">raw_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/getter_functions.html#GetterFunctions._build_rna_masses"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.GetterFunctions._build_rna_masses" title="Link to this definition"></a></dt>
<dd><p>Builds dictionary of molecular weights of RNAs keyed with the RNA IDs.
Molecular weights are calculated from the RNA sequence and the weights
of polymerized NTPs.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.GetterFunctions._build_rna_sequences">
<span class="sig-name descname"><span class="pre">_build_rna_sequences</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">raw_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/getter_functions.html#GetterFunctions._build_rna_sequences"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.GetterFunctions._build_rna_sequences" title="Link to this definition"></a></dt>
<dd><p>Builds nucleotide sequences of each transcription unit using the genome
sequence and the left and right end positions.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.GetterFunctions._build_sequences">
<span class="sig-name descname"><span class="pre">_build_sequences</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">raw_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/getter_functions.html#GetterFunctions._build_sequences"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.GetterFunctions._build_sequences" title="Link to this definition"></a></dt>
<dd><p>Builds sequences of RNAs and proteins.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.GetterFunctions._build_submass_array">
<span class="sig-name descname"><span class="pre">_build_submass_array</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mw</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">submass_name</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/getter_functions.html#GetterFunctions._build_submass_array"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.GetterFunctions._build_submass_array" title="Link to this definition"></a></dt>
<dd><p>Converts a scalar molecular weight value to an array of submasses,
given the name of the submass category that the molecule belongs to.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>mw</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)"><em>float</em></a>)</p></li>
<li><p><strong>submass_name</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>ndarray</em></a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.14)">tuple</a>[<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)">int</a>, …], <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.dtype.html#numpy.dtype" title="(in NumPy v2.4)"><em>dtype</em></a>[<em>float64</em>]]</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.GetterFunctions._compartment_tag">
<span class="sig-name descname"><span class="pre">_compartment_tag</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">re.compile('\\[[a-z]]')</span></em><a class="headerlink" href="#validation.ecoli.validation_data.GetterFunctions._compartment_tag" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.GetterFunctions.get_all_valid_molecules">
<span class="sig-name descname"><span class="pre">get_all_valid_molecules</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/getter_functions.html#GetterFunctions.get_all_valid_molecules"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.GetterFunctions.get_all_valid_molecules" title="Link to this definition"></a></dt>
<dd><p>Returns a list of all molecule IDs with assigned submass arrays and
compartments.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)">list</a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)">str</a>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.GetterFunctions.get_compartment">
<span class="sig-name descname"><span class="pre">get_compartment</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mol_id</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/getter_functions.html#GetterFunctions.get_compartment"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.GetterFunctions.get_compartment" title="Link to this definition"></a></dt>
<dd><p>Returns the list of one-letter codes for the compartments that the
molecule with the given ID can exist in.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>mol_id</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)">list</a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)">list</a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)">str</a>]]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.GetterFunctions.get_compartment_tag">
<span class="sig-name descname"><span class="pre">get_compartment_tag</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">id_</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/getter_functions.html#GetterFunctions.get_compartment_tag"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.GetterFunctions.get_compartment_tag" title="Link to this definition"></a></dt>
<dd><p>Look up a molecule id and return a compartment suffix tag like ‘[c]’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>id_</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)">str</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.GetterFunctions.get_compartments">
<span class="sig-name descname"><span class="pre">get_compartments</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ids</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/getter_functions.html#GetterFunctions.get_compartments"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.GetterFunctions.get_compartments" title="Link to this definition"></a></dt>
<dd><p>Returns a list of the list of one-letter codes for the compartments
that each of the molecules with the given IDs can exist in.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>ids</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a><em>] </em><em>| </em><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>ndarray</em></a>)</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)">list</a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)">list</a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)">str</a>]]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.GetterFunctions.get_genomic_coordinates">
<span class="sig-name descname"><span class="pre">get_genomic_coordinates</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">site_id</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/getter_functions.html#GetterFunctions.get_genomic_coordinates"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.GetterFunctions.get_genomic_coordinates" title="Link to this definition"></a></dt>
<dd><p>Returns the genomic coordinates of the left and right ends of a DNA site
given the ID of the site.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>site_id</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.14)">tuple</a>[<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)">int</a>, <a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)">int</a>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.GetterFunctions.get_mass">
<span class="sig-name descname"><span class="pre">get_mass</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mol_id</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/getter_functions.html#GetterFunctions.get_mass"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.GetterFunctions.get_mass" title="Link to this definition"></a></dt>
<dd><p>Return the total mass of the molecule with a given ID.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>mol_id</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.GetterFunctions.get_masses">
<span class="sig-name descname"><span class="pre">get_masses</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mol_ids</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/getter_functions.html#GetterFunctions.get_masses"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.GetterFunctions.get_masses" title="Link to this definition"></a></dt>
<dd><p>Return an array of total masses of the molecules with the given IDs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>mol_ids</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a><em> | </em><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>ndarray</em></a>)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.GetterFunctions.get_miscrnas_with_singleton_tus">
<span class="sig-name descname"><span class="pre">get_miscrnas_with_singleton_tus</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/getter_functions.html#GetterFunctions.get_miscrnas_with_singleton_tus"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.GetterFunctions.get_miscrnas_with_singleton_tus" title="Link to this definition"></a></dt>
<dd><p>Returns a list of all miscRNA IDs with corresponding single-gene
transcription units.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)">list</a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)">str</a>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.GetterFunctions.get_sequences">
<span class="sig-name descname"><span class="pre">get_sequences</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ids</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/getter_functions.html#GetterFunctions.get_sequences"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.GetterFunctions.get_sequences" title="Link to this definition"></a></dt>
<dd><p>Return a list of sequences of the molecules with the given IDs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>ids</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a><em>] </em><em>| </em><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>ndarray</em></a>)</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)">list</a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)">str</a>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.GetterFunctions.get_singleton_tu_id">
<span class="sig-name descname"><span class="pre">get_singleton_tu_id</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rna_id</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/getter_functions.html#GetterFunctions.get_singleton_tu_id"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.GetterFunctions.get_singleton_tu_id" title="Link to this definition"></a></dt>
<dd><p>Returns the ID of the single-gene transcription unit corresponding to
the given miscRNA ID, if such a transcription unit exists. This is
necessary to replace some references to miscRNA IDs in complexation or
equilibrium reactions with their corresponding TU IDs, which are the
molecules that are actually transcribed.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>rna_id</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)">str</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.GetterFunctions.get_submass_array">
<span class="sig-name descname"><span class="pre">get_submass_array</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mol_id</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/getter_functions.html#GetterFunctions.get_submass_array"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.GetterFunctions.get_submass_array" title="Link to this definition"></a></dt>
<dd><p>Return the submass array of the molecule with a given ID.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>mol_id</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.GetterFunctions.is_valid_molecule">
<span class="sig-name descname"><span class="pre">is_valid_molecule</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mol_id</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/getter_functions.html#GetterFunctions.is_valid_molecule"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.GetterFunctions.is_valid_molecule" title="Link to this definition"></a></dt>
<dd><p>Returns True if the molecule with the given ID is a valid molecule (has
both a submass array and a compartment tag).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>mol_id</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)">bool</a></p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.GrowthRateParameters">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">validation.ecoli.validation_data.</span></span><span class="sig-name descname"><span class="pre">GrowthRateParameters</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">raw_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/growth_rate_dependent_parameters.html#GrowthRateParameters"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.GrowthRateParameters" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.GrowthRateParameters._fit_ribosome_elongation_rate_by_ppgpp">
<span class="sig-name descname"><span class="pre">_fit_ribosome_elongation_rate_by_ppgpp</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ppgpp</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rate</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/growth_rate_dependent_parameters.html#GrowthRateParameters._fit_ribosome_elongation_rate_by_ppgpp"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.GrowthRateParameters._fit_ribosome_elongation_rate_by_ppgpp" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.GrowthRateParameters.get_fraction_active_ribosome">
<span class="sig-name descname"><span class="pre">get_fraction_active_ribosome</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">doubling_time</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/growth_rate_dependent_parameters.html#GrowthRateParameters.get_fraction_active_ribosome"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.GrowthRateParameters.get_fraction_active_ribosome" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.GrowthRateParameters.get_fraction_active_rnap">
<span class="sig-name descname"><span class="pre">get_fraction_active_rnap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">doubling_time</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/growth_rate_dependent_parameters.html#GrowthRateParameters.get_fraction_active_rnap"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.GrowthRateParameters.get_fraction_active_rnap" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.GrowthRateParameters.get_ppGpp_conc">
<span class="sig-name descname"><span class="pre">get_ppGpp_conc</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">doubling_time</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/growth_rate_dependent_parameters.html#GrowthRateParameters.get_ppGpp_conc"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.GrowthRateParameters.get_ppGpp_conc" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.GrowthRateParameters.get_ribosome_elongation_rate">
<span class="sig-name descname"><span class="pre">get_ribosome_elongation_rate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">doubling_time</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/growth_rate_dependent_parameters.html#GrowthRateParameters.get_ribosome_elongation_rate"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.GrowthRateParameters.get_ribosome_elongation_rate" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.GrowthRateParameters.get_ribosome_elongation_rate_by_ppgpp">
<span class="sig-name descname"><span class="pre">get_ribosome_elongation_rate_by_ppgpp</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ppgpp</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_rate</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/growth_rate_dependent_parameters.html#GrowthRateParameters.get_ribosome_elongation_rate_by_ppgpp"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.GrowthRateParameters.get_ribosome_elongation_rate_by_ppgpp" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.GrowthRateParameters.get_rnap_elongation_rate">
<span class="sig-name descname"><span class="pre">get_rnap_elongation_rate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">doubling_time</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/growth_rate_dependent_parameters.html#GrowthRateParameters.get_rnap_elongation_rate"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.GrowthRateParameters.get_rnap_elongation_rate" title="Link to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.InternalState">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">validation.ecoli.validation_data.</span></span><span class="sig-name descname"><span class="pre">InternalState</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">raw_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/state/internal_state.html#InternalState"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.InternalState" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>Internal State</p>
<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.InternalState._build_bulk_molecule_specs">
<span class="sig-name descname"><span class="pre">_build_bulk_molecule_specs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">molecule_ids</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/state/internal_state.html#InternalState._build_bulk_molecule_specs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.InternalState._build_bulk_molecule_specs" title="Link to this definition"></a></dt>
<dd><p>Builds a list of molecule IDs with compartment tags and a corresponding
array of molecular masses to add to the bulk state.
:param molecule_ids: List of molecule IDs w/o compartment tags
:type molecule_ids: List[str]</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><dl class="simple">
<dt>List of molecule IDs</dt><dd><p>with compartment tags</p>
</dd>
<dt>masses (np.ndarray): Array of molecular masses divided into</dt><dd><p>submasses</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>molecule_ids_with_compartments (List[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)">str</a>])</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.InternalState._build_bulk_molecules">
<span class="sig-name descname"><span class="pre">_build_bulk_molecules</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/state/internal_state.html#InternalState._build_bulk_molecules"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.InternalState._build_bulk_molecules" title="Link to this definition"></a></dt>
<dd><p>Add data (IDs and mass) for all classes of bulk molecules.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.InternalState._build_compartments">
<span class="sig-name descname"><span class="pre">_build_compartments</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">raw_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/state/internal_state.html#InternalState._build_compartments"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.InternalState._build_compartments" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.InternalState._build_unique_molecules">
<span class="sig-name descname"><span class="pre">_build_unique_molecules</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/state/internal_state.html#InternalState._build_unique_molecules"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.InternalState._build_unique_molecules" title="Link to this definition"></a></dt>
<dd><p>Add data (name, mass, and attribute data structure) for all classes of
unique molecules.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.KnowledgeBaseEcoli">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">validation.ecoli.validation_data.</span></span><span class="sig-name descname"><span class="pre">KnowledgeBaseEcoli</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">operons_on</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">remove_rrna_operons</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">remove_rrff</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stable_rrna</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">new_genes_option</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'off'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/knowledge_base_raw.html#KnowledgeBaseEcoli"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.KnowledgeBaseEcoli" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>operons_on</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)</p></li>
<li><p><strong>remove_rrna_operons</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)</p></li>
<li><p><strong>remove_rrff</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)</p></li>
<li><p><strong>stable_rrna</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)</p></li>
<li><p><strong>new_genes_option</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.KnowledgeBaseEcoli._check_new_gene_ids">
<span class="sig-name descname"><span class="pre">_check_new_gene_ids</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">nested_attr</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/knowledge_base_raw.html#KnowledgeBaseEcoli._check_new_gene_ids"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.KnowledgeBaseEcoli._check_new_gene_ids" title="Link to this definition"></a></dt>
<dd><p>Check to ensure each new gene, RNA, and protein id starts with NG.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.KnowledgeBaseEcoli._get_new_gene_sequence">
<span class="sig-name descname"><span class="pre">_get_new_gene_sequence</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">nested_attr</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/knowledge_base_raw.html#KnowledgeBaseEcoli._get_new_gene_sequence"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.KnowledgeBaseEcoli._get_new_gene_sequence" title="Link to this definition"></a></dt>
<dd><p>Determine genome sequnce for insertion using the sequences and
relative locations of the new genes.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.KnowledgeBaseEcoli._join_data">
<span class="sig-name descname"><span class="pre">_join_data</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/knowledge_base_raw.html#KnowledgeBaseEcoli._join_data"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.KnowledgeBaseEcoli._join_data" title="Link to this definition"></a></dt>
<dd><p>Add rows that are specified in additional files. Data will only be added
if all the loaded columns from both datasets match.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.KnowledgeBaseEcoli._load_parameters">
<span class="sig-name descname"><span class="pre">_load_parameters</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dir_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file_name</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/knowledge_base_raw.html#KnowledgeBaseEcoli._load_parameters"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.KnowledgeBaseEcoli._load_parameters" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.KnowledgeBaseEcoli._load_sequence">
<span class="sig-name descname"><span class="pre">_load_sequence</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_path</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/knowledge_base_raw.html#KnowledgeBaseEcoli._load_sequence"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.KnowledgeBaseEcoli._load_sequence" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.KnowledgeBaseEcoli._load_tsv">
<span class="sig-name descname"><span class="pre">_load_tsv</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dir_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file_name</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/knowledge_base_raw.html#KnowledgeBaseEcoli._load_tsv"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.KnowledgeBaseEcoli._load_tsv" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.KnowledgeBaseEcoli._modify_data">
<span class="sig-name descname"><span class="pre">_modify_data</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/knowledge_base_raw.html#KnowledgeBaseEcoli._modify_data"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.KnowledgeBaseEcoli._modify_data" title="Link to this definition"></a></dt>
<dd><p>Modify entires in rows that are specified to be modified. Rows must be
identified by their entries in the first column (usually the ID column).</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.KnowledgeBaseEcoli._prune_data">
<span class="sig-name descname"><span class="pre">_prune_data</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/knowledge_base_raw.html#KnowledgeBaseEcoli._prune_data"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.KnowledgeBaseEcoli._prune_data" title="Link to this definition"></a></dt>
<dd><p>Remove rows that are specified to be removed. Data will only be removed
if all data in a row in the file specifying rows to be removed matches
the same columns in the raw data file.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.KnowledgeBaseEcoli._update_gene_insertion_location">
<span class="sig-name descname"><span class="pre">_update_gene_insertion_location</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">nested_attr</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/knowledge_base_raw.html#KnowledgeBaseEcoli._update_gene_insertion_location"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.KnowledgeBaseEcoli._update_gene_insertion_location" title="Link to this definition"></a></dt>
<dd><p>Update insertion location of new genes to prevent conflicts.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.KnowledgeBaseEcoli._update_gene_locations">
<span class="sig-name descname"><span class="pre">_update_gene_locations</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">nested_attr</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">insert_pos</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/knowledge_base_raw.html#KnowledgeBaseEcoli._update_gene_locations"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.KnowledgeBaseEcoli._update_gene_locations" title="Link to this definition"></a></dt>
<dd><p>Modify positions of original genes based upon the insertion location
of new genes. Returns end position of the gene insertion.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.KnowledgeBaseEcoli._update_global_coordinates">
<span class="sig-name descname"><span class="pre">_update_global_coordinates</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">insert_pos</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">insert_len</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/knowledge_base_raw.html#KnowledgeBaseEcoli._update_global_coordinates"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.KnowledgeBaseEcoli._update_global_coordinates" title="Link to this definition"></a></dt>
<dd><p>Updates the left and right end positions for all elements in data if
their positions will be impacted by the new gene insertion.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> – Data attribute to update</p></li>
<li><p><strong>insert_pos</strong> – Location of new gene insertion</p></li>
<li><p><strong>insert_len</strong> – Length of new gene insertion</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.Mass">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">validation.ecoli.validation_data.</span></span><span class="sig-name descname"><span class="pre">Mass</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">raw_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/growth_rate_dependent_parameters.html#Mass"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.Mass" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.Mass._build_CD_periods">
<span class="sig-name descname"><span class="pre">_build_CD_periods</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">raw_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/growth_rate_dependent_parameters.html#Mass._build_CD_periods"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.Mass._build_CD_periods" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.Mass._build_constants">
<span class="sig-name descname"><span class="pre">_build_constants</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">raw_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/growth_rate_dependent_parameters.html#Mass._build_constants"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.Mass._build_constants" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.Mass._build_dependent_constants">
<span class="sig-name descname"><span class="pre">_build_dependent_constants</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/growth_rate_dependent_parameters.html#Mass._build_dependent_constants"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.Mass._build_dependent_constants" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.Mass._build_submasses">
<span class="sig-name descname"><span class="pre">_build_submasses</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">raw_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/growth_rate_dependent_parameters.html#Mass._build_submasses"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.Mass._build_submasses" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.Mass._build_trna_data">
<span class="sig-name descname"><span class="pre">_build_trna_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">raw_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/growth_rate_dependent_parameters.html#Mass._build_trna_data"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.Mass._build_trna_data" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.Mass._calculateGrowthRateDependentDnaMass">
<span class="sig-name descname"><span class="pre">_calculateGrowthRateDependentDnaMass</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">doubling_time</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/growth_rate_dependent_parameters.html#Mass._calculateGrowthRateDependentDnaMass"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.Mass._calculateGrowthRateDependentDnaMass" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.Mass._clipTau_d">
<span class="sig-name descname"><span class="pre">_clipTau_d</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">doubling_time</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/growth_rate_dependent_parameters.html#Mass._clipTau_d"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.Mass._clipTau_d" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.Mass._getFitParameters">
<span class="sig-name descname"><span class="pre">_getFitParameters</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dry_mass_composition</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mass_fraction_name</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/growth_rate_dependent_parameters.html#Mass._getFitParameters"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.Mass._getFitParameters" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.Mass.getBiomassAsConcentrations">
<span class="sig-name descname"><span class="pre">getBiomassAsConcentrations</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">doubling_time</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rp_ratio</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/growth_rate_dependent_parameters.html#Mass.getBiomassAsConcentrations"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.Mass.getBiomassAsConcentrations" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.Mass.get_avg_cell_dry_mass">
<span class="sig-name descname"><span class="pre">get_avg_cell_dry_mass</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">doubling_time</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/growth_rate_dependent_parameters.html#Mass.get_avg_cell_dry_mass"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.Mass.get_avg_cell_dry_mass" title="Link to this definition"></a></dt>
<dd><p>Gets the dry mass for an average cell at the given doubling time.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>doubling_time</strong> (<em>Unum</em>) – expected doubling time</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>average cell dry mass</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><em>Unum</em></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.Mass.get_basal_rna_fractions">
<span class="sig-name descname"><span class="pre">get_basal_rna_fractions</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/growth_rate_dependent_parameters.html#Mass.get_basal_rna_fractions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.Mass.get_basal_rna_fractions" title="Link to this definition"></a></dt>
<dd><p>Measured RNA subgroup mass fractions. Fractions should change in other
conditions with growth rate (see transcription.get_rna_fractions()).</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.Mass.get_component_masses">
<span class="sig-name descname"><span class="pre">get_component_masses</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">doubling_time</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/growth_rate_dependent_parameters.html#Mass.get_component_masses"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.Mass.get_component_masses" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.Mass.get_dna_critical_mass">
<span class="sig-name descname"><span class="pre">get_dna_critical_mass</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">doubling_time</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/growth_rate_dependent_parameters.html#Mass.get_dna_critical_mass"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.Mass.get_dna_critical_mass" title="Link to this definition"></a></dt>
<dd><p>Returns the critical mass for replication initiation.  Faster growing
cells maintain a consistent initiation mass but slower growing cells
are smaller and will never reach this mass so it needs to be adjusted
lower for them.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>doubling_time</strong> (<em>Unum</em>) – expected doubling time of cell</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Critical mass for DNA replication initiation</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><em>Unum</em></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.Mass.get_mass_fractions">
<span class="sig-name descname"><span class="pre">get_mass_fractions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">doubling_time</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/growth_rate_dependent_parameters.html#Mass.get_mass_fractions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.Mass.get_mass_fractions" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.Mass.get_mass_fractions_from_rna_protein_ratio">
<span class="sig-name descname"><span class="pre">get_mass_fractions_from_rna_protein_ratio</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ratio</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/growth_rate_dependent_parameters.html#Mass.get_mass_fractions_from_rna_protein_ratio"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.Mass.get_mass_fractions_from_rna_protein_ratio" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.Mass.get_trna_distribution">
<span class="sig-name descname"><span class="pre">get_trna_distribution</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">doubling_time</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/growth_rate_dependent_parameters.html#Mass.get_trna_distribution"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.Mass.get_trna_distribution" title="Link to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.MoleculeGroups">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">validation.ecoli.validation_data.</span></span><span class="sig-name descname"><span class="pre">MoleculeGroups</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">raw_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/molecule_groups.html#MoleculeGroups"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.MoleculeGroups" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>Helper class to extract molecule IDs of “special” groups of molecules. All
values returned are lists of strings.</p>
<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.MoleculeGroups._build_molecule_groups">
<span class="sig-name descname"><span class="pre">_build_molecule_groups</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">raw_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/molecule_groups.html#MoleculeGroups._build_molecule_groups"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.MoleculeGroups._build_molecule_groups" title="Link to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.MoleculeIds">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">validation.ecoli.validation_data.</span></span><span class="sig-name descname"><span class="pre">MoleculeIds</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">raw_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/molecule_ids.html#MoleculeIds"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.MoleculeIds" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>Helper class to extract molecule IDs of “special” molecules. All values
returned are strings.</p>
<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.MoleculeIds._buildMoleculeIds">
<span class="sig-name descname"><span class="pre">_buildMoleculeIds</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/molecule_ids.html#MoleculeIds._buildMoleculeIds"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.MoleculeIds._buildMoleculeIds" title="Link to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.Process">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">validation.ecoli.validation_data.</span></span><span class="sig-name descname"><span class="pre">Process</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">raw_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/process/process.html#Process"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.Process" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.Protein">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">validation.ecoli.validation_data.</span></span><span class="sig-name descname"><span class="pre">Protein</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">validation_data_raw</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">knowledge_base_raw</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/validation/ecoli/validation_data.html#Protein"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.Protein" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.Protein._loadHouser2015Counts">
<span class="sig-name descname"><span class="pre">_loadHouser2015Counts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">validation_data_raw</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/validation/ecoli/validation_data.html#Protein._loadHouser2015Counts"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.Protein._loadHouser2015Counts" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.Protein._loadSchmidt2015Counts">
<span class="sig-name descname"><span class="pre">_loadSchmidt2015Counts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">validation_data_raw</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/validation/ecoli/validation_data.html#Protein._loadSchmidt2015Counts"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.Protein._loadSchmidt2015Counts" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.Protein._loadTaniguchi2010Counts">
<span class="sig-name descname"><span class="pre">_loadTaniguchi2010Counts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">validation_data_raw</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/validation/ecoli/validation_data.html#Protein._loadTaniguchi2010Counts"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.Protein._loadTaniguchi2010Counts" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.Protein._loadWisniewski2014Counts">
<span class="sig-name descname"><span class="pre">_loadWisniewski2014Counts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">validation_data_raw</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">knowledge_base_raw</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/validation/ecoli/validation_data.html#Protein._loadWisniewski2014Counts"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.Protein._loadWisniewski2014Counts" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.Protein._load_li">
<span class="sig-name descname"><span class="pre">_load_li</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">validation_data_raw</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/validation/ecoli/validation_data.html#Protein._load_li"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.Protein._load_li" title="Link to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.ReactionFlux">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">validation.ecoli.validation_data.</span></span><span class="sig-name descname"><span class="pre">ReactionFlux</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">validation_data_raw</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">knowledge_base_raw</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/validation/ecoli/validation_data.html#ReactionFlux"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.ReactionFlux" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.ReactionFlux._loadToya2010Fluxes">
<span class="sig-name descname"><span class="pre">_loadToya2010Fluxes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">validation_data_raw</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/validation/ecoli/validation_data.html#ReactionFlux._loadToya2010Fluxes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.ReactionFlux._loadToya2010Fluxes" title="Link to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.Relation">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">validation.ecoli.validation_data.</span></span><span class="sig-name descname"><span class="pre">Relation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">raw_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/relation.html#Relation"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.Relation" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.Relation._build_RNA_to_tf_mapping">
<span class="sig-name descname"><span class="pre">_build_RNA_to_tf_mapping</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">raw_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/relation.html#Relation._build_RNA_to_tf_mapping"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.Relation._build_RNA_to_tf_mapping" title="Link to this definition"></a></dt>
<dd><p>Builds a dictionary that maps RNA IDs to a list of all transcription
factor IDs that regulate the given RNA. All TFs that target any of the
constituent cistrons in the RNA are added to each list.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.Relation._build_cistron_to_monomer_mapping">
<span class="sig-name descname"><span class="pre">_build_cistron_to_monomer_mapping</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">raw_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/relation.html#Relation._build_cistron_to_monomer_mapping"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.Relation._build_cistron_to_monomer_mapping" title="Link to this definition"></a></dt>
<dd><p>Build a vector that can map vectors that describe a property for RNA
cistrons into a vector that describes the same property for the
corresponding monomers if used as an index array. Assumes that each
monomer maps to a single RNA cistron (A single RNA can map to multiple
monomers).</p>
<p>e.g.
monomer_property = RNA_cistron_property[</p>
<blockquote>
<div><p>sim_data.relation.cistron_to_monomer_mapping]</p>
</div></blockquote>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.Relation._build_monomer_to_mRNA_cistron_mapping">
<span class="sig-name descname"><span class="pre">_build_monomer_to_mRNA_cistron_mapping</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">raw_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/relation.html#Relation._build_monomer_to_mRNA_cistron_mapping"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.Relation._build_monomer_to_mRNA_cistron_mapping" title="Link to this definition"></a></dt>
<dd><p>Builds a sparse matrix that can map vectors that describe a property
for protein monomers into a vector that describes the same property for
the corresponding mRNA cistrons if multiplied to the right of the
original vector. The transformed property must be additive (i.e. if two
proteins map to the same cistron, the values given for the two proteins
are added to yield a value for the cistron).</p>
<p>The full matrix can be returned by calling
monomer_to_mRNA_cistron_mapping().</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.Relation._build_monomer_to_tu_mapping">
<span class="sig-name descname"><span class="pre">_build_monomer_to_tu_mapping</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">raw_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/relation.html#Relation._build_monomer_to_tu_mapping"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.Relation._build_monomer_to_tu_mapping" title="Link to this definition"></a></dt>
<dd><p>Builds a dictionary that maps monomer IDs to a list of all transcription
unit IDs that the monomer can be translated from.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.Relation._build_tf_to_RNA_mapping">
<span class="sig-name descname"><span class="pre">_build_tf_to_RNA_mapping</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">raw_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sim_data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/relation.html#Relation._build_tf_to_RNA_mapping"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.Relation._build_tf_to_RNA_mapping" title="Link to this definition"></a></dt>
<dd><p>Builds a dictionary that maps transcription factor IDs to a list of all
RNA IDs that are targeted by the given TF. All RNA transcription units
that contain any of the cistrons regulated by the TF are added to each
list.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.Relation.monomer_to_mRNA_cistron_mapping">
<span class="sig-name descname"><span class="pre">monomer_to_mRNA_cistron_mapping</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/reconstruction/ecoli/dataclasses/relation.html#Relation.monomer_to_mRNA_cistron_mapping"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.Relation.monomer_to_mRNA_cistron_mapping" title="Link to this definition"></a></dt>
<dd><p>Returns the full version of the sparse matrix built by
_build_monomer_to_mRNA_cistron_mapping().</p>
<p>e.g.
mRNA_property = sim_data.relation.monomer_to_mRNA_cistron_mapping().T.dot(</p>
<blockquote>
<div><p>monomer_property)</p>
</div></blockquote>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.ValidationDataEcoli">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">validation.ecoli.validation_data.</span></span><span class="sig-name descname"><span class="pre">ValidationDataEcoli</span></span><a class="reference internal" href="../../../_modules/validation/ecoli/validation_data.html#ValidationDataEcoli"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.ValidationDataEcoli" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.ValidationDataEcoli._add_amino_acid_growth_rates">
<span class="sig-name descname"><span class="pre">_add_amino_acid_growth_rates</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">validation_data_raw</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/validation/ecoli/validation_data.html#ValidationDataEcoli._add_amino_acid_growth_rates"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.ValidationDataEcoli._add_amino_acid_growth_rates" title="Link to this definition"></a></dt>
<dd><p>Loads growth rates with single amino acids supplemented in media.</p>
<dl>
<dt>amino_acid_media_growth_rates: dict with data from 4 replicates</dt><dd><dl>
<dt>{</dt><dd><p>media ID (str):
{</p>
<blockquote>
<div><p>‘mean’: mean max growth rate from 4 replicates (float with units per time)
‘std’: standard deviation for max growth rate from 4 replicates (float with units per time)</p>
</div></blockquote>
<p>}</p>
</dd>
</dl>
<p>}</p>
</dd>
<dt>amino_acid_media_dose_dependent_growth_rates: dict with data from single measurements from 4 concentrations</dt><dd><dl>
<dt>{</dt><dd><p>media ID (str):
{</p>
<blockquote>
<div><p>‘conc’: concentration of the amino acid in media (np.ndarray[float] with units mol/volume)
‘growth’: max growth rate corresponding to each media concentration (float with units per time)</p>
</div></blockquote>
<p>}</p>
</dd>
</dl>
<p>}</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="validation.ecoli.validation_data.ValidationDataEcoli.initialize">
<span class="sig-name descname"><span class="pre">initialize</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">validation_data_raw</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">knowledge_base_raw</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/validation/ecoli/validation_data.html#ValidationDataEcoli.initialize"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#validation.ecoli.validation_data.ValidationDataEcoli.initialize" title="Link to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="validation.ecoli.scripts.rnaDegRates.html" class="btn btn-neutral float-left" title="validation.ecoli.scripts.rnaDegRates" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="validation.ecoli.validation_data_raw.html" class="btn btn-neutral float-right" title="validation.ecoli.validation_data_raw" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2021-2026, The Vivarium E. coli Authors.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>