

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../../../../">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>reconstruction.ecoli.dataclasses.process.metabolism &mdash; Vivarium E. coli 1.0.0 documentation</title>
      <link rel="stylesheet" type="text/css" href="../../../../../_static/pygments.css?v=b86133f3" />
      <link rel="stylesheet" type="text/css" href="../../../../../_static/css/theme.css?v=e59714d7" />

  
      <script src="../../../../../_static/jquery.js?v=5d32c60e"></script>
      <script src="../../../../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../../../../../_static/documentation_options.js?v=8d563738"></script>
      <script src="../../../../../_static/doctools.js?v=9bcbadda"></script>
      <script src="../../../../../_static/sphinx_highlight.js?v=dc90522c"></script>
      <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../../../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../../../genindex.html" />
    <link rel="search" title="Search" href="../../../../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../../../index.html" class="icon icon-home">
            Vivarium E. coli
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../../../../stores.html">Stores</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../processes.html">Processes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../composites.html">Composites</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../experiments.html">Experiments</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../workflows.html">Workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../output.html">Output</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../tutorial.html">Tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../docs.html">Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../hpc.html">HPC Clusters</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../gcloud.html">Google Cloud</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../ci.html">Continuous Integration</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../pycharm.html">PyCharm Setup</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../diffs.html">Model Differences From wcEcoli</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../reference/api_ref.html">API Reference</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../../../index.html">Vivarium E. coli</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../../../index.html">Module code</a></li>
      <li class="breadcrumb-item active">reconstruction.ecoli.dataclasses.process.metabolism</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for reconstruction.ecoli.dataclasses.process.metabolism</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">SimulationData for metabolism process</span>

<span class="sd">TODO:</span>

<span class="sd">- improved estimate of ILE/LEU abundance or some external data point</span>
<span class="sd">- implement L1-norm minimization for AA concentrations</span>
<span class="sd">- find concentration for PI[c]</span>
<span class="sd">- add (d)NTP byproduct concentrations</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="c1"># mypy: disable-error-code=attr-defined</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">copy</span><span class="w"> </span><span class="kn">import</span> <span class="n">copy</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">itertools</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">re</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">typing</span><span class="w"> </span><span class="kn">import</span> <span class="p">(</span>
    <span class="n">Any</span><span class="p">,</span>
    <span class="n">Callable</span><span class="p">,</span>
    <span class="n">cast</span><span class="p">,</span>
    <span class="n">Iterable</span><span class="p">,</span>
    <span class="n">Optional</span><span class="p">,</span>
    <span class="n">Set</span><span class="p">,</span>
    <span class="n">Tuple</span><span class="p">,</span>
    <span class="n">Union</span><span class="p">,</span>
    <span class="n">TYPE_CHECKING</span><span class="p">,</span>
<span class="p">)</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">unum</span><span class="w"> </span><span class="kn">import</span> <span class="n">Unum</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">numba</span><span class="w"> </span><span class="kn">import</span> <span class="n">njit</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy.typing</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">npt</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">sympy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sp</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">sympy.parsing.sympy_parser</span><span class="w"> </span><span class="kn">import</span> <span class="n">parse_expr</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">ecoli.library.schema</span><span class="w"> </span><span class="kn">import</span> <span class="n">bulk_name_to_idx</span><span class="p">,</span> <span class="n">counts</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">reconstruction.ecoli.dataclasses.getter_functions</span><span class="w"> </span><span class="kn">import</span> <span class="p">(</span>
    <span class="n">UNDEFINED_COMPARTMENT_IDS_TO_ABBREVS</span><span class="p">,</span>
<span class="p">)</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">reconstruction.ecoli.knowledge_base_raw</span><span class="w"> </span><span class="kn">import</span> <span class="n">KnowledgeBaseEcoli</span>

<span class="k">if</span> <span class="n">TYPE_CHECKING</span><span class="p">:</span>
    <span class="kn">from</span><span class="w"> </span><span class="nn">reconstruction.ecoli.simulation_data</span><span class="w"> </span><span class="kn">import</span> <span class="n">SimulationDataEcoli</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">reconstruction.ecoli.dataclasses.constants</span><span class="w"> </span><span class="kn">import</span> <span class="n">Constants</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">wholecell.utils</span><span class="w"> </span><span class="kn">import</span> <span class="n">units</span>

<span class="n">KINETIC_CONSTRAINT_CONC_UNITS</span> <span class="o">=</span> <span class="n">units</span><span class="o">.</span><span class="n">umol</span> <span class="o">/</span> <span class="n">units</span><span class="o">.</span><span class="n">L</span>
<span class="n">K_CAT_UNITS</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">/</span> <span class="n">units</span><span class="o">.</span><span class="n">s</span>
<span class="sd">&quot;&quot;&quot;Units for k :sub:`cat` values&quot;&quot;&quot;</span>
<span class="n">METABOLITE_CONCENTRATION_UNITS</span> <span class="o">=</span> <span class="n">units</span><span class="o">.</span><span class="n">mol</span> <span class="o">/</span> <span class="n">units</span><span class="o">.</span><span class="n">L</span>
<span class="sd">&quot;&quot;&quot;Units for metabolite concentrations&quot;&quot;&quot;</span>
<span class="n">DRY_MASS_UNITS</span> <span class="o">=</span> <span class="n">units</span><span class="o">.</span><span class="n">fg</span>
<span class="sd">&quot;&quot;&quot;Units for dry mass&quot;&quot;&quot;</span>

<span class="n">USE_ALL_CONSTRAINTS</span> <span class="o">=</span> <span class="kc">False</span>  <span class="c1"># False will remove defined constraints from objective</span>

<span class="n">REVERSE_TAG</span> <span class="o">=</span> <span class="s2">&quot; (reverse)&quot;</span>
<span class="n">REVERSE_REACTION_ID</span> <span class="o">=</span> <span class="s2">&quot;{{}}</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">REVERSE_TAG</span><span class="p">)</span>
<span class="n">ENZYME_REACTION_ID</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">{}</span><span class="s2">__</span><span class="si">{}</span><span class="s2">&quot;</span>

<span class="n">VERBOSE</span> <span class="o">=</span> <span class="kc">False</span>


<div class="viewcode-block" id="InvalidReactionDirectionError">
<a class="viewcode-back" href="../../../../../reference/api/reconstruction/reconstruction.ecoli.dataclasses.process.metabolism.html#reconstruction.ecoli.dataclasses.process.metabolism.InvalidReactionDirectionError">[docs]</a>
<span class="k">class</span><span class="w"> </span><span class="nc">InvalidReactionDirectionError</span><span class="p">(</span><span class="ne">Exception</span><span class="p">):</span>
    <span class="k">pass</span></div>



<div class="viewcode-block" id="Metabolism">
<a class="viewcode-back" href="../../../../../reference/api/reconstruction/reconstruction.ecoli.dataclasses.process.metabolism.html#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism">[docs]</a>
<span class="k">class</span><span class="w"> </span><span class="nc">Metabolism</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Metabolism</span>

<span class="sd">    Args:</span>
<span class="sd">            raw_data: Raw data object</span>
<span class="sd">            sim_data: Simulation data object</span>

<span class="sd">    Attributes:</span>
<span class="sd">            solver (str): solver ID, should match a value in modular_fba.py,</span>
<span class="sd">                    set by :py:meth:`~._set_solver_values`</span>
<span class="sd">            kinetic_objective_weight (float): weighting for the kinetic objective,</span>
<span class="sd">                    1-weighting for the homeostatic objective, set by</span>
<span class="sd">                    :py:meth:`~._set_solver_values`</span>
<span class="sd">            kinetic_objective_weight_in_range (float): weighting for deviations</span>
<span class="sd">                    from the kinetic target within min and max ranges, set by</span>
<span class="sd">                    :py:meth:`~._set_solver_values`</span>
<span class="sd">            secretion_penalty_coeff (float): penalty on secretion fluxes, set by</span>
<span class="sd">                    :py:meth:`~._set_solver_values`</span>
<span class="sd">            metabolite_charge (dict[str, int]): mapping of metabolite IDs to charge,</span>
<span class="sd">                    set by :py:meth:`~._add_metabolite_charge`</span>

<span class="sd">            concentration_updates</span>
<span class="sd">            conc_dict (dict):</span>
<span class="sd">            nutrients_to_internal_conc (dict[str, dict[str, Unum]]):</span>

<span class="sd">            kinetic_constraint_reactions:</span>
<span class="sd">            kinetic_constraint_enzymes:</span>
<span class="sd">            kinetic_constraint_substrates:</span>
<span class="sd">            _kcats:</span>
<span class="sd">            _saturations:</span>
<span class="sd">            _enzymes:</span>
<span class="sd">            constraint_is_kcat_only:</span>
<span class="sd">            _compiled_enzymes:</span>
<span class="sd">            _compiled_saturation:</span>
<span class="sd">            reaction_stoich:</span>
<span class="sd">            maintenance_reaction:</span>
<span class="sd">            reaction_catalysts:</span>
<span class="sd">            catalyst_ids:</span>
<span class="sd">            reactions_with_catalyst:</span>
<span class="sd">            catalysis_matrix_I:</span>
<span class="sd">            catalysis_matrix_J:</span>
<span class="sd">            catalysis_matrix_V:</span>
<span class="sd">            use_all_constraints:</span>
<span class="sd">            constraints_to_disable:</span>
<span class="sd">            base_reaction_ids:</span>
<span class="sd">            reaction_id_to_base_reaction_id:</span>
<span class="sd">            amino_acid_export_kms:</span>

<span class="sd">            transport_reactions (list[str]): transport reaction IDs in the</span>
<span class="sd">                    metabolic network (includes reverse reactions and reactions</span>
<span class="sd">                    with kinetic constraints), set by</span>
<span class="sd">                    :py:meth:`~._build_transport_reactions`</span>
<span class="sd">            aa_synthesis_pathways (dict[str, dict]): data for allosteric</span>
<span class="sd">                    inhibition of amino acid pathways indexed by amino acid ID with</span>
<span class="sd">                    location tag and nested dictionary with the following keys::</span>

<span class="sd">                            {&#39;enzymes&#39; (str): limiting/regulated enzyme ID in synthesis</span>
<span class="sd">                                    pathway with location tag,</span>
<span class="sd">                            &#39;kcat_data&#39; (units.Unum): kcat associated with enzyme</span>
<span class="sd">                                    reaction with units of 1/time,</span>
<span class="sd">                            &#39;ki&#39; (Tuple[units.Unum, units.Unum]]): lower and upper</span>
<span class="sd">                                    limits of KI associated with enzyme reaction with units</span>
<span class="sd">                                    of mol/volume}</span>

<span class="sd">                    Set by :py:meth:`~._build_amino_acid_pathways`</span>
<span class="sd">            KI_aa_synthesis (numpy.ndarray[float]): KI for each AA for synthesis</span>
<span class="sd">                    portion of supply (in units of</span>
<span class="sd">                    :py:data:`~.METABOLITE_CONCENTRATION_UNITS`), set by</span>
<span class="sd">                    :py:meth:`~.set_phenomological_supply_constants`</span>
<span class="sd">            KM_aa_export (numpy.ndarray[float]): KM for each AA for export portion</span>
<span class="sd">                    of supply (in units of</span>
<span class="sd">                    :py:data:`~.METABOLITE_CONCENTRATION_UNITS`), set by</span>
<span class="sd">                    :py:meth:`~.set_phenomological_supply_constants`</span>
<span class="sd">            fraction_supply_rate (float): fraction of AA supply that comes from</span>
<span class="sd">                    a base synthesis rate, set by</span>
<span class="sd">                    :py:meth:`~.set_phenomological_supply_constants`</span>
<span class="sd">            fraction_import_rate (numpy.ndarray[float]): fraction of AA supply that</span>
<span class="sd">                    comes from AA import if nutrients are present, set by</span>
<span class="sd">                    :py:meth:`~.set_phenomological_supply_constants`</span>
<span class="sd">            ppgpp_synthesis_reaction (str): reaction ID for ppGpp synthesis</span>
<span class="sd">                    (catalyzed by RelA and SpoT), set by :py:meth:`~._build_ppgpp_reactions`</span>
<span class="sd">            ppgpp_degradation_reaction (str): reaction ID for ppGpp degradation</span>
<span class="sd">                    (catalyzed by SpoT), set by :py:meth:`~._build_ppgpp_reactions`</span>
<span class="sd">            ppgpp_reaction_names (list[str]): names of reaction involved in ppGpp,</span>
<span class="sd">                    set by :py:meth:`~._build_ppgpp_reactions`</span>
<span class="sd">            ppgpp_reaction_metabolites (list[str]): names of metabolites in</span>
<span class="sd">                    ppGpp reactions, set by :py:meth:`~._build_ppgpp_reactions`</span>
<span class="sd">            ppgpp_reaction_stoich (numpy.ndarray[int]): 2D array with metabolites</span>
<span class="sd">                    on rows and reactions on columns containing the stoichiometric</span>
<span class="sd">                    coefficient, set by :py:meth:`~._build_ppgpp_reactions`</span>
<span class="sd">            aa_to_exporters (dict[str, list]): dictonary that maps aa to</span>
<span class="sd">                    transporters involved in export reactions. Set by</span>
<span class="sd">                    :py:meth:`~.set_mechanistic_export_constants`.</span>
<span class="sd">            aa_to_exporters_matrix (numpy.ndarray[int]): correlation matrix.</span>
<span class="sd">                    Columns correspond to exporting enzymes and rows to amino acids.</span>
<span class="sd">                    Set by :py:meth:`~.set_mechanistic_export_constants`.</span>
<span class="sd">            aa_exporter_names (numpy.ndarray[str]): names of all exporters. Set by</span>
<span class="sd">                    :py:meth:`~.set_mechanistic_export_constants`.</span>
<span class="sd">            aa_export_kms (numpy.ndarray[float]): kms corresponding to generic</span>
<span class="sd">                    transport/enzyme reactions for each AA in concentration units.</span>
<span class="sd">                    Set by :py:meth:`~.set_mechanistic_export_constants`.</span>
<span class="sd">            export_kcats_per_aa (numpy.ndarray[float]): kcats corresponding to</span>
<span class="sd">                    generic export reactions for each AA. Units in counts/second.</span>
<span class="sd">                    Set by :py:meth:`~.set_mechanistic_export_constants` and</span>
<span class="sd">                    :py:meth:`~.set_mechanistic_export_constants`.</span>
<span class="sd">            aa_to_importers (dict[str, list]): dictonary that maps aa to</span>
<span class="sd">                    transporters involved in import reactions. Set by</span>
<span class="sd">                    :py:meth:`~.set_mechanistic_uptake_constants`.</span>
<span class="sd">            aa_to_importers_matrix (numpy.ndarray[int]): correlation matrix.</span>
<span class="sd">                    Columns correspond to importing enzymes and rows to amino acids.</span>
<span class="sd">                    Set by :py:meth:`~.set_mechanistic_export_constants`.</span>
<span class="sd">            aa_importer_names (numpy.ndarray[str]): names of all importers.</span>
<span class="sd">                    Set by :py:meth:`~.set_mechanistic_export_constants`.</span>
<span class="sd">            import_kcats_per_aa (numpy.ndarray[float]): kcats corresponding</span>
<span class="sd">                    to generic import reactions for each AA. Units in counts/second.</span>
<span class="sd">                    Set by :py:meth:`~.set_mechanistic_export_constants`.</span>
<span class="sd">            aa_enzymes (numpy.ndarray[str]): enzyme ID with location tag for each</span>
<span class="sd">                    enzyme that can catalyze an amino acid pathway with</span>
<span class="sd">                    :py:attr:`~.enzyme_to_amino_acid` mapping these to each amino acid.</span>
<span class="sd">                    Set by :py:meth:`~.set_mechanistic_supply_constants`.</span>
<span class="sd">            aa_kcats_fwd (numpy.ndarray[float]): forward kcat value for each</span>
<span class="sd">                    synthesis pathway in units of :py:data:`~.K_CAT_UNITS`, ordered by</span>
<span class="sd">                    amino acid molecule group. Set by</span>
<span class="sd">                    :py:meth:`~.set_mechanistic_supply_constants`.</span>
<span class="sd">            aa_kcats_rev (numpy.ndarray[float]): reverse kcat value for each</span>
<span class="sd">                    synthesis pathway in units of :py:data:`~.K_CAT_UNITS`, ordered by</span>
<span class="sd">                    amino acid molecule group. Set by</span>
<span class="sd">                    :py:meth:`~.set_mechanistic_supply_constants`.</span>
<span class="sd">            aa_kis (numpy.ndarray[float]): KI value for each synthesis pathway</span>
<span class="sd">                    in units of :py:data:`~.METABOLITE_CONCENTRATION_UNITS`, ordered</span>
<span class="sd">                    by amino acid molecule group. Will be inf if there is no inhibitory</span>
<span class="sd">                    control. Set by :py:meth:`~.set_mechanistic_supply_constants`.</span>
<span class="sd">            aa_upstream_kms (list[list[float]]): KM value associated with the</span>
<span class="sd">                    amino acid that feeds into each synthesis pathway in units of</span>
<span class="sd">                    :py:data:`~.METABOLITE_CONCENTRATION_UNITS`, ordered by amino</span>
<span class="sd">                    acid molecule group. Will be 0 if there is no upstream amino</span>
<span class="sd">                    acid considered. Set by</span>
<span class="sd">                    :py:meth:`~.set_mechanistic_supply_constants`.</span>
<span class="sd">            aa_reverse_kms (numpy.ndarray[float]): KM value associated with the</span>
<span class="sd">                    amino acid in each synthesis pathway in units of</span>
<span class="sd">                    :py:data:`~.METABOLITE_CONCENTRATION_UNITS`, ordered by amino acid</span>
<span class="sd">                    molecule group. Will be inf if the synthesis pathway is not</span>
<span class="sd">                    reversible. Set by :py:meth:`~.set_mechanistic_supply_constants`.</span>
<span class="sd">            aa_upstream_mapping (numpy.ndarray[int]): index of the upstream amino</span>
<span class="sd">                    acid that feeds into each synthesis pathway, ordered by amino</span>
<span class="sd">                    acid molecule group. Set by</span>
<span class="sd">                    :py:meth:`~.set_mechanistic_supply_constants`.</span>
<span class="sd">            enzyme_to_amino_acid (numpy.ndarray[float]): relationship mapping from</span>
<span class="sd">                    aa_enzymes to amino acids (n enzymes, m amino acids).  Will</span>
<span class="sd">                    contain a 1 if the enzyme associated with the row can catalyze</span>
<span class="sd">                    the pathway for the amino acid associated with the column.</span>
<span class="sd">                    Set by :py:meth:`~.set_mechanistic_supply_constants`.</span>
<span class="sd">            aa_forward_stoich (numpy.ndarray[float]): relationship mapping from</span>
<span class="sd">                    upstream amino acids to downstream amino acids (n upstream,</span>
<span class="sd">                    m downstream).  Will contain a -1 if the amino acid associated</span>
<span class="sd">                    with the row is required for synthesis of the amino acid</span>
<span class="sd">                    associated with the column. Set by</span>
<span class="sd">                    :py:meth:`~.set_mechanistic_supply_constants`.</span>
<span class="sd">            aa_reverse_stoich (numpy.ndarray[float]): relationship mapping from</span>
<span class="sd">                    upstream amino acids to downstream amino acids (n downstream,</span>
<span class="sd">                    m upstream).  Will contain a -1 if the amino acid associated</span>
<span class="sd">                    with the row is produced through a reverse reaction from</span>
<span class="sd">                    the amino acid associated with the column. Set by</span>
<span class="sd">                    :py:meth:`~.set_mechanistic_supply_constants`.</span>
<span class="sd">            aa_import_kis (numpy.ndarray[float]): inhibition constants for amino</span>
<span class="sd">                    acid import based on the internal amino acid concentration</span>
<span class="sd">            specific_import_rates (numpy.ndarray[float]): import rates expected</span>
<span class="sd">                    in rich media conditions for each amino acid normalized by dry</span>
<span class="sd">                    cell mass in units of :py:data:`~.K_CAT_UNITS` /</span>
<span class="sd">                    :py:data:`~.DRY_MASS_UNITS`, ordered by amino acid molecule group.</span>
<span class="sd">                    Set by :py:meth:`~.set_mechanistic_supply_constants`.</span>
<span class="sd">            max_specific_import_rates (numpy.ndarray[float]): max import rates</span>
<span class="sd">                    for each amino acid without including internal concentration</span>
<span class="sd">                    inhibition normalized by dry cell mass in units of</span>
<span class="sd">                    :py:data:`~.K_CAT_UNITS` / :py:data:`~.DRY_MASS_UNITS`, ordered by</span>
<span class="sd">                    amino acid molecule group. Set by</span>
<span class="sd">                    :py:meth:`~.set_mechanistic_supply_constants`.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">raw_data</span><span class="p">:</span> <span class="n">KnowledgeBaseEcoli</span><span class="p">,</span> <span class="n">sim_data</span><span class="p">:</span> <span class="s2">&quot;SimulationDataEcoli&quot;</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_set_solver_values</span><span class="p">(</span><span class="n">sim_data</span><span class="o">.</span><span class="n">constants</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_build_biomass</span><span class="p">(</span><span class="n">raw_data</span><span class="p">,</span> <span class="n">sim_data</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_build_metabolism</span><span class="p">(</span><span class="n">raw_data</span><span class="p">,</span> <span class="n">sim_data</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_build_ppgpp_reactions</span><span class="p">(</span><span class="n">raw_data</span><span class="p">,</span> <span class="n">sim_data</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_build_transport_reactions</span><span class="p">(</span><span class="n">raw_data</span><span class="p">,</span> <span class="n">sim_data</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_build_amino_acid_pathways</span><span class="p">(</span><span class="n">raw_data</span><span class="p">,</span> <span class="n">sim_data</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_add_metabolite_charge</span><span class="p">(</span><span class="n">raw_data</span><span class="p">)</span>

<div class="viewcode-block" id="Metabolism._add_metabolite_charge">
<a class="viewcode-back" href="../../../../../reference/api/reconstruction/reconstruction.ecoli.dataclasses.process.metabolism.html#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._add_metabolite_charge">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">_add_metabolite_charge</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">raw_data</span><span class="p">:</span> <span class="n">KnowledgeBaseEcoli</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Compiles all metabolite charges.</span>

<span class="sd">        Args:</span>
<span class="sd">                raw_data: Raw data object</span>

<span class="sd">        Attributes set:</span>
<span class="sd">                - :py:attr:`~.metabolite_charge`</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">metabolite_charge</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">met</span> <span class="ow">in</span> <span class="n">raw_data</span><span class="o">.</span><span class="n">metabolites</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">metabolite_charge</span><span class="p">[</span><span class="n">met</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]]</span> <span class="o">=</span> <span class="n">met</span><span class="p">[</span><span class="s2">&quot;molecular_charge&quot;</span><span class="p">]</span></div>


<div class="viewcode-block" id="Metabolism._set_solver_values">
<a class="viewcode-back" href="../../../../../reference/api/reconstruction/reconstruction.ecoli.dataclasses.process.metabolism.html#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._set_solver_values">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">_set_solver_values</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">constants</span><span class="p">:</span> <span class="n">Constants</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Sets values to be used in the FBA solver.</span>

<span class="sd">        Attributes set:</span>
<span class="sd">                - :py:attr:`~.solver`</span>
<span class="sd">                - :py:attr:`~.kinetic_objective_weight`</span>
<span class="sd">                - :py:attr:`~.secretion_penalty_coeff`</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">solver</span> <span class="o">=</span> <span class="s2">&quot;glpk-linear&quot;</span>
        <span class="k">if</span> <span class="s2">&quot;linear&quot;</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">solver</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">kinetic_objective_weight</span> <span class="o">=</span> <span class="p">(</span>
                <span class="n">constants</span><span class="o">.</span><span class="n">metabolism_kinetic_objective_weight_linear</span>
            <span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">kinetic_objective_weight</span> <span class="o">=</span> <span class="p">(</span>
                <span class="n">constants</span><span class="o">.</span><span class="n">metabolism_kinetic_objective_weight_quadratic</span>
            <span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">kinetic_objective_weight_in_range</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">constants</span><span class="o">.</span><span class="n">metabolism_kinetic_objective_weight_in_range</span>
        <span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">secretion_penalty_coeff</span> <span class="o">=</span> <span class="n">constants</span><span class="o">.</span><span class="n">secretion_penalty_coeff</span></div>


<div class="viewcode-block" id="Metabolism._build_biomass">
<a class="viewcode-back" href="../../../../../reference/api/reconstruction/reconstruction.ecoli.dataclasses.process.metabolism.html#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._build_biomass">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">_build_biomass</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">raw_data</span><span class="p">:</span> <span class="n">KnowledgeBaseEcoli</span><span class="p">,</span> <span class="n">sim_data</span><span class="p">:</span> <span class="s2">&quot;SimulationDataEcoli&quot;</span>
    <span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Calculates metabolite concentration targets.</span>

<span class="sd">        Args:</span>
<span class="sd">                raw_data: Raw data object</span>
<span class="sd">                sim_data: Simulation data object</span>

<span class="sd">        Attributes set:</span>
<span class="sd">                - :py:attr:`~.concentration_updates`</span>
<span class="sd">                - :py:attr:`~.conc_dict`</span>
<span class="sd">                - :py:attr:`~.nutrients_to_internal_conc`</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">wildtypeIDs</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">entry</span><span class="p">[</span><span class="s2">&quot;molecule id&quot;</span><span class="p">]</span> <span class="k">for</span> <span class="n">entry</span> <span class="ow">in</span> <span class="n">raw_data</span><span class="o">.</span><span class="n">biomass</span><span class="p">)</span>

        <span class="c1"># Create vector of metabolite target concentrations</span>

        <span class="c1"># Since the data only covers certain metabolites, we need to rationally</span>
        <span class="c1"># expand the dataset to include the other molecules in the biomass</span>
        <span class="c1"># function.</span>

        <span class="c1"># First, load in metabolites that do have concentrations, then assign</span>
        <span class="c1"># compartments according to those given in the biomass objective.  Or,</span>
        <span class="c1"># if there is no compartment, assign it to the cytoplasm.</span>

        <span class="n">concentration_sources</span> <span class="o">=</span> <span class="p">[</span>
            <span class="s2">&quot;Park Concentration&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Lempp Concentration&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Kochanowski Concentration&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Sander Concentration&quot;</span><span class="p">,</span>
        <span class="p">]</span>
        <span class="n">excluded</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;Park Concentration&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;GLT&quot;</span><span class="p">,</span>  <span class="c1"># Steady state concentration reached with tRNA charging is much lower than Park</span>
                <span class="s2">&quot;THR&quot;</span><span class="p">,</span>  <span class="c1"># Attenuation needs concentration to be lower to match validation data</span>
                <span class="s2">&quot;VAL&quot;</span><span class="p">,</span>  <span class="c1"># Synthesis pathway kcat needs concentration to be lower and closer to KI</span>
            <span class="p">},</span>
            <span class="s2">&quot;Lempp Concentration&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;ATP&quot;</span><span class="p">,</span>  <span class="c1"># TF binding does not solve with average concentration</span>
                <span class="s2">&quot;VAL&quot;</span><span class="p">,</span>  <span class="c1"># Synthesis pathway kcat needs concentration to be lower and closer to KI</span>
            <span class="p">},</span>
            <span class="s2">&quot;Kochanowski Concentration&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;ATP&quot;</span><span class="p">,</span>  <span class="c1"># TF binding does not solve with average concentration</span>
            <span class="p">},</span>
            <span class="s2">&quot;Sander Concentration&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;GLT&quot;</span><span class="p">,</span>  <span class="c1"># Steady state concentration reached with tRNA charging is much lower than Sander</span>
            <span class="p">},</span>
        <span class="p">}</span>
        <span class="n">metaboliteIDs</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">metaboliteConcentrations</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="n">wildtypeIDtoCompartment</span> <span class="o">=</span> <span class="p">{</span>
            <span class="n">wildtypeID</span><span class="p">[:</span><span class="o">-</span><span class="mi">3</span><span class="p">]:</span> <span class="n">wildtypeID</span><span class="p">[</span><span class="o">-</span><span class="mi">3</span><span class="p">:]</span> <span class="k">for</span> <span class="n">wildtypeID</span> <span class="ow">in</span> <span class="n">wildtypeIDs</span>
        <span class="p">}</span>  <span class="c1"># this assumes biomass reaction components only exist in a single compartment</span>

        <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">raw_data</span><span class="o">.</span><span class="n">metabolite_concentrations</span><span class="p">:</span>
            <span class="n">metabolite_id</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;Metabolite&quot;</span><span class="p">]</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">sim_data</span><span class="o">.</span><span class="n">getter</span><span class="o">.</span><span class="n">is_valid_molecule</span><span class="p">(</span><span class="n">metabolite_id</span><span class="p">):</span>
                <span class="k">if</span> <span class="n">VERBOSE</span><span class="p">:</span>
                    <span class="nb">print</span><span class="p">(</span>
                        <span class="s2">&quot;Metabolite concentration for unknown molecule: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                            <span class="n">metabolite_id</span>
                        <span class="p">)</span>
                    <span class="p">)</span>
                <span class="k">continue</span>

            <span class="c1"># Use average of both sources</span>
            <span class="c1"># TODO (Travis): geometric mean?</span>
            <span class="n">conc</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nanmean</span><span class="p">(</span>
                <span class="p">[</span>
                    <span class="n">row</span><span class="p">[</span><span class="n">source</span><span class="p">]</span><span class="o">.</span><span class="n">asNumber</span><span class="p">(</span><span class="n">METABOLITE_CONCENTRATION_UNITS</span><span class="p">)</span>
                    <span class="k">for</span> <span class="n">source</span> <span class="ow">in</span> <span class="n">concentration_sources</span>
                    <span class="k">if</span> <span class="n">metabolite_id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">excluded</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">source</span><span class="p">,</span> <span class="nb">set</span><span class="p">())</span>
                <span class="p">]</span>
            <span class="p">)</span>

            <span class="c1"># Check that a value was in the datasets being used</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">np</span><span class="o">.</span><span class="n">isfinite</span><span class="p">(</span><span class="n">conc</span><span class="p">):</span>
                <span class="k">if</span> <span class="n">VERBOSE</span><span class="p">:</span>
                    <span class="nb">print</span><span class="p">(</span>
                        <span class="s2">&quot;No concentration in active datasets for </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                            <span class="n">metabolite_id</span>
                        <span class="p">)</span>
                    <span class="p">)</span>
                <span class="k">continue</span>

            <span class="k">if</span> <span class="n">metabolite_id</span> <span class="ow">in</span> <span class="n">wildtypeIDtoCompartment</span><span class="p">:</span>
                <span class="n">metaboliteIDs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                    <span class="n">metabolite_id</span> <span class="o">+</span> <span class="n">wildtypeIDtoCompartment</span><span class="p">[</span><span class="n">metabolite_id</span><span class="p">]</span>
                <span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">metaboliteIDs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">metabolite_id</span> <span class="o">+</span> <span class="s2">&quot;[c]&quot;</span><span class="p">)</span>

            <span class="n">metaboliteConcentrations</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">conc</span><span class="p">)</span>

        <span class="c1"># CYS/SEL: concentration based on other amino acids</span>
        <span class="n">aaConcentrations</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">aaIndex</span><span class="p">,</span> <span class="n">aaID</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">sim_data</span><span class="o">.</span><span class="n">amino_acid_code_to_id_ordered</span><span class="o">.</span><span class="n">values</span><span class="p">()):</span>
            <span class="k">if</span> <span class="n">aaID</span> <span class="ow">in</span> <span class="n">metaboliteIDs</span><span class="p">:</span>
                <span class="n">metIndex</span> <span class="o">=</span> <span class="n">metaboliteIDs</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">aaID</span><span class="p">)</span>
                <span class="n">aaConcentrations</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">metaboliteConcentrations</span><span class="p">[</span><span class="n">metIndex</span><span class="p">])</span>
        <span class="n">aaSmallestConc</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">aaConcentrations</span><span class="p">)</span>

        <span class="n">metaboliteIDs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;CYS[c]&quot;</span><span class="p">)</span>
        <span class="n">metaboliteConcentrations</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">aaSmallestConc</span><span class="p">)</span>

        <span class="n">metaboliteIDs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;L-SELENOCYSTEINE[c]&quot;</span><span class="p">)</span>
        <span class="n">metaboliteConcentrations</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">aaSmallestConc</span><span class="p">)</span>

        <span class="c1"># DGTP: set to smallest of all other DNTP concentrations</span>
        <span class="n">dntpConcentrations</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">dntpIndex</span><span class="p">,</span> <span class="n">dntpID</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">sim_data</span><span class="o">.</span><span class="n">molecule_groups</span><span class="o">.</span><span class="n">dntps</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">dntpID</span> <span class="ow">in</span> <span class="n">metaboliteIDs</span><span class="p">:</span>
                <span class="n">metIndex</span> <span class="o">=</span> <span class="n">metaboliteIDs</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">dntpID</span><span class="p">)</span>
                <span class="n">dntpConcentrations</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">metaboliteConcentrations</span><span class="p">[</span><span class="n">metIndex</span><span class="p">])</span>
        <span class="n">dntpSmallestConc</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">dntpConcentrations</span><span class="p">)</span>

        <span class="n">metaboliteIDs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;DGTP[c]&quot;</span><span class="p">)</span>
        <span class="n">metaboliteConcentrations</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">dntpSmallestConc</span><span class="p">)</span>

        <span class="c1"># H: from reported pH</span>
        <span class="n">hydrogenConcentration</span> <span class="o">=</span> <span class="mi">10</span> <span class="o">**</span> <span class="p">(</span><span class="o">-</span><span class="n">sim_data</span><span class="o">.</span><span class="n">constants</span><span class="o">.</span><span class="n">pH</span><span class="p">)</span>

        <span class="n">metaboliteIDs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">sim_data</span><span class="o">.</span><span class="n">molecule_ids</span><span class="o">.</span><span class="n">proton</span><span class="p">)</span>
        <span class="n">metaboliteConcentrations</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">hydrogenConcentration</span><span class="p">)</span>

        <span class="c1"># PPI</span>
        <span class="n">ppi_conc</span> <span class="o">=</span> <span class="n">sim_data</span><span class="o">.</span><span class="n">constants</span><span class="o">.</span><span class="n">ppi_concentration</span><span class="o">.</span><span class="n">asNumber</span><span class="p">(</span>
            <span class="n">METABOLITE_CONCENTRATION_UNITS</span>
        <span class="p">)</span>
        <span class="n">metaboliteIDs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">sim_data</span><span class="o">.</span><span class="n">molecule_ids</span><span class="o">.</span><span class="n">ppi</span><span class="p">)</span>
        <span class="n">metaboliteConcentrations</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ppi_conc</span><span class="p">)</span>

        <span class="n">metaboliteIDs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;Pi[c]&quot;</span><span class="p">)</span>
        <span class="n">metaboliteConcentrations</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ppi_conc</span><span class="p">)</span>

        <span class="c1"># include metabolites that are part of biomass</span>
        <span class="n">d</span> <span class="o">=</span> <span class="n">sim_data</span><span class="o">.</span><span class="n">mass</span><span class="o">.</span><span class="n">getBiomassAsConcentrations</span><span class="p">(</span><span class="n">sim_data</span><span class="o">.</span><span class="n">doubling_time</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">d</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">metaboliteIDs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">key</span><span class="p">)</span>
            <span class="n">metaboliteConcentrations</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                <span class="n">value</span><span class="o">.</span><span class="n">asNumber</span><span class="p">(</span><span class="n">METABOLITE_CONCENTRATION_UNITS</span><span class="p">)</span>
            <span class="p">)</span>

        <span class="c1"># Load relative metabolite changes</span>
        <span class="n">relative_changes</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">float</span><span class="p">]]</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">raw_data</span><span class="o">.</span><span class="n">relative_metabolite_concentrations</span><span class="p">:</span>
            <span class="n">met</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;Metabolite&quot;</span><span class="p">]</span>
            <span class="n">met_id</span> <span class="o">=</span> <span class="n">met</span> <span class="o">+</span> <span class="n">wildtypeIDtoCompartment</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">met</span><span class="p">,</span> <span class="s2">&quot;[c]&quot;</span><span class="p">)</span>

            <span class="c1"># AA concentrations are determined through charging</span>
            <span class="k">if</span> <span class="n">met_id</span> <span class="ow">in</span> <span class="n">sim_data</span><span class="o">.</span><span class="n">molecule_groups</span><span class="o">.</span><span class="n">amino_acids</span><span class="p">:</span>
                <span class="k">continue</span>

            <span class="c1"># Get relative metabolite change in each media condition</span>
            <span class="k">for</span> <span class="n">col</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">row</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                <span class="c1"># Skip the ID column and minimal column (only has values of 1)</span>
                <span class="c1"># or skip invalid values</span>
                <span class="k">if</span> <span class="n">col</span> <span class="o">==</span> <span class="s2">&quot;Metabolite&quot;</span> <span class="ow">or</span> <span class="n">col</span> <span class="o">==</span> <span class="s2">&quot;minimal&quot;</span> <span class="ow">or</span> <span class="ow">not</span> <span class="n">np</span><span class="o">.</span><span class="n">isfinite</span><span class="p">(</span><span class="n">value</span><span class="p">):</span>
                    <span class="k">continue</span>

                <span class="k">if</span> <span class="n">col</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">relative_changes</span><span class="p">:</span>
                    <span class="n">relative_changes</span><span class="p">[</span><span class="n">col</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
                <span class="n">relative_changes</span><span class="p">[</span><span class="n">col</span><span class="p">][</span><span class="n">met_id</span><span class="p">]</span> <span class="o">=</span> <span class="n">value</span>

        <span class="c1">## Add manually curated values for other media</span>
        <span class="k">for</span> <span class="p">(</span>
            <span class="n">media</span><span class="p">,</span>
            <span class="n">data</span><span class="p">,</span>
        <span class="p">)</span> <span class="ow">in</span> <span class="n">sim_data</span><span class="o">.</span><span class="n">adjustments</span><span class="o">.</span><span class="n">relative_metabolite_concentrations_changes</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="k">if</span> <span class="n">media</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">relative_changes</span><span class="p">:</span>
                <span class="n">relative_changes</span><span class="p">[</span><span class="n">media</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="k">for</span> <span class="n">met</span><span class="p">,</span> <span class="n">change</span> <span class="ow">in</span> <span class="n">data</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                <span class="k">if</span> <span class="n">met</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">relative_changes</span><span class="p">[</span><span class="n">media</span><span class="p">]:</span>
                    <span class="n">relative_changes</span><span class="p">[</span><span class="n">media</span><span class="p">][</span><span class="n">met</span><span class="p">]</span> <span class="o">=</span> <span class="n">change</span>

        <span class="c1"># save concentrations as class variables</span>
        <span class="n">unique_ids</span><span class="p">,</span> <span class="n">counts</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">metaboliteIDs</span><span class="p">,</span> <span class="n">return_counts</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">np</span><span class="o">.</span><span class="n">any</span><span class="p">(</span><span class="n">counts</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;Multiple concentrations for metabolite(s): </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                    <span class="s2">&quot;, &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">unique_ids</span><span class="p">[</span><span class="n">counts</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">])</span>
                <span class="p">)</span>
            <span class="p">)</span>

        <span class="c1"># TODO (Travis): only pass raw_data and sim_data and create functions to load absolute and relative concentrations</span>
        <span class="n">conc_dict</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span>
            <span class="nb">zip</span><span class="p">(</span>
                <span class="n">metaboliteIDs</span><span class="p">,</span>
                <span class="n">METABOLITE_CONCENTRATION_UNITS</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">metaboliteConcentrations</span><span class="p">),</span>
            <span class="p">)</span>
        <span class="p">)</span>
        <span class="n">all_metabolite_ids</span> <span class="o">=</span> <span class="p">{</span><span class="n">met</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]</span> <span class="k">for</span> <span class="n">met</span> <span class="ow">in</span> <span class="n">raw_data</span><span class="o">.</span><span class="n">metabolites</span><span class="p">}</span>
        <span class="n">linked_metabolites</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_build_linked_metabolites</span><span class="p">(</span><span class="n">raw_data</span><span class="p">,</span> <span class="n">conc_dict</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">concentration_updates</span> <span class="o">=</span> <span class="n">ConcentrationUpdates</span><span class="p">(</span>
            <span class="n">conc_dict</span><span class="p">,</span>
            <span class="n">relative_changes</span><span class="p">,</span>
            <span class="n">raw_data</span><span class="o">.</span><span class="n">equilibrium_reactions</span><span class="p">,</span>
            <span class="n">sim_data</span><span class="o">.</span><span class="n">external_state</span><span class="o">.</span><span class="n">exchange_dict</span><span class="p">,</span>
            <span class="n">all_metabolite_ids</span><span class="p">,</span>
            <span class="n">linked_metabolites</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">conc_dict</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">concentration_updates</span><span class="o">.</span><span class="n">concentrations_based_on_nutrients</span><span class="p">(</span>
            <span class="s2">&quot;minimal&quot;</span>
        <span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">nutrients_to_internal_conc</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">nutrients_to_internal_conc</span><span class="p">[</span><span class="s2">&quot;minimal&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">conc_dict</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span></div>


<div class="viewcode-block" id="Metabolism._build_linked_metabolites">
<a class="viewcode-back" href="../../../../../reference/api/reconstruction/reconstruction.ecoli.dataclasses.process.metabolism.html#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._build_linked_metabolites">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">_build_linked_metabolites</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">raw_data</span><span class="p">:</span> <span class="n">KnowledgeBaseEcoli</span><span class="p">,</span> <span class="n">conc_dict</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Unum</span><span class="p">]</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Calculates ratio between linked metabolites to keep it constant</span>
<span class="sd">        throughout a simulation.</span>

<span class="sd">        Args:</span>
<span class="sd">                raw_data: Raw data object</span>
<span class="sd">                conc_dict: Mapping of metabolite IDs to homeostatic concentrations</span>
<span class="sd">                        calculated by :py:meth:`~_build_biomass`</span>

<span class="sd">        Returns:</span>
<span class="sd">                Mapping from a linked metabolite to its lead</span>
<span class="sd">                metabolite and concentration ratio to be maintained::</span>

<span class="sd">                                {&#39;lead&#39; (str): metabolite to link the concentration to,</span>
<span class="sd">                                &#39;ratio&#39; (float): ratio to multiply the lead concentration by}</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">linked_metabolites</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">raw_data</span><span class="o">.</span><span class="n">linked_metabolites</span><span class="p">:</span>
            <span class="n">lead</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;Lead metabolite&quot;</span><span class="p">]</span>
            <span class="n">linked</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;Linked metabolite&quot;</span><span class="p">]</span>
            <span class="n">ratio</span> <span class="o">=</span> <span class="n">units</span><span class="o">.</span><span class="n">strip_empty_units</span><span class="p">(</span><span class="n">conc_dict</span><span class="p">[</span><span class="n">linked</span><span class="p">]</span> <span class="o">/</span> <span class="n">conc_dict</span><span class="p">[</span><span class="n">lead</span><span class="p">])</span>

            <span class="n">linked_metabolites</span><span class="p">[</span><span class="n">linked</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;lead&quot;</span><span class="p">:</span> <span class="n">lead</span><span class="p">,</span> <span class="s2">&quot;ratio&quot;</span><span class="p">:</span> <span class="n">ratio</span><span class="p">}</span>

        <span class="k">return</span> <span class="n">linked_metabolites</span></div>


<div class="viewcode-block" id="Metabolism._build_metabolism">
<a class="viewcode-back" href="../../../../../reference/api/reconstruction/reconstruction.ecoli.dataclasses.process.metabolism.html#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._build_metabolism">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">_build_metabolism</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">raw_data</span><span class="p">:</span> <span class="n">KnowledgeBaseEcoli</span><span class="p">,</span> <span class="n">sim_data</span><span class="p">:</span> <span class="s2">&quot;SimulationDataEcoli&quot;</span>
    <span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Build the matrices/vectors for metabolism (FBA)</span>
<span class="sd">        Reads in and stores reaction and kinetic constraint information</span>

<span class="sd">        Args:</span>
<span class="sd">                raw_data: Raw data object</span>
<span class="sd">                sim_data: Simulation data object</span>

<span class="sd">        Attributes set:</span>
<span class="sd">                - :py:attr:`~.kinetic_constraint_reactions`</span>
<span class="sd">                - :py:attr:`~.kinetic_constraint_enzymes`</span>
<span class="sd">                - :py:attr:`~.kinetic_constraint_substrates`</span>
<span class="sd">                - :py:attr:`~._kcats`</span>
<span class="sd">                - :py:attr:`~._saturations`</span>
<span class="sd">                - :py:attr:`~._enzymes`</span>
<span class="sd">                - :py:attr:`~.constraint_is_kcat_only`</span>
<span class="sd">                - :py:attr:`~._compiled_enzymes`</span>
<span class="sd">                - :py:attr:`~._compiled_saturation`</span>
<span class="sd">                - :py:attr:`~.reaction_stoich`</span>
<span class="sd">                - :py:attr:`~.maintenance_reaction`</span>
<span class="sd">                - :py:attr:`~.reaction_catalysts`</span>
<span class="sd">                - :py:attr:`~.catalyst_ids`</span>
<span class="sd">                - :py:attr:`~.reactions_with_catalyst`</span>
<span class="sd">                - :py:attr:`~.catalysis_matrix_I`</span>
<span class="sd">                - :py:attr:`~.catalysis_matrix_J`</span>
<span class="sd">                - :py:attr:`~.catalysis_matrix_V`</span>
<span class="sd">                - :py:attr:`~.use_all_constraints`</span>
<span class="sd">                - :py:attr:`~.constraints_to_disable`</span>
<span class="sd">                - :py:attr:`~.base_reaction_ids`</span>
<span class="sd">                - :py:attr:`~.reaction_id_to_base_reaction_id`</span>
<span class="sd">                - :py:attr:`~.amino_acid_export_kms`</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="p">(</span>
            <span class="n">base_rxn_ids</span><span class="p">,</span>
            <span class="n">reaction_stoich</span><span class="p">,</span>
            <span class="n">reversible_reactions</span><span class="p">,</span>
            <span class="n">catalysts</span><span class="p">,</span>
            <span class="n">rxn_id_to_base_rxn_id</span><span class="p">,</span>
        <span class="p">)</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">extract_reactions</span><span class="p">(</span><span class="n">raw_data</span><span class="p">,</span> <span class="n">sim_data</span><span class="p">)</span>

        <span class="c1"># Load kinetic reaction constraints from raw_data</span>
        <span class="n">known_metabolites</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">conc_dict</span><span class="p">)</span>
        <span class="n">raw_constraints</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">extract_kinetic_constraints</span><span class="p">(</span>
            <span class="n">raw_data</span><span class="p">,</span>
            <span class="n">sim_data</span><span class="p">,</span>
            <span class="n">stoich</span><span class="o">=</span><span class="n">reaction_stoich</span><span class="p">,</span>
            <span class="n">catalysts</span><span class="o">=</span><span class="n">catalysts</span><span class="p">,</span>
            <span class="n">known_metabolites</span><span class="o">=</span><span class="n">known_metabolites</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="c1"># Make modifications from kinetics data</span>
        <span class="p">(</span>
            <span class="n">constraints</span><span class="p">,</span>
            <span class="n">reaction_stoich</span><span class="p">,</span>
            <span class="n">catalysts</span><span class="p">,</span>
            <span class="n">reversible_reactions</span><span class="p">,</span>
            <span class="n">rxn_id_to_base_rxn_id</span><span class="p">,</span>
        <span class="p">)</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_replace_enzyme_reactions</span><span class="p">(</span>
            <span class="n">raw_constraints</span><span class="p">,</span>
            <span class="n">reaction_stoich</span><span class="p">,</span>
            <span class="n">catalysts</span><span class="p">,</span>
            <span class="n">reversible_reactions</span><span class="p">,</span>
            <span class="n">rxn_id_to_base_rxn_id</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="c1"># Create symbolic kinetic equations</span>
        <span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">kinetic_constraint_reactions</span><span class="p">,</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">kinetic_constraint_enzymes</span><span class="p">,</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">kinetic_constraint_substrates</span><span class="p">,</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_kcats</span><span class="p">,</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_saturations</span><span class="p">,</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_enzymes</span><span class="p">,</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">constraint_is_kcat_only</span><span class="p">,</span>
        <span class="p">)</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_lambdify_constraints</span><span class="p">(</span><span class="n">constraints</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_compiled_enzymes</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span>
            <span class="n">Callable</span><span class="p">[[</span><span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">]],</span> <span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">]]</span>
        <span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_compiled_saturation</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span>
            <span class="n">Callable</span><span class="p">[[</span><span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">]],</span> <span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">]]</span>
        <span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>

        <span class="c1"># TODO: move this to a sim_data analysis script</span>
        <span class="k">if</span> <span class="n">VERBOSE</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Summary of included metabolism kinetics:&quot;</span><span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span>
                <span class="s2">&quot;Reactions with kinetics: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                    <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">kinetic_constraint_reactions</span><span class="p">)</span>
                <span class="p">)</span>
            <span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span>
                <span class="s2">&quot;Enzymes with kinetics: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">kinetic_constraint_enzymes</span><span class="p">))</span>
            <span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span>
                <span class="s2">&quot;Metabolites in kinetics: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                    <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">kinetic_constraint_substrates</span><span class="p">)</span>
                <span class="p">)</span>
            <span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span>
                <span class="s2">&quot;Number of kcat values: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                    <span class="nb">len</span><span class="p">([</span><span class="n">k</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">constraints</span><span class="o">.</span><span class="n">values</span><span class="p">()</span> <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">c</span><span class="p">[</span><span class="s2">&quot;kcat&quot;</span><span class="p">]])</span>
                <span class="p">)</span>
            <span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span>
                <span class="s2">&quot;Number of saturation terms: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                    <span class="nb">len</span><span class="p">([</span><span class="n">s</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">constraints</span><span class="o">.</span><span class="n">values</span><span class="p">()</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">c</span><span class="p">[</span><span class="s2">&quot;saturation&quot;</span><span class="p">]])</span>
                <span class="p">)</span>
            <span class="p">)</span>

        <span class="c1"># Verify no substrates with unknown concentrations have been added</span>
        <span class="n">unknown</span> <span class="o">=</span> <span class="p">{</span>
            <span class="n">m</span> <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">kinetic_constraint_substrates</span> <span class="k">if</span> <span class="n">m</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">known_metabolites</span>
        <span class="p">}</span>
        <span class="k">if</span> <span class="n">unknown</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;Unknown concentration for </span><span class="si">{}</span><span class="s2">. Need to remove&quot;</span>
                <span class="s2">&quot; kinetics saturation term.&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s2">&quot;, &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">unknown</span><span class="p">))</span>
            <span class="p">)</span>

        <span class="c1"># Extract data</span>
        <span class="n">reactions_with_catalyst</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">catalysts</span><span class="p">)</span>
        <span class="n">catalyst_ids</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">({</span><span class="n">c</span> <span class="k">for</span> <span class="n">all_cat</span> <span class="ow">in</span> <span class="n">catalysts</span><span class="o">.</span><span class="n">values</span><span class="p">()</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">all_cat</span><span class="p">})</span>

        <span class="c1"># Create catalysis matrix (to be used in the simulation)</span>
        <span class="n">catalysisMatrixI</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">catalysisMatrixJ</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">catalysisMatrixV</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">for</span> <span class="n">row</span><span class="p">,</span> <span class="n">reaction</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">reactions_with_catalyst</span><span class="p">):</span>
            <span class="k">for</span> <span class="n">catalyst</span> <span class="ow">in</span> <span class="n">catalysts</span><span class="p">[</span><span class="n">reaction</span><span class="p">]:</span>
                <span class="n">col</span> <span class="o">=</span> <span class="n">catalyst_ids</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">catalyst</span><span class="p">)</span>
                <span class="n">catalysisMatrixI</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">row</span><span class="p">)</span>
                <span class="n">catalysisMatrixJ</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">col</span><span class="p">)</span>
                <span class="n">catalysisMatrixV</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>

        <span class="n">catalysisMatrixI</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">catalysisMatrixI</span><span class="p">)</span>
        <span class="n">catalysisMatrixJ</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">catalysisMatrixJ</span><span class="p">)</span>
        <span class="n">catalysisMatrixV</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">catalysisMatrixV</span><span class="p">)</span>

        <span class="c1"># Properties for FBA reconstruction</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">reaction_stoich</span> <span class="o">=</span> <span class="n">reaction_stoich</span>
        <span class="c1"># TODO (ggsun): add this as a raw .tsv file</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">maintenance_reaction</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;ATP[c]&quot;</span><span class="p">:</span> <span class="o">-</span><span class="mi">1</span><span class="p">,</span>
            <span class="s2">&quot;WATER[c]&quot;</span><span class="p">:</span> <span class="o">-</span><span class="mi">1</span><span class="p">,</span>
            <span class="s2">&quot;ADP[c]&quot;</span><span class="p">:</span> <span class="o">+</span><span class="mi">1</span><span class="p">,</span>
            <span class="s2">&quot;Pi[c]&quot;</span><span class="p">:</span> <span class="o">+</span><span class="mi">1</span><span class="p">,</span>
            <span class="s2">&quot;PROTON[c]&quot;</span><span class="p">:</span> <span class="o">+</span><span class="mi">1</span><span class="p">,</span>
        <span class="p">}</span>

        <span class="c1"># Properties for catalysis matrix (to set hard bounds)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">reaction_catalysts</span> <span class="o">=</span> <span class="n">catalysts</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">catalyst_ids</span> <span class="o">=</span> <span class="n">catalyst_ids</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">reactions_with_catalyst</span> <span class="o">=</span> <span class="n">reactions_with_catalyst</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">catalysis_matrix_I</span> <span class="o">=</span> <span class="n">catalysisMatrixI</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">catalysis_matrix_J</span> <span class="o">=</span> <span class="n">catalysisMatrixJ</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">catalysis_matrix_V</span> <span class="o">=</span> <span class="n">catalysisMatrixV</span>

        <span class="c1"># Properties for setting flux targets</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">use_all_constraints</span> <span class="o">=</span> <span class="n">USE_ALL_CONSTRAINTS</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">constraints_to_disable</span> <span class="o">=</span> <span class="p">[</span>
            <span class="n">rxn</span><span class="p">[</span><span class="s2">&quot;disabled reaction&quot;</span><span class="p">]</span> <span class="k">for</span> <span class="n">rxn</span> <span class="ow">in</span> <span class="n">raw_data</span><span class="o">.</span><span class="n">disabled_kinetic_reactions</span>
        <span class="p">]</span>

        <span class="c1"># Properties for conversion of fluxes to those for base reaction IDs</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">base_reaction_ids</span> <span class="o">=</span> <span class="n">base_rxn_ids</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">reaction_id_to_base_reaction_id</span> <span class="o">=</span> <span class="n">rxn_id_to_base_rxn_id</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">amino_acid_export_kms</span> <span class="o">=</span> <span class="n">raw_data</span><span class="o">.</span><span class="n">amino_acid_export_kms</span></div>


<div class="viewcode-block" id="Metabolism._build_ppgpp_reactions">
<a class="viewcode-back" href="../../../../../reference/api/reconstruction/reconstruction.ecoli.dataclasses.process.metabolism.html#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._build_ppgpp_reactions">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">_build_ppgpp_reactions</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">raw_data</span><span class="p">:</span> <span class="n">KnowledgeBaseEcoli</span><span class="p">,</span> <span class="n">sim_data</span><span class="p">:</span> <span class="s2">&quot;SimulationDataEcoli&quot;</span>
    <span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Creates structures for ppGpp reactions for use in polypeptide_elongation.</span>

<span class="sd">        Args:</span>
<span class="sd">                raw_data: Raw data object</span>
<span class="sd">                sim_data: Simulation data object</span>

<span class="sd">        Attributes set:</span>
<span class="sd">                - :py:attr:`~.ppgpp_synthesis_reaction`</span>
<span class="sd">                - :py:attr:`~.ppgpp_degradation_reaction`</span>
<span class="sd">                - :py:attr:`~.ppgpp_reaction_names`</span>
<span class="sd">                - :py:attr:`~.ppgpp_reaction_metabolites`</span>
<span class="sd">                - :py:attr:`~.ppgpp_reaction_stoich`</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">ppgpp_synthesis_reaction</span> <span class="o">=</span> <span class="s2">&quot;GDPPYPHOSKIN-RXN&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">ppgpp_degradation_reaction</span> <span class="o">=</span> <span class="s2">&quot;PPGPPSYN-RXN&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">ppgpp_reaction_names</span> <span class="o">=</span> <span class="p">[</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">ppgpp_synthesis_reaction</span><span class="p">,</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">ppgpp_degradation_reaction</span><span class="p">,</span>
        <span class="p">]</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">ppgpp_reaction_metabolites</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="c1"># Indices (i: metabolite, j: reaction) and values (v: stoichiometry)</span>
        <span class="c1"># for sparse reaction matrix</span>
        <span class="n">metabolite_indices</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">new_index</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">rxn_i</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">rxn_j</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">rxn_v</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="c1"># Record sparse indices in the matrix</span>
        <span class="k">for</span> <span class="n">j</span><span class="p">,</span> <span class="n">rxn</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">ppgpp_reaction_names</span><span class="p">):</span>
            <span class="k">for</span> <span class="n">met</span><span class="p">,</span> <span class="n">stoich</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">reaction_stoich</span><span class="p">[</span><span class="n">rxn</span><span class="p">]</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                <span class="n">idx</span> <span class="o">=</span> <span class="n">metabolite_indices</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">met</span><span class="p">,</span> <span class="n">new_index</span><span class="p">)</span>

                <span class="k">if</span> <span class="n">idx</span> <span class="o">==</span> <span class="n">new_index</span><span class="p">:</span>
                    <span class="n">metabolite_indices</span><span class="p">[</span><span class="n">met</span><span class="p">]</span> <span class="o">=</span> <span class="n">new_index</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">ppgpp_reaction_metabolites</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">met</span><span class="p">)</span>
                    <span class="n">new_index</span> <span class="o">+=</span> <span class="mi">1</span>

                <span class="n">rxn_i</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">idx</span><span class="p">)</span>
                <span class="n">rxn_j</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">j</span><span class="p">)</span>
                <span class="n">rxn_v</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">stoich</span><span class="p">)</span>

        <span class="c1"># Assemble matrix based on indices</span>
        <span class="c1"># new_index is number of metabolites, j+1 is number of reactions</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">ppgpp_reaction_stoich</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">new_index</span><span class="p">,</span> <span class="n">j</span> <span class="o">+</span> <span class="mi">1</span><span class="p">),</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int32</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">ppgpp_reaction_stoich</span><span class="p">[</span><span class="n">rxn_i</span><span class="p">,</span> <span class="n">rxn_j</span><span class="p">]</span> <span class="o">=</span> <span class="n">rxn_v</span></div>


<div class="viewcode-block" id="Metabolism._build_transport_reactions">
<a class="viewcode-back" href="../../../../../reference/api/reconstruction/reconstruction.ecoli.dataclasses.process.metabolism.html#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._build_transport_reactions">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">_build_transport_reactions</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">raw_data</span><span class="p">:</span> <span class="n">KnowledgeBaseEcoli</span><span class="p">,</span> <span class="n">sim_data</span><span class="p">:</span> <span class="s2">&quot;SimulationDataEcoli&quot;</span>
    <span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Creates list of transport reactions that are included in the</span>
<span class="sd">        reaction network.</span>

<span class="sd">        Args:</span>
<span class="sd">                raw_data: Raw data object</span>
<span class="sd">                sim_data: Simulation data object</span>

<span class="sd">        Attributes set:</span>
<span class="sd">                - :py:attr:`~.transport_reactions`</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">transport_reactions</span> <span class="o">=</span> <span class="p">[</span>
            <span class="n">rxn_id</span>
            <span class="k">for</span> <span class="n">rxn_id</span><span class="p">,</span> <span class="n">stoich</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">reaction_stoich</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_is_transport_rxn</span><span class="p">(</span><span class="n">stoich</span><span class="p">)</span>
        <span class="p">]</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">transport_reactions</span> <span class="o">=</span> <span class="n">transport_reactions</span></div>


<div class="viewcode-block" id="Metabolism._build_amino_acid_pathways">
<a class="viewcode-back" href="../../../../../reference/api/reconstruction/reconstruction.ecoli.dataclasses.process.metabolism.html#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._build_amino_acid_pathways">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">_build_amino_acid_pathways</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">raw_data</span><span class="p">:</span> <span class="n">KnowledgeBaseEcoli</span><span class="p">,</span> <span class="n">sim_data</span><span class="p">:</span> <span class="s2">&quot;SimulationDataEcoli&quot;</span>
    <span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Creates mapping between enzymes and amino acid pathways with</span>
<span class="sd">        allosteric inhibition feedback from the amino acid.</span>

<span class="sd">        Args:</span>
<span class="sd">                raw_data: Raw data object</span>
<span class="sd">                sim_data: Simulation data object</span>

<span class="sd">        Attributes set:</span>
<span class="sd">                - :py:attr:`~.aa_synthesis_pathways`</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">aa_synthesis_pathways</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">cytoplasm_tag</span> <span class="o">=</span> <span class="s2">&quot;[c]&quot;</span>

        <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">raw_data</span><span class="o">.</span><span class="n">amino_acid_pathways</span><span class="p">:</span>
            <span class="n">data</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="n">data</span><span class="p">[</span><span class="s2">&quot;enzymes&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span>
                <span class="n">e</span> <span class="o">+</span> <span class="n">sim_data</span><span class="o">.</span><span class="n">getter</span><span class="o">.</span><span class="n">get_compartment_tag</span><span class="p">(</span><span class="n">e</span><span class="p">)</span> <span class="k">for</span> <span class="n">e</span> <span class="ow">in</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;Enzymes&quot;</span><span class="p">]</span>
            <span class="p">]</span>
            <span class="n">data</span><span class="p">[</span><span class="s2">&quot;reverse enzymes&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span>
                <span class="n">e</span> <span class="o">+</span> <span class="n">sim_data</span><span class="o">.</span><span class="n">getter</span><span class="o">.</span><span class="n">get_compartment_tag</span><span class="p">(</span><span class="n">e</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">e</span> <span class="ow">in</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;Reverse enzymes&quot;</span><span class="p">]</span>
            <span class="p">]</span>
            <span class="n">data</span><span class="p">[</span><span class="s2">&quot;kcat_data&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span> <span class="o">/</span> <span class="n">units</span><span class="o">.</span><span class="n">s</span> <span class="k">if</span> <span class="n">units</span><span class="o">.</span><span class="n">isnan</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;kcat&quot;</span><span class="p">])</span> <span class="k">else</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;kcat&quot;</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">units</span><span class="o">.</span><span class="n">isnan</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;KI, lower bound&quot;</span><span class="p">])</span> <span class="ow">or</span> <span class="n">units</span><span class="o">.</span><span class="n">isnan</span><span class="p">(</span>
                <span class="n">row</span><span class="p">[</span><span class="s2">&quot;KI, lower bound&quot;</span><span class="p">]</span>
            <span class="p">):</span>
                <span class="n">data</span><span class="p">[</span><span class="s2">&quot;ki&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">data</span><span class="p">[</span><span class="s2">&quot;ki&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;KI, lower bound&quot;</span><span class="p">],</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;KI, upper bound&quot;</span><span class="p">])</span>
            <span class="n">data</span><span class="p">[</span><span class="s2">&quot;upstream&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span>
                <span class="n">k</span> <span class="o">+</span> <span class="n">cytoplasm_tag</span><span class="p">:</span> <span class="n">v</span> <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;Upstream amino acids&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
            <span class="p">}</span>
            <span class="n">data</span><span class="p">[</span><span class="s2">&quot;reverse&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span>
                <span class="n">k</span> <span class="o">+</span> <span class="n">cytoplasm_tag</span><span class="p">:</span> <span class="n">v</span> <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;Reverse amino acids&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
            <span class="p">}</span>
            <span class="n">data</span><span class="p">[</span><span class="s2">&quot;km, upstream&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span>
                <span class="n">k</span> <span class="o">+</span> <span class="n">cytoplasm_tag</span><span class="p">:</span> <span class="n">v</span> <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;KM, upstream&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
            <span class="p">}</span>
            <span class="n">data</span><span class="p">[</span><span class="s2">&quot;km, reverse&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;KM, reverse&quot;</span><span class="p">]</span>
            <span class="n">data</span><span class="p">[</span><span class="s2">&quot;km, degradation&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
                <span class="n">np</span><span class="o">.</span><span class="n">inf</span> <span class="o">*</span> <span class="n">units</span><span class="o">.</span><span class="n">mol</span> <span class="o">/</span> <span class="n">units</span><span class="o">.</span><span class="n">L</span>
                <span class="k">if</span> <span class="n">units</span><span class="o">.</span><span class="n">isnan</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;KM, degradation&quot;</span><span class="p">])</span>
                <span class="k">else</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;KM, degradation&quot;</span><span class="p">]</span>
            <span class="p">)</span>
            <span class="n">data</span><span class="p">[</span><span class="s2">&quot;downstream&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span>
                <span class="n">k</span> <span class="o">+</span> <span class="n">cytoplasm_tag</span><span class="p">:</span> <span class="n">v</span> <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;Downstream amino acids&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
            <span class="p">}</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">aa_synthesis_pathways</span><span class="p">[</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;Amino acid&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="n">cytoplasm_tag</span><span class="p">]</span> <span class="o">=</span> <span class="n">data</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">aa_synthesis_pathway_adjustments</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">float</span><span class="p">]]</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">raw_data</span><span class="o">.</span><span class="n">adjustments</span><span class="o">.</span><span class="n">amino_acid_pathways</span><span class="p">:</span>
            <span class="c1"># Read data from row</span>
            <span class="n">aa</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;Amino acid&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="n">cytoplasm_tag</span>
            <span class="n">parameter</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;Parameter&quot;</span><span class="p">]</span>
            <span class="n">factor</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;Factor&quot;</span><span class="p">]</span>

            <span class="c1"># Store adjustments to be used later</span>
            <span class="n">adjustments</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">aa_synthesis_pathway_adjustments</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">aa</span><span class="p">,</span> <span class="p">{})</span>
            <span class="n">adjustments</span><span class="p">[</span><span class="n">parameter</span><span class="p">]</span> <span class="o">=</span> <span class="n">factor</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">aa_synthesis_pathway_adjustments</span><span class="p">[</span><span class="n">aa</span><span class="p">]</span> <span class="o">=</span> <span class="n">adjustments</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">amino_acid_uptake_rates</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">raw_data</span><span class="o">.</span><span class="n">amino_acid_uptake_rates</span><span class="p">:</span>
            <span class="n">rates</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="n">rates</span><span class="p">[</span><span class="s2">&quot;uptake&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;Uptake&quot;</span><span class="p">]</span>
            <span class="n">rates</span><span class="p">[</span><span class="s2">&quot;LB&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;Uptake, LB&quot;</span><span class="p">]</span>
            <span class="n">rates</span><span class="p">[</span><span class="s2">&quot;UB&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;Uptake, UB&quot;</span><span class="p">]</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">amino_acid_uptake_rates</span><span class="p">[</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;Amino acid&quot;</span><span class="p">]]</span> <span class="o">=</span> <span class="n">rates</span></div>


<div class="viewcode-block" id="Metabolism.get_kinetic_constraints">
<a class="viewcode-back" href="../../../../../reference/api/reconstruction/reconstruction.ecoli.dataclasses.process.metabolism.html#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.get_kinetic_constraints">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">get_kinetic_constraints</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">enzymes</span><span class="p">:</span> <span class="n">Unum</span><span class="p">,</span> <span class="n">substrates</span><span class="p">:</span> <span class="n">Unum</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Unum</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Allows for dynamic code generation for kinetic constraint calculation</span>
<span class="sd">        for use in Metabolism process. Inputs should be unitless but the order</span>
<span class="sd">        of magnitude should match the kinetics parameters (umol/L/s).</span>

<span class="sd">        If trying to pickle sim_data object after function has been called,</span>
<span class="sd">        _compiled_enzymes and _compiled_saturation might not be able to be pickled.</span>
<span class="sd">        See __getstate__(), __setstate__() comments on PR 111 to address.</span>

<span class="sd">        Returns np.array of floats of the kinetic constraint target for each</span>
<span class="sd">        reaction with kinetic parameters</span>

<span class="sd">        Args:</span>
<span class="sd">                enzymes: concentrations of enzymes associated with kinetic</span>
<span class="sd">                        constraints (mol / volume units)</span>
<span class="sd">                substrates: concentrations of substrates associated with kinetic</span>
<span class="sd">                        constraints (mol / volume units)</span>

<span class="sd">        Returns:</span>
<span class="sd">                Array of dimensions (n reactions, 3) where each row contains the</span>
<span class="sd">                min, mean and max kinetic constraints for each reaction with kinetic</span>
<span class="sd">                constraints (mol / volume / time units)</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_compiled_enzymes</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_compiled_enzymes</span> <span class="o">=</span> <span class="nb">eval</span><span class="p">(</span><span class="s2">&quot;lambda e: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_enzymes</span><span class="p">))</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_compiled_saturation</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_compiled_saturation</span> <span class="o">=</span> <span class="nb">eval</span><span class="p">(</span><span class="s2">&quot;lambda s: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_saturations</span><span class="p">))</span>

        <span class="c1"># Strip units from args</span>
        <span class="n">enzs</span> <span class="o">=</span> <span class="n">enzymes</span><span class="o">.</span><span class="n">asNumber</span><span class="p">(</span><span class="n">KINETIC_CONSTRAINT_CONC_UNITS</span><span class="p">)</span>
        <span class="n">subs</span> <span class="o">=</span> <span class="n">substrates</span><span class="o">.</span><span class="n">asNumber</span><span class="p">(</span><span class="n">KINETIC_CONSTRAINT_CONC_UNITS</span><span class="p">)</span>

        <span class="n">capacity</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_compiled_enzymes</span><span class="p">(</span><span class="n">enzs</span><span class="p">))[:,</span> <span class="kc">None</span><span class="p">]</span> <span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">_kcats</span>
        <span class="n">saturation</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span>
            <span class="p">[[</span><span class="nb">min</span><span class="p">(</span><span class="n">v</span><span class="p">),</span> <span class="nb">sum</span><span class="p">(</span><span class="n">v</span><span class="p">)</span> <span class="o">/</span> <span class="nb">len</span><span class="p">(</span><span class="n">v</span><span class="p">),</span> <span class="nb">max</span><span class="p">(</span><span class="n">v</span><span class="p">)]</span> <span class="k">for</span> <span class="n">v</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_compiled_saturation</span><span class="p">(</span><span class="n">subs</span><span class="p">)]</span>
        <span class="p">)</span>

        <span class="k">return</span> <span class="n">KINETIC_CONSTRAINT_CONC_UNITS</span> <span class="o">*</span> <span class="n">K_CAT_UNITS</span> <span class="o">*</span> <span class="n">capacity</span> <span class="o">*</span> <span class="n">saturation</span></div>


<div class="viewcode-block" id="Metabolism.exchange_constraints">
<a class="viewcode-back" href="../../../../../reference/api/reconstruction/reconstruction.ecoli.dataclasses.process.metabolism.html#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.exchange_constraints">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">exchange_constraints</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">exchangeIDs</span><span class="p">,</span>
        <span class="n">coefficient</span><span class="p">,</span>
        <span class="n">targetUnits</span><span class="p">,</span>
        <span class="n">media_id</span><span class="p">,</span>
        <span class="n">unconstrained</span><span class="p">,</span>
        <span class="n">constrained</span><span class="p">,</span>
        <span class="n">concModificationsBasedOnCondition</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Called during Metabolism process</span>
<span class="sd">        Returns the homeostatic objective concentrations based on the current nutrients</span>
<span class="sd">        Returns levels for external molecules available to exchange based on the current nutrients</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">newObjective</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">concentration_updates</span><span class="o">.</span><span class="n">concentrations_based_on_nutrients</span><span class="p">(</span>
            <span class="n">imports</span><span class="o">=</span><span class="n">unconstrained</span><span class="o">.</span><span class="n">union</span><span class="p">(</span><span class="n">constrained</span><span class="p">),</span>
            <span class="n">media_id</span><span class="o">=</span><span class="n">media_id</span><span class="p">,</span>
            <span class="n">conversion_units</span><span class="o">=</span><span class="n">targetUnits</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="k">if</span> <span class="n">concModificationsBasedOnCondition</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">newObjective</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">concModificationsBasedOnCondition</span><span class="p">)</span>

        <span class="n">externalMoleculeLevels</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">exchangeIDs</span><span class="p">),</span> <span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">index</span><span class="p">,</span> <span class="n">moleculeID</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">exchangeIDs</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">moleculeID</span> <span class="ow">in</span> <span class="n">unconstrained</span><span class="p">:</span>
                <span class="n">externalMoleculeLevels</span><span class="p">[</span><span class="n">index</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">inf</span>
            <span class="k">elif</span> <span class="n">moleculeID</span> <span class="ow">in</span> <span class="n">constrained</span><span class="p">:</span>
                <span class="n">externalMoleculeLevels</span><span class="p">[</span><span class="n">index</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
                    <span class="n">constrained</span><span class="p">[</span><span class="n">moleculeID</span><span class="p">]</span> <span class="o">*</span> <span class="n">coefficient</span>
                <span class="p">)</span><span class="o">.</span><span class="n">asNumber</span><span class="p">(</span><span class="n">targetUnits</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">externalMoleculeLevels</span><span class="p">[</span><span class="n">index</span><span class="p">]</span> <span class="o">=</span> <span class="mf">0.0</span>

        <span class="k">return</span> <span class="n">externalMoleculeLevels</span><span class="p">,</span> <span class="n">newObjective</span></div>


<div class="viewcode-block" id="Metabolism.set_phenomological_supply_constants">
<a class="viewcode-back" href="../../../../../reference/api/reconstruction/reconstruction.ecoli.dataclasses.process.metabolism.html#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.set_phenomological_supply_constants">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">set_phenomological_supply_constants</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">sim_data</span><span class="p">:</span> <span class="s2">&quot;SimulationDataEcoli&quot;</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Sets constants to determine amino acid supply during translation.  Used</span>
<span class="sd">        with aa_supply_scaling() during simulations but supply can</span>
<span class="sd">        alternatively be determined mechanistically.  This approach may require</span>
<span class="sd">        manually adjusting constants (fraction_supply_inhibited and</span>
<span class="sd">        fraction_supply_exported) but has less variability related to gene</span>
<span class="sd">        expression and regulation.</span>

<span class="sd">        Args:</span>
<span class="sd">                sim_data: simulation data</span>

<span class="sd">        Attributes set:</span>
<span class="sd">                - :py:attr:`~.KI_aa_synthesis`</span>
<span class="sd">                - :py:attr:`~.KM_aa_export`</span>
<span class="sd">                - :py:attr:`~.fraction_supply_rate`</span>
<span class="sd">                - :py:attr:`~.fraction_import_rate`</span>

<span class="sd">        Assumptions:</span>
<span class="sd">                - Each internal amino acid concentration in &#39;minimal_plus_amino_acids&#39;</span>
<span class="sd">                  media is not lower than in &#39;minimal&#39; media</span>

<span class="sd">        TODO (Travis):</span>
<span class="sd">                Better handling of concentration assumption</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">aa_ids</span> <span class="o">=</span> <span class="n">sim_data</span><span class="o">.</span><span class="n">molecule_groups</span><span class="o">.</span><span class="n">amino_acids</span>
        <span class="n">conc</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">concentration_updates</span><span class="o">.</span><span class="n">concentrations_based_on_nutrients</span>

        <span class="n">aa_conc_basal</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span>
            <span class="p">[</span>
                <span class="n">conc</span><span class="p">(</span><span class="n">media_id</span><span class="o">=</span><span class="s2">&quot;minimal&quot;</span><span class="p">)[</span><span class="n">aa</span><span class="p">]</span><span class="o">.</span><span class="n">asNumber</span><span class="p">(</span><span class="n">METABOLITE_CONCENTRATION_UNITS</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">aa</span> <span class="ow">in</span> <span class="n">aa_ids</span>
            <span class="p">]</span>
        <span class="p">)</span>
        <span class="n">aa_conc_aa_media</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span>
            <span class="p">[</span>
                <span class="n">conc</span><span class="p">(</span><span class="n">media_id</span><span class="o">=</span><span class="s2">&quot;minimal_plus_amino_acids&quot;</span><span class="p">)[</span><span class="n">aa</span><span class="p">]</span><span class="o">.</span><span class="n">asNumber</span><span class="p">(</span>
                    <span class="n">METABOLITE_CONCENTRATION_UNITS</span>
                <span class="p">)</span>
                <span class="k">for</span> <span class="n">aa</span> <span class="ow">in</span> <span class="n">aa_ids</span>
            <span class="p">]</span>
        <span class="p">)</span>

        <span class="c1"># Lower concentrations might produce strange rates (excess supply or</span>
        <span class="c1"># negative import when present externally) and constants so raise</span>
        <span class="c1"># to double check the implementation</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">np</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="n">aa_conc_basal</span> <span class="o">&lt;=</span> <span class="n">aa_conc_aa_media</span><span class="p">):</span>
            <span class="n">aas</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">aa_ids</span><span class="p">)[</span><span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">aa_conc_basal</span> <span class="o">&gt;</span> <span class="n">aa_conc_aa_media</span><span class="p">)]</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;Check that amino acid concentrations should be lower in amino acid media for </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                    <span class="n">aas</span>
                <span class="p">)</span>
            <span class="p">)</span>

        <span class="n">f_inhibited</span> <span class="o">=</span> <span class="n">sim_data</span><span class="o">.</span><span class="n">constants</span><span class="o">.</span><span class="n">fraction_supply_inhibited</span>
        <span class="n">f_exported</span> <span class="o">=</span> <span class="n">sim_data</span><span class="o">.</span><span class="n">constants</span><span class="o">.</span><span class="n">fraction_supply_exported</span>

        <span class="c1"># Assumed units of METABOLITE_CONCENTRATION_UNITS for KI and KM</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">KI_aa_synthesis</span> <span class="o">=</span> <span class="n">f_inhibited</span> <span class="o">*</span> <span class="n">aa_conc_basal</span> <span class="o">/</span> <span class="p">(</span><span class="mi">1</span> <span class="o">-</span> <span class="n">f_inhibited</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">KM_aa_export</span> <span class="o">=</span> <span class="p">(</span><span class="mi">1</span> <span class="o">/</span> <span class="n">f_exported</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="o">*</span> <span class="n">aa_conc_aa_media</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">fraction_supply_rate</span> <span class="o">=</span> <span class="p">(</span>
            <span class="mi">1</span> <span class="o">-</span> <span class="n">f_inhibited</span> <span class="o">+</span> <span class="n">aa_conc_basal</span> <span class="o">/</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">KM_aa_export</span> <span class="o">+</span> <span class="n">aa_conc_basal</span><span class="p">)</span>
        <span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">fraction_import_rate</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">-</span> <span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">fraction_supply_rate</span>
            <span class="o">+</span> <span class="mi">1</span> <span class="o">/</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="n">aa_conc_aa_media</span> <span class="o">/</span> <span class="bp">self</span><span class="o">.</span><span class="n">KI_aa_synthesis</span><span class="p">)</span>
            <span class="o">-</span> <span class="n">f_exported</span>
        <span class="p">)</span></div>


<div class="viewcode-block" id="Metabolism.aa_supply_scaling">
<a class="viewcode-back" href="../../../../../reference/api/reconstruction/reconstruction.ecoli.dataclasses.process.metabolism.html#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.aa_supply_scaling">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">aa_supply_scaling</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">aa_conc</span><span class="p">:</span> <span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">],</span> <span class="n">aa_present</span><span class="p">:</span> <span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">bool_</span><span class="p">]</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Called during polypeptide_elongation process</span>
<span class="sd">        Determine amino acid supply rate scaling based on current amino acid</span>
<span class="sd">        concentrations.</span>

<span class="sd">        Args:</span>
<span class="sd">                aa_conc: internal concentration for each amino acid (ndarray[float])</span>
<span class="sd">                aa_present: whether each amino acid is in the</span>
<span class="sd">                        external environment or not (ndarray[bool])</span>

<span class="sd">        Returns:</span>
<span class="sd">                Scaling for the supply of each amino acid with</span>
<span class="sd">                higher supply rate if &gt;1, lower supply rate if &lt;1</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">aa_supply</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">fraction_supply_rate</span>
        <span class="n">aa_import</span> <span class="o">=</span> <span class="n">aa_present</span> <span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">fraction_import_rate</span>
        <span class="n">aa_synthesis</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">/</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="n">aa_conc</span> <span class="o">/</span> <span class="bp">self</span><span class="o">.</span><span class="n">KI_aa_synthesis</span><span class="p">)</span>
        <span class="n">aa_export</span> <span class="o">=</span> <span class="n">aa_conc</span> <span class="o">/</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">KM_aa_export</span> <span class="o">+</span> <span class="n">aa_conc</span><span class="p">)</span>
        <span class="n">supply_scaling</span> <span class="o">=</span> <span class="n">aa_supply</span> <span class="o">+</span> <span class="n">aa_import</span> <span class="o">+</span> <span class="n">aa_synthesis</span> <span class="o">-</span> <span class="n">aa_export</span>

        <span class="k">return</span> <span class="n">supply_scaling</span></div>


<div class="viewcode-block" id="Metabolism.get_aa_to_transporters_mapping_data">
<a class="viewcode-back" href="../../../../../reference/api/reconstruction/reconstruction.ecoli.dataclasses.process.metabolism.html#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.get_aa_to_transporters_mapping_data">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">get_aa_to_transporters_mapping_data</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">sim_data</span><span class="p">:</span> <span class="s2">&quot;SimulationDataEcoli&quot;</span><span class="p">,</span> <span class="n">export</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">tuple</span><span class="p">[</span><span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">list</span><span class="p">],</span> <span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">],</span> <span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">str_</span><span class="p">]]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Creates a dictionary that maps amino acids with their transporters.</span>
<span class="sd">        Based on this dictionary, it creates a correlation matrix with rows</span>
<span class="sd">        as AA and columns as transporters.</span>

<span class="sd">        Args:</span>
<span class="sd">                sim_data: simulation data</span>
<span class="sd">                export: if True, the parameters calculated are for mechanistic</span>
<span class="sd">                        export instead of uptake</span>

<span class="sd">        Returns:</span>
<span class="sd">                3-element tuple containing</span>

<span class="sd">                        - aa_to_transporters: dictonary that maps aa to</span>
<span class="sd">                          transporters involved in transport reactions</span>
<span class="sd">                        - aa_to_transporters_matrix: correlation matrix.</span>
<span class="sd">                          Columns correspond to transporter enzymes and rows to</span>
<span class="sd">                          amino acids</span>
<span class="sd">                        - aa_transporters_names: names of all transporters</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">def</span><span class="w"> </span><span class="nf">matches_direction</span><span class="p">(</span><span class="n">direction</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">export</span><span class="p">:</span>
                <span class="k">return</span> <span class="n">direction</span> <span class="o">&lt;</span> <span class="mi">0</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">return</span> <span class="n">direction</span> <span class="o">&gt;</span> <span class="mi">0</span>

        <span class="c1"># Mapping aminoacids to their transporters</span>
        <span class="c1"># CYS does not have any uptake reaction, so we initialize the dict with it to ensure</span>
        <span class="c1"># the presence of the 21 AAs</span>
        <span class="c1"># TODO (Santiago): Reversible reactions?</span>
        <span class="n">aa_to_transporters</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;CYS[c]&quot;</span><span class="p">:</span> <span class="p">[]}</span>
        <span class="k">for</span> <span class="n">reaction</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">transport_reactions</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">aa</span> <span class="ow">in</span> <span class="n">sim_data</span><span class="o">.</span><span class="n">molecule_groups</span><span class="o">.</span><span class="n">amino_acids</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">aa</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">reaction_stoich</span><span class="p">[</span><span class="n">reaction</span><span class="p">]</span> <span class="ow">and</span> <span class="n">matches_direction</span><span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">reaction_stoich</span><span class="p">[</span><span class="n">reaction</span><span class="p">][</span><span class="n">aa</span><span class="p">]</span>
                <span class="p">):</span>
                    <span class="k">if</span> <span class="n">aa</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">aa_to_transporters</span><span class="p">:</span>
                        <span class="n">aa_to_transporters</span><span class="p">[</span><span class="n">aa</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
                    <span class="n">aa_to_transporters</span><span class="p">[</span><span class="n">aa</span><span class="p">]</span> <span class="o">+=</span> <span class="bp">self</span><span class="o">.</span><span class="n">reaction_catalysts</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">reaction</span><span class="p">,</span> <span class="p">[])</span>

        <span class="n">aa_to_transporters</span> <span class="o">=</span> <span class="p">{</span>
            <span class="n">aa</span><span class="p">:</span> <span class="n">aa_to_transporters</span><span class="p">[</span><span class="n">aa</span><span class="p">]</span> <span class="k">for</span> <span class="n">aa</span> <span class="ow">in</span> <span class="n">sim_data</span><span class="o">.</span><span class="n">molecule_groups</span><span class="o">.</span><span class="n">amino_acids</span>
        <span class="p">}</span>

        <span class="n">c</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">transporters_to_idx</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">aa</span><span class="p">,</span> <span class="n">transporters</span> <span class="ow">in</span> <span class="n">aa_to_transporters</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="k">for</span> <span class="n">transporter</span> <span class="ow">in</span> <span class="n">transporters</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">transporter</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">transporters_to_idx</span><span class="p">:</span>
                    <span class="n">transporters_to_idx</span><span class="p">[</span><span class="n">transporter</span><span class="p">]</span> <span class="o">=</span> <span class="n">c</span>
                    <span class="n">c</span> <span class="o">+=</span> <span class="mi">1</span>

        <span class="n">aa_to_transporters_matrix</span> <span class="o">=</span> <span class="p">[[</span><span class="mi">0</span><span class="p">]]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">aa_to_transporters</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">trnspts</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">aa_to_transporters</span><span class="o">.</span><span class="n">values</span><span class="p">()):</span>
            <span class="n">temp</span> <span class="o">=</span> <span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">transporters_to_idx</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">tr</span> <span class="ow">in</span> <span class="n">trnspts</span><span class="p">:</span>
                <span class="n">temp</span><span class="p">[</span><span class="n">transporters_to_idx</span><span class="p">[</span><span class="n">tr</span><span class="p">]]</span> <span class="o">=</span> <span class="mi">1</span>
            <span class="n">aa_to_transporters_matrix</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">temp</span>

        <span class="n">aa_transporters_names</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">transporters_to_idx</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>

        <span class="k">return</span> <span class="p">(</span>
            <span class="n">aa_to_transporters</span><span class="p">,</span>
            <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">aa_to_transporters_matrix</span><span class="p">),</span>
            <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">aa_transporters_names</span><span class="p">),</span>
        <span class="p">)</span></div>


<div class="viewcode-block" id="Metabolism.set_mechanistic_export_constants">
<a class="viewcode-back" href="../../../../../reference/api/reconstruction/reconstruction.ecoli.dataclasses.process.metabolism.html#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.set_mechanistic_export_constants">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">set_mechanistic_export_constants</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">sim_data</span><span class="p">:</span> <span class="s2">&quot;SimulationDataEcoli&quot;</span><span class="p">,</span>
        <span class="n">cell_specs</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">dict</span><span class="p">],</span>
        <span class="n">basal_container</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
    <span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Calls get_aa_to_transporters_mapping_data() for AA export, which calculates</span>
<span class="sd">        the total amount of export transporter counts per AA. Kcats are calculated using</span>
<span class="sd">        the same exchange rates as for uptake and transporter counts. Missing KMs are calculated</span>
<span class="sd">        based on present KMs. This is done by calculating the average factor for</span>
<span class="sd">        KMs compared to estimated concentration (av_factor = sum(KM / concentration) / n_aa_with_kms).</span>
<span class="sd">        ** KM = av_factor * concentration</span>


<span class="sd">        Args:</span>
<span class="sd">                sim_data: simulation data</span>
<span class="sd">                cell_specs: mapping from condition to calculated cell properties</span>
<span class="sd">                basal_container: average initial bulk molecule counts in the basal</span>
<span class="sd">                        condition (structured Numpy array, see :ref:`bulk`)</span>

<span class="sd">        Attributes set:</span>
<span class="sd">                - :py:attr:`~.aa_to_exporters`</span>
<span class="sd">                - :py:attr:`~.aa_to_exporters_matrix`</span>
<span class="sd">                - :py:attr:`~.aa_exporter_names`</span>
<span class="sd">                - :py:attr:`~.aa_export_kms`</span>
<span class="sd">                - :py:attr:`~.export_kcats_per_aa`</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">aa_to_exporters</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">aa_to_exporters_matrix</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">aa_exporter_names</span> <span class="o">=</span> <span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">get_aa_to_transporters_mapping_data</span><span class="p">(</span><span class="n">sim_data</span><span class="p">,</span> <span class="n">export</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="p">)</span>

        <span class="n">aa_names</span> <span class="o">=</span> <span class="n">sim_data</span><span class="o">.</span><span class="n">molecule_groups</span><span class="o">.</span><span class="n">amino_acids</span>
        <span class="n">aa_idx</span> <span class="o">=</span> <span class="n">bulk_name_to_idx</span><span class="p">(</span><span class="n">aa_names</span><span class="p">,</span> <span class="n">basal_container</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">])</span>
        <span class="n">counts_to_molar</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">sim_data</span><span class="o">.</span><span class="n">constants</span><span class="o">.</span><span class="n">cell_density</span> <span class="o">/</span> <span class="n">cell_specs</span><span class="p">[</span><span class="s2">&quot;basal&quot;</span><span class="p">][</span><span class="s2">&quot;avgCellDryMassInit&quot;</span><span class="p">]</span>
        <span class="p">)</span> <span class="o">/</span> <span class="n">sim_data</span><span class="o">.</span><span class="n">constants</span><span class="o">.</span><span class="n">n_avogadro</span>
        <span class="n">aa_conc</span> <span class="o">=</span> <span class="p">{</span>
            <span class="n">aa</span><span class="p">:</span> <span class="n">counts</span> <span class="o">*</span> <span class="n">counts_to_molar</span>
            <span class="k">for</span> <span class="n">aa</span><span class="p">,</span> <span class="n">counts</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">aa_names</span><span class="p">,</span> <span class="n">counts</span><span class="p">(</span><span class="n">basal_container</span><span class="p">,</span> <span class="n">aa_idx</span><span class="p">))</span>
        <span class="p">}</span>
        <span class="n">aa_with_km</span> <span class="o">=</span> <span class="p">{}</span>

        <span class="c1"># Calculate average factor to estimate missing KMs</span>
        <span class="n">coeff_estimate_kms</span> <span class="o">=</span> <span class="mf">0.0</span>
        <span class="k">for</span> <span class="n">export_kms</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">amino_acid_export_kms</span><span class="p">:</span>
            <span class="n">aa_with_km</span><span class="p">[</span><span class="n">export_kms</span><span class="p">[</span><span class="s2">&quot;Amino Acid&quot;</span><span class="p">]]</span> <span class="o">=</span> <span class="n">export_kms</span><span class="p">[</span><span class="s2">&quot;KM&quot;</span><span class="p">]</span>
            <span class="n">coef_per_aa</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="k">for</span> <span class="n">km</span> <span class="ow">in</span> <span class="n">export_kms</span><span class="p">[</span><span class="s2">&quot;KM&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
                <span class="n">coef_per_aa</span> <span class="o">+=</span> <span class="p">(</span>
                    <span class="n">km</span><span class="o">.</span><span class="n">asUnit</span><span class="p">(</span><span class="n">METABOLITE_CONCENTRATION_UNITS</span><span class="p">)</span>
                    <span class="o">/</span> <span class="n">aa_conc</span><span class="p">[</span><span class="n">export_kms</span><span class="p">[</span><span class="s2">&quot;Amino Acid&quot;</span><span class="p">]]</span>
                <span class="p">)</span>
            <span class="n">coeff_estimate_kms</span> <span class="o">+=</span> <span class="n">coef_per_aa</span> <span class="o">/</span> <span class="nb">len</span><span class="p">(</span><span class="n">export_kms</span><span class="p">[</span><span class="s2">&quot;KM&quot;</span><span class="p">])</span>
        <span class="n">coeff_estimate_kms</span> <span class="o">=</span> <span class="n">coeff_estimate_kms</span> <span class="o">/</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">amino_acid_export_kms</span><span class="p">)</span>

        <span class="c1"># Calculate estimated KMs for each AA</span>
        <span class="n">single_kms</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">aa</span> <span class="ow">in</span> <span class="n">aa_names</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">aa</span> <span class="ow">in</span> <span class="n">aa_with_km</span><span class="p">:</span>
                <span class="n">single_kms</span><span class="p">[</span><span class="n">aa</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">aa_with_km</span><span class="p">[</span><span class="n">aa</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">()))</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">single_kms</span><span class="p">[</span><span class="n">aa</span><span class="p">]</span> <span class="o">=</span> <span class="n">coeff_estimate_kms</span> <span class="o">*</span> <span class="n">aa_conc</span><span class="p">[</span><span class="n">aa</span><span class="p">]</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">aa_export_kms</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span>
            <span class="p">[</span><span class="n">single_kms</span><span class="p">[</span><span class="n">aa</span><span class="p">]</span><span class="o">.</span><span class="n">asNumber</span><span class="p">(</span><span class="n">METABOLITE_CONCENTRATION_UNITS</span><span class="p">)</span> <span class="k">for</span> <span class="n">aa</span> <span class="ow">in</span> <span class="n">aa_names</span><span class="p">]</span>
        <span class="p">)</span></div>


<div class="viewcode-block" id="Metabolism.set_mechanistic_uptake_constants">
<a class="viewcode-back" href="../../../../../reference/api/reconstruction/reconstruction.ecoli.dataclasses.process.metabolism.html#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.set_mechanistic_uptake_constants">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">set_mechanistic_uptake_constants</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">sim_data</span><span class="p">:</span> <span class="s2">&quot;SimulationDataEcoli&quot;</span><span class="p">,</span>
        <span class="n">cell_specs</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">dict</span><span class="p">],</span>
        <span class="n">with_aa_container</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
    <span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Based on the matrix calculated in get_aa_to_transporters_mapping_data(),</span>
<span class="sd">        we calculate the total amount of transporter counts per AA.</span>

<span class="sd">        Args:</span>
<span class="sd">                sim_data: simulation data</span>
<span class="sd">                cell_specs: mapping from condition to calculated cell properties</span>
<span class="sd">                with_aa_container: average initial bulk molecule counts in the</span>
<span class="sd">                        ``with_aa`` condition (structured Numpy array, see :ref:`bulk`)</span>

<span class="sd">        Attributes set:</span>
<span class="sd">                - :py:attr:`~.aa_to_importers`</span>
<span class="sd">                - :py:attr:`~.aa_to_importers_matrix`</span>
<span class="sd">                - :py:attr:`~.aa_importer_names`</span>
<span class="sd">                - :py:attr:`~.import_kcats_per_aa`</span>
<span class="sd">                - :py:attr:`~.export_kcats_per_aa`</span>

<span class="sd">        TODO:</span>
<span class="sd">                - Include external amino acid concentrations and KM values</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">aa_names</span> <span class="o">=</span> <span class="n">sim_data</span><span class="o">.</span><span class="n">molecule_groups</span><span class="o">.</span><span class="n">amino_acids</span>
        <span class="n">aa_idx</span> <span class="o">=</span> <span class="n">bulk_name_to_idx</span><span class="p">(</span><span class="n">aa_names</span><span class="p">,</span> <span class="n">with_aa_container</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">])</span>
        <span class="n">counts_to_molar</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">sim_data</span><span class="o">.</span><span class="n">constants</span><span class="o">.</span><span class="n">cell_density</span>
            <span class="o">/</span> <span class="n">cell_specs</span><span class="p">[</span><span class="s2">&quot;with_aa&quot;</span><span class="p">][</span><span class="s2">&quot;avgCellDryMassInit&quot;</span><span class="p">]</span>
        <span class="p">)</span> <span class="o">/</span> <span class="n">sim_data</span><span class="o">.</span><span class="n">constants</span><span class="o">.</span><span class="n">n_avogadro</span>
        <span class="n">aa_conc</span> <span class="o">=</span> <span class="n">counts</span><span class="p">(</span><span class="n">with_aa_container</span><span class="p">,</span> <span class="n">aa_idx</span><span class="p">)</span> <span class="o">*</span> <span class="n">counts_to_molar</span><span class="o">.</span><span class="n">asNumber</span><span class="p">(</span>
            <span class="n">METABOLITE_CONCENTRATION_UNITS</span>
        <span class="p">)</span>
        <span class="n">exchange_rates</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">specific_import_rates</span> <span class="o">*</span> <span class="n">cell_specs</span><span class="p">[</span><span class="s2">&quot;with_aa&quot;</span><span class="p">][</span>
            <span class="s2">&quot;avgCellDryMassInit&quot;</span>
        <span class="p">]</span><span class="o">.</span><span class="n">asNumber</span><span class="p">(</span><span class="n">units</span><span class="o">.</span><span class="n">fg</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">aa_to_importers</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">aa_to_importers_matrix</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">aa_importer_names</span> <span class="o">=</span> <span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">get_aa_to_transporters_mapping_data</span><span class="p">(</span><span class="n">sim_data</span><span class="p">)</span>
        <span class="p">)</span>

        <span class="n">aa_importer_idx</span> <span class="o">=</span> <span class="n">bulk_name_to_idx</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">aa_importer_names</span><span class="p">,</span> <span class="n">with_aa_container</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]</span>
        <span class="p">)</span>
        <span class="n">importer_counts</span> <span class="o">=</span> <span class="n">counts</span><span class="p">(</span><span class="n">with_aa_container</span><span class="p">,</span> <span class="n">aa_importer_idx</span><span class="p">)</span>
        <span class="n">aa_exporter_idx</span> <span class="o">=</span> <span class="n">bulk_name_to_idx</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">aa_exporter_names</span><span class="p">,</span> <span class="n">with_aa_container</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]</span>
        <span class="p">)</span>
        <span class="n">exporter_counts</span> <span class="o">=</span> <span class="n">counts</span><span class="p">(</span><span class="n">with_aa_container</span><span class="p">,</span> <span class="n">aa_exporter_idx</span><span class="p">)</span>
        <span class="n">counts_per_aa_import</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">aa_to_importers_matrix</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">importer_counts</span><span class="p">)</span>
        <span class="n">counts_per_aa_export</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">aa_to_exporters_matrix</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">exporter_counts</span><span class="p">)</span>

        <span class="c1"># Solve for the two unknown kcats with the calculated net exchange rate</span>
        <span class="c1"># in rich media conditions and the assumption that import and export</span>
        <span class="c1"># rates are equal at the export KM based on how the export KM values</span>
        <span class="c1"># were curated.</span>
        <span class="c1"># Import will decrease and export will increase with higher amino acids</span>
        <span class="c1"># for stable amino acid concentrations.</span>
        <span class="n">import_saturation_in_rich</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">/</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="n">aa_conc</span> <span class="o">/</span> <span class="bp">self</span><span class="o">.</span><span class="n">aa_import_kis</span><span class="p">)</span>
        <span class="n">export_saturation_in_rich</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">/</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">aa_export_kms</span> <span class="o">/</span> <span class="n">aa_conc</span><span class="p">)</span>
        <span class="n">import_saturation_at_km</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">/</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">aa_export_kms</span> <span class="o">/</span> <span class="bp">self</span><span class="o">.</span><span class="n">aa_import_kis</span><span class="p">)</span>
        <span class="n">export_saturation_at_km</span> <span class="o">=</span> <span class="mf">0.5</span>

        <span class="n">import_capacity_at_km</span> <span class="o">=</span> <span class="n">counts_per_aa_import</span> <span class="o">*</span> <span class="n">import_saturation_at_km</span>
        <span class="n">export_capacity_at_km</span> <span class="o">=</span> <span class="n">counts_per_aa_export</span> <span class="o">*</span> <span class="n">export_saturation_at_km</span>
        <span class="k">with</span> <span class="n">np</span><span class="o">.</span><span class="n">errstate</span><span class="p">(</span><span class="n">divide</span><span class="o">=</span><span class="s2">&quot;ignore&quot;</span><span class="p">,</span> <span class="n">invalid</span><span class="o">=</span><span class="s2">&quot;ignore&quot;</span><span class="p">):</span>
            <span class="n">import_vs_export_kcat</span> <span class="o">=</span> <span class="n">export_capacity_at_km</span> <span class="o">/</span> <span class="n">import_capacity_at_km</span>
            <span class="n">kcat_export</span> <span class="o">=</span> <span class="n">exchange_rates</span> <span class="o">/</span> <span class="p">(</span>
                <span class="n">import_vs_export_kcat</span> <span class="o">*</span> <span class="n">counts_per_aa_import</span> <span class="o">*</span> <span class="n">import_saturation_in_rich</span>
                <span class="o">-</span> <span class="n">counts_per_aa_export</span> <span class="o">*</span> <span class="n">export_saturation_in_rich</span>
            <span class="p">)</span>
            <span class="n">kcat_export</span><span class="p">[</span><span class="o">~</span><span class="n">np</span><span class="o">.</span><span class="n">isfinite</span><span class="p">(</span><span class="n">kcat_export</span><span class="p">)]</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="n">kcat_import</span> <span class="o">=</span> <span class="n">import_vs_export_kcat</span> <span class="o">*</span> <span class="n">kcat_export</span>
            <span class="n">kcat_import</span><span class="p">[</span><span class="o">~</span><span class="n">np</span><span class="o">.</span><span class="n">isfinite</span><span class="p">(</span><span class="n">kcat_import</span><span class="p">)]</span> <span class="o">=</span> <span class="mi">0</span>

        <span class="k">if</span> <span class="n">np</span><span class="o">.</span><span class="n">any</span><span class="p">(</span><span class="n">kcat_export</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">)</span> <span class="ow">or</span> <span class="n">np</span><span class="o">.</span><span class="n">any</span><span class="p">(</span><span class="n">kcat_import</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;Could not solve for positive transport kcat.&quot;</span>
                <span class="s2">&quot; Check assumptions or amino acid concentrations compared to KMs.&quot;</span>
            <span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">export_kcats_per_aa</span> <span class="o">=</span> <span class="n">kcat_export</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">import_kcats_per_aa</span> <span class="o">=</span> <span class="n">kcat_import</span></div>


<div class="viewcode-block" id="Metabolism.set_mechanistic_supply_constants">
<a class="viewcode-back" href="../../../../../reference/api/reconstruction/reconstruction.ecoli.dataclasses.process.metabolism.html#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.set_mechanistic_supply_constants">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">set_mechanistic_supply_constants</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">sim_data</span><span class="p">:</span> <span class="s2">&quot;SimulationDataEcoli&quot;</span><span class="p">,</span>
        <span class="n">cell_specs</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">dict</span><span class="p">],</span>
        <span class="n">basal_container</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
        <span class="n">with_aa_container</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
    <span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Sets constants to determine amino acid supply during translation.  Used</span>
<span class="sd">        with amino_acid_synthesis() and amino_acid_import() during simulations</span>
<span class="sd">        but supply can alternatively be determined phenomologically.  This</span>
<span class="sd">        approach is more detailed and should better respond to environmental</span>
<span class="sd">        changes and perturbations but has more variability related to gene</span>
<span class="sd">        expression and regulation.</span>

<span class="sd">        Args:</span>
<span class="sd">                sim_data: simulation data</span>
<span class="sd">                cell_specs: mapping from condition to calculated cell properties</span>
<span class="sd">                basal_container: average initial bulk molecule counts in the basal</span>
<span class="sd">                        condition (structured Numpy array, see :ref:`bulk`)</span>
<span class="sd">                with_aa_container: average initial bulk molecule counts in the</span>
<span class="sd">                        ``with_aa`` condition</span>

<span class="sd">        Sets class attributes:</span>
<span class="sd">                - :py:attr:`~.aa_enzymes`</span>
<span class="sd">                - :py:attr:`~.aa_kcats_fwd`</span>
<span class="sd">                - :py:attr:`~.aa_kcats_rev`</span>
<span class="sd">                - :py:attr:`~.aa_kis`</span>
<span class="sd">                - :py:attr:`~.aa_upstream_kms`</span>
<span class="sd">                - :py:attr:`~.aa_reverse_kms`</span>
<span class="sd">                - :py:attr:`~.aa_upstream_mapping`</span>
<span class="sd">                - :py:attr:`~.enzyme_to_amino_acid`</span>
<span class="sd">                - :py:attr:`~.aa_forward_stoich`</span>
<span class="sd">                - :py:attr:`~.aa_reverse_stoich`</span>
<span class="sd">                - :py:attr:`~.aa_import_kis`</span>
<span class="sd">                - :py:attr:`~.specific_import_rates`</span>
<span class="sd">                - :py:attr:`~.max_specific_import_rates`</span>

<span class="sd">        Assumptions:</span>

<span class="sd">                - Only one reaction is limiting in an amino acid pathway (typically</span>
<span class="sd">                  the first and one with KI) and the kcat for forward or reverse</span>
<span class="sd">                  directions will apply to all enzymes that can catalyze that step</span>
<span class="sd">                - kcat for reverse and degradation reactions is the same (each amino</span>
<span class="sd">                  acid only has reverse or degradation at this point but that could</span>
<span class="sd">                  change with modifications to the amino_acid_pathways flat file)</span>

<span class="sd">        TODO:</span>

<span class="sd">                - Search for new kcat/KM values in literature or use metabolism_kinetics.tsv</span>
<span class="sd">                - Consider multiple reaction steps</span>
<span class="sd">                - Include mulitple amino acid inhibition on importers (currently</span>
<span class="sd">                  amino acids only inhibit their own import but some transporters</span>
<span class="sd">                  import multiple amino acids and will be inhibited by all of the</span>
<span class="sd">                  amino acids for the import of each amino acid)</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">aa_ids</span> <span class="o">=</span> <span class="n">sim_data</span><span class="o">.</span><span class="n">molecule_groups</span><span class="o">.</span><span class="n">amino_acids</span>
        <span class="n">n_aas</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">aa_ids</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">aa_to_index</span> <span class="o">=</span> <span class="p">{</span><span class="n">aa</span><span class="p">:</span> <span class="n">i</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">aa</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">aa_ids</span><span class="p">)}</span>
        <span class="n">conc</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">concentration_updates</span><span class="o">.</span><span class="n">concentrations_based_on_nutrients</span>

        <span class="c1"># Measured data used as targets for calculations</span>
        <span class="n">measured_uptake_rates</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">fwd_kcat_targets</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">aa</span> <span class="ow">in</span> <span class="n">aa_ids</span><span class="p">:</span>
            <span class="n">aa_no_tag</span> <span class="o">=</span> <span class="n">aa</span><span class="p">[:</span><span class="o">-</span><span class="mi">3</span><span class="p">]</span>
            <span class="n">measured_uptake_rates</span><span class="p">[</span><span class="n">aa_no_tag</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">amino_acid_uptake_rates</span><span class="p">[</span><span class="n">aa_no_tag</span><span class="p">][</span><span class="s2">&quot;uptake&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">asNumber</span><span class="p">(</span>
                    <span class="n">units</span><span class="o">.</span><span class="n">mmol</span> <span class="o">/</span> <span class="n">units</span><span class="o">.</span><span class="n">g</span> <span class="o">/</span> <span class="n">units</span><span class="o">.</span><span class="n">h</span>
                <span class="p">)</span>
                <span class="k">if</span> <span class="n">aa_no_tag</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">amino_acid_uptake_rates</span>
                <span class="k">else</span> <span class="mi">0</span>
            <span class="p">)</span>
            <span class="k">if</span> <span class="p">(</span>
                <span class="n">kcat</span> <span class="o">:=</span> <span class="bp">self</span><span class="o">.</span><span class="n">aa_synthesis_pathways</span><span class="p">[</span><span class="n">aa</span><span class="p">][</span><span class="s2">&quot;kcat_data&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">asNumber</span><span class="p">(</span>
                    <span class="n">K_CAT_UNITS</span>
                <span class="p">)</span>
            <span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">fwd_kcat_targets</span><span class="p">[</span><span class="n">aa_no_tag</span><span class="p">]</span> <span class="o">=</span> <span class="n">kcat</span>
        <span class="n">default_fwd_target</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">fwd_kcat_targets</span><span class="o">.</span><span class="n">values</span><span class="p">()))</span>

        <span class="c1"># Allosteric inhibition constants to match required supply rate</span>
        <span class="n">basal_rates</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">sim_data</span><span class="o">.</span><span class="n">translation_supply_rate</span><span class="p">[</span><span class="s2">&quot;minimal&quot;</span><span class="p">]</span>
            <span class="o">*</span> <span class="n">cell_specs</span><span class="p">[</span><span class="s2">&quot;basal&quot;</span><span class="p">][</span><span class="s2">&quot;avgCellDryMassInit&quot;</span><span class="p">]</span>
            <span class="o">*</span> <span class="n">sim_data</span><span class="o">.</span><span class="n">constants</span><span class="o">.</span><span class="n">n_avogadro</span>
        <span class="p">)</span><span class="o">.</span><span class="n">asNumber</span><span class="p">(</span><span class="n">K_CAT_UNITS</span><span class="p">)</span>
        <span class="n">with_aa_rates</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">sim_data</span><span class="o">.</span><span class="n">translation_supply_rate</span><span class="p">[</span><span class="s2">&quot;minimal_plus_amino_acids&quot;</span><span class="p">]</span>
            <span class="o">*</span> <span class="n">cell_specs</span><span class="p">[</span><span class="s2">&quot;with_aa&quot;</span><span class="p">][</span><span class="s2">&quot;avgCellDryMassInit&quot;</span><span class="p">]</span>
            <span class="o">*</span> <span class="n">sim_data</span><span class="o">.</span><span class="n">constants</span><span class="o">.</span><span class="n">n_avogadro</span>
        <span class="p">)</span><span class="o">.</span><span class="n">asNumber</span><span class="p">(</span><span class="n">K_CAT_UNITS</span><span class="p">)</span>
        <span class="n">basal_supply_mapping</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">aa_ids</span><span class="p">,</span> <span class="n">basal_rates</span><span class="p">))</span>
        <span class="n">with_aa_supply_mapping</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">aa_ids</span><span class="p">,</span> <span class="n">with_aa_rates</span><span class="p">))</span>
        <span class="n">aa_enzymes</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">enzyme_to_aa_fwd</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">enzyme_to_aa_rev</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">aa_kcats_fwd</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">aa_kcats_rev</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">aa_kis</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">upstream_aas_for_km</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">aa_upstream_kms</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">aa_reverse_kms</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">aa_degradation_kms</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">fwd_rates</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">float</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">rev_rates</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">float</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">deg_rates</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">float</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">calculated_uptake_rates</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">minimal_conc</span> <span class="o">=</span> <span class="n">conc</span><span class="p">(</span><span class="s2">&quot;minimal&quot;</span><span class="p">)</span>
        <span class="n">with_aa_conc</span> <span class="o">=</span> <span class="n">conc</span><span class="p">(</span><span class="s2">&quot;minimal_plus_amino_acids&quot;</span><span class="p">)</span>

        <span class="c1"># Get order of amino acids to calculate parameters for to ensure that</span>
        <span class="c1"># parameters that are dependent on other amino acids are run after</span>
        <span class="c1"># those calculations have completed</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">aa_forward_stoich</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">eye</span><span class="p">(</span><span class="n">n_aas</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">aa_reverse_stoich</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">eye</span><span class="p">(</span><span class="n">n_aas</span><span class="p">)</span>
        <span class="n">dependencies</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">set</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">aa</span> <span class="ow">in</span> <span class="n">aa_ids</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">downstream_aa</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">aa_synthesis_pathways</span><span class="p">[</span><span class="n">aa</span><span class="p">][</span><span class="s2">&quot;downstream&quot;</span><span class="p">]:</span>
                <span class="k">if</span> <span class="n">units</span><span class="o">.</span><span class="n">isfinite</span><span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">aa_synthesis_pathways</span><span class="p">[</span><span class="n">downstream_aa</span><span class="p">][</span><span class="s2">&quot;km, degradation&quot;</span><span class="p">]</span>
                <span class="p">):</span>
                    <span class="n">dependencies</span><span class="o">.</span><span class="n">setdefault</span><span class="p">(</span><span class="n">aa</span><span class="p">,</span> <span class="nb">set</span><span class="p">())</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">downstream_aa</span><span class="p">)</span>

            <span class="c1"># Convert individual supply calculations to overall supply based on dependencies</span>
            <span class="c1"># via dot product (self.aa_forward_stoich @ supply)</span>
            <span class="k">for</span> <span class="n">upstream_aa</span><span class="p">,</span> <span class="n">stoich</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">aa_synthesis_pathways</span><span class="p">[</span><span class="n">aa</span><span class="p">][</span>
                <span class="s2">&quot;upstream&quot;</span>
            <span class="p">]</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">aa_forward_stoich</span><span class="p">[</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">aa_to_index</span><span class="p">[</span><span class="n">upstream_aa</span><span class="p">],</span> <span class="bp">self</span><span class="o">.</span><span class="n">aa_to_index</span><span class="p">[</span><span class="n">aa</span><span class="p">]</span>
                <span class="p">]</span> <span class="o">=</span> <span class="o">-</span><span class="n">stoich</span>
                <span class="n">dependencies</span><span class="o">.</span><span class="n">setdefault</span><span class="p">(</span><span class="n">upstream_aa</span><span class="p">,</span> <span class="nb">set</span><span class="p">())</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">aa</span><span class="p">)</span>

            <span class="k">for</span> <span class="n">reverse_aa</span><span class="p">,</span> <span class="n">stoich</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">aa_synthesis_pathways</span><span class="p">[</span><span class="n">aa</span><span class="p">][</span><span class="s2">&quot;reverse&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">aa_reverse_stoich</span><span class="p">[</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">aa_to_index</span><span class="p">[</span><span class="n">reverse_aa</span><span class="p">],</span> <span class="bp">self</span><span class="o">.</span><span class="n">aa_to_index</span><span class="p">[</span><span class="n">aa</span><span class="p">]</span>
                <span class="p">]</span> <span class="o">=</span> <span class="o">-</span><span class="n">stoich</span>
                <span class="n">dependencies</span><span class="o">.</span><span class="n">setdefault</span><span class="p">(</span><span class="n">reverse_aa</span><span class="p">,</span> <span class="nb">set</span><span class="p">())</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">aa</span><span class="p">)</span>

        <span class="n">ordered_aa_ids</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="n">aa_ids</span><span class="p">:</span>  <span class="c1"># limit number of iterations number of amino acids in case there are cyclic links</span>
            <span class="k">for</span> <span class="n">aa</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">aa_ids</span><span class="p">)</span> <span class="o">-</span> <span class="nb">set</span><span class="p">(</span><span class="n">ordered_aa_ids</span><span class="p">)):</span>
                <span class="k">for</span> <span class="n">downstream_aa</span> <span class="ow">in</span> <span class="n">dependencies</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">aa</span><span class="p">,</span> <span class="nb">set</span><span class="p">()):</span>
                    <span class="k">if</span> <span class="n">downstream_aa</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">ordered_aa_ids</span><span class="p">:</span>
                        <span class="k">break</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">ordered_aa_ids</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">aa</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">ordered_aa_ids</span><span class="p">)</span> <span class="o">!=</span> <span class="n">n_aas</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span>
                <span class="s2">&quot;Could not determine amino acid order to calculate dependencies first.&quot;</span>
                <span class="s2">&quot; Make sure there are no cyclical pathways for amino acids that can degrade.&quot;</span>
            <span class="p">)</span>

        <span class="k">for</span> <span class="n">amino_acid</span> <span class="ow">in</span> <span class="n">ordered_aa_ids</span><span class="p">:</span>
            <span class="n">data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">aa_synthesis_pathways</span><span class="p">[</span><span class="n">amino_acid</span><span class="p">]</span>
            <span class="n">fwd_enzymes</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;enzymes&quot;</span><span class="p">]</span>
            <span class="n">fwd_enzymes_basal_idx</span> <span class="o">=</span> <span class="n">bulk_name_to_idx</span><span class="p">(</span><span class="n">fwd_enzymes</span><span class="p">,</span> <span class="n">basal_container</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">])</span>
            <span class="n">fwd_enzymes_basal</span> <span class="o">=</span> <span class="n">counts</span><span class="p">(</span><span class="n">basal_container</span><span class="p">,</span> <span class="n">fwd_enzymes_basal_idx</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
            <span class="n">fwd_enzymes_with_aa_idx</span> <span class="o">=</span> <span class="n">bulk_name_to_idx</span><span class="p">(</span>
                <span class="n">fwd_enzymes</span><span class="p">,</span> <span class="n">with_aa_container</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]</span>
            <span class="p">)</span>
            <span class="n">fwd_enzymes_with_aa</span> <span class="o">=</span> <span class="n">counts</span><span class="p">(</span>
                <span class="n">with_aa_container</span><span class="p">,</span> <span class="n">fwd_enzymes_with_aa_idx</span>
            <span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
            <span class="n">rev_enzymes</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;reverse enzymes&quot;</span><span class="p">]</span>
            <span class="n">rev_enzymes_basal_idx</span> <span class="o">=</span> <span class="n">bulk_name_to_idx</span><span class="p">(</span><span class="n">rev_enzymes</span><span class="p">,</span> <span class="n">basal_container</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">])</span>
            <span class="n">rev_enzymes_basal</span> <span class="o">=</span> <span class="n">counts</span><span class="p">(</span><span class="n">basal_container</span><span class="p">,</span> <span class="n">rev_enzymes_basal_idx</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
            <span class="n">rev_enzymes_with_aa_idx</span> <span class="o">=</span> <span class="n">bulk_name_to_idx</span><span class="p">(</span>
                <span class="n">rev_enzymes</span><span class="p">,</span> <span class="n">with_aa_container</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]</span>
            <span class="p">)</span>
            <span class="n">rev_enzymes_with_aa</span> <span class="o">=</span> <span class="n">counts</span><span class="p">(</span>
                <span class="n">with_aa_container</span><span class="p">,</span> <span class="n">rev_enzymes_with_aa_idx</span>
            <span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>

            <span class="n">aa_conc_basal</span> <span class="o">=</span> <span class="n">minimal_conc</span><span class="p">[</span><span class="n">amino_acid</span><span class="p">]</span>
            <span class="n">aa_conc_with_aa</span> <span class="o">=</span> <span class="n">with_aa_conc</span><span class="p">[</span><span class="n">amino_acid</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;ki&quot;</span><span class="p">]</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">ki</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">inf</span> <span class="o">*</span> <span class="n">units</span><span class="o">.</span><span class="n">mol</span> <span class="o">/</span> <span class="n">units</span><span class="o">.</span><span class="n">L</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="c1"># Get largest dynamic range possible given the range of measured KIs</span>
                <span class="n">lower_limit</span><span class="p">,</span> <span class="n">upper_limit</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;ki&quot;</span><span class="p">]</span>
                <span class="k">if</span> <span class="n">aa_conc_basal</span> <span class="o">&lt;</span> <span class="n">lower_limit</span><span class="p">:</span>
                    <span class="n">ki</span> <span class="o">=</span> <span class="n">lower_limit</span>
                <span class="k">elif</span> <span class="n">aa_conc_basal</span> <span class="o">&gt;</span> <span class="n">upper_limit</span><span class="p">:</span>
                    <span class="n">ki</span> <span class="o">=</span> <span class="n">upper_limit</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">ki</span> <span class="o">=</span> <span class="n">aa_conc_basal</span>
            <span class="n">upstream_aa</span> <span class="o">=</span> <span class="p">[</span><span class="n">aa</span> <span class="k">for</span> <span class="n">aa</span> <span class="ow">in</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;upstream&quot;</span><span class="p">]]</span>
            <span class="n">km_conc_basal</span> <span class="o">=</span> <span class="n">METABOLITE_CONCENTRATION_UNITS</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span>
                <span class="p">[</span>
                    <span class="n">minimal_conc</span><span class="p">[</span><span class="n">aa</span><span class="p">]</span><span class="o">.</span><span class="n">asNumber</span><span class="p">(</span><span class="n">METABOLITE_CONCENTRATION_UNITS</span><span class="p">)</span>
                    <span class="k">for</span> <span class="n">aa</span> <span class="ow">in</span> <span class="n">upstream_aa</span>
                <span class="p">]</span>
            <span class="p">)</span>
            <span class="n">km_conc_with_aa</span> <span class="o">=</span> <span class="n">METABOLITE_CONCENTRATION_UNITS</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span>
                <span class="p">[</span>
                    <span class="n">with_aa_conc</span><span class="p">[</span><span class="n">aa</span><span class="p">]</span><span class="o">.</span><span class="n">asNumber</span><span class="p">(</span><span class="n">METABOLITE_CONCENTRATION_UNITS</span><span class="p">)</span>
                    <span class="k">for</span> <span class="n">aa</span> <span class="ow">in</span> <span class="n">upstream_aa</span>
                <span class="p">]</span>
            <span class="p">)</span>
            <span class="n">kms_upstream</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;km, upstream&quot;</span><span class="p">]</span>
            <span class="n">kms</span> <span class="o">=</span> <span class="n">METABOLITE_CONCENTRATION_UNITS</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span>
                <span class="p">[</span>
                    <span class="n">kms_upstream</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">aa</span><span class="p">,</span> <span class="n">minimal_conc</span><span class="p">[</span><span class="n">aa</span><span class="p">])</span><span class="o">.</span><span class="n">asNumber</span><span class="p">(</span>
                        <span class="n">METABOLITE_CONCENTRATION_UNITS</span>
                    <span class="p">)</span>
                    <span class="k">for</span> <span class="n">aa</span> <span class="ow">in</span> <span class="n">upstream_aa</span>
                <span class="p">]</span>
            <span class="p">)</span>  <span class="c1"># TODO: better way to fill in this missing data</span>
            <span class="k">if</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;reverse&quot;</span><span class="p">]:</span>
                <span class="k">if</span> <span class="n">units</span><span class="o">.</span><span class="n">isnan</span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;km, reverse&quot;</span><span class="p">]):</span>
                    <span class="n">km_reverse</span> <span class="o">=</span> <span class="p">(</span>
                        <span class="n">minimal_conc</span><span class="p">[</span><span class="n">amino_acid</span><span class="p">]</span> <span class="o">*</span> <span class="mi">10</span>
                    <span class="p">)</span>  <span class="c1"># TODO: better way to fill in this missing data</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">km_reverse</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;km, reverse&quot;</span><span class="p">]</span>
            <span class="c1"># TODO: remove this if separating reverse and deg enzymes</span>
            <span class="k">elif</span> <span class="p">(</span>
                <span class="n">data</span><span class="p">[</span><span class="s2">&quot;reverse enzymes&quot;</span><span class="p">]</span>
                <span class="ow">and</span> <span class="ow">not</span> <span class="n">units</span><span class="o">.</span><span class="n">isfinite</span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;km, degradation&quot;</span><span class="p">])</span>
                <span class="ow">and</span> <span class="n">amino_acid</span> <span class="o">!=</span> <span class="s2">&quot;L-SELENOCYSTEINE[c]&quot;</span>
            <span class="p">):</span>
                <span class="n">km_reverse</span> <span class="o">=</span> <span class="p">(</span>
                    <span class="n">minimal_conc</span><span class="p">[</span><span class="n">amino_acid</span><span class="p">]</span> <span class="o">*</span> <span class="mi">10</span>
                <span class="p">)</span>  <span class="c1"># TODO: better way to fill in this missing data</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">km_reverse</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">inf</span> <span class="o">*</span> <span class="n">units</span><span class="o">.</span><span class="n">mol</span> <span class="o">/</span> <span class="n">units</span><span class="o">.</span><span class="n">L</span>
            <span class="n">km_degradation</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;km, degradation&quot;</span><span class="p">]</span>

            <span class="c1"># Make required adjustments in order to get positive kcats and import rates</span>
            <span class="k">for</span> <span class="n">parameter</span><span class="p">,</span> <span class="n">factor</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">aa_synthesis_pathway_adjustments</span><span class="o">.</span><span class="n">get</span><span class="p">(</span>
                <span class="n">amino_acid</span><span class="p">,</span> <span class="p">{}</span>
            <span class="p">)</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                <span class="k">if</span> <span class="n">parameter</span> <span class="o">==</span> <span class="s2">&quot;ki&quot;</span><span class="p">:</span>
                    <span class="n">ki</span> <span class="o">*=</span> <span class="n">factor</span>
                <span class="k">elif</span> <span class="n">parameter</span> <span class="o">==</span> <span class="s2">&quot;km_degradation&quot;</span><span class="p">:</span>
                    <span class="n">km_degradation</span> <span class="o">*=</span> <span class="n">factor</span>
                <span class="k">elif</span> <span class="n">parameter</span> <span class="o">==</span> <span class="s2">&quot;km_reverse&quot;</span><span class="p">:</span>
                    <span class="n">km_reverse</span> <span class="o">*=</span> <span class="n">factor</span>
                <span class="k">elif</span> <span class="n">parameter</span> <span class="o">==</span> <span class="s2">&quot;kms&quot;</span><span class="p">:</span>
                    <span class="n">kms</span> <span class="o">*=</span> <span class="n">factor</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                        <span class="sa">f</span><span class="s2">&quot;Unexpected parameter adjustment (</span><span class="si">{</span><span class="n">parameter</span><span class="si">}</span><span class="s2">) for </span><span class="si">{</span><span class="n">amino_acid</span><span class="si">}</span><span class="s2">.&quot;</span>
                    <span class="p">)</span>

            <span class="k">if</span> <span class="n">units</span><span class="o">.</span><span class="n">isfinite</span><span class="p">(</span><span class="n">km_reverse</span><span class="p">)</span> <span class="ow">and</span> <span class="n">units</span><span class="o">.</span><span class="n">isfinite</span><span class="p">(</span><span class="n">km_degradation</span><span class="p">):</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                    <span class="s2">&quot;Currently cannot have a reverse and degradation KM for amino acid&quot;</span>
                    <span class="sa">f</span><span class="s2">&quot; synthesis pathways (</span><span class="si">{</span><span class="n">amino_acid</span><span class="si">}</span><span class="s2">).  Consider a method to solve for separate&quot;</span>
                    <span class="s2">&quot; kcats and implement matching saturation calculation in amino_acid_synthesis()&quot;</span>
                    <span class="s2">&quot; and calc_kcats()&quot;</span>
                <span class="p">)</span>

            <span class="k">def</span><span class="w"> </span><span class="nf">calc_kcats</span><span class="p">(</span>
                <span class="n">aa_conc_basal</span><span class="p">,</span>
                <span class="n">km_conc_basal</span><span class="p">,</span>
                <span class="n">aa_conc_with_aa</span><span class="p">,</span>
                <span class="n">km_conc_with_aa</span><span class="p">,</span>
                <span class="n">kms</span><span class="p">,</span>
                <span class="n">km_reverse</span><span class="p">,</span>
                <span class="n">km_degradation</span><span class="p">,</span>
                <span class="n">ki</span><span class="p">,</span>
                <span class="n">uptake_rate</span><span class="p">,</span>
            <span class="p">):</span>
                <span class="c1"># Calculate kcat value to ensure sufficient supply to double</span>
                <span class="n">fwd_fraction_basal</span> <span class="o">=</span> <span class="n">units</span><span class="o">.</span><span class="n">strip_empty_units</span><span class="p">(</span>
                    <span class="mi">1</span>
                    <span class="o">/</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="n">aa_conc_basal</span> <span class="o">/</span> <span class="n">ki</span><span class="p">)</span>
                    <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">prod</span><span class="p">(</span><span class="mi">1</span> <span class="o">/</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="n">kms</span> <span class="o">/</span> <span class="n">km_conc_basal</span><span class="p">))</span>
                <span class="p">)</span>
                <span class="n">rev_fraction_basal</span> <span class="o">=</span> <span class="n">units</span><span class="o">.</span><span class="n">strip_empty_units</span><span class="p">(</span>
                    <span class="mi">1</span>
                    <span class="o">/</span> <span class="p">(</span>
                        <span class="mi">1</span>
                        <span class="o">+</span> <span class="n">km_reverse</span>
                        <span class="o">/</span> <span class="n">aa_conc_basal</span>
                        <span class="o">*</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="n">aa_conc_basal</span> <span class="o">/</span> <span class="n">km_degradation</span><span class="p">)</span>
                    <span class="p">)</span>
                <span class="p">)</span>
                <span class="n">deg_fraction_basal</span> <span class="o">=</span> <span class="n">units</span><span class="o">.</span><span class="n">strip_empty_units</span><span class="p">(</span>
                    <span class="mi">1</span>
                    <span class="o">/</span> <span class="p">(</span>
                        <span class="mi">1</span>
                        <span class="o">+</span> <span class="n">km_degradation</span>
                        <span class="o">/</span> <span class="n">aa_conc_basal</span>
                        <span class="o">*</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="n">aa_conc_basal</span> <span class="o">/</span> <span class="n">km_reverse</span><span class="p">)</span>
                    <span class="p">)</span>
                <span class="p">)</span>
                <span class="n">loss_fraction_basal</span> <span class="o">=</span> <span class="n">rev_fraction_basal</span> <span class="o">+</span> <span class="n">deg_fraction_basal</span>
                <span class="n">fwd_capacity_basal</span> <span class="o">=</span> <span class="n">fwd_enzymes_basal</span> <span class="o">*</span> <span class="n">fwd_fraction_basal</span>
                <span class="n">rev_capacity_basal</span> <span class="o">=</span> <span class="n">rev_enzymes_basal</span> <span class="o">*</span> <span class="n">loss_fraction_basal</span>

                <span class="n">fwd_fraction_with_aa</span> <span class="o">=</span> <span class="n">units</span><span class="o">.</span><span class="n">strip_empty_units</span><span class="p">(</span>
                    <span class="mi">1</span>
                    <span class="o">/</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="n">aa_conc_with_aa</span> <span class="o">/</span> <span class="n">ki</span><span class="p">)</span>
                    <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">prod</span><span class="p">(</span><span class="mi">1</span> <span class="o">/</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="n">kms</span> <span class="o">/</span> <span class="n">km_conc_with_aa</span><span class="p">))</span>
                <span class="p">)</span>
                <span class="n">rev_fraction_with_aa</span> <span class="o">=</span> <span class="n">units</span><span class="o">.</span><span class="n">strip_empty_units</span><span class="p">(</span>
                    <span class="mi">1</span>
                    <span class="o">/</span> <span class="p">(</span>
                        <span class="mi">1</span>
                        <span class="o">+</span> <span class="n">km_reverse</span>
                        <span class="o">/</span> <span class="n">aa_conc_with_aa</span>
                        <span class="o">*</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="n">aa_conc_with_aa</span> <span class="o">/</span> <span class="n">km_degradation</span><span class="p">)</span>
                    <span class="p">)</span>
                <span class="p">)</span>
                <span class="n">deg_fraction_with_aa</span> <span class="o">=</span> <span class="n">units</span><span class="o">.</span><span class="n">strip_empty_units</span><span class="p">(</span>
                    <span class="mi">1</span>
                    <span class="o">/</span> <span class="p">(</span>
                        <span class="mi">1</span>
                        <span class="o">+</span> <span class="n">km_degradation</span>
                        <span class="o">/</span> <span class="n">aa_conc_with_aa</span>
                        <span class="o">*</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="n">aa_conc_with_aa</span> <span class="o">/</span> <span class="n">km_reverse</span><span class="p">)</span>
                    <span class="p">)</span>
                <span class="p">)</span>
                <span class="n">loss_fraction_with_aa</span> <span class="o">=</span> <span class="n">rev_fraction_with_aa</span> <span class="o">+</span> <span class="n">deg_fraction_with_aa</span>
                <span class="n">fwd_capacity_with_aa</span> <span class="o">=</span> <span class="n">fwd_enzymes_with_aa</span> <span class="o">*</span> <span class="n">fwd_fraction_with_aa</span>
                <span class="n">rev_capacity_with_aa</span> <span class="o">=</span> <span class="n">rev_enzymes_with_aa</span> <span class="o">*</span> <span class="n">loss_fraction_with_aa</span>

                <span class="n">supply_basal</span> <span class="o">=</span> <span class="n">basal_supply_mapping</span><span class="p">[</span><span class="n">amino_acid</span><span class="p">]</span>
                <span class="n">supply_with_aa</span> <span class="o">=</span> <span class="n">with_aa_supply_mapping</span><span class="p">[</span><span class="n">amino_acid</span><span class="p">]</span>
                <span class="n">downstream_basal</span> <span class="o">=</span> <span class="mi">0</span>
                <span class="n">downstream_with_aa</span> <span class="o">=</span> <span class="mi">0</span>
                <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">stoich</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">aa_forward_stoich</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">aa_to_index</span><span class="p">[</span><span class="n">amino_acid</span><span class="p">],</span> <span class="p">:]</span>
                <span class="p">):</span>
                    <span class="k">if</span> <span class="n">stoich</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>
                        <span class="n">downstream_aa</span> <span class="o">=</span> <span class="n">aa_ids</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
                        <span class="n">downstream_basal</span> <span class="o">+=</span> <span class="o">-</span><span class="n">stoich</span> <span class="o">*</span> <span class="n">fwd_rates</span><span class="p">[</span><span class="n">downstream_aa</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
                        <span class="n">downstream_with_aa</span> <span class="o">+=</span> <span class="o">-</span><span class="n">stoich</span> <span class="o">*</span> <span class="n">fwd_rates</span><span class="p">[</span><span class="n">downstream_aa</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span>
                <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">stoich</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">aa_reverse_stoich</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">aa_to_index</span><span class="p">[</span><span class="n">amino_acid</span><span class="p">],</span> <span class="p">:]</span>
                <span class="p">):</span>
                    <span class="k">if</span> <span class="n">stoich</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>
                        <span class="n">downstream_aa</span> <span class="o">=</span> <span class="n">aa_ids</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
                        <span class="n">downstream_basal</span> <span class="o">+=</span> <span class="n">stoich</span> <span class="o">*</span> <span class="n">rev_rates</span><span class="p">[</span><span class="n">downstream_aa</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
                        <span class="n">downstream_with_aa</span> <span class="o">+=</span> <span class="n">stoich</span> <span class="o">*</span> <span class="n">rev_rates</span><span class="p">[</span><span class="n">downstream_aa</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span>
                <span class="n">uptake</span> <span class="o">=</span> <span class="p">(</span>
                    <span class="n">units</span><span class="o">.</span><span class="n">mmol</span>
                    <span class="o">/</span> <span class="n">units</span><span class="o">.</span><span class="n">g</span>
                    <span class="o">/</span> <span class="n">units</span><span class="o">.</span><span class="n">h</span>
                    <span class="o">*</span> <span class="n">uptake_rate</span>
                    <span class="o">*</span> <span class="n">cell_specs</span><span class="p">[</span><span class="s2">&quot;with_aa&quot;</span><span class="p">][</span><span class="s2">&quot;avgCellDryMassInit&quot;</span><span class="p">]</span>
                <span class="p">)</span><span class="o">.</span><span class="n">asNumber</span><span class="p">(</span><span class="n">units</span><span class="o">.</span><span class="n">count</span> <span class="o">*</span> <span class="n">K_CAT_UNITS</span><span class="p">)</span>

                <span class="n">balance_basal</span> <span class="o">=</span> <span class="n">supply_basal</span> <span class="o">+</span> <span class="n">downstream_basal</span>
                <span class="n">balance_with_aa</span> <span class="o">=</span> <span class="n">supply_with_aa</span> <span class="o">+</span> <span class="n">downstream_with_aa</span> <span class="o">-</span> <span class="n">uptake</span>
                <span class="n">A</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span>
                    <span class="p">[</span>
                        <span class="p">[</span><span class="n">fwd_capacity_basal</span><span class="p">,</span> <span class="o">-</span><span class="n">rev_capacity_basal</span><span class="p">],</span>
                        <span class="p">[</span><span class="n">fwd_capacity_with_aa</span><span class="p">,</span> <span class="o">-</span><span class="n">rev_capacity_with_aa</span><span class="p">],</span>
                    <span class="p">]</span>
                <span class="p">)</span>
                <span class="n">b</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">balance_basal</span><span class="p">,</span> <span class="n">balance_with_aa</span><span class="p">])</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="n">kcat_fwd</span><span class="p">,</span> <span class="n">kcat_rev</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">linalg</span><span class="o">.</span><span class="n">solve</span><span class="p">(</span><span class="n">A</span><span class="p">,</span> <span class="n">b</span><span class="p">)</span>
                <span class="k">except</span> <span class="n">np</span><span class="o">.</span><span class="n">linalg</span><span class="o">.</span><span class="n">LinAlgError</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">VERBOSE</span><span class="p">:</span>
                        <span class="nb">print</span><span class="p">(</span>
                            <span class="sa">f</span><span class="s2">&quot;Warning: could not solve directly for </span><span class="si">{</span><span class="n">amino_acid</span><span class="si">}</span><span class="s2"> kcats - switching to least squares&quot;</span>
                        <span class="p">)</span>
                    <span class="n">kcat_fwd</span><span class="p">,</span> <span class="n">kcat_rev</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">linalg</span><span class="o">.</span><span class="n">lstsq</span><span class="p">(</span><span class="n">A</span><span class="p">,</span> <span class="n">b</span><span class="p">,</span> <span class="n">rcond</span><span class="o">=</span><span class="kc">None</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>

                <span class="n">fwd_rate</span> <span class="o">=</span> <span class="p">(</span>
                    <span class="n">kcat_fwd</span> <span class="o">*</span> <span class="n">fwd_enzymes_basal</span> <span class="o">*</span> <span class="n">fwd_fraction_basal</span><span class="p">,</span>
                    <span class="n">kcat_fwd</span> <span class="o">*</span> <span class="n">fwd_enzymes_with_aa</span> <span class="o">*</span> <span class="n">fwd_fraction_with_aa</span><span class="p">,</span>
                <span class="p">)</span>
                <span class="n">rev_rate</span> <span class="o">=</span> <span class="p">(</span>
                    <span class="n">kcat_rev</span> <span class="o">*</span> <span class="n">rev_enzymes_basal</span> <span class="o">*</span> <span class="n">rev_fraction_basal</span><span class="p">,</span>
                    <span class="n">kcat_rev</span> <span class="o">*</span> <span class="n">rev_enzymes_with_aa</span> <span class="o">*</span> <span class="n">rev_fraction_with_aa</span><span class="p">,</span>
                <span class="p">)</span>
                <span class="n">deg_rate</span> <span class="o">=</span> <span class="p">(</span>
                    <span class="n">kcat_rev</span> <span class="o">*</span> <span class="n">rev_enzymes_basal</span> <span class="o">*</span> <span class="n">deg_fraction_basal</span><span class="p">,</span>
                    <span class="n">kcat_rev</span> <span class="o">*</span> <span class="n">rev_enzymes_with_aa</span> <span class="o">*</span> <span class="n">deg_fraction_with_aa</span><span class="p">,</span>
                <span class="p">)</span>

                <span class="k">return</span> <span class="n">kcat_fwd</span><span class="p">,</span> <span class="n">kcat_rev</span><span class="p">,</span> <span class="n">fwd_rate</span><span class="p">,</span> <span class="n">rev_rate</span><span class="p">,</span> <span class="n">deg_rate</span><span class="p">,</span> <span class="n">uptake</span>

            <span class="c1"># Fit forward and reverse kcats by adjusting the uptake rate</span>
            <span class="k">def</span><span class="w"> </span><span class="nf">objective</span><span class="p">(</span><span class="n">aa</span><span class="p">,</span> <span class="n">uptake</span><span class="p">,</span> <span class="n">kcat_fwd</span><span class="p">,</span> <span class="n">kcat_rev</span><span class="p">):</span>
                <span class="n">aa</span> <span class="o">=</span> <span class="n">aa</span><span class="p">[:</span><span class="o">-</span><span class="mi">3</span><span class="p">]</span>
                <span class="n">diffs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span>
                    <span class="p">[</span>
                        <span class="n">measured_uptake_rates</span><span class="p">[</span><span class="n">aa</span><span class="p">]</span> <span class="o">-</span> <span class="n">uptake</span><span class="p">,</span>
                        <span class="n">fwd_kcat_targets</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">aa</span><span class="p">,</span> <span class="n">default_fwd_target</span><span class="p">)</span> <span class="o">-</span> <span class="n">kcat_fwd</span><span class="p">,</span>
                        <span class="mi">0</span>
                        <span class="k">if</span> <span class="n">units</span><span class="o">.</span><span class="n">isfinite</span><span class="p">(</span><span class="n">km_reverse</span><span class="p">)</span>
                        <span class="k">else</span> <span class="n">kcat_rev</span><span class="p">,</span>  <span class="c1"># no penalty if reverse, minimize if degradation</span>
                    <span class="p">]</span>
                <span class="p">)</span>
                <span class="n">weights</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">1000</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
                <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">linalg</span><span class="o">.</span><span class="n">norm</span><span class="p">(</span><span class="n">weights</span> <span class="o">*</span> <span class="n">diffs</span><span class="p">)</span>

            <span class="c1"># TODO (travis): use a more rigorous method to fit the kcats (eg. gradient descent)</span>
            <span class="c1"># It would be better to include all amino acids in an objective and solve for all</span>
            <span class="c1"># kcats iteratively instead of solving for kcats for a single amino acid and then the next</span>
            <span class="n">best_objective</span> <span class="o">=</span> <span class="kc">None</span>
            <span class="n">kcat_fwd</span> <span class="o">=</span> <span class="kc">None</span>
            <span class="n">n_factors</span> <span class="o">=</span> <span class="mi">500</span>
            <span class="k">if</span> <span class="n">VERBOSE</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;uptake:&quot;</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">factor</span> <span class="ow">in</span> <span class="n">np</span><span class="o">.</span><span class="n">logspace</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">n_factors</span><span class="p">):</span>
                <span class="n">results</span> <span class="o">=</span> <span class="n">calc_kcats</span><span class="p">(</span>
                    <span class="n">aa_conc_basal</span><span class="p">,</span>
                    <span class="n">km_conc_basal</span><span class="p">,</span>
                    <span class="n">aa_conc_with_aa</span><span class="p">,</span>
                    <span class="n">km_conc_with_aa</span><span class="p">,</span>
                    <span class="n">kms</span><span class="p">,</span>
                    <span class="n">km_reverse</span><span class="p">,</span>
                    <span class="n">km_degradation</span><span class="p">,</span>
                    <span class="n">ki</span><span class="p">,</span>
                    <span class="n">factor</span> <span class="o">*</span> <span class="n">measured_uptake_rates</span><span class="p">[</span><span class="n">amino_acid</span><span class="p">[:</span><span class="o">-</span><span class="mi">3</span><span class="p">]],</span>
                <span class="p">)</span>

                <span class="n">new_kcat_fwd</span><span class="p">,</span> <span class="n">new_kcat_rev</span><span class="p">,</span> <span class="o">*</span><span class="n">_</span> <span class="o">=</span> <span class="n">results</span>
                <span class="k">if</span> <span class="n">VERBOSE</span><span class="p">:</span>
                    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\t</span><span class="si">{</span><span class="n">factor</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2">:</span><span class="se">\t</span><span class="si">{</span><span class="n">new_kcat_fwd</span><span class="si">:</span><span class="s2">5.1f</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">new_kcat_rev</span><span class="si">:</span><span class="s2">5.1f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">new_kcat_fwd</span> <span class="o">&gt;=</span> <span class="mi">0</span> <span class="ow">and</span> <span class="n">new_kcat_rev</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="n">new_objective</span> <span class="o">=</span> <span class="n">objective</span><span class="p">(</span>
                        <span class="n">amino_acid</span><span class="p">,</span>
                        <span class="n">factor</span> <span class="o">*</span> <span class="n">measured_uptake_rates</span><span class="p">[</span><span class="n">amino_acid</span><span class="p">[:</span><span class="o">-</span><span class="mi">3</span><span class="p">]],</span>
                        <span class="n">new_kcat_fwd</span><span class="p">,</span>
                        <span class="n">new_kcat_rev</span><span class="p">,</span>
                    <span class="p">)</span>
                    <span class="k">if</span> <span class="n">best_objective</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">new_objective</span> <span class="o">&lt;</span> <span class="n">best_objective</span><span class="p">:</span>
                        <span class="n">kcat_fwd</span><span class="p">,</span> <span class="n">kcat_rev</span><span class="p">,</span> <span class="n">fwd_rate</span><span class="p">,</span> <span class="n">rev_rate</span><span class="p">,</span> <span class="n">deg_rate</span><span class="p">,</span> <span class="n">uptake</span> <span class="o">=</span> <span class="p">(</span>
                            <span class="n">results</span>
                        <span class="p">)</span>
                        <span class="n">data</span><span class="p">[</span><span class="s2">&quot;kcat&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">kcat_fwd</span> <span class="o">*</span> <span class="n">K_CAT_UNITS</span>
                        <span class="n">best_objective</span> <span class="o">=</span> <span class="n">new_objective</span>

            <span class="c1"># Vary input parameters for kcat calculations for debugging purposes</span>
            <span class="k">if</span> <span class="n">VERBOSE</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;KMs:&quot;</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">factor</span> <span class="ow">in</span> <span class="n">np</span><span class="o">.</span><span class="n">logspace</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">n_factors</span><span class="p">):</span>
                    <span class="n">new_kcat_fwd</span><span class="p">,</span> <span class="n">new_kcat_rev</span><span class="p">,</span> <span class="o">*</span><span class="n">_</span> <span class="o">=</span> <span class="n">calc_kcats</span><span class="p">(</span>
                        <span class="n">aa_conc_basal</span><span class="p">,</span>
                        <span class="n">km_conc_basal</span><span class="p">,</span>
                        <span class="n">aa_conc_with_aa</span><span class="p">,</span>
                        <span class="n">km_conc_with_aa</span><span class="p">,</span>
                        <span class="n">factor</span> <span class="o">*</span> <span class="n">kms</span><span class="p">,</span>
                        <span class="n">km_reverse</span><span class="p">,</span>
                        <span class="n">km_degradation</span><span class="p">,</span>
                        <span class="n">ki</span><span class="p">,</span>
                        <span class="n">measured_uptake_rates</span><span class="p">[</span><span class="n">amino_acid</span><span class="p">[:</span><span class="o">-</span><span class="mi">3</span><span class="p">]],</span>
                    <span class="p">)</span>

                    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\t</span><span class="si">{</span><span class="n">factor</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2">:</span><span class="se">\t</span><span class="si">{</span><span class="n">new_kcat_fwd</span><span class="si">:</span><span class="s2">5.1f</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">new_kcat_rev</span><span class="si">:</span><span class="s2">5.1f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

                <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;km_reverse:&quot;</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">factor</span> <span class="ow">in</span> <span class="n">np</span><span class="o">.</span><span class="n">logspace</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">n_factors</span><span class="p">):</span>
                    <span class="n">new_kcat_fwd</span><span class="p">,</span> <span class="n">new_kcat_rev</span><span class="p">,</span> <span class="o">*</span><span class="n">_</span> <span class="o">=</span> <span class="n">calc_kcats</span><span class="p">(</span>
                        <span class="n">aa_conc_basal</span><span class="p">,</span>
                        <span class="n">km_conc_basal</span><span class="p">,</span>
                        <span class="n">aa_conc_with_aa</span><span class="p">,</span>
                        <span class="n">km_conc_with_aa</span><span class="p">,</span>
                        <span class="n">kms</span><span class="p">,</span>
                        <span class="n">factor</span> <span class="o">*</span> <span class="n">km_reverse</span><span class="p">,</span>
                        <span class="n">km_degradation</span><span class="p">,</span>
                        <span class="n">ki</span><span class="p">,</span>
                        <span class="n">measured_uptake_rates</span><span class="p">[</span><span class="n">amino_acid</span><span class="p">[:</span><span class="o">-</span><span class="mi">3</span><span class="p">]],</span>
                    <span class="p">)</span>

                    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\t</span><span class="si">{</span><span class="n">factor</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2">:</span><span class="se">\t</span><span class="si">{</span><span class="n">new_kcat_fwd</span><span class="si">:</span><span class="s2">5.1f</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">new_kcat_rev</span><span class="si">:</span><span class="s2">5.1f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

                <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;km_degradation:&quot;</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">factor</span> <span class="ow">in</span> <span class="n">np</span><span class="o">.</span><span class="n">logspace</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">n_factors</span><span class="p">):</span>
                    <span class="n">new_kcat_fwd</span><span class="p">,</span> <span class="n">new_kcat_rev</span><span class="p">,</span> <span class="o">*</span><span class="n">_</span> <span class="o">=</span> <span class="n">calc_kcats</span><span class="p">(</span>
                        <span class="n">aa_conc_basal</span><span class="p">,</span>
                        <span class="n">km_conc_basal</span><span class="p">,</span>
                        <span class="n">aa_conc_with_aa</span><span class="p">,</span>
                        <span class="n">km_conc_with_aa</span><span class="p">,</span>
                        <span class="n">kms</span><span class="p">,</span>
                        <span class="n">km_reverse</span><span class="p">,</span>
                        <span class="n">factor</span> <span class="o">*</span> <span class="n">km_degradation</span><span class="p">,</span>
                        <span class="n">ki</span><span class="p">,</span>
                        <span class="n">measured_uptake_rates</span><span class="p">[</span><span class="n">amino_acid</span><span class="p">[:</span><span class="o">-</span><span class="mi">3</span><span class="p">]],</span>
                    <span class="p">)</span>

                    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\t</span><span class="si">{</span><span class="n">factor</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2">:</span><span class="se">\t</span><span class="si">{</span><span class="n">new_kcat_fwd</span><span class="si">:</span><span class="s2">5.1f</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">new_kcat_rev</span><span class="si">:</span><span class="s2">5.1f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

                <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;ki:&quot;</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">factor</span> <span class="ow">in</span> <span class="n">np</span><span class="o">.</span><span class="n">logspace</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">n_factors</span><span class="p">):</span>
                    <span class="n">new_kcat_fwd</span><span class="p">,</span> <span class="n">new_kcat_rev</span><span class="p">,</span> <span class="o">*</span><span class="n">_</span> <span class="o">=</span> <span class="n">calc_kcats</span><span class="p">(</span>
                        <span class="n">aa_conc_basal</span><span class="p">,</span>
                        <span class="n">km_conc_basal</span><span class="p">,</span>
                        <span class="n">aa_conc_with_aa</span><span class="p">,</span>
                        <span class="n">km_conc_with_aa</span><span class="p">,</span>
                        <span class="n">kms</span><span class="p">,</span>
                        <span class="n">km_reverse</span><span class="p">,</span>
                        <span class="n">km_degradation</span><span class="p">,</span>
                        <span class="n">factor</span> <span class="o">*</span> <span class="n">ki</span><span class="p">,</span>
                        <span class="n">measured_uptake_rates</span><span class="p">[</span><span class="n">amino_acid</span><span class="p">[:</span><span class="o">-</span><span class="mi">3</span><span class="p">]],</span>
                    <span class="p">)</span>

                    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\t</span><span class="si">{</span><span class="n">factor</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2">:</span><span class="se">\t</span><span class="si">{</span><span class="n">new_kcat_fwd</span><span class="si">:</span><span class="s2">5.1f</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">new_kcat_rev</span><span class="si">:</span><span class="s2">5.1f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;*** </span><span class="si">{</span><span class="n">amino_acid</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">kcat_fwd</span><span class="si">:</span><span class="s2">5.1f</span><span class="si">}</span><span class="s2"> </span><span class="si">{</span><span class="n">kcat_rev</span><span class="si">:</span><span class="s2">5.1f</span><span class="si">}</span><span class="s2"> ***&quot;</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">kcat_fwd</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                    <span class="s2">&quot;Could not find positive forward and reverse&quot;</span>
                    <span class="sa">f</span><span class="s2">&quot; kcat for </span><span class="si">{</span><span class="n">amino_acid</span><span class="si">}</span><span class="s2">. Run with VERBOSE to check input&quot;</span>
                    <span class="s2">&quot; parameters like KM and KI or check concentrations.&quot;</span>
                <span class="p">)</span>

            <span class="n">aa_enzymes</span> <span class="o">+=</span> <span class="n">fwd_enzymes</span> <span class="o">+</span> <span class="n">rev_enzymes</span>
            <span class="n">enzyme_to_aa_fwd</span> <span class="o">+=</span> <span class="p">[</span><span class="n">amino_acid</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">fwd_enzymes</span><span class="p">)</span> <span class="o">+</span> <span class="p">[</span><span class="kc">None</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span>
                <span class="n">rev_enzymes</span>
            <span class="p">)</span>
            <span class="n">enzyme_to_aa_rev</span> <span class="o">+=</span> <span class="p">[</span><span class="kc">None</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">fwd_enzymes</span><span class="p">)</span> <span class="o">+</span> <span class="p">[</span><span class="n">amino_acid</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span>
                <span class="n">rev_enzymes</span>
            <span class="p">)</span>
            <span class="n">aa_kcats_fwd</span><span class="p">[</span><span class="n">amino_acid</span><span class="p">]</span> <span class="o">=</span> <span class="n">kcat_fwd</span>
            <span class="n">aa_kcats_rev</span><span class="p">[</span><span class="n">amino_acid</span><span class="p">]</span> <span class="o">=</span> <span class="n">kcat_rev</span>
            <span class="n">aa_kis</span><span class="p">[</span><span class="n">amino_acid</span><span class="p">]</span> <span class="o">=</span> <span class="n">ki</span><span class="o">.</span><span class="n">asNumber</span><span class="p">(</span><span class="n">METABOLITE_CONCENTRATION_UNITS</span><span class="p">)</span>
            <span class="n">upstream_aas_for_km</span><span class="p">[</span><span class="n">amino_acid</span><span class="p">]</span> <span class="o">=</span> <span class="n">upstream_aa</span>
            <span class="n">aa_upstream_kms</span><span class="p">[</span><span class="n">amino_acid</span><span class="p">]</span> <span class="o">=</span> <span class="n">kms</span><span class="o">.</span><span class="n">asNumber</span><span class="p">(</span><span class="n">METABOLITE_CONCENTRATION_UNITS</span><span class="p">)</span>
            <span class="n">aa_reverse_kms</span><span class="p">[</span><span class="n">amino_acid</span><span class="p">]</span> <span class="o">=</span> <span class="n">km_reverse</span><span class="o">.</span><span class="n">asNumber</span><span class="p">(</span>
                <span class="n">METABOLITE_CONCENTRATION_UNITS</span>
            <span class="p">)</span>
            <span class="n">aa_degradation_kms</span><span class="p">[</span><span class="n">amino_acid</span><span class="p">]</span> <span class="o">=</span> <span class="n">km_degradation</span><span class="o">.</span><span class="n">asNumber</span><span class="p">(</span>
                <span class="n">METABOLITE_CONCENTRATION_UNITS</span>
            <span class="p">)</span>
            <span class="n">fwd_rates</span><span class="p">[</span><span class="n">amino_acid</span><span class="p">]</span> <span class="o">=</span> <span class="n">fwd_rate</span>
            <span class="n">rev_rates</span><span class="p">[</span><span class="n">amino_acid</span><span class="p">]</span> <span class="o">=</span> <span class="n">rev_rate</span>
            <span class="n">deg_rates</span><span class="p">[</span><span class="n">amino_acid</span><span class="p">]</span> <span class="o">=</span> <span class="n">deg_rate</span>
            <span class="n">calculated_uptake_rates</span><span class="p">[</span><span class="n">amino_acid</span><span class="p">]</span> <span class="o">=</span> <span class="n">uptake</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">aa_enzymes</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">aa_enzymes</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">aa_kcats_fwd</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">aa_kcats_fwd</span><span class="p">[</span><span class="n">aa</span><span class="p">]</span> <span class="k">for</span> <span class="n">aa</span> <span class="ow">in</span> <span class="n">aa_ids</span><span class="p">])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">aa_kcats_rev</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">aa_kcats_rev</span><span class="p">[</span><span class="n">aa</span><span class="p">]</span> <span class="k">for</span> <span class="n">aa</span> <span class="ow">in</span> <span class="n">aa_ids</span><span class="p">])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">aa_kis</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">aa_kis</span><span class="p">[</span><span class="n">aa</span><span class="p">]</span> <span class="k">for</span> <span class="n">aa</span> <span class="ow">in</span> <span class="n">aa_ids</span><span class="p">])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">aa_reverse_kms</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">aa_reverse_kms</span><span class="p">[</span><span class="n">aa</span><span class="p">]</span> <span class="k">for</span> <span class="n">aa</span> <span class="ow">in</span> <span class="n">aa_ids</span><span class="p">])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">aa_degradation_kms</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">aa_degradation_kms</span><span class="p">[</span><span class="n">aa</span><span class="p">]</span> <span class="k">for</span> <span class="n">aa</span> <span class="ow">in</span> <span class="n">aa_ids</span><span class="p">])</span>

        <span class="c1"># Import inhibition of transporters</span>
        <span class="n">rich_conc</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span>
            <span class="p">[</span><span class="n">with_aa_conc</span><span class="p">[</span><span class="n">aa</span><span class="p">]</span><span class="o">.</span><span class="n">asNumber</span><span class="p">(</span><span class="n">METABOLITE_CONCENTRATION_UNITS</span><span class="p">)</span> <span class="k">for</span> <span class="n">aa</span> <span class="ow">in</span> <span class="n">aa_ids</span><span class="p">]</span>
        <span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">aa_import_kis</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">rich_conc</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="p">)</span>  <span class="c1"># Assume this conc is the inhibition constant: TODO: find KIs</span>
        <span class="n">saturation</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">/</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="n">rich_conc</span> <span class="o">/</span> <span class="bp">self</span><span class="o">.</span><span class="n">aa_import_kis</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">specific_import_rates</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span>
            <span class="p">[</span><span class="n">calculated_uptake_rates</span><span class="p">[</span><span class="n">aa</span><span class="p">]</span> <span class="k">for</span> <span class="n">aa</span> <span class="ow">in</span> <span class="n">aa_ids</span><span class="p">]</span>
        <span class="p">)</span> <span class="o">/</span> <span class="n">cell_specs</span><span class="p">[</span><span class="s2">&quot;with_aa&quot;</span><span class="p">][</span><span class="s2">&quot;avgCellDryMassInit&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">asNumber</span><span class="p">(</span><span class="n">DRY_MASS_UNITS</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">max_specific_import_rates</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">specific_import_rates</span> <span class="o">/</span> <span class="n">saturation</span>

        <span class="c1"># KMs for upstream amino acids</span>
        <span class="n">upstream_kms</span> <span class="o">=</span> <span class="p">[</span><span class="n">aa_upstream_kms</span><span class="p">[</span><span class="n">aa</span><span class="p">]</span> <span class="k">for</span> <span class="n">aa</span> <span class="ow">in</span> <span class="n">aa_ids</span><span class="p">]</span>
        <span class="n">upstream_aas</span> <span class="o">=</span> <span class="p">[</span><span class="n">upstream_aas_for_km</span><span class="p">[</span><span class="n">aa</span><span class="p">]</span> <span class="k">for</span> <span class="n">aa</span> <span class="ow">in</span> <span class="n">aa_ids</span><span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">aa_upstream_kms</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">n_aas</span><span class="p">,</span> <span class="n">n_aas</span><span class="p">))</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="p">(</span><span class="n">kms</span><span class="p">,</span> <span class="n">aas</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">upstream_kms</span><span class="p">,</span> <span class="n">upstream_aas</span><span class="p">)):</span>
            <span class="k">for</span> <span class="n">km</span><span class="p">,</span> <span class="n">aa</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">kms</span><span class="p">,</span> <span class="n">aas</span><span class="p">):</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">aa_upstream_kms</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">aa_to_index</span><span class="p">[</span><span class="n">aa</span><span class="p">]]</span> <span class="o">=</span> <span class="n">km</span>

        <span class="c1"># Convert enzyme counts to an amino acid basis via dot product (counts @ self.enzyme_to_amino_acid)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">enzyme_to_amino_acid_fwd</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">aa_enzymes</span><span class="p">),</span> <span class="n">n_aas</span><span class="p">))</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">enzyme_to_amino_acid_rev</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">aa_enzymes</span><span class="p">),</span> <span class="n">n_aas</span><span class="p">))</span>
        <span class="n">enzyme_mapping</span> <span class="o">=</span> <span class="p">{</span><span class="n">e</span><span class="p">:</span> <span class="n">i</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">e</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">aa_enzymes</span><span class="p">)}</span>
        <span class="n">aa_mapping</span> <span class="o">=</span> <span class="p">{</span><span class="n">a</span><span class="p">:</span> <span class="n">i</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">a</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">aa_ids</span><span class="p">)}</span>
        <span class="k">for</span> <span class="n">enzyme</span><span class="p">,</span> <span class="n">fwd</span><span class="p">,</span> <span class="n">rev</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">aa_enzymes</span><span class="p">,</span> <span class="n">enzyme_to_aa_fwd</span><span class="p">,</span> <span class="n">enzyme_to_aa_rev</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">fwd</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">enzyme_to_amino_acid_fwd</span><span class="p">[</span>
                    <span class="n">enzyme_mapping</span><span class="p">[</span><span class="n">enzyme</span><span class="p">],</span> <span class="n">aa_mapping</span><span class="p">[</span><span class="n">fwd</span><span class="p">]</span>
                <span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>
            <span class="k">if</span> <span class="n">rev</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">enzyme_to_amino_acid_rev</span><span class="p">[</span>
                    <span class="n">enzyme_mapping</span><span class="p">[</span><span class="n">enzyme</span><span class="p">],</span> <span class="n">aa_mapping</span><span class="p">[</span><span class="n">rev</span><span class="p">]</span>
                <span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>

        <span class="c1"># Concentrations for reference in analysis plot</span>
        <span class="n">conversion</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">sim_data</span><span class="o">.</span><span class="n">constants</span><span class="o">.</span><span class="n">cell_density</span>
            <span class="o">/</span> <span class="n">sim_data</span><span class="o">.</span><span class="n">constants</span><span class="o">.</span><span class="n">n_avogadro</span>
            <span class="o">*</span> <span class="n">sim_data</span><span class="o">.</span><span class="n">mass</span><span class="o">.</span><span class="n">cell_dry_mass_fraction</span>
        <span class="p">)</span>
        <span class="n">aa_enzymes_basal_idx</span> <span class="o">=</span> <span class="n">bulk_name_to_idx</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">aa_enzymes</span><span class="p">,</span> <span class="n">basal_container</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">])</span>
        <span class="n">basal_counts</span> <span class="o">=</span> <span class="n">counts</span><span class="p">(</span><span class="n">basal_container</span><span class="p">,</span> <span class="n">aa_enzymes_basal_idx</span><span class="p">)</span>
        <span class="n">aa_enzymes_with_aa_idx</span> <span class="o">=</span> <span class="n">bulk_name_to_idx</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">aa_enzymes</span><span class="p">,</span> <span class="n">with_aa_container</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]</span>
        <span class="p">)</span>
        <span class="n">with_aa_counts</span> <span class="o">=</span> <span class="n">counts</span><span class="p">(</span><span class="n">with_aa_container</span><span class="p">,</span> <span class="n">aa_enzymes_with_aa_idx</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">aa_supply_enzyme_conc_with_aa</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">conversion</span> <span class="o">*</span> <span class="n">with_aa_counts</span> <span class="o">/</span> <span class="n">cell_specs</span><span class="p">[</span><span class="s2">&quot;with_aa&quot;</span><span class="p">][</span><span class="s2">&quot;avgCellDryMassInit&quot;</span><span class="p">]</span>
        <span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">aa_supply_enzyme_conc_basal</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">conversion</span> <span class="o">*</span> <span class="n">basal_counts</span> <span class="o">/</span> <span class="n">cell_specs</span><span class="p">[</span><span class="s2">&quot;basal&quot;</span><span class="p">][</span><span class="s2">&quot;avgCellDryMassInit&quot;</span><span class="p">]</span>
        <span class="p">)</span>

        <span class="c1"># Check calculations that could end up negative</span>
        <span class="n">neg_idx</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">max_specific_import_rates</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">neg_idx</span><span class="p">):</span>
            <span class="n">bad_aas</span> <span class="o">=</span> <span class="s2">&quot;, &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">aa_ids</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span> <span class="k">for</span> <span class="n">idx</span> <span class="ow">in</span> <span class="n">neg_idx</span><span class="p">])</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">max_specific_import_rates</span><span class="w"> </span><span class="si">= }</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Import rate was determined to be negative for </span><span class="si">{</span><span class="n">bad_aas</span><span class="si">}</span><span class="s2">.&quot;</span>
                <span class="s2">&quot; Check input parameters like supply and synthesis or enzyme expression.&quot;</span>
            <span class="p">)</span></div>


<div class="viewcode-block" id="Metabolism.get_pathway_enzyme_counts_per_aa">
<a class="viewcode-back" href="../../../../../reference/api/reconstruction/reconstruction.ecoli.dataclasses.process.metabolism.html#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.get_pathway_enzyme_counts_per_aa">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">get_pathway_enzyme_counts_per_aa</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">enzyme_counts</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Get the counts of enzymes for forward and reverse reactions in the</span>
<span class="sd">        amino acid synthesis network based on all of the enzymes used in the</span>
<span class="sd">        network.  Useful to get the counts to pass to amino_acid_synthesis()</span>
<span class="sd">        from counts based on self.aa_enzymes.</span>

<span class="sd">        Args:</span>
<span class="sd">                enzyme_counts: counts of all enzymes in the amino acid network</span>

<span class="sd">        Returns:</span>
<span class="sd">                2-element tuple containing</span>
<span class="sd">                        - counts_per_aa_fwd: counts of enzymes for the forward reaction</span>
<span class="sd">                          for each amino acid</span>
<span class="sd">                        - counts_per_aa_rev: counts of enzymes for the reverse reaction</span>
<span class="sd">                          for each amino acid</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">counts_per_aa_fwd</span> <span class="o">=</span> <span class="n">enzyme_counts</span> <span class="o">@</span> <span class="bp">self</span><span class="o">.</span><span class="n">enzyme_to_amino_acid_fwd</span>
        <span class="n">counts_per_aa_rev</span> <span class="o">=</span> <span class="n">enzyme_counts</span> <span class="o">@</span> <span class="bp">self</span><span class="o">.</span><span class="n">enzyme_to_amino_acid_rev</span>
        <span class="k">return</span> <span class="n">counts_per_aa_fwd</span><span class="p">,</span> <span class="n">counts_per_aa_rev</span></div>


<div class="viewcode-block" id="Metabolism.amino_acid_synthesis">
<a class="viewcode-back" href="../../../../../reference/api/reconstruction/reconstruction.ecoli.dataclasses.process.metabolism.html#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.amino_acid_synthesis">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">amino_acid_synthesis</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">counts_per_aa_fwd</span><span class="p">:</span> <span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">int64</span><span class="p">],</span>
        <span class="n">counts_per_aa_rev</span><span class="p">:</span> <span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">int64</span><span class="p">],</span>
        <span class="n">aa_conc</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">units</span><span class="o">.</span><span class="n">Unum</span><span class="p">,</span> <span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">]],</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">tuple</span><span class="p">[</span>
        <span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">],</span> <span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">],</span> <span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">]</span>
    <span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Calculate the net rate of synthesis for amino acid pathways (can be</span>
<span class="sd">        negative with reverse reactions).</span>

<span class="sd">        Args:</span>
<span class="sd">                counts_per_aa_fwd: counts for enzymes in forward reactions for</span>
<span class="sd">                        each amino acid</span>
<span class="sd">                counts_per_aa_rev: counts for enzymes in loss reactions for each</span>
<span class="sd">                        amino acid</span>
<span class="sd">                aa_conc: concentrations of each amino acid with mol/volume units</span>

<span class="sd">        Returns:</span>
<span class="sd">                3-element tuple containing</span>

<span class="sd">                        - synthesis: net rate of synthesis for each amino acid pathway.</span>
<span class="sd">                          array is unitless but represents counts of amino acid per second</span>
<span class="sd">                        - forward_fraction: saturated fraction for forward reactions</span>
<span class="sd">                        - loss_fraction: saturated fraction for loss reactions</span>

<span class="sd">        .. note::</span>
<span class="sd">                Currently does not match saturation terms used in calc_kcats since</span>
<span class="sd">                it assumes only a reverse or degradation KM exists for simpler calculations</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Convert to appropriate arrays</span>
        <span class="k">if</span> <span class="n">units</span><span class="o">.</span><span class="n">hasUnit</span><span class="p">(</span><span class="n">aa_conc</span><span class="p">):</span>
            <span class="n">aa_conc</span> <span class="o">=</span> <span class="n">aa_conc</span><span class="o">.</span><span class="n">asNumber</span><span class="p">(</span><span class="n">METABOLITE_CONCENTRATION_UNITS</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">amino_acid_synthesis_jit</span><span class="p">(</span>
            <span class="n">counts_per_aa_fwd</span><span class="p">,</span>
            <span class="n">counts_per_aa_rev</span><span class="p">,</span>
            <span class="n">aa_conc</span><span class="p">,</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">aa_upstream_kms</span><span class="p">,</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">aa_kis</span><span class="p">,</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">aa_reverse_kms</span><span class="p">,</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">aa_degradation_kms</span><span class="p">,</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">aa_forward_stoich</span><span class="p">,</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">aa_kcats_fwd</span><span class="p">,</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">aa_reverse_stoich</span><span class="p">,</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">aa_kcats_rev</span><span class="p">,</span>
        <span class="p">)</span></div>


<div class="viewcode-block" id="Metabolism.amino_acid_export">
<a class="viewcode-back" href="../../../../../reference/api/reconstruction/reconstruction.ecoli.dataclasses.process.metabolism.html#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.amino_acid_export">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">amino_acid_export</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">aa_transporters_counts</span><span class="p">:</span> <span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">int64</span><span class="p">],</span>
        <span class="n">aa_conc</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">units</span><span class="o">.</span><span class="n">Unum</span><span class="p">,</span> <span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">]],</span>
        <span class="n">mechanistic_uptake</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Calculate the rate of amino acid export.</span>

<span class="sd">        Args:</span>
<span class="sd">                aa_transporters_counts: counts of each transporter</span>
<span class="sd">                aa_conc: concentrations of each amino acid with mol/volume units</span>
<span class="sd">                mechanistic_uptake: if true, the uptake is calculated based on transporters</span>

<span class="sd">        Returns:</span>
<span class="sd">                Rate of export for each amino acid (unitless but</span>
<span class="sd">                represents counts of amino acid per second)</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">units</span><span class="o">.</span><span class="n">hasUnit</span><span class="p">(</span><span class="n">aa_conc</span><span class="p">):</span>
            <span class="n">aa_conc</span> <span class="o">=</span> <span class="n">aa_conc</span><span class="o">.</span><span class="n">asNumber</span><span class="p">(</span><span class="n">METABOLITE_CONCENTRATION_UNITS</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">amino_acid_export_jit</span><span class="p">(</span>
            <span class="n">aa_transporters_counts</span><span class="p">,</span>
            <span class="n">aa_conc</span><span class="p">,</span>
            <span class="n">mechanistic_uptake</span><span class="p">,</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">aa_to_exporters_matrix</span><span class="p">,</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">export_kcats_per_aa</span><span class="p">,</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">aa_export_kms</span><span class="p">,</span>
        <span class="p">)</span></div>


<div class="viewcode-block" id="Metabolism.amino_acid_import">
<a class="viewcode-back" href="../../../../../reference/api/reconstruction/reconstruction.ecoli.dataclasses.process.metabolism.html#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.amino_acid_import">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">amino_acid_import</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">aa_in_media</span><span class="p">:</span> <span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">bool_</span><span class="p">],</span>
        <span class="n">dry_mass</span><span class="p">:</span> <span class="n">units</span><span class="o">.</span><span class="n">Unum</span><span class="p">,</span>
        <span class="n">internal_aa_conc</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">units</span><span class="o">.</span><span class="n">Unum</span><span class="p">,</span> <span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">]],</span>
        <span class="n">aa_transporters_counts</span><span class="p">:</span> <span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">int64</span><span class="p">],</span>
        <span class="n">mechanistic_uptake</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Calculate the rate of amino acid uptake.</span>

<span class="sd">        Args:</span>
<span class="sd">                aa_in_media: bool for each amino acid being present in current media</span>
<span class="sd">                dry_mass: current dry mass of the cell, with mass units</span>
<span class="sd">                internal_aa_conc: internal concentrations of amino acids</span>
<span class="sd">                aa_transporters_counts: counts of each transporter</span>
<span class="sd">                mechanistic_uptake: if true, the uptake is calculated based on</span>
<span class="sd">                        transporters</span>

<span class="sd">        Returns:</span>
<span class="sd">                Rate of uptake for each amino acid (unitless but</span>
<span class="sd">                represents counts of amino acid per second)</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="n">units</span><span class="o">.</span><span class="n">hasUnit</span><span class="p">(</span><span class="n">internal_aa_conc</span><span class="p">):</span>
            <span class="n">internal_aa_conc</span> <span class="o">=</span> <span class="n">internal_aa_conc</span><span class="o">.</span><span class="n">asNumber</span><span class="p">(</span><span class="n">METABOLITE_CONCENTRATION_UNITS</span><span class="p">)</span>

        <span class="n">dry_mass</span> <span class="o">=</span> <span class="n">dry_mass</span><span class="o">.</span><span class="n">asNumber</span><span class="p">(</span><span class="n">DRY_MASS_UNITS</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">amino_acid_import_jit</span><span class="p">(</span>
            <span class="n">aa_in_media</span><span class="p">,</span>
            <span class="n">dry_mass</span><span class="p">,</span>
            <span class="n">internal_aa_conc</span><span class="p">,</span>
            <span class="n">aa_transporters_counts</span><span class="p">,</span>
            <span class="n">mechanistic_uptake</span><span class="p">,</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">aa_import_kis</span><span class="p">,</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">aa_to_importers_matrix</span><span class="p">,</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">import_kcats_per_aa</span><span class="p">,</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">max_specific_import_rates</span><span class="p">,</span>
        <span class="p">)</span></div>


<div class="viewcode-block" id="Metabolism.get_amino_acid_conc_conversion">
<a class="viewcode-back" href="../../../../../reference/api/reconstruction/reconstruction.ecoli.dataclasses.process.metabolism.html#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.get_amino_acid_conc_conversion">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">get_amino_acid_conc_conversion</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">conc_units</span><span class="p">):</span>
        <span class="k">return</span> <span class="n">units</span><span class="o">.</span><span class="n">strip_empty_units</span><span class="p">(</span><span class="n">conc_units</span> <span class="o">/</span> <span class="n">METABOLITE_CONCENTRATION_UNITS</span><span class="p">)</span></div>


<div class="viewcode-block" id="Metabolism.extract_reactions">
<a class="viewcode-back" href="../../../../../reference/api/reconstruction/reconstruction.ecoli.dataclasses.process.metabolism.html#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.extract_reactions">[docs]</a>
    <span class="nd">@staticmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">extract_reactions</span><span class="p">(</span>
        <span class="n">raw_data</span><span class="p">:</span> <span class="n">KnowledgeBaseEcoli</span><span class="p">,</span> <span class="n">sim_data</span><span class="p">:</span> <span class="s2">&quot;SimulationDataEcoli&quot;</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">tuple</span><span class="p">[</span>
        <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
        <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">]],</span>
        <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
        <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]],</span>
        <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">],</span>
    <span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Extracts reaction data from raw_data to build metabolism reaction</span>
<span class="sd">        network with stoichiometry, reversibility and enzyme catalysts.</span>

<span class="sd">        Args:</span>
<span class="sd">                raw_data: knowledge base data</span>
<span class="sd">                sim_data: simulation data</span>

<span class="sd">        Returns:</span>
<span class="sd">                5-element tuple containing</span>

<span class="sd">                        - base_rxn_ids: list of base reaction IDs from which reaction</span>
<span class="sd">                          IDs were derived from</span>
<span class="sd">                        - reaction_stoich: stoichiometry of metabolites for each reaction::</span>

<span class="sd">                                {reaction ID: {metabolite ID with location tag: stoichiometry}}</span>

<span class="sd">                        - reversible_reactions: reaction IDs for reactions that have a</span>
<span class="sd">                          reverse complement, does not have reverse tag</span>
<span class="sd">                        - reaction_catalysts: enzyme catalysts for each reaction with known</span>
<span class="sd">                          catalysts, likely a subset of reactions in stoich::</span>

<span class="sd">                                {reaction ID: enzyme IDs with location tag}</span>

<span class="sd">                        - rxn_id_to_base_rxn_id: mapping from reaction IDs to the IDs of</span>
<span class="sd">                          the base reactions they were derived from::</span>

<span class="sd">                                {reaction ID: base ID}</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">compartment_ids_to_abbreviations</span> <span class="o">=</span> <span class="p">{</span>
            <span class="n">comp</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]:</span> <span class="n">comp</span><span class="p">[</span><span class="s2">&quot;abbrev&quot;</span><span class="p">]</span> <span class="k">for</span> <span class="n">comp</span> <span class="ow">in</span> <span class="n">raw_data</span><span class="o">.</span><span class="n">compartments</span>
        <span class="p">}</span>
        <span class="n">compartment_ids_to_abbreviations</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">UNDEFINED_COMPARTMENT_IDS_TO_ABBREVS</span><span class="p">)</span>

        <span class="n">valid_directions</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;L2R&quot;</span><span class="p">,</span> <span class="s2">&quot;R2L&quot;</span><span class="p">,</span> <span class="s2">&quot;BOTH&quot;</span><span class="p">}</span>
        <span class="n">forward_directions</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;L2R&quot;</span><span class="p">,</span> <span class="s2">&quot;BOTH&quot;</span><span class="p">}</span>
        <span class="n">reverse_directions</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;R2L&quot;</span><span class="p">,</span> <span class="s2">&quot;BOTH&quot;</span><span class="p">}</span>

        <span class="n">metabolite_ids</span> <span class="o">=</span> <span class="p">{</span><span class="n">met</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]</span> <span class="k">for</span> <span class="n">met</span> <span class="ow">in</span> <span class="n">cast</span><span class="p">(</span><span class="n">Any</span><span class="p">,</span> <span class="n">raw_data</span><span class="p">)</span><span class="o">.</span><span class="n">metabolites</span><span class="p">}</span>

        <span class="c1"># Build mapping from each complexation subunit to all downstream</span>
        <span class="c1"># complexes containing the subunit, including itself</span>
        <span class="c1"># Start by building mappings from subunits to complexes that are</span>
        <span class="c1"># directly formed from the subunit through a single reaction</span>
        <span class="n">subunit_id_to_parent_complexes</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="p">{}</span>

        <span class="k">for</span> <span class="n">comp_reaction</span> <span class="ow">in</span> <span class="n">itertools</span><span class="o">.</span><span class="n">chain</span><span class="p">(</span>
            <span class="n">cast</span><span class="p">(</span><span class="n">Any</span><span class="p">,</span> <span class="n">raw_data</span><span class="p">)</span><span class="o">.</span><span class="n">complexation_reactions</span><span class="p">,</span>
            <span class="n">cast</span><span class="p">(</span><span class="n">Any</span><span class="p">,</span> <span class="n">raw_data</span><span class="p">)</span><span class="o">.</span><span class="n">equilibrium_reactions</span><span class="p">,</span>
        <span class="p">):</span>
            <span class="n">complex_id</span> <span class="o">=</span> <span class="kc">None</span>

            <span class="c1"># Find ID of complex</span>
            <span class="k">for</span> <span class="n">mol_id</span><span class="p">,</span> <span class="n">coeff</span> <span class="ow">in</span> <span class="n">comp_reaction</span><span class="p">[</span><span class="s2">&quot;stoichiometry&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                <span class="k">if</span> <span class="n">coeff</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="n">complex_id</span> <span class="o">=</span> <span class="n">mol_id</span>
                    <span class="k">break</span>

            <span class="k">assert</span> <span class="n">complex_id</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>

            <span class="c1"># Map each subunit to found complex</span>
            <span class="k">for</span> <span class="n">mol_id</span><span class="p">,</span> <span class="n">coeff</span> <span class="ow">in</span> <span class="n">comp_reaction</span><span class="p">[</span><span class="s2">&quot;stoichiometry&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                <span class="k">if</span> <span class="n">mol_id</span> <span class="o">==</span> <span class="n">complex_id</span> <span class="ow">or</span> <span class="n">mol_id</span> <span class="ow">in</span> <span class="n">metabolite_ids</span><span class="p">:</span>
                    <span class="k">continue</span>
                <span class="k">elif</span> <span class="n">mol_id</span> <span class="ow">in</span> <span class="n">subunit_id_to_parent_complexes</span><span class="p">:</span>
                    <span class="n">subunit_id_to_parent_complexes</span><span class="p">[</span><span class="n">mol_id</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">complex_id</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">subunit_id_to_parent_complexes</span><span class="p">[</span><span class="n">mol_id</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">complex_id</span><span class="p">]</span>

        <span class="c1"># Recursive function that returns a list of all downstream complexes</span>
        <span class="c1"># containing the given subunit, including itself</span>
        <span class="k">def</span><span class="w"> </span><span class="nf">get_all_complexes</span><span class="p">(</span><span class="n">subunit_id</span><span class="p">):</span>
            <span class="n">all_downstream_complex_ids</span> <span class="o">=</span> <span class="p">[</span><span class="n">subunit_id</span><span class="p">]</span>

            <span class="k">if</span> <span class="n">subunit_id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">subunit_id_to_parent_complexes</span><span class="p">:</span>
                <span class="k">return</span> <span class="n">all_downstream_complex_ids</span>

            <span class="c1"># Get downstream complexes of all parent complexes</span>
            <span class="k">for</span> <span class="n">parent_complex_id</span> <span class="ow">in</span> <span class="n">subunit_id_to_parent_complexes</span><span class="p">[</span><span class="n">subunit_id</span><span class="p">]:</span>
                <span class="n">all_downstream_complex_ids</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">get_all_complexes</span><span class="p">(</span><span class="n">parent_complex_id</span><span class="p">))</span>

            <span class="c1"># Remove duplicates</span>
            <span class="k">return</span> <span class="nb">sorted</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">all_downstream_complex_ids</span><span class="p">))</span>

        <span class="n">subunit_id_to_all_downstream_complexes</span> <span class="o">=</span> <span class="p">{</span>
            <span class="n">subunit_id</span><span class="p">:</span> <span class="n">get_all_complexes</span><span class="p">(</span><span class="n">subunit_id</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">subunit_id</span> <span class="ow">in</span> <span class="n">subunit_id_to_parent_complexes</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
        <span class="p">}</span>

        <span class="c1"># Initialize variables to store reaction information</span>
        <span class="n">all_base_rxns</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
        <span class="n">reaction_stoich</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">reversible_reactions</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">reaction_catalysts</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">rxn_id_to_base_rxn_id</span> <span class="o">=</span> <span class="p">{}</span>

        <span class="c1"># Load and parse reaction information from raw_data</span>
        <span class="k">for</span> <span class="n">reaction</span> <span class="ow">in</span> <span class="n">cast</span><span class="p">(</span><span class="n">Any</span><span class="p">,</span> <span class="n">raw_data</span><span class="p">)</span><span class="o">.</span><span class="n">metabolic_reactions</span><span class="p">:</span>
            <span class="n">reaction_id</span> <span class="o">=</span> <span class="n">reaction</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]</span>
            <span class="n">stoich</span> <span class="o">=</span> <span class="n">reaction</span><span class="p">[</span><span class="s2">&quot;stoichiometry&quot;</span><span class="p">]</span>
            <span class="n">direction</span> <span class="o">=</span> <span class="n">reaction</span><span class="p">[</span><span class="s2">&quot;direction&quot;</span><span class="p">]</span>

            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">stoich</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="mi">1</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span>
                    <span class="s2">&quot;Invalid biochemical reaction: </span><span class="si">{}</span><span class="s2">, </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">reaction_id</span><span class="p">,</span> <span class="n">stoich</span><span class="p">)</span>
                <span class="p">)</span>

            <span class="k">if</span> <span class="n">direction</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">valid_directions</span><span class="p">:</span>
                <span class="k">raise</span> <span class="n">InvalidReactionDirectionError</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;The direction </span><span class="si">{</span><span class="n">direction</span><span class="si">}</span><span class="s2"> given for reaction </span><span class="si">{</span><span class="n">reaction_id</span><span class="si">}</span><span class="s2"> is invalid.&quot;</span>
                <span class="p">)</span>

            <span class="n">forward</span> <span class="o">=</span> <span class="n">direction</span> <span class="ow">in</span> <span class="n">forward_directions</span>
            <span class="n">reverse</span> <span class="o">=</span> <span class="n">direction</span> <span class="ow">in</span> <span class="n">reverse_directions</span>

            <span class="k">def</span><span class="w"> </span><span class="nf">convert_compartment_tags</span><span class="p">(</span><span class="n">met_id</span><span class="p">):</span>
                <span class="n">new_met_id</span> <span class="o">=</span> <span class="n">met_id</span>

                <span class="k">for</span> <span class="n">comp_id</span><span class="p">,</span> <span class="n">comp_abbrev</span> <span class="ow">in</span> <span class="n">compartment_ids_to_abbreviations</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                    <span class="n">new_met_id</span> <span class="o">=</span> <span class="n">new_met_id</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;[</span><span class="si">{</span><span class="n">comp_id</span><span class="si">}</span><span class="s2">]&quot;</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;[</span><span class="si">{</span><span class="n">comp_abbrev</span><span class="si">}</span><span class="s2">]&quot;</span><span class="p">)</span>

                <span class="k">return</span> <span class="n">new_met_id</span>

            <span class="c1"># All protein complexes that contain an enzyme subunit are assumed</span>
            <span class="c1"># to retain the enzyme&#39;s catalytic activity</span>
            <span class="n">catalysts_for_this_rxn</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="n">all_potential_catalysts</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">catalyst</span> <span class="ow">in</span> <span class="n">reaction</span><span class="p">[</span><span class="s2">&quot;catalyzed_by&quot;</span><span class="p">]:</span>
                <span class="n">all_potential_catalysts</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span>
                    <span class="n">subunit_id_to_all_downstream_complexes</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">catalyst</span><span class="p">,</span> <span class="p">[</span><span class="n">catalyst</span><span class="p">])</span>
                <span class="p">)</span>

            <span class="k">for</span> <span class="n">catalyst</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">all_potential_catalysts</span><span class="p">)):</span>
                <span class="k">if</span> <span class="n">sim_data</span><span class="o">.</span><span class="n">getter</span><span class="o">.</span><span class="n">is_valid_molecule</span><span class="p">(</span><span class="n">catalyst</span><span class="p">):</span>
                    <span class="n">catalysts_with_loc</span> <span class="o">=</span> <span class="n">catalyst</span> <span class="o">+</span> <span class="n">sim_data</span><span class="o">.</span><span class="n">getter</span><span class="o">.</span><span class="n">get_compartment_tag</span><span class="p">(</span>
                        <span class="n">catalyst</span>
                    <span class="p">)</span>
                    <span class="n">catalysts_for_this_rxn</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">catalysts_with_loc</span><span class="p">)</span>
                <span class="c1"># If we don&#39;t have the catalyst in our reconstruction, drop it</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">VERBOSE</span><span class="p">:</span>
                        <span class="nb">print</span><span class="p">(</span>
                            <span class="s2">&quot;Skipping catalyst </span><span class="si">{}</span><span class="s2"> for </span><span class="si">{}</span><span class="s2"> since it is not in the model&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                                <span class="n">catalyst</span><span class="p">,</span> <span class="n">reaction_id</span>
                            <span class="p">)</span>
                        <span class="p">)</span>

            <span class="c1"># Get base reaction ID of this reaction</span>
            <span class="c1"># If reaction ID does not end with a dot, the given reaction ID is</span>
            <span class="c1"># already a base reaction ID</span>
            <span class="k">if</span> <span class="n">reaction_id</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">!=</span> <span class="s2">&quot;.&quot;</span><span class="p">:</span>
                <span class="n">base_reaction_id</span> <span class="o">=</span> <span class="n">reaction_id</span>
            <span class="c1"># If reaction ID ends with a dot, find the base reaction ID based</span>
            <span class="c1"># on the following rules (provided by EcoCyc):</span>
            <span class="c1">#   The parsing instructions for obtaining the base rxn-ID are, effectively:</span>
            <span class="c1">#   1: Check whether the full rxn-ID ends with a dot.</span>
            <span class="c1">#   2: If there is no dot at the end, it is already a base rxn-ID.</span>
            <span class="c1">#       Otherwise, find the position of a second dot, to the left of the dot at the end.</span>
            <span class="c1">#   3: Extract the string between the last dot and the second to last dot.</span>
            <span class="c1">#       If this intervening string consists of only digits, then convert this string to an integer.</span>
            <span class="c1">#       In this case, this rxn-ID stands for a generic rxn, and has extra suffixes</span>
            <span class="c1">#       that need to be trimmed off, to retrieve the base rxn-ID.</span>
            <span class="c1">#       The extracted integer indicates the length of the suffixes to be trimmed off.</span>
            <span class="c1">#   4: To find the end position of the base rxn-ID, subtract the integer (obtained by 3: )</span>
            <span class="c1">#       from the position of the second dot (obtained by 2: ) .</span>
            <span class="c1">#   5: Retrieve the base rxn-ID, which is the substring from the very left (position 0)</span>
            <span class="c1">#       to the end position (obtained by 4: ) .</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">reaction_id_split</span> <span class="o">=</span> <span class="n">reaction_id</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">)</span>
                <span class="n">suffix_length</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">reaction_id_split</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                <span class="n">base_reaction_id</span> <span class="o">=</span> <span class="s2">&quot;.&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">reaction_id_split</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">])[:</span><span class="o">-</span><span class="n">suffix_length</span><span class="p">]</span>

            <span class="k">if</span> <span class="n">forward</span><span class="p">:</span>
                <span class="n">reaction_stoich</span><span class="p">[</span><span class="n">reaction_id</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span>
                    <span class="n">convert_compartment_tags</span><span class="p">(</span><span class="n">moleculeID</span><span class="p">):</span> <span class="n">stoichCoeff</span>
                    <span class="k">for</span> <span class="n">moleculeID</span><span class="p">,</span> <span class="n">stoichCoeff</span> <span class="ow">in</span> <span class="n">stoich</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
                <span class="p">}</span>
                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">catalysts_for_this_rxn</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="n">reaction_catalysts</span><span class="p">[</span><span class="n">reaction_id</span><span class="p">]</span> <span class="o">=</span> <span class="n">catalysts_for_this_rxn</span>
                <span class="n">rxn_id_to_base_rxn_id</span><span class="p">[</span><span class="n">reaction_id</span><span class="p">]</span> <span class="o">=</span> <span class="n">base_reaction_id</span>

            <span class="k">if</span> <span class="n">reverse</span><span class="p">:</span>
                <span class="n">reverse_reaction_id</span> <span class="o">=</span> <span class="n">REVERSE_REACTION_ID</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">reaction_id</span><span class="p">)</span>
                <span class="n">reaction_stoich</span><span class="p">[</span><span class="n">reverse_reaction_id</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span>
                    <span class="n">convert_compartment_tags</span><span class="p">(</span><span class="n">moleculeID</span><span class="p">):</span> <span class="o">-</span><span class="n">stoichCoeff</span>
                    <span class="k">for</span> <span class="n">moleculeID</span><span class="p">,</span> <span class="n">stoichCoeff</span> <span class="ow">in</span> <span class="n">stoich</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
                <span class="p">}</span>
                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">catalysts_for_this_rxn</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="n">reaction_catalysts</span><span class="p">[</span><span class="n">reverse_reaction_id</span><span class="p">]</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span>
                        <span class="n">catalysts_for_this_rxn</span>
                    <span class="p">)</span>
                <span class="n">rxn_id_to_base_rxn_id</span><span class="p">[</span><span class="n">reverse_reaction_id</span><span class="p">]</span> <span class="o">=</span> <span class="n">base_reaction_id</span>

            <span class="k">if</span> <span class="n">forward</span> <span class="ow">and</span> <span class="n">reverse</span><span class="p">:</span>
                <span class="n">reversible_reactions</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">reaction_id</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">base_reaction_id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">all_base_rxns</span><span class="p">:</span>
                <span class="n">all_base_rxns</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">base_reaction_id</span><span class="p">)</span>

        <span class="n">base_rxn_ids</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">all_base_rxns</span><span class="p">))</span>

        <span class="k">return</span> <span class="p">(</span>
            <span class="n">base_rxn_ids</span><span class="p">,</span>
            <span class="n">reaction_stoich</span><span class="p">,</span>
            <span class="n">reversible_reactions</span><span class="p">,</span>
            <span class="n">reaction_catalysts</span><span class="p">,</span>
            <span class="n">rxn_id_to_base_rxn_id</span><span class="p">,</span>
        <span class="p">)</span></div>


<div class="viewcode-block" id="Metabolism.match_reaction">
<a class="viewcode-back" href="../../../../../reference/api/reconstruction/reconstruction.ecoli.dataclasses.process.metabolism.html#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.match_reaction">[docs]</a>
    <span class="nd">@staticmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">match_reaction</span><span class="p">(</span>
        <span class="n">stoich</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">]],</span>
        <span class="n">catalysts</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]],</span>
        <span class="n">rxn_to_match</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
        <span class="n">enz</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
        <span class="n">mets</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
        <span class="n">direction</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Matches a given reaction (rxn_to_match) to reactions that exist in</span>
<span class="sd">        stoich given that enz is known to catalyze the reaction and mets are</span>
<span class="sd">        reactants in the reaction. Can perform a fuzzy reaction match since</span>
<span class="sd">        rxn_to_match just needs to be part of the actual reaction name to match</span>
<span class="sd">        specific instances of a reaction.</span>
<span class="sd">        (eg. rxn_to_match=&quot;ALCOHOL-DEHYDROG-GENERIC-RXN&quot; can match</span>
<span class="sd">        &quot;ALCOHOL-DEHYDROG-GENERIC-RXN-ETOH/NAD//ACETALD/NADH/PROTON.30.&quot;).</span>

<span class="sd">        Args:</span>
<span class="sd">                stoich: stoichiometry of metabolites for each reaction::</span>

<span class="sd">                        {reaction ID: {metabolite ID with location tag: stoichiometry}}</span>

<span class="sd">                catalysts: enzyme catalysts for each reaction with known catalysts,</span>
<span class="sd">                        likely a subset of reactions in stoich::</span>

<span class="sd">                        {reaction ID: enzyme IDs with location tag}</span>

<span class="sd">                rxn_to_match: reaction ID from kinetics to match to existing reactions</span>
<span class="sd">                enz: enzyme ID with location tag</span>
<span class="sd">                mets: metabolite IDs with no location tag from kinetics</span>
<span class="sd">                direction: reaction directionality, ``&#39;forward&#39;`` or ``&#39;reverse&#39;``</span>
<span class="sd">                        or ``None``</span>

<span class="sd">        Returns:</span>
<span class="sd">                Matched reaction IDs in stoich</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Mapping to handle instances of metabolite classes in kinetics</span>
        <span class="c1"># Keys: specific molecules in kinetics file</span>
        <span class="c1"># Values: class of molecules in reactions file that contain the key</span>
        <span class="n">class_mets</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;RED-THIOREDOXIN-MONOMER&quot;</span><span class="p">:</span> <span class="s2">&quot;Red-Thioredoxin&quot;</span><span class="p">,</span>
            <span class="s2">&quot;RED-THIOREDOXIN2-MONOMER&quot;</span><span class="p">:</span> <span class="s2">&quot;Red-Thioredoxin&quot;</span><span class="p">,</span>
            <span class="s2">&quot;RED-GLUTAREDOXIN&quot;</span><span class="p">:</span> <span class="s2">&quot;Red-Glutaredoxins&quot;</span><span class="p">,</span>
            <span class="s2">&quot;GRXB-MONOMER&quot;</span><span class="p">:</span> <span class="s2">&quot;Red-Glutaredoxins&quot;</span><span class="p">,</span>
            <span class="s2">&quot;GRXC-MONOMER&quot;</span><span class="p">:</span> <span class="s2">&quot;Red-Glutaredoxins&quot;</span><span class="p">,</span>
            <span class="s2">&quot;OX-FLAVODOXIN1&quot;</span><span class="p">:</span> <span class="s2">&quot;Oxidized-flavodoxins&quot;</span><span class="p">,</span>
            <span class="s2">&quot;OX-FLAVODOXIN2&quot;</span><span class="p">:</span> <span class="s2">&quot;Oxidized-flavodoxins&quot;</span><span class="p">,</span>
        <span class="p">}</span>

        <span class="c1"># Match full reaction name from partial reaction in kinetics. Must</span>
        <span class="c1"># also match metabolites since there can be multiple reaction instances.</span>
        <span class="n">match</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="n">match_candidates</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">if</span> <span class="n">rxn_to_match</span> <span class="ow">in</span> <span class="n">stoich</span><span class="p">:</span>
            <span class="n">match_candidates</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">rxn_to_match</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">long_rxn</span><span class="p">,</span> <span class="n">long_mets</span> <span class="ow">in</span> <span class="n">stoich</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                <span class="k">if</span> <span class="n">rxn_to_match</span> <span class="ow">in</span> <span class="n">long_rxn</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">long_rxn</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="n">REVERSE_TAG</span><span class="p">):</span>
                    <span class="n">match</span> <span class="o">=</span> <span class="kc">True</span>
                    <span class="n">stripped_enzs</span> <span class="o">=</span> <span class="p">{</span><span class="n">e</span><span class="p">[:</span><span class="o">-</span><span class="mi">3</span><span class="p">]</span> <span class="k">for</span> <span class="n">e</span> <span class="ow">in</span> <span class="n">catalysts</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">long_rxn</span><span class="p">,</span> <span class="p">[])}</span>
                    <span class="n">stripped_mets</span> <span class="o">=</span> <span class="p">{</span><span class="n">m</span><span class="p">[:</span><span class="o">-</span><span class="mi">3</span><span class="p">]</span> <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">long_mets</span><span class="p">}</span>
                    <span class="k">if</span> <span class="p">(</span>
                        <span class="n">np</span><span class="o">.</span><span class="n">all</span><span class="p">([</span><span class="n">class_mets</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">m</span><span class="p">,</span> <span class="n">m</span><span class="p">)</span> <span class="ow">in</span> <span class="n">stripped_mets</span> <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">mets</span><span class="p">])</span>
                        <span class="ow">and</span> <span class="n">enz</span> <span class="ow">in</span> <span class="n">stripped_enzs</span>
                    <span class="p">):</span>
                        <span class="n">match_candidates</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">long_rxn</span><span class="p">)</span>

        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">match_candidates</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">VERBOSE</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">match</span><span class="p">:</span>
                    <span class="nb">print</span><span class="p">(</span>
                        <span class="s2">&quot;Partial reaction match: </span><span class="si">{}</span><span class="s2"> </span><span class="si">{}</span><span class="s2"> </span><span class="si">{}</span><span class="s2"> </span><span class="si">{}</span><span class="s2"> </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                            <span class="n">rxn_to_match</span><span class="p">,</span> <span class="n">enz</span><span class="p">,</span> <span class="n">stripped_enzs</span><span class="p">,</span> <span class="n">mets</span><span class="p">,</span> <span class="n">stripped_mets</span>
                        <span class="p">)</span>
                    <span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;No reaction match: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">rxn_to_match</span><span class="p">))</span>

        <span class="c1"># Determine direction of kinetic reaction from annotation or</span>
        <span class="c1"># metabolite stoichiometry.</span>
        <span class="n">rxn_matches</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">rxn</span> <span class="ow">in</span> <span class="n">match_candidates</span><span class="p">:</span>
            <span class="n">reverse_rxn</span> <span class="o">=</span> <span class="n">REVERSE_REACTION_ID</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">rxn</span><span class="p">)</span>
            <span class="n">reverse_rxn_exists</span> <span class="o">=</span> <span class="n">reverse_rxn</span> <span class="ow">in</span> <span class="n">stoich</span>
            <span class="k">if</span> <span class="n">direction</span><span class="p">:</span>
                <span class="n">reverse</span> <span class="o">=</span> <span class="n">direction</span> <span class="o">==</span> <span class="s2">&quot;reverse&quot;</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">s</span> <span class="o">=</span> <span class="p">{</span><span class="n">k</span><span class="p">[:</span><span class="o">-</span><span class="mi">3</span><span class="p">]:</span> <span class="n">v</span> <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">stoich</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">rxn</span><span class="p">,</span> <span class="p">{})</span><span class="o">.</span><span class="n">items</span><span class="p">()}</span>
                <span class="n">direction_</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span>
                    <span class="n">np</span><span class="o">.</span><span class="n">sign</span><span class="p">([</span><span class="n">s</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">class_mets</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">m</span><span class="p">,</span> <span class="n">m</span><span class="p">),</span> <span class="mi">0</span><span class="p">)</span> <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">mets</span><span class="p">])</span>
                <span class="p">)</span>
                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">direction_</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">reverse_rxn_exists</span><span class="p">:</span>
                    <span class="n">reverse</span> <span class="o">=</span> <span class="kc">False</span>
                <span class="k">elif</span> <span class="nb">len</span><span class="p">(</span><span class="n">direction_</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">1</span> <span class="ow">or</span> <span class="n">direction_</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">VERBOSE</span><span class="p">:</span>
                        <span class="nb">print</span><span class="p">(</span>
                            <span class="s2">&quot;Conflicting directionality: </span><span class="si">{}</span><span class="s2"> </span><span class="si">{}</span><span class="s2"> </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                                <span class="n">rxn</span><span class="p">,</span> <span class="n">mets</span><span class="p">,</span> <span class="n">direction_</span>
                            <span class="p">)</span>
                        <span class="p">)</span>
                    <span class="k">continue</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">reverse</span> <span class="o">=</span> <span class="n">direction_</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span>

            <span class="c1"># Verify a reverse reaction exists in the model</span>
            <span class="k">if</span> <span class="n">reverse</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">reverse_rxn_exists</span><span class="p">:</span>
                    <span class="n">rxn_matches</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">reverse_rxn</span><span class="p">)</span>
                    <span class="k">continue</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">VERBOSE</span><span class="p">:</span>
                        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;No reverse reaction: </span><span class="si">{}</span><span class="s2"> </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">rxn</span><span class="p">,</span> <span class="n">mets</span><span class="p">))</span>
                    <span class="k">continue</span>

            <span class="n">rxn_matches</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">rxn</span><span class="p">)</span>

        <span class="k">return</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">rxn_matches</span><span class="p">)</span></div>


<div class="viewcode-block" id="Metabolism.temperature_adjusted_kcat">
<a class="viewcode-back" href="../../../../../reference/api/reconstruction/reconstruction.ecoli.dataclasses.process.metabolism.html#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.temperature_adjusted_kcat">[docs]</a>
    <span class="nd">@staticmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">temperature_adjusted_kcat</span><span class="p">(</span>
        <span class="n">kcat</span><span class="p">:</span> <span class="n">Unum</span><span class="p">,</span> <span class="n">temp</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">float</span><span class="p">,</span> <span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Args:</span>
<span class="sd">                kcat: enzyme turnover number(s) (1 / time)</span>
<span class="sd">                temp: temperature of measurement, defaults to 25 if &#39;&#39;</span>

<span class="sd">        Returns:</span>
<span class="sd">                Temperature adjusted kcat values, in units of 1/s</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">temp</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="n">temp</span> <span class="o">=</span> <span class="mi">25</span>
        <span class="k">return</span> <span class="mi">2</span> <span class="o">**</span> <span class="p">((</span><span class="mf">37.0</span> <span class="o">-</span> <span class="n">temp</span><span class="p">)</span> <span class="o">/</span> <span class="mf">10.0</span><span class="p">)</span> <span class="o">*</span> <span class="n">kcat</span><span class="o">.</span><span class="n">asNumber</span><span class="p">(</span><span class="n">K_CAT_UNITS</span><span class="p">)</span></div>


<div class="viewcode-block" id="Metabolism._construct_default_saturation_equation">
<a class="viewcode-back" href="../../../../../reference/api/reconstruction/reconstruction.ecoli.dataclasses.process.metabolism.html#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._construct_default_saturation_equation">[docs]</a>
    <span class="nd">@staticmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">_construct_default_saturation_equation</span><span class="p">(</span>
        <span class="n">mets</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span> <span class="n">kms</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">float</span><span class="p">],</span> <span class="n">kis</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">float</span><span class="p">],</span> <span class="n">known_mets</span><span class="p">:</span> <span class="n">Iterable</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Args:</span>
<span class="sd">                mets: metabolite IDs with location tag for KM and KI</span>
<span class="sd">                        parameters ordered to match order of kms then kis</span>
<span class="sd">                kms: KM parameters associated with mets</span>
<span class="sd">                kis: KI parameters associated with mets</span>
<span class="sd">                known_mets: metabolite IDs with location tag with known</span>
<span class="sd">                        concentrations</span>

<span class="sd">        Returns:</span>
<span class="sd">                Saturation equation with metabolites to replace delimited</span>
<span class="sd">                by double quote (e.g. &quot;metabolite&quot;)</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Check input dimensions</span>
        <span class="n">n_params</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">kms</span><span class="p">)</span> <span class="o">+</span> <span class="nb">len</span><span class="p">(</span><span class="n">kis</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">n_params</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">return</span> <span class="s2">&quot;1&quot;</span>
        <span class="k">if</span> <span class="n">n_params</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="n">mets</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">VERBOSE</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Saturation parameter mismatch: </span><span class="si">{}</span><span class="s2"> </span><span class="si">{}</span><span class="s2"> </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">mets</span><span class="p">,</span> <span class="n">kms</span><span class="p">,</span> <span class="n">kis</span><span class="p">))</span>
            <span class="k">return</span> <span class="s2">&quot;1&quot;</span>

        <span class="n">terms</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="c1"># Add KM terms</span>
        <span class="k">for</span> <span class="n">m</span><span class="p">,</span> <span class="n">k</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">mets</span><span class="p">,</span> <span class="n">kms</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">known_mets</span><span class="p">:</span>
                <span class="n">terms</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;1+</span><span class="si">{}</span><span class="s1">/&quot;</span><span class="si">{}</span><span class="s1">&quot;&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">k</span><span class="p">,</span> <span class="n">m</span><span class="p">))</span>
            <span class="k">elif</span> <span class="n">VERBOSE</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Do not have concentration for </span><span class="si">{}</span><span class="s2"> with KM=</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">m</span><span class="p">,</span> <span class="n">k</span><span class="p">))</span>
        <span class="c1"># Add KI terms</span>
        <span class="k">for</span> <span class="n">m</span><span class="p">,</span> <span class="n">k</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">mets</span><span class="p">[</span><span class="nb">len</span><span class="p">(</span><span class="n">kms</span><span class="p">)</span> <span class="p">:],</span> <span class="n">kis</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">known_mets</span><span class="p">:</span>
                <span class="n">terms</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;1+&quot;</span><span class="si">{}</span><span class="s1">&quot;/</span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">m</span><span class="p">,</span> <span class="n">k</span><span class="p">))</span>
            <span class="k">elif</span> <span class="n">VERBOSE</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Do not have concentration for </span><span class="si">{}</span><span class="s2"> with KI=</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">m</span><span class="p">,</span> <span class="n">k</span><span class="p">))</span>

        <span class="c1"># Enclose groupings if being multiplied together</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">terms</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
            <span class="n">terms</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;(&quot;</span> <span class="o">+</span> <span class="n">terms</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">terms</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">+=</span> <span class="s2">&quot;)&quot;</span>
        <span class="k">elif</span> <span class="nb">len</span><span class="p">(</span><span class="n">terms</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">return</span> <span class="s2">&quot;1&quot;</span>

        <span class="k">return</span> <span class="s2">&quot;1/(</span><span class="si">{}</span><span class="s2">)&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s2">&quot;)*(&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">terms</span><span class="p">))</span></div>


<div class="viewcode-block" id="Metabolism._extract_custom_constraint">
<a class="viewcode-back" href="../../../../../reference/api/reconstruction/reconstruction.ecoli.dataclasses.process.metabolism.html#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._extract_custom_constraint">[docs]</a>
    <span class="nd">@staticmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">_extract_custom_constraint</span><span class="p">(</span>
        <span class="n">constraint</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">],</span>
        <span class="n">reactant_tags</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">],</span>
        <span class="n">product_tags</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">],</span>
        <span class="n">known_mets</span><span class="p">:</span> <span class="nb">set</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">tuple</span><span class="p">[</span><span class="n">Optional</span><span class="p">[</span><span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">]],</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Args:</span>
<span class="sd">                constraint: values defining a kinetic constraint::</span>

<span class="sd">                        {&#39;customRateEquation&#39; (str): mathematical representation of</span>
<span class="sd">                                rate (must contain &#39;kcat*E&#39;),</span>
<span class="sd">                        &#39;customParameterVariables&#39; (dict[str, str]): mapping of</span>
<span class="sd">                                variable names in the rate equation to metabolite IDs</span>
<span class="sd">                                without location tags (must contain &#39;E&#39;),</span>
<span class="sd">                        &#39;customParameterConstants&#39; (list[str]): constant strings</span>
<span class="sd">                                in the rate equation that correspond to values (must</span>
<span class="sd">                                contain &#39;kcat&#39;),</span>
<span class="sd">                        &#39;customParameterConstantValues&#39; (list[float]): values for</span>
<span class="sd">                                each of the constant strings,</span>
<span class="sd">                        &#39;Temp&#39; (float or &#39;&#39;): temperature of measurement}</span>

<span class="sd">                reactant_tags: mapping of molecule IDs without a location tag to</span>
<span class="sd">                        molecule IDs with a location tag for all reactants</span>
<span class="sd">                product_tags: mapping of molecule IDs without a location tag to</span>
<span class="sd">                        molecule IDs with a location tag for all products</span>
<span class="sd">                known_mets: molecule IDs with a location tag for molecules with</span>
<span class="sd">                        known concentrations</span>

<span class="sd">        Returns:</span>
<span class="sd">                2-element tuple containing</span>

<span class="sd">                        - kcats: temperature adjusted kcat value, in units of 1/s</span>
<span class="sd">                        - saturation: saturation equation with metabolites to replace</span>
<span class="sd">                          delimited by double quote (eg. &quot;metabolite&quot;)</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">equation</span> <span class="o">=</span> <span class="n">constraint</span><span class="p">[</span><span class="s2">&quot;customRateEquation&quot;</span><span class="p">]</span>
        <span class="n">variables</span> <span class="o">=</span> <span class="n">constraint</span><span class="p">[</span><span class="s2">&quot;customParameterVariables&quot;</span><span class="p">]</span>
        <span class="n">constant_keys</span> <span class="o">=</span> <span class="n">constraint</span><span class="p">[</span><span class="s2">&quot;customParameterConstants&quot;</span><span class="p">]</span>
        <span class="n">constant_values</span> <span class="o">=</span> <span class="n">constraint</span><span class="p">[</span><span class="s2">&quot;customParameterConstantValues&quot;</span><span class="p">]</span>
        <span class="n">temp</span> <span class="o">=</span> <span class="n">constraint</span><span class="p">[</span><span class="s2">&quot;Temp&quot;</span><span class="p">]</span>

        <span class="c1"># Need to have these in the constraint</span>
        <span class="n">kcat_str</span> <span class="o">=</span> <span class="s2">&quot;kcat&quot;</span>
        <span class="n">enzyme_str</span> <span class="o">=</span> <span class="s2">&quot;E&quot;</span>
        <span class="n">capacity_str</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">{}</span><span class="s2">*</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">kcat_str</span><span class="p">,</span> <span class="n">enzyme_str</span><span class="p">)</span>

        <span class="c1"># Need to replace these symbols in equations</span>
        <span class="n">symbol_sub</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;^&quot;</span><span class="p">:</span> <span class="s2">&quot;**&quot;</span><span class="p">,</span>
        <span class="p">}</span>

        <span class="c1"># Make sure kcat exists</span>
        <span class="k">if</span> <span class="n">kcat_str</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">constant_keys</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">VERBOSE</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span>
                    <span class="s2">&quot;Missing </span><span class="si">{}</span><span class="s2"> in custom constants: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">kcat_str</span><span class="p">,</span> <span class="n">constant_keys</span><span class="p">)</span>
                <span class="p">)</span>
            <span class="k">return</span> <span class="kc">None</span><span class="p">,</span> <span class="p">[]</span>

        <span class="n">custom_kcat</span> <span class="o">=</span> <span class="p">(</span>
            <span class="mi">1</span> <span class="o">/</span> <span class="n">units</span><span class="o">.</span><span class="n">s</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">constant_values</span><span class="p">[</span><span class="n">constant_keys</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">kcat_str</span><span class="p">)]])</span>
        <span class="p">)</span>
        <span class="n">kcats</span> <span class="o">=</span> <span class="n">Metabolism</span><span class="o">.</span><span class="n">temperature_adjusted_kcat</span><span class="p">(</span><span class="n">custom_kcat</span><span class="p">,</span> <span class="n">temp</span><span class="p">)</span>

        <span class="c1"># Make sure equation can be parsed, otherwise just return kcat</span>
        <span class="k">if</span> <span class="n">enzyme_str</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">variables</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">VERBOSE</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span>
                    <span class="s2">&quot;Missing enzyme key (</span><span class="si">{}</span><span class="s2">) in custom variables: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                        <span class="n">enzyme_str</span><span class="p">,</span> <span class="n">variables</span>
                    <span class="p">)</span>
                <span class="p">)</span>
            <span class="k">return</span> <span class="n">kcats</span><span class="p">,</span> <span class="p">[]</span>
        <span class="k">if</span> <span class="n">capacity_str</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">equation</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">VERBOSE</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span>
                    <span class="s2">&quot;Expected to find </span><span class="si">{}</span><span class="s2"> in custom equation: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                        <span class="n">capacity_str</span><span class="p">,</span> <span class="n">equation</span>
                    <span class="p">)</span>
                <span class="p">)</span>
            <span class="k">return</span> <span class="n">kcats</span><span class="p">,</span> <span class="p">[]</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">constant_keys</span><span class="p">)</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="n">constant_values</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">VERBOSE</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span>
                    <span class="s2">&quot;Mismatch between constants: </span><span class="si">{}</span><span class="s2"> </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                        <span class="n">constant_keys</span><span class="p">,</span> <span class="n">constant_values</span>
                    <span class="p">)</span>
                <span class="p">)</span>
            <span class="k">return</span> <span class="n">kcats</span><span class="p">,</span> <span class="p">[]</span>

        <span class="n">variables_with_tags</span> <span class="o">=</span> <span class="p">{</span>
            <span class="n">k</span><span class="p">:</span> <span class="n">reactant_tags</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">v</span><span class="p">,</span> <span class="n">product_tags</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">v</span><span class="p">,</span> <span class="kc">None</span><span class="p">))</span>
            <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">variables</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
            <span class="k">if</span> <span class="n">k</span> <span class="o">!=</span> <span class="n">enzyme_str</span> <span class="ow">and</span> <span class="p">(</span><span class="n">v</span> <span class="ow">in</span> <span class="n">reactant_tags</span> <span class="ow">or</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">product_tags</span><span class="p">)</span>
        <span class="p">}</span>

        <span class="c1"># Substitute values into custom equations</span>
        <span class="c1">## Replace terms with known constant values or sim molecule IDs with concentrations</span>
        <span class="n">custom_subs</span> <span class="o">=</span> <span class="p">{</span><span class="n">k</span><span class="p">:</span> <span class="nb">str</span><span class="p">(</span><span class="n">v</span><span class="p">)</span> <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">constant_keys</span><span class="p">,</span> <span class="n">constant_values</span><span class="p">)}</span>
        <span class="n">custom_subs</span><span class="o">.</span><span class="n">update</span><span class="p">(</span>
            <span class="p">{</span>
                <span class="n">k</span><span class="p">:</span> <span class="s1">&#39;&quot;</span><span class="si">{}</span><span class="s1">&quot;&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">v</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">variables_with_tags</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
                <span class="k">if</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">known_mets</span>
            <span class="p">}</span>
        <span class="p">)</span>

        <span class="c1">## Remove capacity to get only saturation term</span>
        <span class="n">new_equation</span> <span class="o">=</span> <span class="n">equation</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">capacity_str</span><span class="p">,</span> <span class="s2">&quot;1&quot;</span><span class="p">)</span>

        <span class="c1">## Tokenize equation to terms and symbols</span>
        <span class="n">parsed_variables</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;\w*&quot;</span><span class="p">,</span> <span class="n">new_equation</span><span class="p">)[</span>
            <span class="p">:</span><span class="o">-</span><span class="mi">1</span>
        <span class="p">]</span>  <span class="c1"># Remove trailing empty match</span>
        <span class="c1">## Ensure valid input of known variables or a float term</span>
        <span class="k">for</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">parsed_variables</span><span class="p">:</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="p">(</span><span class="n">v</span> <span class="o">==</span> <span class="s2">&quot;&quot;</span> <span class="ow">or</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">custom_subs</span><span class="p">):</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="nb">float</span><span class="p">(</span><span class="n">v</span><span class="p">)</span>
                <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">VERBOSE</span><span class="p">:</span>
                        <span class="nb">print</span><span class="p">(</span>
                            <span class="s2">&quot;Unknown value encountered in custom equation </span><span class="si">{}</span><span class="s2">: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                                <span class="n">equation</span><span class="p">,</span> <span class="n">v</span>
                            <span class="p">)</span>
                        <span class="p">)</span>
                    <span class="k">return</span> <span class="n">kcats</span><span class="p">,</span> <span class="p">[]</span>
        <span class="n">parsed_symbols</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;\W&quot;</span><span class="p">,</span> <span class="n">new_equation</span><span class="p">)</span>
        <span class="n">tokenized_equation</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">parsed_variables</span><span class="p">)</span>
        <span class="n">symbol_idx_mask</span> <span class="o">=</span> <span class="n">tokenized_equation</span> <span class="o">==</span> <span class="s2">&quot;&quot;</span>

        <span class="c1">## Verify tokenized equation matches original before replacements</span>
        <span class="n">tokenized_equation</span><span class="p">[</span><span class="n">symbol_idx_mask</span><span class="p">]</span> <span class="o">=</span> <span class="n">parsed_symbols</span>
        <span class="k">if</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">tokenized_equation</span><span class="p">)</span> <span class="o">!=</span> <span class="n">new_equation</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">VERBOSE</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Error parsing custom equation: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">equation</span><span class="p">))</span>
            <span class="k">return</span> <span class="n">kcats</span><span class="p">,</span> <span class="p">[]</span>

        <span class="c1">## Perform replacement of symbols</span>
        <span class="n">tokenized_equation</span><span class="p">[</span><span class="n">symbol_idx_mask</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span>
            <span class="n">symbol_sub</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">s</span><span class="p">,</span> <span class="n">s</span><span class="p">)</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">parsed_symbols</span>
        <span class="p">]</span>

        <span class="c1"># Reconstruct saturation equation with replacements</span>
        <span class="n">saturation</span> <span class="o">=</span> <span class="p">[</span>
            <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">custom_subs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">token</span><span class="p">,</span> <span class="n">token</span><span class="p">)</span> <span class="k">for</span> <span class="n">token</span> <span class="ow">in</span> <span class="n">tokenized_equation</span><span class="p">])</span>
        <span class="p">]</span>

        <span class="k">return</span> <span class="n">kcats</span><span class="p">,</span> <span class="n">saturation</span></div>


<div class="viewcode-block" id="Metabolism.extract_kinetic_constraints">
<a class="viewcode-back" href="../../../../../reference/api/reconstruction/reconstruction.ecoli.dataclasses.process.metabolism.html#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism.extract_kinetic_constraints">[docs]</a>
    <span class="nd">@staticmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">extract_kinetic_constraints</span><span class="p">(</span>
        <span class="n">raw_data</span><span class="p">:</span> <span class="n">KnowledgeBaseEcoli</span><span class="p">,</span>
        <span class="n">sim_data</span><span class="p">:</span> <span class="s2">&quot;SimulationDataEcoli&quot;</span><span class="p">,</span>
        <span class="n">stoich</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">]]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">catalysts</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">known_metabolites</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Set</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">],</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">list</span><span class="p">[</span><span class="n">Any</span><span class="p">]]]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Load and parse kinetic constraint information from raw_data</span>

<span class="sd">        Args:</span>
<span class="sd">                raw_data: knowledge base data</span>
<span class="sd">                sim_data: simulation data</span>
<span class="sd">                stoich: stoichiometry of metabolites for each reaction (if ``None``,</span>
<span class="sd">                        data is loaded from ``raw_data`` and ``sim_data``)::</span>

<span class="sd">                        {reaction ID: {metabolite ID with location tag: stoichiometry}}</span>

<span class="sd">                catalysts: enzyme catalysts for each reaction with known catalysts,</span>
<span class="sd">                        likely a subset of reactions in ``stoich`` (if ``None``, data</span>
<span class="sd">                        is loaded from ``raw_data`` and ``sim_data``::</span>

<span class="sd">                        {reaction ID: enzyme IDs with location tag}</span>

<span class="sd">                known_metabolites: metabolites with known concentrations</span>

<span class="sd">        Returns:</span>
<span class="sd">                Valid kinetic constraints for each reaction/enzyme pair::</span>

<span class="sd">                        {(reaction ID, enzyme with location tag): {</span>
<span class="sd">                                &#39;kcat&#39;: kcat values (list[float]),</span>
<span class="sd">                                &#39;saturation&#39;: saturation equations (list[str])</span>
<span class="sd">                        }}</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Load data for optional args if needed</span>
        <span class="k">if</span> <span class="n">stoich</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">catalysts</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">_</span><span class="p">,</span> <span class="n">loaded_stoich</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">loaded_catalysts</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">Metabolism</span><span class="o">.</span><span class="n">extract_reactions</span><span class="p">(</span>
                <span class="n">raw_data</span><span class="p">,</span> <span class="n">sim_data</span>
            <span class="p">)</span>

            <span class="k">if</span> <span class="n">stoich</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">stoich</span> <span class="o">=</span> <span class="n">loaded_stoich</span>
            <span class="k">if</span> <span class="n">catalysts</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">catalysts</span> <span class="o">=</span> <span class="n">loaded_catalysts</span>

        <span class="n">known_metabolites_</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span> <span class="k">if</span> <span class="n">known_metabolites</span> <span class="ow">is</span> <span class="kc">None</span> <span class="k">else</span> <span class="n">known_metabolites</span>

        <span class="n">constraints</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">],</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">list</span><span class="p">]]</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">constraint</span> <span class="ow">in</span> <span class="n">cast</span><span class="p">(</span><span class="n">Any</span><span class="p">,</span> <span class="n">raw_data</span><span class="p">)</span><span class="o">.</span><span class="n">metabolism_kinetics</span><span class="p">:</span>
            <span class="n">rxn</span> <span class="o">=</span> <span class="n">constraint</span><span class="p">[</span><span class="s2">&quot;reactionID&quot;</span><span class="p">]</span>
            <span class="n">enzyme</span> <span class="o">=</span> <span class="n">constraint</span><span class="p">[</span><span class="s2">&quot;enzymeID&quot;</span><span class="p">]</span>
            <span class="n">metabolites</span> <span class="o">=</span> <span class="n">constraint</span><span class="p">[</span><span class="s2">&quot;substrateIDs&quot;</span><span class="p">]</span>
            <span class="n">direction</span> <span class="o">=</span> <span class="n">constraint</span><span class="p">[</span><span class="s2">&quot;direction&quot;</span><span class="p">]</span>
            <span class="n">kms</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">constraint</span><span class="p">[</span><span class="s2">&quot;kM&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">asNumber</span><span class="p">(</span><span class="n">KINETIC_CONSTRAINT_CONC_UNITS</span><span class="p">))</span>
            <span class="n">kis</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">constraint</span><span class="p">[</span><span class="s2">&quot;kI&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">asNumber</span><span class="p">(</span><span class="n">KINETIC_CONSTRAINT_CONC_UNITS</span><span class="p">))</span>
            <span class="n">n_reactants</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">metabolites</span><span class="p">)</span> <span class="o">-</span> <span class="nb">len</span><span class="p">(</span><span class="n">kis</span><span class="p">)</span>
            <span class="n">matched_rxns</span> <span class="o">=</span> <span class="n">Metabolism</span><span class="o">.</span><span class="n">match_reaction</span><span class="p">(</span>
                <span class="n">stoich</span><span class="p">,</span> <span class="n">catalysts</span><span class="p">,</span> <span class="n">rxn</span><span class="p">,</span> <span class="n">enzyme</span><span class="p">,</span> <span class="n">metabolites</span><span class="p">[:</span><span class="n">n_reactants</span><span class="p">],</span> <span class="n">direction</span>
            <span class="p">)</span>

            <span class="k">for</span> <span class="n">matched_rxn</span> <span class="ow">in</span> <span class="n">matched_rxns</span><span class="p">:</span>
                <span class="c1"># Ensure enzyme catalyzes reaction in model</span>
                <span class="n">enzymes_tag_conversion</span> <span class="o">=</span> <span class="p">{</span>
                    <span class="n">e</span><span class="p">[:</span><span class="o">-</span><span class="mi">3</span><span class="p">]:</span> <span class="n">e</span> <span class="k">for</span> <span class="n">e</span> <span class="ow">in</span> <span class="n">catalysts</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">matched_rxn</span><span class="p">,</span> <span class="p">[])</span>
                <span class="p">}</span>
                <span class="k">if</span> <span class="n">enzyme</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">enzymes_tag_conversion</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">VERBOSE</span><span class="p">:</span>
                        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2"> does not catalyze </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">enzyme</span><span class="p">,</span> <span class="n">matched_rxn</span><span class="p">))</span>
                    <span class="k">continue</span>

                <span class="c1"># Update metabolites with a location tag from the reaction</span>
                <span class="c1"># First look in reactants but some products can inhibit</span>
                <span class="n">reactant_tags</span> <span class="o">=</span> <span class="p">{</span>
                    <span class="n">k</span><span class="p">[:</span><span class="o">-</span><span class="mi">3</span><span class="p">]:</span> <span class="n">k</span> <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">stoich</span><span class="p">[</span><span class="n">matched_rxn</span><span class="p">]</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> <span class="k">if</span> <span class="n">v</span> <span class="o">&lt;</span> <span class="mi">0</span>
                <span class="p">}</span>
                <span class="n">product_tags</span> <span class="o">=</span> <span class="p">{</span>
                    <span class="n">k</span><span class="p">[:</span><span class="o">-</span><span class="mi">3</span><span class="p">]:</span> <span class="n">k</span> <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">stoich</span><span class="p">[</span><span class="n">matched_rxn</span><span class="p">]</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> <span class="k">if</span> <span class="n">v</span> <span class="o">&gt;</span> <span class="mi">0</span>
                <span class="p">}</span>
                <span class="n">mets_with_tag</span> <span class="o">=</span> <span class="p">[</span>
                    <span class="n">reactant_tags</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">met</span><span class="p">,</span> <span class="n">product_tags</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">met</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">))</span>
                    <span class="k">for</span> <span class="n">met</span> <span class="ow">in</span> <span class="n">metabolites</span>
                    <span class="k">if</span> <span class="n">met</span> <span class="ow">in</span> <span class="n">reactant_tags</span> <span class="ow">or</span> <span class="n">met</span> <span class="ow">in</span> <span class="n">product_tags</span>
                <span class="p">]</span>
                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">mets_with_tag</span><span class="p">)</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="n">metabolites</span><span class="p">):</span>
                    <span class="c1"># Warn if verbose but no continue since we can still use kcat</span>
                    <span class="k">if</span> <span class="n">VERBOSE</span><span class="p">:</span>
                        <span class="nb">print</span><span class="p">(</span>
                            <span class="s2">&quot;Could not match all metabolites: </span><span class="si">{}</span><span class="s2"> </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                                <span class="n">metabolites</span><span class="p">,</span> <span class="n">mets_with_tag</span>
                            <span class="p">)</span>
                        <span class="p">)</span>

                <span class="c1"># Extract kcat and saturation parameters</span>
                <span class="k">if</span> <span class="n">constraint</span><span class="p">[</span><span class="s2">&quot;rateEquationType&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;custom&quot;</span><span class="p">:</span>
                    <span class="n">kcats</span><span class="p">,</span> <span class="n">saturation</span> <span class="o">=</span> <span class="n">Metabolism</span><span class="o">.</span><span class="n">_extract_custom_constraint</span><span class="p">(</span>
                        <span class="n">constraint</span><span class="p">,</span> <span class="n">reactant_tags</span><span class="p">,</span> <span class="n">product_tags</span><span class="p">,</span> <span class="n">known_metabolites_</span>
                    <span class="p">)</span>
                    <span class="k">if</span> <span class="n">kcats</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                        <span class="k">continue</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">kcats</span> <span class="o">=</span> <span class="n">Metabolism</span><span class="o">.</span><span class="n">temperature_adjusted_kcat</span><span class="p">(</span>
                        <span class="n">constraint</span><span class="p">[</span><span class="s2">&quot;kcat&quot;</span><span class="p">],</span> <span class="n">constraint</span><span class="p">[</span><span class="s2">&quot;Temp&quot;</span><span class="p">]</span>
                    <span class="p">)</span>
                    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">kcats</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
                        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">kcats</span><span class="p">)</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="n">kms</span><span class="p">)</span> <span class="ow">or</span> <span class="nb">len</span><span class="p">(</span><span class="n">kms</span><span class="p">)</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="n">mets_with_tag</span><span class="p">):</span>
                            <span class="k">if</span> <span class="n">VERBOSE</span><span class="p">:</span>
                                <span class="nb">print</span><span class="p">(</span>
                                    <span class="s2">&quot;Could not align kcats and kms: </span><span class="si">{}</span><span class="s2"> </span><span class="si">{}</span><span class="s2"> </span><span class="si">{}</span><span class="s2"> </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                                        <span class="n">rxn</span><span class="p">,</span> <span class="n">kcats</span><span class="p">,</span> <span class="n">kms</span><span class="p">,</span> <span class="n">mets_with_tag</span>
                                    <span class="p">)</span>
                                <span class="p">)</span>
                            <span class="k">continue</span>

                        <span class="n">saturation</span> <span class="o">=</span> <span class="p">[</span>
                            <span class="n">Metabolism</span><span class="o">.</span><span class="n">_construct_default_saturation_equation</span><span class="p">(</span>
                                <span class="p">[</span><span class="n">m</span><span class="p">],</span> <span class="p">[</span><span class="n">km</span><span class="p">],</span> <span class="p">[],</span> <span class="n">known_metabolites_</span>
                            <span class="p">)</span>
                            <span class="k">for</span> <span class="n">m</span><span class="p">,</span> <span class="n">km</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">mets_with_tag</span><span class="p">,</span> <span class="n">kms</span><span class="p">)</span>
                        <span class="p">]</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">saturation</span> <span class="o">=</span> <span class="p">[</span>
                            <span class="n">Metabolism</span><span class="o">.</span><span class="n">_construct_default_saturation_equation</span><span class="p">(</span>
                                <span class="n">mets_with_tag</span><span class="p">,</span> <span class="n">kms</span><span class="p">,</span> <span class="n">kis</span><span class="p">,</span> <span class="n">known_metabolites_</span>
                            <span class="p">)</span>
                        <span class="p">]</span>

                    <span class="n">saturation</span> <span class="o">=</span> <span class="p">[</span><span class="n">s</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">saturation</span> <span class="k">if</span> <span class="n">s</span> <span class="o">!=</span> <span class="s2">&quot;1&quot;</span><span class="p">]</span>

                <span class="c1"># Add new kcats and saturation terms for the enzymatic reaction</span>
                <span class="n">key</span> <span class="o">=</span> <span class="p">(</span><span class="n">matched_rxn</span><span class="p">,</span> <span class="n">enzymes_tag_conversion</span><span class="p">[</span><span class="n">enzyme</span><span class="p">])</span>
                <span class="n">entries</span> <span class="o">=</span> <span class="n">constraints</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">key</span><span class="p">,</span> <span class="p">{})</span>
                <span class="n">entries</span><span class="p">[</span><span class="s2">&quot;kcat&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">entries</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;kcat&quot;</span><span class="p">,</span> <span class="p">[])</span> <span class="o">+</span> <span class="nb">list</span><span class="p">(</span><span class="n">kcats</span><span class="p">)</span>
                <span class="n">entries</span><span class="p">[</span><span class="s2">&quot;saturation&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">entries</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;saturation&quot;</span><span class="p">,</span> <span class="p">[])</span> <span class="o">+</span> <span class="n">saturation</span>
                <span class="n">constraints</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="n">entries</span>

        <span class="k">return</span> <span class="n">constraints</span></div>


<div class="viewcode-block" id="Metabolism._replace_enzyme_reactions">
<a class="viewcode-back" href="../../../../../reference/api/reconstruction/reconstruction.ecoli.dataclasses.process.metabolism.html#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._replace_enzyme_reactions">[docs]</a>
    <span class="nd">@staticmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">_replace_enzyme_reactions</span><span class="p">(</span>
        <span class="n">constraints</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">],</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">list</span><span class="p">[</span><span class="n">Any</span><span class="p">]]],</span>
        <span class="n">stoich</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">]],</span>
        <span class="n">rxn_catalysts</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]],</span>
        <span class="n">reversible_rxns</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
        <span class="n">rxn_id_to_compiled_id</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">],</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">tuple</span><span class="p">[</span>
        <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">],</span>
        <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">]],</span>
        <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]],</span>
        <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
        <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">],</span>
    <span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Modifies reaction IDs in data structures to duplicate reactions with</span>
<span class="sd">        kinetic constraints and multiple enzymes.</span>

<span class="sd">        Args:</span>
<span class="sd">                constraints: valid kinetic constraints for each reaction/enzyme pair::</span>

<span class="sd">                        {(reaction ID, enzyme with location tag): {</span>
<span class="sd">                                &#39;kcat&#39;: kcat values (list[float]),</span>
<span class="sd">                                &#39;saturation&#39;: saturation equations (list[str])</span>
<span class="sd">                        }}</span>

<span class="sd">                stoich: stoichiometry of metabolites for each reaction (if None, data</span>
<span class="sd">                        is loaded from raw_data and sim_data)::</span>

<span class="sd">                        {reaction ID: {metabolite ID with location tag: stoichiometry}}</span>

<span class="sd">                rxn_catalysts: enzyme catalysts for each reaction with known catalysts,</span>
<span class="sd">                        likely a subset of reactions in stoich (if None, data is loaded</span>
<span class="sd">                        from raw_data and sim_data)::</span>

<span class="sd">                        {reaction ID: enzyme IDs with location tag}</span>

<span class="sd">                reversible_rxns: reaction IDs for reactions that have a reverse</span>
<span class="sd">                        complement, does not have reverse tag</span>
<span class="sd">                rxn_id_to_compiled_id: mapping from reaction IDs to the IDs of the</span>
<span class="sd">                        original reactions they were derived from</span>

<span class="sd">        Returns:</span>
<span class="sd">                5-element tuple containing</span>

<span class="sd">                        - new_constraints: valid kinetic constraints for each reaction::</span>

<span class="sd">                                {reaction ID: {</span>
<span class="sd">                                        &#39;enzyme&#39;: enzyme catalyst (str),</span>
<span class="sd">                                        &#39;kcat&#39;: kcat values (list[float]),</span>
<span class="sd">                                        &#39;saturation&#39;: saturation equations (list[str])</span>
<span class="sd">                                }}</span>

<span class="sd">                        - stoich: stoichiometry of metabolites for each reaction with</span>
<span class="sd">                          updated reactions for enzyme catalyzed kinetic reactions::</span>

<span class="sd">                                {reaction ID: {metabolite ID with location tag: stoichiometry}}</span>

<span class="sd">                        - rxn_catalysts: enzyme catalysts for each reaction with known</span>
<span class="sd">                          catalysts, likely a subset of reactions in stoich with</span>
<span class="sd">                          updated reactions for enzyme catalyzed kinetic reactions::</span>

<span class="sd">                                {reaction ID: enzyme IDs with location tag}</span>

<span class="sd">                        - reversible_rxns: reaction IDs for reactions that have a reverse</span>
<span class="sd">                          complement with updated reactions for enzyme catalyzed kinetic</span>
<span class="sd">                          reactions, does not have reverse tag</span>
<span class="sd">                        - rxn_id_to_compiled_id: mapping from reaction IDs to the IDs</span>
<span class="sd">                          of the original reactions they were derived from, with updated</span>
<span class="sd">                          reactions for enzyme catalyzed kinetic reactions</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">new_constraints</span> <span class="o">=</span> <span class="p">{}</span>

        <span class="n">n_catalysts</span> <span class="o">=</span> <span class="p">{</span><span class="n">rxn</span><span class="p">:</span> <span class="nb">len</span><span class="p">(</span><span class="n">catalysts</span><span class="p">)</span> <span class="k">for</span> <span class="n">rxn</span><span class="p">,</span> <span class="n">catalysts</span> <span class="ow">in</span> <span class="n">rxn_catalysts</span><span class="o">.</span><span class="n">items</span><span class="p">()}</span>

        <span class="c1"># Split out reactions that are kinetically constrained and that have</span>
        <span class="c1"># more than one enzyme that catalyzes the reaction</span>
        <span class="k">for</span> <span class="p">(</span><span class="n">rxn</span><span class="p">,</span> <span class="n">enzyme</span><span class="p">),</span> <span class="n">constraint</span> <span class="ow">in</span> <span class="n">constraints</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="k">if</span> <span class="n">n_catalysts</span><span class="p">[</span><span class="n">rxn</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
                <span class="c1"># Create new reaction name with enzyme appended to the end</span>
                <span class="k">if</span> <span class="n">rxn</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="n">REVERSE_TAG</span><span class="p">):</span>
                    <span class="n">new_rxn</span> <span class="o">=</span> <span class="n">REVERSE_REACTION_ID</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                        <span class="n">ENZYME_REACTION_ID</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">rxn</span><span class="p">[:</span> <span class="o">-</span><span class="nb">len</span><span class="p">(</span><span class="n">REVERSE_TAG</span><span class="p">)],</span> <span class="n">enzyme</span><span class="p">[:</span><span class="o">-</span><span class="mi">3</span><span class="p">])</span>
                    <span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">new_rxn</span> <span class="o">=</span> <span class="n">ENZYME_REACTION_ID</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">rxn</span><span class="p">,</span> <span class="n">enzyme</span><span class="p">[:</span><span class="o">-</span><span class="mi">3</span><span class="p">])</span>

                <span class="c1"># Add the new reaction to appropriate lists and dicts</span>
                <span class="n">stoich</span><span class="p">[</span><span class="n">new_rxn</span><span class="p">]</span> <span class="o">=</span> <span class="n">copy</span><span class="p">(</span><span class="n">stoich</span><span class="p">[</span><span class="n">rxn</span><span class="p">])</span>
                <span class="n">rxn_catalysts</span><span class="p">[</span><span class="n">new_rxn</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">enzyme</span><span class="p">]</span>
                <span class="k">if</span> <span class="n">rxn</span> <span class="ow">in</span> <span class="n">reversible_rxns</span><span class="p">:</span>
                    <span class="n">reversible_rxns</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">new_rxn</span><span class="p">)</span>

                <span class="c1"># Remove enzyme from old reaction and remove old reaction if no</span>
                <span class="c1"># more enzyme catalysts</span>
                <span class="n">rxn_catalysts</span><span class="p">[</span><span class="n">rxn</span><span class="p">]</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="n">rxn_catalysts</span><span class="p">[</span><span class="n">rxn</span><span class="p">]</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">enzyme</span><span class="p">))</span>

                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">rxn_catalysts</span><span class="p">[</span><span class="n">rxn</span><span class="p">])</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="n">stoich</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="n">rxn</span><span class="p">)</span>
                    <span class="n">rxn_catalysts</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="n">rxn</span><span class="p">)</span>
                    <span class="k">if</span> <span class="n">rxn</span> <span class="ow">in</span> <span class="n">reversible_rxns</span><span class="p">:</span>
                        <span class="n">reversible_rxns</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="n">reversible_rxns</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">rxn</span><span class="p">))</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">new_rxn</span> <span class="o">=</span> <span class="n">rxn</span>

            <span class="n">rxn_id_to_compiled_id</span><span class="p">[</span><span class="n">new_rxn</span><span class="p">]</span> <span class="o">=</span> <span class="n">rxn_id_to_compiled_id</span><span class="p">[</span><span class="n">rxn</span><span class="p">]</span>

            <span class="c1"># noinspection PyTypeChecker</span>
            <span class="n">new_constraints</span><span class="p">[</span><span class="n">new_rxn</span><span class="p">]</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="n">constraints</span><span class="p">[(</span><span class="n">rxn</span><span class="p">,</span> <span class="n">enzyme</span><span class="p">)],</span> <span class="n">enzyme</span><span class="o">=</span><span class="n">enzyme</span><span class="p">)</span>

        <span class="k">return</span> <span class="p">(</span>
            <span class="n">new_constraints</span><span class="p">,</span>
            <span class="n">stoich</span><span class="p">,</span>
            <span class="n">rxn_catalysts</span><span class="p">,</span>
            <span class="n">reversible_rxns</span><span class="p">,</span>
            <span class="n">rxn_id_to_compiled_id</span><span class="p">,</span>
        <span class="p">)</span></div>


<div class="viewcode-block" id="Metabolism._lambdify_constraints">
<a class="viewcode-back" href="../../../../../reference/api/reconstruction/reconstruction.ecoli.dataclasses.process.metabolism.html#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._lambdify_constraints">[docs]</a>
    <span class="nd">@staticmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">_lambdify_constraints</span><span class="p">(</span>
        <span class="n">constraints</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">],</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">tuple</span><span class="p">[</span>
        <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
        <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
        <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
        <span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">],</span>
        <span class="nb">str</span><span class="p">,</span>
        <span class="nb">str</span><span class="p">,</span>
        <span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">bool_</span><span class="p">],</span>
    <span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Creates str representations of kinetic terms to be used to create</span>
<span class="sd">        kinetic constraints that are returned with getKineticConstraints().</span>

<span class="sd">        Args:</span>
<span class="sd">                constraints: valid kinetic constraints for each reaction::</span>

<span class="sd">                        {reaction ID: {</span>
<span class="sd">                                &#39;enzyme&#39;: enzyme catalyst (str),</span>
<span class="sd">                                &#39;kcat&#39;: kcat values (list[float]),</span>
<span class="sd">                                &#39;saturation&#39;: saturation equations (list[str])</span>
<span class="sd">                        }}</span>

<span class="sd">        Returns:</span>
<span class="sd">                7-element tuple containing</span>

<span class="sd">                        - rxns: sorted reaction IDs for reactions with a kinetic</span>
<span class="sd">                          constraint</span>
<span class="sd">                        - enzymes: sorted enzyme IDs for enzymes that catalyze a</span>
<span class="sd">                          kinetic reaction</span>
<span class="sd">                        - substrates: sorted substrate IDs for substrates that are</span>
<span class="sd">                          needed for kinetic saturation terms</span>
<span class="sd">                        - all_kcats: (n rxns, 3) min, mean and max kcat value for</span>
<span class="sd">                          each reaction</span>
<span class="sd">                        - all_saturations: sympy str representation of a list of</span>
<span class="sd">                          saturation terms (eg. &#39;[s[0] / (1 + s[0]), 2 / (2 + s[1])]&#39;)</span>
<span class="sd">                        - all_enzymes: sympy str representation of enzymes for each</span>
<span class="sd">                          reaction (e.g. &#39;[e[0], e[2], e[1]]&#39;)</span>
<span class="sd">                        - constraint_is_kcat_only: True if reaction only has kcat</span>
<span class="sd">                          values and no saturation terms</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Ordered lists of constraint related IDs</span>
        <span class="n">rxns</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">constraints</span><span class="p">)</span>
        <span class="n">enzymes</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">({</span><span class="n">c</span><span class="p">[</span><span class="s2">&quot;enzyme&quot;</span><span class="p">]</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">constraints</span><span class="o">.</span><span class="n">values</span><span class="p">()})</span>
        <span class="n">substrates</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span>
            <span class="p">{</span>
                <span class="k">match</span><span class="o">.</span><span class="n">strip</span><span class="p">(</span><span class="s1">&#39;&quot;&#39;</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">constraints</span><span class="o">.</span><span class="n">values</span><span class="p">()</span>
                <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">c</span><span class="p">[</span><span class="s2">&quot;saturation&quot;</span><span class="p">]</span>
                <span class="k">for</span> <span class="n">match</span> <span class="ow">in</span> <span class="n">re</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="s1">&#39;&quot;.+?&quot;&#39;</span><span class="p">,</span> <span class="n">s</span><span class="p">)</span>
            <span class="p">}</span>
        <span class="p">)</span>

        <span class="c1"># Mapping to replace molecule IDs with generic list strings</span>
        <span class="n">enzyme_sub</span> <span class="o">=</span> <span class="p">{</span><span class="n">e</span><span class="p">:</span> <span class="s2">&quot;e[</span><span class="si">{}</span><span class="s2">]&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">e</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">enzymes</span><span class="p">)}</span>
        <span class="n">substrate_sub</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s1">&#39;&quot;</span><span class="si">{}</span><span class="s1">&quot;&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">s</span><span class="p">):</span> <span class="s2">&quot;s[</span><span class="si">{}</span><span class="s2">]&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">s</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">substrates</span><span class="p">)</span>
        <span class="p">}</span>

        <span class="c1"># Mapping to replace generic list strings with sympy variables.</span>
        <span class="c1"># Need separate mapping from above because sympy handles &#39;[]&#39; as indexing</span>
        <span class="c1"># so location tags are not parsed properly.</span>
        <span class="n">enzyme_symbols</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;e&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">sp</span><span class="o">.</span><span class="n">symbols</span><span class="p">(</span><span class="s2">&quot;e[</span><span class="si">{}</span><span class="s2">]&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span><span class="p">))</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">enzymes</span><span class="p">))]</span>
        <span class="p">}</span>
        <span class="n">substrate_symbols</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;s&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">sp</span><span class="o">.</span><span class="n">symbols</span><span class="p">(</span><span class="s2">&quot;s[</span><span class="si">{}</span><span class="s2">]&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span><span class="p">))</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">substrates</span><span class="p">))]</span>
        <span class="p">}</span>

        <span class="c1"># Values to return</span>
        <span class="n">all_kcats</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="nb">len</span><span class="p">(</span><span class="n">rxns</span><span class="p">),</span> <span class="mi">3</span><span class="p">))</span>
        <span class="n">all_saturations</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">all_enzymes</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">constraint_is_kcat_only</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="c1"># Extract out data from each constraint</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">rxn</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">rxns</span><span class="p">):</span>
            <span class="n">kcats</span> <span class="o">=</span> <span class="n">constraints</span><span class="p">[</span><span class="n">rxn</span><span class="p">][</span><span class="s2">&quot;kcat&quot;</span><span class="p">]</span>
            <span class="n">saturation</span> <span class="o">=</span> <span class="n">constraints</span><span class="p">[</span><span class="n">rxn</span><span class="p">][</span><span class="s2">&quot;saturation&quot;</span><span class="p">]</span>
            <span class="n">enzyme</span> <span class="o">=</span> <span class="n">constraints</span><span class="p">[</span><span class="n">rxn</span><span class="p">][</span><span class="s2">&quot;enzyme&quot;</span><span class="p">]</span>

            <span class="c1"># Parse saturation equations into sympy format</span>
            <span class="c1"># If no saturation data is known, assume open range from 0 to 1</span>
            <span class="n">saturations</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">sat</span> <span class="ow">in</span> <span class="n">saturation</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">sat</span> <span class="o">==</span> <span class="s2">&quot;1&quot;</span><span class="p">:</span>
                    <span class="k">continue</span>

                <span class="k">for</span> <span class="n">token</span><span class="p">,</span> <span class="n">replace</span> <span class="ow">in</span> <span class="n">substrate_sub</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                    <span class="n">sat</span> <span class="o">=</span> <span class="n">sat</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">token</span><span class="p">,</span> <span class="n">replace</span><span class="p">)</span>
                <span class="n">saturations</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">parse_expr</span><span class="p">(</span><span class="n">sat</span><span class="p">,</span> <span class="n">local_dict</span><span class="o">=</span><span class="n">substrate_symbols</span><span class="p">))</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">saturations</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">saturations</span> <span class="o">=</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span>
                <span class="n">constraint_is_kcat_only</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">constraint_is_kcat_only</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>

            <span class="c1"># Save values for this constraint</span>
            <span class="n">all_kcats</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="p">:]</span> <span class="o">=</span> <span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">min</span><span class="p">(</span><span class="n">kcats</span><span class="p">),</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">kcats</span><span class="p">),</span> <span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">kcats</span><span class="p">)]</span>
            <span class="n">all_saturations</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">saturations</span><span class="p">)</span>
            <span class="n">all_enzymes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                <span class="n">parse_expr</span><span class="p">(</span><span class="n">enzyme_sub</span><span class="p">[</span><span class="n">enzyme</span><span class="p">],</span> <span class="n">local_dict</span><span class="o">=</span><span class="n">enzyme_symbols</span><span class="p">)</span>
            <span class="p">)</span>

        <span class="c1"># Convert to str to save as class attr to be executed</span>
        <span class="n">all_saturations</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">all_saturations</span><span class="p">)</span>
        <span class="n">all_enzymes</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">all_enzymes</span><span class="p">)</span>
        <span class="n">constraint_is_kcat_only</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">constraint_is_kcat_only</span><span class="p">)</span>

        <span class="k">return</span> <span class="p">(</span>
            <span class="n">rxns</span><span class="p">,</span>
            <span class="n">enzymes</span><span class="p">,</span>
            <span class="n">substrates</span><span class="p">,</span>
            <span class="n">all_kcats</span><span class="p">,</span>
            <span class="n">all_saturations</span><span class="p">,</span>
            <span class="n">all_enzymes</span><span class="p">,</span>
            <span class="n">constraint_is_kcat_only</span><span class="p">,</span>
        <span class="p">)</span></div>


<div class="viewcode-block" id="Metabolism._is_transport_rxn">
<a class="viewcode-back" href="../../../../../reference/api/reconstruction/reconstruction.ecoli.dataclasses.process.metabolism.html#reconstruction.ecoli.dataclasses.process.metabolism.Metabolism._is_transport_rxn">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">_is_transport_rxn</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">stoich</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Determines if the metabolic reaction with a given stoichiometry is a</span>
<span class="sd">        transport reactions that transports metabolites between different</span>
<span class="sd">        compartments. A metabolic reaction is considered to be a transport</span>
<span class="sd">        reaction if the substrate set and the product share the same metabolite</span>
<span class="sd">        tagged into different compartments.</span>

<span class="sd">        Args:</span>
<span class="sd">                stoich: Stoichiometry of the metabolic reaction::</span>

<span class="sd">                        {</span>
<span class="sd">                        metabolite ID (str): stoichiometric coefficient (int)</span>
<span class="sd">                        }</span>

<span class="sd">        Returns:</span>
<span class="sd">                True if given stoichiometry is a transport reaction</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">is_transport_rxn</span> <span class="o">=</span> <span class="kc">False</span>

        <span class="c1"># Get IDs of all substrates and products</span>
        <span class="n">substrates</span><span class="p">,</span> <span class="n">products</span> <span class="o">=</span> <span class="p">[],</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">mol_id</span><span class="p">,</span> <span class="n">coeff</span> <span class="ow">in</span> <span class="n">stoich</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="k">if</span> <span class="n">coeff</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">substrates</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">mol_id</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">products</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">mol_id</span><span class="p">)</span>

        <span class="c1"># Get mapping from IDs to IDs without compartments</span>
        <span class="n">substrates_tagged_to_no_tag</span> <span class="o">=</span> <span class="p">{</span>
            <span class="n">mol_id</span><span class="p">:</span> <span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;\[.*]&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">mol_id</span><span class="p">)</span> <span class="k">for</span> <span class="n">mol_id</span> <span class="ow">in</span> <span class="n">substrates</span>
        <span class="p">}</span>
        <span class="n">products_tagged_to_no_tag</span> <span class="o">=</span> <span class="p">{</span>
            <span class="n">mol_id</span><span class="p">:</span> <span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;\[.*]&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">mol_id</span><span class="p">)</span> <span class="k">for</span> <span class="n">mol_id</span> <span class="ow">in</span> <span class="n">products</span>
        <span class="p">}</span>

        <span class="n">overlap_no_tag</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">substrates_tagged_to_no_tag</span><span class="o">.</span><span class="n">values</span><span class="p">())</span> <span class="o">&amp;</span> <span class="nb">set</span><span class="p">(</span>
            <span class="n">products_tagged_to_no_tag</span><span class="o">.</span><span class="n">values</span><span class="p">()</span>
        <span class="p">)</span>

        <span class="k">for</span> <span class="n">mol_id_no_tag</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="n">overlap_no_tag</span><span class="p">):</span>
            <span class="n">substrates_tagged</span> <span class="o">=</span> <span class="p">[</span>
                <span class="n">mol_tagged</span>
                <span class="k">for</span> <span class="n">mol_tagged</span> <span class="ow">in</span> <span class="n">substrates</span>
                <span class="k">if</span> <span class="n">substrates_tagged_to_no_tag</span><span class="p">[</span><span class="n">mol_tagged</span><span class="p">]</span> <span class="o">==</span> <span class="n">mol_id_no_tag</span>
            <span class="p">]</span>
            <span class="n">products_tagged</span> <span class="o">=</span> <span class="p">[</span>
                <span class="n">mol_tagged</span>
                <span class="k">for</span> <span class="n">mol_tagged</span> <span class="ow">in</span> <span class="n">products</span>
                <span class="k">if</span> <span class="n">products_tagged_to_no_tag</span><span class="p">[</span><span class="n">mol_tagged</span><span class="p">]</span> <span class="o">==</span> <span class="n">mol_id_no_tag</span>
            <span class="p">]</span>

            <span class="n">overlap_tagged</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">substrates_tagged</span><span class="p">)</span> <span class="o">&amp;</span> <span class="nb">set</span><span class="p">(</span><span class="n">products_tagged</span><span class="p">)</span>

            <span class="c1"># Tag reaction as a transport reaction if there is no overlap</span>
            <span class="c1"># between those substrates and products with locations included</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">overlap_tagged</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">is_transport_rxn</span> <span class="o">=</span> <span class="kc">True</span>
                <span class="k">break</span>

        <span class="k">return</span> <span class="n">is_transport_rxn</span></div>
</div>



<div class="viewcode-block" id="np_apply_along_axis">
<a class="viewcode-back" href="../../../../../reference/api/reconstruction/reconstruction.ecoli.dataclasses.process.metabolism.html#reconstruction.ecoli.dataclasses.process.metabolism.np_apply_along_axis">[docs]</a>
<span class="nd">@njit</span>
<span class="k">def</span><span class="w"> </span><span class="nf">np_apply_along_axis</span><span class="p">(</span><span class="n">func1d</span><span class="p">,</span> <span class="n">axis</span><span class="p">,</span> <span class="n">arr</span><span class="p">):</span>
    <span class="k">if</span> <span class="n">arr</span><span class="o">.</span><span class="n">ndim</span> <span class="o">!=</span> <span class="mi">2</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="s2">&quot;Array must have 2 dimensions.&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">axis</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">]:</span>
        <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="s2">&quot;Axis must be 0 or 1.&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">axis</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">result</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">empty</span><span class="p">(</span><span class="n">arr</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">result</span><span class="p">)):</span>
            <span class="n">result</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">func1d</span><span class="p">(</span><span class="n">arr</span><span class="p">[:,</span> <span class="n">i</span><span class="p">])</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">result</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">empty</span><span class="p">(</span><span class="n">arr</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">result</span><span class="p">)):</span>
            <span class="n">result</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">func1d</span><span class="p">(</span><span class="n">arr</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="p">:])</span>
    <span class="k">return</span> <span class="n">result</span></div>



<div class="viewcode-block" id="np_prod">
<a class="viewcode-back" href="../../../../../reference/api/reconstruction/reconstruction.ecoli.dataclasses.process.metabolism.html#reconstruction.ecoli.dataclasses.process.metabolism.np_prod">[docs]</a>
<span class="nd">@njit</span>
<span class="k">def</span><span class="w"> </span><span class="nf">np_prod</span><span class="p">(</span><span class="n">array</span><span class="p">,</span> <span class="n">axis</span><span class="p">):</span>
    <span class="k">return</span> <span class="n">np_apply_along_axis</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">prod</span><span class="p">,</span> <span class="n">axis</span><span class="p">,</span> <span class="n">array</span><span class="p">)</span></div>



<div class="viewcode-block" id="amino_acid_synthesis_jit">
<a class="viewcode-back" href="../../../../../reference/api/reconstruction/reconstruction.ecoli.dataclasses.process.metabolism.html#reconstruction.ecoli.dataclasses.process.metabolism.amino_acid_synthesis_jit">[docs]</a>
<span class="nd">@njit</span>
<span class="k">def</span><span class="w"> </span><span class="nf">amino_acid_synthesis_jit</span><span class="p">(</span>
    <span class="n">counts_per_aa_fwd</span><span class="p">,</span>
    <span class="n">counts_per_aa_rev</span><span class="p">,</span>
    <span class="n">aa_conc</span><span class="p">,</span>
    <span class="n">aa_upstream_kms</span><span class="p">,</span>
    <span class="n">aa_kis</span><span class="p">,</span>
    <span class="n">aa_reverse_kms</span><span class="p">,</span>
    <span class="n">aa_degradation_kms</span><span class="p">,</span>
    <span class="n">aa_forward_stoich</span><span class="p">,</span>
    <span class="n">aa_kcats_fwd</span><span class="p">,</span>
    <span class="n">aa_reverse_stoich</span><span class="p">,</span>
    <span class="n">aa_kcats_rev</span><span class="p">,</span>
<span class="p">):</span>
    <span class="n">km_saturation</span> <span class="o">=</span> <span class="n">np_prod</span><span class="p">(</span><span class="mi">1</span> <span class="o">/</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="n">aa_upstream_kms</span> <span class="o">/</span> <span class="n">aa_conc</span><span class="p">),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

    <span class="c1"># Determine saturation fraction for reactions</span>
    <span class="n">forward_fraction</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">/</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="n">aa_conc</span> <span class="o">/</span> <span class="n">aa_kis</span><span class="p">)</span> <span class="o">*</span> <span class="n">km_saturation</span>
    <span class="n">reverse_fraction</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">/</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="n">aa_reverse_kms</span> <span class="o">/</span> <span class="n">aa_conc</span><span class="p">)</span>
    <span class="n">deg_fraction</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">/</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="n">aa_degradation_kms</span> <span class="o">/</span> <span class="n">aa_conc</span><span class="p">)</span>
    <span class="n">loss_fraction</span> <span class="o">=</span> <span class="n">reverse_fraction</span> <span class="o">+</span> <span class="n">deg_fraction</span>

    <span class="c1"># Calculate synthesis rate</span>
    <span class="n">synthesis</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">aa_forward_stoich</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">)</span>
<div class="viewcode-block" id="amino_acid_export_jit">
<a class="viewcode-back" href="../../../../../reference/api/reconstruction/reconstruction.ecoli.dataclasses.process.metabolism.html#reconstruction.ecoli.dataclasses.process.metabolism.amino_acid_export_jit">[docs]</a>
        <span class="o">@</span> <span class="p">(</span><span class="n">aa_kcats_fwd</span> <span class="o">*</span> <span class="n">counts_per_aa_fwd</span> <span class="o">*</span> <span class="n">forward_fraction</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">)</span>
        <span class="o">-</span> <span class="n">aa_reverse_stoich</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">)</span>
        <span class="o">@</span> <span class="p">(</span><span class="n">aa_kcats_rev</span> <span class="o">*</span> <span class="n">counts_per_aa_rev</span> <span class="o">*</span> <span class="n">reverse_fraction</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">)</span>
        <span class="o">-</span> <span class="n">aa_kcats_rev</span> <span class="o">*</span> <span class="n">counts_per_aa_rev</span> <span class="o">*</span> <span class="n">deg_fraction</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="n">synthesis</span><span class="p">,</span> <span class="n">forward_fraction</span><span class="p">,</span> <span class="n">loss_fraction</span></div>



<span class="nd">@njit</span>
<span class="k">def</span><span class="w"> </span><span class="nf">amino_acid_export_jit</span><span class="p">(</span>
    <span class="n">aa_transporters_counts</span><span class="p">,</span>
    <span class="n">aa_conc</span><span class="p">,</span>
    <span class="n">mechanistic_uptake</span><span class="p">,</span>
    <span class="n">aa_to_exporters_matrix</span><span class="p">,</span>
    <span class="n">export_kcats_per_aa</span><span class="p">,</span>
    <span class="n">aa_export_kms</span><span class="p">,</span>
<span class="p">):</span>
    <span class="k">if</span> <span class="n">mechanistic_uptake</span><span class="p">:</span>
        <span class="c1"># Export based on mechanistic model</span>
        <span class="n">trans_counts_per_aa</span> <span class="o">=</span> <span class="n">aa_to_exporters_matrix</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span>
            <span class="n">np</span><span class="o">.</span><span class="n">float64</span>
        <span class="p">)</span> <span class="o">@</span> <span class="n">aa_transporters_counts</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">)</span>
        <span class="n">export_rates</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">export_kcats_per_aa</span> <span class="o">*</span> <span class="n">trans_counts_per_aa</span> <span class="o">/</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="n">aa_export_kms</span> <span class="o">/</span> <span class="n">aa_conc</span><span class="p">)</span>
        <span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="c1"># Export is lumped with specific uptake rates in amino_acid_import</span>
        <span class="c1"># and not dependent on internal amino acid concentrations or</span>
        <span class="c1"># explicitly considered here</span>
        <span class="n">export_rates</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">aa_conc</span><span class="p">))</span>
    <span class="k">return</span> <span class="n">export_rates</span></div>



<div class="viewcode-block" id="amino_acid_import_jit">
<a class="viewcode-back" href="../../../../../reference/api/reconstruction/reconstruction.ecoli.dataclasses.process.metabolism.html#reconstruction.ecoli.dataclasses.process.metabolism.amino_acid_import_jit">[docs]</a>
<span class="nd">@njit</span>
<span class="k">def</span><span class="w"> </span><span class="nf">amino_acid_import_jit</span><span class="p">(</span>
    <span class="n">aa_in_media</span><span class="p">,</span>
    <span class="n">dry_mass</span><span class="p">,</span>
    <span class="n">internal_aa_conc</span><span class="p">,</span>
    <span class="n">aa_transporters_counts</span><span class="p">,</span>
    <span class="n">mechanistic_uptake</span><span class="p">,</span>
    <span class="n">aa_import_kis</span><span class="p">,</span>
    <span class="n">aa_to_importers_matrix</span><span class="p">,</span>
    <span class="n">import_kcats_per_aa</span><span class="p">,</span>
    <span class="n">max_specific_import_rates</span><span class="p">,</span>
<span class="p">):</span>
    <span class="n">saturation</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">/</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="n">internal_aa_conc</span> <span class="o">/</span> <span class="n">aa_import_kis</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">mechanistic_uptake</span><span class="p">:</span>
        <span class="c1"># Uptake based on mechanistic model</span>
        <span class="n">counts_per_aa</span> <span class="o">=</span> <span class="n">aa_to_importers_matrix</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span>
            <span class="n">np</span><span class="o">.</span><span class="n">float64</span>
        <span class="p">)</span> <span class="o">@</span> <span class="n">aa_transporters_counts</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">)</span>
        <span class="n">import_rates</span> <span class="o">=</span> <span class="n">import_kcats_per_aa</span> <span class="o">*</span> <span class="n">counts_per_aa</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">import_rates</span> <span class="o">=</span> <span class="n">max_specific_import_rates</span> <span class="o">*</span> <span class="n">dry_mass</span>

    <span class="k">return</span> <span class="n">import_rates</span> <span class="o">*</span> <span class="n">saturation</span> <span class="o">*</span> <span class="n">aa_in_media</span></div>



<span class="c1"># Class used to update metabolite concentrations based on the current nutrient conditions</span>
<div class="viewcode-block" id="ConcentrationUpdates">
<a class="viewcode-back" href="../../../../../reference/api/reconstruction/reconstruction.ecoli.dataclasses.process.metabolism.html#reconstruction.ecoli.dataclasses.process.metabolism.ConcentrationUpdates">[docs]</a>
<span class="k">class</span><span class="w"> </span><span class="nc">ConcentrationUpdates</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
    <span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">concDict</span><span class="p">,</span>
        <span class="n">relative_changes</span><span class="p">,</span>
        <span class="n">equilibriumReactions</span><span class="p">,</span>
        <span class="n">exchange_data_dict</span><span class="p">,</span>
        <span class="n">all_metabolite_ids</span><span class="p">,</span>
        <span class="n">linked_metabolites</span><span class="p">,</span>
    <span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">units</span> <span class="o">=</span> <span class="n">units</span><span class="o">.</span><span class="n">getUnit</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">concDict</span><span class="o">.</span><span class="n">values</span><span class="p">())[</span><span class="mi">0</span><span class="p">])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">default_concentrations_dict</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span>
            <span class="p">(</span><span class="n">key</span><span class="p">,</span> <span class="n">concDict</span><span class="p">[</span><span class="n">key</span><span class="p">]</span><span class="o">.</span><span class="n">asNumber</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">units</span><span class="p">))</span> <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">concDict</span>
        <span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">exchange_fluxes</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_exchange_flux_present</span><span class="p">(</span><span class="n">exchange_data_dict</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">relative_changes</span> <span class="o">=</span> <span class="n">relative_changes</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_all_metabolite_ids</span> <span class="o">=</span> <span class="n">all_metabolite_ids</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">linked_metabolites</span> <span class="o">=</span> <span class="n">linked_metabolites</span>

        <span class="c1"># factor of internal amino acid increase if amino acids present in nutrients</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">molecule_scale_factors</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;L-ALPHA-ALANINE[c]&quot;</span><span class="p">:</span> <span class="mf">2.0</span><span class="p">,</span>
            <span class="s2">&quot;ARG[c]&quot;</span><span class="p">:</span> <span class="mf">2.0</span><span class="p">,</span>
            <span class="s2">&quot;ASN[c]&quot;</span><span class="p">:</span> <span class="mf">2.0</span><span class="p">,</span>
            <span class="s2">&quot;L-ASPARTATE[c]&quot;</span><span class="p">:</span> <span class="mf">2.0</span><span class="p">,</span>
            <span class="s2">&quot;CYS[c]&quot;</span><span class="p">:</span> <span class="mf">2.0</span><span class="p">,</span>
            <span class="s2">&quot;GLT[c]&quot;</span><span class="p">:</span> <span class="mf">2.0</span><span class="p">,</span>
            <span class="s2">&quot;GLN[c]&quot;</span><span class="p">:</span> <span class="mf">2.0</span><span class="p">,</span>
            <span class="s2">&quot;GLY[c]&quot;</span><span class="p">:</span> <span class="mf">2.0</span><span class="p">,</span>
            <span class="s2">&quot;HIS[c]&quot;</span><span class="p">:</span> <span class="mf">2.0</span><span class="p">,</span>
            <span class="s2">&quot;ILE[c]&quot;</span><span class="p">:</span> <span class="mf">2.0</span><span class="p">,</span>
            <span class="s2">&quot;LEU[c]&quot;</span><span class="p">:</span> <span class="mf">2.0</span><span class="p">,</span>
            <span class="s2">&quot;LYS[c]&quot;</span><span class="p">:</span> <span class="mf">2.0</span><span class="p">,</span>
            <span class="s2">&quot;MET[c]&quot;</span><span class="p">:</span> <span class="mf">2.0</span><span class="p">,</span>
            <span class="s2">&quot;PHE[c]&quot;</span><span class="p">:</span> <span class="mf">2.0</span><span class="p">,</span>
            <span class="s2">&quot;PRO[c]&quot;</span><span class="p">:</span> <span class="mf">2.0</span><span class="p">,</span>
            <span class="s2">&quot;SER[c]&quot;</span><span class="p">:</span> <span class="mf">2.0</span><span class="p">,</span>
            <span class="s2">&quot;THR[c]&quot;</span><span class="p">:</span> <span class="mf">2.0</span><span class="p">,</span>
            <span class="s2">&quot;TRP[c]&quot;</span><span class="p">:</span> <span class="mf">2.0</span><span class="p">,</span>
            <span class="s2">&quot;TYR[c]&quot;</span><span class="p">:</span> <span class="mf">2.0</span><span class="p">,</span>
            <span class="s2">&quot;L-SELENOCYSTEINE[c]&quot;</span><span class="p">:</span> <span class="mf">2.0</span><span class="p">,</span>
            <span class="s2">&quot;VAL[c]&quot;</span><span class="p">:</span> <span class="mf">2.0</span><span class="p">,</span>
        <span class="p">}</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">molecule_set_amounts</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_add_molecule_amounts</span><span class="p">(</span>
            <span class="n">equilibriumReactions</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">default_concentrations_dict</span>
        <span class="p">)</span>

    <span class="c1"># return adjustments to concDict based on nutrient conditions</span>
<div class="viewcode-block" id="ConcentrationUpdates.concentrations_based_on_nutrients">
<a class="viewcode-back" href="../../../../../reference/api/reconstruction/reconstruction.ecoli.dataclasses.process.metabolism.html#reconstruction.ecoli.dataclasses.process.metabolism.ConcentrationUpdates.concentrations_based_on_nutrients">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">concentrations_based_on_nutrients</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">media_id</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">imports</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">conversion_units</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">units</span><span class="o">.</span><span class="n">Unum</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]:</span>
        <span class="k">if</span> <span class="n">conversion_units</span><span class="p">:</span>
            <span class="n">conversion</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">units</span><span class="o">.</span><span class="n">asNumber</span><span class="p">(</span><span class="n">conversion_units</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">conversion</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">units</span>

        <span class="k">if</span> <span class="n">imports</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">media_id</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">imports</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">exchange_fluxes</span><span class="p">[</span><span class="n">media_id</span><span class="p">]</span>

        <span class="n">concentrationsDict</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">default_concentrations_dict</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

        <span class="n">metaboliteTargetIds</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">concentrationsDict</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
        <span class="n">concentrations</span> <span class="o">=</span> <span class="n">conversion</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span>
            <span class="p">[</span><span class="n">concentrationsDict</span><span class="p">[</span><span class="n">k</span><span class="p">]</span> <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">metaboliteTargetIds</span><span class="p">]</span>
        <span class="p">)</span>
        <span class="n">concDict</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">metaboliteTargetIds</span><span class="p">,</span> <span class="n">concentrations</span><span class="p">))</span>

        <span class="k">if</span> <span class="n">imports</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="c1"># For faster conversions than .asNumber(conversion_units) for each setAmount</span>
            <span class="k">if</span> <span class="n">conversion_units</span><span class="p">:</span>
                <span class="n">conversion_to_no_units</span> <span class="o">=</span> <span class="n">conversion_units</span><span class="o">.</span><span class="n">asUnit</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">units</span><span class="p">)</span>

            <span class="c1"># Adjust for measured concentration changes in different media</span>
            <span class="k">if</span> <span class="n">media_id</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">relative_changes</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">mol_id</span><span class="p">,</span> <span class="n">conc_change</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">relative_changes</span><span class="p">[</span><span class="n">media_id</span><span class="p">]</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                    <span class="k">if</span> <span class="n">mol_id</span> <span class="ow">in</span> <span class="n">concDict</span><span class="p">:</span>
                        <span class="n">concDict</span><span class="p">[</span><span class="n">mol_id</span><span class="p">]</span> <span class="o">*=</span> <span class="n">conc_change</span>

            <span class="k">for</span> <span class="n">moleculeName</span><span class="p">,</span> <span class="n">setAmount</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">molecule_set_amounts</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                <span class="k">if</span> <span class="p">(</span>
                    <span class="n">moleculeName</span> <span class="ow">in</span> <span class="n">imports</span>
                    <span class="ow">and</span> <span class="p">(</span>
                        <span class="n">moleculeName</span><span class="p">[:</span><span class="o">-</span><span class="mi">3</span><span class="p">]</span> <span class="o">+</span> <span class="s2">&quot;[c]&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">molecule_scale_factors</span>
                        <span class="ow">or</span> <span class="n">moleculeName</span> <span class="o">==</span> <span class="s2">&quot;L-SELENOCYSTEINE[c]&quot;</span>
                    <span class="p">)</span>
                <span class="p">)</span> <span class="ow">or</span> <span class="p">(</span>
                    <span class="n">moleculeName</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">molecule_scale_factors</span>
                    <span class="ow">and</span> <span class="n">moleculeName</span><span class="p">[:</span><span class="o">-</span><span class="mi">3</span><span class="p">]</span> <span class="o">+</span> <span class="s2">&quot;[p]&quot;</span> <span class="ow">in</span> <span class="n">imports</span>
                <span class="p">):</span>
                    <span class="k">if</span> <span class="n">conversion_units</span><span class="p">:</span>
                        <span class="n">setAmount</span> <span class="o">=</span> <span class="p">(</span><span class="n">setAmount</span> <span class="o">/</span> <span class="n">conversion_to_no_units</span><span class="p">)</span><span class="o">.</span><span class="n">asNumber</span><span class="p">()</span>
                    <span class="n">concDict</span><span class="p">[</span><span class="n">moleculeName</span><span class="p">]</span> <span class="o">=</span> <span class="n">setAmount</span>

        <span class="k">for</span> <span class="n">met</span><span class="p">,</span> <span class="n">linked</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">linked_metabolites</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">concDict</span><span class="p">[</span><span class="n">met</span><span class="p">]</span> <span class="o">=</span> <span class="n">concDict</span><span class="p">[</span><span class="n">linked</span><span class="p">[</span><span class="s2">&quot;lead&quot;</span><span class="p">]]</span> <span class="o">*</span> <span class="n">linked</span><span class="p">[</span><span class="s2">&quot;ratio&quot;</span><span class="p">]</span>

        <span class="k">return</span> <span class="n">concDict</span></div>


<div class="viewcode-block" id="ConcentrationUpdates._exchange_flux_present">
<a class="viewcode-back" href="../../../../../reference/api/reconstruction/reconstruction.ecoli.dataclasses.process.metabolism.html#reconstruction.ecoli.dataclasses.process.metabolism.ConcentrationUpdates._exchange_flux_present">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">_exchange_flux_present</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">exchange_data</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">set</span><span class="p">[</span><span class="nb">str</span><span class="p">]]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Caches the presence of exchanges in each media condition based on</span>
<span class="sd">        exchange_data to set concentrations in concentrations_based_on_nutrients().</span>

<span class="sd">        Args:</span>
<span class="sd">                exchange_data: dictionary of exchange data for all media conditions</span>
<span class="sd">                        with keys::</span>

<span class="sd">                                {importUnconstrainedExchangeMolecules (dict[str, set[str]]):</span>
<span class="sd">                                        exchange molecules (with location tag) for each media key</span>
<span class="sd">                                        that do not have an upper bound on their flux,</span>
<span class="sd">                                importConstrainedExchangeMolecules (dict[str,</span>
<span class="sd">                                        dict[str, float with mol/mass/time units]]): constrained</span>
<span class="sd">                                        molecules (with location tag) for each media key with upper</span>
<span class="sd">                                        bound flux constraints}</span>

<span class="sd">        Returns:</span>
<span class="sd">                Sets of molecules IDs (with location tags) that can be imported</span>
<span class="sd">                for each media ID</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">exchange_fluxes</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">media</span><span class="p">,</span> <span class="n">env</span> <span class="ow">in</span> <span class="n">exchange_data</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">exchange_fluxes</span><span class="p">[</span><span class="n">media</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="n">mol</span> <span class="k">for</span> <span class="n">mol</span><span class="p">,</span> <span class="n">conc</span> <span class="ow">in</span> <span class="n">env</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> <span class="k">if</span> <span class="n">conc</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">}</span>

        <span class="k">return</span> <span class="n">exchange_fluxes</span></div>


<div class="viewcode-block" id="ConcentrationUpdates._add_molecule_amounts">
<a class="viewcode-back" href="../../../../../reference/api/reconstruction/reconstruction.ecoli.dataclasses.process.metabolism.html#reconstruction.ecoli.dataclasses.process.metabolism.ConcentrationUpdates._add_molecule_amounts">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">_add_molecule_amounts</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">equilibriumReactions</span><span class="p">,</span> <span class="n">concDict</span><span class="p">):</span>
        <span class="n">moleculeSetAmounts</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">reaction</span> <span class="ow">in</span> <span class="n">equilibriumReactions</span><span class="p">:</span>
            <span class="c1"># We only want to do this for species with standard Michaelis-Menten kinetics initially</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">reaction</span><span class="p">[</span><span class="s2">&quot;stoichiometry&quot;</span><span class="p">])</span> <span class="o">!=</span> <span class="mi">3</span><span class="p">:</span>
                <span class="k">continue</span>

            <span class="n">moleculeName</span> <span class="o">=</span> <span class="p">[</span>
                <span class="n">mol_id</span>
                <span class="k">for</span> <span class="n">mol_id</span> <span class="ow">in</span> <span class="n">reaction</span><span class="p">[</span><span class="s2">&quot;stoichiometry&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
                <span class="k">if</span> <span class="n">mol_id</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_all_metabolite_ids</span>
            <span class="p">][</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">amountToSet</span> <span class="o">=</span> <span class="mf">1e-4</span>
            <span class="n">moleculeSetAmounts</span><span class="p">[</span><span class="n">moleculeName</span> <span class="o">+</span> <span class="s2">&quot;[p]&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">amountToSet</span> <span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">units</span>
            <span class="n">moleculeSetAmounts</span><span class="p">[</span><span class="n">moleculeName</span> <span class="o">+</span> <span class="s2">&quot;[c]&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">amountToSet</span> <span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">units</span>

        <span class="k">for</span> <span class="n">moleculeName</span><span class="p">,</span> <span class="n">scaleFactor</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">molecule_scale_factors</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">moleculeSetAmounts</span><span class="p">[</span><span class="n">moleculeName</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
                <span class="n">scaleFactor</span> <span class="o">*</span> <span class="n">concDict</span><span class="p">[</span><span class="n">moleculeName</span><span class="p">]</span> <span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">units</span>
            <span class="p">)</span>
        <span class="k">return</span> <span class="n">moleculeSetAmounts</span></div>
</div>

</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2021-2026, The Vivarium E. coli Authors.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>