

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../../../../">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>reconstruction.ecoli.dataclasses.process.replication &mdash; Vivarium E. coli 1.0.0 documentation</title>
      <link rel="stylesheet" type="text/css" href="../../../../../_static/pygments.css?v=b86133f3" />
      <link rel="stylesheet" type="text/css" href="../../../../../_static/css/theme.css?v=e59714d7" />

  
      <script src="../../../../../_static/jquery.js?v=5d32c60e"></script>
      <script src="../../../../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../../../../../_static/documentation_options.js?v=8d563738"></script>
      <script src="../../../../../_static/doctools.js?v=9bcbadda"></script>
      <script src="../../../../../_static/sphinx_highlight.js?v=dc90522c"></script>
      <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../../../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../../../genindex.html" />
    <link rel="search" title="Search" href="../../../../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../../../index.html" class="icon icon-home">
            Vivarium E. coli
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../../../../stores.html">Stores</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../processes.html">Processes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../composites.html">Composites</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../experiments.html">Experiments</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../workflows.html">Workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../output.html">Output</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../tutorial.html">Tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../docs.html">Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../hpc.html">HPC Clusters</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../gcloud.html">Google Cloud</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../ci.html">Continuous Integration</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../pycharm.html">PyCharm Setup</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../diffs.html">Model Differences From wcEcoli</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../reference/api_ref.html">API Reference</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../../../index.html">Vivarium E. coli</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../../../index.html">Module code</a></li>
      <li class="breadcrumb-item active">reconstruction.ecoli.dataclasses.process.replication</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for reconstruction.ecoli.dataclasses.process.replication</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">SimulationData for replication process</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">collections</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">wholecell.utils</span><span class="w"> </span><span class="kn">import</span> <span class="n">units</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">wholecell.utils.polymerize</span><span class="w"> </span><span class="kn">import</span> <span class="n">polymerize</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">wholecell.utils.random</span><span class="w"> </span><span class="kn">import</span> <span class="n">stochasticRound</span>


<span class="c1"># Maximum allowed simulation time step. Used to create buffer for sequences.</span>
<span class="n">MAX_TIMESTEP</span> <span class="o">=</span> <span class="mi">2</span>


<div class="viewcode-block" id="SiteNotFoundError">
<a class="viewcode-back" href="../../../../../reference/api/reconstruction/reconstruction.ecoli.dataclasses.process.replication.html#reconstruction.ecoli.dataclasses.process.replication.SiteNotFoundError">[docs]</a>
<span class="k">class</span><span class="w"> </span><span class="nc">SiteNotFoundError</span><span class="p">(</span><span class="ne">Exception</span><span class="p">):</span>
    <span class="k">pass</span></div>



<div class="viewcode-block" id="Replication">
<a class="viewcode-back" href="../../../../../reference/api/reconstruction/reconstruction.ecoli.dataclasses.process.replication.html#reconstruction.ecoli.dataclasses.process.replication.Replication">[docs]</a>
<span class="k">class</span><span class="w"> </span><span class="nc">Replication</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    SimulationData for the replication process</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">raw_data</span><span class="p">,</span> <span class="n">sim_data</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_n_nt_types</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">sim_data</span><span class="o">.</span><span class="n">dntp_code_to_id_ordered</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_build_sequence</span><span class="p">(</span><span class="n">raw_data</span><span class="p">,</span> <span class="n">sim_data</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_build_gene_data</span><span class="p">(</span><span class="n">raw_data</span><span class="p">,</span> <span class="n">sim_data</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_build_sites</span><span class="p">(</span><span class="n">raw_data</span><span class="p">,</span> <span class="n">sim_data</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_build_replication</span><span class="p">(</span><span class="n">raw_data</span><span class="p">,</span> <span class="n">sim_data</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_build_motifs</span><span class="p">(</span><span class="n">raw_data</span><span class="p">,</span> <span class="n">sim_data</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_build_elongation_rates</span><span class="p">(</span><span class="n">raw_data</span><span class="p">,</span> <span class="n">sim_data</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">c_period</span> <span class="o">=</span> <span class="n">sim_data</span><span class="o">.</span><span class="n">constants</span><span class="o">.</span><span class="n">c_period</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">d_period</span> <span class="o">=</span> <span class="n">sim_data</span><span class="o">.</span><span class="n">constants</span><span class="o">.</span><span class="n">d_period</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">c_period_in_mins</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">c_period</span><span class="o">.</span><span class="n">asNumber</span><span class="p">(</span><span class="n">units</span><span class="o">.</span><span class="n">min</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">d_period_in_mins</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">d_period</span><span class="o">.</span><span class="n">asNumber</span><span class="p">(</span><span class="n">units</span><span class="o">.</span><span class="n">min</span><span class="p">)</span>

<div class="viewcode-block" id="Replication._build_sequence">
<a class="viewcode-back" href="../../../../../reference/api/reconstruction/reconstruction.ecoli.dataclasses.process.replication.html#reconstruction.ecoli.dataclasses.process.replication.Replication._build_sequence">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">_build_sequence</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">raw_data</span><span class="p">,</span> <span class="n">sim_data</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">genome_sequence</span> <span class="o">=</span> <span class="n">raw_data</span><span class="o">.</span><span class="n">genome_sequence</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">genome_sequence_rc</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome_sequence</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">genome_length</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genome_sequence</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">genome_A_count</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome_sequence</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;A&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">genome_T_count</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome_sequence</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;T&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">genome_G_count</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome_sequence</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;G&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">genome_C_count</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome_sequence</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;C&quot;</span><span class="p">)</span></div>


<div class="viewcode-block" id="Replication._build_gene_data">
<a class="viewcode-back" href="../../../../../reference/api/reconstruction/reconstruction.ecoli.dataclasses.process.replication.html#reconstruction.ecoli.dataclasses.process.replication.Replication._build_gene_data">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">_build_gene_data</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">raw_data</span><span class="p">,</span> <span class="n">sim_data</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Build gene-associated simulation data from raw data.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">def</span><span class="w"> </span><span class="nf">extract_data</span><span class="p">(</span><span class="n">raw</span><span class="p">,</span> <span class="n">key</span><span class="p">,</span> <span class="n">use_first_from_list</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">use_first_from_list</span><span class="p">:</span>
                <span class="n">data</span> <span class="o">=</span> <span class="p">[</span><span class="n">row</span><span class="p">[</span><span class="n">key</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">raw</span><span class="p">]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">data</span> <span class="o">=</span> <span class="p">[</span><span class="n">row</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">raw</span><span class="p">]</span>
            <span class="n">dtype</span> <span class="o">=</span> <span class="s2">&quot;U</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">max</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">d</span><span class="p">)</span> <span class="k">for</span> <span class="n">d</span> <span class="ow">in</span> <span class="n">data</span> <span class="k">if</span> <span class="n">d</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">))</span>
            <span class="k">return</span> <span class="n">data</span><span class="p">,</span> <span class="n">dtype</span>

        <span class="n">names</span><span class="p">,</span> <span class="n">name_dtype</span> <span class="o">=</span> <span class="n">extract_data</span><span class="p">(</span><span class="n">raw_data</span><span class="o">.</span><span class="n">genes</span><span class="p">,</span> <span class="s2">&quot;id&quot;</span><span class="p">)</span>
        <span class="n">symbols</span><span class="p">,</span> <span class="n">symbol_dtype</span> <span class="o">=</span> <span class="n">extract_data</span><span class="p">(</span><span class="n">raw_data</span><span class="o">.</span><span class="n">genes</span><span class="p">,</span> <span class="s2">&quot;symbol&quot;</span><span class="p">)</span>
        <span class="n">cistron_ids</span><span class="p">,</span> <span class="n">cistron_dtype</span> <span class="o">=</span> <span class="n">extract_data</span><span class="p">(</span>
            <span class="n">raw_data</span><span class="o">.</span><span class="n">genes</span><span class="p">,</span> <span class="s2">&quot;rna_ids&quot;</span><span class="p">,</span> <span class="n">use_first_from_list</span><span class="o">=</span><span class="kc">True</span>
        <span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">gene_data</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span>
            <span class="nb">len</span><span class="p">(</span><span class="n">raw_data</span><span class="o">.</span><span class="n">genes</span><span class="p">),</span>
            <span class="n">dtype</span><span class="o">=</span><span class="p">[</span>
                <span class="p">(</span><span class="s2">&quot;name&quot;</span><span class="p">,</span> <span class="n">name_dtype</span><span class="p">),</span>
                <span class="p">(</span><span class="s2">&quot;symbol&quot;</span><span class="p">,</span> <span class="n">symbol_dtype</span><span class="p">),</span>
                <span class="p">(</span><span class="s2">&quot;cistron_id&quot;</span><span class="p">,</span> <span class="n">cistron_dtype</span><span class="p">),</span>
            <span class="p">],</span>
        <span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">gene_data</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">names</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">gene_data</span><span class="p">[</span><span class="s2">&quot;symbol&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">symbols</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">gene_data</span><span class="p">[</span><span class="s2">&quot;cistron_id&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">cistron_ids</span></div>


<div class="viewcode-block" id="Replication._build_sites">
<a class="viewcode-back" href="../../../../../reference/api/reconstruction/reconstruction.ecoli.dataclasses.process.replication.html#reconstruction.ecoli.dataclasses.process.replication.Replication._build_sites">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">_build_sites</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">raw_data</span><span class="p">,</span> <span class="n">sim_data</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Build simulation data associated with DNA sites from raw_data.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">def</span><span class="w"> </span><span class="nf">get_site_center_coordinates</span><span class="p">(</span><span class="n">site_id</span><span class="p">):</span>
<span class="w">            </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">            Calculate the center coordinates (rounded average of left and right</span>
<span class="sd">            end coordinates) of a given DNA site.</span>
<span class="sd">            &quot;&quot;&quot;</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">left</span><span class="p">,</span> <span class="n">right</span> <span class="o">=</span> <span class="n">sim_data</span><span class="o">.</span><span class="n">getter</span><span class="o">.</span><span class="n">get_genomic_coordinates</span><span class="p">(</span><span class="n">site_id</span><span class="p">)</span>
                <span class="n">center_coordinate</span> <span class="o">=</span> <span class="nb">round</span><span class="p">((</span><span class="n">left</span> <span class="o">+</span> <span class="n">right</span><span class="p">)</span> <span class="o">/</span> <span class="mi">2</span><span class="p">)</span>
            <span class="k">except</span> <span class="p">(</span><span class="ne">KeyError</span><span class="p">,</span> <span class="ne">TypeError</span><span class="p">):</span>
                <span class="k">raise</span> <span class="n">SiteNotFoundError</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;Coordinates of DNA site with ID </span><span class="si">{</span><span class="n">site_id</span><span class="si">}</span><span class="s2"> were not found in raw_data.&quot;</span>
                <span class="p">)</span>

            <span class="k">return</span> <span class="n">center_coordinate</span>

        <span class="c1"># Get coordinates of oriC and terC</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">oric_coordinate</span> <span class="o">=</span> <span class="n">get_site_center_coordinates</span><span class="p">(</span>
            <span class="n">sim_data</span><span class="o">.</span><span class="n">molecule_ids</span><span class="o">.</span><span class="n">oriC_site</span>
        <span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">terc_coordinate</span> <span class="o">=</span> <span class="n">get_site_center_coordinates</span><span class="p">(</span>
            <span class="n">sim_data</span><span class="o">.</span><span class="n">molecule_ids</span><span class="o">.</span><span class="n">terC_site</span>
        <span class="p">)</span></div>


<div class="viewcode-block" id="Replication._build_replication">
<a class="viewcode-back" href="../../../../../reference/api/reconstruction/reconstruction.ecoli.dataclasses.process.replication.html#reconstruction.ecoli.dataclasses.process.replication.Replication._build_replication">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">_build_replication</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">raw_data</span><span class="p">,</span> <span class="n">sim_data</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Build replication-associated simulation data from raw data.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Map ATGC to 8 bit integers</span>
        <span class="n">numerical_sequence</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">empty</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genome_length</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">int8</span><span class="p">)</span>
        <span class="n">ntMapping</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">OrderedDict</span><span class="p">(</span>
            <span class="p">[</span>
                <span class="p">(</span><span class="n">ntp_code</span><span class="p">,</span> <span class="n">i</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">ntp_code</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">sim_data</span><span class="o">.</span><span class="n">dntp_code_to_id_ordered</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
            <span class="p">]</span>
        <span class="p">)</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">letter</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">raw_data</span><span class="o">.</span><span class="n">genome_sequence</span><span class="p">):</span>
            <span class="n">numerical_sequence</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">ntMapping</span><span class="p">[</span>
                <span class="n">letter</span>
            <span class="p">]</span>  <span class="c1"># Build genome sequence as small integers</span>

        <span class="c1"># Create 4 possible polymerization sequences</span>
        <span class="c1"># Forward sequence includes oriC</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">forward_sequence</span> <span class="o">=</span> <span class="n">numerical_sequence</span><span class="p">[</span>
            <span class="n">np</span><span class="o">.</span><span class="n">hstack</span><span class="p">(</span>
                <span class="p">(</span>
                    <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">oric_coordinate</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome_length</span><span class="p">),</span>
                    <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">terc_coordinate</span><span class="p">),</span>
                <span class="p">)</span>
            <span class="p">)</span>
        <span class="p">]</span>

        <span class="c1"># Reverse sequence includes terC</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">reverse_sequence</span> <span class="o">=</span> <span class="n">numerical_sequence</span><span class="p">[</span>
            <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">oric_coordinate</span> <span class="o">-</span> <span class="mi">1</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">terc_coordinate</span> <span class="o">-</span> <span class="mi">1</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">)</span>
        <span class="p">]</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">forward_complement_sequence</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_complement_sequence</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">forward_sequence</span>
        <span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">reverse_complement_sequence</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_complement_sequence</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">reverse_sequence</span>
        <span class="p">)</span>

        <span class="k">assert</span> <span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">forward_sequence</span><span class="o">.</span><span class="n">size</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">reverse_sequence</span><span class="o">.</span><span class="n">size</span>
            <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome_length</span>
        <span class="p">)</span>

        <span class="c1"># Log array of lengths of each replichore</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">replichore_lengths</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span>
            <span class="p">[</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">forward_sequence</span><span class="o">.</span><span class="n">size</span><span class="p">,</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">reverse_sequence</span><span class="o">.</span><span class="n">size</span><span class="p">,</span>
            <span class="p">]</span>
        <span class="p">)</span>

        <span class="c1"># Determine size of the matrix used by polymerize function</span>
        <span class="n">maxLen</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">int64</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">replichore_lengths</span><span class="o">.</span><span class="n">max</span><span class="p">()</span>
            <span class="c1"># Add buffer to account for elongation by max timestep</span>
            <span class="o">+</span> <span class="n">MAX_TIMESTEP</span>
            <span class="o">*</span> <span class="n">sim_data</span><span class="o">.</span><span class="n">constants</span><span class="o">.</span><span class="n">replisome_elongation_rate</span><span class="o">.</span><span class="n">asNumber</span><span class="p">(</span><span class="n">units</span><span class="o">.</span><span class="n">nt</span> <span class="o">/</span> <span class="n">units</span><span class="o">.</span><span class="n">s</span><span class="p">)</span>
        <span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">replication_sequences</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">empty</span><span class="p">((</span><span class="mi">4</span><span class="p">,</span> <span class="n">maxLen</span><span class="p">),</span> <span class="n">np</span><span class="o">.</span><span class="n">int8</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">replication_sequences</span><span class="o">.</span><span class="n">fill</span><span class="p">(</span><span class="n">polymerize</span><span class="o">.</span><span class="n">PAD_VALUE</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">replication_sequences</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">forward_sequence</span><span class="o">.</span><span class="n">size</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">forward_sequence</span>
        <span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">replication_sequences</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">forward_complement_sequence</span><span class="o">.</span><span class="n">size</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">forward_complement_sequence</span>
        <span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">replication_sequences</span><span class="p">[</span><span class="mi">2</span><span class="p">,</span> <span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">reverse_sequence</span><span class="o">.</span><span class="n">size</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">reverse_sequence</span>
        <span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">replication_sequences</span><span class="p">[</span><span class="mi">3</span><span class="p">,</span> <span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">reverse_complement_sequence</span><span class="o">.</span><span class="n">size</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">reverse_complement_sequence</span>
        <span class="p">)</span>

        <span class="c1"># Get polymerized nucleotide weights</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">replication_monomer_weights</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">sim_data</span><span class="o">.</span><span class="n">getter</span><span class="o">.</span><span class="n">get_masses</span><span class="p">(</span><span class="n">sim_data</span><span class="o">.</span><span class="n">molecule_groups</span><span class="o">.</span><span class="n">dntps</span><span class="p">)</span>
            <span class="o">-</span> <span class="n">sim_data</span><span class="o">.</span><span class="n">getter</span><span class="o">.</span><span class="n">get_masses</span><span class="p">([</span><span class="n">sim_data</span><span class="o">.</span><span class="n">molecule_ids</span><span class="o">.</span><span class="n">ppi</span><span class="p">])</span>
        <span class="p">)</span> <span class="o">/</span> <span class="n">sim_data</span><span class="o">.</span><span class="n">constants</span><span class="o">.</span><span class="n">n_avogadro</span>

        <span class="c1"># Placeholder value for &quot;child_domains&quot; attribute of domains without</span>
        <span class="c1"># children domains</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">no_child_place_holder</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span></div>


<div class="viewcode-block" id="Replication._build_motifs">
<a class="viewcode-back" href="../../../../../reference/api/reconstruction/reconstruction.ecoli.dataclasses.process.replication.html#reconstruction.ecoli.dataclasses.process.replication.Replication._build_motifs">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">_build_motifs</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">raw_data</span><span class="p">,</span> <span class="n">sim_data</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Build simulation data associated with sequence motifs from raw_data.</span>
<span class="sd">        Coordinates of all motifs are calculated based on the given sequences</span>
<span class="sd">        of the genome and the motifs.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Initialize dictionary of motif coordinates</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">motif_coordinates</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>

        <span class="k">for</span> <span class="n">motif</span> <span class="ow">in</span> <span class="n">raw_data</span><span class="o">.</span><span class="n">sequence_motifs</span><span class="p">:</span>
            <span class="c1"># Keys are the IDs of the motifs; Values are arrays of the motif&#39;s</span>
            <span class="c1"># coordinates.</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">motif_coordinates</span><span class="p">[</span><span class="n">motif</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_motif_coordinates</span><span class="p">(</span>
                <span class="n">motif</span><span class="p">[</span><span class="s2">&quot;length&quot;</span><span class="p">],</span> <span class="n">motif</span><span class="p">[</span><span class="s2">&quot;sequences&quot;</span><span class="p">]</span>
            <span class="p">)</span></div>


<div class="viewcode-block" id="Replication._get_complement_sequence">
<a class="viewcode-back" href="../../../../../reference/api/reconstruction/reconstruction.ecoli.dataclasses.process.replication.html#reconstruction.ecoli.dataclasses.process.replication.Replication._get_complement_sequence">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">_get_complement_sequence</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">sequenceVector</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Calculates the vector for a complement sequence of a DNA sequence given</span>
<span class="sd">        in vector form.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_n_nt_types</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="o">-</span> <span class="n">sequenceVector</span></div>


<div class="viewcode-block" id="Replication._get_motif_coordinates">
<a class="viewcode-back" href="../../../../../reference/api/reconstruction/reconstruction.ecoli.dataclasses.process.replication.html#reconstruction.ecoli.dataclasses.process.replication.Replication._get_motif_coordinates">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">_get_motif_coordinates</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">motif_length</span><span class="p">,</span> <span class="n">motif_sequences</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Finds the coordinates of all sequence motifs of a specific type. The</span>
<span class="sd">        coordinates are given as the positions of the midpoint of the motif</span>
<span class="sd">        relative to the oriC in base pairs.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Append the first n bases to the end of the sequences to account for</span>
        <span class="c1"># motifs that span the two endpoints</span>
        <span class="n">extended_sequence</span> <span class="o">=</span> <span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">genome_sequence</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome_sequence</span><span class="p">[:</span> <span class="p">(</span><span class="n">motif_length</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)]</span>
        <span class="p">)</span>
        <span class="n">extended_rc_sequence</span> <span class="o">=</span> <span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">genome_sequence_rc</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome_sequence_rc</span><span class="p">[:</span> <span class="p">(</span><span class="n">motif_length</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)]</span>
        <span class="p">)</span>

        <span class="c1"># Initialize list for coordinates of motifs</span>
        <span class="n">coordinates</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="c1"># Loop through all possible motif sequences</span>
        <span class="k">for</span> <span class="n">sequence</span> <span class="ow">in</span> <span class="n">motif_sequences</span><span class="p">:</span>
            <span class="c1"># Find occurrences of the motif in the original sequence</span>
            <span class="n">loc</span> <span class="o">=</span> <span class="n">extended_sequence</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>  <span class="c1"># Returns -1 if not found</span>

            <span class="k">while</span> <span class="n">loc</span> <span class="o">!=</span> <span class="o">-</span><span class="mi">1</span><span class="p">:</span>
                <span class="n">coordinates</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">loc</span> <span class="o">+</span> <span class="n">motif_length</span> <span class="o">//</span> <span class="mi">2</span><span class="p">)</span>
                <span class="n">loc</span> <span class="o">=</span> <span class="n">extended_sequence</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="n">sequence</span><span class="p">,</span> <span class="n">loc</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>

            <span class="c1"># Find occurrences of the motif in the reverse complement</span>
            <span class="n">loc</span> <span class="o">=</span> <span class="n">extended_rc_sequence</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>

            <span class="k">while</span> <span class="n">loc</span> <span class="o">!=</span> <span class="o">-</span><span class="mi">1</span><span class="p">:</span>
                <span class="n">coordinates</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">genome_length</span> <span class="o">-</span> <span class="n">loc</span> <span class="o">-</span> <span class="n">motif_length</span> <span class="o">+</span> <span class="n">motif_length</span> <span class="o">//</span> <span class="mi">2</span>
                <span class="p">)</span>
                <span class="n">loc</span> <span class="o">=</span> <span class="n">extended_rc_sequence</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="n">sequence</span><span class="p">,</span> <span class="n">loc</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>

        <span class="c1"># Compute coordinates relative to oriC</span>
        <span class="n">motif_coordinates</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_relative_coordinates</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">coordinates</span><span class="p">))</span>
        <span class="n">motif_coordinates</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>

        <span class="k">return</span> <span class="n">motif_coordinates</span></div>


<div class="viewcode-block" id="Replication._get_relative_coordinates">
<a class="viewcode-back" href="../../../../../reference/api/reconstruction/reconstruction.ecoli.dataclasses.process.replication.html#reconstruction.ecoli.dataclasses.process.replication.Replication._get_relative_coordinates">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">_get_relative_coordinates</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">coordinates</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Converts an array of genomic coordinates into coordinates relative to</span>
<span class="sd">        the origin of replication.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">relative_coordinates</span> <span class="o">=</span> <span class="p">(</span>
            <span class="p">(</span><span class="n">coordinates</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">terc_coordinate</span><span class="p">)</span> <span class="o">%</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome_length</span>
            <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">terc_coordinate</span>
            <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">oric_coordinate</span>
        <span class="p">)</span>

        <span class="n">relative_coordinates</span><span class="p">[</span><span class="n">relative_coordinates</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>

        <span class="k">return</span> <span class="n">relative_coordinates</span></div>


<div class="viewcode-block" id="Replication._build_elongation_rates">
<a class="viewcode-back" href="../../../../../reference/api/reconstruction/reconstruction.ecoli.dataclasses.process.replication.html#reconstruction.ecoli.dataclasses.process.replication.Replication._build_elongation_rates">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">_build_elongation_rates</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">raw_data</span><span class="p">,</span> <span class="n">sim_data</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">basal_elongation_rate</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span>
            <span class="nb">round</span><span class="p">(</span>
                <span class="n">sim_data</span><span class="o">.</span><span class="n">constants</span><span class="o">.</span><span class="n">replisome_elongation_rate</span><span class="o">.</span><span class="n">asNumber</span><span class="p">(</span>
                    <span class="n">units</span><span class="o">.</span><span class="n">nt</span> <span class="o">/</span> <span class="n">units</span><span class="o">.</span><span class="n">s</span>
                <span class="p">)</span>
            <span class="p">)</span>
        <span class="p">)</span></div>


<div class="viewcode-block" id="Replication.make_elongation_rates">
<a class="viewcode-back" href="../../../../../reference/api/reconstruction/reconstruction.ecoli.dataclasses.process.replication.html#reconstruction.ecoli.dataclasses.process.replication.Replication.make_elongation_rates">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">make_elongation_rates</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">random</span><span class="p">,</span> <span class="n">replisomes</span><span class="p">,</span> <span class="n">base</span><span class="p">,</span> <span class="n">time_step</span><span class="p">):</span>
        <span class="n">rates</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">full</span><span class="p">(</span>
            <span class="n">replisomes</span><span class="p">,</span> <span class="n">stochasticRound</span><span class="p">(</span><span class="n">random</span><span class="p">,</span> <span class="n">base</span> <span class="o">*</span> <span class="n">time_step</span><span class="p">),</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int64</span>
        <span class="p">)</span>

        <span class="k">return</span> <span class="n">rates</span></div>


<div class="viewcode-block" id="Replication.get_average_copy_number">
<a class="viewcode-back" href="../../../../../reference/api/reconstruction/reconstruction.ecoli.dataclasses.process.replication.html#reconstruction.ecoli.dataclasses.process.replication.Replication.get_average_copy_number">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">get_average_copy_number</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">tau</span><span class="p">,</span> <span class="n">coords</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Calculates the average copy number of a gene throughout the cell cycle</span>
<span class="sd">        given the location of the gene in coordinates.</span>

<span class="sd">        Args:</span>
<span class="sd">                tau (float): expected doubling time in minutes</span>
<span class="sd">                coords (int or ndarray[int]): chromosome coordinates of genes</span>

<span class="sd">        Returns:</span>
<span class="sd">                float or ndarray[float] (matches length of coords): average copy</span>
<span class="sd">                number of each gene expected at a doubling time, tau</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">right_replichore_length</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">replichore_lengths</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">left_replichore_length</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">replichore_lengths</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>

        <span class="c1"># Calculate the relative position of the gene along the chromosome</span>
        <span class="c1"># from its coordinate</span>
        <span class="n">relative_pos</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">coords</span><span class="p">,</span> <span class="nb">float</span><span class="p">)</span>
        <span class="n">relative_pos</span><span class="p">[</span><span class="n">coords</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="n">relative_pos</span><span class="p">[</span><span class="n">coords</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">]</span> <span class="o">/</span> <span class="n">right_replichore_length</span>
        <span class="n">relative_pos</span><span class="p">[</span><span class="n">coords</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="o">-</span><span class="n">relative_pos</span><span class="p">[</span><span class="n">coords</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">]</span> <span class="o">/</span> <span class="n">left_replichore_length</span>

        <span class="c1"># Return the predicted average copy number</span>
        <span class="n">n_avg_copy</span> <span class="o">=</span> <span class="mi">2</span> <span class="o">**</span> <span class="p">(</span>
            <span class="p">((</span><span class="mi">1</span> <span class="o">-</span> <span class="n">relative_pos</span><span class="p">)</span> <span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">c_period_in_mins</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">d_period_in_mins</span><span class="p">)</span> <span class="o">/</span> <span class="n">tau</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="n">n_avg_copy</span></div>
</div>

</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2021-2026, The Vivarium E. coli Authors.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>