

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../../../../">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>reconstruction.ecoli.dataclasses.process.translation &mdash; Vivarium E. coli 1.0.0 documentation</title>
      <link rel="stylesheet" type="text/css" href="../../../../../_static/pygments.css?v=b86133f3" />
      <link rel="stylesheet" type="text/css" href="../../../../../_static/css/theme.css?v=e59714d7" />

  
      <script src="../../../../../_static/jquery.js?v=5d32c60e"></script>
      <script src="../../../../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../../../../../_static/documentation_options.js?v=8d563738"></script>
      <script src="../../../../../_static/doctools.js?v=9bcbadda"></script>
      <script src="../../../../../_static/sphinx_highlight.js?v=dc90522c"></script>
      <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../../../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../../../genindex.html" />
    <link rel="search" title="Search" href="../../../../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../../../index.html" class="icon icon-home">
            Vivarium E. coli
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../../../../stores.html">Stores</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../processes.html">Processes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../composites.html">Composites</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../experiments.html">Experiments</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../workflows.html">Workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../output.html">Output</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../tutorial.html">Tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../docs.html">Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../hpc.html">HPC Clusters</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../gcloud.html">Google Cloud</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../ci.html">Continuous Integration</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../pycharm.html">PyCharm Setup</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../diffs.html">Model Differences From wcEcoli</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../reference/api_ref.html">API Reference</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../../../index.html">Vivarium E. coli</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../../../index.html">Module code</a></li>
      <li class="breadcrumb-item active">reconstruction.ecoli.dataclasses.process.translation</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for reconstruction.ecoli.dataclasses.process.translation</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">SimulationData for translation process</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">.replication</span><span class="w"> </span><span class="kn">import</span> <span class="n">MAX_TIMESTEP</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">wholecell.utils</span><span class="w"> </span><span class="kn">import</span> <span class="n">data</span><span class="p">,</span> <span class="n">units</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">wholecell.utils.unit_struct_array</span><span class="w"> </span><span class="kn">import</span> <span class="n">UnitStructArray</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">wholecell.utils.polymerize</span><span class="w"> </span><span class="kn">import</span> <span class="n">polymerize</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">wholecell.utils.random</span><span class="w"> </span><span class="kn">import</span> <span class="n">make_elongation_rates</span>


<div class="viewcode-block" id="Translation">
<a class="viewcode-back" href="../../../../../reference/api/reconstruction/reconstruction.ecoli.dataclasses.process.translation.html#reconstruction.ecoli.dataclasses.process.translation.Translation">[docs]</a>
<span class="k">class</span><span class="w"> </span><span class="nc">Translation</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Translation&quot;&quot;&quot;</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">raw_data</span><span class="p">,</span> <span class="n">sim_data</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">next_aa_pad</span> <span class="o">=</span> <span class="p">(</span>
            <span class="mi">1</span>  <span class="c1"># Need an extra amino acid in sequences lengths to find next one</span>
        <span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_build_monomer_data</span><span class="p">(</span><span class="n">raw_data</span><span class="p">,</span> <span class="n">sim_data</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_build_translation</span><span class="p">(</span><span class="n">raw_data</span><span class="p">,</span> <span class="n">sim_data</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_build_translation_efficiency</span><span class="p">(</span><span class="n">raw_data</span><span class="p">,</span> <span class="n">sim_data</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_build_elongation_rates</span><span class="p">(</span><span class="n">raw_data</span><span class="p">,</span> <span class="n">sim_data</span><span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">__getstate__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Return the state to pickle with translationSequences removed and</span>
<span class="sd">        only storing data from translationSequences with pad values stripped.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">state</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">dissoc_strict</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="vm">__dict__</span><span class="p">,</span> <span class="p">(</span><span class="s2">&quot;translation_sequences&quot;</span><span class="p">,))</span>
        <span class="n">state</span><span class="p">[</span><span class="s2">&quot;sequences&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span>
            <span class="p">[</span><span class="n">seq</span><span class="p">[</span><span class="n">seq</span> <span class="o">!=</span> <span class="n">polymerize</span><span class="o">.</span><span class="n">PAD_VALUE</span><span class="p">]</span> <span class="k">for</span> <span class="n">seq</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">translation_sequences</span><span class="p">],</span>
            <span class="n">dtype</span><span class="o">=</span><span class="nb">object</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">state</span><span class="p">[</span><span class="s2">&quot;sequence_shape&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">translation_sequences</span><span class="o">.</span><span class="n">shape</span>
        <span class="k">return</span> <span class="n">state</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">__setstate__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">state</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Restore translationSequences and remove processed versions of the data.&quot;&quot;&quot;</span>
        <span class="n">sequences</span> <span class="o">=</span> <span class="n">state</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s2">&quot;sequences&quot;</span><span class="p">)</span>
        <span class="n">sequence_shape</span> <span class="o">=</span> <span class="n">state</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s2">&quot;sequence_shape&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="vm">__dict__</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">state</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">translation_sequences</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">full</span><span class="p">(</span>
            <span class="n">sequence_shape</span><span class="p">,</span> <span class="n">polymerize</span><span class="o">.</span><span class="n">PAD_VALUE</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int8</span>
        <span class="p">)</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">seq</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">sequences</span><span class="p">):</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">translation_sequences</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="p">:</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">)]</span> <span class="o">=</span> <span class="n">seq</span>

<div class="viewcode-block" id="Translation._build_monomer_data">
<a class="viewcode-back" href="../../../../../reference/api/reconstruction/reconstruction.ecoli.dataclasses.process.translation.html#reconstruction.ecoli.dataclasses.process.translation.Translation._build_monomer_data">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">_build_monomer_data</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">raw_data</span><span class="p">,</span> <span class="n">sim_data</span><span class="p">):</span>
        <span class="c1"># Get set of all cistrons IDs with an associated gene and right and left</span>
        <span class="c1"># end positions</span>
        <span class="n">rna_id_to_gene_id</span> <span class="o">=</span> <span class="p">{</span><span class="n">gene</span><span class="p">[</span><span class="s2">&quot;rna_ids&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]:</span> <span class="n">gene</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]</span> <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">raw_data</span><span class="o">.</span><span class="n">genes</span><span class="p">}</span>
        <span class="n">gene_id_to_left_end_pos</span> <span class="o">=</span> <span class="p">{</span>
            <span class="n">gene</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]:</span> <span class="n">gene</span><span class="p">[</span><span class="s2">&quot;left_end_pos&quot;</span><span class="p">]</span> <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">raw_data</span><span class="o">.</span><span class="n">genes</span>
        <span class="p">}</span>
        <span class="n">gene_id_to_right_end_pos</span> <span class="o">=</span> <span class="p">{</span>
            <span class="n">gene</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]:</span> <span class="n">gene</span><span class="p">[</span><span class="s2">&quot;right_end_pos&quot;</span><span class="p">]</span> <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">raw_data</span><span class="o">.</span><span class="n">genes</span>
        <span class="p">}</span>

        <span class="n">all_mRNA_cistrons</span> <span class="o">=</span> <span class="p">{</span>
            <span class="n">rna</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]</span>
            <span class="k">for</span> <span class="n">rna</span> <span class="ow">in</span> <span class="n">raw_data</span><span class="o">.</span><span class="n">rnas</span>
            <span class="k">if</span> <span class="n">rna</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]</span> <span class="ow">in</span> <span class="n">rna_id_to_gene_id</span>
            <span class="ow">and</span> <span class="n">gene_id_to_left_end_pos</span><span class="p">[</span><span class="n">rna_id_to_gene_id</span><span class="p">[</span><span class="n">rna</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]]]</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
            <span class="ow">and</span> <span class="n">gene_id_to_right_end_pos</span><span class="p">[</span><span class="n">rna_id_to_gene_id</span><span class="p">[</span><span class="n">rna</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]]]</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
            <span class="ow">and</span> <span class="n">rna</span><span class="p">[</span><span class="s2">&quot;type&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;mRNA&quot;</span>
        <span class="p">}</span>

        <span class="c1"># Get mappings from monomer IDs to cistron IDs</span>
        <span class="c1"># TODO (ggsun): Handle cistrons with two or more associated proteins</span>
        <span class="n">monomer_id_to_cistron_id</span> <span class="o">=</span> <span class="p">{</span>
            <span class="n">rna</span><span class="p">[</span><span class="s2">&quot;monomer_ids&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]:</span> <span class="n">rna</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]</span>
            <span class="k">for</span> <span class="n">rna</span> <span class="ow">in</span> <span class="n">raw_data</span><span class="o">.</span><span class="n">rnas</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">rna</span><span class="p">[</span><span class="s2">&quot;monomer_ids&quot;</span><span class="p">])</span> <span class="o">&gt;</span> <span class="mi">0</span>
        <span class="p">}</span>

        <span class="c1"># Select proteins with valid sequences and mappings to valid mRNAs</span>
        <span class="n">all_proteins</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">protein</span> <span class="ow">in</span> <span class="n">raw_data</span><span class="o">.</span><span class="n">proteins</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">sim_data</span><span class="o">.</span><span class="n">getter</span><span class="o">.</span><span class="n">is_valid_molecule</span><span class="p">(</span><span class="n">protein</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]):</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="n">rna_id</span> <span class="o">=</span> <span class="n">monomer_id_to_cistron_id</span><span class="p">[</span><span class="n">protein</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]]</span>
                <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                    <span class="k">continue</span>
                <span class="k">if</span> <span class="n">rna_id</span> <span class="ow">in</span> <span class="n">all_mRNA_cistrons</span><span class="p">:</span>
                    <span class="n">all_proteins</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">protein</span><span class="p">)</span>

        <span class="c1"># Get protein IDs with compartments</span>
        <span class="n">protein_ids</span> <span class="o">=</span> <span class="p">[</span><span class="n">protein</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]</span> <span class="k">for</span> <span class="n">protein</span> <span class="ow">in</span> <span class="n">all_proteins</span><span class="p">]</span>
        <span class="n">protein_compartments</span> <span class="o">=</span> <span class="n">sim_data</span><span class="o">.</span><span class="n">getter</span><span class="o">.</span><span class="n">get_compartments</span><span class="p">(</span><span class="n">protein_ids</span><span class="p">)</span>
        <span class="k">assert</span> <span class="nb">all</span><span class="p">([</span><span class="nb">len</span><span class="p">(</span><span class="n">loc</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span> <span class="k">for</span> <span class="n">loc</span> <span class="ow">in</span> <span class="n">protein_compartments</span><span class="p">])</span>
        <span class="n">protein_ids_with_compartments</span> <span class="o">=</span> <span class="p">[</span>
            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">protein_id</span><span class="si">}</span><span class="s2">[</span><span class="si">{</span><span class="n">loc</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2">]&quot;</span>
            <span class="k">for</span> <span class="p">(</span><span class="n">protein_id</span><span class="p">,</span> <span class="n">loc</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">protein_ids</span><span class="p">,</span> <span class="n">protein_compartments</span><span class="p">)</span>
        <span class="p">]</span>
        <span class="n">n_proteins</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">protein_ids</span><span class="p">)</span>

        <span class="c1"># Get cistron IDs associated to each monomer</span>
        <span class="n">cistron_ids</span> <span class="o">=</span> <span class="p">[</span>
            <span class="n">monomer_id_to_cistron_id</span><span class="p">[</span><span class="n">protein</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]]</span> <span class="k">for</span> <span class="n">protein</span> <span class="ow">in</span> <span class="n">all_proteins</span>
        <span class="p">]</span>

        <span class="c1"># Get lengths and amino acids counts of each protein</span>
        <span class="n">protein_seqs</span> <span class="o">=</span> <span class="n">sim_data</span><span class="o">.</span><span class="n">getter</span><span class="o">.</span><span class="n">get_sequences</span><span class="p">(</span><span class="n">protein_ids</span><span class="p">)</span>
        <span class="n">lengths</span> <span class="o">=</span> <span class="p">[</span><span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span> <span class="k">for</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">protein_seqs</span><span class="p">]</span>
        <span class="n">aa_counts</span> <span class="o">=</span> <span class="p">[</span>
            <span class="p">[</span><span class="n">seq</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="n">aa</span><span class="p">)</span> <span class="k">for</span> <span class="n">aa</span> <span class="ow">in</span> <span class="n">sim_data</span><span class="o">.</span><span class="n">amino_acid_code_to_id_ordered</span><span class="o">.</span><span class="n">keys</span><span class="p">()]</span>
            <span class="k">for</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">protein_seqs</span>
        <span class="p">]</span>
        <span class="n">n_amino_acids</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">sim_data</span><span class="o">.</span><span class="n">amino_acid_code_to_id_ordered</span><span class="p">)</span>

        <span class="c1"># Get molecular weights</span>
        <span class="n">mws</span> <span class="o">=</span> <span class="n">sim_data</span><span class="o">.</span><span class="n">getter</span><span class="o">.</span><span class="n">get_masses</span><span class="p">(</span><span class="n">protein_ids</span><span class="p">)</span><span class="o">.</span><span class="n">asNumber</span><span class="p">(</span><span class="n">units</span><span class="o">.</span><span class="n">g</span> <span class="o">/</span> <span class="n">units</span><span class="o">.</span><span class="n">mol</span><span class="p">)</span>

        <span class="c1"># Calculate degradation rates based on N-rule</span>
        <span class="n">deg_rate_units</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">/</span> <span class="n">units</span><span class="o">.</span><span class="n">s</span>
        <span class="n">n_end_rule_deg_rates</span> <span class="o">=</span> <span class="p">{</span>
            <span class="n">row</span><span class="p">[</span><span class="s2">&quot;aa_code&quot;</span><span class="p">]:</span> <span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span> <span class="o">/</span> <span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;half life&quot;</span><span class="p">]))</span><span class="o">.</span><span class="n">asNumber</span><span class="p">(</span><span class="n">deg_rate_units</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">raw_data</span><span class="o">.</span><span class="n">protein_half_lives_n_end_rule</span>
        <span class="p">}</span>

        <span class="c1"># Get degradation rates from measured protein half lives</span>
        <span class="n">measured_deg_rates</span> <span class="o">=</span> <span class="p">{</span>
            <span class="n">p</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]:</span> <span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span> <span class="o">/</span> <span class="n">p</span><span class="p">[</span><span class="s2">&quot;half life&quot;</span><span class="p">])</span><span class="o">.</span><span class="n">asNumber</span><span class="p">(</span><span class="n">deg_rate_units</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">raw_data</span><span class="o">.</span><span class="n">protein_half_lives_measured</span>
        <span class="p">}</span>

        <span class="c1"># Get degradation rates from Nagar (2022) pulsed-SILAC half lives</span>
        <span class="n">pulsed_silac_deg_rates</span> <span class="o">=</span> <span class="p">{</span>
            <span class="n">p</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]:</span> <span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span> <span class="o">/</span> <span class="n">p</span><span class="p">[</span><span class="s2">&quot;half_life&quot;</span><span class="p">])</span><span class="o">.</span><span class="n">asNumber</span><span class="p">(</span><span class="n">deg_rate_units</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">raw_data</span><span class="o">.</span><span class="n">protein_half_lives_pulsed_silac</span>
        <span class="p">}</span>

        <span class="n">deg_rate</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">all_proteins</span><span class="p">))</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">protein</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">all_proteins</span><span class="p">):</span>
            <span class="c1"># Use measured degradation rates if available</span>
            <span class="k">if</span> <span class="n">protein</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]</span> <span class="ow">in</span> <span class="n">measured_deg_rates</span><span class="p">:</span>
                <span class="n">deg_rate</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">measured_deg_rates</span><span class="p">[</span><span class="n">protein</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]]</span>
            <span class="k">elif</span> <span class="n">protein</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]</span> <span class="ow">in</span> <span class="n">pulsed_silac_deg_rates</span><span class="p">:</span>
                <span class="n">deg_rate</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">pulsed_silac_deg_rates</span><span class="p">[</span><span class="n">protein</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]]</span>
            <span class="c1"># If measured rates are unavailable, use N-end rule</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">seq</span> <span class="o">=</span> <span class="n">protein</span><span class="p">[</span><span class="s2">&quot;seq&quot;</span><span class="p">]</span>
                <span class="k">assert</span> <span class="p">(</span>
                    <span class="n">seq</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;M&quot;</span>
                <span class="p">)</span>  <span class="c1"># All protein sequences should start with methionine</span>

                <span class="c1"># Set N-end residue as second amino acid if initial methionine</span>
                <span class="c1"># is cleaved</span>
                <span class="n">n_end_residue</span> <span class="o">=</span> <span class="n">seq</span><span class="p">[</span><span class="n">protein</span><span class="p">[</span><span class="s2">&quot;cleavage_of_initial_methionine&quot;</span><span class="p">]]</span>
                <span class="n">deg_rate</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">n_end_rule_deg_rates</span><span class="p">[</span><span class="n">n_end_residue</span><span class="p">]</span>

        <span class="n">max_protein_id_length</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span>
            <span class="nb">len</span><span class="p">(</span><span class="n">protein_id</span><span class="p">)</span> <span class="k">for</span> <span class="n">protein_id</span> <span class="ow">in</span> <span class="n">protein_ids_with_compartments</span>
        <span class="p">)</span>
        <span class="n">max_cistron_id_length</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">cistron_id</span><span class="p">)</span> <span class="k">for</span> <span class="n">cistron_id</span> <span class="ow">in</span> <span class="n">cistron_ids</span><span class="p">)</span>
        <span class="n">monomer_data</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span>
            <span class="n">n_proteins</span><span class="p">,</span>
            <span class="n">dtype</span><span class="o">=</span><span class="p">[</span>
                <span class="p">(</span><span class="s2">&quot;id&quot;</span><span class="p">,</span> <span class="s2">&quot;U</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">max_protein_id_length</span><span class="p">)),</span>
                <span class="p">(</span><span class="s2">&quot;cistron_id&quot;</span><span class="p">,</span> <span class="s2">&quot;U</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">max_cistron_id_length</span><span class="p">)),</span>
                <span class="p">(</span><span class="s2">&quot;deg_rate&quot;</span><span class="p">,</span> <span class="s2">&quot;f8&quot;</span><span class="p">),</span>
                <span class="p">(</span><span class="s2">&quot;length&quot;</span><span class="p">,</span> <span class="s2">&quot;i8&quot;</span><span class="p">),</span>
                <span class="p">(</span><span class="s2">&quot;aa_counts&quot;</span><span class="p">,</span> <span class="s2">&quot;</span><span class="si">{}</span><span class="s2">i8&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">n_amino_acids</span><span class="p">)),</span>
                <span class="p">(</span><span class="s2">&quot;mw&quot;</span><span class="p">,</span> <span class="s2">&quot;f8&quot;</span><span class="p">),</span>
            <span class="p">],</span>
        <span class="p">)</span>

        <span class="n">monomer_data</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">protein_ids_with_compartments</span>
        <span class="n">monomer_data</span><span class="p">[</span><span class="s2">&quot;cistron_id&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">cistron_ids</span>
        <span class="n">monomer_data</span><span class="p">[</span><span class="s2">&quot;deg_rate&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">deg_rate</span>
        <span class="n">monomer_data</span><span class="p">[</span><span class="s2">&quot;length&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">lengths</span>
        <span class="n">monomer_data</span><span class="p">[</span><span class="s2">&quot;aa_counts&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">aa_counts</span>
        <span class="n">monomer_data</span><span class="p">[</span><span class="s2">&quot;mw&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">mws</span>

        <span class="n">field_units</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;id&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
            <span class="s2">&quot;cistron_id&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
            <span class="s2">&quot;deg_rate&quot;</span><span class="p">:</span> <span class="n">deg_rate_units</span><span class="p">,</span>
            <span class="s2">&quot;length&quot;</span><span class="p">:</span> <span class="n">units</span><span class="o">.</span><span class="n">aa</span><span class="p">,</span>
            <span class="s2">&quot;aa_counts&quot;</span><span class="p">:</span> <span class="n">units</span><span class="o">.</span><span class="n">aa</span><span class="p">,</span>
            <span class="s2">&quot;mw&quot;</span><span class="p">:</span> <span class="n">units</span><span class="o">.</span><span class="n">g</span> <span class="o">/</span> <span class="n">units</span><span class="o">.</span><span class="n">mol</span><span class="p">,</span>
        <span class="p">}</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">monomer_data</span> <span class="o">=</span> <span class="n">UnitStructArray</span><span class="p">(</span><span class="n">monomer_data</span><span class="p">,</span> <span class="n">field_units</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">n_monomers</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">monomer_data</span><span class="p">)</span></div>


<div class="viewcode-block" id="Translation._build_translation">
<a class="viewcode-back" href="../../../../../reference/api/reconstruction/reconstruction.ecoli.dataclasses.process.translation.html#reconstruction.ecoli.dataclasses.process.translation.Translation._build_translation">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">_build_translation</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">raw_data</span><span class="p">,</span> <span class="n">sim_data</span><span class="p">):</span>
        <span class="n">sequences</span> <span class="o">=</span> <span class="n">sim_data</span><span class="o">.</span><span class="n">getter</span><span class="o">.</span><span class="n">get_sequences</span><span class="p">(</span>
            <span class="p">[</span><span class="n">protein_id</span><span class="p">[:</span><span class="o">-</span><span class="mi">3</span><span class="p">]</span> <span class="k">for</span> <span class="n">protein_id</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">monomer_data</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]]</span>
        <span class="p">)</span>

        <span class="n">max_len</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">np</span><span class="o">.</span><span class="n">int64</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">monomer_data</span><span class="p">[</span><span class="s2">&quot;length&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">asNumber</span><span class="p">()</span><span class="o">.</span><span class="n">max</span><span class="p">()</span>
                <span class="c1"># Add buffer to account for elongation by max timestep</span>
                <span class="o">+</span> <span class="n">MAX_TIMESTEP</span>
                <span class="o">*</span> <span class="n">sim_data</span><span class="o">.</span><span class="n">constants</span><span class="o">.</span><span class="n">ribosome_elongation_rate_max</span><span class="o">.</span><span class="n">asNumber</span><span class="p">(</span>
                    <span class="n">units</span><span class="o">.</span><span class="n">aa</span> <span class="o">/</span> <span class="n">units</span><span class="o">.</span><span class="n">s</span>
                <span class="p">)</span>
            <span class="p">)</span>
            <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">next_aa_pad</span>
        <span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">translation_sequences</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">full</span><span class="p">(</span>
            <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">sequences</span><span class="p">),</span> <span class="n">max_len</span><span class="p">),</span> <span class="n">polymerize</span><span class="o">.</span><span class="n">PAD_VALUE</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int8</span>
        <span class="p">)</span>
        <span class="n">aa_ids_single_letter</span> <span class="o">=</span> <span class="n">sim_data</span><span class="o">.</span><span class="n">amino_acid_code_to_id_ordered</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
        <span class="n">aaMapping</span> <span class="o">=</span> <span class="p">{</span><span class="n">aa</span><span class="p">:</span> <span class="n">i</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">aa</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">aa_ids_single_letter</span><span class="p">)}</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">sequence</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">sequences</span><span class="p">):</span>
            <span class="k">for</span> <span class="n">j</span><span class="p">,</span> <span class="n">letter</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">sequence</span><span class="p">):</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">translation_sequences</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span> <span class="o">=</span> <span class="n">aaMapping</span><span class="p">[</span><span class="n">letter</span><span class="p">]</span>

        <span class="n">aaIDs</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">sim_data</span><span class="o">.</span><span class="n">amino_acid_code_to_id_ordered</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">translation_monomer_weights</span> <span class="o">=</span> <span class="p">(</span>
            <span class="p">(</span>
                <span class="n">sim_data</span><span class="o">.</span><span class="n">getter</span><span class="o">.</span><span class="n">get_masses</span><span class="p">(</span><span class="n">aaIDs</span><span class="p">)</span>
                <span class="o">-</span> <span class="n">sim_data</span><span class="o">.</span><span class="n">getter</span><span class="o">.</span><span class="n">get_masses</span><span class="p">([</span><span class="n">sim_data</span><span class="o">.</span><span class="n">molecule_ids</span><span class="o">.</span><span class="n">water</span><span class="p">])</span>
            <span class="p">)</span>
            <span class="o">/</span> <span class="n">sim_data</span><span class="o">.</span><span class="n">constants</span><span class="o">.</span><span class="n">n_avogadro</span>
        <span class="p">)</span><span class="o">.</span><span class="n">asNumber</span><span class="p">(</span><span class="n">units</span><span class="o">.</span><span class="n">fg</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">translation_end_weight</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">sim_data</span><span class="o">.</span><span class="n">getter</span><span class="o">.</span><span class="n">get_masses</span><span class="p">([</span><span class="n">sim_data</span><span class="o">.</span><span class="n">molecule_ids</span><span class="o">.</span><span class="n">water</span><span class="p">])</span>
            <span class="o">/</span> <span class="n">sim_data</span><span class="o">.</span><span class="n">constants</span><span class="o">.</span><span class="n">n_avogadro</span>
        <span class="p">)</span><span class="o">.</span><span class="n">asNumber</span><span class="p">(</span><span class="n">units</span><span class="o">.</span><span class="n">fg</span><span class="p">)</span>

        <span class="c1"># Load active ribosome footprint on RNA</span>
        <span class="n">molecule_id_to_footprint_sizes</span> <span class="o">=</span> <span class="p">{</span>
            <span class="n">row</span><span class="p">[</span><span class="s2">&quot;molecule_id&quot;</span><span class="p">]:</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;footprint_size&quot;</span><span class="p">]</span>
            <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">raw_data</span><span class="o">.</span><span class="n">footprint_sizes</span>
        <span class="p">}</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">active_ribosome_footprint_size</span> <span class="o">=</span> <span class="n">molecule_id_to_footprint_sizes</span><span class="p">[</span>
                <span class="s2">&quot;active_ribosome&quot;</span>
            <span class="p">]</span>
        <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;RNA footprint size for ribosomes not found.&quot;</span><span class="p">)</span></div>


<div class="viewcode-block" id="Translation._build_translation_efficiency">
<a class="viewcode-back" href="../../../../../reference/api/reconstruction/reconstruction.ecoli.dataclasses.process.translation.html#reconstruction.ecoli.dataclasses.process.translation.Translation._build_translation_efficiency">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">_build_translation_efficiency</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">raw_data</span><span class="p">,</span> <span class="n">sim_data</span><span class="p">):</span>
        <span class="n">monomer_ids</span> <span class="o">=</span> <span class="p">[</span><span class="n">protein_id</span><span class="p">[:</span><span class="o">-</span><span class="mi">3</span><span class="p">]</span> <span class="k">for</span> <span class="n">protein_id</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">monomer_data</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]]</span>

        <span class="c1"># Get mappings from monomer IDs to gene IDs</span>
        <span class="n">monomer_id_to_rna_id</span> <span class="o">=</span> <span class="p">{</span>
            <span class="n">rna</span><span class="p">[</span><span class="s2">&quot;monomer_ids&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]:</span> <span class="n">rna</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]</span>
            <span class="k">for</span> <span class="n">rna</span> <span class="ow">in</span> <span class="n">raw_data</span><span class="o">.</span><span class="n">rnas</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">rna</span><span class="p">[</span><span class="s2">&quot;monomer_ids&quot;</span><span class="p">])</span> <span class="o">&gt;</span> <span class="mi">0</span>
        <span class="p">}</span>
        <span class="n">rna_id_to_gene_id</span> <span class="o">=</span> <span class="p">{</span><span class="n">gene</span><span class="p">[</span><span class="s2">&quot;rna_ids&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]:</span> <span class="n">gene</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]</span> <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">raw_data</span><span class="o">.</span><span class="n">genes</span><span class="p">}</span>
        <span class="n">monomer_id_to_gene_id</span> <span class="o">=</span> <span class="p">{</span>
            <span class="n">monomer_id</span><span class="p">:</span> <span class="n">rna_id_to_gene_id</span><span class="p">[</span><span class="n">monomer_id_to_rna_id</span><span class="p">[</span><span class="n">monomer_id</span><span class="p">]]</span>
            <span class="k">for</span> <span class="n">monomer_id</span> <span class="ow">in</span> <span class="n">monomer_ids</span>
        <span class="p">}</span>

        <span class="c1"># Get mappings from gene IDs to translation efficiencies</span>
        <span class="n">gene_id_to_trl_eff</span> <span class="o">=</span> <span class="p">{</span>
            <span class="n">x</span><span class="p">[</span><span class="s2">&quot;geneId&quot;</span><span class="p">]:</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;translationEfficiency&quot;</span><span class="p">]</span>
            <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">raw_data</span><span class="o">.</span><span class="n">translation_efficiency</span>
            <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="s2">&quot;translationEfficiency&quot;</span><span class="p">],</span> <span class="nb">float</span><span class="p">)</span>
        <span class="p">}</span>

        <span class="n">trl_effs</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">monomer_id</span> <span class="ow">in</span> <span class="n">monomer_ids</span><span class="p">:</span>
            <span class="n">gene_id</span> <span class="o">=</span> <span class="n">monomer_id_to_gene_id</span><span class="p">[</span><span class="n">monomer_id</span><span class="p">]</span>

            <span class="k">if</span> <span class="n">gene_id</span> <span class="ow">in</span> <span class="n">gene_id_to_trl_eff</span><span class="p">:</span>
                <span class="n">trl_effs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gene_id_to_trl_eff</span><span class="p">[</span><span class="n">gene_id</span><span class="p">])</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">trl_effs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">)</span>

        <span class="c1"># If efficiency is unavailable, the average of existing effciencies</span>
        <span class="c1"># is used</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">translation_efficiencies_by_monomer</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">trl_effs</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">translation_efficiencies_by_monomer</span><span class="p">[</span>
            <span class="n">np</span><span class="o">.</span><span class="n">isnan</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">translation_efficiencies_by_monomer</span><span class="p">)</span>
        <span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nanmean</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">translation_efficiencies_by_monomer</span><span class="p">)</span></div>


<div class="viewcode-block" id="Translation._build_elongation_rates">
<a class="viewcode-back" href="../../../../../reference/api/reconstruction/reconstruction.ecoli.dataclasses.process.translation.html#reconstruction.ecoli.dataclasses.process.translation.Translation._build_elongation_rates">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">_build_elongation_rates</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">raw_data</span><span class="p">,</span> <span class="n">sim_data</span><span class="p">):</span>
        <span class="n">protein_ids</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">monomer_data</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]</span>
        <span class="n">ribosomal_protein_ids</span> <span class="o">=</span> <span class="n">sim_data</span><span class="o">.</span><span class="n">molecule_groups</span><span class="o">.</span><span class="n">ribosomal_proteins</span>

        <span class="n">protein_indexes</span> <span class="o">=</span> <span class="p">{</span><span class="n">protein</span><span class="p">:</span> <span class="n">index</span> <span class="k">for</span> <span class="n">index</span><span class="p">,</span> <span class="n">protein</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">protein_ids</span><span class="p">)}</span>

        <span class="n">ribosomal_proteins</span> <span class="o">=</span> <span class="p">{</span>
            <span class="n">rprotein</span><span class="p">:</span> <span class="n">protein_indexes</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">rprotein</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">rprotein</span> <span class="ow">in</span> <span class="n">ribosomal_protein_ids</span>
        <span class="p">}</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">ribosomal_protein_indexes</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span>
            <span class="p">[</span><span class="n">index</span> <span class="k">for</span> <span class="n">index</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">ribosomal_proteins</span><span class="o">.</span><span class="n">values</span><span class="p">())</span> <span class="k">if</span> <span class="n">index</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">],</span>
            <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int64</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">basal_elongation_rate</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">sim_data</span><span class="o">.</span><span class="n">constants</span><span class="o">.</span><span class="n">ribosome_elongation_rate_basal</span><span class="o">.</span><span class="n">asNumber</span><span class="p">(</span>
                <span class="n">units</span><span class="o">.</span><span class="n">aa</span> <span class="o">/</span> <span class="n">units</span><span class="o">.</span><span class="n">s</span>
            <span class="p">)</span>
        <span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">max_elongation_rate</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">sim_data</span><span class="o">.</span><span class="n">constants</span><span class="o">.</span><span class="n">ribosome_elongation_rate_max</span><span class="o">.</span><span class="n">asNumber</span><span class="p">(</span><span class="n">units</span><span class="o">.</span><span class="n">aa</span> <span class="o">/</span> <span class="n">units</span><span class="o">.</span><span class="n">s</span><span class="p">)</span>
        <span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">elongation_rates</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">full</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">n_monomers</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">basal_elongation_rate</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int64</span>
        <span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">elongation_rates</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">ribosomal_protein_indexes</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">max_elongation_rate</span></div>


<div class="viewcode-block" id="Translation.make_elongation_rates">
<a class="viewcode-back" href="../../../../../reference/api/reconstruction/reconstruction.ecoli.dataclasses.process.translation.html#reconstruction.ecoli.dataclasses.process.translation.Translation.make_elongation_rates">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">make_elongation_rates</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">random</span><span class="p">,</span> <span class="n">base</span><span class="p">,</span> <span class="n">time_step</span><span class="p">,</span> <span class="n">variable_elongation</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
        <span class="k">return</span> <span class="n">make_elongation_rates</span><span class="p">(</span>
            <span class="n">random</span><span class="p">,</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">n_monomers</span><span class="p">,</span>
            <span class="n">base</span><span class="p">,</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">ribosomal_protein_indexes</span><span class="p">,</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">max_elongation_rate</span><span class="p">,</span>
            <span class="n">time_step</span><span class="p">,</span>
            <span class="n">variable_elongation</span><span class="p">,</span>
        <span class="p">)</span></div>
</div>

</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2021-2026, The Vivarium E. coli Authors.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>