

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>wholecell.utils.spatial_tool &mdash; Vivarium E. coli 1.0.0 documentation</title>
      <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=b86133f3" />
      <link rel="stylesheet" type="text/css" href="../../../_static/css/theme.css?v=e59714d7" />

  
      <script src="../../../_static/jquery.js?v=5d32c60e"></script>
      <script src="../../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../../../_static/documentation_options.js?v=8d563738"></script>
      <script src="../../../_static/doctools.js?v=9bcbadda"></script>
      <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
      <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="wholecell.utils.test_parallelization" href="wholecell.utils.test_parallelization.html" />
    <link rel="prev" title="wholecell.utils.sparkline" href="wholecell.utils.sparkline.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../index.html" class="icon icon-home">
            Vivarium E. coli
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../../stores.html">Stores</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../processes.html">Processes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../composites.html">Composites</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../experiments.html">Experiments</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../workflows.html">Workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../output.html">Output</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../tutorial.html">Tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../docs.html">Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../hpc.html">HPC Clusters</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../gcloud.html">Google Cloud</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../ci.html">Continuous Integration</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../pycharm.html">PyCharm Setup</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../../api_ref.html">API Reference</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../ecoli/ecoli.html"><code class="docutils literal notranslate"><span class="pre">ecoli</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="../reconstruction/reconstruction.html"><code class="docutils literal notranslate"><span class="pre">reconstruction</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="../runscripts/runscripts.html"><code class="docutils literal notranslate"><span class="pre">runscripts</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="../validation/validation.html"><code class="docutils literal notranslate"><span class="pre">validation</span></code></a></li>
<li class="toctree-l2 current"><a class="reference internal" href="wholecell.html"><code class="docutils literal notranslate"><span class="pre">wholecell</span></code></a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="wholecell.io.html"><code class="docutils literal notranslate"><span class="pre">wholecell.io</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="wholecell.tests.html"><code class="docutils literal notranslate"><span class="pre">wholecell.tests</span></code></a></li>
<li class="toctree-l3 current"><a class="reference internal" href="wholecell.utils.html"><code class="docutils literal notranslate"><span class="pre">wholecell.utils</span></code></a><ul class="current">
<li class="toctree-l4"><a class="reference internal" href="wholecell.utils.build_ode.html"><code class="docutils literal notranslate"><span class="pre">wholecell.utils.build_ode</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="wholecell.utils.constants.html"><code class="docutils literal notranslate"><span class="pre">wholecell.utils.constants</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="wholecell.utils.data.html"><code class="docutils literal notranslate"><span class="pre">wholecell.utils.data</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="wholecell.utils.dependency_graph.html"><code class="docutils literal notranslate"><span class="pre">wholecell.utils.dependency_graph</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="wholecell.utils.fast_nonnegative_least_squares.html"><code class="docutils literal notranslate"><span class="pre">wholecell.utils.fast_nonnegative_least_squares</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="wholecell.utils.filepath.html"><code class="docutils literal notranslate"><span class="pre">wholecell.utils.filepath</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="wholecell.utils.fitting.html"><code class="docutils literal notranslate"><span class="pre">wholecell.utils.fitting</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="wholecell.utils.forkedPdb.html"><code class="docutils literal notranslate"><span class="pre">wholecell.utils.forkedPdb</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="wholecell.utils.make_media.html"><code class="docutils literal notranslate"><span class="pre">wholecell.utils.make_media</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="wholecell.utils.mc_complexation.html"><code class="docutils literal notranslate"><span class="pre">wholecell.utils.mc_complexation</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="wholecell.utils.memory_debug.html"><code class="docutils literal notranslate"><span class="pre">wholecell.utils.memory_debug</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="wholecell.utils.modular_fba.html"><code class="docutils literal notranslate"><span class="pre">wholecell.utils.modular_fba</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="wholecell.utils.parallelization.html"><code class="docutils literal notranslate"><span class="pre">wholecell.utils.parallelization</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="wholecell.utils.plotting_tools.html"><code class="docutils literal notranslate"><span class="pre">wholecell.utils.plotting_tools</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="wholecell.utils.polymerize.html"><code class="docutils literal notranslate"><span class="pre">wholecell.utils.polymerize</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="wholecell.utils.profiler.html"><code class="docutils literal notranslate"><span class="pre">wholecell.utils.profiler</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="wholecell.utils.protein_counts.html"><code class="docutils literal notranslate"><span class="pre">wholecell.utils.protein_counts</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="wholecell.utils.random.html"><code class="docutils literal notranslate"><span class="pre">wholecell.utils.random</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="wholecell.utils.sparkline.html"><code class="docutils literal notranslate"><span class="pre">wholecell.utils.sparkline</span></code></a></li>
<li class="toctree-l4 current"><a class="current reference internal" href="#"><code class="docutils literal notranslate"><span class="pre">wholecell.utils.spatial_tool</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="wholecell.utils.test_parallelization.html"><code class="docutils literal notranslate"><span class="pre">wholecell.utils.test_parallelization</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="wholecell.utils.toya.html"><code class="docutils literal notranslate"><span class="pre">wholecell.utils.toya</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="wholecell.utils.unit_struct_array.html"><code class="docutils literal notranslate"><span class="pre">wholecell.utils.unit_struct_array</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="wholecell.utils.units.html"><code class="docutils literal notranslate"><span class="pre">wholecell.utils.units</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="wholecell.utils.voronoi_plot_main.html"><code class="docutils literal notranslate"><span class="pre">wholecell.utils.voronoi_plot_main</span></code></a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">Vivarium E. coli</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../api_ref.html">API Reference</a></li>
          <li class="breadcrumb-item"><a href="wholecell.html"><code class="docutils literal notranslate"><span class="pre">wholecell</span></code></a></li>
          <li class="breadcrumb-item"><a href="wholecell.utils.html"><code class="docutils literal notranslate"><span class="pre">wholecell.utils</span></code></a></li>
      <li class="breadcrumb-item active"><code class="docutils literal notranslate"><span class="pre">wholecell.utils.spatial_tool</span></code></li>
      <li class="wy-breadcrumbs-aside">
            <a href="../../../_sources/reference/api/wholecell/wholecell.utils.spatial_tool.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="module-wholecell.utils.spatial_tool">
<span id="wholecell-utils-spatial-tool"></span><h1><code class="docutils literal notranslate"><span class="pre">wholecell.utils.spatial_tool</span></code><a class="headerlink" href="#module-wholecell.utils.spatial_tool" title="Link to this heading"></a></h1>
<p>Functions that may be useful for future investigation in spatial model.
Some of the functions contain default values specific to E coli. Please read
before you use. References for the numbers are included in each function.</p>
<dl class="py function">
<dt class="sig sig-object py" id="wholecell.utils.spatial_tool.compute_diffusion_constant_from_mw">
<span class="sig-prename descclassname"><span class="pre">wholecell.utils.spatial_tool.</span></span><span class="sig-name descname"><span class="pre">compute_diffusion_constant_from_mw</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mw</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mtype</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">loc</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">temp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parameters</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/wholecell/utils/spatial_tool.html#compute_diffusion_constant_from_mw"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#wholecell.utils.spatial_tool.compute_diffusion_constant_from_mw" title="Link to this definition"></a></dt>
<dd><p>Warning: The default values of the ‘parameters’ are E coli specific.</p>
<p>This function computes the hypothesized diffusion constant of
macromolecules within the nucleoid and the cytoplasm region.
In literature, there is no known differentiation between the diffusion
constant of a molecule in the nucleoid and in the cytoplasm up to the
best of our knowledge in 2019. However, there is a good reason why we
can assume that previously reported diffusion constant are in fact the
diffusion constant of a protein in the nucleoid region:
(1) The image traces of a protein within a bacteria usually cross the</p>
<blockquote>
<div><p>nucleoid regions.</p>
</div></blockquote>
<p>(2) The nucleoid region, compared to the cytoplasm, should be the main
limiting factor restricting the magnitude of diffusion constant.
(3) The same theory of diffusion constant has been implemented to
mammalian cells, and the term ‘rh’, the average hydrodynamic radius of
the biggest crowders, are different in mammalian cytoplasm, and it seems
to reflect the hydrodynamic radius of the actin filament (note: the
hydrodynamic radius of actin filament should be computed based on the
average length of actin fiber, and is not equal to the radius of the
actin filament itself.) (ref: Nano Lett. 2011, 11, 2157-2163).
As for E coli, the ‘rh’ term = 40nm, which may correspond to the 80nm
DNA fiber. On the other hand, for the diffusion constant of E coli in
the true cytoplasm, we will expect the value of ‘rh’ term to be
approximately 10 nm, which correspond to the radius of active ribosomes.</p>
<p>However, all the above statements are just hypothesis. If you want to
compute the diffusion constant of a macromolecule in the whole E coli
cell, you should set loc = ‘nucleoid’: the formula for this is obtained
from actual experimental data set. When you set loc = ‘cytoplasm’, the
entire results are merely hypothesis.</p>
<p>Ref: Kalwarczyk, T., Tabaka, M. &amp; Holyst, R.
Bioinformatics (2012). doi:10.1093/bioinformatics/bts537</p>
<p>D_0 = K_B*T/(6*pi*eta_0*rp)
ln(D_0/D_cyto) = ln(eta/eta_0) = (xi^2/Rh^2 + xi^2/rp^2)^(-a/2)
D_0 = the diffusion constant of a macromolecule in pure solvent
eta_0 = the viscosity of pure solvent, in this case, water
eta = the size-dependent viscosity experienced by the macromolecule.
xi = average distance between the surface of proteins
rh = average hydrodynamic radius of the biggest crowders
a = some constant of the order of 1
rp = hydrodynamic radius of probed molecule</p>
<p>In this formula, since we allow the changes in temperature, we also
consider the viscosity changes of water under different temperature:
Ref: Dortmund Data Bank
eta_0 = A*10^(B/(T-C))
A = 2.414*10^(-5) Pa*sec
B = 247.8 K
C = 140 K</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>mw</strong> – molecular weight(unit: Da) of the macromolecule</p></li>
<li><p><strong>mtype</strong> – There are 5 possible mtype options: protein, RNA, linear_DNA,</p></li>
<li><p><strong>circular_DNA</strong></p></li>
<li><p><strong>supercoiled_DNA.</strong> (<em>and</em>)</p></li>
<li><p><strong>loc</strong> – The location of the molecule: ‘nucleoid’ or ‘cytoplasm’.</p></li>
<li><p><strong>temp</strong> – The temperature of interest. unit: K.</p></li>
<li><p><strong>parameters</strong> – The 4 parameters required to compute the diffusion
constant: xi, a, rh_nuc, rh_cyto.
The default values are E coli specific.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>the diffusion constant of the macromolecule, units: um**2/sec</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dc</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="wholecell.utils.spatial_tool.compute_diffusion_constant_from_rp">
<span class="sig-prename descclassname"><span class="pre">wholecell.utils.spatial_tool.</span></span><span class="sig-name descname"><span class="pre">compute_diffusion_constant_from_rp</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rp</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">loc</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">temp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parameters</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/wholecell/utils/spatial_tool.html#compute_diffusion_constant_from_rp"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#wholecell.utils.spatial_tool.compute_diffusion_constant_from_rp" title="Link to this definition"></a></dt>
<dd><p>Warning: The default values of the ‘parameters’ are E coli specific.</p>
<p>This is the same function as ‘compute_diffusion_constant_from_mw’
except that it accepts the hydrodynamic radius of the macromolecules
as the input. All hypothesis and statements in
‘compute_diffusion_constant_from_mw’ apply here as well.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>rp</strong> – the hydrodynamic radius of the macromolecule, units: nm.</p></li>
<li><p><strong>loc</strong> – The location of the molecule: ‘nucleoid’ or ‘cytoplasm’.</p></li>
<li><p><strong>temp</strong> – The temperature of interest. unit: K.</p></li>
<li><p><strong>parameters</strong> – The 4 parameters required to compute the diffusion
constant: xi, a, rh_nuc, rh_cyto.
The default values are E coli specific.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>the diffusion constant of the macromolecule, units: um**2/sec</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dc</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="wholecell.utils.spatial_tool.compute_free_volume_ratio">
<span class="sig-prename descclassname"><span class="pre">wholecell.utils.spatial_tool.</span></span><span class="sig-name descname"><span class="pre">compute_free_volume_ratio</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">q</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">eta</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/wholecell/utils/spatial_tool.html#compute_free_volume_ratio"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#wholecell.utils.spatial_tool.compute_free_volume_ratio" title="Link to this definition"></a></dt>
<dd><p>This is a function that computes the free volume ratio of existing
macromolecules with respect to a new molecule based on the scaled
particle theory. This is not E coli specific. It is important to note
that the results may not be accurate for large q or eta &gt; 0.50.</p>
<p>References:
Malloggi, F. Soft Matter at Aqueous Interfaces.
Lecture Notes in Physics (2016). doi:10.1007/978-3-319-24502-7.
Chapter 3-4-2.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>q</strong> – the size ratio between the new particle and existing particles.
For example, if the radius of the new particle = delta, and the
radius of the existing particle = R, then q = delta/R.</p></li>
<li><p><strong>eta</strong> – the compaction ratio, or the volume occupancy of the existing
particle. eta = n*v_particle/v_total</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>the free volume ratio of the space = v_free/v_total.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>alpha</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="wholecell.utils.spatial_tool.compute_hydrodynamic_radius">
<span class="sig-prename descclassname"><span class="pre">wholecell.utils.spatial_tool.</span></span><span class="sig-name descname"><span class="pre">compute_hydrodynamic_radius</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mw</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mtype</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/wholecell/utils/spatial_tool.html#compute_hydrodynamic_radius"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#wholecell.utils.spatial_tool.compute_hydrodynamic_radius" title="Link to this definition"></a></dt>
<dd><p>This function compute the hydrodynamic diameter of a macromolecules from
its molecular weight. It is important to note that the hydrodynamic
diameter is mainly used for computation of diffusion constant, and can
be different from the observed diameter under microscopes or the radius
of gyration, especially for loose polymers such as RNAs. This function
is not E coli specific.</p>
<p>References: Bioinformatics (2012). doi:10.1093/bioinformatics/bts537</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>mw</strong> – molecular weight of the macromolecules, units: Daltons.</p></li>
<li><p><strong>mtype</strong> – There are 5 possible mtype options: protein, RNA, linear_DNA,
circular_DNA, and supercoiled_DNA.</p></li>
</ul>
</dd>
</dl>
<dl class="simple">
<dt>Returns: the hydrodynamic radius (in unit of nm) of the macromolecules</dt><dd><p>using the following formula
- rp = 0.0515*MW^(0.392) nm (Hong &amp; Lei 2008) (protein)
- rp = 0.0566*MW^(0.38) nm (Werner 2011) (RNA)
- rp = 0.024*MW^(0.57) nm (Robertson et al 2006) (linear DNA)
- rp = 0.0125*MW^(0.59) nm (Robertson et al 2006) (circular DNA)
- rp = 0.0145*MW^(0.57) nm (Robertson et al 2006) (supercoiled DNA)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="wholecell.utils.spatial_tool.compute_n_blob">
<span class="sig-prename descclassname"><span class="pre">wholecell.utils.spatial_tool.</span></span><span class="sig-name descname"><span class="pre">compute_n_blob</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">choice_model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">choice_unit</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bp_dna</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/wholecell/utils/spatial_tool.html#compute_n_blob"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#wholecell.utils.spatial_tool.compute_n_blob" title="Link to this definition"></a></dt>
<dd><p>This function computes the number of DNA blobs within a bacteria. This
functions contain multiple assumptions that demand careful examination.</p>
<p>Previous research (Ref: Skoko D, Wong B, Johnson R, Marko J (2004)
Biochemistry 43: 13867-13874.) directly observed the structure of
chromosomal DNA of E coli under AFM. They observed that the DNA are
composed of 40nm &amp; 80nm fibers. The 80nm fibers are the main type within
the cells, but it is hypothesized to be folded from 2 threads of 40nm
fibers. Therefore, we hypothesized that chromosomal DNA are composed of
DNA blobs with diameter of 40nm.</p>
<p>For the model choice, there are 2 possible options. In the Flory chain
option, the DNA within a blob is assumed to be self-avoidant. In the
ideal chain option, the DNA within a blob is assumed to be in a melted
state.</p>
<p>For the unit choice, there are 3 possible options. You can choose single
base pair of DNA as the smallest unit in the formation of DNA blob.
In conventional physics, 2*persistence length is regarded as the
standard choice of the smallest unit in a polymer. It is important to
note that, the persistence length of DNA can be drastically different
under in vitro &amp; in vivo conditions. The persistence length of DNA in
vitro is ~ 50nm, while the persistence length of DNA in vivo can be
down to 20nm. This is because the binding of DNA binding proteins &amp;
DNA supercoiling can decrease the energy of DNA and makes it softer.
Another possible choice of the smallest unit of DNA is 10bp, which
corresponds to the average binding distance between HU protein, a
structural DNA binding protein.</p>
<p>Choosing different model setting and different smallest unit of DNA can
result in drastically different results.</p>
<p>The default value of genomic size is set to be E coli specific.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>choice_model</strong> – the physical model of DNA within a blob. The 2
possible options are ‘ideal’ &amp; ‘flory’.</p></li>
<li><p><strong>choice_unit</strong> – the choice of the smallest unit within a blob. The 3
possible options are ‘single’, ‘lp’, or an integer number
indicating the number of base pairs within a single unit.</p></li>
<li><p><strong>bp_dna</strong> – the number of base pairs in the chromosome. The default
value is set to be the chromosome size of E coli.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>the number of DNA blobs in a bacteria.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>n_blob</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="wholecell.utils.spatial_tool.compute_nucleoid_size">
<span class="sig-prename descclassname"><span class="pre">wholecell.utils.spatial_tool.</span></span><span class="sig-name descname"><span class="pre">compute_nucleoid_size</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">l_cell</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">d_cell</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">length_scaling_parameters</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nucleoid_area_ratio</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/wholecell/utils/spatial_tool.html#compute_nucleoid_size"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#wholecell.utils.spatial_tool.compute_nucleoid_size" title="Link to this definition"></a></dt>
<dd><p>Warning 1: This function contains default values that are specific to
E coli grew on M9 media under aerobic condition only. The shape and size
of nucleoid of a bacteria can be very different across species.
Warning 2: This function is not suitable to compute the nucleoid size of
E coli when its shape turn filamentaous. This is because the scaling
formula of the length of the nucleoid is obtained from a dnaC mutant
E coli strain. According to our reference, when the cells are allowed
to replicate their DNA normally, a constant N/C area ratio is maintained
even for filamentous variants (treated with cephalexin). However, if the
DNA is not allowed to replicate, the N/C area ratio will decrease as the
cell elongate. It is therefore recommended to carefully examine the
condition when the length of the cells grow beyond 3 um.
Warning 3: the default values of length_scaling_parameters and
nucleoid_area_ratio are set to be E coli specific, grew on M9 media
under aerobic condition. The N/C area ratio for E coli grew on LB under
aerobic condition is close to 0.4. For E coli grew under anaerobic
condition it is close to 0.5.
Warning 4: It is also important to note that a single cell can contain
more than 1 nucleoid, and this formula may not be suitable for these
cases.</p>
<p>Reference on nucleoid length:
Wu, F. et al. Curr. Biol. (2019). doi:10.1016/j.cub.2019.05.015
Reference on nucleoid/cytoplasm area ratio:
Gray, W. T. et al. Cell (2019). doi:10.1016/j.cell.2019.05.017</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>l_cell</strong> – the length of the cell, in units of um</p></li>
<li><p><strong>d_cell</strong> – the width of the cell, in units of um</p></li>
<li><p><strong>length_scaling_parameters</strong> – the parameters used for the scaling
formula of the length of the nucleoid with respect to the length
of the whole cell.</p></li>
<li><p><strong>nucleoid_area_ratio</strong> – the nucleoid/cytoplasm area ratio measured
under microscope.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>the length of the nucleoid
d_nuc: the diameter of the nucleoid</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>l_nuc</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="wholecell.utils.spatial_tool.compute_rg_rna">
<span class="sig-prename descclassname"><span class="pre">wholecell.utils.spatial_tool.</span></span><span class="sig-name descname"><span class="pre">compute_rg_rna</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n_nt</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/wholecell/utils/spatial_tool.html#compute_rg_rna"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#wholecell.utils.spatial_tool.compute_rg_rna" title="Link to this definition"></a></dt>
<dd><p>This function computes the radius of gyration (in units of nm) of an RNA
with length n_nt(in units of nt). This function is not E coli specific.
It is important to note that radius of gyration can be very different
from the hydrodynamic radius or the radius you can observe under
microscopes, especially for RNAs.</p>
<dl class="simple">
<dt>References: Nucleic Acids Research, 2017, vol 45, 5: 2919-2934.</dt><dd><p>doi: 10.1093/nar/gkx023</p>
</dd>
</dl>
<p>Formula: Rg = a*N**nu. a = 0.366 +/- 0.146 nm, nu = 0.50 +/- 0.05</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>n_nt</strong> – the number of nucleotide in an RNA.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The radius of gyration of an RNA in the unit of nm.</p>
</dd>
</dl>
</dd></dl>

</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="wholecell.utils.sparkline.html" class="btn btn-neutral float-left" title="wholecell.utils.sparkline" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="wholecell.utils.test_parallelization.html" class="btn btn-neutral float-right" title="wholecell.utils.test_parallelization" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2021-2025, The Vivarium E. coli Authors.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>